{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "78b0d1fd",
   "metadata": {},
   "source": [
    "### Moved a lot to snakemake for speed, memory and readabilitiy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d165efb4",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "vcf = \"/data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/variants.all.vcf\"\n",
    "bam_fs = \"/data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc\"\n",
    "outdir = \"/data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap\"\n",
    "genome_fa = \"/data/Mito_Trace/data/processed/genomes/mtMasked/GRCh38_MT_blacklist_A2_2020/fasta/genome.fa\"\n",
    "scripts_dir = \"/data/Mito_Trace/src/mtpreproc/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9da443f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fig_utils from mplh\n"
     ]
    }
   ],
   "source": [
    "from glob import glob\n",
    "from os.path import basename, join, exists\n",
    "import os\n",
    "from src.utils.data_io import read_csv_multichar\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "47e47540",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_outdir = join(outdir, \"regions\")\n",
    "if not exists(regions_outdir):\n",
    "    os.mkdir(regions_outdir)\n",
    "    \n",
    "# sc_regions_outdir = join(regions_outdir, \"sc\")\n",
    "# if not exists(sc_regions_outdir):\n",
    "#     os.mkdir(sc_regions_outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3076ac30",
   "metadata": {},
   "outputs": [],
   "source": [
    "b = \"/data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8f8492a0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#for b in glob(bam_fs+\"/*.bam\"):\n",
    "#    cmd = f\"bedtools intersect -wa -a {b} -b {vcf}  > {outdir}/{basename(b)}\"\n",
    "#    print(cmd)\n",
    "#    !{cmd}\n",
    "#    cmd = f\"samtools index {outdir}/{basename(b)}\"\n",
    "#    print(cmd)\n",
    "#    !{cmd}\n",
    "if not exists(join(outdir,basename(b)+\".bai\")):\n",
    "    cmd = f\"samtools index {outdir}/{basename(b)}\"\n",
    "    print(cmd)\n",
    "    !{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6377ab31",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python /data/Mito_Trace/src/mtpreproc//split_by_CB.py /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/CHIP_b1_Input.bam /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/CHIP_b1_Input_pileups\n",
      "/data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/CHIP_b1_Input.bam /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/CHIP_b1_Input.CB.bam\n",
      "samtools sort -t CB /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/CHIP_b1_Input.bam > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/CHIP_b1_Input.CB.bam\n",
      "[bam_sort_core] merging from 80 files and 1 in-memory blocks...\n",
      "150133000it [44:57, 55655.99it/s] \n",
      "no CBs:  8487066\n",
      "no CBs permission 0\n"
     ]
    }
   ],
   "source": [
    "curr_b = f\"{outdir}/{basename(b)}\"\n",
    "curr_b_outdir = f'{curr_b.replace(\".bam\", \"\")}_pileups'\n",
    "if not exists(curr_b_outdir):\n",
    "    os.mkdir(curr_b_outdir)\n",
    "cmd = f\"python {scripts_dir}/split_by_CB.py {curr_b} {curr_b_outdir}\"\n",
    "print(cmd)\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5a3a921",
   "metadata": {},
   "source": [
    "## Bedtools intersect for each specific region\n",
    "### Then add each variant for counts and variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ad611bbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_f = vcf+\".region.txt\"\n",
    "vcf_df = read_csv_multichar(vcf, sep=\"\\t\", multicomment=\"##\")\n",
    "vcf_df[[\"#CHROM\", \"POS\"]].to_csv(region_f, index=None, sep=\"\\t\", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "09e58375",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>CHIP_b1_Control</th>\n",
       "      <th>...</th>\n",
       "      <th>CHIP_b1_Input</th>\n",
       "      <th>CHIP_b2_Control</th>\n",
       "      <th>CHIP_b2_Flt3l</th>\n",
       "      <th>CHIP_b2_Input</th>\n",
       "      <th>CHIP_b3_Control</th>\n",
       "      <th>CHIP_b3_Flt3l</th>\n",
       "      <th>CHIP_b3_Input</th>\n",
       "      <th>CHIP_b4_Control</th>\n",
       "      <th>CHIP_b4_Flt3l</th>\n",
       "      <th>CHIP_b4_Input</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1000</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10018</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>97.47380</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=snv;NS=12;AF=0.08333333;DP=4744;RO=4722;A...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/0:0.2387064:681:678:1:0.001468429:402,276,1,...</td>\n",
       "      <td>...</td>\n",
       "      <td>0/1:97.4738:1249:1237:12:0.009607686:740,497,7...</td>\n",
       "      <td>0/0:0:234:233:0:0:113,120,0,0:-1:-1:0:944.2852:.</td>\n",
       "      <td>0/0:0:225:225:0:0:113,112,0,0:-1:-1:0:926.5604:.</td>\n",
       "      <td>0/0:0:343:343:0:0:176,167,0,0:-1:-1:0:1432.95:.</td>\n",
       "      <td>0/0:16.98098:186:184:2:0.01075269:100,84,0,2:9...</td>\n",
       "      <td>0/0:0:269:269:0:0:130,139,0,0:-1:-1:0:1116.99:.</td>\n",
       "      <td>0/0:0:108:108:0:0:53,55,0,0:-1:-1:0:416.3386:.</td>\n",
       "      <td>0/0:21.71786:356:353:3:0.008426966:162,191,2,1...</td>\n",
       "      <td>0/0:0:315:314:0:0:160,154,0,0:-1:-1:0:1305.324:.</td>\n",
       "      <td>0/0:0:118:118:0:0:67,51,0,0:-1:-1:0:458.7915:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10031</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>65.50475</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=ins;NS=12;AF=0.08333333;DP=6095;RO=6077;A...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/1:65.50475:856:853:5:0.005841121:532,321,3,2...</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0:1640:1639:0:0:1029,610,0,0:-1:-1:0:Inf:.</td>\n",
       "      <td>0/0:0:315:315:0:0:163,152,0,0:-1:-1:0:1790.363:.</td>\n",
       "      <td>0/0:0:286:286:0:0:146,140,0,0:-1:-1:0:1616.495:.</td>\n",
       "      <td>0/0:0:432:431:0:0:237,194,0,0:-1:-1:0:2483.383:.</td>\n",
       "      <td>0/0:0:245:245:0:0:141,104,0,0:-1:-1:0:1386.389:.</td>\n",
       "      <td>0/0:0:357:357:0:0:189,168,0,0:-1:-1:0:2052.156:.</td>\n",
       "      <td>0/0:0:141:141:0:0:70,71,0,0:-1:-1:0:744.5862:.</td>\n",
       "      <td>0/0:0:446:444:0:0:230,214,0,0:-1:-1:0:2570.377:.</td>\n",
       "      <td>0/0:0:381:379:0:0:208,171,0,0:-1:-1:0:2196.873:.</td>\n",
       "      <td>0/0:0:154:153:0:0:94,59,0,0:-1:-1:0:833.2017:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1002</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10039</td>\n",
       "      <td>.</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>87.23672</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=del;NS=12;AF=0.08333333;DP=6678;RO=6620;A...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/0:0:946:939:0:0:633,306,0,0:-1:-1:0:Inf:.</td>\n",
       "      <td>...</td>\n",
       "      <td>0/1:87.23672:1822:1808:8:0.004390779:1257,551,...</td>\n",
       "      <td>0/0:0:341:340:0:0:195,145,0,0:-1:-1:0:1873.462:.</td>\n",
       "      <td>0/0:0:308:306:0:0:178,128,0,0:-1:-1:0:1679.652:.</td>\n",
       "      <td>0/0:0:472:470:0:0:277,193,0,0:-1:-1:0:2615.803:.</td>\n",
       "      <td>0/0:0:259:259:0:0:168,91,0,0:-1:-1:0:1403.992:.</td>\n",
       "      <td>0/0:22.05355:368:361:2:0.005434783:218,143,0,2...</td>\n",
       "      <td>0/0:0:161:160:0:0:94,66,0,0:-1:-1:0:819.4477:.</td>\n",
       "      <td>0/0:21.33713:492:489:2:0.004065041:288,201,2,0...</td>\n",
       "      <td>0/0:0:422:422:0:0:251,171,0,0:-1:-1:0:2310.822:.</td>\n",
       "      <td>0/0:0:176:174:0:0:124,50,0,0:-1:-1:0:933.9806:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1003</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10051</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>ACCG</td>\n",
       "      <td>67.72425</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=ins;NS=12;AF=0.08333333;DP=7730;RO=7680;A...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/0:0:1123:1117:0:0:754,363,0,0:-1:-1:0:Inf:.</td>\n",
       "      <td>...</td>\n",
       "      <td>0/1:67.72425:2162:2147:6:0.002775208:1497,650,...</td>\n",
       "      <td>0/0:0:384:382:0:0:230,152,0,0:-1:-1:0:2286.389:.</td>\n",
       "      <td>0/0:0:348:348:0:0:212,136,0,0:-1:-1:0:2077.343:.</td>\n",
       "      <td>0/0:0:535:532:0:0:317,215,0,0:-1:-1:0:Inf:.</td>\n",
       "      <td>0/0:0:293:291:0:0:191,100,0,0:-1:-1:0:1747.041:.</td>\n",
       "      <td>0/0:0:419:416:0:0:272,144,0,0:-1:-1:0:2530.343:.</td>\n",
       "      <td>0/0:0:190:190:0:0:113,77,0,0:-1:-1:0:1078.799:.</td>\n",
       "      <td>0/0:0:562:561:0:0:346,215,0,0:-1:-1:0:Inf:.</td>\n",
       "      <td>0/0:0:456:456:0:0:289,167,0,0:-1:-1:0:2738.056:.</td>\n",
       "      <td>0/0:0:208:206:0:0:137,69,0,0:-1:-1:0:1201.061:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1004</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10055</td>\n",
       "      <td>.</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>60.01402</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=del;NS=12;AF=0.1666667;DP=7734;RO=7677;AO...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/1:60.01402:1122:1113:9:0.00802139:707,406,4,...</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0:2135:2124:1:0.0004683841:1442,682,1,0:99...</td>\n",
       "      <td>0/0:0:389:388:0:0:236,152,0,0:-1:-1:0:1570.522:.</td>\n",
       "      <td>0/0:0:352:352:0:0:216,136,0,0:-1:-1:0:1430.16:.</td>\n",
       "      <td>0/0:0.5660855:535:535:1:0.001869159:314,221,0,...</td>\n",
       "      <td>0/0:0:295:294:0:0:196,98,0,0:-1:-1:0:1183.517:.</td>\n",
       "      <td>0/0:0.5660855:422:421:1:0.002369668:270,151,0,...</td>\n",
       "      <td>0/0:0:189:186:0:0:107,79,0,0:-1:-1:0:730.9136:.</td>\n",
       "      <td>0/0:0:557:555:0:0:335,220,0,0:-1:-1:0:2283.826:.</td>\n",
       "      <td>0/0:0:451:449:0:0:288,161,0,0:-1:-1:0:1836.398:.</td>\n",
       "      <td>0/0:0:198:197:0:0:128,69,0,0:-1:-1:0:772.6831:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6200</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56767254</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>56.78491</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=snv;NS=12;AF=0.08333333;DP=1759;RO=1744;A...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/0:20.33698:408:405:3:0.007352941:277,128,0,3...</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0:380:380:0:0:252,128,0,0:-1:-1:0:1563.988:.</td>\n",
       "      <td>0/0:0:67:66:0:0:36,30,0,0:-1:-1:0:216.2226:.</td>\n",
       "      <td>0/0:0:24:24:0:0:15,9,0,0:-1:-1:0:51.69794:.</td>\n",
       "      <td>0/0:20.33698:376:372:3:0.007978723:213,159,3,0...</td>\n",
       "      <td>./.:0:1:1:0:0:1,0,0,0:-1:-1:0:0:.</td>\n",
       "      <td>./.:0:6:6:0:0:6,0,0,0:-1:-1:0:0:.</td>\n",
       "      <td>0/0:0:26:26:0:0:18,8,0,0:-1:-1:0:50.65918:.</td>\n",
       "      <td>0/0:0:53:53:0:0:24,29,0,0:-1:-1:0:153.1909:.</td>\n",
       "      <td>0/0:0:38:38:0:0:22,16,0,0:-1:-1:0:90.83723:.</td>\n",
       "      <td>./.:0:12:12:0:0:6,6,0,0:-1:-1:0:11.25068:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6201</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56768565</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>51.13507</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=snv;NS=12;AF=0.08333333;DP=3111;RO=3102;A...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/0:0:829:828:0:0:551,277,0,0:-1:-1:0:Inf:.</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0:594:593:0:0:389,204,0,0:-1:-1:0:Inf:.</td>\n",
       "      <td>0/0:0:110:110:0:0:61,49,0,0:-1:-1:0:566.8315:.</td>\n",
       "      <td>0/0:0:31:31:0:0:24,7,0,0:-1:-1:0:103.6253:.</td>\n",
       "      <td>0/1:51.13507:657:652:4:0.00608828:436,216,3,1:...</td>\n",
       "      <td>./.:0:9:9:0:0:7,2,0,0:-1:-1:0:0:.</td>\n",
       "      <td>./.:0:12:12:0:0:8,4,0,0:-1:-1:0:18.57706:.</td>\n",
       "      <td>0/0:0:46:46:0:0:25,21,0,0:-1:-1:0:190.525:.</td>\n",
       "      <td>0/0:12.14579:106:105:1:0.009433962:51,54,1,0:9...</td>\n",
       "      <td>0/0:0:55:55:0:0:28,27,0,0:-1:-1:0:249.1141:.</td>\n",
       "      <td>./.:0:20:20:0:0:13,7,0,0:-1:-1:0:46.51547:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6202</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56768763</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>67.89379</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=snv;NS=12;AF=0.08333333;DP=1164;RO=1144;A...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/0:0:265:264:0:0:80,184,0,0:-1:-1:0:1241.999:.</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0:220:215:0:0:58,157,0,0:-1:-1:0:1026.952:.</td>\n",
       "      <td>0/0:10.07019:57:55:1:0.01754386:22,33,1,0:99:1...</td>\n",
       "      <td>./.:0:14:14:0:0:4,10,0,0:-1:-1:0:13.13029:.</td>\n",
       "      <td>0/1:67.89379:242:231:5:0.02066116:77,154,2,3:0...</td>\n",
       "      <td>./.:0:5:5:0:0:1,4,0,0:-1:-1:0:0:.</td>\n",
       "      <td>./.:0:7:7:0:0:2,5,0,0:-1:-1:0:0:.</td>\n",
       "      <td>0/0:0:30:30:0:0:8,22,0,0:-1:-1:0:85.0804:.</td>\n",
       "      <td>0/0:0:58:58:0:0:23,35,0,0:-1:-1:0:204.0312:.</td>\n",
       "      <td>0/0:0:29:28:0:0:8,20,0,0:-1:-1:0:85.66831:.</td>\n",
       "      <td>./.:0:8:8:0:0:2,6,0,0:-1:-1:0:0:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6203</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56768805</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>77.00362</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=snv;NS=12;AF=0.08333333;DP=1131;RO=1125;A...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/0:0:269:269:0:0:67,202,0,0:-1:-1:0:1316.794:.</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0:209:209:0:0:41,168,0,0:-1:-1:0:989.6153:.</td>\n",
       "      <td>0/0:0:80:80:0:0:30,50,0,0:-1:-1:0:340.0043:.</td>\n",
       "      <td>./.:0:11:11:0:0:2,9,0,0:-1:-1:0:0:.</td>\n",
       "      <td>0/0:0:227:226:0:0:56,170,0,0:-1:-1:0:1090.748:.</td>\n",
       "      <td>./.:0:2:2:0:0:0,2,0,0:-1:-1:0:0:.</td>\n",
       "      <td>./.:0:8:8:0:0:2,6,0,0:-1:-1:0:0:.</td>\n",
       "      <td>0/0:0:24:24:0:0:5,19,0,0:-1:-1:0:64.55056:.</td>\n",
       "      <td>0/0:0:60:60:0:0:21,39,0,0:-1:-1:0:239.7105:.</td>\n",
       "      <td>0/0:0:32:32:0:0:11,21,0,0:-1:-1:0:87.73926:.</td>\n",
       "      <td>./.:0:12:12:0:0:3,9,0,0:-1:-1:0:14.7431:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6204</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56770041</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>51.62626</td>\n",
       "      <td>.</td>\n",
       "      <td>TYPE=snv;NS=12;AF=0.08333333;DP=763;RO=750;AO=...</td>\n",
       "      <td>GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...</td>\n",
       "      <td>0/0:0:187:186:0:0:82,104,0,0:-1:-1:0:795.0247:.</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:0:137:137:0:0:70,67,0,0:-1:-1:0:546.2925:.</td>\n",
       "      <td>0/0:0:34:33:0:0:16,17,0,0:-1:-1:0:101.2374:.</td>\n",
       "      <td>./.:0:7:7:0:0:3,4,0,0:-1:-1:0:0:.</td>\n",
       "      <td>0/0:0:178:177:0:0:97,80,0,0:-1:-1:0:749.758:.</td>\n",
       "      <td>./.:0:1:1:0:0:0,1,0,0:-1:-1:0:0:.</td>\n",
       "      <td>./.:0:1:1:0:0:0,1,0,0:-1:-1:0:0:.</td>\n",
       "      <td>./.:0:12:12:0:0:5,7,0,0:-1:-1:0:12.84924:.</td>\n",
       "      <td>0/0:33.65802:34:32:2:0.05882353:15,17,2,0:99:1...</td>\n",
       "      <td>./.:36.06597:21:19:2:0.0952381:11,8,1,1:0.4245...</td>\n",
       "      <td>./.:0:9:9:0:0:6,3,0,0:-1:-1:0:0:.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5205 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #CHROM       POS ID REF   ALT      QUAL FILTER  \\\n",
       "1000   chr1     10018  .   C     G  97.47380      .   \n",
       "1001   chr1     10031  .   T    TA  65.50475      .   \n",
       "1002   chr1     10039  .  AC     A  87.23672      .   \n",
       "1003   chr1     10051  .   A  ACCG  67.72425      .   \n",
       "1004   chr1     10055  .  TA     T  60.01402      .   \n",
       "...     ...       ... ..  ..   ...       ...    ...   \n",
       "6200   chrY  56767254  .   T     A  56.78491      .   \n",
       "6201   chrY  56768565  .   G     A  51.13507      .   \n",
       "6202   chrY  56768763  .   T     A  67.89379      .   \n",
       "6203   chrY  56768805  .   C     T  77.00362      .   \n",
       "6204   chrY  56770041  .   G     A  51.62626      .   \n",
       "\n",
       "                                                   INFO  \\\n",
       "1000  TYPE=snv;NS=12;AF=0.08333333;DP=4744;RO=4722;A...   \n",
       "1001  TYPE=ins;NS=12;AF=0.08333333;DP=6095;RO=6077;A...   \n",
       "1002  TYPE=del;NS=12;AF=0.08333333;DP=6678;RO=6620;A...   \n",
       "1003  TYPE=ins;NS=12;AF=0.08333333;DP=7730;RO=7680;A...   \n",
       "1004  TYPE=del;NS=12;AF=0.1666667;DP=7734;RO=7677;AO...   \n",
       "...                                                 ...   \n",
       "6200  TYPE=snv;NS=12;AF=0.08333333;DP=1759;RO=1744;A...   \n",
       "6201  TYPE=snv;NS=12;AF=0.08333333;DP=3111;RO=3102;A...   \n",
       "6202  TYPE=snv;NS=12;AF=0.08333333;DP=1164;RO=1144;A...   \n",
       "6203  TYPE=snv;NS=12;AF=0.08333333;DP=1131;RO=1125;A...   \n",
       "6204  TYPE=snv;NS=12;AF=0.08333333;DP=763;RO=750;AO=...   \n",
       "\n",
       "                                                 FORMAT  \\\n",
       "1000  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "1001  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "1002  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "1003  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "1004  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "...                                                 ...   \n",
       "6200  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "6201  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "6202  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "6203  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "6204  GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:...   \n",
       "\n",
       "                                        CHIP_b1_Control  ...  \\\n",
       "1000  0/0:0.2387064:681:678:1:0.001468429:402,276,1,...  ...   \n",
       "1001  0/1:65.50475:856:853:5:0.005841121:532,321,3,2...  ...   \n",
       "1002        0/0:0:946:939:0:0:633,306,0,0:-1:-1:0:Inf:.  ...   \n",
       "1003      0/0:0:1123:1117:0:0:754,363,0,0:-1:-1:0:Inf:.  ...   \n",
       "1004  0/1:60.01402:1122:1113:9:0.00802139:707,406,4,...  ...   \n",
       "...                                                 ...  ...   \n",
       "6200  0/0:20.33698:408:405:3:0.007352941:277,128,0,3...  ...   \n",
       "6201        0/0:0:829:828:0:0:551,277,0,0:-1:-1:0:Inf:.  ...   \n",
       "6202    0/0:0:265:264:0:0:80,184,0,0:-1:-1:0:1241.999:.  ...   \n",
       "6203    0/0:0:269:269:0:0:67,202,0,0:-1:-1:0:1316.794:.  ...   \n",
       "6204    0/0:0:187:186:0:0:82,104,0,0:-1:-1:0:795.0247:.  ...   \n",
       "\n",
       "                                          CHIP_b1_Input  \\\n",
       "1000  0/1:97.4738:1249:1237:12:0.009607686:740,497,7...   \n",
       "1001     0/0:0:1640:1639:0:0:1029,610,0,0:-1:-1:0:Inf:.   \n",
       "1002  0/1:87.23672:1822:1808:8:0.004390779:1257,551,...   \n",
       "1003  0/1:67.72425:2162:2147:6:0.002775208:1497,650,...   \n",
       "1004  0/0:0:2135:2124:1:0.0004683841:1442,682,1,0:99...   \n",
       "...                                                 ...   \n",
       "6200   0/0:0:380:380:0:0:252,128,0,0:-1:-1:0:1563.988:.   \n",
       "6201        0/0:0:594:593:0:0:389,204,0,0:-1:-1:0:Inf:.   \n",
       "6202    0/0:0:220:215:0:0:58,157,0,0:-1:-1:0:1026.952:.   \n",
       "6203    0/0:0:209:209:0:0:41,168,0,0:-1:-1:0:989.6153:.   \n",
       "6204     0/0:0:137:137:0:0:70,67,0,0:-1:-1:0:546.2925:.   \n",
       "\n",
       "                                        CHIP_b2_Control  \\\n",
       "1000   0/0:0:234:233:0:0:113,120,0,0:-1:-1:0:944.2852:.   \n",
       "1001   0/0:0:315:315:0:0:163,152,0,0:-1:-1:0:1790.363:.   \n",
       "1002   0/0:0:341:340:0:0:195,145,0,0:-1:-1:0:1873.462:.   \n",
       "1003   0/0:0:384:382:0:0:230,152,0,0:-1:-1:0:2286.389:.   \n",
       "1004   0/0:0:389:388:0:0:236,152,0,0:-1:-1:0:1570.522:.   \n",
       "...                                                 ...   \n",
       "6200       0/0:0:67:66:0:0:36,30,0,0:-1:-1:0:216.2226:.   \n",
       "6201     0/0:0:110:110:0:0:61,49,0,0:-1:-1:0:566.8315:.   \n",
       "6202  0/0:10.07019:57:55:1:0.01754386:22,33,1,0:99:1...   \n",
       "6203       0/0:0:80:80:0:0:30,50,0,0:-1:-1:0:340.0043:.   \n",
       "6204       0/0:0:34:33:0:0:16,17,0,0:-1:-1:0:101.2374:.   \n",
       "\n",
       "                                         CHIP_b2_Flt3l  \\\n",
       "1000  0/0:0:225:225:0:0:113,112,0,0:-1:-1:0:926.5604:.   \n",
       "1001  0/0:0:286:286:0:0:146,140,0,0:-1:-1:0:1616.495:.   \n",
       "1002  0/0:0:308:306:0:0:178,128,0,0:-1:-1:0:1679.652:.   \n",
       "1003  0/0:0:348:348:0:0:212,136,0,0:-1:-1:0:2077.343:.   \n",
       "1004   0/0:0:352:352:0:0:216,136,0,0:-1:-1:0:1430.16:.   \n",
       "...                                                ...   \n",
       "6200       0/0:0:24:24:0:0:15,9,0,0:-1:-1:0:51.69794:.   \n",
       "6201       0/0:0:31:31:0:0:24,7,0,0:-1:-1:0:103.6253:.   \n",
       "6202       ./.:0:14:14:0:0:4,10,0,0:-1:-1:0:13.13029:.   \n",
       "6203               ./.:0:11:11:0:0:2,9,0,0:-1:-1:0:0:.   \n",
       "6204                 ./.:0:7:7:0:0:3,4,0,0:-1:-1:0:0:.   \n",
       "\n",
       "                                          CHIP_b2_Input  \\\n",
       "1000    0/0:0:343:343:0:0:176,167,0,0:-1:-1:0:1432.95:.   \n",
       "1001   0/0:0:432:431:0:0:237,194,0,0:-1:-1:0:2483.383:.   \n",
       "1002   0/0:0:472:470:0:0:277,193,0,0:-1:-1:0:2615.803:.   \n",
       "1003        0/0:0:535:532:0:0:317,215,0,0:-1:-1:0:Inf:.   \n",
       "1004  0/0:0.5660855:535:535:1:0.001869159:314,221,0,...   \n",
       "...                                                 ...   \n",
       "6200  0/0:20.33698:376:372:3:0.007978723:213,159,3,0...   \n",
       "6201  0/1:51.13507:657:652:4:0.00608828:436,216,3,1:...   \n",
       "6202  0/1:67.89379:242:231:5:0.02066116:77,154,2,3:0...   \n",
       "6203    0/0:0:227:226:0:0:56,170,0,0:-1:-1:0:1090.748:.   \n",
       "6204      0/0:0:178:177:0:0:97,80,0,0:-1:-1:0:749.758:.   \n",
       "\n",
       "                                        CHIP_b3_Control  \\\n",
       "1000  0/0:16.98098:186:184:2:0.01075269:100,84,0,2:9...   \n",
       "1001   0/0:0:245:245:0:0:141,104,0,0:-1:-1:0:1386.389:.   \n",
       "1002    0/0:0:259:259:0:0:168,91,0,0:-1:-1:0:1403.992:.   \n",
       "1003   0/0:0:293:291:0:0:191,100,0,0:-1:-1:0:1747.041:.   \n",
       "1004    0/0:0:295:294:0:0:196,98,0,0:-1:-1:0:1183.517:.   \n",
       "...                                                 ...   \n",
       "6200                  ./.:0:1:1:0:0:1,0,0,0:-1:-1:0:0:.   \n",
       "6201                  ./.:0:9:9:0:0:7,2,0,0:-1:-1:0:0:.   \n",
       "6202                  ./.:0:5:5:0:0:1,4,0,0:-1:-1:0:0:.   \n",
       "6203                  ./.:0:2:2:0:0:0,2,0,0:-1:-1:0:0:.   \n",
       "6204                  ./.:0:1:1:0:0:0,1,0,0:-1:-1:0:0:.   \n",
       "\n",
       "                                          CHIP_b3_Flt3l  \\\n",
       "1000    0/0:0:269:269:0:0:130,139,0,0:-1:-1:0:1116.99:.   \n",
       "1001   0/0:0:357:357:0:0:189,168,0,0:-1:-1:0:2052.156:.   \n",
       "1002  0/0:22.05355:368:361:2:0.005434783:218,143,0,2...   \n",
       "1003   0/0:0:419:416:0:0:272,144,0,0:-1:-1:0:2530.343:.   \n",
       "1004  0/0:0.5660855:422:421:1:0.002369668:270,151,0,...   \n",
       "...                                                 ...   \n",
       "6200                  ./.:0:6:6:0:0:6,0,0,0:-1:-1:0:0:.   \n",
       "6201         ./.:0:12:12:0:0:8,4,0,0:-1:-1:0:18.57706:.   \n",
       "6202                  ./.:0:7:7:0:0:2,5,0,0:-1:-1:0:0:.   \n",
       "6203                  ./.:0:8:8:0:0:2,6,0,0:-1:-1:0:0:.   \n",
       "6204                  ./.:0:1:1:0:0:0,1,0,0:-1:-1:0:0:.   \n",
       "\n",
       "                                        CHIP_b3_Input  \\\n",
       "1000   0/0:0:108:108:0:0:53,55,0,0:-1:-1:0:416.3386:.   \n",
       "1001   0/0:0:141:141:0:0:70,71,0,0:-1:-1:0:744.5862:.   \n",
       "1002   0/0:0:161:160:0:0:94,66,0,0:-1:-1:0:819.4477:.   \n",
       "1003  0/0:0:190:190:0:0:113,77,0,0:-1:-1:0:1078.799:.   \n",
       "1004  0/0:0:189:186:0:0:107,79,0,0:-1:-1:0:730.9136:.   \n",
       "...                                               ...   \n",
       "6200      0/0:0:26:26:0:0:18,8,0,0:-1:-1:0:50.65918:.   \n",
       "6201      0/0:0:46:46:0:0:25,21,0,0:-1:-1:0:190.525:.   \n",
       "6202       0/0:0:30:30:0:0:8,22,0,0:-1:-1:0:85.0804:.   \n",
       "6203      0/0:0:24:24:0:0:5,19,0,0:-1:-1:0:64.55056:.   \n",
       "6204       ./.:0:12:12:0:0:5,7,0,0:-1:-1:0:12.84924:.   \n",
       "\n",
       "                                        CHIP_b4_Control  \\\n",
       "1000  0/0:21.71786:356:353:3:0.008426966:162,191,2,1...   \n",
       "1001   0/0:0:446:444:0:0:230,214,0,0:-1:-1:0:2570.377:.   \n",
       "1002  0/0:21.33713:492:489:2:0.004065041:288,201,2,0...   \n",
       "1003        0/0:0:562:561:0:0:346,215,0,0:-1:-1:0:Inf:.   \n",
       "1004   0/0:0:557:555:0:0:335,220,0,0:-1:-1:0:2283.826:.   \n",
       "...                                                 ...   \n",
       "6200       0/0:0:53:53:0:0:24,29,0,0:-1:-1:0:153.1909:.   \n",
       "6201  0/0:12.14579:106:105:1:0.009433962:51,54,1,0:9...   \n",
       "6202       0/0:0:58:58:0:0:23,35,0,0:-1:-1:0:204.0312:.   \n",
       "6203       0/0:0:60:60:0:0:21,39,0,0:-1:-1:0:239.7105:.   \n",
       "6204  0/0:33.65802:34:32:2:0.05882353:15,17,2,0:99:1...   \n",
       "\n",
       "                                          CHIP_b4_Flt3l  \\\n",
       "1000   0/0:0:315:314:0:0:160,154,0,0:-1:-1:0:1305.324:.   \n",
       "1001   0/0:0:381:379:0:0:208,171,0,0:-1:-1:0:2196.873:.   \n",
       "1002   0/0:0:422:422:0:0:251,171,0,0:-1:-1:0:2310.822:.   \n",
       "1003   0/0:0:456:456:0:0:289,167,0,0:-1:-1:0:2738.056:.   \n",
       "1004   0/0:0:451:449:0:0:288,161,0,0:-1:-1:0:1836.398:.   \n",
       "...                                                 ...   \n",
       "6200       0/0:0:38:38:0:0:22,16,0,0:-1:-1:0:90.83723:.   \n",
       "6201       0/0:0:55:55:0:0:28,27,0,0:-1:-1:0:249.1141:.   \n",
       "6202        0/0:0:29:28:0:0:8,20,0,0:-1:-1:0:85.66831:.   \n",
       "6203       0/0:0:32:32:0:0:11,21,0,0:-1:-1:0:87.73926:.   \n",
       "6204  ./.:36.06597:21:19:2:0.0952381:11,8,1,1:0.4245...   \n",
       "\n",
       "                                        CHIP_b4_Input  \n",
       "1000   0/0:0:118:118:0:0:67,51,0,0:-1:-1:0:458.7915:.  \n",
       "1001   0/0:0:154:153:0:0:94,59,0,0:-1:-1:0:833.2017:.  \n",
       "1002  0/0:0:176:174:0:0:124,50,0,0:-1:-1:0:933.9806:.  \n",
       "1003  0/0:0:208:206:0:0:137,69,0,0:-1:-1:0:1201.061:.  \n",
       "1004  0/0:0:198:197:0:0:128,69,0,0:-1:-1:0:772.6831:.  \n",
       "...                                               ...  \n",
       "6200       ./.:0:12:12:0:0:6,6,0,0:-1:-1:0:11.25068:.  \n",
       "6201      ./.:0:20:20:0:0:13,7,0,0:-1:-1:0:46.51547:.  \n",
       "6202                ./.:0:8:8:0:0:2,6,0,0:-1:-1:0:0:.  \n",
       "6203        ./.:0:12:12:0:0:3,9,0,0:-1:-1:0:14.7431:.  \n",
       "6204                ./.:0:9:9:0:0:6,3,0,0:-1:-1:0:0:.  \n",
       "\n",
       "[5205 rows x 21 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_df = vcf_df.loc[vcf_df[\"#CHROM\"].str.contains(\"^chr\")]\n",
    "vcf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "03ac1156",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10018-10018  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10018_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10031-10031  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10031_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10039-10039  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10039_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10051-10051  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10051_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10055-10055  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10055_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10057-10057  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10057_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10060-10060  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10060_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10065-10065  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10065_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10106-10106  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10106_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10109-10109  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10109_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10110-10110  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10110_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10122-10122  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10122_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10130-10130  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10130_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10131-10131  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10131_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10135-10135  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10135_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10135-10135  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10135_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10138-10138  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10138_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10140-10140  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10140_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10142-10142  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10142_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10158-10158  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10158_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10170-10170  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10170_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10177-10177  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10177_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10178-10178  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10178_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10192-10192  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10192_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10197-10197  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10197_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10218-10218  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10218_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10232-10232  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10232_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10234-10234  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10234_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10241-10241  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10241_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10250-10250  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10250_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10263-10263  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10263_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10273-10273  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10273_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10365-10365  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10365_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10366-10366  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10366_CHIP_b1_Input.bam\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10377-10377  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10377_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10377-10377  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10377_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10403-10403  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10403_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10421-10421  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10421_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10439-10439  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10439_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10443-10443  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10443_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:180766-180766  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_180766_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:180813-180813  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_180813_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:180816-180816  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_180816_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:180828-180828  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_180828_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:180880-180880  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_180880_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:180915-180915  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_180915_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:180930-180930  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_180930_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:180960-180960  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_180960_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:180961-180961  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_180961_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:5992521-5992521  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_5992521_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:8427760-8427760  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_8427760_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10032940-10032940  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10032940_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10032942-10032942  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10032942_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:10032945-10032945  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_10032945_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:11262677-11262677  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_11262677_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:11908266-11908266  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_11908266_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16513836-16513836  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16513836_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16513925-16513925  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16513925_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514077-16514077  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514077_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514101-16514101  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514101_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514106-16514106  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514106_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514116-16514116  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514116_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514125-16514125  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514125_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514129-16514129  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514129_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514148-16514148  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514148_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514179-16514179  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514179_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514194-16514194  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514194_CHIP_b1_Input.bam\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514226-16514226  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514226_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514262-16514262  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514262_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16514273-16514273  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16514273_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16644730-16644730  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16644730_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666652-16666652  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666652_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666693-16666693  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666693_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666730-16666730  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666730_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666742-16666742  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666742_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666747-16666747  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666747_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666754-16666754  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666754_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666792-16666792  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666792_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666811-16666811  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666811_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666842-16666842  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666842_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666887-16666887  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666887_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666889-16666889  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666889_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666928-16666928  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666928_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666932-16666932  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666932_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666957-16666957  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666957_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666957-16666957  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666957_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666960-16666960  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666960_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16666961-16666961  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16666961_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16667079-16667079  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16667079_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16667109-16667109  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16667109_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16667114-16667114  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16667114_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16667223-16667223  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16667223_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16732718-16732718  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16732718_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740127-16740127  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740127_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740241-16740241  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740241_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740385-16740385  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740385_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740503-16740503  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740503_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740504-16740504  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740504_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740507-16740507  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740507_CHIP_b1_Input.bam\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740507-16740507  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740507_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740536-16740536  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740536_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740562-16740562  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740562_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740575-16740575  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740575_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740577-16740577  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740577_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740622-16740622  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740622_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740672-16740672  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740672_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740709-16740709  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740709_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740710-16740710  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740710_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740771-16740771  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740771_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740812-16740812  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740812_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16740863-16740863  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16740863_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16895977-16895977  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16895977_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16895996-16895996  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16895996_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16896000-16896000  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16896000_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16896039-16896039  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16896039_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16896041-16896041  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16896041_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16896050-16896050  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16896050_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16896136-16896136  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16896136_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16896149-16896149  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16896149_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16896289-16896289  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16896289_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16896340-16896340  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16896340_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16896390-16896390  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16896390_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16904916-16904916  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16904916_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16904997-16904997  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16904997_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16905010-16905010  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16905010_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16905232-16905232  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16905232_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16905350-16905350  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16905350_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:16905351-16905351  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_16905351_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:23344492-23344492  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_23344492_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:23344503-23344503  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_23344503_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:23344504-23344504  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_23344504_CHIP_b1_Input.bam\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:23344514-23344514  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_23344514_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:23344522-23344522  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_23344522_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:23980384-23980384  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_23980384_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:27490267-27490267  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_27490267_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:27725677-27725677  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_27725677_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:28247247-28247247  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_28247247_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:28927037-28927037  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_28927037_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:28927060-28927060  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_28927060_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:30718246-30718246  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_30718246_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:32291949-32291949  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_32291949_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:32291955-32291955  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_32291955_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:32992687-32992687  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_32992687_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:32992695-32992695  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_32992695_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:32992698-32992698  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_32992698_CHIP_b1_Input.bam\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:32992719-32992719  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_32992719_CHIP_b1_Input.bam\n",
      "^C\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:32992725-32992725  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_32992725_CHIP_b1_Input.bam\n",
      "^C\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:36224267-36224267  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_36224267_CHIP_b1_Input.bam\n",
      "^C\n",
      "samtools view -b /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b1_Input.bam chr1:38225218-38225218  > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/regions/chr1_38225218_CHIP_b1_Input.bam\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'child' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/mttrace/lib/python3.7/site-packages/IPython/utils/_process_posix.py\u001b[0m in \u001b[0;36msystem\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    156\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 157\u001b[0;31m                 \u001b[0mchild\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpexpect\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspawn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msh\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'-c'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcmd\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# Vanilla Pexpect\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    158\u001b[0m             \u001b[0mflush\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mttrace/lib/python3.7/site-packages/pexpect/pty_spawn.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, command, args, timeout, maxread, searchwindowsize, logfile, cwd, env, ignore_sighup, echo, preexec_fn, encoding, codec_errors, dimensions, use_poll)\u001b[0m\n\u001b[1;32m    204\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 205\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_spawn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpreexec_fn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdimensions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    206\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muse_poll\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0muse_poll\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mttrace/lib/python3.7/site-packages/pexpect/pty_spawn.py\u001b[0m in \u001b[0;36m_spawn\u001b[0;34m(self, command, args, preexec_fn, dimensions)\u001b[0m\n\u001b[1;32m    303\u001b[0m         self.ptyproc = self._spawnpty(self.args, env=self.env,\n\u001b[0;32m--> 304\u001b[0;31m                                      cwd=self.cwd, **kwargs)\n\u001b[0m\u001b[1;32m    305\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mttrace/lib/python3.7/site-packages/pexpect/pty_spawn.py\u001b[0m in \u001b[0;36m_spawnpty\u001b[0;34m(self, args, **kwargs)\u001b[0m\n\u001b[1;32m    314\u001b[0m         \u001b[0;34m'''Spawn a pty and return an instance of PtyProcess.'''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 315\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mptyprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPtyProcess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspawn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    316\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mttrace/lib/python3.7/site-packages/ptyprocess/ptyprocess.py\u001b[0m in \u001b[0;36mspawn\u001b[0;34m(cls, argv, cwd, env, echo, preexec_fn, dimensions, pass_fds)\u001b[0m\n\u001b[1;32m    314\u001b[0m         \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexec_err_pipe_write\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 315\u001b[0;31m         \u001b[0mexec_err_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexec_err_pipe_read\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m4096\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    316\u001b[0m         \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexec_err_pipe_read\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_2676642/3911683652.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mcmd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"\"\"samtools view -b {b} {region_str}  > {regions_outdir}/{region_f}_{basename(b)}\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m     \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'{cmd}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/mttrace/lib/python3.7/site-packages/ipykernel/zmqshell.py\u001b[0m in \u001b[0;36msystem_piped\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    635\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muser_ns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'_exit_code'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    636\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 637\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muser_ns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'_exit_code'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvar_expand\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdepth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    638\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    639\u001b[0m     \u001b[0;31m# Ensure new system_piped implementation is used\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mttrace/lib/python3.7/site-packages/IPython/utils/_process_posix.py\u001b[0m in \u001b[0;36msystem\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    171\u001b[0m             \u001b[0;31m# (the character is known as ETX for 'End of Text', see\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    172\u001b[0m             \u001b[0;31m# curses.ascii.ETX).\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 173\u001b[0;31m             \u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msendline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    174\u001b[0m             \u001b[0;31m# Read and print any more output the program might produce on its\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    175\u001b[0m             \u001b[0;31m# way out.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'child' referenced before assignment"
     ]
    }
   ],
   "source": [
    "for ind, region in vcf_df.iterrows(): \n",
    "    #for b in glob(bam_fs+\"/*.bam\"):\n",
    "    region_str = f\"{region['#CHROM']}:{region['POS']}-{region['POS']}\"\n",
    "    region_f =  f\"{region['#CHROM']}_{region['POS']}\" #region_str.replace(\":\", \"_\").replace(\"-\", \"_\")\n",
    "    cmd = f\"\"\"samtools view -b {b} {region_str}  > {regions_outdir}/{region_f}_{basename(b)}\"\"\"\n",
    "    print(cmd)\n",
    "    !{cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4401662",
   "metadata": {},
   "source": [
    "## Break-up bam into single-cell. Then do pileup for each. \n",
    "a. Sort by CB\n",
    "b. Loop "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f592184",
   "metadata": {},
   "outputs": [],
   "source": [
    "### a. split_by_cb but add the read group or the id to the name (have a dictionary)\n",
    "### b. run the pileup for snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4632da11",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\"samtools mpileup -f reference.fa -r chr22:425236-425236 alignments.bam | cut -f 5 | tr '[a-z]' '[A-Z]' | fold -w 1 | sort | uniq -c\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2d292d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_f = vcf+\".region.txt\"\n",
    "vcf_df = read_csv_multichar(vcf, sep=\"\\t\", multicomment=\"##\")\n",
    "vcf_df[[\"#CHROM\", \"POS\"]].to_csv(region_f, index=None, sep=\"\\t\", header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b020df2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "07c5d6fa",
   "metadata": {},
   "source": [
    "## pileup for each cb and position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c089112",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for ind, region in vcf_df.iterrows(): \n",
    "#     #for b in glob(bam_fs+\"/*.bam\"):\n",
    "#     region_str = f\"{region['#CHROM']}:{region['POS']}-{region['POS']}\"\n",
    "#     region_f =  f\"{region['#CHROM']}_{region['POS']}\" #region_str.replace(\":\", \"_\").replace(\"-\", \"_\")\n",
    "#     cmd = f\"\"\"samtools view -b {b} {region_str} | samtools fillmd -e - {genome_fa} | grep -v \"^@\"| awk -v pos=\"{region['POS']}\" 'BEGIN {{OFS = FS = \"\\t\" }} ; {{n=split($10,a,\"\") ; if(a[(pos-$4)+1] != \"=\" ) print pos,(pos-$4)+1, a[(pos-$4)+1], $1, $4, $10 }}'  > {outdir}/{basename(b)}_{region_f}\"\"\"\n",
    "#     print(cmd)\n",
    "#         #!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d1ba37c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools mpileup -t AD -f /data/Mito_Trace/data/processed/genomes/mtMasked/GRCh38_MT_blacklist_A2_2020/fasta/genome.fa -l /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/variants.all.vcf.region.txt /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/preproc/CHIP_b3_Control.bam > /data/Mito_Trace/output/somatic_variants/CHIP_dec172021/aggregate/bam_vcf_overlap/CHIP_b3_Control.bam.pileup.txt\n"
     ]
    }
   ],
   "source": [
    "cmd = f\"samtools mpileup -t AD -f {genome_fa} -l {region_f} {b} > {join(outdir, basename(b)+'.pileup.txt')}\"\n",
    "print(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0a443d3f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10018</td>\n",
       "      <td>C</td>\n",
       "      <td>186</td>\n",
       "      <td>,,,,,,,g,,.,.,,,,,,,,,,,,,,,,,,,,,,,,,,,g,,......</td>\n",
       "      <td>AjkFCC.;FQ\\FkFFQFFFkkFFFFkFFFkkFFFkQkFF:&gt;kk_kk...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10031</td>\n",
       "      <td>T</td>\n",
       "      <td>245</td>\n",
       "      <td>,,,,,,,,,..,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,......</td>\n",
       "      <td>kkkCC.;kk^&lt;kiFkFkkFFFFk_FF_kFFkk_kFkFk_::k_Ekk...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10039</td>\n",
       "      <td>A</td>\n",
       "      <td>259</td>\n",
       "      <td>,,,,,,,,,,,,,,,,,,,,,,,+1c,,,,,,,,,,,....,,,,,...</td>\n",
       "      <td>kkCC.kkkikkFkQFk_Q:kkFFkkkkkFk_EkE&gt;kk`BFF:F^:g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10051</td>\n",
       "      <td>A</td>\n",
       "      <td>293</td>\n",
       "      <td>,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,....,,,,.........</td>\n",
       "      <td>khCSkkk:k_kkk_FkkFF__kkQkkkk9]kkkkkFFF:^:gFFQF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10055</td>\n",
       "      <td>T</td>\n",
       "      <td>295</td>\n",
       "      <td>,,,,,,,,,,,,,,,,,,,,,,,,,,,,,...,,,,,,,.....,,...</td>\n",
       "      <td>Sk_kFk_kkkQFEk&gt;FkQQkkQkkk9ikkkk`FF:kF:^gFkFFi:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6045</th>\n",
       "      <td>KI270337.1</td>\n",
       "      <td>852</td>\n",
       "      <td>C</td>\n",
       "      <td>12</td>\n",
       "      <td>,$,.......,,.</td>\n",
       "      <td>DFFkFFFkFFFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6046</th>\n",
       "      <td>KI270337.1</td>\n",
       "      <td>894</td>\n",
       "      <td>C</td>\n",
       "      <td>7</td>\n",
       "      <td>...,,..</td>\n",
       "      <td>:FFFFF&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6047</th>\n",
       "      <td>KI270336.1</td>\n",
       "      <td>923</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>,,,,,.</td>\n",
       "      <td>k:Fkkk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6048</th>\n",
       "      <td>KI270336.1</td>\n",
       "      <td>937</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>,,,,,.</td>\n",
       "      <td>_FFkQk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6049</th>\n",
       "      <td>KI270336.1</td>\n",
       "      <td>969</td>\n",
       "      <td>C</td>\n",
       "      <td>5</td>\n",
       "      <td>,...,</td>\n",
       "      <td>FFkkk</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6050 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0  ...                                                  5\n",
       "0           chr1  ...  AjkFCC.;FQ\\FkFFQFFFkkFFFFkFFFkkFFFkQkFF:>kk_kk...\n",
       "1           chr1  ...  kkkCC.;kk^<kiFkFkkFFFFk_FF_kFFkk_kFkFk_::k_Ekk...\n",
       "2           chr1  ...  kkCC.kkkikkFkQFk_Q:kkFFkkkkkFk_EkE>kk`BFF:F^:g...\n",
       "3           chr1  ...  khCSkkk:k_kkk_FkkFF__kkQkkkk9]kkkkkFFF:^:gFFQF...\n",
       "4           chr1  ...  Sk_kFk_kkkQFEk>FkQQkkQkkk9ikkkk`FF:kF:^gFkFFi:...\n",
       "...          ...  ...                                                ...\n",
       "6045  KI270337.1  ...                                       DFFkFFFkFFFF\n",
       "6046  KI270337.1  ...                                            :FFFFF>\n",
       "6047  KI270336.1  ...                                             k:Fkkk\n",
       "6048  KI270336.1  ...                                             _FFkQk\n",
       "6049  KI270336.1  ...                                              FFkkk\n",
       "\n",
       "[6050 rows x 6 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampile = pd.read_csv(join(outdir, basename(b)+'.pileup.txt'), sep=\"\\t\", header=None)\n",
    "sampile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "02927af3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[warning] samtools mpileup option `t` is functional, but deprecated. Please switch to using bcftools mpileup in future.\n",
      "[mpileup] 1 samples in 1 input files\n"
     ]
    }
   ],
   "source": [
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90a08b69",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr1    143215235       .       G       C       57.48579        .       TYPE=snv;NS=12;AF=0.08333333;DP=2830;RO=2787;AO=10;SRF=1557;SRR=1230;SAF=5;SAR=5;SOR=0.4849351;RVSB=0.5584555;FS=1.206317;ERR=0.0002171339;SIG=0.001;CONT=ATCxAAT       GT:QVAL:DP:RO:AO:AF:SB:SOR:RVSB:FS:QVAL_minAF:STATUS    0/1:57.48579:526:517:5:0.009505703:289,228,2,3:1.05515:0.6539326:1.809933:2665.46:.     0/0:30.6072:450:442:3:0.006666667:260,182,2,1:1.085189:0.5828025:0:2263.195:.   0/0:17.91017:427:417:2:0.004683841:243,174,1,1:0.4139138:0.5823389:0:2136.623:. 0/0:0:47:46:0:0:24,22,0,0:-1:-1:0:165.2151:.    ./.:0:23:23:0:0:13,10,0,0:-1:-1:0:38.25952:.    0/0:0:1083:1072:0:0:576,496,0,0:-1:-1:0:Inf:.   ./.:0:10:9:0:0:4,5,0,0:-1:-1:0:0:.      ./.:0:15:15:0:0:11,4,0,0:-1:-1:0:14.46496:.     0/0:0:55:55:0:0:29,26,0,0:-1:-1:0:217.4113:.    0/0:0:62:61:0:0:35,26,0,0:-1:-1:0:241.4051:.    0/0:0:77:75:0:0:45,30,0,0:-1:-1:0:317.5159:.    0/0:0:55:55:0:0:28,27,0,0:-1:-1:0:217.4113:.\n",
    "chr1    143215344       .       G       C       78.56018        .       TYPE=snv;NS=12;AF=0.08333333;DP=2838;RO=2814;AO=12;SRF=996;SRR=1818;SAF=4;SAR=8;SOR=0.7887207;RVSB=0.5227392;FS=0;ERR=0.0002879004;SIG=0.00"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
