{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d165efb4",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# indir = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/somatic_variants/needle/peaks_all/\"\n",
    "# outdir = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/somatic_variants/needle/peaks_all/som_dendro_0.6/clones_variants_summary\"\n",
    "# cells_meta_f = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/cells_meta.tsv\"\n",
    "# dendro_indir = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/barcodes/btwnClones_dendro_dt_0.6\"\n",
    "# to_chip = False\n",
    "# chip_genes_f = \"/data/Mito_Trace/output/geneRegions/gff_A2/chip_genes.bed\"\n",
    "\n",
    "dendro_indir = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/barcodes/btwnClones_dendro_dt_0.6\"\n",
    "cells_meta_f = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/cells_meta.tsv\"\n",
    "indir = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/somatic_variants/gatk_mutect/peaks_atac_peaks\" \n",
    "outdir = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/somatic_variants/gatk_mutect/peaks_atac_peaks/som_dendro_0.6/chip_clones_variants_summary\"\n",
    "to_chip = True \n",
    "chip_genes_f = \"/data/Mito_Trace/output/geneRegions/gff_A2/chip_genes.bed\" \n",
    "n_cpu =16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "21ae2b63",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fig_utils from mplh\n"
     ]
    }
   ],
   "source": [
    "from src.utils.data_io import read_csv_multichar\n",
    "import mplh.cluster_help as ch\n",
    "import mplh.fig_utils as fu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9da443f6",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "from os.path import basename, join, dirname, exists\n",
    "import pandas as pd\n",
    "from pandarallel import pandarallel\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c6d2e7fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 16 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "pandarallel.initialize(nb_workers=n_cpu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "891ea687",
   "metadata": {},
   "outputs": [],
   "source": [
    "import src.utils.dendro_clust as dc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "49c011fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#outdir = join(indir, \"som_dendro\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "366a1217",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>donor</th>\n",
       "      <th>condition</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGAATCAAC-1_Control</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Control</td>\n",
       "      <td>0_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAATGAGAGAGACTCG-1_Control</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Control</td>\n",
       "      <td>0_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACAAAGTCAAGTTGC-1_Control</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Control</td>\n",
       "      <td>0_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACAAAGTCACATCCC-1_Control</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Control</td>\n",
       "      <td>0_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACCAACGTTCATTTC-1_Control</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Control</td>\n",
       "      <td>0_0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            lineage  donor condition name\n",
       "ID                                                       \n",
       "AAACGAAAGAATCAAC-1_Control        0      0   Control  0_0\n",
       "AAATGAGAGAGACTCG-1_Control        0      0   Control  0_0\n",
       "AACAAAGTCAAGTTGC-1_Control        0      0   Control  0_0\n",
       "AACAAAGTCACATCCC-1_Control        0      0   Control  0_0\n",
       "AACCAACGTTCATTTC-1_Control        0      0   Control  0_0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cells_meta = pd.read_csv(cells_meta_f,sep=\"\\t\").set_index(\"ID\")\n",
    "cells_meta[\"name\"] = cells_meta.apply(lambda x: f\"{x['donor']}_{x['lineage']}\", axis=1)\n",
    "donors = cells_meta[\"donor\"].unique()\n",
    "conditions = cells_meta[\"condition\"].unique()\n",
    "\n",
    "\n",
    "cells_meta.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "530f6cc6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      chr      start        end name score strand       6     7  8  \\\n",
      "0    chr1   43332848   43357772    .     .      +  HAVANA  gene  .   \n",
      "1   chr11  119201275  119318926    .     .      +  HAVANA  gene  .   \n",
      "2   chr15   90078044   90107504    .     .      -  HAVANA  gene  .   \n",
      "3   chr17    7656778    7692550    .     .      -  HAVANA  gene  .   \n",
      "4   chr17   76729114   76742374    .     .      -  HAVANA  gene  .   \n",
      "5    chr2   25222854   25347590    .     .      -  HAVANA  gene  .   \n",
      "6    chr2  197383514  197440091    .     .      -  HAVANA  gene  .   \n",
      "7    chr2  208231226  208271074    .     .      -  HAVANA  gene  .   \n",
      "8   chr20   32353329   32444319    .     .      +  HAVANA  gene  .   \n",
      "9   chr21   43087955   43112587    .     .      -  HAVANA  gene  .   \n",
      "10   chr4  105140874  105284816    .     .      +  HAVANA  gene  .   \n",
      "11   chr7  148802382  148889321    .     .      -  HAVANA  gene  .   \n",
      "12   chr8  116840934  116879866    .     .      -  HAVANA  gene  .   \n",
      "\n",
      "                                                    9    10  \n",
      "0   gene_id \"ENSG00000117400\"; gene_version \"17\"; ...  True  \n",
      "1   gene_id \"ENSG00000110395\"; gene_version \"6\"; g...  True  \n",
      "2   gene_id \"ENSG00000182054\"; gene_version \"9\"; g...  True  \n",
      "3   gene_id \"ENSG00000141510\"; gene_version \"17\"; ...  True  \n",
      "4   gene_id \"ENSG00000161547\"; gene_version \"16\"; ...  True  \n",
      "5   gene_id \"ENSG00000119772\"; gene_version \"16\"; ...  True  \n",
      "6   gene_id \"ENSG00000115524\"; gene_version \"16\"; ...  True  \n",
      "7   gene_id \"ENSG00000138413\"; gene_version \"13\"; ...  True  \n",
      "8   gene_id \"ENSG00000171456\"; gene_version \"19\"; ...  True  \n",
      "9   gene_id \"ENSG00000160201\"; gene_version \"11\"; ...  True  \n",
      "10  gene_id \"ENSG00000168769\"; gene_version \"13\"; ...  True  \n",
      "11  gene_id \"ENSG00000106462\"; gene_version \"10\"; ...  True  \n",
      "12  gene_id \"ENSG00000164754\"; gene_version \"14\"; ...  True  \n"
     ]
    }
   ],
   "source": [
    "if to_chip:\n",
    "    chip_df = pd.read_csv(chip_genes_f,sep=\"\\t\", header=None)\n",
    "    cols = chip_df.columns.astype(\"object\").values \n",
    "    cols[:6] = [\"chr\", \"start\", \"end\", \"name\",\"score\",\"strand\"]\n",
    "    chip_df.columns = cols\n",
    "    print(chip_df)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5522a5ec",
   "metadata": {},
   "source": [
    "## Load the somatic variants number of cells and merge across conditions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f231210e",
   "metadata": {},
   "source": [
    "## Load dendro labels and barcodes mean af"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcd874fd",
   "metadata": {},
   "source": [
    "### Merge the variants as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4e78ae71",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_chip_gene(chip_gene, curr_var ):\n",
    "    if (curr_var[\"chr\"] == chip_gene[\"chr\"]):\n",
    "        if (chip_gene[\"strand\"] == \"+\"):\n",
    "            if curr_var[\"pos\"]>=chip_gene[\"start\"] & curr_var[\"pos\"]<=chip_gene[\"end\"]:\n",
    "                return True    \n",
    "        elif (chip_gene[\"strand\"] == \"-\"):\n",
    "            if curr_var[\"pos\"]<=chip_gene[\"start\"] & curr_var[\"pos\"]>=chip_gene[\"end\"]:\n",
    "                return True\n",
    "        else: \n",
    "            print(\"No strand on gene. Returning False\")\n",
    "    return False\n",
    "\n",
    "def is_in_any_chip_gene(curr_var, chip_bed):\n",
    "    for ind, curr_gene in chip_bed.iterrows():\n",
    "        #print('curr_gene')\n",
    "        #print(curr_gene)\n",
    "        if is_chip_gene(curr_gene, curr_var):\n",
    "            return True\n",
    "    return False\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d14d4af",
   "metadata": {},
   "source": [
    "#### Get dendro cluster and mean af for each donor and concat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "af749eb4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "clone_som_d = {}\n",
    "clone_som_vars_d = {}\n",
    "\n",
    "# for c in conditions:\n",
    "#     try:\n",
    "#         print(c)\n",
    "#         clone_som_d[c] = pd.read_csv(join(indir, c, \"clones_alt_numCells.csv\"), index_col=0)\n",
    "#         clone_som_vars_d[c] = pd.read_csv(join(indir, c, \"clone_alt_numCells.variants.tsv\"), sep=\"\\t\")\n",
    "#         clone_som_d[c][\"condition\"] = c\n",
    "#         clone_som_d[c][\"name\"] = clone_som_d[c].index\n",
    "#     except:\n",
    "#         print(f\"{c} not here\")\n",
    "\n",
    "for c in conditions:\n",
    "    print(c)\n",
    "    if not exists(join(indir, c, \"clones_alt_numCells.csv\")):\n",
    "        print(f\"{c} numCells not here\")\n",
    "    elif not exists(join(indir, c, \"clone_alt_numCells.variants.tsv\")):\n",
    "        print(f\"{c} numCell variants not here\")    \n",
    "    else:\n",
    "        clone_som_d[c] = pd.read_csv(join(indir, c, \"clones_alt_numCells.csv\"), index_col=0)\n",
    "        clone_som_vars_d[c] = pd.read_csv(join(indir, c, \"clone_alt_numCells.variants.tsv\"), sep=\"\\t\")\n",
    "        clone_som_d[c][\"condition\"] = c\n",
    "        clone_som_d[c][\"name\"] = clone_som_d[c].index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f7ef3e0d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>var</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>posID</th>\n",
       "      <th>ref</th>\n",
       "      <th>Allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Control</th>\n",
       "      <th>0</th>\n",
       "      <td>chr10_100267146_C</td>\n",
       "      <td>CWF19L1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant;i...</td>\n",
       "      <td>chr10_100267146</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10_102504349_T</td>\n",
       "      <td>SUFU;ACTR1A</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant;intron_variant</td>\n",
       "      <td>chr10_102504349</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10_102646092_A</td>\n",
       "      <td>ENSG00000272933;TRIM8</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant;intron_variant,non_codin...</td>\n",
       "      <td>chr10_102646092</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10_102963521_G</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>chr10_102963521</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10_110211097_A</td>\n",
       "      <td>ENSG00000228417;MXI1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant;intron_variant</td>\n",
       "      <td>chr10_110211097</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Input</th>\n",
       "      <th>765</th>\n",
       "      <td>chrY_11333113_A</td>\n",
       "      <td>ENSG00000258991;ENSG00000270455</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>non_coding_transcript_exon_variant;upstream_ge...</td>\n",
       "      <td>chrY_11333113</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>chrY_11333831_T</td>\n",
       "      <td>ENSG00000258991;ENSG00000270455</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant;upstream_gene_variant</td>\n",
       "      <td>chrY_11333831</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>767</th>\n",
       "      <td>chrY_11334230_T</td>\n",
       "      <td>ENSG00000258991;ENSG00000270455</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant;upstream_gene_variant</td>\n",
       "      <td>chrY_11334230</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>chrY_11334344_A</td>\n",
       "      <td>ENSG00000258991;ENSG00000270455</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant;upstream_gene_variant</td>\n",
       "      <td>chrY_11334344</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769</th>\n",
       "      <td>chrY_11334378_A</td>\n",
       "      <td>ENSG00000258991;ENSG00000270455</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant;upstream_gene_variant</td>\n",
       "      <td>chrY_11334378</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1519 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           var                        Gene Name Feature_type  \\\n",
       "Control 0    chr10_100267146_C                          CWF19L1   Transcript   \n",
       "        1    chr10_102504349_T                      SUFU;ACTR1A   Transcript   \n",
       "        2    chr10_102646092_A            ENSG00000272933;TRIM8   Transcript   \n",
       "        3    chr10_102963521_G                            CNNM2   Transcript   \n",
       "        4    chr10_110211097_A             ENSG00000228417;MXI1   Transcript   \n",
       "...                        ...                              ...          ...   \n",
       "Input   765    chrY_11333113_A  ENSG00000258991;ENSG00000270455   Transcript   \n",
       "        766    chrY_11333831_T  ENSG00000258991;ENSG00000270455   Transcript   \n",
       "        767    chrY_11334230_T  ENSG00000258991;ENSG00000270455   Transcript   \n",
       "        768    chrY_11334344_A  ENSG00000258991;ENSG00000270455   Transcript   \n",
       "        769    chrY_11334378_A  ENSG00000258991;ENSG00000270455   Transcript   \n",
       "\n",
       "                                                   Consequence  \\\n",
       "Control 0    intron_variant,non_coding_transcript_variant;i...   \n",
       "        1                 upstream_gene_variant;intron_variant   \n",
       "        2    upstream_gene_variant;intron_variant,non_codin...   \n",
       "        3                                       intron_variant   \n",
       "        4                 upstream_gene_variant;intron_variant   \n",
       "...                                                        ...   \n",
       "Input   765  non_coding_transcript_exon_variant;upstream_ge...   \n",
       "        766      downstream_gene_variant;upstream_gene_variant   \n",
       "        767      downstream_gene_variant;upstream_gene_variant   \n",
       "        768      downstream_gene_variant;upstream_gene_variant   \n",
       "        769      downstream_gene_variant;upstream_gene_variant   \n",
       "\n",
       "                       posID ref Allele  \n",
       "Control 0    chr10_100267146   G      C  \n",
       "        1    chr10_102504349   G      T  \n",
       "        2    chr10_102646092   C      A  \n",
       "        3    chr10_102963521   T      G  \n",
       "        4    chr10_110211097   C      A  \n",
       "...                      ...  ..    ...  \n",
       "Input   765    chrY_11333113   G      A  \n",
       "        766    chrY_11333831   G      T  \n",
       "        767    chrY_11334230   A      T  \n",
       "        768    chrY_11334344   C      A  \n",
       "        769    chrY_11334378   C      A  \n",
       "\n",
       "[1519 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clone_som_df = pd.concat(clone_som_d, axis=0).reset_index(drop=True).fillna(0)\n",
    "clone_som_merge_df = clone_som_df.drop(\"condition\",axis=1).groupby(\"name\").agg(sum).astype(int)\n",
    "clone_som_merge_df\n",
    "\n",
    "\n",
    "## Process the variants \n",
    "clone_som_vars_df = pd.concat(clone_som_vars_d) #[conditions[0]].reset_index()\n",
    "# for c in conditions[1:]:\n",
    "#     clone_som_vars_df = pd.merge(clone_som_vars_df, clone_som_vars_d[c].reset_index(), how=\"outer\" ) #, left_index=True, right_index=True, \n",
    "clone_som_vars_df = clone_som_vars_df.drop_duplicates()\n",
    "clone_som_vars_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "dfcbb803",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def add_chr_pos(x):\n",
    "    spl = x.name.split(\"_\")\n",
    "    return [spl[0], spl[1]]\n",
    "\n",
    "\n",
    "clone_som_vars_df = clone_som_vars_df.set_index(\"posID\")\n",
    "clone_som_vars_df[[\"chr\", \"pos\"]] = clone_som_vars_df.parallel_apply(add_chr_pos , axis=1, result_type=\"expand\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6bf3205d",
   "metadata": {},
   "outputs": [],
   "source": [
    "clone_som_vars_df[\"pos\"] =  clone_som_vars_df[\"pos\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7e9dfa4b",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if to_chip:\n",
    "    clone_som_vars_df[\"is_chip\"] = clone_som_vars_df.apply(is_in_any_chip_gene, args=(chip_df,), axis=1)\n",
    "    clone_som_vars_df = clone_som_vars_df[clone_som_vars_df[\"is_chip\"]].set_index(\"var\")\n",
    "\n",
    "clone_som_vars_df[\"Gene Name\"] = clone_som_vars_df[\"Gene Name\"].fillna(\"\")\n",
    "clone_som_vars_df[\"Gene Name\"] = clone_som_vars_df.apply(lambda x: x[\"Gene\"] if x['Gene Name']==\"\" else x['Gene Name'] , axis=1)\n",
    "\n",
    "\n",
    "clone_som_vars_df = clone_som_vars_df.reset_index().drop_duplicates(subset=[\"var\", \"Gene Name\", \"Feature_type\", \"Consequence\"]).sort_values(\"Gene Name\").iloc[::-1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9ddd51bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## For duplicated variants, add to gene\n",
    "def dups(df):\n",
    "    if df.shape[0] == 0:\n",
    "        return df.iloc[0]\n",
    "\n",
    "    cons = \";\".join(list(set(df[\"Consequence\"].values)))\n",
    "    gene_name = \";\".join(list(set(df[\"Gene Name\"].values)))\n",
    "    out = df.iloc[0]\n",
    "    out[\"Consequence\"] = cons\n",
    "    out[\"Gene Name\"] = gene_name\n",
    "    out.name = None\n",
    "    return out\n",
    "\n",
    "clone_som_vars_df_rmdups = clone_som_vars_df.groupby(\"var\").apply(dups)\n",
    "clone_som_vars_df_rmdups = clone_som_vars_df_rmdups[[\"Gene Name\", \"Feature_type\", \"Consequence\"]]\n",
    "\n",
    "\n",
    "\n",
    "if to_chip:\n",
    "    clone_som_merge_df = clone_som_merge_df.loc[:, clone_som_vars_df_rmdups.index]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3be9785c",
   "metadata": {},
   "outputs": [],
   "source": [
    "donors = cells_meta[\"donor\"].unique()\n",
    "conditions = cells_meta[\"condition\"].unique()\n",
    "\n",
    "clones = []\n",
    "for d in donors:\n",
    "    clones.append(pd.read_csv(join(dendro_indir, f\"donor{d}.clones_dendro.csv\"), index_col=0))\n",
    "    clones[-1][\"den_clust\"] = str(d) + \"_\" + clones[-1][\"den_clust\"] \n",
    "clones = pd.concat(clones)\n",
    "\n",
    "# mean_af = pd.concat(mean_af, axis=0, ignore_index=False).fillna(0)\n",
    "# mean_af"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ea0df4f",
   "metadata": {},
   "source": [
    "## Split by donor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6234b06d",
   "metadata": {},
   "source": [
    "# Calculate enrichment for each clone\n",
    "A. Fisher's exact test for variant in clone vs not (account for clone sizes)  \n",
    "B. Remove variants seen in >50% of clones    \n",
    "C. Entropy for the variant num cells in each clone  \n",
    "C. Rank by (%Cells in clone with variant)/(%cells not in clone with variant) .. calculate percent by adding pseudocount of 1 to both\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2faea31b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from scipy.stats import fisher_exact\n",
    "\n",
    "def fish_test(variant_ser, clone_id, clones_meta):\n",
    "    x_pos = variant_ser.loc[clone_id]\n",
    "\n",
    "    x_neg = clones_meta.loc[clone_id, 'size'] - x_pos\n",
    "    \n",
    "    y_pos = sum(variant_ser.loc[variant_ser.index!=clone_id])\n",
    "    y_neg = clones_meta['size'].sum() - x_pos - x_neg - y_pos\n",
    "#     print('clone_id', clone_id)\n",
    "#     print('var', variant_ser.name)\n",
    "#     print(f\"cells in clone with var: {x_pos} ; cells in clone: {x_neg}\")\n",
    "#     print(f\"cells not in clone with var: {y_pos};  cells not in clone: {y_neg} \")\n",
    "    oddsratio, pval = fisher_exact([[x_pos,x_neg],[y_pos,y_neg]], alternative='greater')\n",
    "    return pval\n",
    "\n",
    "def clone_fish(var_ser, clones_meta):\n",
    "    #print(var_ser)\n",
    "    pval_ser = pd.Series(index=var_ser.index, dtype=float)\n",
    "    for clone_id in var_ser.index:\n",
    "        pval_ser.loc[clone_id] = fish_test(var_ser, clone_id, clones_meta)\n",
    "    #print(pval_ser)\n",
    "    return pval_ser.fillna(1)\n",
    "\n",
    "def var_fish(df, clones_meta):\n",
    "    pval_df = df.apply(clone_fish, args=(clones_meta,), axis=0)\n",
    "    return pval_df\n",
    "\n",
    "def get_sig_var_clone(sig_inds, pval_df, ncells_var_df, curr_cl):\n",
    "    \"\"\" Filter for significant p-values and add number of cells information \n",
    "    \"\"\"\n",
    "    sig_var_clone = []\n",
    "    clones = pval_df.index\n",
    "    variants = pval_df.columns\n",
    "    for cl_ind, vrt_ind in sig_inds:\n",
    "        cl = clones[cl_ind]\n",
    "        vrt = variants[vrt_ind]\n",
    "        ncells_cl = curr_cl.loc[cl, \"size\"]\n",
    "        ncells_var_cl = ncells_var_df.loc[cl, vrt]\n",
    "        pct = ncells_var_cl/ncells_cl\n",
    "        \n",
    "        ncells_oth_cl = curr_cl[\"size\"].sum() - ncells_cl\n",
    "        ncells_var_oth_cl = sum(ncells_var_df[vrt]) - ncells_var_cl\n",
    "        pct_oth = ncells_var_oth_cl/ncells_oth_cl\n",
    "        \n",
    "        var_pct = ncells_var_cl/(ncells_var_cl+ncells_var_oth_cl)\n",
    "        \n",
    "        diff = pct-pct_oth\n",
    "        \n",
    "        sig_var_clone.append((cl, vrt, pval_df.loc[cl, vrt], var_pct, diff,\n",
    "                              pct, ncells_var_cl, ncells_cl, \n",
    "                              pct_oth, ncells_var_oth_cl, ncells_oth_cl)) #, curr_cl)\n",
    "        \n",
    "        \n",
    "    return pd.DataFrame(sig_var_clone, columns=[\"clone\", \"variant\", \"Fisher's exact pval\", \n",
    "                                                \"var_pct_in_clone\",\"diff\",\n",
    "                                                \"ncells_pct\", \"ncells_var_cl\", \"ncells_cl\", \n",
    "                                                \"ncells_oth_pct\", \"ncells_var_oth\", \"ncells_oth\" ])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "57eaca26",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_top_clone_vars_donor(curr_som, curr_cl):\n",
    "    pval_df = var_fish(curr_som, curr_cl)\n",
    "    sig_inds = np.argwhere(pval_df.values<0.05)\n",
    "\n",
    "    sig_var_clone = get_sig_var_clone(sig_inds, pval_df, curr_som, curr_cl)\n",
    "    sig_var_clone\n",
    "\n",
    "    ### Filter (add column) where ncells oth clones is high (>100)\n",
    "\n",
    "    sig_var_clone[\"oth_100\"] = sig_var_clone[\"ncells_var_oth\"]>100\n",
    "    sum(sig_var_clone[\"oth_100\"])\n",
    "\n",
    "    sig_var_clone_nooth = sig_var_clone.loc[~(sig_var_clone[\"oth_100\"])]\n",
    "    sig_var_clone_nooth.sort_values([\"var_pct_in_clone\",\"ncells_var_cl\"])[::-1].head(10)\n",
    "\n",
    "    sig_var_clone_nooth = sig_var_clone.loc[~(sig_var_clone[\"oth_100\"])]\n",
    "    sig_var_clone_nooth.sort_values([\"ncells_var_cl\"])[::-1].head(10)\n",
    "\n",
    "    return sig_var_clone_nooth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "22251cfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_top_dendro_vars_donor(curr_som, curr_cl):\n",
    "    ## Do for dendro\n",
    "    den_som_df = pd.concat((curr_som, curr_cl[[\"den_clust\"]]), axis=1, ignore_index=False).groupby(\"den_clust\").agg(sum)\n",
    "    den_som_df\n",
    "\n",
    "    den_clone_df = pd.DataFrame(curr_cl.groupby(\"den_clust\").sum()[\"size\"])\n",
    "    den_clone_df\n",
    "\n",
    "\n",
    "    den_pval_df = var_fish(den_som_df, den_clone_df)\n",
    "    sig_inds = np.argwhere(den_pval_df.values<0.05)\n",
    "\n",
    "    sig_var_dendro = get_sig_var_clone(sig_inds, den_pval_df, den_som_df, den_clone_df)\n",
    "    sig_var_dendro\n",
    "\n",
    "    ### Filter (add column) where ncells oth dendros is high (>100)\n",
    "\n",
    "    sig_var_dendro[\"oth_100\"] = sig_var_dendro[\"ncells_var_oth\"]>100\n",
    "    sum(sig_var_dendro[\"oth_100\"])\n",
    "\n",
    "    sig_var_dendro_nooth = sig_var_dendro.loc[~(sig_var_dendro[\"oth_100\"])]\n",
    "    sig_var_dendro_nooth.sort_values([\"diff\"])[::-1].head(10)\n",
    "    return sig_var_dendro_nooth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "73a0310f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "d 0\n",
      "curr_som (34, 161)\n",
      "d 1\n",
      "curr_som (33, 162)\n"
     ]
    }
   ],
   "source": [
    "donors_clones = {}\n",
    "donors_dendro = {}\n",
    "\n",
    "for d, curr_cl in clones.groupby(\"donor\"):\n",
    "    print('d', d)\n",
    "    curr_som = clone_som_merge_df.loc[clone_som_merge_df.index.isin(curr_cl.index)]\n",
    "    curr_som = curr_som.loc[:, ~((curr_som==0).all())]\n",
    "    print('curr_som', curr_som.shape)\n",
    "    donors_clones[d] = get_top_clone_vars_donor(curr_som, curr_cl)\n",
    "    donors_dendro[d] = get_top_dendro_vars_donor(curr_som, curr_cl)\n",
    "#     if curr_som.shape[0] == 1:\n",
    "#         f = plt.figure()\n",
    "#         sns.heatmap(np.log10(curr_som+1))\n",
    "#     else:\n",
    "#         ch.plot_cluster(np.log10(curr_som+1), row_meta=curr_cl.drop(\"donor\",axis=1), \n",
    "#                    row_clr_schemes={\"size\":\"sequential\", \"den_clust\": \"categorical\", \"donor\":\"categorical\"})\n",
    "\n",
    "#     plt.suptitle(f\"Donor {d}\")\n",
    "    #fu.helper_save(join(outdir, f\"donor{d}_som_vars_dendro.png\"), to_pdf=False)\n",
    "    #fu.helper_save(join(outdir, f\"donor{d}_som_vars_dendro.png\"), to_pdf=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f352e250",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr11_1016905_T</th>\n",
       "      <td>AP2A2;MUC6;MUC6;AP2A2</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant;upstream_gene_variant;downstr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11_10607775_T</th>\n",
       "      <td>LYVE1;IRAG1;IRAG1;LYVE1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant;d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11_108120896_G</th>\n",
       "      <td>ENSG00000255467;ACAT1;ACAT1;ENSG00000255467</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant;i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11_122506084_T</th>\n",
       "      <td>ENSG00000255248</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11_126586049_C</th>\n",
       "      <td>ENSG00000257271;KIRREL3;KIRREL3;ENSG00000257271</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant;i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr4_74189490_C</th>\n",
       "      <td>MTHFD2L</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant;upstream_gene_variant;intron_va...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr4_75672581_C</th>\n",
       "      <td>G3BP2</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant;i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr4_82892864_A</th>\n",
       "      <td>SEC31A;ENSG00000251022;ENSG00000251022;SEC31A</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant;intron_variant;upstrea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr4_84482595_C</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intergenic_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr4_88378194_A</th>\n",
       "      <td>HERC6</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>181 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         Gene Name  \\\n",
       "var                                                                  \n",
       "chr11_1016905_T                              AP2A2;MUC6;MUC6;AP2A2   \n",
       "chr11_10607775_T                           LYVE1;IRAG1;IRAG1;LYVE1   \n",
       "chr11_108120896_G      ENSG00000255467;ACAT1;ACAT1;ENSG00000255467   \n",
       "chr11_122506084_T                                  ENSG00000255248   \n",
       "chr11_126586049_C  ENSG00000257271;KIRREL3;KIRREL3;ENSG00000257271   \n",
       "...                                                            ...   \n",
       "chr4_74189490_C                                            MTHFD2L   \n",
       "chr4_75672581_C                                              G3BP2   \n",
       "chr4_82892864_A      SEC31A;ENSG00000251022;ENSG00000251022;SEC31A   \n",
       "chr4_84482595_C                                                  -   \n",
       "chr4_88378194_A                                              HERC6   \n",
       "\n",
       "                  Feature_type  \\\n",
       "var                              \n",
       "chr11_1016905_T     Transcript   \n",
       "chr11_10607775_T    Transcript   \n",
       "chr11_108120896_G   Transcript   \n",
       "chr11_122506084_T   Transcript   \n",
       "chr11_126586049_C   Transcript   \n",
       "...                        ...   \n",
       "chr4_74189490_C     Transcript   \n",
       "chr4_75672581_C     Transcript   \n",
       "chr4_82892864_A     Transcript   \n",
       "chr4_84482595_C              -   \n",
       "chr4_88378194_A     Transcript   \n",
       "\n",
       "                                                         Consequence  \n",
       "var                                                                   \n",
       "chr11_1016905_T    missense_variant;upstream_gene_variant;downstr...  \n",
       "chr11_10607775_T   intron_variant,non_coding_transcript_variant;d...  \n",
       "chr11_108120896_G  intron_variant,non_coding_transcript_variant;i...  \n",
       "chr11_122506084_T       intron_variant,non_coding_transcript_variant  \n",
       "chr11_126586049_C  intron_variant,non_coding_transcript_variant;i...  \n",
       "...                                                              ...  \n",
       "chr4_74189490_C    intron_variant;upstream_gene_variant;intron_va...  \n",
       "chr4_75672581_C    intron_variant,non_coding_transcript_variant;i...  \n",
       "chr4_82892864_A    downstream_gene_variant;intron_variant;upstrea...  \n",
       "chr4_84482595_C                                   intergenic_variant  \n",
       "chr4_88378194_A                                upstream_gene_variant  \n",
       "\n",
       "[181 rows x 3 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clone_som_vars_df_rmdups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e94d5942",
   "metadata": {},
   "outputs": [],
   "source": [
    "donors_dendro_df = pd.concat(donors_dendro).reset_index().rename({\"level_0\":\"donor\"},axis=1)\n",
    "donors_dendro_df\n",
    "donors_dendro_df = pd.merge(donors_dendro_df,clone_som_vars_df_rmdups, left_on=\"variant\", right_index=True)\n",
    "donors_dendro_df.sort_values([ \"var_pct_in_clone\", \"ncells_var_cl\"], ascending=False).to_csv(join(outdir,\"top_dendro_somvars.csv\"),\n",
    "                                                                                             index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6550a54b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "donors_clones_df = pd.concat(donors_clones).reset_index(0).rename({\"level_0\":\"donor\"},axis=1)\n",
    "donors_clones_df = pd.merge(donors_clones_df,clone_som_vars_df_rmdups, left_on=\"variant\", right_index=True)\n",
    "donors_clones_df.sort_values([\"var_pct_in_clone\", \"Fisher's exact pval\", \"diff\"], ascending=[False, True, False])\n",
    "\n",
    "donors_clones_df.to_csv(join(outdir,\"top_clone_somvars.csv\"),index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7e2fc465",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/somatic_variants/gatk_mutect/peaks_atac_peaks/som_dendro_0.6/chip_clones_variants_summary'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outdir"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
