{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "223d6f46",
   "metadata": {},
   "source": [
    "## Plot markers umap embeddings\n",
    "Does this over each donor and separate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "235ae72f",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# Input info\n",
    "allSamples_f <- \"/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/allSamples.rds\"\n",
    "#se_f = \"/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/annotation_clones/SE.rds\"\n",
    "sample_names = \"preB,postB\"\n",
    "outdir = \"/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome\"\n",
    "#\"/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/annotation_clones/umap_markers\"\n",
    "\n",
    "# Parameters\n",
    "assay=\"RNA\"\n",
    "\n",
    "# Parallel processing\n",
    "cores = 4\n",
    "\n",
    "# R utils script\n",
    "utils_script <- \"/data/Mito_Trace/R_scripts/seurat_utils.R\"\n",
    "\n",
    "gene_file = \"/data/Mito_Trace/data/processed/genomes/mtMasked/GRCh38_MT_blacklist/genes/genes.gtf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e663aab6",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source(utils_script)\n",
    "dir.create(outdir, showWarnings = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "34253229",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'preB'</li><li>'postB'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'preB'\n",
       "\\item 'postB'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'preB'\n",
       "2. 'postB'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"preB\"  \"postB\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sample_names <- unlist(strsplit(sample_names, \",\")[[1]])\n",
    "sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4f435a9b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "An object of class Seurat \n",
       "40000 features across 1351 samples within 1 assay \n",
       "Active assay: ATAC (40000 features, 0 variable features)\n",
       "\n",
       "[[2]]\n",
       "An object of class Seurat \n",
       "41874 features across 2890 samples within 1 assay \n",
       "Active assay: ATAC (41874 features, 0 variable features)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allSamples <- readRDS(allSamples_f)\n",
    "allSamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aad4f638",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'data.table'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'magrittr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    not\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'cowplot'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:patchwork':\n",
      "\n",
      "    align_plots\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'future'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    value\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    values\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>structure(function (..., envir = parent.frame()) \n",
       "{\n",
       "<span style=white-space:pre-wrap>    future &lt;- SequentialFuture(..., envir = envir)</span>\n",
       "<span style=white-space:pre-wrap>    if (!future$lazy) </span>\n",
       "<span style=white-space:pre-wrap>        future &lt;- run(future)</span>\n",
       "<span style=white-space:pre-wrap>    invisible(future)</span>\n",
       "}, class = c(\"FutureStrategy\", \"sequential\", \"uniprocess\", \"future\", \n",
       "\"function\"))</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "structure(function (..., envir = parent.frame()) \n",
       "\\{\n",
       "    future <- SequentialFuture(..., envir = envir)\n",
       "    if (!future\\$lazy) \n",
       "        future <- run(future)\n",
       "    invisible(future)\n",
       "\\}, class = c(\"FutureStrategy\", \"sequential\", \"uniprocess\", \"future\", \n",
       "\"function\"))\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "structure(function (..., envir = parent.frame()) \n",
       "{\n",
       "    future <- SequentialFuture(..., envir = envir)\n",
       "    if (!future$lazy) \n",
       "        future <- run(future)\n",
       "    invisible(future)\n",
       "}, class = c(\"FutureStrategy\", \"sequential\", \"uniprocess\", \"future\", \n",
       "\"function\"))\n",
       "```"
      ],
      "text/plain": [
       "sequential:\n",
       "- args: function (..., envir = parent.frame())\n",
       "- tweaked: FALSE\n",
       "- call: NULL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Strategy 'multiprocess' is deprecated in future (>= 1.20.0). Instead, explicitly specify either 'multisession' or 'multicore'. In the current R session, 'multiprocess' equals 'multicore'.\"\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ensembldb':\n",
      "\n",
      "    filter, select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(GenomicRanges)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(GenomeInfoDb)\n",
    "library(EnsDb.Hsapiens.v75)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "set.seed(1234)\n",
    "library(data.table)\n",
    "library(magrittr)\n",
    "library(cowplot)\n",
    "library(metap)\n",
    "library(\"RColorBrewer\")\n",
    "\n",
    "library(future)\n",
    "plan()\n",
    "\n",
    "plan(\"multiprocess\", workers = cores)\n",
    "#options(future.globals.maxSize = 50000 * 1024^2) # for 50 Gb RAM\n",
    "options(future.globals.maxSize = 8000 * 1024^2)\n",
    "\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a1494722",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# annotations <- makeGRangesFromGTF(\n",
    "#       gene_file,\n",
    "#       level = c(\"genes\", \"transcripts\"),\n",
    "#       ignoreTxVersion = TRUE,\n",
    "#       .checkAgainstTxDb = FALSE\n",
    "#     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "54f52a80",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Warning message in readGFFAsGRanges(resource, version = version, colnames = colnames, :\n",
      "\"cannot set the seqlengths or circularity flags on the GRanges object to\n",
      "  return because the sequence names in the GTF or GFF file are in\n",
      "  disagreement with the sequence names implied by the genome assembly\n",
      "  (hg38) specified via the 'genome' argument\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 2765969 ranges and 25 metadata columns:\n",
       "              seqnames        ranges strand |   source       type     score\n",
       "                 <Rle>     <IRanges>  <Rle> | <factor>   <factor> <numeric>\n",
       "        [1]       chr1   29554-31109      + |   HAVANA gene              NA\n",
       "        [2]       chr1   29554-31097      + |   HAVANA transcript        NA\n",
       "        [3]       chr1   29554-30039      + |   HAVANA exon              NA\n",
       "        [4]       chr1   30564-30667      + |   HAVANA exon              NA\n",
       "        [5]       chr1   30976-31097      + |   HAVANA exon              NA\n",
       "        ...        ...           ...    ... .      ...        ...       ...\n",
       "  [2765965] KI270734.1 138483-138667      - |  ENSEMBL CDS               NA\n",
       "  [2765966] KI270734.1 138480-138482      - |  ENSEMBL stop_codon        NA\n",
       "  [2765967] KI270734.1 161689-161852      - |  ENSEMBL UTR               NA\n",
       "  [2765968] KI270734.1 161587-161626      - |  ENSEMBL UTR               NA\n",
       "  [2765969] KI270734.1 138082-138482      - |  ENSEMBL UTR               NA\n",
       "                phase         gene_id gene_version      gene_type   gene_name\n",
       "            <integer>     <character>  <character>    <character> <character>\n",
       "        [1]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        [2]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        [3]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        [4]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        [5]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        ...       ...             ...          ...            ...         ...\n",
       "  [2765965]         2 ENSG00000277196            4 protein_coding  AC007325.2\n",
       "  [2765966]         0 ENSG00000277196            4 protein_coding  AC007325.2\n",
       "  [2765967]      <NA> ENSG00000277196            4 protein_coding  AC007325.2\n",
       "  [2765968]      <NA> ENSG00000277196            4 protein_coding  AC007325.2\n",
       "  [2765969]      <NA> ENSG00000277196            4 protein_coding  AC007325.2\n",
       "                  level     hgnc_id         tag          havana_gene\n",
       "            <character> <character> <character>          <character>\n",
       "        [1]           2  HGNC:52482  ncRNA_host OTTHUMG00000000959.2\n",
       "        [2]           2  HGNC:52482       basic OTTHUMG00000000959.2\n",
       "        [3]           2  HGNC:52482       basic OTTHUMG00000000959.2\n",
       "        [4]           2  HGNC:52482       basic OTTHUMG00000000959.2\n",
       "        [5]           2  HGNC:52482       basic OTTHUMG00000000959.2\n",
       "        ...         ...         ...         ...                  ...\n",
       "  [2765965]           3        <NA>       basic                 <NA>\n",
       "  [2765966]           3        <NA>       basic                 <NA>\n",
       "  [2765967]           3        <NA>       basic                 <NA>\n",
       "  [2765968]           3        <NA>       basic                 <NA>\n",
       "  [2765969]           3        <NA>       basic                 <NA>\n",
       "              transcript_id transcript_version transcript_type transcript_name\n",
       "                <character>        <character>     <character>     <character>\n",
       "        [1]            <NA>               <NA>            <NA>            <NA>\n",
       "        [2] ENST00000473358                  1          lncRNA MIR1302-2HG-202\n",
       "        [3] ENST00000473358                  1          lncRNA MIR1302-2HG-202\n",
       "        [4] ENST00000473358                  1          lncRNA MIR1302-2HG-202\n",
       "        [5] ENST00000473358                  1          lncRNA MIR1302-2HG-202\n",
       "        ...             ...                ...             ...             ...\n",
       "  [2765965] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "  [2765966] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "  [2765967] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "  [2765968] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "  [2765969] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "            transcript_support_level    havana_transcript exon_number\n",
       "                         <character>          <character> <character>\n",
       "        [1]                     <NA>                 <NA>        <NA>\n",
       "        [2]                        5 OTTHUMT00000002840.1        <NA>\n",
       "        [3]                        5 OTTHUMT00000002840.1           1\n",
       "        [4]                        5 OTTHUMT00000002840.1           2\n",
       "        [5]                        5 OTTHUMT00000002840.1           3\n",
       "        ...                      ...                  ...         ...\n",
       "  [2765965]                        1                 <NA>          15\n",
       "  [2765966]                        1                 <NA>          15\n",
       "  [2765967]                        1                 <NA>           1\n",
       "  [2765968]                        1                 <NA>           2\n",
       "  [2765969]                        1                 <NA>          15\n",
       "                    exon_id exon_version        protein_id      ccdsid\n",
       "                <character>  <character>       <character> <character>\n",
       "        [1]            <NA>         <NA>              <NA>        <NA>\n",
       "        [2]            <NA>         <NA>              <NA>        <NA>\n",
       "        [3] ENSE00001947070            1              <NA>        <NA>\n",
       "        [4] ENSE00001922571            1              <NA>        <NA>\n",
       "        [5] ENSE00001827679            1              <NA>        <NA>\n",
       "        ...             ...          ...               ...         ...\n",
       "  [2765965] ENSE00003753010            1 ENSP00000481127.1        <NA>\n",
       "  [2765966] ENSE00003753010            1 ENSP00000481127.1        <NA>\n",
       "  [2765967] ENSE00003746084            1 ENSP00000481127.1        <NA>\n",
       "  [2765968] ENSE00003719550            1 ENSP00000481127.1        <NA>\n",
       "  [2765969] ENSE00003753010            1 ENSP00000481127.1        <NA>\n",
       "                    ont   gene_biotype\n",
       "            <character>    <character>\n",
       "        [1]        <NA>         lncRNA\n",
       "        [2]        <NA>         lncRNA\n",
       "        [3]        <NA>         lncRNA\n",
       "        [4]        <NA>         lncRNA\n",
       "        [5]        <NA>         lncRNA\n",
       "        ...         ...            ...\n",
       "  [2765965]        <NA> protein_coding\n",
       "  [2765966]        <NA> protein_coding\n",
       "  [2765967]        <NA> protein_coding\n",
       "  [2765968]        <NA> protein_coding\n",
       "  [2765969]        <NA> protein_coding\n",
       "  -------\n",
       "  seqinfo: 40 sequences from hg38 genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# extract gene annotations from EnsDb\n",
    "#annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)\n",
    "gffRangedData<-rtracklayer::import.gff(gene_file, format=\"gtf\", genome=\"hg38\")\n",
    "annotations<-as(gffRangedData, \"GRanges\")\n",
    "values(annotations)$gene_biotype <- values(annotations)$gene_type\n",
    "annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6fb0f463",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ensdb.annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)\n",
    "# ensdb.annotations\n",
    "\n",
    "# seqlevelsStyle(ensdb.annotations) <- 'UCSC'\n",
    "# ensdb.annotations\n",
    "\n",
    "# sort(ensdb.annotations[(elementMetadata(ensdb.annotations)[, \"gene_name\"] %in% c(\"LYZ\"))])\n",
    "\n",
    "\n",
    "# sort(annotations[(elementMetadata(annotations)[, \"gene_name\"] %in% c(\"LYZ\"))])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d5f295e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers_f <- \"/data/Mito_Trace/data/processed/pbmc_markers.txt\"\n",
    "markers <- read.csv(markers_f, header=F)[[1]]\n",
    "\n",
    "#small.umap.outdir <- file.path(outdir, \"markersDimRed\")\n",
    "#dir.create(small.umap.outdir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "45959956",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    }
   ],
   "source": [
    "curr.se <- allSamples[[1]]\n",
    "# add the gene information to the object\n",
    "Annotation(curr.se) <- annotations\n",
    "gene.activities <- GeneActivity(curr.se)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "90b714de",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"outdir\"\n",
      "[1] \"/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n",
      "Warning message in dir.create(curr.outdir):\n",
      "\"'/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB' already exists\"\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "Warning message:\n",
      "\"The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session\"\n",
      "14:48:47 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "14:48:47 Read 1351 rows and found 49 numeric columns\n",
      "\n",
      "14:48:47 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "14:48:47 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "14:48:47 Writing NN index file to temp file /tmp/Rtmp5gbbMq/file1f506147c795ed\n",
      "\n",
      "14:48:47 Searching Annoy index using 4 threads, search_k = 3000\n",
      "\n",
      "14:48:47 Annoy recall = 100%\n",
      "\n",
      "14:48:49 Commencing smooth kNN distance calibration using 4 threads\n",
      "\n",
      "14:48:52 Initializing from normalized Laplacian + noise\n",
      "\n",
      "14:48:52 Commencing optimization for 500 epochs, with 52544 positive edges\n",
      "\n",
      "14:48:55 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n",
      "Warning message:\n",
      "\"UNRELIABLE VALUE: One of the 'future.apply' iterations ('future_lapply-1') unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Feature no in object: TCRA\"\n",
      "[1] \"Feature no in object: CD56\"\n",
      "[1] \"Feature no in object: CD161\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Feature no in object: IGHM\"\n",
      "[1] \"Feature no in object: TCRA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"outdir\"\n",
      "[1] \"/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "14:53:29 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "14:53:29 Read 2890 rows and found 49 numeric columns\n",
      "\n",
      "14:53:29 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "14:53:29 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "14:53:30 Writing NN index file to temp file /tmp/Rtmp5gbbMq/file1f50612938b607\n",
      "\n",
      "14:53:30 Searching Annoy index using 4 threads, search_k = 3000\n",
      "\n",
      "14:53:30 Annoy recall = 100%\n",
      "\n",
      "14:53:32 Commencing smooth kNN distance calibration using 4 threads\n",
      "\n",
      "14:53:35 Initializing from normalized Laplacian + noise\n",
      "\n",
      "14:53:36 Commencing optimization for 500 epochs, with 116380 positive edges\n",
      "\n",
      "14:53:41 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n",
      "Warning message:\n",
      "\"UNRELIABLE VALUE: One of the 'future.apply' iterations ('future_lapply-1') unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to \"ignore\".\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Feature no in object: TCRA\"\n",
      "[1] \"Feature no in object: CD56\"\n",
      "[1] \"Feature no in object: CD161\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Feature no in object: IGHM\"\n",
      "[1] \"Feature no in object: TCRA\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 24 × 2 of type chr</caption>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>TCRA</th><td>Feature no in object: TCRA                                                                                                                                      </td><td>Feature no in object: TCRA                                                                                                                                       </td></tr>\n",
       "\t<tr><th scope=row>CD56</th><td>Feature no in object: CD56                                                                                                                                      </td><td>Feature no in object: CD56                                                                                                                                       </td></tr>\n",
       "\t<tr><th scope=row>CD161</th><td>Feature no in object: CD161                                                                                                                                     </td><td>Feature no in object: CD161                                                                                                                                      </td></tr>\n",
       "\t<tr><th scope=row>NCR1</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/NCR1.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/NCR1.embedFeat.top.png   </td></tr>\n",
       "\t<tr><th scope=row>NCAM1</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/NCAM1.embedFeat.top.png  </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/NCAM1.embedFeat.top.png  </td></tr>\n",
       "\t<tr><th scope=row>CEACAM8</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CEACAM8.embedFeat.top.png</td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CEACAM8.embedFeat.top.png</td></tr>\n",
       "\t<tr><th scope=row>IGHM</th><td>Feature no in object: IGHM                                                                                                                                      </td><td>Feature no in object: IGHM                                                                                                                                       </td></tr>\n",
       "\t<tr><th scope=row>TCRA</th><td>Feature no in object: TCRA                                                                                                                                      </td><td>Feature no in object: TCRA                                                                                                                                       </td></tr>\n",
       "\t<tr><th scope=row>IL7R</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/IL7R.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/IL7R.embedFeat.top.png   </td></tr>\n",
       "\t<tr><th scope=row>CCR7</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CCR7.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CCR7.embedFeat.top.png   </td></tr>\n",
       "\t<tr><th scope=row>CD14</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CD14.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CD14.embedFeat.top.png   </td></tr>\n",
       "\t<tr><th scope=row>LYZ</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/LYZ.embedFeat.top.png    </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/LYZ.embedFeat.top.png    </td></tr>\n",
       "\t<tr><th scope=row>IL7R</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/IL7R.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/IL7R.embedFeat.top.png   </td></tr>\n",
       "\t<tr><th scope=row>S100A4</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/S100A4.embedFeat.top.png </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/S100A4.embedFeat.top.png </td></tr>\n",
       "\t<tr><th scope=row>MS4A1</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/MS4A1.embedFeat.top.png  </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/MS4A1.embedFeat.top.png  </td></tr>\n",
       "\t<tr><th scope=row>CD8A</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CD8A.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CD8A.embedFeat.top.png   </td></tr>\n",
       "\t<tr><th scope=row>FCGR3A</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/FCGR3A.embedFeat.top.png </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/FCGR3A.embedFeat.top.png </td></tr>\n",
       "\t<tr><th scope=row>MS4A7</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/MS4A7.embedFeat.top.png  </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/MS4A7.embedFeat.top.png  </td></tr>\n",
       "\t<tr><th scope=row>FCGR3A</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/FCGR3A.embedFeat.top.png </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/FCGR3A.embedFeat.top.png </td></tr>\n",
       "\t<tr><th scope=row>GNLY</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/GNLY.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/GNLY.embedFeat.top.png   </td></tr>\n",
       "\t<tr><th scope=row>NKG7</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/NKG7.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/NKG7.embedFeat.top.png   </td></tr>\n",
       "\t<tr><th scope=row>FCER1A</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/FCER1A.embedFeat.top.png </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/FCER1A.embedFeat.top.png </td></tr>\n",
       "\t<tr><th scope=row>CST3</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CST3.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CST3.embedFeat.top.png   </td></tr>\n",
       "\t<tr><th scope=row>PPBP</th><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/PPBP.embedFeat.top.png   </td><td>/data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/PPBP.embedFeat.top.png   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 24 × 2 of type chr\n",
       "\\begin{tabular}{r|ll}\n",
       "\tTCRA & Feature no in object: TCRA                                                                                                                                       & Feature no in object: TCRA                                                                                                                                       \\\\\n",
       "\tCD56 & Feature no in object: CD56                                                                                                                                       & Feature no in object: CD56                                                                                                                                       \\\\\n",
       "\tCD161 & Feature no in object: CD161                                                                                                                                      & Feature no in object: CD161                                                                                                                                      \\\\\n",
       "\tNCR1 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/NCR1.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/NCR1.embedFeat.top.png   \\\\\n",
       "\tNCAM1 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/NCAM1.embedFeat.top.png   & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/NCAM1.embedFeat.top.png  \\\\\n",
       "\tCEACAM8 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/CEACAM8.embedFeat.top.png & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/CEACAM8.embedFeat.top.png\\\\\n",
       "\tIGHM & Feature no in object: IGHM                                                                                                                                       & Feature no in object: IGHM                                                                                                                                       \\\\\n",
       "\tTCRA & Feature no in object: TCRA                                                                                                                                       & Feature no in object: TCRA                                                                                                                                       \\\\\n",
       "\tIL7R & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/IL7R.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/IL7R.embedFeat.top.png   \\\\\n",
       "\tCCR7 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/CCR7.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/CCR7.embedFeat.top.png   \\\\\n",
       "\tCD14 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/CD14.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/CD14.embedFeat.top.png   \\\\\n",
       "\tLYZ & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/LYZ.embedFeat.top.png     & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/LYZ.embedFeat.top.png    \\\\\n",
       "\tIL7R & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/IL7R.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/IL7R.embedFeat.top.png   \\\\\n",
       "\tS100A4 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/S100A4.embedFeat.top.png  & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/S100A4.embedFeat.top.png \\\\\n",
       "\tMS4A1 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/MS4A1.embedFeat.top.png   & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/MS4A1.embedFeat.top.png  \\\\\n",
       "\tCD8A & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/CD8A.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/CD8A.embedFeat.top.png   \\\\\n",
       "\tFCGR3A & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/FCGR3A.embedFeat.top.png  & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/FCGR3A.embedFeat.top.png \\\\\n",
       "\tMS4A7 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/MS4A7.embedFeat.top.png   & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/MS4A7.embedFeat.top.png  \\\\\n",
       "\tFCGR3A & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/FCGR3A.embedFeat.top.png  & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/FCGR3A.embedFeat.top.png \\\\\n",
       "\tGNLY & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/GNLY.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/GNLY.embedFeat.top.png   \\\\\n",
       "\tNKG7 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/NKG7.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/NKG7.embedFeat.top.png   \\\\\n",
       "\tFCER1A & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/FCER1A.embedFeat.top.png  & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/FCER1A.embedFeat.top.png \\\\\n",
       "\tCST3 & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/CST3.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/CST3.embedFeat.top.png   \\\\\n",
       "\tPPBP & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/preB/PPBP.embedFeat.top.png    & /data/Mito\\_Trace/output/pipeline/DUPI\\_april08\\_2021/MTblacklist\\_mtasnucl\\_Bonly/data/annotation/mergedSamples/sepConds\\_customgenome/postB/PPBP.embedFeat.top.png   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 24 × 2 of type chr\n",
       "\n",
       "| TCRA | Feature no in object: TCRA                                                                                                                                       | Feature no in object: TCRA                                                                                                                                        |\n",
       "| CD56 | Feature no in object: CD56                                                                                                                                       | Feature no in object: CD56                                                                                                                                        |\n",
       "| CD161 | Feature no in object: CD161                                                                                                                                      | Feature no in object: CD161                                                                                                                                       |\n",
       "| NCR1 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/NCR1.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/NCR1.embedFeat.top.png    |\n",
       "| NCAM1 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/NCAM1.embedFeat.top.png   | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/NCAM1.embedFeat.top.png   |\n",
       "| CEACAM8 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CEACAM8.embedFeat.top.png | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CEACAM8.embedFeat.top.png |\n",
       "| IGHM | Feature no in object: IGHM                                                                                                                                       | Feature no in object: IGHM                                                                                                                                        |\n",
       "| TCRA | Feature no in object: TCRA                                                                                                                                       | Feature no in object: TCRA                                                                                                                                        |\n",
       "| IL7R | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/IL7R.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/IL7R.embedFeat.top.png    |\n",
       "| CCR7 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CCR7.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CCR7.embedFeat.top.png    |\n",
       "| CD14 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CD14.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CD14.embedFeat.top.png    |\n",
       "| LYZ | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/LYZ.embedFeat.top.png     | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/LYZ.embedFeat.top.png     |\n",
       "| IL7R | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/IL7R.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/IL7R.embedFeat.top.png    |\n",
       "| S100A4 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/S100A4.embedFeat.top.png  | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/S100A4.embedFeat.top.png  |\n",
       "| MS4A1 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/MS4A1.embedFeat.top.png   | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/MS4A1.embedFeat.top.png   |\n",
       "| CD8A | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CD8A.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CD8A.embedFeat.top.png    |\n",
       "| FCGR3A | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/FCGR3A.embedFeat.top.png  | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/FCGR3A.embedFeat.top.png  |\n",
       "| MS4A7 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/MS4A7.embedFeat.top.png   | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/MS4A7.embedFeat.top.png   |\n",
       "| FCGR3A | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/FCGR3A.embedFeat.top.png  | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/FCGR3A.embedFeat.top.png  |\n",
       "| GNLY | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/GNLY.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/GNLY.embedFeat.top.png    |\n",
       "| NKG7 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/NKG7.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/NKG7.embedFeat.top.png    |\n",
       "| FCER1A | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/FCER1A.embedFeat.top.png  | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/FCER1A.embedFeat.top.png  |\n",
       "| CST3 | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CST3.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CST3.embedFeat.top.png    |\n",
       "| PPBP | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/PPBP.embedFeat.top.png    | /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/PPBP.embedFeat.top.png    |\n",
       "\n"
      ],
      "text/plain": [
       "        [,1]                                                                                                                                                            \n",
       "TCRA    Feature no in object: TCRA                                                                                                                                      \n",
       "CD56    Feature no in object: CD56                                                                                                                                      \n",
       "CD161   Feature no in object: CD161                                                                                                                                     \n",
       "NCR1    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/NCR1.embedFeat.top.png   \n",
       "NCAM1   /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/NCAM1.embedFeat.top.png  \n",
       "CEACAM8 /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CEACAM8.embedFeat.top.png\n",
       "IGHM    Feature no in object: IGHM                                                                                                                                      \n",
       "TCRA    Feature no in object: TCRA                                                                                                                                      \n",
       "IL7R    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/IL7R.embedFeat.top.png   \n",
       "CCR7    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CCR7.embedFeat.top.png   \n",
       "CD14    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CD14.embedFeat.top.png   \n",
       "LYZ     /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/LYZ.embedFeat.top.png    \n",
       "IL7R    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/IL7R.embedFeat.top.png   \n",
       "S100A4  /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/S100A4.embedFeat.top.png \n",
       "MS4A1   /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/MS4A1.embedFeat.top.png  \n",
       "CD8A    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CD8A.embedFeat.top.png   \n",
       "FCGR3A  /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/FCGR3A.embedFeat.top.png \n",
       "MS4A7   /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/MS4A7.embedFeat.top.png  \n",
       "FCGR3A  /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/FCGR3A.embedFeat.top.png \n",
       "GNLY    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/GNLY.embedFeat.top.png   \n",
       "NKG7    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/NKG7.embedFeat.top.png   \n",
       "FCER1A  /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/FCER1A.embedFeat.top.png \n",
       "CST3    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/CST3.embedFeat.top.png   \n",
       "PPBP    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/preB/PPBP.embedFeat.top.png   \n",
       "        [,2]                                                                                                                                                             \n",
       "TCRA    Feature no in object: TCRA                                                                                                                                       \n",
       "CD56    Feature no in object: CD56                                                                                                                                       \n",
       "CD161   Feature no in object: CD161                                                                                                                                      \n",
       "NCR1    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/NCR1.embedFeat.top.png   \n",
       "NCAM1   /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/NCAM1.embedFeat.top.png  \n",
       "CEACAM8 /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CEACAM8.embedFeat.top.png\n",
       "IGHM    Feature no in object: IGHM                                                                                                                                       \n",
       "TCRA    Feature no in object: TCRA                                                                                                                                       \n",
       "IL7R    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/IL7R.embedFeat.top.png   \n",
       "CCR7    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CCR7.embedFeat.top.png   \n",
       "CD14    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CD14.embedFeat.top.png   \n",
       "LYZ     /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/LYZ.embedFeat.top.png    \n",
       "IL7R    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/IL7R.embedFeat.top.png   \n",
       "S100A4  /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/S100A4.embedFeat.top.png \n",
       "MS4A1   /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/MS4A1.embedFeat.top.png  \n",
       "CD8A    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CD8A.embedFeat.top.png   \n",
       "FCGR3A  /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/FCGR3A.embedFeat.top.png \n",
       "MS4A7   /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/MS4A7.embedFeat.top.png  \n",
       "FCGR3A  /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/FCGR3A.embedFeat.top.png \n",
       "GNLY    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/GNLY.embedFeat.top.png   \n",
       "NKG7    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/NKG7.embedFeat.top.png   \n",
       "FCER1A  /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/FCER1A.embedFeat.top.png \n",
       "CST3    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/CST3.embedFeat.top.png   \n",
       "PPBP    /data/Mito_Trace/output/pipeline/DUPI_april08_2021/MTblacklist_mtasnucl_Bonly/data/annotation/mergedSamples/sepConds_customgenome/postB/PPBP.embedFeat.top.png   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "source(utils_script)\n",
    "#curr.sample <- curr.sample\n",
    "\n",
    "run.sample.sep <- function(curr.sample, outdir){\n",
    "    print(\"outdir\")\n",
    "    print(outdir)\n",
    "    DefaultAssay(curr.sample) <- \"ATAC\"\n",
    "\n",
    "    Annotation(curr.sample) <- annotations\n",
    "    gene.activities <- GeneActivity(curr.sample)\n",
    "\n",
    "    curr.outdir <- file.path(outdir, curr.sample$orig.ident[[1]])\n",
    "    dir.create(curr.outdir)\n",
    "    out <- embed.atac(curr.sample)\n",
    "    curr.sample <- out[[1]]\n",
    "    # add the gene activity matrix to the Seurat object as a new assay and normalize it\n",
    "    curr.sample[['RNA']] <- CreateAssayObject(counts = gene.activities)\n",
    "    curr.sample <- NormalizeData(\n",
    "      object = curr.sample,\n",
    "      assay = 'RNA',\n",
    "      normalization.method = 'LogNormalize',\n",
    "      scale.factor = median(curr.sample$nCount_RNA)\n",
    "    )\n",
    "\n",
    "    saveRDS(curr.sample, file.path(curr.outdir, paste0(\"se.rds\")))\n",
    "    DimPlot(curr.sample)\n",
    "    ggsave(file.path(curr.outdir, \"dimPlot.png\"))\n",
    "    \n",
    "    feature.names <- ClosestFeature(curr.sample, rownames(curr.sample))\n",
    "    marker.feature.names <- feature.names %>% filter(gene_name %in% markers)\n",
    "    rownames(marker.feature.names) <- marker.feature.names[,\"query_region\"]\n",
    "    # marker.feature.names\n",
    "    marker.feature.names$name <- paste0(marker.feature.names$gene_name, \"_\", marker.feature.names$type)\n",
    "    sapply(rownames(marker.feature.names), \n",
    "            FUN=featplot, se=curr.sample, curr.outdir=curr.outdir, feat.names=marker.feature.names[\"name\"])\n",
    "\n",
    "\n",
    "    DefaultAssay(curr.sample) <- \"RNA\"\n",
    "    sapply(markers, \n",
    "        FUN=featplot, se=curr.sample, curr.outdir=curr.outdir, feat.names=NULL)\n",
    "    \n",
    "}\n",
    "\n",
    "sapply(allSamples, FUN=run.sample.sep, outdir=outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32edb325",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7ae1698",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
