{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "41fdb08b",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "DE.out.path = \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust\"\n",
    "export.path = \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1/\"\n",
    "gsea_dir = \"/mnt/md0/isshamie/Projects/Mito_Trace/software/Bioinformatics_Tools/\"\n",
    "#pthresh = 0.1\n",
    "\n",
    "\n",
    "gsea_pthresh = 0.1 \n",
    "prefilter = FALSE \n",
    "stat_col = \"log2FoldChange\"\n",
    "padjmethod = \"BH\"\n",
    "\n",
    "pct_thresh = 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d59d5e98",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "export.path.prep <- file.path(export.path, \"prep\")\n",
    "dir.create(export.path.prep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ea4fc8d1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: clusterProfiler\n",
      "\n",
      "\n",
      "\n",
      "clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:clusterProfiler\"\n",
      "[1] \"package already installed:clusterProfiler\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GSEABase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: annotate\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Loading required package: XML\n",
      "\n",
      "Loading required package: graph\n",
      "\n",
      "\n",
      "Attaching package: ‘graph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XML’:\n",
      "\n",
      "    addNode\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:GSEABase\"\n",
      "[1] \"package already installed:GSEABase\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: biomaRt\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:biomaRt\"\n",
      "[1] \"package already installed:biomaRt\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: org.Hs.eg.db\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:org.Hs.eg.db\"\n",
      "[1] \"package already installed:org.Hs.eg.db\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: DOSE\n",
      "\n",
      "DOSE v3.20.1  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use DOSE in published research, please cite:\n",
      "Guangchuang Yu, Li-Gen Wang, Guang-Rong Yan, Qing-Yu He. DOSE: an R/Bioconductor package for Disease Ontology Semantic and Enrichment analysis. Bioinformatics 2015, 31(4):608-609\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:DOSE\"\n",
      "[1] \"package already installed:DOSE\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: gdata\n",
      "\n",
      "gdata: read.xls support for 'XLS' (Excel 97-2004) files ENABLED.\n",
      "\n",
      "\n",
      "\n",
      "gdata: Unable to load perl libaries needed by read.xls()\n",
      "gdata: to support 'XLSX' (Excel 2007+) files.\n",
      "\n",
      "\n",
      "\n",
      "gdata: Run the function 'installXLSXsupport()'\n",
      "gdata: to automatically download and install the perl\n",
      "gdata: libaries needed to support Excel XLS and XLSX formats.\n",
      "\n",
      "\n",
      "Attaching package: ‘gdata’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    startsWith, trim\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, first<-\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats4’:\n",
      "\n",
      "    nobs\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    nobs\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    object.size\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    startsWith\n",
      "\n",
      "\n",
      "Loading required package: data.table\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:gdata’:\n",
      "\n",
      "    first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "Loading required package: openxlsx\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:gdata’:\n",
      "\n",
      "    combine, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GSEABase’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graph’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘stringr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graph’:\n",
      "\n",
      "    boundary\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# library(\"AnnotationDbi\")\n",
    "#library(\"org.Hs.eg.db\")\n",
    "# library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "source(paste0(gsea_dir,'R/Gsea.R'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "895b2b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols.prior <- c(\"Gene\",\"baseMean\",\"log2FoldChange\",\"lfcSE\",\"stat\",\"pvalue\",\"padj\")\n",
    "cols.needed <- c(\"Gene\",\"log2FoldChange\")\n",
    "cols.curr <- c(\"X\",\"p_val\",\"avg_log2FC\",\"pct.1\",\"pct.2\",\"p_val_adj_BH\") #first col is empty in csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9270a190",
   "metadata": {},
   "source": [
    "## Prep files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "541f8ea4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster0.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster1.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster10.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster11.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster12.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster13.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster14.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster2.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster3.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster4.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster5.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster6.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster7.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster8.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/cluster9.DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#for f in list.files(DE.path, pattern=\"*.csv\"){\n",
    "\n",
    "DE_files <- list.files(path = DE.out.path,pattern=\"*.csv\", full.names=TRUE)\n",
    "for(f in DE_files){\n",
    "    print(f)\n",
    "    curr <- read.csv(f)\n",
    "    \n",
    "    if (!(nrow(curr)==0)){\n",
    "        colnames(curr) <- c(\"Gene\",\"pvalue\",\"log2FoldChange\",\"pct.1\",\"pct.2\",\"padj_bonf\", \"padj\")\n",
    "        curr$ensg <- mapIds(org.Hs.eg.db,\n",
    "                    keys=curr$Gene, \n",
    "                    column=\"ENSEMBL\",\n",
    "                    keytype=\"SYMBOL\",\n",
    "                    multiVals=\"first\")\n",
    "        write.table(curr, file.path(export.path.prep, paste0(strsplit(basename(f), \".csv\")[[1]], \".tsv\")),quote = FALSE,\n",
    "                    row.names=FALSE, sep='\\t')\n",
    "    }\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c641fe0f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# #rm(list=ls())\n",
    "\n",
    "# ########################################################################################\n",
    "# ##  gsea_ex1\n",
    "# #---------------------------------------------------------------------------------------\n",
    "# #' @description This example is designed to demonstrate how to run gsea analysis on the \n",
    "# #'              input files (DESeq2 results) put in the same directory\n",
    "# #' @details All the parameter definitions is in the below section of \"Setting Parameter\"\n",
    "# #' @return out a list that summarizes the GSEA enrichment analysis\n",
    "# #' \n",
    "# #' @export exportExcel_fp an excel file of the 'out' can be exported\n",
    "# #' @export exportPDF_fp an figure of the GSEA enrichment analysis\n",
    "# #' --------------------------------------------------------------------------------------------------\n",
    "# #' @author Austin W.T. Chiang\n",
    "# #' @note   v1.4 -- 2017.02.15 Add description of ex1, ex2, and idmapping file\n",
    "# #' @note   v1.3 -- 2017.02.08 Add function of displaying topN pathways and optimze code\n",
    "# #' @note   v1.2 -- 2017.02.06 Add zebrafish2 human id mapping\n",
    "# #' @note   v1.1 -- 2017.01.19 Midified for BEN\n",
    "# #' @note   v1.0 -- 2016.12.15 File created\n",
    "# #' --------------------------------------------------------------------------------------------------\n",
    "# ########################################################################################\n",
    "# gsea_ex1 <- function(DE_out_path = NULL, export_path = NULL,\n",
    "#                      IDmappingfile = NULL, convert = FALSE,\n",
    "#                      gene_name = TRUE, GSEA_set = c('c0', 'c2', 'c3'),\n",
    "#                      minGSSize = 30, pAdjustMethod = \"none\",\n",
    "#                      pvalueCutoff = 0.01, export = \"BOTH\",\n",
    "#                      ShowTop = TRUE, TopN = 5, gene_sets_dir=NULL){\n",
    "#   DE_out_files <- list.files(path = DE_out_path,pattern=\"*.tsv\", full.names = TRUE)\n",
    "#   for(f in DE_out_files){\n",
    "#     print(paste0(\"Here we go .... :\", f))\n",
    "#     #f <- \"res_geno_dobs_removed_C57BL6_J__M.tsv\"\n",
    "#     ## Input the DESeq result file\n",
    "#     myClass <- read.csv(f, row.names = 1,sep=\"\\t\")\n",
    "#     if(convert == TRUE){\n",
    "#       geneMapTable <- read.csv(IDmappingfile, sep = \"\\t\")\n",
    "#       myClass <- useIDmappingFile(myClass=myClass,\n",
    "#                                   geneMapTable=geneMapTable) #,\n",
    "#                                   #Original_attribute=Original_attribute,\n",
    "#                                  # Convert_attribute=Co nvert_attribute)\n",
    "#     }\n",
    "    \n",
    "#     #DEGTable <- subset(myClass, abs(log2FoldChange) >= .58 & (padj < 0.05) )\n",
    "#     #DEGTable <- subset(myClass, (padj < 0.1) )\n",
    "#     print(\"Myclass\")\n",
    "#     print(dim(myClass))\n",
    "#     DEGTable <- subset(myClass, (pct.1 > 0 & pct.2 > 0) )\n",
    "#     print(dim(DEGTable))\n",
    "\n",
    "#     DEGTable <- DEGTable[with(DEGTable, order(-log2FoldChange)), ]\n",
    "#     print(colnames(DEGTable))\n",
    "#     if(gene_name==TRUE){\n",
    "#       DEGTable$Gene <- DEGTable$gene_name\n",
    "#     }else{\n",
    "#       print('here')\n",
    "#       DEGTable$Gene <- row.names(DEGTable)\n",
    "#     }\n",
    "\n",
    "#     DEGTable <- DEGTable[,c(\"Gene\",\"log2FoldChange\")]\n",
    "    \n",
    "#     ## do the GSEA analysis\n",
    "#     f1 <- gsub(\".csv\", \"\", basename(f))\n",
    "#     exportExcel_fp = paste0(export_path,f1,\"_gsea.xlsx\")\n",
    "#     exportPDF_fp = paste0(export_path,f1,\"_gsea.pdf\")\n",
    "#     GSEA_res <-  AnnotDEG_byGSEA(DEGTable = DEGTable, GSEA_set = GSEA_set,\n",
    "#                                  export = export, ShowTop = ShowTop, TopN = TopN,\n",
    "#                                  exportExcel_fp = exportExcel_fp,\n",
    "#                                  exportPDF_fp = exportPDF_fp,\n",
    "#                                  minGSSize = minGSSize, \n",
    "#                                  pAdjustMethod = pAdjustMethod, \n",
    "#                                  pvalueCutoff = pvalueCutoff, gene_sets_dir = gene_sets_dir)\n",
    "    \n",
    "#   }\n",
    "# }\n",
    "\n",
    "\n",
    "# ########################################################################################\n",
    "# ##  Setting Parameter\n",
    "# ########################################################################################\n",
    "# #' @param DE_out_path the path of the DESeq2 output files \n",
    "# #' @param export_path the path that user want to export the GSEA results\n",
    "# #' @param convert 'TRUE': need to do id converion; 'FALSE': no id conversion needed  \n",
    "# #' @param IDmappingfile the idmapping file show that user's id to human gene name\n",
    "# #' @param gene_name 'TRUE': there is a columan \"gene_name\" in the input files (i.e. DESeq2 output files)\n",
    "# #' @param GSEA_set = c('c0', 'c2', 'c3')\n",
    "# #'                 co - hallmark gene sets;       c1 - positional gene sets; \n",
    "# #'                 c2 - curated gene sets;        c3 - motif gene sets;\n",
    "# #'                 c4 - computational gene sets;  c5 - GO gene sets;\n",
    "# #'                 c6 - oncogenic signatures;     c7 - immunologic signatures. \n",
    "# #'                 (http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C2)\n",
    "# #' @param minGSSize minimum size of genes for a specific GSEA gene set\n",
    "# #' @param pAdjustMethod one of \"holm\", \"hochberg\", \"hommel\", \"bonferroni\", \"BH\", \"BY\", \"fdr\", \"none\"\n",
    "# #' @param pvalueCutoff cutoff of pvalue in the GSEA enrichment analysis \n",
    "# #' @param export option for exporting GSEA result: \"EXCEL\", \"PDF\", or \"BOTH\"\n",
    "# #' @param ShowTop 'TRUE': the visualization will display only the top and bottom N gene sets; \n",
    "# #'                'FALSE': all of the significant gsea gene sets will be displayed \n",
    "# #'                        --> Note that: it is usually results in too many gene sets, which are not easy to see anything    \n",
    "# #' @param TopN this parameter indicate how many gene sets need to be displayed in the plots\n",
    "# ########################################################################################\n",
    "\n",
    "# # DE_out_path <- DE.out.path\n",
    "# # export_path <- \n",
    "# #IDmappingfile <- \"\" #\"mouse2human.txt\" #\"Data/zebrafish2human_idmapping.txt\"\n",
    "# convert = FALSE\n",
    "# gene_name = FALSE \n",
    "# GSEA_set = c('c0', 'c2', 'c5', 'c7')\n",
    "# minGSSize = 10 #20 #30\n",
    "# pAdjustMethod = \"BH\"\n",
    "# pvalueCutoff = gsea_pthresh\n",
    "# export = \"BOTH\"\n",
    "# ShowTop = TRUE \n",
    "# TopN = 10\n",
    "\n",
    "\n",
    "# ########################################################################################\n",
    "# ##  RUN gsea_ex1\n",
    "# ########################################################################################\n",
    "# gsea_ex1(DE_out_path = export.path.prep, export_path = export.path, gene_name = gene_name,\n",
    "#          convert = convert, GSEA_set = GSEA_set, minGSSize = minGSSize,\n",
    "#          pAdjustMethod = pAdjustMethod,  pvalueCutoff = pvalueCutoff,\n",
    "#          export = export, ShowTop = ShowTop,   \n",
    "#          TopN = TopN, gene_sets_dir=gsea_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5a2b671b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster0.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 2748    7\n",
      "[1] 2655    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "              pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "SHC3    1.300202e-11      0.3594911 0.694 0.290 2.475455e-07 1.190985e-08\n",
      "GPR88   1.007255e-10      0.1921534 0.408 0.112 1.917713e-06 6.919842e-08\n",
      "RBM47   6.770215e-10      0.6175955 0.878 0.593 1.288981e-05 3.100759e-07\n",
      "SFT2D2  8.738095e-10      0.2699131 0.592 0.221 1.663646e-05 3.430326e-07\n",
      "FAM107B 1.604936e-09      0.5684002 0.898 0.646 3.055637e-05 5.512954e-07\n",
      "STX18   4.715542e-09      0.2808526 0.796 0.354 8.977920e-05 1.295831e-06\n",
      "                   ensg\n",
      "SHC3    ENSG00000148082\n",
      "GPR88   ENSG00000181656\n",
      "RBM47   ENSG00000163694\n",
      "SFT2D2  ENSG00000213064\n",
      "FAM107B ENSG00000065809\n",
      "STX18   ENSG00000168818\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "RBM47     RBM47      0.6175955\n",
      "EDNRB     EDNRB      0.6110585\n",
      "FAM107B FAM107B      0.5684002\n",
      "FNDC3B   FNDC3B      0.5559874\n",
      "UBAC2     UBAC2      0.5320749\n",
      "LRRC8C   LRRC8C      0.5090904\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster1.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 1922    7\n",
      "[1] 1860    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "             pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "TREM1  1.692755e-09      0.1967330 0.461 0.199 3.222837e-05 1.445301e-06\n",
      "CDA    2.255932e-09      0.2286145 0.303 0.109 4.295069e-05 1.445301e-06\n",
      "SRPK1  1.218518e-08      0.2297985 0.449 0.209 2.319937e-04 5.805216e-06\n",
      "ZNF274 3.027105e-08      0.1400484 0.348 0.138 5.763305e-04 9.687150e-06\n",
      "FNDC3B 4.032113e-08      0.3919171 0.843 0.812 7.676739e-04 9.687150e-06\n",
      "FCER1G 1.101348e-07      0.1546290 0.281 0.106 2.096856e-03 2.351989e-05\n",
      "                  ensg\n",
      "TREM1  ENSG00000124731\n",
      "CDA    ENSG00000158825\n",
      "SRPK1  ENSG00000096063\n",
      "ZNF274 ENSG00000171606\n",
      "FNDC3B ENSG00000075420\n",
      "FCER1G ENSG00000158869\n",
      "[1] \"here\"\n",
      "         Gene log2FoldChange\n",
      "SPIDR   SPIDR      0.4351778\n",
      "TAOK3   TAOK3      0.4262597\n",
      "FNDC3B FNDC3B      0.3919171\n",
      "SYNPO2 SYNPO2      0.3803389\n",
      "RAB27A RAB27A      0.3639749\n",
      "MYO9B   MYO9B      0.3568084\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.05% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.05% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.05% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.05% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster10.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 2729    7\n",
      "[1] 2578    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "             pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "DUSP22 3.755870e-14      0.6074315 0.895 0.553 7.150801e-10 3.416590e-11\n",
      "P2RY6  1.180151e-13      0.4248014 0.579 0.217 2.246890e-09 6.441266e-11\n",
      "SAMHD1 3.746000e-12      0.4266320 0.614 0.259 7.132010e-08 1.703806e-09\n",
      "SULF2  8.657784e-12      0.4876427 0.807 0.417 1.648355e-07 2.953387e-09\n",
      "FGD2   4.239535e-11      0.3159547 0.544 0.213 8.071651e-07 1.285521e-08\n",
      "ZNF366 3.031251e-10      0.2811687 0.596 0.243 5.771199e-06 7.204697e-08\n",
      "                  ensg\n",
      "DUSP22 ENSG00000112679\n",
      "P2RY6  ENSG00000171631\n",
      "SAMHD1 ENSG00000101347\n",
      "SULF2  ENSG00000196562\n",
      "FGD2   ENSG00000146192\n",
      "ZNF366 ENSG00000178175\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "DUSP22   DUSP22      0.6074315\n",
      "UBAC2     UBAC2      0.5910751\n",
      "XYLT1     XYLT1      0.5183923\n",
      "SULF2     SULF2      0.4876427\n",
      "FAM53B   FAM53B      0.4748454\n",
      "RAPGEF4 RAPGEF4      0.4666756\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.04% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.04% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.04% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.04% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster11.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 1773    7\n",
      "[1] 1713    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "               pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "CXADR    5.893781e-93      0.4315471 0.621 0.274 1.122117e-88 2.612419e-90\n",
      "GAS7     2.739031e-92     -0.6763486 0.384 0.728 5.214841e-88 9.712603e-90\n",
      "KIAA1211 3.511034e-89      0.5242064 0.817 0.532 6.684657e-85 7.914606e-87\n",
      "ABCC4    3.571170e-89      0.5722032 0.915 0.692 6.799151e-85 7.914606e-87\n",
      "PHTF1    1.340226e-84      0.5338288 0.705 0.400 2.551657e-80 2.640246e-82\n",
      "NBAS     4.790408e-72      0.4916605 0.816 0.572 9.120457e-68 6.533379e-70\n",
      "                    ensg\n",
      "CXADR    ENSG00000154639\n",
      "GAS7     ENSG00000007237\n",
      "KIAA1211            <NA>\n",
      "ABCC4    ENSG00000125257\n",
      "PHTF1    ENSG00000116793\n",
      "NBAS     ENSG00000151779\n",
      "[1] \"here\"\n",
      "             Gene log2FoldChange\n",
      "ABCC4       ABCC4      0.5722032\n",
      "PHTF1       PHTF1      0.5338288\n",
      "KIAA1211 KIAA1211      0.5242064\n",
      "NBAS         NBAS      0.4916605\n",
      "DNAJC6     DNAJC6      0.4724268\n",
      "SGMS1       SGMS1      0.4722091\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster12.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 1804    7\n",
      "[1] 1758    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "               pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "PDZRN3   1.754121e-52      0.5426902 0.711 0.368 3.339670e-48 6.328867e-50\n",
      "FAM160A1 4.643906e-52      0.5162730 0.797 0.455 8.841532e-48 1.396268e-49\n",
      "GABRR3   1.867852e-51      0.2983251 0.447 0.141 3.556204e-47 4.813722e-49\n",
      "PALD1    4.164597e-51      0.5882127 0.810 0.507 7.928976e-47 9.391166e-49\n",
      "VASH2    9.521400e-48      0.3465417 0.550 0.219 1.812779e-43 1.908512e-45\n",
      "ADGRD1   1.242133e-46      0.4606054 0.682 0.344 2.364897e-42 2.240808e-44\n",
      "                    ensg\n",
      "PDZRN3   ENSG00000121440\n",
      "FAM160A1            <NA>\n",
      "GABRR3   ENSG00000183185\n",
      "PALD1    ENSG00000107719\n",
      "VASH2    ENSG00000143494\n",
      "ADGRD1   ENSG00000111452\n",
      "[1] \"here\"\n",
      "             Gene log2FoldChange\n",
      "PALD1       PALD1      0.5882127\n",
      "DUSP22     DUSP22      0.5842464\n",
      "PDZRN3     PDZRN3      0.5426902\n",
      "FAM160A1 FAM160A1      0.5162730\n",
      "ESR2         ESR2      0.5105183\n",
      "SH3RF3     SH3RF3      0.5003210\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster13.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 2349    7\n",
      "[1] 2243    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "               pvalue log2FoldChange pct.1 pct.2     padj_bonf          padj\n",
      "ADPRHL1 1.476829e-114      0.5112066 0.512 0.101 2.811735e-110 3.469071e-111\n",
      "CD81    3.418258e-105      0.6129103 0.573 0.139 6.508021e-101 2.676496e-102\n",
      "GSDME    1.228116e-97      0.6343366 0.698 0.225  2.338209e-93  7.212109e-95\n",
      "IQSEC1   6.828346e-89      1.0754666 0.963 0.839  1.300049e-84  2.673298e-86\n",
      "SH2D4B   1.807914e-85      0.7044860 0.725 0.282  3.442087e-81  6.066842e-83\n",
      "POU2AF1  5.857634e-69      0.7229606 0.797 0.416  1.115235e-64  1.528843e-66\n",
      "                   ensg\n",
      "ADPRHL1 ENSG00000153531\n",
      "CD81    ENSG00000110651\n",
      "GSDME   ENSG00000105928\n",
      "IQSEC1  ENSG00000144711\n",
      "SH2D4B  ENSG00000178217\n",
      "POU2AF1 ENSG00000110777\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "IQSEC1   IQSEC1      1.0754666\n",
      "POU2AF1 POU2AF1      0.7229606\n",
      "STK32B   STK32B      0.7164427\n",
      "EBF1       EBF1      0.7151201\n",
      "SH2D4B   SH2D4B      0.7044860\n",
      "HS6ST1   HS6ST1      0.6650850\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster14.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 4421    7\n",
      "[1] 4066    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "               pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "SH3RF3   2.015704e-11     -0.6023952 0.284 0.728 3.837700e-07 8.911429e-08\n",
      "PLCB4    2.850679e-10     -0.5207638 0.254 0.658 5.427408e-06 6.301427e-07\n",
      "TTC7A    6.256168e-10     -0.5413430 0.403 0.790 1.191112e-05 9.219507e-07\n",
      "TG       1.333314e-09     -0.5328141 0.448 0.829 2.538496e-05 1.473645e-06\n",
      "ARHGAP26 2.180425e-09     -0.5474291 0.642 0.921 4.151312e-05 1.698968e-06\n",
      "CAMK1D   2.305770e-09     -0.5142398 0.373 0.736 4.389955e-05 1.698968e-06\n",
      "                    ensg\n",
      "SH3RF3   ENSG00000172985\n",
      "PLCB4    ENSG00000101333\n",
      "TTC7A    ENSG00000068724\n",
      "TG       ENSG00000042832\n",
      "ARHGAP26 ENSG00000145819\n",
      "CAMK1D   ENSG00000183049\n",
      "[1] \"here\"\n",
      "             Gene log2FoldChange\n",
      "PNKP         PNKP      0.5200417\n",
      "AKT1S1     AKT1S1      0.5035193\n",
      "ARL6IP6   ARL6IP6      0.4979031\n",
      "TTC9C       TTC9C      0.4796567\n",
      "BRD2         BRD2      0.4667659\n",
      "HNRNPUL2 HNRNPUL2      0.4542660\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.02% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.02% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.02% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.02% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster2.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 536   7\n",
      "[1] 534   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "             pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "STXBP4 1.444796e-70      0.3234813 0.524 0.336 2.750747e-66 7.744106e-68\n",
      "GBP5   5.771940e-62      0.2345411 0.371 0.197 1.098920e-57 1.546880e-59\n",
      "CDH7   6.296195e-61      0.2826013 0.443 0.269 1.198733e-56 1.124920e-58\n",
      "PRKCD  3.637440e-57     -0.2061194 0.194 0.397 6.925321e-53 4.874169e-55\n",
      "FNDC3B 3.422179e-56     -0.3336139 0.752 0.840 6.515486e-52 3.668576e-54\n",
      "PAG1   9.640002e-54     -0.2507497 0.473 0.669 1.835360e-49 8.611735e-52\n",
      "                  ensg\n",
      "STXBP4 ENSG00000166263\n",
      "GBP5   ENSG00000154451\n",
      "CDH7   ENSG00000081138\n",
      "PRKCD  ENSG00000163932\n",
      "FNDC3B ENSG00000075420\n",
      "PAG1   ENSG00000076641\n",
      "[1] \"here\"\n",
      "         Gene log2FoldChange\n",
      "STXBP4 STXBP4      0.3234813\n",
      "IDS       IDS      0.2945213\n",
      "PREX2   PREX2      0.2847497\n",
      "CDH7     CDH7      0.2826013\n",
      "RBPMS   RBPMS      0.2622928\n",
      "LIMCH1 LIMCH1      0.2609779\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster3.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 515   7\n",
      "[1] 514   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "                pvalue log2FoldChange pct.1 pct.2     padj_bonf          padj\n",
      "FNDC3B   2.681228e-116      0.4199917 0.910 0.768 5.104790e-112 1.380832e-113\n",
      "HRH2     3.277582e-115      0.2490083 0.541 0.239 6.240189e-111 8.439774e-113\n",
      "ARHGAP26  2.370965e-82      0.3295712 0.953 0.902  4.514079e-78  3.052617e-80\n",
      "PLEK      4.039935e-80      0.2124438 0.578 0.316  7.691633e-76  4.161134e-78\n",
      "NEDD9     7.789418e-78      0.2771485 0.769 0.544  1.483027e-73  6.685917e-76\n",
      "TLR4      7.395997e-77      0.1381588 0.350 0.138  1.408124e-72  5.441341e-75\n",
      "                    ensg\n",
      "FNDC3B   ENSG00000075420\n",
      "HRH2     ENSG00000113749\n",
      "ARHGAP26 ENSG00000145819\n",
      "PLEK     ENSG00000115956\n",
      "NEDD9    ENSG00000111859\n",
      "TLR4     ENSG00000136869\n",
      "[1] \"here\"\n",
      "             Gene log2FoldChange\n",
      "FNDC3B     FNDC3B      0.4199917\n",
      "ARHGAP26 ARHGAP26      0.3295712\n",
      "SPIDR       SPIDR      0.2910866\n",
      "NEDD9       NEDD9      0.2771485\n",
      "CRYBG1     CRYBG1      0.2726461\n",
      "NIBAN1     NIBAN1      0.2636502\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster4.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 3723    7\n",
      "[1] 3300    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "             pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "PAPPA  2.175191e-19      0.6933287 0.726 0.285 4.141345e-15 2.699412e-16\n",
      "DNAJC6 1.873663e-14      0.8099137 0.758 0.438 3.567267e-10 1.003790e-11\n",
      "GAS7   1.990829e-14     -0.8079114 0.210 0.697 3.790340e-10 1.003790e-11\n",
      "HBS1L  2.297521e-14      0.7968706 0.839 0.569 4.374251e-10 1.003790e-11\n",
      "RSRC1  2.723428e-14      0.7346708 0.887 0.645 5.185135e-10 1.013932e-11\n",
      "NBAS   3.030861e-12      0.7378054 0.806 0.595 5.770457e-08 7.052435e-10\n",
      "                  ensg\n",
      "PAPPA  ENSG00000182752\n",
      "DNAJC6 ENSG00000116675\n",
      "GAS7   ENSG00000007237\n",
      "HBS1L  ENSG00000112339\n",
      "RSRC1  ENSG00000174891\n",
      "NBAS   ENSG00000151779\n",
      "[1] \"here\"\n",
      "         Gene log2FoldChange\n",
      "DNAJC6 DNAJC6      0.8099137\n",
      "HBS1L   HBS1L      0.7968706\n",
      "NBAS     NBAS      0.7378054\n",
      "RSRC1   RSRC1      0.7346708\n",
      "PAPPA   PAPPA      0.6933287\n",
      "KDM4C   KDM4C      0.6622285\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.03% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.03% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.03% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.03% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster5.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 1490    7\n",
      "[1] 1456    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "               pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "ABCC4    1.179951e-16      0.4494771 0.911 0.712 2.246509e-12 1.758127e-13\n",
      "ARHGAP26 1.057611e-13     -0.4951176 0.887 0.919 2.013586e-09 5.252803e-11\n",
      "STAU2    1.083234e-11      0.3605069 0.806 0.595 2.062369e-07 3.228036e-09\n",
      "TRPC6    1.506367e-11      0.3791280 0.677 0.448 2.867973e-07 3.740812e-09\n",
      "FBN1     8.073139e-11      0.4046036 0.589 0.362 1.537045e-06 1.503622e-08\n",
      "STAB2    1.061231e-10      0.2646638 0.484 0.249 2.020477e-06 1.756926e-08\n",
      "                    ensg\n",
      "ABCC4    ENSG00000125257\n",
      "ARHGAP26 ENSG00000145819\n",
      "STAU2    ENSG00000040341\n",
      "TRPC6    ENSG00000137672\n",
      "FBN1     ENSG00000166147\n",
      "STAB2    ENSG00000136011\n",
      "[1] \"here\"\n",
      "         Gene log2FoldChange\n",
      "ABCC4   ABCC4      0.4494771\n",
      "FBN1     FBN1      0.4046036\n",
      "TRPC6   TRPC6      0.3791280\n",
      "ZBTB16 ZBTB16      0.3628065\n",
      "STAU2   STAU2      0.3605069\n",
      "ACKR2   ACKR2      0.3179753\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster6.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 2059    7\n",
      "[1] 1945    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "             pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "IL1RL1 2.595089e-17      0.3893434 0.397 0.146 4.940789e-13 5.343287e-14\n",
      "NIBAN1 1.612850e-12      0.5372001 0.836 0.680 3.070704e-08 1.106952e-09\n",
      "FNDC3B 3.622886e-12      0.4667811 0.914 0.811 6.897613e-08 1.864881e-09\n",
      "MTCL1  8.060460e-11      0.4374754 0.638 0.437 1.534631e-06 3.319298e-08\n",
      "HDC    1.239793e-10      0.3880010 0.491 0.277 2.360442e-06 3.926606e-08\n",
      "IL21R  1.334932e-10     -0.3109422 0.121 0.443 2.541576e-06 3.926606e-08\n",
      "                  ensg\n",
      "IL1RL1 ENSG00000115602\n",
      "NIBAN1 ENSG00000135842\n",
      "FNDC3B ENSG00000075420\n",
      "MTCL1  ENSG00000168502\n",
      "HDC    ENSG00000140287\n",
      "IL21R  ENSG00000103522\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "NIBAN1   NIBAN1      0.5372001\n",
      "FAM107B FAM107B      0.4699492\n",
      "FNDC3B   FNDC3B      0.4667811\n",
      "MTCL1     MTCL1      0.4374754\n",
      "CACNA1D CACNA1D      0.4155360\n",
      "MRPS22   MRPS22      0.4047123\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster7.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 295   7\n",
      "[1] 295   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "             pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "KIFAP3 7.601911e-12     -0.1814480 0.294 0.499 1.447328e-07 1.639523e-09\n",
      "PTPN14 1.111541e-11      0.2697832 0.638 0.567 2.116262e-07 1.639523e-09\n",
      "PDE6A  9.907833e-11     -0.1033735 0.121 0.274 1.886352e-06 8.704012e-09\n",
      "CTDP1  1.428232e-10     -0.1279875 0.297 0.495 2.719211e-06 8.704012e-09\n",
      "MICU1  1.475256e-10     -0.1266285 0.217 0.392 2.808740e-06 8.704012e-09\n",
      "VAV2   4.180937e-10     -0.1268003 0.222 0.396 7.960086e-06 2.055627e-08\n",
      "                  ensg\n",
      "KIFAP3 ENSG00000075945\n",
      "PTPN14 ENSG00000152104\n",
      "PDE6A  ENSG00000132915\n",
      "CTDP1  ENSG00000060069\n",
      "MICU1  ENSG00000107745\n",
      "VAV2   ENSG00000160293\n",
      "[1] \"here\"\n",
      "         Gene log2FoldChange\n",
      "PTPN14 PTPN14      0.2697832\n",
      "TGFBR2 TGFBR2      0.2519881\n",
      "XYLT1   XYLT1      0.2387782\n",
      "GSAP     GSAP      0.2272137\n",
      "SPIDR   SPIDR      0.2270040\n",
      "KCNIP1 KCNIP1      0.2186043\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster8.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 1648    7\n",
      "[1] 1642    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "              pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "OR9Q1   1.082304e-08      0.1610702 0.327 0.132 0.0002060599 1.783637e-05\n",
      "XKR4    2.882887e-08      0.2648601 0.673 0.426 0.0005488728 2.166714e-05\n",
      "CACNA1E 3.944262e-08      0.3045096 0.663 0.429 0.0007509480 2.166714e-05\n",
      "DSCAML1 6.347643e-08      0.2996004 0.894 0.656 0.0012085278 2.615229e-05\n",
      "NOS1    8.083100e-08      0.2095611 0.433 0.214 0.0015389414 2.664190e-05\n",
      "PPP3CA  1.476629e-07     -0.3848197 0.721 0.776 0.0028113547 3.477235e-05\n",
      "                   ensg\n",
      "OR9Q1   ENSG00000186509\n",
      "XKR4    ENSG00000206579\n",
      "CACNA1E ENSG00000198216\n",
      "DSCAML1 ENSG00000177103\n",
      "NOS1    ENSG00000089250\n",
      "PPP3CA  ENSG00000138814\n",
      "[1] \"here\"\n",
      "             Gene log2FoldChange\n",
      "NCOR2       NCOR2      0.3289826\n",
      "FNDC3B     FNDC3B      0.3053310\n",
      "CACNA1E   CACNA1E      0.3045096\n",
      "ARHGAP26 ARHGAP26      0.3031086\n",
      "DSCAML1   DSCAML1      0.2996004\n",
      "SH3PXD2A SH3PXD2A      0.2779532\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some pathways, in reality P-values are less than 1e-10. You can set the `eps` argument to zero for better estimation.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1//prep/cluster9.DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 10172     7\n",
      "[1] 3524    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "              pvalue log2FoldChange pct.1 pct.2   padj_bonf        padj\n",
      "ANKRD44 3.511464e-07     -0.9269130 0.118 0.885 0.006685477 0.001190620\n",
      "CDH23   1.994185e-06     -0.9767119 0.176 0.847 0.037967292 0.001970732\n",
      "KDM4C   2.131150e-06     -0.7285555 0.118 0.808 0.040574962 0.001970732\n",
      "ZBTB16  6.818995e-06     -0.8635996 0.176 0.833 0.129826840 0.003009636\n",
      "OSBPL3  1.071772e-05     -0.7372069 0.118 0.755 0.204054608 0.003692687\n",
      "GAB2    1.122044e-05     -0.6473667 0.118 0.761 0.213625881 0.003692687\n",
      "                   ensg\n",
      "ANKRD44 ENSG00000065413\n",
      "CDH23   ENSG00000107736\n",
      "KDM4C   ENSG00000107077\n",
      "ZBTB16  ENSG00000109906\n",
      "OSBPL3  ENSG00000070882\n",
      "GAB2    ENSG00000033327\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "PLD5       PLD5       1.606308\n",
      "CDCP1     CDCP1       1.350161\n",
      "EXOC5     EXOC5       1.326394\n",
      "RASGRP3 RASGRP3       1.285011\n",
      "ARL6       ARL6       1.252783\n",
      "ALDH5A1 ALDH5A1       1.187421\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.03% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.03% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“There were 12 pathways for which P-values were not calculated properly due to unbalanced (positive and negative) gene-level statistic values. For such pathways pval, padj, NES, log2err are set to NA. You can try to increase the value of the argument nPermSimple (for example set it nPermSimple = 10000)”\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some of the pathways the P-values were likely overestimated. For such pathways log2err is set to NA.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.03% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.03% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "Warning message in fgseaMultilevel(...):\n",
      "“For some of the pathways the P-values were likely overestimated. For such pathways log2err is set to NA.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>.Primitive(\"return\")</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       ".Primitive(\"return\")\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       ".Primitive(\"return\")\n",
       "```"
      ],
      "text/plain": [
       ".Primitive(\"return\")"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#rm(list=ls())\n",
    "\n",
    "########################################################################################\n",
    "##  gsea_ex1\n",
    "#---------------------------------------------------------------------------------------\n",
    "#' @description This example is designed to demonstrate how to run gsea analysis on the \n",
    "#'              input files (DESeq2 results) put in the same directory\n",
    "#' @details All the parameter definitions is in the below section of \"Setting Parameter\"\n",
    "#' @return out a list that summarizes the GSEA enrichment analysis\n",
    "#' \n",
    "#' @export exportExcel_fp an excel file of the 'out' can be exported\n",
    "#' @export exportPDF_fp an figure of the GSEA enrichment analysis\n",
    "#' --------------------------------------------------------------------------------------------------\n",
    "#' @author Austin W.T. Chiang\n",
    "#' @note   v1.4 -- 2017.02.15 Add description of ex1, ex2, and idmapping file\n",
    "#' @note   v1.3 -- 2017.02.08 Add function of displaying topN pathways and optimze code\n",
    "#' @note   v1.2 -- 2017.02.06 Add zebrafish2 human id mapping\n",
    "#' @note   v1.1 -- 2017.01.19 Midified for BEN\n",
    "#' @note   v1.0 -- 2016.12.15 File created\n",
    "#' --------------------------------------------------------------------------------------------------\n",
    "########################################################################################\n",
    "gsea_ex1 <- function(DE_out_path = NULL, export_path = NULL,\n",
    "                     IDmappingfile = NULL, convert = FALSE,\n",
    "                     gene_name = TRUE, stat.col=\"log2FoldChange\", GSEA_set = c('c0', 'c2', 'c3'),\n",
    "                     minGSSize = 30, pAdjustMethod = \"none\",\n",
    "                     pvalueCutoff = 0.01, export = \"BOTH\",\n",
    "                     ShowTop = TRUE, TopN = 5, gene_sets_dir=NULL){\n",
    "  DE_out_files <- list.files(path = DE_out_path,pattern=\"*.tsv\", full.names = TRUE)\n",
    "  for(f in DE_out_files){\n",
    "    print(paste0(\"Here we go .... :\", f))\n",
    "    #f <- \"res_geno_dobs_removed_C57BL6_J__M.tsv\"\n",
    "    ## Input the DESeq result file\n",
    "    myClass <- read.csv(f, row.names = 1,sep=\"\\t\")\n",
    "    if(convert == TRUE){\n",
    "      geneMapTable <- read.csv(IDmappingfile, sep = \"\\t\")\n",
    "      myClass <- useIDmappingFile(myClass=myClass,\n",
    "                                  geneMapTable=geneMapTable) #,\n",
    "                                  #Original_attribute=Original_attribute,\n",
    "                                 # Convert_attribute=Co nvert_attribute)\n",
    "    }\n",
    "    \n",
    "    #DEGTable <- subset(myClass, abs(log2FoldChange) >= .58 & (padj < 0.05) )\n",
    "    if (prefilter){\n",
    "        DEGTable <- subset(myClass, (padj < pthresh) )\n",
    "        }\n",
    "      \n",
    "    print(\"Myclass\")\n",
    "    print(dim(myClass))\n",
    "    DEGTable <- subset(myClass, (pct.1 > pct_thresh & pct.2 > pct_thresh) )\n",
    "    print(dim(DEGTable))\n",
    "\n",
    "    print(colnames(DEGTable))\n",
    "    print(head(DEGTable))\n",
    "    if(gene_name==TRUE){\n",
    "      DEGTable$Gene <- DEGTable$gene_name\n",
    "    }else{\n",
    "      print('here')\n",
    "      DEGTable$Gene <- row.names(DEGTable)\n",
    "    }\n",
    "    \n",
    "    DEGTable <- DEGTable[with(DEGTable, order(-log2FoldChange)), ] #order\n",
    "    DEGTable <- DEGTable[,c(\"Gene\",stat.col)]\n",
    "    if (stat.col==\"padj\"){\n",
    "        DEGTable[,stat.col] <- -log10(DEGTable[,stat.col])\n",
    "        DEGTable <- DEGTable[with(DEGTable, order(-padj)), ] #order\n",
    "    }\n",
    "    \n",
    "    \n",
    "\n",
    "    #DEGTable <- DEGTable[order(stat.col, ), ]\n",
    "      \n",
    "    print(head(DEGTable))\n",
    "    ## do the GSEA analysis\n",
    "    f1 <- gsub(\".csv\", \"\", basename(f))\n",
    "    exportExcel_fp = paste0(export_path,f1,\"_gsea.xlsx\")\n",
    "    exportPDF_fp = paste0(export_path,f1,\"_gsea.pdf\")\n",
    "    exportDOT_fp = file.path(export_path,paste0(f1,\"_gsea.dot\"))\n",
    "    GSEA_res <-  AnnotDEG_byGSEA(DEGTable = DEGTable, GSEA_set = GSEA_set,\n",
    "                                 export = export, ShowTop = ShowTop, TopN = TopN,\n",
    "                                 exportExcel_fp = exportExcel_fp,\n",
    "                                 exportPDF_fp = exportPDF_fp,\n",
    "                                 minGSSize = minGSSize, exportDOT_fp=exportDOT_fp,\n",
    "                                 pAdjustMethod = pAdjustMethod, stat.col=stat.col,\n",
    "                                 pvalueCutoff = pvalueCutoff, gene_sets_dir = gene_sets_dir)\n",
    "    \n",
    "    #dotplot(GSEA_res)  \n",
    "    #ggsave(file.path(export_path, paste0(f1, \"_gsea.dot.pdf\")))\n",
    "    \n",
    "  }\n",
    "    return\n",
    "}\n",
    "\n",
    "\n",
    "########################################################################################\n",
    "##  Setting Parameter\n",
    "########################################################################################\n",
    "#' @param DE_out_path the path of the DESeq2 output files \n",
    "#' @param export_path the path that user want to export the GSEA results\n",
    "#' @param convert 'TRUE': need to do id converion; 'FALSE': no id conversion needed  \n",
    "#' @param IDmappingfile the idmapping file show that user's id to human gene name\n",
    "#' @param gene_name 'TRUE': there is a columan \"gene_name\" in the input files (i.e. DESeq2 output files)\n",
    "#' @param GSEA_set = c('c0', 'c2', 'c3')\n",
    "#'                 co - hallmark gene sets;       c1 - positional gene sets; \n",
    "#'                 c2 - curated gene sets;        c3 - motif gene sets;\n",
    "#'                 c4 - computational gene sets;  c5 - GO gene sets;\n",
    "#'                 c6 - oncogenic signatures;     c7 - immunologic signatures. \n",
    "#'                 (http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C2)\n",
    "#' @param minGSSize minimum size of genes for a specific GSEA gene set\n",
    "#' @param pAdjustMethod one of \"holm\", \"hochberg\", \"hommel\", \"bonferroni\", \"BH\", \"BY\", \"fdr\", \"none\"\n",
    "#' @param pvalueCutoff cutoff of pvalue in the GSEA enrichment analysis \n",
    "#' @param export option for exporting GSEA result: \"EXCEL\", \"PDF\", or \"BOTH\"\n",
    "#' @param ShowTop 'TRUE': the visualization will display only the top and bottom N gene sets; \n",
    "#'                'FALSE': all of the significant gsea gene sets will be displayed \n",
    "#'                        --> Note that: it is usually results in too many gene sets, which are not easy to see anything    \n",
    "#' @param TopN this parameter indicate how many gene sets need to be displayed in the plots\n",
    "########################################################################################\n",
    "\n",
    "# DE_out_path <- DE.out.path\n",
    "# export_path <- \n",
    "#IDmappingfile <- \"\" #\"mouse2human.txt\" #\"Data/zebrafish2human_idmapping.txt\"\n",
    "convert = FALSE\n",
    "gene_name = FALSE \n",
    "GSEA_set = c('c0', 'c2', 'c5', 'c7')\n",
    "minGSSize = 10 #20 #30\n",
    "pAdjustMethod = \"BH\"\n",
    "#pvalueCutoff = pthresh\n",
    "export = \"BOTH\"\n",
    "ShowTop = TRUE \n",
    "TopN = 10\n",
    "\n",
    "\n",
    "########################################################################################\n",
    "##  RUN gsea_ex1\n",
    "########################################################################################\n",
    "gsea_ex1(DE_out_path = export.path.prep, export_path = export.path, gene_name = gene_name,\n",
    "         convert = convert, GSEA_set = GSEA_set, minGSSize = minGSSize,\n",
    "         pAdjustMethod = padjmethod,  pvalueCutoff = gsea_pthresh,\n",
    "         export = export, ShowTop = ShowTop,  stat.col=stat_col ,\n",
    "         TopN = TopN, gene_sets_dir=gsea_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c8624952",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1/'"
      ],
      "text/latex": [
       "'/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/input\\_cond/DE\\_minPct\\_00\\_pval0.1/btwnClust/GSEA\\_pthresh\\_0.1/'"
      ],
      "text/markdown": [
       "'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1/'"
      ],
      "text/plain": [
       "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/input_cond/DE_minPct_00_pval0.1/btwnClust/GSEA_pthresh_0.1/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "export.path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8e3f179e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.3 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/isaac/miniconda3/envs/mttrace/lib/libopenblasp-r0.3.18.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] stringr_1.4.0         ggplot2_3.3.5         dplyr_1.0.7          \n",
       " [4] openxlsx_4.2.4        data.table_1.14.2     gdata_2.18.0         \n",
       " [7] DOSE_3.20.1           org.Hs.eg.db_3.14.0   biomaRt_2.50.0       \n",
       "[10] GSEABase_1.56.0       graph_1.72.0          annotate_1.72.0      \n",
       "[13] XML_3.99-0.8          AnnotationDbi_1.56.1  IRanges_2.28.0       \n",
       "[16] S4Vectors_0.32.0      Biobase_2.54.0        BiocGenerics_0.40.0  \n",
       "[19] clusterProfiler_4.2.2\n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] fgsea_1.20.0           colorspace_2.0-2       ggtree_3.2.1          \n",
       "  [4] ellipsis_0.3.2         IRdisplay_1.0          qvalue_2.26.0         \n",
       "  [7] XVector_0.34.0         base64enc_0.1-3        aplot_0.1.2           \n",
       " [10] farver_2.1.0           graphlayouts_0.8.0     ggrepel_0.9.1         \n",
       " [13] bit64_4.0.5            fansi_0.4.2            scatterpie_0.1.7      \n",
       " [16] xml2_1.3.2             splines_4.1.1          cachem_1.0.6          \n",
       " [19] GOSemSim_2.20.0        polyclip_1.10-0        IRkernel_1.2          \n",
       " [22] jsonlite_1.7.2         dbplyr_2.1.1           GO.db_3.14.0          \n",
       " [25] png_0.1-7              ggforce_0.3.3          BiocManager_1.30.16   \n",
       " [28] compiler_4.1.1         httr_1.4.2             assertthat_0.2.1      \n",
       " [31] Matrix_1.3-4           fastmap_1.1.0          lazyeval_0.2.2        \n",
       " [34] tweenr_1.0.2           prettyunits_1.1.1      htmltools_0.5.2       \n",
       " [37] tools_4.1.1            igraph_1.2.9           gtable_0.3.0          \n",
       " [40] glue_1.5.1             GenomeInfoDbData_1.2.7 reshape2_1.4.4        \n",
       " [43] DO.db_2.9              rappdirs_0.3.3         fastmatch_1.1-3       \n",
       " [46] Rcpp_1.0.7             enrichplot_1.14.1      vctrs_0.3.8           \n",
       " [49] Biostrings_2.62.0      ape_5.6-1              nlme_3.1-153          \n",
       " [52] ggraph_2.0.5           lifecycle_1.0.1        gtools_3.9.2          \n",
       " [55] zlibbioc_1.40.0        MASS_7.3-54            scales_1.1.1          \n",
       " [58] tidygraph_1.2.0        hms_1.1.1              parallel_4.1.1        \n",
       " [61] RColorBrewer_1.1-2     curl_4.3.2             memoise_2.0.1         \n",
       " [64] gridExtra_2.3          downloader_0.4         ggfun_0.0.5           \n",
       " [67] yulab.utils_0.0.4      stringi_1.7.6          RSQLite_2.2.8         \n",
       " [70] tidytree_0.3.7         filelock_1.0.2         zip_2.2.0             \n",
       " [73] BiocParallel_1.28.0    repr_1.1.3             GenomeInfoDb_1.30.0   \n",
       " [76] rlang_0.4.12           pkgconfig_2.0.3        bitops_1.0-7          \n",
       " [79] evaluate_0.14          lattice_0.20-45        purrr_0.3.4           \n",
       " [82] labeling_0.4.2         treeio_1.18.1          patchwork_1.1.1       \n",
       " [85] shadowtext_0.1.1       bit_4.0.4              tidyselect_1.1.1      \n",
       " [88] plyr_1.8.6             magrittr_2.0.1         R6_2.5.1              \n",
       " [91] generics_0.1.1         pbdZMQ_0.3-6           DBI_1.1.1             \n",
       " [94] withr_2.4.2            pillar_1.6.4           KEGGREST_1.34.0       \n",
       " [97] RCurl_1.98-1.5         tibble_3.1.6           crayon_1.4.2          \n",
       "[100] uuid_1.0-3             utf8_1.2.2             BiocFileCache_2.2.0   \n",
       "[103] progress_1.2.2         viridis_0.6.2          grid_4.1.1            \n",
       "[106] blob_1.2.2             digest_0.6.28          xtable_1.8-4          \n",
       "[109] tidyr_1.1.4            gridGraphics_0.5-1     munsell_0.5.0         \n",
       "[112] viridisLite_0.4.0      ggplotify_0.1.0       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686d6487",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d0c01837",
   "metadata": {},
   "source": [
    "### "
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
