{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d7254983",
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = \"/mnt/md0/isshamie/Projects/Mito_Trace/data/processed/annotation_granges/\"\n",
    "\n",
    "gene_file = \"ucsc\"\n",
    "gene_type=\"custom\" #\"UCSC\" #or custom\n",
    "\n",
    "gene_files = c(ucsc=\"UCSC\",\n",
    "               A2_black=\"/data/Mito_Trace/data/processed/genomes/mtMasked/GRCh38_MT_blacklist_A2_2020/genes/genes.gtf\",\n",
    "               hg38_1_2=\"/data/Mito_Trace/data/external/GRCh38_1.2.0/refdata-cellranger-atac-GRCh38-1.2.0/genes/genes.gtf\",\n",
    "               hg38_ucsc=\"hg38_UCSC\"\n",
    "              )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "61864810",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(outdir):\n",
      "“'/mnt/md0/isshamie/Projects/Mito_Trace/data/processed/annotation_granges' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "264c29ad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'magrittr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    not\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ensembldb':\n",
      "\n",
      "    filter, select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'future'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    value\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    values\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(GenomicRanges)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(GenomeInfoDb)\n",
    "library(EnsDb.Hsapiens.v75)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "\n",
    "set.seed(1234)\n",
    "library(magrittr)\n",
    "library(metap)\n",
    "library(dplyr)\n",
    "library(future)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e68dc0db",
   "metadata": {},
   "outputs": [],
   "source": [
    "get.annotations <- function(gene_file, gene_files){\n",
    "    print(gene_file)\n",
    "    if (gene_file==\"ucsc\"){\n",
    "        annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)\n",
    "        seqlevelsStyle(annotations) <- 'UCSC'\n",
    "        genome(annotations) <- \"hg38\"\n",
    "    }else if(gene_file==\"hg38_ucsc\"){\n",
    "        annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)\n",
    "        seqlevelsStyle(annotations) <- 'UCSC'\n",
    "        genome(annotations) <- \"hg38\"        \n",
    "    }else{\n",
    "        gffRangedData<-rtracklayer::import.gff(gene_files[[gene_file]], format=\"gtf\", genome=\"hg38\")\n",
    "        annotations<-as(gffRangedData, \"GRanges\")\n",
    "        if (! \"gene_biotype\" %in% colnames(values(annotations)) ){\n",
    "            values(annotations)$gene_biotype <- values(annotations)$gene_type\n",
    "        }\n",
    "    }\n",
    "    saveRDS(annotations, file.path(outdir, paste0(gene_file, \".annotationGranges.rds\")))\n",
    "    fileConn<-file(file.path(outdir, paste0(gene_file, \".annotationGranges.txt\")))\n",
    "    writeLines(gene_files[[gene_file]], fileConn)\n",
    "    close(fileConn)\n",
    "    return(annotations)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5f1b4ec6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ucsc\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"A2_black\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Warning message in readGFFAsGRanges(resource, version = version, colnames = colnames, :\n",
      "\"cannot set the seqlengths or circularity flags on the GRanges object to\n",
      "  return because the sequence names in the GTF or GFF file are in\n",
      "  disagreement with the sequence names implied by the genome assembly\n",
      "  (hg38) specified via the 'genome' argument\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"hg38_1_2\"\n",
      "[1] \"hg38_ucsc\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "$ucsc\n",
       "GRanges object with 3072120 ranges and 5 metadata columns:\n",
       "                  seqnames        ranges strand |           tx_id   gene_name\n",
       "                     <Rle>     <IRanges>  <Rle> |     <character> <character>\n",
       "  ENSE00001489430     chrX 192989-193061      + | ENST00000399012      PLCXD1\n",
       "  ENSE00001536003     chrX 192991-193061      + | ENST00000484611      PLCXD1\n",
       "  ENSE00002160563     chrX 193020-193061      + | ENST00000430923      PLCXD1\n",
       "  ENSE00001750899     chrX 197722-197788      + | ENST00000445062      PLCXD1\n",
       "  ENSE00001489388     chrX 197859-198351      + | ENST00000381657      PLCXD1\n",
       "              ...      ...           ...    ... .             ...         ...\n",
       "  ENST00000361739    chrMT     7586-8269      + | ENST00000361739      MT-CO2\n",
       "  ENST00000361789    chrMT   14747-15887      + | ENST00000361789      MT-CYB\n",
       "  ENST00000361851    chrMT     8366-8572      + | ENST00000361851     MT-ATP8\n",
       "  ENST00000361899    chrMT     8527-9207      + | ENST00000361899     MT-ATP6\n",
       "  ENST00000362079    chrMT     9207-9990      + | ENST00000362079      MT-CO3\n",
       "                          gene_id   gene_biotype     type\n",
       "                      <character>    <character> <factor>\n",
       "  ENSE00001489430 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001536003 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00002160563 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001750899 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001489388 ENSG00000182378 protein_coding     exon\n",
       "              ...             ...            ...      ...\n",
       "  ENST00000361739 ENSG00000198712 protein_coding      cds\n",
       "  ENST00000361789 ENSG00000198727 protein_coding      cds\n",
       "  ENST00000361851 ENSG00000228253 protein_coding      cds\n",
       "  ENST00000361899 ENSG00000198899 protein_coding      cds\n",
       "  ENST00000362079 ENSG00000198938 protein_coding      cds\n",
       "  -------\n",
       "  seqinfo: 25 sequences from hg38 genome\n",
       "\n",
       "$A2_black\n",
       "GRanges object with 2765969 ranges and 25 metadata columns:\n",
       "              seqnames        ranges strand |   source       type     score\n",
       "                 <Rle>     <IRanges>  <Rle> | <factor>   <factor> <numeric>\n",
       "        [1]       chr1   29554-31109      + |   HAVANA gene              NA\n",
       "        [2]       chr1   29554-31097      + |   HAVANA transcript        NA\n",
       "        [3]       chr1   29554-30039      + |   HAVANA exon              NA\n",
       "        [4]       chr1   30564-30667      + |   HAVANA exon              NA\n",
       "        [5]       chr1   30976-31097      + |   HAVANA exon              NA\n",
       "        ...        ...           ...    ... .      ...        ...       ...\n",
       "  [2765965] KI270734.1 138483-138667      - |  ENSEMBL CDS               NA\n",
       "  [2765966] KI270734.1 138480-138482      - |  ENSEMBL stop_codon        NA\n",
       "  [2765967] KI270734.1 161689-161852      - |  ENSEMBL UTR               NA\n",
       "  [2765968] KI270734.1 161587-161626      - |  ENSEMBL UTR               NA\n",
       "  [2765969] KI270734.1 138082-138482      - |  ENSEMBL UTR               NA\n",
       "                phase         gene_id gene_version      gene_type   gene_name\n",
       "            <integer>     <character>  <character>    <character> <character>\n",
       "        [1]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        [2]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        [3]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        [4]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        [5]      <NA> ENSG00000243485            5         lncRNA MIR1302-2HG\n",
       "        ...       ...             ...          ...            ...         ...\n",
       "  [2765965]         2 ENSG00000277196            4 protein_coding  AC007325.2\n",
       "  [2765966]         0 ENSG00000277196            4 protein_coding  AC007325.2\n",
       "  [2765967]      <NA> ENSG00000277196            4 protein_coding  AC007325.2\n",
       "  [2765968]      <NA> ENSG00000277196            4 protein_coding  AC007325.2\n",
       "  [2765969]      <NA> ENSG00000277196            4 protein_coding  AC007325.2\n",
       "                  level     hgnc_id         tag          havana_gene\n",
       "            <character> <character> <character>          <character>\n",
       "        [1]           2  HGNC:52482  ncRNA_host OTTHUMG00000000959.2\n",
       "        [2]           2  HGNC:52482       basic OTTHUMG00000000959.2\n",
       "        [3]           2  HGNC:52482       basic OTTHUMG00000000959.2\n",
       "        [4]           2  HGNC:52482       basic OTTHUMG00000000959.2\n",
       "        [5]           2  HGNC:52482       basic OTTHUMG00000000959.2\n",
       "        ...         ...         ...         ...                  ...\n",
       "  [2765965]           3        <NA>       basic                 <NA>\n",
       "  [2765966]           3        <NA>       basic                 <NA>\n",
       "  [2765967]           3        <NA>       basic                 <NA>\n",
       "  [2765968]           3        <NA>       basic                 <NA>\n",
       "  [2765969]           3        <NA>       basic                 <NA>\n",
       "              transcript_id transcript_version transcript_type transcript_name\n",
       "                <character>        <character>     <character>     <character>\n",
       "        [1]            <NA>               <NA>            <NA>            <NA>\n",
       "        [2] ENST00000473358                  1          lncRNA MIR1302-2HG-202\n",
       "        [3] ENST00000473358                  1          lncRNA MIR1302-2HG-202\n",
       "        [4] ENST00000473358                  1          lncRNA MIR1302-2HG-202\n",
       "        [5] ENST00000473358                  1          lncRNA MIR1302-2HG-202\n",
       "        ...             ...                ...             ...             ...\n",
       "  [2765965] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "  [2765966] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "  [2765967] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "  [2765968] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "  [2765969] ENST00000621424                  4  protein_coding  AC007325.2-201\n",
       "            transcript_support_level    havana_transcript exon_number\n",
       "                         <character>          <character> <character>\n",
       "        [1]                     <NA>                 <NA>        <NA>\n",
       "        [2]                        5 OTTHUMT00000002840.1        <NA>\n",
       "        [3]                        5 OTTHUMT00000002840.1           1\n",
       "        [4]                        5 OTTHUMT00000002840.1           2\n",
       "        [5]                        5 OTTHUMT00000002840.1           3\n",
       "        ...                      ...                  ...         ...\n",
       "  [2765965]                        1                 <NA>          15\n",
       "  [2765966]                        1                 <NA>          15\n",
       "  [2765967]                        1                 <NA>           1\n",
       "  [2765968]                        1                 <NA>           2\n",
       "  [2765969]                        1                 <NA>          15\n",
       "                    exon_id exon_version        protein_id      ccdsid\n",
       "                <character>  <character>       <character> <character>\n",
       "        [1]            <NA>         <NA>              <NA>        <NA>\n",
       "        [2]            <NA>         <NA>              <NA>        <NA>\n",
       "        [3] ENSE00001947070            1              <NA>        <NA>\n",
       "        [4] ENSE00001922571            1              <NA>        <NA>\n",
       "        [5] ENSE00001827679            1              <NA>        <NA>\n",
       "        ...             ...          ...               ...         ...\n",
       "  [2765965] ENSE00003753010            1 ENSP00000481127.1        <NA>\n",
       "  [2765966] ENSE00003753010            1 ENSP00000481127.1        <NA>\n",
       "  [2765967] ENSE00003746084            1 ENSP00000481127.1        <NA>\n",
       "  [2765968] ENSE00003719550            1 ENSP00000481127.1        <NA>\n",
       "  [2765969] ENSE00003753010            1 ENSP00000481127.1        <NA>\n",
       "                    ont   gene_biotype\n",
       "            <character>    <character>\n",
       "        [1]        <NA>         lncRNA\n",
       "        [2]        <NA>         lncRNA\n",
       "        [3]        <NA>         lncRNA\n",
       "        [4]        <NA>         lncRNA\n",
       "        [5]        <NA>         lncRNA\n",
       "        ...         ...            ...\n",
       "  [2765965]        <NA> protein_coding\n",
       "  [2765966]        <NA> protein_coding\n",
       "  [2765967]        <NA> protein_coding\n",
       "  [2765968]        <NA> protein_coding\n",
       "  [2765969]        <NA> protein_coding\n",
       "  -------\n",
       "  seqinfo: 40 sequences from hg38 genome; no seqlengths\n",
       "\n",
       "$hg38_1_2\n",
       "GRanges object with 1684537 ranges and 21 metadata columns:\n",
       "            seqnames      ranges strand |   source       type     score\n",
       "               <Rle>   <IRanges>  <Rle> | <factor>   <factor> <numeric>\n",
       "        [1]     chr1 11869-14409      + |   HAVANA gene              NA\n",
       "        [2]     chr1 11869-14409      + |   HAVANA transcript        NA\n",
       "        [3]     chr1 11869-12227      + |   HAVANA exon              NA\n",
       "        [4]     chr1 12613-12721      + |   HAVANA exon              NA\n",
       "        [5]     chr1 13221-14409      + |   HAVANA exon              NA\n",
       "        ...      ...         ...    ... .      ...        ...       ...\n",
       "  [1684533]     chrM 15888-15953      + |  ENSEMBL transcript        NA\n",
       "  [1684534]     chrM 15888-15953      + |  ENSEMBL exon              NA\n",
       "  [1684535]     chrM 15956-16023      - |  ENSEMBL gene              NA\n",
       "  [1684536]     chrM 15956-16023      - |  ENSEMBL transcript        NA\n",
       "  [1684537]     chrM 15956-16023      - |  ENSEMBL exon              NA\n",
       "                phase           gene_id              gene_type   gene_name\n",
       "            <integer>       <character>            <character> <character>\n",
       "        [1]      <NA> ENSG00000223972.5 transcribed_unproces..     DDX11L1\n",
       "        [2]      <NA> ENSG00000223972.5 transcribed_unproces..     DDX11L1\n",
       "        [3]      <NA> ENSG00000223972.5 transcribed_unproces..     DDX11L1\n",
       "        [4]      <NA> ENSG00000223972.5 transcribed_unproces..     DDX11L1\n",
       "        [5]      <NA> ENSG00000223972.5 transcribed_unproces..     DDX11L1\n",
       "        ...       ...               ...                    ...         ...\n",
       "  [1684533]      <NA> ENSG00000210195.2                Mt_tRNA       MT-TT\n",
       "  [1684534]      <NA> ENSG00000210195.2                Mt_tRNA       MT-TT\n",
       "  [1684535]      <NA> ENSG00000210196.2                Mt_tRNA       MT-TP\n",
       "  [1684536]      <NA> ENSG00000210196.2                Mt_tRNA       MT-TP\n",
       "  [1684537]      <NA> ENSG00000210196.2                Mt_tRNA       MT-TP\n",
       "                  level          havana_gene     transcript_id\n",
       "            <character>          <character>       <character>\n",
       "        [1]           2 OTTHUMG00000000961.2              <NA>\n",
       "        [2]           2 OTTHUMG00000000961.2 ENST00000456328.2\n",
       "        [3]           2 OTTHUMG00000000961.2 ENST00000456328.2\n",
       "        [4]           2 OTTHUMG00000000961.2 ENST00000456328.2\n",
       "        [5]           2 OTTHUMG00000000961.2 ENST00000456328.2\n",
       "        ...         ...                  ...               ...\n",
       "  [1684533]           3                 <NA> ENST00000387460.2\n",
       "  [1684534]           3                 <NA> ENST00000387460.2\n",
       "  [1684535]           3                 <NA>              <NA>\n",
       "  [1684536]           3                 <NA> ENST00000387461.2\n",
       "  [1684537]           3                 <NA> ENST00000387461.2\n",
       "                 transcript_type  transcript_name transcript_support_level\n",
       "                     <character>      <character>              <character>\n",
       "        [1]                 <NA>             <NA>                     <NA>\n",
       "        [2] processed_transcript RP11-34P13.1-002                        1\n",
       "        [3] processed_transcript RP11-34P13.1-002                        1\n",
       "        [4] processed_transcript RP11-34P13.1-002                        1\n",
       "        [5] processed_transcript RP11-34P13.1-002                        1\n",
       "        ...                  ...              ...                      ...\n",
       "  [1684533]              Mt_tRNA        MT-TT-201                       NA\n",
       "  [1684534]              Mt_tRNA        MT-TT-201                       NA\n",
       "  [1684535]                 <NA>             <NA>                     <NA>\n",
       "  [1684536]              Mt_tRNA        MT-TP-201                       NA\n",
       "  [1684537]              Mt_tRNA        MT-TP-201                       NA\n",
       "                    tag    havana_transcript exon_number           exon_id\n",
       "            <character>          <character> <character>       <character>\n",
       "        [1]        <NA>                 <NA>        <NA>              <NA>\n",
       "        [2]       basic OTTHUMT00000362751.1        <NA>              <NA>\n",
       "        [3]       basic OTTHUMT00000362751.1           1 ENSE00002234944.1\n",
       "        [4]       basic OTTHUMT00000362751.1           2 ENSE00003582793.1\n",
       "        [5]       basic OTTHUMT00000362751.1           3 ENSE00002312635.1\n",
       "        ...         ...                  ...         ...               ...\n",
       "  [1684533]       basic                 <NA>        <NA>              <NA>\n",
       "  [1684534]       basic                 <NA>           1 ENSE00001544475.2\n",
       "  [1684535]        <NA>                 <NA>        <NA>              <NA>\n",
       "  [1684536]       basic                 <NA>        <NA>              <NA>\n",
       "  [1684537]       basic                 <NA>           1 ENSE00001544473.2\n",
       "                    ont  protein_id      ccdsid           gene_biotype\n",
       "            <character> <character> <character>            <character>\n",
       "        [1]        <NA>        <NA>        <NA> transcribed_unproces..\n",
       "        [2]        <NA>        <NA>        <NA> transcribed_unproces..\n",
       "        [3]        <NA>        <NA>        <NA> transcribed_unproces..\n",
       "        [4]        <NA>        <NA>        <NA> transcribed_unproces..\n",
       "        [5]        <NA>        <NA>        <NA> transcribed_unproces..\n",
       "        ...         ...         ...         ...                    ...\n",
       "  [1684533]        <NA>        <NA>        <NA>                Mt_tRNA\n",
       "  [1684534]        <NA>        <NA>        <NA>                Mt_tRNA\n",
       "  [1684535]        <NA>        <NA>        <NA>                Mt_tRNA\n",
       "  [1684536]        <NA>        <NA>        <NA>                Mt_tRNA\n",
       "  [1684537]        <NA>        <NA>        <NA>                Mt_tRNA\n",
       "  -------\n",
       "  seqinfo: 640 sequences (1 circular) from hg38 genome\n",
       "\n",
       "$hg38_ucsc\n",
       "GRanges object with 3021151 ranges and 5 metadata columns:\n",
       "                  seqnames        ranges strand |           tx_id   gene_name\n",
       "                     <Rle>     <IRanges>  <Rle> |     <character> <character>\n",
       "  ENSE00001489430     chrX 276322-276394      + | ENST00000399012      PLCXD1\n",
       "  ENSE00001536003     chrX 276324-276394      + | ENST00000484611      PLCXD1\n",
       "  ENSE00002160563     chrX 276353-276394      + | ENST00000430923      PLCXD1\n",
       "  ENSE00001750899     chrX 281055-281121      + | ENST00000445062      PLCXD1\n",
       "  ENSE00001489388     chrX 281192-281684      + | ENST00000381657      PLCXD1\n",
       "              ...      ...           ...    ... .             ...         ...\n",
       "  ENST00000361739     chrM     7586-8269      + | ENST00000361739      MT-CO2\n",
       "  ENST00000361789     chrM   14747-15887      + | ENST00000361789      MT-CYB\n",
       "  ENST00000361851     chrM     8366-8572      + | ENST00000361851     MT-ATP8\n",
       "  ENST00000361899     chrM     8527-9207      + | ENST00000361899     MT-ATP6\n",
       "  ENST00000362079     chrM     9207-9990      + | ENST00000362079      MT-CO3\n",
       "                          gene_id   gene_biotype     type\n",
       "                      <character>    <character> <factor>\n",
       "  ENSE00001489430 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001536003 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00002160563 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001750899 ENSG00000182378 protein_coding     exon\n",
       "  ENSE00001489388 ENSG00000182378 protein_coding     exon\n",
       "              ...             ...            ...      ...\n",
       "  ENST00000361739 ENSG00000198712 protein_coding      cds\n",
       "  ENST00000361789 ENSG00000198727 protein_coding      cds\n",
       "  ENST00000361851 ENSG00000228253 protein_coding      cds\n",
       "  ENST00000361899 ENSG00000198899 protein_coding      cds\n",
       "  ENST00000362079 ENSG00000198938 protein_coding      cds\n",
       "  -------\n",
       "  seqinfo: 25 sequences from hg38 genome\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sapply(names(gene_files), get.annotations, gene_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ce072db",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcedc6bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
