{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "223d6f46",
   "metadata": {},
   "source": [
    "## Creates binarized Seurat objects, integrates conditions and annotates genes by nearby peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "663e2d4d",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# Input info\n",
    "cellr_in = \"/home/isshamie/data/Projects/Mito_Trace/output/mtscATAC/data/jan21_2021/MTblacklist/\" \n",
    "outdir = \"/home/isshamie/data/Projects/Mito_Trace/output/annotation/data/cd34norm/MTblacklist/mergedSamples/\"\n",
    "\n",
    "sample_names = \"Control,Flt3l\"\n",
    "samples = \"P2,J2\"\n",
    "\n",
    "# Parameters\n",
    "nTop = 25000\n",
    "cores = 24"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b62b75ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(repr)\n",
    "options(repr.plot.width=12, repr.plot.height=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c2c99fc4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'P2'</li><li>'J2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'P2'\n",
       "\\item 'J2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'P2'\n",
       "2. 'J2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"P2\" \"J2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "samples <- unlist(strsplit(samples, \",\")[[1]])\n",
    "sample_names <- unlist(strsplit(sample_names, \",\"))\n",
    "\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41f9cf84",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a14a2f0e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Installing package(s) 'BiocVersion', 'GenomicRanges', 'Signac',\n",
      "  'EnsDb.Hsapiens.v75'\n",
      "\n",
      "also installing the dependencies ‘filelock’, ‘MatrixGenerics’, ‘DelayedArray’, ‘lambda.r’, ‘futile.options’, ‘gtools’, ‘caTools’, ‘sitmo’, ‘BiocFileCache’, ‘SummarizedExperiment’, ‘rjson’, ‘futile.logger’, ‘snow’, ‘BH’, ‘crosstalk’, ‘here’, ‘gplots’, ‘gridExtra’, ‘RcppArmadillo’, ‘spatstat.data’, ‘spatstat.utils’, ‘spatstat.sparse’, ‘abind’, ‘tensor’, ‘goftest’, ‘deldir’, ‘FNN’, ‘RSpectra’, ‘dqrng’, ‘BiocIO’, ‘biomaRt’, ‘memoise’, ‘plogr’, ‘KEGGREST’, ‘XML’, ‘GenomicAlignments’, ‘restfulr’, ‘RCurl’, ‘GenomeInfoDbData’, ‘zlibbioc’, ‘Biostrings’, ‘bitops’, ‘BiocParallel’, ‘Rhtslib’, ‘fitdistrplus’, ‘ggridges’, ‘ica’, ‘igraph’, ‘leiden’, ‘lmtest’, ‘matrixStats’, ‘miniUI’, ‘plotly’, ‘png’, ‘RANN’, ‘RcppAnnoy’, ‘reticulate’, ‘ROCR’, ‘Rtsne’, ‘scattermore’, ‘sctransform’, ‘spatstat.core’, ‘spatstat.geom’, ‘uwot’, ‘RcppEigen’, ‘RcppProgress’, ‘SnowballC’, ‘tweenr’, ‘polyclip’, ‘slam’, ‘sparsesvd’, ‘docopt’, ‘GenomicFeatures’, ‘AnnotationFilter’, ‘RSQLite’, ‘Biobase’, ‘AnnotationDbi’, ‘rtracklayer’, ‘ProtGenerics’, ‘BiocGenerics’, ‘S4Vectors’, ‘IRanges’, ‘GenomeInfoDb’, ‘XVector’, ‘Rsamtools’, ‘Seurat’, ‘SeuratObject’, ‘ggseqlogo’, ‘irlba’, ‘pbapply’, ‘patchwork’, ‘ggrepel’, ‘fastmatch’, ‘lsa’, ‘RcppRoll’, ‘ggforce’, ‘qlcMatrix’, ‘ensembldb’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'fansi', 'glmnet', 'knitr', 'later'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# install.packages(\"cowplot\")\n",
    "# install.packages(\"data.table\")\n",
    "# install.packages(\"dplyr\")\n",
    "# install.packages(\"BiocManager\")\n",
    "# BiocManager::install(c(\"GenomicRanges\", \"Signac\", \"EnsDb.Hsapiens.v75\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a93a0339",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Installing package(s) 'multtest'\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'fansi', 'glmnet', 'knitr', 'later'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#BiocManager::install(\"multtest\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f8400bf3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Installing package(s) 'metap'\n",
      "\n",
      "also installing the dependency 'mutoss'\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'fansi', 'glmnet', 'knitr', 'later'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#BiocManager::install(\"metap\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aad4f638",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'data.table'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'magrittr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    not\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'cowplot'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:patchwork':\n",
      "\n",
      "    align_plots\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ensembldb':\n",
      "\n",
      "    filter, select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'future'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    value\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"Strategy 'multiprocess' is deprecated in future (>= 1.20.0). Instead, explicitly specify either 'multisession' or 'multicore'. In the current R session, 'multiprocess' equals 'multicore'.\"\n"
     ]
    }
   ],
   "source": [
    "library(GenomicRanges)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(GenomeInfoDb)\n",
    "library(EnsDb.Hsapiens.v75)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "set.seed(1234)\n",
    "library(data.table)\n",
    "library(magrittr)\n",
    "library(cowplot)\n",
    "library(metap)\n",
    "library(dplyr)\n",
    "library(future)\n",
    "\n",
    "plan(\"multiprocess\", workers = cores)\n",
    "options(future.globals.maxSize = 8000 * 1024^2)\n",
    "#options(future.globals.maxSize = 50000 * 1024^2) # for 50 Gb RAM\n",
    "#plan(\"multiprocess\", workers = workers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a53a3256",
   "metadata": {},
   "source": [
    "## Merge all peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "62849b8a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"here\"\n",
      "[1] \"/home/isshamie/data/Projects/Mito_Trace/output/mtscATAC/data/jan21_2021/MTblacklist//P2/outs/filtered_peak_bc_matrix/peaks.bed\"\n",
      "[1] \"here\"\n",
      "[1] \"/home/isshamie/data/Projects/Mito_Trace/output/mtscATAC/data/jan21_2021/MTblacklist//J2/outs/filtered_peak_bc_matrix/peaks.bed\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 0 metadata columns:\n",
       "      seqnames        ranges strand\n",
       "         <Rle>     <IRanges>  <Rle>\n",
       "  [1]     chr1    9942-10364      *\n",
       "  [2]     chr1 191685-191736      *\n",
       "  [3]     chr1 267780-268257      *\n",
       "  [4]     chr1 271072-271548      *\n",
       "  [5]     chr1 585995-586411      *\n",
       "  [6]     chr1 605386-605767      *\n",
       "  -------\n",
       "  seqinfo: 24 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "read.peaks <- function(exp, cellr_in){\n",
    "    print('here')\n",
    "    print(file.path(cellr_in, exp, \"outs\", \"filtered_peak_bc_matrix\", \"peaks.bed\"))\n",
    "    peaks <- read.table(\n",
    "      file = file.path(cellr_in, exp, \"outs\", \"filtered_peak_bc_matrix\", \"peaks.bed\"),\n",
    "      col.names = c(\"chr\", \"start\", \"end\")\n",
    "    )\n",
    "    # convert to genomic ranges\n",
    "    gr <- makeGRangesFromDataFrame(peaks)\n",
    "    return(gr)\n",
    "}\n",
    "\n",
    "\n",
    "gr.full <- c(sapply(samples, read.peaks, cellr_in=cellr_in, USE.NAMES=F))\n",
    "\n",
    "gr.full.c <- gr.full[[1]]\n",
    "if (length(gr.full)>1){\n",
    "    for (i in 2:length(gr.full)){\n",
    "      gr.full.c <- c(gr.full.c, gr.full[[i]])\n",
    "    }\n",
    "}\n",
    "combined.peaks <- reduce(x = c(gr.full.c))\n",
    "\n",
    "# Filter out bad peaks based on length\n",
    "peakwidths <- width(combined.peaks)\n",
    "combined.peaks <- combined.peaks[peakwidths  < 10000 & peakwidths > 20]\n",
    "head(combined.peaks)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cae010c",
   "metadata": {},
   "source": [
    "## Create fragment objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "67e92e33",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"samples_df\"\n",
      "     sample_names samples\n",
      "[1,] \"Control\"    \"P2\"   \n",
      "[2,] \"Flt3l\"      \"J2\"   \n",
      "[1] \"exp\"\n",
      "[1] \"P2\"\n",
      "[1] \"name\"\n",
      "[1] \"Control\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'cli':\n",
      "  method     from         \n",
      "  print.boxx spatstat.geom\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m6875\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m1\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): X1\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m332219\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m18\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m  (2): barcode, cell_id\n",
      "\u001b[32mdbl\u001b[39m (16): total, duplicate, chimeric, unmapped, lowmapq, mitochondrial, pass...\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating fragments object\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Quantifying peaks\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating chromatin assay\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CreateSeuratObject.Assay(curr_assay, assay = \"ATAC\", project = name, :\n",
      "\"Some cells in meta.data not present in provided counts matrix.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"curr_assay\"\n",
      "data frame with 0 columns and 10 rows\n",
      "[1] \"curr\"\n",
      "                   orig.ident nCount_ATAC nFeature_ATAC total duplicate\n",
      "AAACGAAAGAGGTCCA-1    Control        1400          1357 16929      2587\n",
      "AAACGAAAGCGATACG-1    Control        3692          3505 45359     10546\n",
      "AAACGAAAGTCGTGAG-1    Control        1031           994 10177      2450\n",
      "AAACGAACAATAGTGA-1    Control        2829          2687 22452      4364\n",
      "AAACGAACACAATAAG-1    Control        1206          1184 11763      2003\n",
      "AAACGAACACTGATAC-1    Control         968           954  9449      1971\n",
      "                   chimeric unmapped lowmapq mitochondrial passed_filters\n",
      "AAACGAAAGAGGTCCA-1       58       70     674          9629           3911\n",
      "AAACGAAAGCGATACG-1      184      184    2302         22247           9896\n",
      "AAACGAAAGTCGTGAG-1       48       60     657          2135           4827\n",
      "AAACGAACAATAGTGA-1      100       83     897          9233           7775\n",
      "AAACGAACACAATAAG-1       43       57     680          5547           3433\n",
      "AAACGAACACTGATAC-1       59       21     489          3785           3124\n",
      "                   cell_id is__cell_barcode TSS_fragments\n",
      "AAACGAAAGAGGTCCA-1 _cell_0                1           740\n",
      "AAACGAAAGCGATACG-1 _cell_1                1          1611\n",
      "AAACGAAAGTCGTGAG-1 _cell_2                1           775\n",
      "AAACGAACAATAGTGA-1 _cell_3                1          1239\n",
      "AAACGAACACAATAAG-1 _cell_4                1           621\n",
      "AAACGAACACTGATAC-1 _cell_5                1           517\n",
      "                   DNase_sensitive_region_fragments enhancer_region_fragments\n",
      "AAACGAAAGAGGTCCA-1                                0                         0\n",
      "AAACGAAAGCGATACG-1                                0                         0\n",
      "AAACGAAAGTCGTGAG-1                                0                         0\n",
      "AAACGAACAATAGTGA-1                                0                         0\n",
      "AAACGAACACAATAAG-1                                0                         0\n",
      "AAACGAACACTGATAC-1                                0                         0\n",
      "                   promoter_region_fragments on_target_fragments\n",
      "AAACGAAAGAGGTCCA-1                         0                 740\n",
      "AAACGAAAGCGATACG-1                         0                1611\n",
      "AAACGAAAGTCGTGAG-1                         0                 775\n",
      "AAACGAACAATAGTGA-1                         0                1239\n",
      "AAACGAACACAATAAG-1                         0                 621\n",
      "AAACGAACACTGATAC-1                         0                 517\n",
      "                   blacklist_region_fragments peak_region_fragments\n",
      "AAACGAAAGAGGTCCA-1                          0                  1330\n",
      "AAACGAAAGCGATACG-1                          0                  3462\n",
      "AAACGAAAGTCGTGAG-1                          0                   966\n",
      "AAACGAACAATAGTGA-1                          0                  2701\n",
      "AAACGAACACAATAAG-1                          0                  1138\n",
      "AAACGAACACTGATAC-1                          0                   878\n",
      "                   peak_region_cutsites    proj\n",
      "AAACGAAAGAGGTCCA-1                 2434 Control\n",
      "AAACGAAAGCGATACG-1                 6167 Control\n",
      "AAACGAAAGTCGTGAG-1                 1768 Control\n",
      "AAACGAACAATAGTGA-1                 4951 Control\n",
      "AAACGAACACAATAAG-1                 2019 Control\n",
      "AAACGAACACTGATAC-1                 1595 Control\n",
      "[1] \"exp\"\n",
      "[1] \"J2\"\n",
      "[1] \"name\"\n",
      "[1] \"Flt3l\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m12009\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m1\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): X1\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m430951\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m18\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m  (2): barcode, cell_id\n",
      "\u001b[32mdbl\u001b[39m (16): total, duplicate, chimeric, unmapped, lowmapq, mitochondrial, pass...\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating fragments object\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Quantifying peaks\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating chromatin assay\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CreateSeuratObject.Assay(curr_assay, assay = \"ATAC\", project = name, :\n",
      "\"Some cells in meta.data not present in provided counts matrix.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"curr_assay\"\n",
      "data frame with 0 columns and 10 rows\n",
      "[1] \"curr\"\n",
      "                   orig.ident nCount_ATAC nFeature_ATAC total duplicate\n",
      "AAACGAAAGAGCTCCC-1      Flt3l        4716          4430 25798      3956\n",
      "AAACGAAAGCGATACG-1      Flt3l        2136          2066 18340      2478\n",
      "AAACGAAAGGCTTCGC-1      Flt3l        3875          3691 28417      4398\n",
      "AAACGAAAGTACAACA-1      Flt3l        1353          1317 12018      1524\n",
      "AAACGAACAACGTACT-1      Flt3l        2973          2864 37690      3034\n",
      "AAACGAACAAGCGGTA-1      Flt3l        1292          1257  9381      1416\n",
      "                   chimeric unmapped lowmapq mitochondrial passed_filters\n",
      "AAACGAAAGAGCTCCC-1      111       95    1643          4740          15253\n",
      "AAACGAAAGCGATACG-1       78       92     794          9325           5573\n",
      "AAACGAAAGGCTTCGC-1      132      122    1771         10224          11770\n",
      "AAACGAAAGTACAACA-1       60       45     585          5688           4116\n",
      "AAACGAACAACGTACT-1       82      125    1680         24974           7795\n",
      "AAACGAACAAGCGGTA-1       44       50     547          3616           3708\n",
      "                   cell_id is__cell_barcode TSS_fragments\n",
      "AAACGAAAGAGCTCCC-1 _cell_0                1          2708\n",
      "AAACGAAAGCGATACG-1 _cell_1                1          1062\n",
      "AAACGAAAGGCTTCGC-1 _cell_2                1          2145\n",
      "AAACGAAAGTACAACA-1 _cell_3                1           727\n",
      "AAACGAACAACGTACT-1 _cell_4                1          1562\n",
      "AAACGAACAAGCGGTA-1 _cell_5                1           705\n",
      "                   DNase_sensitive_region_fragments enhancer_region_fragments\n",
      "AAACGAAAGAGCTCCC-1                                0                         0\n",
      "AAACGAAAGCGATACG-1                                0                         0\n",
      "AAACGAAAGGCTTCGC-1                                0                         0\n",
      "AAACGAAAGTACAACA-1                                0                         0\n",
      "AAACGAACAACGTACT-1                                0                         0\n",
      "AAACGAACAAGCGGTA-1                                0                         0\n",
      "                   promoter_region_fragments on_target_fragments\n",
      "AAACGAAAGAGCTCCC-1                         0                2708\n",
      "AAACGAAAGCGATACG-1                         0                1062\n",
      "AAACGAAAGGCTTCGC-1                         0                2145\n",
      "AAACGAAAGTACAACA-1                         0                 727\n",
      "AAACGAACAACGTACT-1                         0                1562\n",
      "AAACGAACAAGCGGTA-1                         0                 705\n",
      "                   blacklist_region_fragments peak_region_fragments\n",
      "AAACGAAAGAGCTCCC-1                          0                  4667\n",
      "AAACGAAAGCGATACG-1                          0                  2122\n",
      "AAACGAAAGGCTTCGC-1                          0                  3851\n",
      "AAACGAAAGTACAACA-1                          0                  1329\n",
      "AAACGAACAACGTACT-1                          0                  2945\n",
      "AAACGAACAAGCGGTA-1                          0                  1279\n",
      "                   peak_region_cutsites  proj\n",
      "AAACGAAAGAGCTCCC-1                 8189 Flt3l\n",
      "AAACGAAAGCGATACG-1                 3975 Flt3l\n",
      "AAACGAAAGGCTTCGC-1                 6992 Flt3l\n",
      "AAACGAAAGTACAACA-1                 2440 Flt3l\n",
      "AAACGAACAACGTACT-1                 5445 Flt3l\n",
      "AAACGAACAAGCGGTA-1                 2386 Flt3l\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "An object of class Seurat \n",
       "147518 features across 6543 samples within 1 assay \n",
       "Active assay: ATAC (147518 features, 0 variable features)\n",
       "\n",
       "[[2]]\n",
       "An object of class Seurat \n",
       "149495 features across 11949 samples within 1 assay \n",
       "Active assay: ATAC (149495 features, 0 variable features)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allSE = c() \n",
    "\n",
    "samples_df <- cbind(sample_names, samples)\n",
    "print('samples_df')\n",
    "print(samples_df)\n",
    "for (row in 1:nrow(samples_df)){\n",
    "    exp <- (samples_df[[row, \"samples\"]])\n",
    "    name <- (samples_df[[row, \"sample_names\"]]) \n",
    "    \n",
    "#for (exp in samples) {\n",
    "    print('exp')\n",
    "    print(exp)\n",
    "    print('name')\n",
    "    print(name)\n",
    "    barcode_path <- file.path(cellr_in, exp, \"outs\", \"filtered_peak_bc_matrix\", \"barcodes.tsv\")    \n",
    "    barcodes <- readr::read_tsv(barcode_path, col_names = F) # %>% tidyr::unite(barcode)\n",
    "    barcodes <- as.data.frame(barcodes) %>%  tibble::column_to_rownames(var=\"X1\") %>% tibble::add_column(proj=name)\n",
    "    frag_file <- file.path(cellr_in, exp, \"outs\", \"fragments.tsv.gz\")\n",
    "    \n",
    "    cells.meta.f <- file.path(cellr_in, exp, \"outs\", \"singlecell.csv\") \n",
    "    cells.meta <- as.data.frame(readr::read_csv(cells.meta.f)) %>% tibble::column_to_rownames(var=\"barcode\") %>% tibble::add_column(proj=name)\n",
    "    cells.meta <- cells.meta[rownames(cells.meta) %in% rownames(barcodes), ]\n",
    "\n",
    "    # quantify multiome peaks in the scATAC-seq dataset\n",
    "    \n",
    "    \n",
    "    print(\"Creating fragments object\")\n",
    "    frags.curr <- CreateFragmentObject(path = frag_file, cells= rownames(barcodes))\n",
    "    #print(frags.curr)\n",
    "    print(\"Quantifying peaks\")\n",
    "    ## Quantify peaks\n",
    "    curr.counts <- FeatureMatrix(\n",
    "      fragments = frags.curr,\n",
    "      features = combined.peaks,\n",
    "      cells = rownames(barcodes),\n",
    "      process_n = cores\n",
    "    )\n",
    "    \n",
    "    print(\"Creating chromatin assay\")\n",
    "    ## Create the objects and use simple filters\n",
    "    curr_assay <- CreateChromatinAssay(curr.counts, fragments = frags.curr, min.cells = 10, min.features = 200)\n",
    "    curr <- CreateSeuratObject(curr_assay, assay = \"ATAC\", project=name, meta.data=cells.meta)\n",
    "    print('curr_assay')\n",
    "    print(head(curr_assay))\n",
    "    print('curr')\n",
    "    print(head(curr[[]]))\n",
    "    allSE = c(allSE, curr)\n",
    "    #return(curr)\n",
    "}\n",
    "\n",
    "allSE\n",
    "\n",
    "#allSE <- sapply(samples, create_frag, cellr_in=cellr_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c116a7db",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(allSE, file.path(outdir, paste0(\"allSamples.rds\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f9082d43",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "An object of class Seurat \n",
       "147518 features across 6543 samples within 1 assay \n",
       "Active assay: ATAC (147518 features, 0 variable features)\n",
       "\n",
       "[[2]]\n",
       "An object of class Seurat \n",
       "149495 features across 11949 samples within 1 assay \n",
       "Active assay: ATAC (149495 features, 0 variable features)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bc70c919",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6543 × 21</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_ATAC</th><th scope=col>nFeature_ATAC</th><th scope=col>total</th><th scope=col>duplicate</th><th scope=col>chimeric</th><th scope=col>unmapped</th><th scope=col>lowmapq</th><th scope=col>mitochondrial</th><th scope=col>passed_filters</th><th scope=col>⋯</th><th scope=col>is__cell_barcode</th><th scope=col>TSS_fragments</th><th scope=col>DNase_sensitive_region_fragments</th><th scope=col>enhancer_region_fragments</th><th scope=col>promoter_region_fragments</th><th scope=col>on_target_fragments</th><th scope=col>blacklist_region_fragments</th><th scope=col>peak_region_fragments</th><th scope=col>peak_region_cutsites</th><th scope=col>proj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>AAACGAAAGAGGTCCA-1</th><td>Control</td><td>1400</td><td>1357</td><td> 16929</td><td> 2587</td><td> 58</td><td> 70</td><td> 674</td><td> 9629</td><td> 3911</td><td>⋯</td><td>1</td><td> 740</td><td>0</td><td>0</td><td>0</td><td> 740</td><td>0</td><td>1330</td><td> 2434</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGCGATACG-1</th><td>Control</td><td>3692</td><td>3505</td><td> 45359</td><td>10546</td><td>184</td><td>184</td><td>2302</td><td>22247</td><td> 9896</td><td>⋯</td><td>1</td><td>1611</td><td>0</td><td>0</td><td>0</td><td>1611</td><td>0</td><td>3462</td><td> 6167</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAAGTCGTGAG-1</th><td>Control</td><td>1031</td><td> 994</td><td> 10177</td><td> 2450</td><td> 48</td><td> 60</td><td> 657</td><td> 2135</td><td> 4827</td><td>⋯</td><td>1</td><td> 775</td><td>0</td><td>0</td><td>0</td><td> 775</td><td>0</td><td> 966</td><td> 1768</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAACAATAGTGA-1</th><td>Control</td><td>2829</td><td>2687</td><td> 22452</td><td> 4364</td><td>100</td><td> 83</td><td> 897</td><td> 9233</td><td> 7775</td><td>⋯</td><td>1</td><td>1239</td><td>0</td><td>0</td><td>0</td><td>1239</td><td>0</td><td>2701</td><td> 4951</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAACACAATAAG-1</th><td>Control</td><td>1206</td><td>1184</td><td> 11763</td><td> 2003</td><td> 43</td><td> 57</td><td> 680</td><td> 5547</td><td> 3433</td><td>⋯</td><td>1</td><td> 621</td><td>0</td><td>0</td><td>0</td><td> 621</td><td>0</td><td>1138</td><td> 2019</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAACACTGATAC-1</th><td>Control</td><td> 968</td><td> 954</td><td>  9449</td><td> 1971</td><td> 59</td><td> 21</td><td> 489</td><td> 3785</td><td> 3124</td><td>⋯</td><td>1</td><td> 517</td><td>0</td><td>0</td><td>0</td><td> 517</td><td>0</td><td> 878</td><td> 1595</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAACAGTAACCT-1</th><td>Control</td><td> 702</td><td> 693</td><td>  6196</td><td> 1097</td><td> 19</td><td> 17</td><td> 288</td><td> 2635</td><td> 2140</td><td>⋯</td><td>1</td><td> 397</td><td>0</td><td>0</td><td>0</td><td> 397</td><td>0</td><td> 651</td><td> 1170</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAACATCGTGAT-1</th><td>Control</td><td>7716</td><td>6800</td><td>100024</td><td>36721</td><td>588</td><td>531</td><td>5317</td><td>36828</td><td>20039</td><td>⋯</td><td>1</td><td>3467</td><td>0</td><td>0</td><td>0</td><td>3467</td><td>0</td><td>7263</td><td>13261</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAGTAACGGTG-1</th><td>Control</td><td>1111</td><td>1074</td><td>  8454</td><td> 1950</td><td> 24</td><td> 33</td><td> 354</td><td> 3729</td><td> 2364</td><td>⋯</td><td>1</td><td> 451</td><td>0</td><td>0</td><td>0</td><td> 451</td><td>0</td><td>1042</td><td> 1920</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAGTAAGTCTC-1</th><td>Control</td><td>1273</td><td>1262</td><td>  7684</td><td> 1420</td><td> 36</td><td> 23</td><td> 488</td><td> 1787</td><td> 3930</td><td>⋯</td><td>1</td><td> 649</td><td>0</td><td>0</td><td>0</td><td> 649</td><td>0</td><td>1195</td><td> 2115</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAGTACCAAGG-1</th><td>Control</td><td>4175</td><td>3936</td><td> 27865</td><td> 7135</td><td>126</td><td>131</td><td>1330</td><td> 9549</td><td> 9594</td><td>⋯</td><td>1</td><td>1733</td><td>0</td><td>0</td><td>0</td><td>1733</td><td>0</td><td>3887</td><td> 7139</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAGTCCCGTGA-1</th><td>Control</td><td>4720</td><td>4425</td><td> 31805</td><td> 8689</td><td>178</td><td>135</td><td>1649</td><td> 7738</td><td>13416</td><td>⋯</td><td>1</td><td>2181</td><td>0</td><td>0</td><td>0</td><td>2181</td><td>0</td><td>4387</td><td> 7857</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAAGTGGATTCT-1</th><td>Control</td><td> 813</td><td> 805</td><td>  9518</td><td> 1121</td><td> 32</td><td> 37</td><td> 330</td><td> 5983</td><td> 2015</td><td>⋯</td><td>1</td><td> 414</td><td>0</td><td>0</td><td>0</td><td> 414</td><td>0</td><td> 769</td><td> 1431</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAATCGAGCGCT-1</th><td>Control</td><td>1941</td><td>1892</td><td> 11456</td><td> 2558</td><td> 61</td><td> 49</td><td> 687</td><td> 3476</td><td> 4625</td><td>⋯</td><td>1</td><td> 806</td><td>0</td><td>0</td><td>0</td><td> 806</td><td>0</td><td>1807</td><td> 3217</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAATCGCAACTA-1</th><td>Control</td><td>6811</td><td>6272</td><td> 56262</td><td>15793</td><td>214</td><td>374</td><td>3523</td><td>19210</td><td>17148</td><td>⋯</td><td>1</td><td>3042</td><td>0</td><td>0</td><td>0</td><td>3042</td><td>0</td><td>6449</td><td>11150</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACGAATCGCGCTGA-1</th><td>Control</td><td> 621</td><td> 614</td><td>  6366</td><td>  747</td><td> 30</td><td> 20</td><td> 306</td><td> 3521</td><td> 1742</td><td>⋯</td><td>1</td><td> 308</td><td>0</td><td>0</td><td>0</td><td> 308</td><td>0</td><td> 583</td><td> 1076</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGAGATTACGA-1</th><td>Control</td><td>1808</td><td>1763</td><td> 22032</td><td> 3419</td><td> 83</td><td> 85</td><td>1085</td><td>11844</td><td> 5516</td><td>⋯</td><td>1</td><td> 832</td><td>0</td><td>0</td><td>0</td><td> 832</td><td>0</td><td>1710</td><td> 3182</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGCAAGACTTC-1</th><td>Control</td><td> 409</td><td> 402</td><td>  5303</td><td>  728</td><td>  8</td><td> 18</td><td> 232</td><td> 2874</td><td> 1443</td><td>⋯</td><td>1</td><td> 260</td><td>0</td><td>0</td><td>0</td><td> 260</td><td>0</td><td> 378</td><td>  666</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGCACAAGGGT-1</th><td>Control</td><td> 352</td><td> 336</td><td> 18623</td><td>  982</td><td>  9</td><td> 91</td><td> 510</td><td>15503</td><td> 1528</td><td>⋯</td><td>1</td><td> 292</td><td>0</td><td>0</td><td>0</td><td> 292</td><td>0</td><td> 334</td><td>  589</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGCACCTCGTT-1</th><td>Control</td><td>3709</td><td>3499</td><td> 30543</td><td> 6297</td><td>148</td><td>239</td><td>1359</td><td>12689</td><td> 9811</td><td>⋯</td><td>1</td><td>1707</td><td>0</td><td>0</td><td>0</td><td>1707</td><td>0</td><td>3507</td><td> 6132</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGCACTTGCGG-1</th><td>Control</td><td> 646</td><td> 631</td><td>  7038</td><td>  991</td><td> 27</td><td> 31</td><td> 299</td><td> 3787</td><td> 1903</td><td>⋯</td><td>1</td><td> 316</td><td>0</td><td>0</td><td>0</td><td> 316</td><td>0</td><td> 604</td><td> 1102</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGCATCGCCTT-1</th><td>Control</td><td>1514</td><td>1469</td><td> 13537</td><td> 3614</td><td> 56</td><td> 90</td><td> 777</td><td> 3050</td><td> 5950</td><td>⋯</td><td>1</td><td>1098</td><td>0</td><td>0</td><td>0</td><td>1098</td><td>0</td><td>1378</td><td> 2403</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGCATGTGGGA-1</th><td>Control</td><td>3538</td><td>3349</td><td> 48037</td><td> 8932</td><td>171</td><td>241</td><td>1618</td><td>26241</td><td>10834</td><td>⋯</td><td>1</td><td>1726</td><td>0</td><td>0</td><td>0</td><td>1726</td><td>0</td><td>3277</td><td> 5957</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGGTATCCTTT-1</th><td>Control</td><td>2727</td><td>2525</td><td> 21479</td><td> 7157</td><td>101</td><td>159</td><td>1220</td><td> 2182</td><td>10660</td><td>⋯</td><td>1</td><td>2145</td><td>0</td><td>0</td><td>0</td><td>2145</td><td>0</td><td>2456</td><td> 4185</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGGTGCCCTAG-1</th><td>Control</td><td>3607</td><td>3414</td><td> 32698</td><td> 7966</td><td>110</td><td>187</td><td>1479</td><td>13991</td><td> 8965</td><td>⋯</td><td>1</td><td>1591</td><td>0</td><td>0</td><td>0</td><td>1591</td><td>0</td><td>3357</td><td> 6010</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGGTGTAGGAC-1</th><td>Control</td><td> 530</td><td> 517</td><td>  3157</td><td>  501</td><td>  9</td><td> 12</td><td> 201</td><td>  977</td><td> 1457</td><td>⋯</td><td>1</td><td> 196</td><td>0</td><td>0</td><td>0</td><td> 196</td><td>0</td><td> 494</td><td>  854</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGTCACTAGCA-1</th><td>Control</td><td>1628</td><td>1571</td><td> 30764</td><td> 4287</td><td> 64</td><td>125</td><td>1018</td><td>19274</td><td> 5996</td><td>⋯</td><td>1</td><td> 838</td><td>0</td><td>0</td><td>0</td><td> 838</td><td>0</td><td>1532</td><td> 2755</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGTCGAGTGTT-1</th><td>Control</td><td>2872</td><td>2715</td><td> 29393</td><td> 6514</td><td>144</td><td>188</td><td>1598</td><td>10615</td><td>10334</td><td>⋯</td><td>1</td><td>1903</td><td>0</td><td>0</td><td>0</td><td>1903</td><td>0</td><td>2632</td><td> 4689</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTCGTCTATCCTA-1</th><td>Control</td><td>2198</td><td>2111</td><td> 18524</td><td> 5154</td><td>104</td><td> 82</td><td>1150</td><td> 5499</td><td> 6535</td><td>⋯</td><td>1</td><td>1001</td><td>0</td><td>0</td><td>0</td><td>1001</td><td>0</td><td>2046</td><td> 3619</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>AAACTGCAGCATGTGG-1</th><td>Control</td><td>3705</td><td>3475</td><td> 25818</td><td> 5609</td><td>136</td><td>114</td><td>1775</td><td> 7066</td><td>11118</td><td>⋯</td><td>1</td><td>2326</td><td>0</td><td>0</td><td>0</td><td>2326</td><td>0</td><td>3505</td><td> 6308</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>TTTGGCCGTGCGTAGA-1</th><td>Control</td><td> 2638</td><td> 2515</td><td>21042</td><td> 4238</td><td> 90</td><td>146</td><td>1400</td><td> 7990</td><td> 7178</td><td>⋯</td><td>1</td><td>1300</td><td>0</td><td>0</td><td>0</td><td>1300</td><td>0</td><td> 2518</td><td> 4400</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGCCGTTAGGCTT-1</th><td>Control</td><td> 1615</td><td> 1562</td><td>18757</td><td> 3321</td><td> 72</td><td>101</td><td> 855</td><td> 9021</td><td> 5387</td><td>⋯</td><td>1</td><td> 893</td><td>0</td><td>0</td><td>0</td><td> 893</td><td>0</td><td> 1498</td><td> 2682</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGCCTCTGGCTAA-1</th><td>Control</td><td> 3822</td><td> 3614</td><td>44721</td><td> 7325</td><td>135</td><td>203</td><td>1947</td><td>24374</td><td>10737</td><td>⋯</td><td>1</td><td>1884</td><td>0</td><td>0</td><td>0</td><td>1884</td><td>0</td><td> 3653</td><td> 6678</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTAGCCCATTA-1</th><td>Control</td><td>  556</td><td>  550</td><td> 2698</td><td>  581</td><td> 28</td><td>  8</td><td> 113</td><td>  699</td><td> 1269</td><td>⋯</td><td>1</td><td> 291</td><td>0</td><td>0</td><td>0</td><td> 291</td><td>0</td><td>  528</td><td>  972</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTAGCGCCTAC-1</th><td>Control</td><td> 1347</td><td> 1304</td><td>18309</td><td> 3837</td><td> 93</td><td> 80</td><td> 766</td><td> 8860</td><td> 4673</td><td>⋯</td><td>1</td><td> 668</td><td>0</td><td>0</td><td>0</td><td> 668</td><td>0</td><td> 1250</td><td> 2237</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTCAAGAAACT-1</th><td>Control</td><td> 1962</td><td> 1904</td><td>25018</td><td> 3679</td><td>101</td><td>115</td><td>1329</td><td>14418</td><td> 5376</td><td>⋯</td><td>1</td><td> 934</td><td>0</td><td>0</td><td>0</td><td> 934</td><td>0</td><td> 1846</td><td> 3267</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTCACCCTTTG-1</th><td>Control</td><td>  800</td><td>  784</td><td>13165</td><td> 1515</td><td> 56</td><td> 47</td><td> 357</td><td> 8522</td><td> 2668</td><td>⋯</td><td>1</td><td> 459</td><td>0</td><td>0</td><td>0</td><td> 459</td><td>0</td><td>  731</td><td> 1344</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTCAGCAGGTA-1</th><td>Control</td><td>14293</td><td>12406</td><td>75730</td><td>26956</td><td>496</td><td>701</td><td>4313</td><td> 5179</td><td>38085</td><td>⋯</td><td>1</td><td>6629</td><td>0</td><td>0</td><td>0</td><td>6629</td><td>0</td><td>13120</td><td>22372</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTCATGGATGG-1</th><td>Control</td><td> 2718</td><td> 2617</td><td>36709</td><td> 6004</td><td>156</td><td>183</td><td>1688</td><td>20638</td><td> 8040</td><td>⋯</td><td>1</td><td>1283</td><td>0</td><td>0</td><td>0</td><td>1283</td><td>0</td><td> 2556</td><td> 4712</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTGTATCTCAG-1</th><td>Control</td><td>  504</td><td>  497</td><td> 4121</td><td>  963</td><td> 19</td><td> 17</td><td> 208</td><td> 1271</td><td> 1643</td><td>⋯</td><td>1</td><td> 230</td><td>0</td><td>0</td><td>0</td><td> 230</td><td>0</td><td>  462</td><td>  794</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTTCACTGATG-1</th><td>Control</td><td>  960</td><td>  935</td><td>19160</td><td> 2190</td><td> 46</td><td> 80</td><td> 454</td><td>13286</td><td> 3104</td><td>⋯</td><td>1</td><td> 469</td><td>0</td><td>0</td><td>0</td><td> 469</td><td>0</td><td>  908</td><td> 1666</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTTCATCATGG-1</th><td>Control</td><td> 2230</td><td> 2164</td><td>24974</td><td> 4120</td><td>118</td><td>111</td><td>1095</td><td>13520</td><td> 6010</td><td>⋯</td><td>1</td><td> 932</td><td>0</td><td>0</td><td>0</td><td> 932</td><td>0</td><td> 2087</td><td> 3733</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTTCCGGACTG-1</th><td>Control</td><td> 2069</td><td> 1980</td><td>21336</td><td> 4738</td><td>103</td><td>126</td><td>1252</td><td> 8244</td><td> 6873</td><td>⋯</td><td>1</td><td>1219</td><td>0</td><td>0</td><td>0</td><td>1219</td><td>0</td><td> 1910</td><td> 3316</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTTCCTGACCT-1</th><td>Control</td><td>  323</td><td>  318</td><td> 4397</td><td>  660</td><td> 18</td><td> 21</td><td> 115</td><td> 2462</td><td> 1121</td><td>⋯</td><td>1</td><td> 166</td><td>0</td><td>0</td><td>0</td><td> 166</td><td>0</td><td>  304</td><td>  552</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTTCGATGAAA-1</th><td>Control</td><td>  712</td><td>  700</td><td> 3150</td><td>  652</td><td> 27</td><td> 15</td><td> 171</td><td>  771</td><td> 1514</td><td>⋯</td><td>1</td><td> 290</td><td>0</td><td>0</td><td>0</td><td> 290</td><td>0</td><td>  653</td><td> 1169</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGGTTTCGCGGCAT-1</th><td>Control</td><td> 3282</td><td> 3119</td><td>22039</td><td> 5742</td><td>102</td><td>147</td><td>1276</td><td> 6173</td><td> 8599</td><td>⋯</td><td>1</td><td>1516</td><td>0</td><td>0</td><td>0</td><td>1516</td><td>0</td><td> 3134</td><td> 5582</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTAGACACGGT-1</th><td>Control</td><td>  724</td><td>  714</td><td> 7099</td><td> 1101</td><td> 46</td><td> 29</td><td> 398</td><td> 3271</td><td> 2254</td><td>⋯</td><td>1</td><td> 361</td><td>0</td><td>0</td><td>0</td><td> 361</td><td>0</td><td>  682</td><td> 1200</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTAGACCGCAA-1</th><td>Control</td><td> 2144</td><td> 2066</td><td>23619</td><td> 4038</td><td> 75</td><td> 97</td><td> 803</td><td>12287</td><td> 6319</td><td>⋯</td><td>1</td><td>1023</td><td>0</td><td>0</td><td>0</td><td>1023</td><td>0</td><td> 2033</td><td> 3623</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTAGACTAATG-1</th><td>Control</td><td> 4521</td><td> 4282</td><td>44991</td><td> 7595</td><td>195</td><td>171</td><td>1749</td><td>24211</td><td>11070</td><td>⋯</td><td>1</td><td>2037</td><td>0</td><td>0</td><td>0</td><td>2037</td><td>0</td><td> 4260</td><td> 7865</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTAGACTTCCA-1</th><td>Control</td><td> 1697</td><td> 1646</td><td>13085</td><td> 2533</td><td> 57</td><td> 49</td><td> 583</td><td> 5314</td><td> 4549</td><td>⋯</td><td>1</td><td> 959</td><td>0</td><td>0</td><td>0</td><td> 959</td><td>0</td><td> 1594</td><td> 2889</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTAGCCAGAAC-1</th><td>Control</td><td>  454</td><td>  450</td><td> 5687</td><td> 1071</td><td> 25</td><td> 23</td><td> 272</td><td> 2237</td><td> 2059</td><td>⋯</td><td>1</td><td> 326</td><td>0</td><td>0</td><td>0</td><td> 326</td><td>0</td><td>  401</td><td>  676</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTCAGCAACCC-1</th><td>Control</td><td> 2173</td><td> 2091</td><td>21754</td><td> 4090</td><td> 95</td><td>106</td><td> 755</td><td> 9631</td><td> 7077</td><td>⋯</td><td>1</td><td>1245</td><td>0</td><td>0</td><td>0</td><td>1245</td><td>0</td><td> 2005</td><td> 3566</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTCATGTATCG-1</th><td>Control</td><td>  803</td><td>  793</td><td> 5577</td><td> 1166</td><td> 34</td><td> 30</td><td> 305</td><td> 1547</td><td> 2495</td><td>⋯</td><td>1</td><td> 403</td><td>0</td><td>0</td><td>0</td><td> 403</td><td>0</td><td>  742</td><td> 1318</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTGTTTGCATG-1</th><td>Control</td><td> 2345</td><td> 2277</td><td>19469</td><td> 3652</td><td> 98</td><td>134</td><td>1140</td><td> 5904</td><td> 8541</td><td>⋯</td><td>1</td><td>1314</td><td>0</td><td>0</td><td>0</td><td>1314</td><td>0</td><td> 2192</td><td> 3746</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCACTACCC-1</th><td>Control</td><td> 1061</td><td> 1037</td><td>18676</td><td> 1809</td><td> 47</td><td> 72</td><td> 575</td><td>12724</td><td> 3449</td><td>⋯</td><td>1</td><td> 585</td><td>0</td><td>0</td><td>0</td><td> 585</td><td>0</td><td>  994</td><td> 1848</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCCAAACCA-1</th><td>Control</td><td> 8737</td><td> 7785</td><td>90247</td><td>25919</td><td>402</td><td>477</td><td>5068</td><td>23688</td><td>34693</td><td>⋯</td><td>1</td><td>5402</td><td>0</td><td>0</td><td>0</td><td>5402</td><td>0</td><td> 8125</td><td>13977</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCCGATGCG-1</th><td>Control</td><td> 2518</td><td> 2411</td><td>26526</td><td> 3605</td><td>109</td><td>114</td><td>1089</td><td>14768</td><td> 6841</td><td>⋯</td><td>1</td><td>1318</td><td>0</td><td>0</td><td>0</td><td>1318</td><td>0</td><td> 2389</td><td> 4370</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCCTAGAGT-1</th><td>Control</td><td>  676</td><td>  670</td><td> 5693</td><td>  758</td><td> 29</td><td> 30</td><td> 220</td><td> 2742</td><td> 1914</td><td>⋯</td><td>1</td><td> 322</td><td>0</td><td>0</td><td>0</td><td> 322</td><td>0</td><td>  639</td><td> 1144</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCGCTCGGA-1</th><td>Control</td><td> 2023</td><td> 1952</td><td>19572</td><td> 2975</td><td> 79</td><td> 77</td><td>1064</td><td> 8534</td><td> 6843</td><td>⋯</td><td>1</td><td>1240</td><td>0</td><td>0</td><td>0</td><td>1240</td><td>0</td><td> 1859</td><td> 3311</td><td>Control</td></tr>\n",
       "\t<tr><th scope=row>TTTGTGTTCTTCCACG-1</th><td>Control</td><td> 4312</td><td> 4010</td><td>57305</td><td>12924</td><td>203</td><td>273</td><td>2172</td><td>28008</td><td>13725</td><td>⋯</td><td>1</td><td>2275</td><td>0</td><td>0</td><td>0</td><td>2275</td><td>0</td><td> 4090</td><td> 7264</td><td>Control</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6543 × 21\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & orig.ident & nCount\\_ATAC & nFeature\\_ATAC & total & duplicate & chimeric & unmapped & lowmapq & mitochondrial & passed\\_filters & ⋯ & is\\_\\_cell\\_barcode & TSS\\_fragments & DNase\\_sensitive\\_region\\_fragments & enhancer\\_region\\_fragments & promoter\\_region\\_fragments & on\\_target\\_fragments & blacklist\\_region\\_fragments & peak\\_region\\_fragments & peak\\_region\\_cutsites & proj\\\\\n",
       "  & <fct> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\tAAACGAAAGAGGTCCA-1 & Control & 1400 & 1357 &  16929 &  2587 &  58 &  70 &  674 &  9629 &  3911 & ⋯ & 1 &  740 & 0 & 0 & 0 &  740 & 0 & 1330 &  2434 & Control\\\\\n",
       "\tAAACGAAAGCGATACG-1 & Control & 3692 & 3505 &  45359 & 10546 & 184 & 184 & 2302 & 22247 &  9896 & ⋯ & 1 & 1611 & 0 & 0 & 0 & 1611 & 0 & 3462 &  6167 & Control\\\\\n",
       "\tAAACGAAAGTCGTGAG-1 & Control & 1031 &  994 &  10177 &  2450 &  48 &  60 &  657 &  2135 &  4827 & ⋯ & 1 &  775 & 0 & 0 & 0 &  775 & 0 &  966 &  1768 & Control\\\\\n",
       "\tAAACGAACAATAGTGA-1 & Control & 2829 & 2687 &  22452 &  4364 & 100 &  83 &  897 &  9233 &  7775 & ⋯ & 1 & 1239 & 0 & 0 & 0 & 1239 & 0 & 2701 &  4951 & Control\\\\\n",
       "\tAAACGAACACAATAAG-1 & Control & 1206 & 1184 &  11763 &  2003 &  43 &  57 &  680 &  5547 &  3433 & ⋯ & 1 &  621 & 0 & 0 & 0 &  621 & 0 & 1138 &  2019 & Control\\\\\n",
       "\tAAACGAACACTGATAC-1 & Control &  968 &  954 &   9449 &  1971 &  59 &  21 &  489 &  3785 &  3124 & ⋯ & 1 &  517 & 0 & 0 & 0 &  517 & 0 &  878 &  1595 & Control\\\\\n",
       "\tAAACGAACAGTAACCT-1 & Control &  702 &  693 &   6196 &  1097 &  19 &  17 &  288 &  2635 &  2140 & ⋯ & 1 &  397 & 0 & 0 & 0 &  397 & 0 &  651 &  1170 & Control\\\\\n",
       "\tAAACGAACATCGTGAT-1 & Control & 7716 & 6800 & 100024 & 36721 & 588 & 531 & 5317 & 36828 & 20039 & ⋯ & 1 & 3467 & 0 & 0 & 0 & 3467 & 0 & 7263 & 13261 & Control\\\\\n",
       "\tAAACGAAGTAACGGTG-1 & Control & 1111 & 1074 &   8454 &  1950 &  24 &  33 &  354 &  3729 &  2364 & ⋯ & 1 &  451 & 0 & 0 & 0 &  451 & 0 & 1042 &  1920 & Control\\\\\n",
       "\tAAACGAAGTAAGTCTC-1 & Control & 1273 & 1262 &   7684 &  1420 &  36 &  23 &  488 &  1787 &  3930 & ⋯ & 1 &  649 & 0 & 0 & 0 &  649 & 0 & 1195 &  2115 & Control\\\\\n",
       "\tAAACGAAGTACCAAGG-1 & Control & 4175 & 3936 &  27865 &  7135 & 126 & 131 & 1330 &  9549 &  9594 & ⋯ & 1 & 1733 & 0 & 0 & 0 & 1733 & 0 & 3887 &  7139 & Control\\\\\n",
       "\tAAACGAAGTCCCGTGA-1 & Control & 4720 & 4425 &  31805 &  8689 & 178 & 135 & 1649 &  7738 & 13416 & ⋯ & 1 & 2181 & 0 & 0 & 0 & 2181 & 0 & 4387 &  7857 & Control\\\\\n",
       "\tAAACGAAGTGGATTCT-1 & Control &  813 &  805 &   9518 &  1121 &  32 &  37 &  330 &  5983 &  2015 & ⋯ & 1 &  414 & 0 & 0 & 0 &  414 & 0 &  769 &  1431 & Control\\\\\n",
       "\tAAACGAATCGAGCGCT-1 & Control & 1941 & 1892 &  11456 &  2558 &  61 &  49 &  687 &  3476 &  4625 & ⋯ & 1 &  806 & 0 & 0 & 0 &  806 & 0 & 1807 &  3217 & Control\\\\\n",
       "\tAAACGAATCGCAACTA-1 & Control & 6811 & 6272 &  56262 & 15793 & 214 & 374 & 3523 & 19210 & 17148 & ⋯ & 1 & 3042 & 0 & 0 & 0 & 3042 & 0 & 6449 & 11150 & Control\\\\\n",
       "\tAAACGAATCGCGCTGA-1 & Control &  621 &  614 &   6366 &   747 &  30 &  20 &  306 &  3521 &  1742 & ⋯ & 1 &  308 & 0 & 0 & 0 &  308 & 0 &  583 &  1076 & Control\\\\\n",
       "\tAAACTCGAGATTACGA-1 & Control & 1808 & 1763 &  22032 &  3419 &  83 &  85 & 1085 & 11844 &  5516 & ⋯ & 1 &  832 & 0 & 0 & 0 &  832 & 0 & 1710 &  3182 & Control\\\\\n",
       "\tAAACTCGCAAGACTTC-1 & Control &  409 &  402 &   5303 &   728 &   8 &  18 &  232 &  2874 &  1443 & ⋯ & 1 &  260 & 0 & 0 & 0 &  260 & 0 &  378 &   666 & Control\\\\\n",
       "\tAAACTCGCACAAGGGT-1 & Control &  352 &  336 &  18623 &   982 &   9 &  91 &  510 & 15503 &  1528 & ⋯ & 1 &  292 & 0 & 0 & 0 &  292 & 0 &  334 &   589 & Control\\\\\n",
       "\tAAACTCGCACCTCGTT-1 & Control & 3709 & 3499 &  30543 &  6297 & 148 & 239 & 1359 & 12689 &  9811 & ⋯ & 1 & 1707 & 0 & 0 & 0 & 1707 & 0 & 3507 &  6132 & Control\\\\\n",
       "\tAAACTCGCACTTGCGG-1 & Control &  646 &  631 &   7038 &   991 &  27 &  31 &  299 &  3787 &  1903 & ⋯ & 1 &  316 & 0 & 0 & 0 &  316 & 0 &  604 &  1102 & Control\\\\\n",
       "\tAAACTCGCATCGCCTT-1 & Control & 1514 & 1469 &  13537 &  3614 &  56 &  90 &  777 &  3050 &  5950 & ⋯ & 1 & 1098 & 0 & 0 & 0 & 1098 & 0 & 1378 &  2403 & Control\\\\\n",
       "\tAAACTCGCATGTGGGA-1 & Control & 3538 & 3349 &  48037 &  8932 & 171 & 241 & 1618 & 26241 & 10834 & ⋯ & 1 & 1726 & 0 & 0 & 0 & 1726 & 0 & 3277 &  5957 & Control\\\\\n",
       "\tAAACTCGGTATCCTTT-1 & Control & 2727 & 2525 &  21479 &  7157 & 101 & 159 & 1220 &  2182 & 10660 & ⋯ & 1 & 2145 & 0 & 0 & 0 & 2145 & 0 & 2456 &  4185 & Control\\\\\n",
       "\tAAACTCGGTGCCCTAG-1 & Control & 3607 & 3414 &  32698 &  7966 & 110 & 187 & 1479 & 13991 &  8965 & ⋯ & 1 & 1591 & 0 & 0 & 0 & 1591 & 0 & 3357 &  6010 & Control\\\\\n",
       "\tAAACTCGGTGTAGGAC-1 & Control &  530 &  517 &   3157 &   501 &   9 &  12 &  201 &   977 &  1457 & ⋯ & 1 &  196 & 0 & 0 & 0 &  196 & 0 &  494 &   854 & Control\\\\\n",
       "\tAAACTCGTCACTAGCA-1 & Control & 1628 & 1571 &  30764 &  4287 &  64 & 125 & 1018 & 19274 &  5996 & ⋯ & 1 &  838 & 0 & 0 & 0 &  838 & 0 & 1532 &  2755 & Control\\\\\n",
       "\tAAACTCGTCGAGTGTT-1 & Control & 2872 & 2715 &  29393 &  6514 & 144 & 188 & 1598 & 10615 & 10334 & ⋯ & 1 & 1903 & 0 & 0 & 0 & 1903 & 0 & 2632 &  4689 & Control\\\\\n",
       "\tAAACTCGTCTATCCTA-1 & Control & 2198 & 2111 &  18524 &  5154 & 104 &  82 & 1150 &  5499 &  6535 & ⋯ & 1 & 1001 & 0 & 0 & 0 & 1001 & 0 & 2046 &  3619 & Control\\\\\n",
       "\tAAACTGCAGCATGTGG-1 & Control & 3705 & 3475 &  25818 &  5609 & 136 & 114 & 1775 &  7066 & 11118 & ⋯ & 1 & 2326 & 0 & 0 & 0 & 2326 & 0 & 3505 &  6308 & Control\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\tTTTGGCCGTGCGTAGA-1 & Control &  2638 &  2515 & 21042 &  4238 &  90 & 146 & 1400 &  7990 &  7178 & ⋯ & 1 & 1300 & 0 & 0 & 0 & 1300 & 0 &  2518 &  4400 & Control\\\\\n",
       "\tTTTGGCCGTTAGGCTT-1 & Control &  1615 &  1562 & 18757 &  3321 &  72 & 101 &  855 &  9021 &  5387 & ⋯ & 1 &  893 & 0 & 0 & 0 &  893 & 0 &  1498 &  2682 & Control\\\\\n",
       "\tTTTGGCCTCTGGCTAA-1 & Control &  3822 &  3614 & 44721 &  7325 & 135 & 203 & 1947 & 24374 & 10737 & ⋯ & 1 & 1884 & 0 & 0 & 0 & 1884 & 0 &  3653 &  6678 & Control\\\\\n",
       "\tTTTGGTTAGCCCATTA-1 & Control &   556 &   550 &  2698 &   581 &  28 &   8 &  113 &   699 &  1269 & ⋯ & 1 &  291 & 0 & 0 & 0 &  291 & 0 &   528 &   972 & Control\\\\\n",
       "\tTTTGGTTAGCGCCTAC-1 & Control &  1347 &  1304 & 18309 &  3837 &  93 &  80 &  766 &  8860 &  4673 & ⋯ & 1 &  668 & 0 & 0 & 0 &  668 & 0 &  1250 &  2237 & Control\\\\\n",
       "\tTTTGGTTCAAGAAACT-1 & Control &  1962 &  1904 & 25018 &  3679 & 101 & 115 & 1329 & 14418 &  5376 & ⋯ & 1 &  934 & 0 & 0 & 0 &  934 & 0 &  1846 &  3267 & Control\\\\\n",
       "\tTTTGGTTCACCCTTTG-1 & Control &   800 &   784 & 13165 &  1515 &  56 &  47 &  357 &  8522 &  2668 & ⋯ & 1 &  459 & 0 & 0 & 0 &  459 & 0 &   731 &  1344 & Control\\\\\n",
       "\tTTTGGTTCAGCAGGTA-1 & Control & 14293 & 12406 & 75730 & 26956 & 496 & 701 & 4313 &  5179 & 38085 & ⋯ & 1 & 6629 & 0 & 0 & 0 & 6629 & 0 & 13120 & 22372 & Control\\\\\n",
       "\tTTTGGTTCATGGATGG-1 & Control &  2718 &  2617 & 36709 &  6004 & 156 & 183 & 1688 & 20638 &  8040 & ⋯ & 1 & 1283 & 0 & 0 & 0 & 1283 & 0 &  2556 &  4712 & Control\\\\\n",
       "\tTTTGGTTGTATCTCAG-1 & Control &   504 &   497 &  4121 &   963 &  19 &  17 &  208 &  1271 &  1643 & ⋯ & 1 &  230 & 0 & 0 & 0 &  230 & 0 &   462 &   794 & Control\\\\\n",
       "\tTTTGGTTTCACTGATG-1 & Control &   960 &   935 & 19160 &  2190 &  46 &  80 &  454 & 13286 &  3104 & ⋯ & 1 &  469 & 0 & 0 & 0 &  469 & 0 &   908 &  1666 & Control\\\\\n",
       "\tTTTGGTTTCATCATGG-1 & Control &  2230 &  2164 & 24974 &  4120 & 118 & 111 & 1095 & 13520 &  6010 & ⋯ & 1 &  932 & 0 & 0 & 0 &  932 & 0 &  2087 &  3733 & Control\\\\\n",
       "\tTTTGGTTTCCGGACTG-1 & Control &  2069 &  1980 & 21336 &  4738 & 103 & 126 & 1252 &  8244 &  6873 & ⋯ & 1 & 1219 & 0 & 0 & 0 & 1219 & 0 &  1910 &  3316 & Control\\\\\n",
       "\tTTTGGTTTCCTGACCT-1 & Control &   323 &   318 &  4397 &   660 &  18 &  21 &  115 &  2462 &  1121 & ⋯ & 1 &  166 & 0 & 0 & 0 &  166 & 0 &   304 &   552 & Control\\\\\n",
       "\tTTTGGTTTCGATGAAA-1 & Control &   712 &   700 &  3150 &   652 &  27 &  15 &  171 &   771 &  1514 & ⋯ & 1 &  290 & 0 & 0 & 0 &  290 & 0 &   653 &  1169 & Control\\\\\n",
       "\tTTTGGTTTCGCGGCAT-1 & Control &  3282 &  3119 & 22039 &  5742 & 102 & 147 & 1276 &  6173 &  8599 & ⋯ & 1 & 1516 & 0 & 0 & 0 & 1516 & 0 &  3134 &  5582 & Control\\\\\n",
       "\tTTTGTGTAGACACGGT-1 & Control &   724 &   714 &  7099 &  1101 &  46 &  29 &  398 &  3271 &  2254 & ⋯ & 1 &  361 & 0 & 0 & 0 &  361 & 0 &   682 &  1200 & Control\\\\\n",
       "\tTTTGTGTAGACCGCAA-1 & Control &  2144 &  2066 & 23619 &  4038 &  75 &  97 &  803 & 12287 &  6319 & ⋯ & 1 & 1023 & 0 & 0 & 0 & 1023 & 0 &  2033 &  3623 & Control\\\\\n",
       "\tTTTGTGTAGACTAATG-1 & Control &  4521 &  4282 & 44991 &  7595 & 195 & 171 & 1749 & 24211 & 11070 & ⋯ & 1 & 2037 & 0 & 0 & 0 & 2037 & 0 &  4260 &  7865 & Control\\\\\n",
       "\tTTTGTGTAGACTTCCA-1 & Control &  1697 &  1646 & 13085 &  2533 &  57 &  49 &  583 &  5314 &  4549 & ⋯ & 1 &  959 & 0 & 0 & 0 &  959 & 0 &  1594 &  2889 & Control\\\\\n",
       "\tTTTGTGTAGCCAGAAC-1 & Control &   454 &   450 &  5687 &  1071 &  25 &  23 &  272 &  2237 &  2059 & ⋯ & 1 &  326 & 0 & 0 & 0 &  326 & 0 &   401 &   676 & Control\\\\\n",
       "\tTTTGTGTCAGCAACCC-1 & Control &  2173 &  2091 & 21754 &  4090 &  95 & 106 &  755 &  9631 &  7077 & ⋯ & 1 & 1245 & 0 & 0 & 0 & 1245 & 0 &  2005 &  3566 & Control\\\\\n",
       "\tTTTGTGTCATGTATCG-1 & Control &   803 &   793 &  5577 &  1166 &  34 &  30 &  305 &  1547 &  2495 & ⋯ & 1 &  403 & 0 & 0 & 0 &  403 & 0 &   742 &  1318 & Control\\\\\n",
       "\tTTTGTGTGTTTGCATG-1 & Control &  2345 &  2277 & 19469 &  3652 &  98 & 134 & 1140 &  5904 &  8541 & ⋯ & 1 & 1314 & 0 & 0 & 0 & 1314 & 0 &  2192 &  3746 & Control\\\\\n",
       "\tTTTGTGTTCACTACCC-1 & Control &  1061 &  1037 & 18676 &  1809 &  47 &  72 &  575 & 12724 &  3449 & ⋯ & 1 &  585 & 0 & 0 & 0 &  585 & 0 &   994 &  1848 & Control\\\\\n",
       "\tTTTGTGTTCCAAACCA-1 & Control &  8737 &  7785 & 90247 & 25919 & 402 & 477 & 5068 & 23688 & 34693 & ⋯ & 1 & 5402 & 0 & 0 & 0 & 5402 & 0 &  8125 & 13977 & Control\\\\\n",
       "\tTTTGTGTTCCGATGCG-1 & Control &  2518 &  2411 & 26526 &  3605 & 109 & 114 & 1089 & 14768 &  6841 & ⋯ & 1 & 1318 & 0 & 0 & 0 & 1318 & 0 &  2389 &  4370 & Control\\\\\n",
       "\tTTTGTGTTCCTAGAGT-1 & Control &   676 &   670 &  5693 &   758 &  29 &  30 &  220 &  2742 &  1914 & ⋯ & 1 &  322 & 0 & 0 & 0 &  322 & 0 &   639 &  1144 & Control\\\\\n",
       "\tTTTGTGTTCGCTCGGA-1 & Control &  2023 &  1952 & 19572 &  2975 &  79 &  77 & 1064 &  8534 &  6843 & ⋯ & 1 & 1240 & 0 & 0 & 0 & 1240 & 0 &  1859 &  3311 & Control\\\\\n",
       "\tTTTGTGTTCTTCCACG-1 & Control &  4312 &  4010 & 57305 & 12924 & 203 & 273 & 2172 & 28008 & 13725 & ⋯ & 1 & 2275 & 0 & 0 & 0 & 2275 & 0 &  4090 &  7264 & Control\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6543 × 21\n",
       "\n",
       "| <!--/--> | orig.ident &lt;fct&gt; | nCount_ATAC &lt;dbl&gt; | nFeature_ATAC &lt;int&gt; | total &lt;dbl&gt; | duplicate &lt;dbl&gt; | chimeric &lt;dbl&gt; | unmapped &lt;dbl&gt; | lowmapq &lt;dbl&gt; | mitochondrial &lt;dbl&gt; | passed_filters &lt;dbl&gt; | ⋯ ⋯ | is__cell_barcode &lt;dbl&gt; | TSS_fragments &lt;dbl&gt; | DNase_sensitive_region_fragments &lt;dbl&gt; | enhancer_region_fragments &lt;dbl&gt; | promoter_region_fragments &lt;dbl&gt; | on_target_fragments &lt;dbl&gt; | blacklist_region_fragments &lt;dbl&gt; | peak_region_fragments &lt;dbl&gt; | peak_region_cutsites &lt;dbl&gt; | proj &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| AAACGAAAGAGGTCCA-1 | Control | 1400 | 1357 |  16929 |  2587 |  58 |  70 |  674 |  9629 |  3911 | ⋯ | 1 |  740 | 0 | 0 | 0 |  740 | 0 | 1330 |  2434 | Control |\n",
       "| AAACGAAAGCGATACG-1 | Control | 3692 | 3505 |  45359 | 10546 | 184 | 184 | 2302 | 22247 |  9896 | ⋯ | 1 | 1611 | 0 | 0 | 0 | 1611 | 0 | 3462 |  6167 | Control |\n",
       "| AAACGAAAGTCGTGAG-1 | Control | 1031 |  994 |  10177 |  2450 |  48 |  60 |  657 |  2135 |  4827 | ⋯ | 1 |  775 | 0 | 0 | 0 |  775 | 0 |  966 |  1768 | Control |\n",
       "| AAACGAACAATAGTGA-1 | Control | 2829 | 2687 |  22452 |  4364 | 100 |  83 |  897 |  9233 |  7775 | ⋯ | 1 | 1239 | 0 | 0 | 0 | 1239 | 0 | 2701 |  4951 | Control |\n",
       "| AAACGAACACAATAAG-1 | Control | 1206 | 1184 |  11763 |  2003 |  43 |  57 |  680 |  5547 |  3433 | ⋯ | 1 |  621 | 0 | 0 | 0 |  621 | 0 | 1138 |  2019 | Control |\n",
       "| AAACGAACACTGATAC-1 | Control |  968 |  954 |   9449 |  1971 |  59 |  21 |  489 |  3785 |  3124 | ⋯ | 1 |  517 | 0 | 0 | 0 |  517 | 0 |  878 |  1595 | Control |\n",
       "| AAACGAACAGTAACCT-1 | Control |  702 |  693 |   6196 |  1097 |  19 |  17 |  288 |  2635 |  2140 | ⋯ | 1 |  397 | 0 | 0 | 0 |  397 | 0 |  651 |  1170 | Control |\n",
       "| AAACGAACATCGTGAT-1 | Control | 7716 | 6800 | 100024 | 36721 | 588 | 531 | 5317 | 36828 | 20039 | ⋯ | 1 | 3467 | 0 | 0 | 0 | 3467 | 0 | 7263 | 13261 | Control |\n",
       "| AAACGAAGTAACGGTG-1 | Control | 1111 | 1074 |   8454 |  1950 |  24 |  33 |  354 |  3729 |  2364 | ⋯ | 1 |  451 | 0 | 0 | 0 |  451 | 0 | 1042 |  1920 | Control |\n",
       "| AAACGAAGTAAGTCTC-1 | Control | 1273 | 1262 |   7684 |  1420 |  36 |  23 |  488 |  1787 |  3930 | ⋯ | 1 |  649 | 0 | 0 | 0 |  649 | 0 | 1195 |  2115 | Control |\n",
       "| AAACGAAGTACCAAGG-1 | Control | 4175 | 3936 |  27865 |  7135 | 126 | 131 | 1330 |  9549 |  9594 | ⋯ | 1 | 1733 | 0 | 0 | 0 | 1733 | 0 | 3887 |  7139 | Control |\n",
       "| AAACGAAGTCCCGTGA-1 | Control | 4720 | 4425 |  31805 |  8689 | 178 | 135 | 1649 |  7738 | 13416 | ⋯ | 1 | 2181 | 0 | 0 | 0 | 2181 | 0 | 4387 |  7857 | Control |\n",
       "| AAACGAAGTGGATTCT-1 | Control |  813 |  805 |   9518 |  1121 |  32 |  37 |  330 |  5983 |  2015 | ⋯ | 1 |  414 | 0 | 0 | 0 |  414 | 0 |  769 |  1431 | Control |\n",
       "| AAACGAATCGAGCGCT-1 | Control | 1941 | 1892 |  11456 |  2558 |  61 |  49 |  687 |  3476 |  4625 | ⋯ | 1 |  806 | 0 | 0 | 0 |  806 | 0 | 1807 |  3217 | Control |\n",
       "| AAACGAATCGCAACTA-1 | Control | 6811 | 6272 |  56262 | 15793 | 214 | 374 | 3523 | 19210 | 17148 | ⋯ | 1 | 3042 | 0 | 0 | 0 | 3042 | 0 | 6449 | 11150 | Control |\n",
       "| AAACGAATCGCGCTGA-1 | Control |  621 |  614 |   6366 |   747 |  30 |  20 |  306 |  3521 |  1742 | ⋯ | 1 |  308 | 0 | 0 | 0 |  308 | 0 |  583 |  1076 | Control |\n",
       "| AAACTCGAGATTACGA-1 | Control | 1808 | 1763 |  22032 |  3419 |  83 |  85 | 1085 | 11844 |  5516 | ⋯ | 1 |  832 | 0 | 0 | 0 |  832 | 0 | 1710 |  3182 | Control |\n",
       "| AAACTCGCAAGACTTC-1 | Control |  409 |  402 |   5303 |   728 |   8 |  18 |  232 |  2874 |  1443 | ⋯ | 1 |  260 | 0 | 0 | 0 |  260 | 0 |  378 |   666 | Control |\n",
       "| AAACTCGCACAAGGGT-1 | Control |  352 |  336 |  18623 |   982 |   9 |  91 |  510 | 15503 |  1528 | ⋯ | 1 |  292 | 0 | 0 | 0 |  292 | 0 |  334 |   589 | Control |\n",
       "| AAACTCGCACCTCGTT-1 | Control | 3709 | 3499 |  30543 |  6297 | 148 | 239 | 1359 | 12689 |  9811 | ⋯ | 1 | 1707 | 0 | 0 | 0 | 1707 | 0 | 3507 |  6132 | Control |\n",
       "| AAACTCGCACTTGCGG-1 | Control |  646 |  631 |   7038 |   991 |  27 |  31 |  299 |  3787 |  1903 | ⋯ | 1 |  316 | 0 | 0 | 0 |  316 | 0 |  604 |  1102 | Control |\n",
       "| AAACTCGCATCGCCTT-1 | Control | 1514 | 1469 |  13537 |  3614 |  56 |  90 |  777 |  3050 |  5950 | ⋯ | 1 | 1098 | 0 | 0 | 0 | 1098 | 0 | 1378 |  2403 | Control |\n",
       "| AAACTCGCATGTGGGA-1 | Control | 3538 | 3349 |  48037 |  8932 | 171 | 241 | 1618 | 26241 | 10834 | ⋯ | 1 | 1726 | 0 | 0 | 0 | 1726 | 0 | 3277 |  5957 | Control |\n",
       "| AAACTCGGTATCCTTT-1 | Control | 2727 | 2525 |  21479 |  7157 | 101 | 159 | 1220 |  2182 | 10660 | ⋯ | 1 | 2145 | 0 | 0 | 0 | 2145 | 0 | 2456 |  4185 | Control |\n",
       "| AAACTCGGTGCCCTAG-1 | Control | 3607 | 3414 |  32698 |  7966 | 110 | 187 | 1479 | 13991 |  8965 | ⋯ | 1 | 1591 | 0 | 0 | 0 | 1591 | 0 | 3357 |  6010 | Control |\n",
       "| AAACTCGGTGTAGGAC-1 | Control |  530 |  517 |   3157 |   501 |   9 |  12 |  201 |   977 |  1457 | ⋯ | 1 |  196 | 0 | 0 | 0 |  196 | 0 |  494 |   854 | Control |\n",
       "| AAACTCGTCACTAGCA-1 | Control | 1628 | 1571 |  30764 |  4287 |  64 | 125 | 1018 | 19274 |  5996 | ⋯ | 1 |  838 | 0 | 0 | 0 |  838 | 0 | 1532 |  2755 | Control |\n",
       "| AAACTCGTCGAGTGTT-1 | Control | 2872 | 2715 |  29393 |  6514 | 144 | 188 | 1598 | 10615 | 10334 | ⋯ | 1 | 1903 | 0 | 0 | 0 | 1903 | 0 | 2632 |  4689 | Control |\n",
       "| AAACTCGTCTATCCTA-1 | Control | 2198 | 2111 |  18524 |  5154 | 104 |  82 | 1150 |  5499 |  6535 | ⋯ | 1 | 1001 | 0 | 0 | 0 | 1001 | 0 | 2046 |  3619 | Control |\n",
       "| AAACTGCAGCATGTGG-1 | Control | 3705 | 3475 |  25818 |  5609 | 136 | 114 | 1775 |  7066 | 11118 | ⋯ | 1 | 2326 | 0 | 0 | 0 | 2326 | 0 | 3505 |  6308 | Control |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| TTTGGCCGTGCGTAGA-1 | Control |  2638 |  2515 | 21042 |  4238 |  90 | 146 | 1400 |  7990 |  7178 | ⋯ | 1 | 1300 | 0 | 0 | 0 | 1300 | 0 |  2518 |  4400 | Control |\n",
       "| TTTGGCCGTTAGGCTT-1 | Control |  1615 |  1562 | 18757 |  3321 |  72 | 101 |  855 |  9021 |  5387 | ⋯ | 1 |  893 | 0 | 0 | 0 |  893 | 0 |  1498 |  2682 | Control |\n",
       "| TTTGGCCTCTGGCTAA-1 | Control |  3822 |  3614 | 44721 |  7325 | 135 | 203 | 1947 | 24374 | 10737 | ⋯ | 1 | 1884 | 0 | 0 | 0 | 1884 | 0 |  3653 |  6678 | Control |\n",
       "| TTTGGTTAGCCCATTA-1 | Control |   556 |   550 |  2698 |   581 |  28 |   8 |  113 |   699 |  1269 | ⋯ | 1 |  291 | 0 | 0 | 0 |  291 | 0 |   528 |   972 | Control |\n",
       "| TTTGGTTAGCGCCTAC-1 | Control |  1347 |  1304 | 18309 |  3837 |  93 |  80 |  766 |  8860 |  4673 | ⋯ | 1 |  668 | 0 | 0 | 0 |  668 | 0 |  1250 |  2237 | Control |\n",
       "| TTTGGTTCAAGAAACT-1 | Control |  1962 |  1904 | 25018 |  3679 | 101 | 115 | 1329 | 14418 |  5376 | ⋯ | 1 |  934 | 0 | 0 | 0 |  934 | 0 |  1846 |  3267 | Control |\n",
       "| TTTGGTTCACCCTTTG-1 | Control |   800 |   784 | 13165 |  1515 |  56 |  47 |  357 |  8522 |  2668 | ⋯ | 1 |  459 | 0 | 0 | 0 |  459 | 0 |   731 |  1344 | Control |\n",
       "| TTTGGTTCAGCAGGTA-1 | Control | 14293 | 12406 | 75730 | 26956 | 496 | 701 | 4313 |  5179 | 38085 | ⋯ | 1 | 6629 | 0 | 0 | 0 | 6629 | 0 | 13120 | 22372 | Control |\n",
       "| TTTGGTTCATGGATGG-1 | Control |  2718 |  2617 | 36709 |  6004 | 156 | 183 | 1688 | 20638 |  8040 | ⋯ | 1 | 1283 | 0 | 0 | 0 | 1283 | 0 |  2556 |  4712 | Control |\n",
       "| TTTGGTTGTATCTCAG-1 | Control |   504 |   497 |  4121 |   963 |  19 |  17 |  208 |  1271 |  1643 | ⋯ | 1 |  230 | 0 | 0 | 0 |  230 | 0 |   462 |   794 | Control |\n",
       "| TTTGGTTTCACTGATG-1 | Control |   960 |   935 | 19160 |  2190 |  46 |  80 |  454 | 13286 |  3104 | ⋯ | 1 |  469 | 0 | 0 | 0 |  469 | 0 |   908 |  1666 | Control |\n",
       "| TTTGGTTTCATCATGG-1 | Control |  2230 |  2164 | 24974 |  4120 | 118 | 111 | 1095 | 13520 |  6010 | ⋯ | 1 |  932 | 0 | 0 | 0 |  932 | 0 |  2087 |  3733 | Control |\n",
       "| TTTGGTTTCCGGACTG-1 | Control |  2069 |  1980 | 21336 |  4738 | 103 | 126 | 1252 |  8244 |  6873 | ⋯ | 1 | 1219 | 0 | 0 | 0 | 1219 | 0 |  1910 |  3316 | Control |\n",
       "| TTTGGTTTCCTGACCT-1 | Control |   323 |   318 |  4397 |   660 |  18 |  21 |  115 |  2462 |  1121 | ⋯ | 1 |  166 | 0 | 0 | 0 |  166 | 0 |   304 |   552 | Control |\n",
       "| TTTGGTTTCGATGAAA-1 | Control |   712 |   700 |  3150 |   652 |  27 |  15 |  171 |   771 |  1514 | ⋯ | 1 |  290 | 0 | 0 | 0 |  290 | 0 |   653 |  1169 | Control |\n",
       "| TTTGGTTTCGCGGCAT-1 | Control |  3282 |  3119 | 22039 |  5742 | 102 | 147 | 1276 |  6173 |  8599 | ⋯ | 1 | 1516 | 0 | 0 | 0 | 1516 | 0 |  3134 |  5582 | Control |\n",
       "| TTTGTGTAGACACGGT-1 | Control |   724 |   714 |  7099 |  1101 |  46 |  29 |  398 |  3271 |  2254 | ⋯ | 1 |  361 | 0 | 0 | 0 |  361 | 0 |   682 |  1200 | Control |\n",
       "| TTTGTGTAGACCGCAA-1 | Control |  2144 |  2066 | 23619 |  4038 |  75 |  97 |  803 | 12287 |  6319 | ⋯ | 1 | 1023 | 0 | 0 | 0 | 1023 | 0 |  2033 |  3623 | Control |\n",
       "| TTTGTGTAGACTAATG-1 | Control |  4521 |  4282 | 44991 |  7595 | 195 | 171 | 1749 | 24211 | 11070 | ⋯ | 1 | 2037 | 0 | 0 | 0 | 2037 | 0 |  4260 |  7865 | Control |\n",
       "| TTTGTGTAGACTTCCA-1 | Control |  1697 |  1646 | 13085 |  2533 |  57 |  49 |  583 |  5314 |  4549 | ⋯ | 1 |  959 | 0 | 0 | 0 |  959 | 0 |  1594 |  2889 | Control |\n",
       "| TTTGTGTAGCCAGAAC-1 | Control |   454 |   450 |  5687 |  1071 |  25 |  23 |  272 |  2237 |  2059 | ⋯ | 1 |  326 | 0 | 0 | 0 |  326 | 0 |   401 |   676 | Control |\n",
       "| TTTGTGTCAGCAACCC-1 | Control |  2173 |  2091 | 21754 |  4090 |  95 | 106 |  755 |  9631 |  7077 | ⋯ | 1 | 1245 | 0 | 0 | 0 | 1245 | 0 |  2005 |  3566 | Control |\n",
       "| TTTGTGTCATGTATCG-1 | Control |   803 |   793 |  5577 |  1166 |  34 |  30 |  305 |  1547 |  2495 | ⋯ | 1 |  403 | 0 | 0 | 0 |  403 | 0 |   742 |  1318 | Control |\n",
       "| TTTGTGTGTTTGCATG-1 | Control |  2345 |  2277 | 19469 |  3652 |  98 | 134 | 1140 |  5904 |  8541 | ⋯ | 1 | 1314 | 0 | 0 | 0 | 1314 | 0 |  2192 |  3746 | Control |\n",
       "| TTTGTGTTCACTACCC-1 | Control |  1061 |  1037 | 18676 |  1809 |  47 |  72 |  575 | 12724 |  3449 | ⋯ | 1 |  585 | 0 | 0 | 0 |  585 | 0 |   994 |  1848 | Control |\n",
       "| TTTGTGTTCCAAACCA-1 | Control |  8737 |  7785 | 90247 | 25919 | 402 | 477 | 5068 | 23688 | 34693 | ⋯ | 1 | 5402 | 0 | 0 | 0 | 5402 | 0 |  8125 | 13977 | Control |\n",
       "| TTTGTGTTCCGATGCG-1 | Control |  2518 |  2411 | 26526 |  3605 | 109 | 114 | 1089 | 14768 |  6841 | ⋯ | 1 | 1318 | 0 | 0 | 0 | 1318 | 0 |  2389 |  4370 | Control |\n",
       "| TTTGTGTTCCTAGAGT-1 | Control |   676 |   670 |  5693 |   758 |  29 |  30 |  220 |  2742 |  1914 | ⋯ | 1 |  322 | 0 | 0 | 0 |  322 | 0 |   639 |  1144 | Control |\n",
       "| TTTGTGTTCGCTCGGA-1 | Control |  2023 |  1952 | 19572 |  2975 |  79 |  77 | 1064 |  8534 |  6843 | ⋯ | 1 | 1240 | 0 | 0 | 0 | 1240 | 0 |  1859 |  3311 | Control |\n",
       "| TTTGTGTTCTTCCACG-1 | Control |  4312 |  4010 | 57305 | 12924 | 203 | 273 | 2172 | 28008 | 13725 | ⋯ | 1 | 2275 | 0 | 0 | 0 | 2275 | 0 |  4090 |  7264 | Control |\n",
       "\n"
      ],
      "text/plain": [
       "                   orig.ident nCount_ATAC nFeature_ATAC total  duplicate\n",
       "AAACGAAAGAGGTCCA-1 Control    1400        1357           16929  2587    \n",
       "AAACGAAAGCGATACG-1 Control    3692        3505           45359 10546    \n",
       "AAACGAAAGTCGTGAG-1 Control    1031         994           10177  2450    \n",
       "AAACGAACAATAGTGA-1 Control    2829        2687           22452  4364    \n",
       "AAACGAACACAATAAG-1 Control    1206        1184           11763  2003    \n",
       "AAACGAACACTGATAC-1 Control     968         954            9449  1971    \n",
       "AAACGAACAGTAACCT-1 Control     702         693            6196  1097    \n",
       "AAACGAACATCGTGAT-1 Control    7716        6800          100024 36721    \n",
       "AAACGAAGTAACGGTG-1 Control    1111        1074            8454  1950    \n",
       "AAACGAAGTAAGTCTC-1 Control    1273        1262            7684  1420    \n",
       "AAACGAAGTACCAAGG-1 Control    4175        3936           27865  7135    \n",
       "AAACGAAGTCCCGTGA-1 Control    4720        4425           31805  8689    \n",
       "AAACGAAGTGGATTCT-1 Control     813         805            9518  1121    \n",
       "AAACGAATCGAGCGCT-1 Control    1941        1892           11456  2558    \n",
       "AAACGAATCGCAACTA-1 Control    6811        6272           56262 15793    \n",
       "AAACGAATCGCGCTGA-1 Control     621         614            6366   747    \n",
       "AAACTCGAGATTACGA-1 Control    1808        1763           22032  3419    \n",
       "AAACTCGCAAGACTTC-1 Control     409         402            5303   728    \n",
       "AAACTCGCACAAGGGT-1 Control     352         336           18623   982    \n",
       "AAACTCGCACCTCGTT-1 Control    3709        3499           30543  6297    \n",
       "AAACTCGCACTTGCGG-1 Control     646         631            7038   991    \n",
       "AAACTCGCATCGCCTT-1 Control    1514        1469           13537  3614    \n",
       "AAACTCGCATGTGGGA-1 Control    3538        3349           48037  8932    \n",
       "AAACTCGGTATCCTTT-1 Control    2727        2525           21479  7157    \n",
       "AAACTCGGTGCCCTAG-1 Control    3607        3414           32698  7966    \n",
       "AAACTCGGTGTAGGAC-1 Control     530         517            3157   501    \n",
       "AAACTCGTCACTAGCA-1 Control    1628        1571           30764  4287    \n",
       "AAACTCGTCGAGTGTT-1 Control    2872        2715           29393  6514    \n",
       "AAACTCGTCTATCCTA-1 Control    2198        2111           18524  5154    \n",
       "AAACTGCAGCATGTGG-1 Control    3705        3475           25818  5609    \n",
       "⋮                  ⋮          ⋮           ⋮             ⋮      ⋮        \n",
       "TTTGGCCGTGCGTAGA-1 Control     2638        2515         21042   4238    \n",
       "TTTGGCCGTTAGGCTT-1 Control     1615        1562         18757   3321    \n",
       "TTTGGCCTCTGGCTAA-1 Control     3822        3614         44721   7325    \n",
       "TTTGGTTAGCCCATTA-1 Control      556         550          2698    581    \n",
       "TTTGGTTAGCGCCTAC-1 Control     1347        1304         18309   3837    \n",
       "TTTGGTTCAAGAAACT-1 Control     1962        1904         25018   3679    \n",
       "TTTGGTTCACCCTTTG-1 Control      800         784         13165   1515    \n",
       "TTTGGTTCAGCAGGTA-1 Control    14293       12406         75730  26956    \n",
       "TTTGGTTCATGGATGG-1 Control     2718        2617         36709   6004    \n",
       "TTTGGTTGTATCTCAG-1 Control      504         497          4121    963    \n",
       "TTTGGTTTCACTGATG-1 Control      960         935         19160   2190    \n",
       "TTTGGTTTCATCATGG-1 Control     2230        2164         24974   4120    \n",
       "TTTGGTTTCCGGACTG-1 Control     2069        1980         21336   4738    \n",
       "TTTGGTTTCCTGACCT-1 Control      323         318          4397    660    \n",
       "TTTGGTTTCGATGAAA-1 Control      712         700          3150    652    \n",
       "TTTGGTTTCGCGGCAT-1 Control     3282        3119         22039   5742    \n",
       "TTTGTGTAGACACGGT-1 Control      724         714          7099   1101    \n",
       "TTTGTGTAGACCGCAA-1 Control     2144        2066         23619   4038    \n",
       "TTTGTGTAGACTAATG-1 Control     4521        4282         44991   7595    \n",
       "TTTGTGTAGACTTCCA-1 Control     1697        1646         13085   2533    \n",
       "TTTGTGTAGCCAGAAC-1 Control      454         450          5687   1071    \n",
       "TTTGTGTCAGCAACCC-1 Control     2173        2091         21754   4090    \n",
       "TTTGTGTCATGTATCG-1 Control      803         793          5577   1166    \n",
       "TTTGTGTGTTTGCATG-1 Control     2345        2277         19469   3652    \n",
       "TTTGTGTTCACTACCC-1 Control     1061        1037         18676   1809    \n",
       "TTTGTGTTCCAAACCA-1 Control     8737        7785         90247  25919    \n",
       "TTTGTGTTCCGATGCG-1 Control     2518        2411         26526   3605    \n",
       "TTTGTGTTCCTAGAGT-1 Control      676         670          5693    758    \n",
       "TTTGTGTTCGCTCGGA-1 Control     2023        1952         19572   2975    \n",
       "TTTGTGTTCTTCCACG-1 Control     4312        4010         57305  12924    \n",
       "                   chimeric unmapped lowmapq mitochondrial passed_filters ⋯\n",
       "AAACGAAAGAGGTCCA-1  58       70       674     9629          3911          ⋯\n",
       "AAACGAAAGCGATACG-1 184      184      2302    22247          9896          ⋯\n",
       "AAACGAAAGTCGTGAG-1  48       60       657     2135          4827          ⋯\n",
       "AAACGAACAATAGTGA-1 100       83       897     9233          7775          ⋯\n",
       "AAACGAACACAATAAG-1  43       57       680     5547          3433          ⋯\n",
       "AAACGAACACTGATAC-1  59       21       489     3785          3124          ⋯\n",
       "AAACGAACAGTAACCT-1  19       17       288     2635          2140          ⋯\n",
       "AAACGAACATCGTGAT-1 588      531      5317    36828         20039          ⋯\n",
       "AAACGAAGTAACGGTG-1  24       33       354     3729          2364          ⋯\n",
       "AAACGAAGTAAGTCTC-1  36       23       488     1787          3930          ⋯\n",
       "AAACGAAGTACCAAGG-1 126      131      1330     9549          9594          ⋯\n",
       "AAACGAAGTCCCGTGA-1 178      135      1649     7738         13416          ⋯\n",
       "AAACGAAGTGGATTCT-1  32       37       330     5983          2015          ⋯\n",
       "AAACGAATCGAGCGCT-1  61       49       687     3476          4625          ⋯\n",
       "AAACGAATCGCAACTA-1 214      374      3523    19210         17148          ⋯\n",
       "AAACGAATCGCGCTGA-1  30       20       306     3521          1742          ⋯\n",
       "AAACTCGAGATTACGA-1  83       85      1085    11844          5516          ⋯\n",
       "AAACTCGCAAGACTTC-1   8       18       232     2874          1443          ⋯\n",
       "AAACTCGCACAAGGGT-1   9       91       510    15503          1528          ⋯\n",
       "AAACTCGCACCTCGTT-1 148      239      1359    12689          9811          ⋯\n",
       "AAACTCGCACTTGCGG-1  27       31       299     3787          1903          ⋯\n",
       "AAACTCGCATCGCCTT-1  56       90       777     3050          5950          ⋯\n",
       "AAACTCGCATGTGGGA-1 171      241      1618    26241         10834          ⋯\n",
       "AAACTCGGTATCCTTT-1 101      159      1220     2182         10660          ⋯\n",
       "AAACTCGGTGCCCTAG-1 110      187      1479    13991          8965          ⋯\n",
       "AAACTCGGTGTAGGAC-1   9       12       201      977          1457          ⋯\n",
       "AAACTCGTCACTAGCA-1  64      125      1018    19274          5996          ⋯\n",
       "AAACTCGTCGAGTGTT-1 144      188      1598    10615         10334          ⋯\n",
       "AAACTCGTCTATCCTA-1 104       82      1150     5499          6535          ⋯\n",
       "AAACTGCAGCATGTGG-1 136      114      1775     7066         11118          ⋯\n",
       "⋮                  ⋮        ⋮        ⋮       ⋮             ⋮              ⋱\n",
       "TTTGGCCGTGCGTAGA-1  90      146      1400     7990          7178          ⋯\n",
       "TTTGGCCGTTAGGCTT-1  72      101       855     9021          5387          ⋯\n",
       "TTTGGCCTCTGGCTAA-1 135      203      1947    24374         10737          ⋯\n",
       "TTTGGTTAGCCCATTA-1  28        8       113      699          1269          ⋯\n",
       "TTTGGTTAGCGCCTAC-1  93       80       766     8860          4673          ⋯\n",
       "TTTGGTTCAAGAAACT-1 101      115      1329    14418          5376          ⋯\n",
       "TTTGGTTCACCCTTTG-1  56       47       357     8522          2668          ⋯\n",
       "TTTGGTTCAGCAGGTA-1 496      701      4313     5179         38085          ⋯\n",
       "TTTGGTTCATGGATGG-1 156      183      1688    20638          8040          ⋯\n",
       "TTTGGTTGTATCTCAG-1  19       17       208     1271          1643          ⋯\n",
       "TTTGGTTTCACTGATG-1  46       80       454    13286          3104          ⋯\n",
       "TTTGGTTTCATCATGG-1 118      111      1095    13520          6010          ⋯\n",
       "TTTGGTTTCCGGACTG-1 103      126      1252     8244          6873          ⋯\n",
       "TTTGGTTTCCTGACCT-1  18       21       115     2462          1121          ⋯\n",
       "TTTGGTTTCGATGAAA-1  27       15       171      771          1514          ⋯\n",
       "TTTGGTTTCGCGGCAT-1 102      147      1276     6173          8599          ⋯\n",
       "TTTGTGTAGACACGGT-1  46       29       398     3271          2254          ⋯\n",
       "TTTGTGTAGACCGCAA-1  75       97       803    12287          6319          ⋯\n",
       "TTTGTGTAGACTAATG-1 195      171      1749    24211         11070          ⋯\n",
       "TTTGTGTAGACTTCCA-1  57       49       583     5314          4549          ⋯\n",
       "TTTGTGTAGCCAGAAC-1  25       23       272     2237          2059          ⋯\n",
       "TTTGTGTCAGCAACCC-1  95      106       755     9631          7077          ⋯\n",
       "TTTGTGTCATGTATCG-1  34       30       305     1547          2495          ⋯\n",
       "TTTGTGTGTTTGCATG-1  98      134      1140     5904          8541          ⋯\n",
       "TTTGTGTTCACTACCC-1  47       72       575    12724          3449          ⋯\n",
       "TTTGTGTTCCAAACCA-1 402      477      5068    23688         34693          ⋯\n",
       "TTTGTGTTCCGATGCG-1 109      114      1089    14768          6841          ⋯\n",
       "TTTGTGTTCCTAGAGT-1  29       30       220     2742          1914          ⋯\n",
       "TTTGTGTTCGCTCGGA-1  79       77      1064     8534          6843          ⋯\n",
       "TTTGTGTTCTTCCACG-1 203      273      2172    28008         13725          ⋯\n",
       "                   is__cell_barcode TSS_fragments\n",
       "AAACGAAAGAGGTCCA-1 1                 740         \n",
       "AAACGAAAGCGATACG-1 1                1611         \n",
       "AAACGAAAGTCGTGAG-1 1                 775         \n",
       "AAACGAACAATAGTGA-1 1                1239         \n",
       "AAACGAACACAATAAG-1 1                 621         \n",
       "AAACGAACACTGATAC-1 1                 517         \n",
       "AAACGAACAGTAACCT-1 1                 397         \n",
       "AAACGAACATCGTGAT-1 1                3467         \n",
       "AAACGAAGTAACGGTG-1 1                 451         \n",
       "AAACGAAGTAAGTCTC-1 1                 649         \n",
       "AAACGAAGTACCAAGG-1 1                1733         \n",
       "AAACGAAGTCCCGTGA-1 1                2181         \n",
       "AAACGAAGTGGATTCT-1 1                 414         \n",
       "AAACGAATCGAGCGCT-1 1                 806         \n",
       "AAACGAATCGCAACTA-1 1                3042         \n",
       "AAACGAATCGCGCTGA-1 1                 308         \n",
       "AAACTCGAGATTACGA-1 1                 832         \n",
       "AAACTCGCAAGACTTC-1 1                 260         \n",
       "AAACTCGCACAAGGGT-1 1                 292         \n",
       "AAACTCGCACCTCGTT-1 1                1707         \n",
       "AAACTCGCACTTGCGG-1 1                 316         \n",
       "AAACTCGCATCGCCTT-1 1                1098         \n",
       "AAACTCGCATGTGGGA-1 1                1726         \n",
       "AAACTCGGTATCCTTT-1 1                2145         \n",
       "AAACTCGGTGCCCTAG-1 1                1591         \n",
       "AAACTCGGTGTAGGAC-1 1                 196         \n",
       "AAACTCGTCACTAGCA-1 1                 838         \n",
       "AAACTCGTCGAGTGTT-1 1                1903         \n",
       "AAACTCGTCTATCCTA-1 1                1001         \n",
       "AAACTGCAGCATGTGG-1 1                2326         \n",
       "⋮                  ⋮                ⋮            \n",
       "TTTGGCCGTGCGTAGA-1 1                1300         \n",
       "TTTGGCCGTTAGGCTT-1 1                 893         \n",
       "TTTGGCCTCTGGCTAA-1 1                1884         \n",
       "TTTGGTTAGCCCATTA-1 1                 291         \n",
       "TTTGGTTAGCGCCTAC-1 1                 668         \n",
       "TTTGGTTCAAGAAACT-1 1                 934         \n",
       "TTTGGTTCACCCTTTG-1 1                 459         \n",
       "TTTGGTTCAGCAGGTA-1 1                6629         \n",
       "TTTGGTTCATGGATGG-1 1                1283         \n",
       "TTTGGTTGTATCTCAG-1 1                 230         \n",
       "TTTGGTTTCACTGATG-1 1                 469         \n",
       "TTTGGTTTCATCATGG-1 1                 932         \n",
       "TTTGGTTTCCGGACTG-1 1                1219         \n",
       "TTTGGTTTCCTGACCT-1 1                 166         \n",
       "TTTGGTTTCGATGAAA-1 1                 290         \n",
       "TTTGGTTTCGCGGCAT-1 1                1516         \n",
       "TTTGTGTAGACACGGT-1 1                 361         \n",
       "TTTGTGTAGACCGCAA-1 1                1023         \n",
       "TTTGTGTAGACTAATG-1 1                2037         \n",
       "TTTGTGTAGACTTCCA-1 1                 959         \n",
       "TTTGTGTAGCCAGAAC-1 1                 326         \n",
       "TTTGTGTCAGCAACCC-1 1                1245         \n",
       "TTTGTGTCATGTATCG-1 1                 403         \n",
       "TTTGTGTGTTTGCATG-1 1                1314         \n",
       "TTTGTGTTCACTACCC-1 1                 585         \n",
       "TTTGTGTTCCAAACCA-1 1                5402         \n",
       "TTTGTGTTCCGATGCG-1 1                1318         \n",
       "TTTGTGTTCCTAGAGT-1 1                 322         \n",
       "TTTGTGTTCGCTCGGA-1 1                1240         \n",
       "TTTGTGTTCTTCCACG-1 1                2275         \n",
       "                   DNase_sensitive_region_fragments enhancer_region_fragments\n",
       "AAACGAAAGAGGTCCA-1 0                                0                        \n",
       "AAACGAAAGCGATACG-1 0                                0                        \n",
       "AAACGAAAGTCGTGAG-1 0                                0                        \n",
       "AAACGAACAATAGTGA-1 0                                0                        \n",
       "AAACGAACACAATAAG-1 0                                0                        \n",
       "AAACGAACACTGATAC-1 0                                0                        \n",
       "AAACGAACAGTAACCT-1 0                                0                        \n",
       "AAACGAACATCGTGAT-1 0                                0                        \n",
       "AAACGAAGTAACGGTG-1 0                                0                        \n",
       "AAACGAAGTAAGTCTC-1 0                                0                        \n",
       "AAACGAAGTACCAAGG-1 0                                0                        \n",
       "AAACGAAGTCCCGTGA-1 0                                0                        \n",
       "AAACGAAGTGGATTCT-1 0                                0                        \n",
       "AAACGAATCGAGCGCT-1 0                                0                        \n",
       "AAACGAATCGCAACTA-1 0                                0                        \n",
       "AAACGAATCGCGCTGA-1 0                                0                        \n",
       "AAACTCGAGATTACGA-1 0                                0                        \n",
       "AAACTCGCAAGACTTC-1 0                                0                        \n",
       "AAACTCGCACAAGGGT-1 0                                0                        \n",
       "AAACTCGCACCTCGTT-1 0                                0                        \n",
       "AAACTCGCACTTGCGG-1 0                                0                        \n",
       "AAACTCGCATCGCCTT-1 0                                0                        \n",
       "AAACTCGCATGTGGGA-1 0                                0                        \n",
       "AAACTCGGTATCCTTT-1 0                                0                        \n",
       "AAACTCGGTGCCCTAG-1 0                                0                        \n",
       "AAACTCGGTGTAGGAC-1 0                                0                        \n",
       "AAACTCGTCACTAGCA-1 0                                0                        \n",
       "AAACTCGTCGAGTGTT-1 0                                0                        \n",
       "AAACTCGTCTATCCTA-1 0                                0                        \n",
       "AAACTGCAGCATGTGG-1 0                                0                        \n",
       "⋮                  ⋮                                ⋮                        \n",
       "TTTGGCCGTGCGTAGA-1 0                                0                        \n",
       "TTTGGCCGTTAGGCTT-1 0                                0                        \n",
       "TTTGGCCTCTGGCTAA-1 0                                0                        \n",
       "TTTGGTTAGCCCATTA-1 0                                0                        \n",
       "TTTGGTTAGCGCCTAC-1 0                                0                        \n",
       "TTTGGTTCAAGAAACT-1 0                                0                        \n",
       "TTTGGTTCACCCTTTG-1 0                                0                        \n",
       "TTTGGTTCAGCAGGTA-1 0                                0                        \n",
       "TTTGGTTCATGGATGG-1 0                                0                        \n",
       "TTTGGTTGTATCTCAG-1 0                                0                        \n",
       "TTTGGTTTCACTGATG-1 0                                0                        \n",
       "TTTGGTTTCATCATGG-1 0                                0                        \n",
       "TTTGGTTTCCGGACTG-1 0                                0                        \n",
       "TTTGGTTTCCTGACCT-1 0                                0                        \n",
       "TTTGGTTTCGATGAAA-1 0                                0                        \n",
       "TTTGGTTTCGCGGCAT-1 0                                0                        \n",
       "TTTGTGTAGACACGGT-1 0                                0                        \n",
       "TTTGTGTAGACCGCAA-1 0                                0                        \n",
       "TTTGTGTAGACTAATG-1 0                                0                        \n",
       "TTTGTGTAGACTTCCA-1 0                                0                        \n",
       "TTTGTGTAGCCAGAAC-1 0                                0                        \n",
       "TTTGTGTCAGCAACCC-1 0                                0                        \n",
       "TTTGTGTCATGTATCG-1 0                                0                        \n",
       "TTTGTGTGTTTGCATG-1 0                                0                        \n",
       "TTTGTGTTCACTACCC-1 0                                0                        \n",
       "TTTGTGTTCCAAACCA-1 0                                0                        \n",
       "TTTGTGTTCCGATGCG-1 0                                0                        \n",
       "TTTGTGTTCCTAGAGT-1 0                                0                        \n",
       "TTTGTGTTCGCTCGGA-1 0                                0                        \n",
       "TTTGTGTTCTTCCACG-1 0                                0                        \n",
       "                   promoter_region_fragments on_target_fragments\n",
       "AAACGAAAGAGGTCCA-1 0                          740               \n",
       "AAACGAAAGCGATACG-1 0                         1611               \n",
       "AAACGAAAGTCGTGAG-1 0                          775               \n",
       "AAACGAACAATAGTGA-1 0                         1239               \n",
       "AAACGAACACAATAAG-1 0                          621               \n",
       "AAACGAACACTGATAC-1 0                          517               \n",
       "AAACGAACAGTAACCT-1 0                          397               \n",
       "AAACGAACATCGTGAT-1 0                         3467               \n",
       "AAACGAAGTAACGGTG-1 0                          451               \n",
       "AAACGAAGTAAGTCTC-1 0                          649               \n",
       "AAACGAAGTACCAAGG-1 0                         1733               \n",
       "AAACGAAGTCCCGTGA-1 0                         2181               \n",
       "AAACGAAGTGGATTCT-1 0                          414               \n",
       "AAACGAATCGAGCGCT-1 0                          806               \n",
       "AAACGAATCGCAACTA-1 0                         3042               \n",
       "AAACGAATCGCGCTGA-1 0                          308               \n",
       "AAACTCGAGATTACGA-1 0                          832               \n",
       "AAACTCGCAAGACTTC-1 0                          260               \n",
       "AAACTCGCACAAGGGT-1 0                          292               \n",
       "AAACTCGCACCTCGTT-1 0                         1707               \n",
       "AAACTCGCACTTGCGG-1 0                          316               \n",
       "AAACTCGCATCGCCTT-1 0                         1098               \n",
       "AAACTCGCATGTGGGA-1 0                         1726               \n",
       "AAACTCGGTATCCTTT-1 0                         2145               \n",
       "AAACTCGGTGCCCTAG-1 0                         1591               \n",
       "AAACTCGGTGTAGGAC-1 0                          196               \n",
       "AAACTCGTCACTAGCA-1 0                          838               \n",
       "AAACTCGTCGAGTGTT-1 0                         1903               \n",
       "AAACTCGTCTATCCTA-1 0                         1001               \n",
       "AAACTGCAGCATGTGG-1 0                         2326               \n",
       "⋮                  ⋮                         ⋮                  \n",
       "TTTGGCCGTGCGTAGA-1 0                         1300               \n",
       "TTTGGCCGTTAGGCTT-1 0                          893               \n",
       "TTTGGCCTCTGGCTAA-1 0                         1884               \n",
       "TTTGGTTAGCCCATTA-1 0                          291               \n",
       "TTTGGTTAGCGCCTAC-1 0                          668               \n",
       "TTTGGTTCAAGAAACT-1 0                          934               \n",
       "TTTGGTTCACCCTTTG-1 0                          459               \n",
       "TTTGGTTCAGCAGGTA-1 0                         6629               \n",
       "TTTGGTTCATGGATGG-1 0                         1283               \n",
       "TTTGGTTGTATCTCAG-1 0                          230               \n",
       "TTTGGTTTCACTGATG-1 0                          469               \n",
       "TTTGGTTTCATCATGG-1 0                          932               \n",
       "TTTGGTTTCCGGACTG-1 0                         1219               \n",
       "TTTGGTTTCCTGACCT-1 0                          166               \n",
       "TTTGGTTTCGATGAAA-1 0                          290               \n",
       "TTTGGTTTCGCGGCAT-1 0                         1516               \n",
       "TTTGTGTAGACACGGT-1 0                          361               \n",
       "TTTGTGTAGACCGCAA-1 0                         1023               \n",
       "TTTGTGTAGACTAATG-1 0                         2037               \n",
       "TTTGTGTAGACTTCCA-1 0                          959               \n",
       "TTTGTGTAGCCAGAAC-1 0                          326               \n",
       "TTTGTGTCAGCAACCC-1 0                         1245               \n",
       "TTTGTGTCATGTATCG-1 0                          403               \n",
       "TTTGTGTGTTTGCATG-1 0                         1314               \n",
       "TTTGTGTTCACTACCC-1 0                          585               \n",
       "TTTGTGTTCCAAACCA-1 0                         5402               \n",
       "TTTGTGTTCCGATGCG-1 0                         1318               \n",
       "TTTGTGTTCCTAGAGT-1 0                          322               \n",
       "TTTGTGTTCGCTCGGA-1 0                         1240               \n",
       "TTTGTGTTCTTCCACG-1 0                         2275               \n",
       "                   blacklist_region_fragments peak_region_fragments\n",
       "AAACGAAAGAGGTCCA-1 0                          1330                 \n",
       "AAACGAAAGCGATACG-1 0                          3462                 \n",
       "AAACGAAAGTCGTGAG-1 0                           966                 \n",
       "AAACGAACAATAGTGA-1 0                          2701                 \n",
       "AAACGAACACAATAAG-1 0                          1138                 \n",
       "AAACGAACACTGATAC-1 0                           878                 \n",
       "AAACGAACAGTAACCT-1 0                           651                 \n",
       "AAACGAACATCGTGAT-1 0                          7263                 \n",
       "AAACGAAGTAACGGTG-1 0                          1042                 \n",
       "AAACGAAGTAAGTCTC-1 0                          1195                 \n",
       "AAACGAAGTACCAAGG-1 0                          3887                 \n",
       "AAACGAAGTCCCGTGA-1 0                          4387                 \n",
       "AAACGAAGTGGATTCT-1 0                           769                 \n",
       "AAACGAATCGAGCGCT-1 0                          1807                 \n",
       "AAACGAATCGCAACTA-1 0                          6449                 \n",
       "AAACGAATCGCGCTGA-1 0                           583                 \n",
       "AAACTCGAGATTACGA-1 0                          1710                 \n",
       "AAACTCGCAAGACTTC-1 0                           378                 \n",
       "AAACTCGCACAAGGGT-1 0                           334                 \n",
       "AAACTCGCACCTCGTT-1 0                          3507                 \n",
       "AAACTCGCACTTGCGG-1 0                           604                 \n",
       "AAACTCGCATCGCCTT-1 0                          1378                 \n",
       "AAACTCGCATGTGGGA-1 0                          3277                 \n",
       "AAACTCGGTATCCTTT-1 0                          2456                 \n",
       "AAACTCGGTGCCCTAG-1 0                          3357                 \n",
       "AAACTCGGTGTAGGAC-1 0                           494                 \n",
       "AAACTCGTCACTAGCA-1 0                          1532                 \n",
       "AAACTCGTCGAGTGTT-1 0                          2632                 \n",
       "AAACTCGTCTATCCTA-1 0                          2046                 \n",
       "AAACTGCAGCATGTGG-1 0                          3505                 \n",
       "⋮                  ⋮                          ⋮                    \n",
       "TTTGGCCGTGCGTAGA-1 0                           2518                \n",
       "TTTGGCCGTTAGGCTT-1 0                           1498                \n",
       "TTTGGCCTCTGGCTAA-1 0                           3653                \n",
       "TTTGGTTAGCCCATTA-1 0                            528                \n",
       "TTTGGTTAGCGCCTAC-1 0                           1250                \n",
       "TTTGGTTCAAGAAACT-1 0                           1846                \n",
       "TTTGGTTCACCCTTTG-1 0                            731                \n",
       "TTTGGTTCAGCAGGTA-1 0                          13120                \n",
       "TTTGGTTCATGGATGG-1 0                           2556                \n",
       "TTTGGTTGTATCTCAG-1 0                            462                \n",
       "TTTGGTTTCACTGATG-1 0                            908                \n",
       "TTTGGTTTCATCATGG-1 0                           2087                \n",
       "TTTGGTTTCCGGACTG-1 0                           1910                \n",
       "TTTGGTTTCCTGACCT-1 0                            304                \n",
       "TTTGGTTTCGATGAAA-1 0                            653                \n",
       "TTTGGTTTCGCGGCAT-1 0                           3134                \n",
       "TTTGTGTAGACACGGT-1 0                            682                \n",
       "TTTGTGTAGACCGCAA-1 0                           2033                \n",
       "TTTGTGTAGACTAATG-1 0                           4260                \n",
       "TTTGTGTAGACTTCCA-1 0                           1594                \n",
       "TTTGTGTAGCCAGAAC-1 0                            401                \n",
       "TTTGTGTCAGCAACCC-1 0                           2005                \n",
       "TTTGTGTCATGTATCG-1 0                            742                \n",
       "TTTGTGTGTTTGCATG-1 0                           2192                \n",
       "TTTGTGTTCACTACCC-1 0                            994                \n",
       "TTTGTGTTCCAAACCA-1 0                           8125                \n",
       "TTTGTGTTCCGATGCG-1 0                           2389                \n",
       "TTTGTGTTCCTAGAGT-1 0                            639                \n",
       "TTTGTGTTCGCTCGGA-1 0                           1859                \n",
       "TTTGTGTTCTTCCACG-1 0                           4090                \n",
       "                   peak_region_cutsites proj   \n",
       "AAACGAAAGAGGTCCA-1  2434                Control\n",
       "AAACGAAAGCGATACG-1  6167                Control\n",
       "AAACGAAAGTCGTGAG-1  1768                Control\n",
       "AAACGAACAATAGTGA-1  4951                Control\n",
       "AAACGAACACAATAAG-1  2019                Control\n",
       "AAACGAACACTGATAC-1  1595                Control\n",
       "AAACGAACAGTAACCT-1  1170                Control\n",
       "AAACGAACATCGTGAT-1 13261                Control\n",
       "AAACGAAGTAACGGTG-1  1920                Control\n",
       "AAACGAAGTAAGTCTC-1  2115                Control\n",
       "AAACGAAGTACCAAGG-1  7139                Control\n",
       "AAACGAAGTCCCGTGA-1  7857                Control\n",
       "AAACGAAGTGGATTCT-1  1431                Control\n",
       "AAACGAATCGAGCGCT-1  3217                Control\n",
       "AAACGAATCGCAACTA-1 11150                Control\n",
       "AAACGAATCGCGCTGA-1  1076                Control\n",
       "AAACTCGAGATTACGA-1  3182                Control\n",
       "AAACTCGCAAGACTTC-1   666                Control\n",
       "AAACTCGCACAAGGGT-1   589                Control\n",
       "AAACTCGCACCTCGTT-1  6132                Control\n",
       "AAACTCGCACTTGCGG-1  1102                Control\n",
       "AAACTCGCATCGCCTT-1  2403                Control\n",
       "AAACTCGCATGTGGGA-1  5957                Control\n",
       "AAACTCGGTATCCTTT-1  4185                Control\n",
       "AAACTCGGTGCCCTAG-1  6010                Control\n",
       "AAACTCGGTGTAGGAC-1   854                Control\n",
       "AAACTCGTCACTAGCA-1  2755                Control\n",
       "AAACTCGTCGAGTGTT-1  4689                Control\n",
       "AAACTCGTCTATCCTA-1  3619                Control\n",
       "AAACTGCAGCATGTGG-1  6308                Control\n",
       "⋮                  ⋮                    ⋮      \n",
       "TTTGGCCGTGCGTAGA-1  4400                Control\n",
       "TTTGGCCGTTAGGCTT-1  2682                Control\n",
       "TTTGGCCTCTGGCTAA-1  6678                Control\n",
       "TTTGGTTAGCCCATTA-1   972                Control\n",
       "TTTGGTTAGCGCCTAC-1  2237                Control\n",
       "TTTGGTTCAAGAAACT-1  3267                Control\n",
       "TTTGGTTCACCCTTTG-1  1344                Control\n",
       "TTTGGTTCAGCAGGTA-1 22372                Control\n",
       "TTTGGTTCATGGATGG-1  4712                Control\n",
       "TTTGGTTGTATCTCAG-1   794                Control\n",
       "TTTGGTTTCACTGATG-1  1666                Control\n",
       "TTTGGTTTCATCATGG-1  3733                Control\n",
       "TTTGGTTTCCGGACTG-1  3316                Control\n",
       "TTTGGTTTCCTGACCT-1   552                Control\n",
       "TTTGGTTTCGATGAAA-1  1169                Control\n",
       "TTTGGTTTCGCGGCAT-1  5582                Control\n",
       "TTTGTGTAGACACGGT-1  1200                Control\n",
       "TTTGTGTAGACCGCAA-1  3623                Control\n",
       "TTTGTGTAGACTAATG-1  7865                Control\n",
       "TTTGTGTAGACTTCCA-1  2889                Control\n",
       "TTTGTGTAGCCAGAAC-1   676                Control\n",
       "TTTGTGTCAGCAACCC-1  3566                Control\n",
       "TTTGTGTCATGTATCG-1  1318                Control\n",
       "TTTGTGTGTTTGCATG-1  3746                Control\n",
       "TTTGTGTTCACTACCC-1  1848                Control\n",
       "TTTGTGTTCCAAACCA-1 13977                Control\n",
       "TTTGTGTTCCGATGCG-1  4370                Control\n",
       "TTTGTGTTCCTAGAGT-1  1144                Control\n",
       "TTTGTGTTCGCTCGGA-1  3311                Control\n",
       "TTTGTGTTCTTCCACG-1  7264                Control"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allSE[[1]][[]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d7bff2c",
   "metadata": {},
   "source": [
    "## QC Metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b6ddf734",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Rsamtools)\n",
    "\n",
    "qc <- function(se){\n",
    "    # extract gene annotations from EnsDb\n",
    "    annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)\n",
    "\n",
    "    # change to UCSC style since the data was mapped to hg19\n",
    "    seqlevelsStyle(annotations) <- 'NCBI'\n",
    "    genome(annotations) <- \"hg38\"\n",
    "    # add the gene information to the object\n",
    "    Annotation(se) <- annotations\n",
    "    gene.activities <- GeneActivity(se)\n",
    "    # compute nucleosome signal score per cell\n",
    "    se <- NucleosomeSignal(object = se)\n",
    "    \n",
    "    Annotation(se) <- annotations\n",
    "    # compute TSS enrichment score per cell\n",
    "    se <- TSSEnrichment(object = se, fast = FALSE)\n",
    "\n",
    "    # add blacklist ratio and fraction of reads in peaks\n",
    "    se$pct_reads_in_peaks <- se$peak_region_fragments / se$passed_filters * 100\n",
    "    se$blacklist_ratio <- se$blacklist_region_fragments / se$peak_region_fragments\n",
    "    se$high.tss <- ifelse(se$TSS.enrichment > 2, 'High', 'Low')\n",
    "    se$nucleosome_group <- ifelse(se$nucleosome_signal > 4, 'NS > 4', 'NS < 4')\n",
    "\n",
    "    return(se)\n",
    "}\n",
    "vPlot <- function(se){\n",
    "      vPlot <- VlnPlot(\n",
    "      object = se,\n",
    "      features = c('pct_reads_in_peaks', 'peak_region_fragments',\n",
    "                   'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal'),\n",
    "      pt.size = 0.1,\n",
    "      ncol = 5\n",
    "    )  \n",
    "    vPlot <- vPlot +    # Create grid of plots with title\n",
    "             plot_annotation(title = se$orig.ident[[1]]) & \n",
    "             theme(plot.title = element_text(hjust = 0.5, size=15))\n",
    "    #print(vPlot)\n",
    "    return(vPlot)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a74a3529",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n"
     ]
    }
   ],
   "source": [
    "allSE <- lapply(allSE, qc)\n",
    "saveRDS(allSE, file.path(outdir, paste0(\"allSamples.rds\")))\n",
    "lapply(allSE,vPlot)\n",
    "ggsave(file.path(outdir, paste0(\"QC_01.png\")))\n",
    "ggsave(file.path(outdir, paste0(\"QC_01.pdf\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "626a0538",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
