{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "223d6f46",
   "metadata": {},
   "source": [
    "## Creates binarized Seurat objects, integrates conditions and annotates genes by nearby peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "663e2d4d",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# Input info\n",
    "cellr_in = \"/data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/\" \n",
    "outdir =  \"/data2/mito_lineage/output/annotation/data/CHIP_april08_2021/MTblacklist/mergedSamples\" \n",
    "sample_names = \"Control,Flt3l,Input\"\n",
    "samples = \"Control,Flt3l,Input\"\n",
    "\n",
    "# Parameters\n",
    "nTop = 25000\n",
    "cores = 24"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b62b75ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(repr)\n",
    "options(repr.plot.width=12, repr.plot.height=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c2c99fc4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Control'</li><li>'Flt3l'</li><li>'Input'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Control'\n",
       "\\item 'Flt3l'\n",
       "\\item 'Input'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Control'\n",
       "2. 'Flt3l'\n",
       "3. 'Input'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"Control\" \"Flt3l\"   \"Input\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "samples <- unlist(strsplit(samples, \",\")[[1]])\n",
    "sample_names <- unlist(strsplit(sample_names, \",\"))\n",
    "\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aad4f638",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'data.table'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'magrittr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    not\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'cowplot'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:patchwork':\n",
      "\n",
      "    align_plots\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ensembldb':\n",
      "\n",
      "    filter, select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'future'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    value\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"Strategy 'multiprocess' is deprecated in future (>= 1.20.0). Instead, explicitly specify either 'multisession' or 'multicore'. In the current R session, 'multiprocess' equals 'multicore'.\"\n"
     ]
    }
   ],
   "source": [
    "library(GenomicRanges)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(GenomeInfoDb)\n",
    "library(EnsDb.Hsapiens.v75)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "set.seed(1234)\n",
    "library(data.table)\n",
    "library(magrittr)\n",
    "library(cowplot)\n",
    "library(metap)\n",
    "library(dplyr)\n",
    "library(future)\n",
    "\n",
    "plan(\"multiprocess\", workers = cores)\n",
    "options(future.globals.maxSize = 8000 * 1024^2)\n",
    "#options(future.globals.maxSize = 50000 * 1024^2) # for 50 Gb RAM\n",
    "#plan(\"multiprocess\", workers = workers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a53a3256",
   "metadata": {},
   "source": [
    "## Merge all peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "62849b8a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"here\"\n",
      "[1] \"/data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist//Control/outs/filtered_peak_bc_matrix/peaks.bed\"\n",
      "[1] \"here\"\n",
      "[1] \"/data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist//Flt3l/outs/filtered_peak_bc_matrix/peaks.bed\"\n",
      "[1] \"here\"\n",
      "[1] \"/data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist//Input/outs/filtered_peak_bc_matrix/peaks.bed\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 263336 ranges and 0 metadata columns:\n",
       "           seqnames            ranges strand\n",
       "              <Rle>         <IRanges>  <Rle>\n",
       "       [1]     chr1        9872-10620      *\n",
       "       [2]     chr1       96454-96767      *\n",
       "       [3]     chr1     180638-181142      *\n",
       "       [4]     chr1     191518-191976      *\n",
       "       [5]     chr1     267762-268278      *\n",
       "       ...      ...               ...    ...\n",
       "  [263332]     chrY 56872076-56872403      *\n",
       "  [263333]     chrY 56873720-56873847      *\n",
       "  [263334]     chrY 56873946-56874635      *\n",
       "  [263335]     chrY 56879619-56879842      *\n",
       "  [263336]     chrY 56880257-56880611      *\n",
       "  -------\n",
       "  seqinfo: 24 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "read.peaks <- function(exp, cellr_in){\n",
    "    print('here')\n",
    "    print(file.path(cellr_in, exp, \"outs\", \"filtered_peak_bc_matrix\", \"peaks.bed\"))\n",
    "    peaks <- read.table(\n",
    "      file = file.path(cellr_in, exp, \"outs\", \"filtered_peak_bc_matrix\", \"peaks.bed\"),\n",
    "      col.names = c(\"chr\", \"start\", \"end\")\n",
    "    )\n",
    "    # convert to genomic ranges\n",
    "    gr <- makeGRangesFromDataFrame(peaks)\n",
    "    return(gr)\n",
    "}\n",
    "\n",
    "\n",
    "gr.full <- c(sapply(samples, read.peaks, cellr_in=cellr_in, USE.NAMES=F))\n",
    "\n",
    "gr.full.c <- gr.full[[1]]\n",
    "if (length(gr.full)>1){\n",
    "    for (i in 2:length(gr.full)){\n",
    "      gr.full.c <- c(gr.full.c, gr.full[[i]])\n",
    "    }\n",
    "}\n",
    "combined.peaks <- reduce(x = c(gr.full.c))\n",
    "\n",
    "# Filter out bad peaks based on length\n",
    "peakwidths <- width(combined.peaks)\n",
    "combined.peaks <- combined.peaks[peakwidths  < 10000 & peakwidths > 20]\n",
    "combined.peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cae010c",
   "metadata": {},
   "source": [
    "## Create fragment objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "67e92e33",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"samples_df\"\n",
      "     sample_names samples  \n",
      "[1,] \"Control\"    \"Control\"\n",
      "[2,] \"Flt3l\"      \"Flt3l\"  \n",
      "[3,] \"Input\"      \"Input\"  \n",
      "[1] \"exp\"\n",
      "[1] \"Control\"\n",
      "[1] \"name\"\n",
      "[1] \"Control\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'cli':\n",
      "  method     from         \n",
      "  print.boxx spatstat.geom\n",
      "\u001b[1m\u001b[1mRows: \u001b[1m\u001b[22m\u001b[34m\u001b[34m843\u001b[34m\u001b[39m \u001b[1m\u001b[1mColumns: \u001b[1m\u001b[22m\u001b[34m\u001b[34m1\u001b[34m\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): X1\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n",
      "\u001b[1m\u001b[1mRows: \u001b[1m\u001b[22m\u001b[34m\u001b[34m470007\u001b[34m\u001b[39m \u001b[1m\u001b[1mColumns: \u001b[1m\u001b[22m\u001b[34m\u001b[34m18\u001b[34m\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m  (2): barcode, cell_id\n",
      "\u001b[32mdbl\u001b[39m (16): total, duplicate, chimeric, unmapped, lowmapq, mitochondrial, pass...\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating fragments object\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Quantifying peaks\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating chromatin assay\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CreateSeuratObject.Assay(curr_assay, assay = \"ATAC\", project = name, :\n",
      "\"Some cells in meta.data not present in provided counts matrix.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"curr_assay\"\n",
      "data frame with 0 columns and 10 rows\n",
      "[1] \"curr\"\n",
      "                   orig.ident nCount_ATAC nFeature_ATAC  total duplicate\n",
      "AAACGAATCTTACTCA-1    Control         215           215    779        47\n",
      "AAACTCGAGTGGTGTG-1    Control         390           387  16911      5224\n",
      "AAACTCGTCAATGCAC-1    Control        1030          1012  11963      2864\n",
      "AAACTGCTCTCGCGGA-1    Control       21843         15140 265035     83694\n",
      "AAAGATGAGCGCATTT-1    Control         373           368  15367      5530\n",
      "AAAGATGCAGCAACGA-1    Control       15235         10565 250347     85267\n",
      "                   chimeric unmapped lowmapq mitochondrial passed_filters\n",
      "AAACGAATCTTACTCA-1        6        7      37           299            383\n",
      "AAACTCGAGTGGTGTG-1       95      356     690          9717            829\n",
      "AAACTCGTCAATGCAC-1       63       78     583          6889           1486\n",
      "AAACTGCTCTCGCGGA-1     1575     1101    9756        141471          27438\n",
      "AAAGATGAGCGCATTT-1       83      294     733          8051            676\n",
      "AAAGATGCAGCAACGA-1     1663     1230    6598        134365          21224\n",
      "                   cell_id is__cell_barcode TSS_fragments\n",
      "AAACGAATCTTACTCA-1 _cell_0                1           131\n",
      "AAACTCGAGTGGTGTG-1 _cell_1                1           267\n",
      "AAACTCGTCAATGCAC-1 _cell_2                1           638\n",
      "AAACTGCTCTCGCGGA-1 _cell_3                1         12690\n",
      "AAAGATGAGCGCATTT-1 _cell_4                1           265\n",
      "AAAGATGCAGCAACGA-1 _cell_5                1          9560\n",
      "                   DNase_sensitive_region_fragments enhancer_region_fragments\n",
      "AAACGAATCTTACTCA-1                                0                         0\n",
      "AAACTCGAGTGGTGTG-1                                0                         0\n",
      "AAACTCGTCAATGCAC-1                                0                         0\n",
      "AAACTGCTCTCGCGGA-1                                0                         0\n",
      "AAAGATGAGCGCATTT-1                                0                         0\n",
      "AAAGATGCAGCAACGA-1                                0                         0\n",
      "                   promoter_region_fragments on_target_fragments\n",
      "AAACGAATCTTACTCA-1                         0                 131\n",
      "AAACTCGAGTGGTGTG-1                         0                 267\n",
      "AAACTCGTCAATGCAC-1                         0                 638\n",
      "AAACTGCTCTCGCGGA-1                         0               12690\n",
      "AAAGATGAGCGCATTT-1                         0                 265\n",
      "AAAGATGCAGCAACGA-1                         0                9560\n",
      "                   blacklist_region_fragments peak_region_fragments\n",
      "AAACGAATCTTACTCA-1                          0                   196\n",
      "AAACTCGAGTGGTGTG-1                          0                   383\n",
      "AAACTCGTCAATGCAC-1                          0                   996\n",
      "AAACTGCTCTCGCGGA-1                          0                 21401\n",
      "AAAGATGAGCGCATTT-1                          0                   366\n",
      "AAAGATGCAGCAACGA-1                          0                 14580\n",
      "                   peak_region_cutsites    proj\n",
      "AAACGAATCTTACTCA-1                  372 Control\n",
      "AAACTCGAGTGGTGTG-1                  751 Control\n",
      "AAACTCGTCAATGCAC-1                 1923 Control\n",
      "AAACTGCTCTCGCGGA-1                41723 Control\n",
      "AAAGATGAGCGCATTT-1                  718 Control\n",
      "AAAGATGCAGCAACGA-1                28399 Control\n",
      "[1] \"exp\"\n",
      "[1] \"Flt3l\"\n",
      "[1] \"name\"\n",
      "[1] \"Flt3l\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[1mRows: \u001b[1m\u001b[22m\u001b[34m\u001b[34m5313\u001b[34m\u001b[39m \u001b[1m\u001b[1mColumns: \u001b[1m\u001b[22m\u001b[34m\u001b[34m1\u001b[34m\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): X1\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n",
      "\u001b[1m\u001b[1mRows: \u001b[1m\u001b[22m\u001b[34m\u001b[34m532237\u001b[34m\u001b[39m \u001b[1m\u001b[1mColumns: \u001b[1m\u001b[22m\u001b[34m\u001b[34m18\u001b[34m\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m  (2): barcode, cell_id\n",
      "\u001b[32mdbl\u001b[39m (16): total, duplicate, chimeric, unmapped, lowmapq, mitochondrial, pass...\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating fragments object\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Quantifying peaks\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating chromatin assay\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CreateSeuratObject.Assay(curr_assay, assay = \"ATAC\", project = name, :\n",
      "\"Some cells in meta.data not present in provided counts matrix.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"curr_assay\"\n",
      "data frame with 0 columns and 10 rows\n",
      "[1] \"curr\"\n",
      "                   orig.ident nCount_ATAC nFeature_ATAC total duplicate\n",
      "AAACGAAAGGAAACTT-1      Flt3l        3108          2935 49701      5002\n",
      "AAACGAAGTCCAAGAG-1      Flt3l        5131          4737 62869      9096\n",
      "AAACGAAGTCCGAGCT-1      Flt3l        8403          7462 60574     11583\n",
      "AAACGAAGTTCAGAAA-1      Flt3l        3095          2948 32202      4591\n",
      "AAACGAAGTTTGAAGA-1      Flt3l        7979          7225 68829     12010\n",
      "AAACGAATCAGCCGGT-1      Flt3l        1251          1223 13222      1216\n",
      "                   chimeric unmapped lowmapq mitochondrial passed_filters\n",
      "AAACGAAAGGAAACTT-1      164       92    1138         37812           5493\n",
      "AAACGAAGTCCAAGAG-1      249      139    1902         42086           9397\n",
      "AAACGAAGTCCGAGCT-1      352      141    2087         31767          14644\n",
      "AAACGAAGTTCAGAAA-1      146       90     911         19849           6615\n",
      "AAACGAAGTTTGAAGA-1      432      163    2209         39985          14030\n",
      "AAACGAATCAGCCGGT-1       85       25     369          9630           1897\n",
      "                   cell_id is__cell_barcode TSS_fragments\n",
      "AAACGAAAGGAAACTT-1 _cell_0                1          1489\n",
      "AAACGAAGTCCAAGAG-1 _cell_1                1          2713\n",
      "AAACGAAGTCCGAGCT-1 _cell_2                1          4090\n",
      "AAACGAAGTTCAGAAA-1 _cell_3                1          1483\n",
      "AAACGAAGTTTGAAGA-1 _cell_4                1          3862\n",
      "AAACGAATCAGCCGGT-1 _cell_5                1           587\n",
      "                   DNase_sensitive_region_fragments enhancer_region_fragments\n",
      "AAACGAAAGGAAACTT-1                                0                         0\n",
      "AAACGAAGTCCAAGAG-1                                0                         0\n",
      "AAACGAAGTCCGAGCT-1                                0                         0\n",
      "AAACGAAGTTCAGAAA-1                                0                         0\n",
      "AAACGAAGTTTGAAGA-1                                0                         0\n",
      "AAACGAATCAGCCGGT-1                                0                         0\n",
      "                   promoter_region_fragments on_target_fragments\n",
      "AAACGAAAGGAAACTT-1                         0                1489\n",
      "AAACGAAGTCCAAGAG-1                         0                2713\n",
      "AAACGAAGTCCGAGCT-1                         0                4090\n",
      "AAACGAAGTTCAGAAA-1                         0                1483\n",
      "AAACGAAGTTTGAAGA-1                         0                3862\n",
      "AAACGAATCAGCCGGT-1                         0                 587\n",
      "                   blacklist_region_fragments peak_region_fragments\n",
      "AAACGAAAGGAAACTT-1                          0                  2969\n",
      "AAACGAAGTCCAAGAG-1                          0                  4893\n",
      "AAACGAAGTCCGAGCT-1                          0                  8076\n",
      "AAACGAAGTTCAGAAA-1                          0                  2971\n",
      "AAACGAAGTTTGAAGA-1                          0                  7640\n",
      "AAACGAATCAGCCGGT-1                          0                  1208\n",
      "                   peak_region_cutsites  proj\n",
      "AAACGAAAGGAAACTT-1                 5761 Flt3l\n",
      "AAACGAAGTCCAAGAG-1                 9514 Flt3l\n",
      "AAACGAAGTCCGAGCT-1                15616 Flt3l\n",
      "AAACGAAGTTCAGAAA-1                 5760 Flt3l\n",
      "AAACGAAGTTTGAAGA-1                14850 Flt3l\n",
      "AAACGAATCAGCCGGT-1                 2350 Flt3l\n",
      "[1] \"exp\"\n",
      "[1] \"Input\"\n",
      "[1] \"name\"\n",
      "[1] \"Input\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[1mRows: \u001b[1m\u001b[22m\u001b[34m\u001b[34m222\u001b[34m\u001b[39m \u001b[1m\u001b[1mColumns: \u001b[1m\u001b[22m\u001b[34m\u001b[34m1\u001b[34m\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): X1\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n",
      "\u001b[1m\u001b[1mRows: \u001b[1m\u001b[22m\u001b[34m\u001b[34m361348\u001b[34m\u001b[39m \u001b[1m\u001b[1mColumns: \u001b[1m\u001b[22m\u001b[34m\u001b[34m18\u001b[34m\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m  (2): barcode, cell_id\n",
      "\u001b[32mdbl\u001b[39m (16): total, duplicate, chimeric, unmapped, lowmapq, mitochondrial, pass...\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating fragments object\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Quantifying peaks\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating chromatin assay\"\n",
      "[1] \"curr_assay\"\n",
      "data frame with 0 columns and 10 rows\n",
      "[1] \"curr\"\n",
      "                   orig.ident nCount_ATAC nFeature_ATAC   total duplicate\n",
      "AAACGAATCTTACTCA-1      Input       20967         14540 1136869    627713\n",
      "AAAGGATAGTCGTGAG-1      Input       34012         21373 1065254    711897\n",
      "AAATGAGGTCAATACC-1      Input       23267         17558  408417    114699\n",
      "AAATGCCAGGAACCGT-1      Input        6184          5352  312883     92752\n",
      "AAATGCCCACCGAAAG-1      Input       33799         22561  582430    187426\n",
      "AACAAAGCAATTCAGC-1      Input       16096         12306  922273    390846\n",
      "                   chimeric unmapped lowmapq mitochondrial passed_filters\n",
      "AAACGAATCTTACTCA-1     5110     3785   53071        404558          42632\n",
      "AAAGGATAGTCGTGAG-1     5461     3619   57926        218077          68274\n",
      "AAATGAGGTCAATACC-1     1725     1858   17193        216353          56589\n",
      "AAATGCCAGGAACCGT-1     1351     1566   11187        190764          15263\n",
      "AAATGCCCACCGAAAG-1     2379     2668   27520        277501          84936\n",
      "AACAAAGCAATTCAGC-1     4689     2401   37581        450342          36414\n",
      "                   cell_id is__cell_barcode TSS_fragments\n",
      "AAACGAATCTTACTCA-1 _cell_0                1         12571\n",
      "AAAGGATAGTCGTGAG-1 _cell_1                1         20378\n",
      "AAATGAGGTCAATACC-1 _cell_2                1         13821\n",
      "AAATGCCAGGAACCGT-1 _cell_3                1          3684\n",
      "AAATGCCCACCGAAAG-1 _cell_4                1         22252\n",
      "AACAAAGCAATTCAGC-1 _cell_5                1          8656\n",
      "                   DNase_sensitive_region_fragments enhancer_region_fragments\n",
      "AAACGAATCTTACTCA-1                                0                         0\n",
      "AAAGGATAGTCGTGAG-1                                0                         0\n",
      "AAATGAGGTCAATACC-1                                0                         0\n",
      "AAATGCCAGGAACCGT-1                                0                         0\n",
      "AAATGCCCACCGAAAG-1                                0                         0\n",
      "AACAAAGCAATTCAGC-1                                0                         0\n",
      "                   promoter_region_fragments on_target_fragments\n",
      "AAACGAATCTTACTCA-1                         0               12571\n",
      "AAAGGATAGTCGTGAG-1                         0               20378\n",
      "AAATGAGGTCAATACC-1                         0               13821\n",
      "AAATGCCAGGAACCGT-1                         0                3684\n",
      "AAATGCCCACCGAAAG-1                         0               22252\n",
      "AACAAAGCAATTCAGC-1                         0                8656\n",
      "                   blacklist_region_fragments peak_region_fragments\n",
      "AAACGAATCTTACTCA-1                          0                 24988\n",
      "AAAGGATAGTCGTGAG-1                          0                 40005\n",
      "AAATGAGGTCAATACC-1                          0                 27927\n",
      "AAATGCCAGGAACCGT-1                          0                  7424\n",
      "AAATGCCCACCGAAAG-1                          0                 40800\n",
      "AACAAAGCAATTCAGC-1                          0                 19227\n",
      "                   peak_region_cutsites  proj\n",
      "AAACGAATCTTACTCA-1                46372 Input\n",
      "AAAGGATAGTCGTGAG-1                73483 Input\n",
      "AAATGAGGTCAATACC-1                52095 Input\n",
      "AAATGCCAGGAACCGT-1                14091 Input\n",
      "AAATGCCCACCGAAAG-1                75773 Input\n",
      "AACAAAGCAATTCAGC-1                35591 Input\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "An object of class Seurat \n",
       "100522 features across 743 samples within 1 assay \n",
       "Active assay: ATAC (100522 features, 0 variable features)\n",
       "\n",
       "[[2]]\n",
       "An object of class Seurat \n",
       "187490 features across 5312 samples within 1 assay \n",
       "Active assay: ATAC (187490 features, 0 variable features)\n",
       "\n",
       "[[3]]\n",
       "An object of class Seurat \n",
       "80277 features across 222 samples within 1 assay \n",
       "Active assay: ATAC (80277 features, 0 variable features)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allSE = c() \n",
    "\n",
    "samples_df <- cbind(sample_names, samples)\n",
    "print('samples_df')\n",
    "print(samples_df)\n",
    "for (row in 1:nrow(samples_df)){\n",
    "    exp <- (samples_df[[row, \"samples\"]])\n",
    "    name <- (samples_df[[row, \"sample_names\"]]) \n",
    "    \n",
    "#for (exp in samples) {\n",
    "    print('exp')\n",
    "    print(exp)\n",
    "    print('name')\n",
    "    print(name)\n",
    "    barcode_path <- file.path(cellr_in, exp, \"outs\", \"filtered_peak_bc_matrix\", \"barcodes.tsv\")    \n",
    "    barcodes <- readr::read_tsv(barcode_path, col_names = F) # %>% tidyr::unite(barcode)\n",
    "    barcodes <- as.data.frame(barcodes) %>%  tibble::column_to_rownames(var=\"X1\") %>% tibble::add_column(proj=name)\n",
    "    frag_file <- file.path(cellr_in, exp, \"outs\", \"fragments.tsv.gz\")\n",
    "    \n",
    "    cells.meta.f <- file.path(cellr_in, exp, \"outs\", \"singlecell.csv\") \n",
    "    cells.meta <- as.data.frame(readr::read_csv(cells.meta.f)) %>% tibble::column_to_rownames(var=\"barcode\") %>% tibble::add_column(proj=name)\n",
    "    cells.meta <- cells.meta[rownames(cells.meta) %in% rownames(barcodes), ]\n",
    "\n",
    "    # quantify multiome peaks in the scATAC-seq dataset\n",
    "    \n",
    "    \n",
    "    print(\"Creating fragments object\")\n",
    "    frags.curr <- CreateFragmentObject(path = frag_file, cells= rownames(barcodes))\n",
    "    #print(frags.curr)\n",
    "    print(\"Quantifying peaks\")\n",
    "    ## Quantify peaks\n",
    "    curr.counts <- FeatureMatrix(\n",
    "      fragments = frags.curr,\n",
    "      features = combined.peaks,\n",
    "      cells = rownames(barcodes),\n",
    "      process_n = cores\n",
    "    )\n",
    "    \n",
    "    print(\"Creating chromatin assay\")\n",
    "    ## Create the objects and use simple filters\n",
    "    curr_assay <- CreateChromatinAssay(curr.counts, fragments = frags.curr, min.cells = 10, min.features = 200)\n",
    "    curr <- CreateSeuratObject(curr_assay, assay = \"ATAC\", project=name, meta.data=cells.meta)\n",
    "    print('curr_assay')\n",
    "    print(head(curr_assay))\n",
    "    print('curr')\n",
    "    print(head(curr[[]]))\n",
    "    allSE = c(allSE, curr)\n",
    "    #return(curr)\n",
    "}\n",
    "\n",
    "allSE\n",
    "\n",
    "#allSE <- sapply(samples, create_frag, cellr_in=cellr_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c116a7db",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(allSE, file.path(outdir, paste0(\"allSamples.rds\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a2cf886",
   "metadata": {},
   "outputs": [],
   "source": [
    "vPlot <- function(se){\n",
    "      vPlot <- VlnPlot(\n",
    "      object = se,\n",
    "      features = c('pct_reads_in_peaks', 'peak_region_fragments',\n",
    "                   'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal'),\n",
    "      pt.size = 0.1,\n",
    "      ncol = 5\n",
    "    )\n",
    "    vPlot <- vPlot +    # Create grid of plots with title\n",
    "             plot_annotation(title = se$orig.ident[[1]]) &\n",
    "             theme(plot.title = element_text(hjust = 0.5, size=15))\n",
    "    #print(vPlot)\n",
    "    return(vPlot)\n",
    "}\n",
    "lapply(allSE,vPlot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "626a0538",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 18.04.6 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/isshamie/miniconda3/envs/mttrace/lib/libopenblasp-r0.3.17.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] future_1.22.1             dplyr_1.0.7              \n",
       " [3] metap_1.5                 cowplot_1.1.1            \n",
       " [5] magrittr_2.0.1            data.table_1.14.2        \n",
       " [7] patchwork_1.1.1           ggplot2_3.3.5            \n",
       " [9] EnsDb.Hsapiens.v75_2.99.0 ensembldb_2.16.4         \n",
       "[11] AnnotationFilter_1.16.0   GenomicFeatures_1.44.2   \n",
       "[13] AnnotationDbi_1.54.1      Biobase_2.52.0           \n",
       "[15] Signac_1.4.0              SeuratObject_4.0.2       \n",
       "[17] Seurat_4.0.4              GenomicRanges_1.44.0     \n",
       "[19] GenomeInfoDb_1.28.4       IRanges_2.26.0           \n",
       "[21] S4Vectors_0.30.1          BiocGenerics_0.38.0      \n",
       "[23] repr_1.1.3               \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] rappdirs_0.3.3              SnowballC_0.7.0            \n",
       "  [3] rtracklayer_1.52.1          pbdZMQ_0.3-5               \n",
       "  [5] scattermore_0.7             tidyr_1.1.4                \n",
       "  [7] bit64_4.0.5                 knitr_1.36                 \n",
       "  [9] irlba_2.3.3                 multcomp_1.4-17            \n",
       " [11] DelayedArray_0.18.0         rpart_4.1-15               \n",
       " [13] KEGGREST_1.32.0             RCurl_1.98-1.5             \n",
       " [15] generics_0.1.0              TH.data_1.1-0              \n",
       " [17] RSQLite_2.2.8               RANN_2.6.1                 \n",
       " [19] bit_4.0.4                   tzdb_0.1.2                 \n",
       " [21] mutoss_0.1-12               spatstat.data_2.1-0        \n",
       " [23] xml2_1.3.2                  httpuv_1.6.3               \n",
       " [25] SummarizedExperiment_1.22.0 assertthat_0.2.1           \n",
       " [27] xfun_0.26                   hms_1.1.1                  \n",
       " [29] evaluate_0.14               promises_1.2.0.1           \n",
       " [31] fansi_0.5.0                 restfulr_0.0.13            \n",
       " [33] progress_1.2.2              dbplyr_2.1.1               \n",
       " [35] igraph_1.2.6                DBI_1.1.1                  \n",
       " [37] tmvnsim_1.0-2               htmlwidgets_1.5.4          \n",
       " [39] sparsesvd_0.2               spatstat.geom_2.2-2        \n",
       " [41] purrr_0.3.4                 ellipsis_0.3.2             \n",
       " [43] RSpectra_0.16-0             backports_1.2.1            \n",
       " [45] biomaRt_2.48.3              deldir_0.2-10              \n",
       " [47] MatrixGenerics_1.4.3        vctrs_0.3.8                \n",
       " [49] Cairo_1.5-12.2              ROCR_1.0-11                \n",
       " [51] abind_1.4-5                 cachem_1.0.6               \n",
       " [53] withr_2.4.2                 ggforce_0.3.3              \n",
       " [55] BSgenome_1.60.0             checkmate_2.0.0            \n",
       " [57] vroom_1.5.5                 sctransform_0.3.2          \n",
       " [59] GenomicAlignments_1.28.0    prettyunits_1.1.1          \n",
       " [61] goftest_1.2-2               mnormt_2.0.2               \n",
       " [63] cluster_2.1.2               IRdisplay_1.0              \n",
       " [65] lazyeval_0.2.2              crayon_1.4.1               \n",
       " [67] pkgconfig_2.0.3             slam_0.1-48                \n",
       " [69] labeling_0.4.2              tweenr_1.0.2               \n",
       " [71] nlme_3.1-153                ProtGenerics_1.24.0        \n",
       " [73] nnet_7.3-16                 rlang_0.4.11               \n",
       " [75] globals_0.14.0              lifecycle_1.0.1            \n",
       " [77] miniUI_0.1.1.1              sandwich_3.0-1             \n",
       " [79] filelock_1.0.2              BiocFileCache_2.0.0        \n",
       " [81] mathjaxr_1.4-0              dichromat_2.0-0            \n",
       " [83] polyclip_1.10-0             matrixStats_0.61.0         \n",
       " [85] lmtest_0.9-38               Matrix_1.3-4               \n",
       " [87] ggseqlogo_0.1               IRkernel_1.2               \n",
       " [89] zoo_1.8-9                   base64enc_0.1-3            \n",
       " [91] ggridges_0.5.3              png_0.1-7                  \n",
       " [93] viridisLite_0.4.0           rjson_0.2.20               \n",
       " [95] bitops_1.0-7                KernSmooth_2.23-20         \n",
       " [97] Biostrings_2.60.2           blob_1.2.2                 \n",
       " [99] stringr_1.4.0               parallelly_1.28.1          \n",
       "[101] readr_2.0.2                 jpeg_0.1-9                 \n",
       "[103] scales_1.1.1                memoise_2.0.0              \n",
       "[105] plyr_1.8.6                  ica_1.0-2                  \n",
       "[107] zlibbioc_1.38.0             compiler_4.1.1             \n",
       "[109] BiocIO_1.2.0                RColorBrewer_1.1-2         \n",
       "[111] plotrix_3.8-2               fitdistrplus_1.1-6         \n",
       "[113] Rsamtools_2.8.0             cli_3.0.1                  \n",
       "[115] XVector_0.32.0              listenv_0.8.0              \n",
       "[117] pbapply_1.5-0               htmlTable_2.2.1            \n",
       "[119] Formula_1.2-4               MASS_7.3-54                \n",
       "[121] mgcv_1.8-37                 tidyselect_1.1.1           \n",
       "[123] stringi_1.7.4               yaml_2.2.1                 \n",
       "[125] latticeExtra_0.6-29         ggrepel_0.9.1              \n",
       "[127] grid_4.1.1                  VariantAnnotation_1.38.0   \n",
       "[129] fastmatch_1.1-3             tools_4.1.1                \n",
       "[131] future.apply_1.8.1          rstudioapi_0.13            \n",
       "[133] uuid_0.1-4                  foreign_0.8-81             \n",
       "[135] lsa_0.73.2                  gridExtra_2.3              \n",
       "[137] farver_2.1.0                Rtsne_0.15                 \n",
       "[139] digest_0.6.28               shiny_1.7.0                \n",
       "[141] qlcMatrix_0.9.7             Rcpp_1.0.7                 \n",
       "[143] later_1.3.0                 RcppAnnoy_0.0.19           \n",
       "[145] httr_1.4.2                  biovizBase_1.40.0          \n",
       "[147] Rdpack_2.1.2                colorspace_2.0-2           \n",
       "[149] XML_3.99-0.8                tensor_1.5                 \n",
       "[151] reticulate_1.22             splines_4.1.1              \n",
       "[153] uwot_0.1.10                 RcppRoll_0.3.0             \n",
       "[155] sn_2.0.0                    spatstat.utils_2.2-0       \n",
       "[157] multtest_2.48.0             plotly_4.9.4.1             \n",
       "[159] xtable_1.8-4                jsonlite_1.7.2             \n",
       "[161] R6_2.5.1                    TFisher_0.2.0              \n",
       "[163] Hmisc_4.5-0                 pillar_1.6.3               \n",
       "[165] htmltools_0.5.2             mime_0.12                  \n",
       "[167] glue_1.4.2                  fastmap_1.1.0              \n",
       "[169] BiocParallel_1.26.2         codetools_0.2-18           \n",
       "[171] mvtnorm_1.1-2               utf8_1.2.2                 \n",
       "[173] lattice_0.20-45             spatstat.sparse_2.0-0      \n",
       "[175] tibble_3.1.5                numDeriv_2016.8-1.1        \n",
       "[177] curl_4.3.2                  leiden_0.3.9               \n",
       "[179] survival_3.2-13             docopt_0.7.1               \n",
       "[181] munsell_0.5.0               GenomeInfoDbData_1.2.6     \n",
       "[183] reshape2_1.4.4              gtable_0.3.0               \n",
       "[185] rbibutils_2.2.3             spatstat.core_2.3-0        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb0cc51b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
