{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "223d6f46",
   "metadata": {},
   "source": [
    "## Creates binarized Seurat objects, integrates conditions and annotates genes by nearby peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "663e2d4d",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# Input info\n",
    "cellr_in = \"/home/isshamie/lewis/mito_lineage/output/mtscATAC/data/jan21_2021/MTblacklist/\" \n",
    "outdir = \"/home/isshamie/data/Projects/Mito_Trace/output/annotation/data/cd34norm/MTblacklist/mergedSamples/\"\n",
    "\n",
    "sample_names = \"Control,Flt3l\"\n",
    "samples = \"P2,J2\"\n",
    "\n",
    "# Parameters\n",
    "nTop = 25000\n",
    "cores = 24"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b62b75ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(repr)\n",
    "options(repr.plot.width=12, repr.plot.height=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c2c99fc4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'P2'</li><li>'J2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'P2'\n",
       "\\item 'J2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'P2'\n",
       "2. 'J2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"P2\" \"J2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "samples <- unlist(strsplit(samples, \",\")[[1]])\n",
    "sample_names <- unlist(strsplit(sample_names, \",\"))\n",
    "\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aad4f638",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'data.table'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'magrittr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    not\n",
      "\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'cowplot'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:patchwork':\n",
      "\n",
      "    align_plots\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biostrings':\n",
      "\n",
      "    collapse, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:XVector':\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:ensembldb':\n",
      "\n",
      "    filter, select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'future'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    value\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomicRanges':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"Strategy 'multiprocess' is deprecated in future (>= 1.20.0). Instead, explicitly specify either 'multisession' or 'multicore'. In the current R session, 'multiprocess' equals 'multicore'.\"\n"
     ]
    }
   ],
   "source": [
    "library(GenomicRanges)\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(GenomeInfoDb)\n",
    "library(EnsDb.Hsapiens.v75)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "set.seed(1234)\n",
    "library(data.table)\n",
    "library(magrittr)\n",
    "library(Rsamtools)\n",
    "library(cowplot)\n",
    "library(metap)\n",
    "library(dplyr)\n",
    "library(future)\n",
    "\n",
    "plan(\"multiprocess\", workers = cores)\n",
    "options(future.globals.maxSize = 8000 * 1024^2)\n",
    "#options(future.globals.maxSize = 50000 * 1024^2) # for 50 Gb RAM\n",
    "#plan(\"multiprocess\", workers = workers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a53a3256",
   "metadata": {},
   "source": [
    "## Merge all peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "62849b8a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"here\"\n",
      "[1] \"/home/isshamie/lewis/mito_lineage/output/mtscATAC/data/jan21_2021/MTblacklist//P2/outs/filtered_peak_bc_matrix/peaks.bed\"\n",
      "[1] \"here\"\n",
      "[1] \"/home/isshamie/lewis/mito_lineage/output/mtscATAC/data/jan21_2021/MTblacklist//J2/outs/filtered_peak_bc_matrix/peaks.bed\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 149551 ranges and 0 metadata columns:\n",
       "           seqnames            ranges strand\n",
       "              <Rle>         <IRanges>  <Rle>\n",
       "       [1]     chr1        9942-10364      *\n",
       "       [2]     chr1     191685-191736      *\n",
       "       [3]     chr1     267780-268257      *\n",
       "       [4]     chr1     271072-271548      *\n",
       "       [5]     chr1     585995-586411      *\n",
       "       ...      ...               ...    ...\n",
       "  [149547]     chrY 56844769-56845155      *\n",
       "  [149548]     chrY 56846033-56848664      *\n",
       "  [149549]     chrY 56849234-56851581      *\n",
       "  [149550]     chrY 56857506-56857613      *\n",
       "  [149551]     chrY 56873729-56874140      *\n",
       "  -------\n",
       "  seqinfo: 24 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "read.peaks <- function(exp, cellr_in){\n",
    "    print('here')\n",
    "    print(file.path(cellr_in, exp, \"outs\", \"filtered_peak_bc_matrix\", \"peaks.bed\"))\n",
    "    peaks <- read.table(\n",
    "      file = file.path(cellr_in, exp, \"outs\", \"filtered_peak_bc_matrix\", \"peaks.bed\"),\n",
    "      col.names = c(\"chr\", \"start\", \"end\")\n",
    "    )\n",
    "    # convert to genomic ranges\n",
    "    gr <- makeGRangesFromDataFrame(peaks)\n",
    "    return(gr)\n",
    "}\n",
    "\n",
    "\n",
    "gr.full <- c(sapply(samples, read.peaks, cellr_in=cellr_in, USE.NAMES=F))\n",
    "\n",
    "gr.full.c <- gr.full[[1]]\n",
    "if (length(gr.full)>1){\n",
    "    for (i in 2:length(gr.full)){\n",
    "      gr.full.c <- c(gr.full.c, gr.full[[i]])\n",
    "    }\n",
    "}\n",
    "combined.peaks <- reduce(x = c(gr.full.c))\n",
    "\n",
    "# Filter out bad peaks based on length\n",
    "peakwidths <- width(combined.peaks)\n",
    "combined.peaks <- combined.peaks[peakwidths  < 10000 & peakwidths > 20]\n",
    "combined.peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cae010c",
   "metadata": {},
   "source": [
    "## Create fragment objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "67e92e33",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"samples_df\"\n",
      "     sample_names samples\n",
      "[1,] \"Control\"    \"P2\"   \n",
      "[2,] \"Flt3l\"      \"J2\"   \n",
      "[1] \"exp\"\n",
      "[1] \"P2\"\n",
      "[1] \"name\"\n",
      "[1] \"Control\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'cli':\n",
      "  method     from         \n",
      "  print.boxx spatstat.geom\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m6875\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m1\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): X1\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m332219\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m18\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m  (2): barcode, cell_id\n",
      "\u001b[32mdbl\u001b[39m (16): total, duplicate, chimeric, unmapped, lowmapq, mitochondrial, pass...\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating fragments object\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Quantifying peaks\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating chromatin assay\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CreateSeuratObject.Assay(curr_assay, assay = \"ATAC\", project = name, :\n",
      "\"Some cells in meta.data not present in provided counts matrix.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"curr_assay\"\n",
      "data frame with 0 columns and 10 rows\n",
      "[1] \"curr\"\n",
      "                   orig.ident nCount_ATAC nFeature_ATAC total duplicate\n",
      "AAACGAAAGAGGTCCA-1    Control        1400          1357 16929      2587\n",
      "AAACGAAAGCGATACG-1    Control        3692          3505 45359     10546\n",
      "AAACGAAAGTCGTGAG-1    Control        1031           994 10177      2450\n",
      "AAACGAACAATAGTGA-1    Control        2829          2687 22452      4364\n",
      "AAACGAACACAATAAG-1    Control        1206          1184 11763      2003\n",
      "AAACGAACACTGATAC-1    Control         968           954  9449      1971\n",
      "                   chimeric unmapped lowmapq mitochondrial passed_filters\n",
      "AAACGAAAGAGGTCCA-1       58       70     674          9629           3911\n",
      "AAACGAAAGCGATACG-1      184      184    2302         22247           9896\n",
      "AAACGAAAGTCGTGAG-1       48       60     657          2135           4827\n",
      "AAACGAACAATAGTGA-1      100       83     897          9233           7775\n",
      "AAACGAACACAATAAG-1       43       57     680          5547           3433\n",
      "AAACGAACACTGATAC-1       59       21     489          3785           3124\n",
      "                   cell_id is__cell_barcode TSS_fragments\n",
      "AAACGAAAGAGGTCCA-1 _cell_0                1           740\n",
      "AAACGAAAGCGATACG-1 _cell_1                1          1611\n",
      "AAACGAAAGTCGTGAG-1 _cell_2                1           775\n",
      "AAACGAACAATAGTGA-1 _cell_3                1          1239\n",
      "AAACGAACACAATAAG-1 _cell_4                1           621\n",
      "AAACGAACACTGATAC-1 _cell_5                1           517\n",
      "                   DNase_sensitive_region_fragments enhancer_region_fragments\n",
      "AAACGAAAGAGGTCCA-1                                0                         0\n",
      "AAACGAAAGCGATACG-1                                0                         0\n",
      "AAACGAAAGTCGTGAG-1                                0                         0\n",
      "AAACGAACAATAGTGA-1                                0                         0\n",
      "AAACGAACACAATAAG-1                                0                         0\n",
      "AAACGAACACTGATAC-1                                0                         0\n",
      "                   promoter_region_fragments on_target_fragments\n",
      "AAACGAAAGAGGTCCA-1                         0                 740\n",
      "AAACGAAAGCGATACG-1                         0                1611\n",
      "AAACGAAAGTCGTGAG-1                         0                 775\n",
      "AAACGAACAATAGTGA-1                         0                1239\n",
      "AAACGAACACAATAAG-1                         0                 621\n",
      "AAACGAACACTGATAC-1                         0                 517\n",
      "                   blacklist_region_fragments peak_region_fragments\n",
      "AAACGAAAGAGGTCCA-1                          0                  1330\n",
      "AAACGAAAGCGATACG-1                          0                  3462\n",
      "AAACGAAAGTCGTGAG-1                          0                   966\n",
      "AAACGAACAATAGTGA-1                          0                  2701\n",
      "AAACGAACACAATAAG-1                          0                  1138\n",
      "AAACGAACACTGATAC-1                          0                   878\n",
      "                   peak_region_cutsites    proj\n",
      "AAACGAAAGAGGTCCA-1                 2434 Control\n",
      "AAACGAAAGCGATACG-1                 6167 Control\n",
      "AAACGAAAGTCGTGAG-1                 1768 Control\n",
      "AAACGAACAATAGTGA-1                 4951 Control\n",
      "AAACGAACACAATAAG-1                 2019 Control\n",
      "AAACGAACACTGATAC-1                 1595 Control\n",
      "[1] \"exp\"\n",
      "[1] \"J2\"\n",
      "[1] \"name\"\n",
      "[1] \"Flt3l\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m12009\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m1\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): X1\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m430951\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m18\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m  (2): barcode, cell_id\n",
      "\u001b[32mdbl\u001b[39m (16): total, duplicate, chimeric, unmapped, lowmapq, mitochondrial, pass...\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating fragments object\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Quantifying peaks\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating chromatin assay\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CreateSeuratObject.Assay(curr_assay, assay = \"ATAC\", project = name, :\n",
      "\"Some cells in meta.data not present in provided counts matrix.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"curr_assay\"\n",
      "data frame with 0 columns and 10 rows\n",
      "[1] \"curr\"\n",
      "                   orig.ident nCount_ATAC nFeature_ATAC total duplicate\n",
      "AAACGAAAGAGCTCCC-1      Flt3l        4716          4430 25798      3956\n",
      "AAACGAAAGCGATACG-1      Flt3l        2136          2066 18340      2478\n",
      "AAACGAAAGGCTTCGC-1      Flt3l        3875          3691 28417      4398\n",
      "AAACGAAAGTACAACA-1      Flt3l        1353          1317 12018      1524\n",
      "AAACGAACAACGTACT-1      Flt3l        2973          2864 37690      3034\n",
      "AAACGAACAAGCGGTA-1      Flt3l        1292          1257  9381      1416\n",
      "                   chimeric unmapped lowmapq mitochondrial passed_filters\n",
      "AAACGAAAGAGCTCCC-1      111       95    1643          4740          15253\n",
      "AAACGAAAGCGATACG-1       78       92     794          9325           5573\n",
      "AAACGAAAGGCTTCGC-1      132      122    1771         10224          11770\n",
      "AAACGAAAGTACAACA-1       60       45     585          5688           4116\n",
      "AAACGAACAACGTACT-1       82      125    1680         24974           7795\n",
      "AAACGAACAAGCGGTA-1       44       50     547          3616           3708\n",
      "                   cell_id is__cell_barcode TSS_fragments\n",
      "AAACGAAAGAGCTCCC-1 _cell_0                1          2708\n",
      "AAACGAAAGCGATACG-1 _cell_1                1          1062\n",
      "AAACGAAAGGCTTCGC-1 _cell_2                1          2145\n",
      "AAACGAAAGTACAACA-1 _cell_3                1           727\n",
      "AAACGAACAACGTACT-1 _cell_4                1          1562\n",
      "AAACGAACAAGCGGTA-1 _cell_5                1           705\n",
      "                   DNase_sensitive_region_fragments enhancer_region_fragments\n",
      "AAACGAAAGAGCTCCC-1                                0                         0\n",
      "AAACGAAAGCGATACG-1                                0                         0\n",
      "AAACGAAAGGCTTCGC-1                                0                         0\n",
      "AAACGAAAGTACAACA-1                                0                         0\n",
      "AAACGAACAACGTACT-1                                0                         0\n",
      "AAACGAACAAGCGGTA-1                                0                         0\n",
      "                   promoter_region_fragments on_target_fragments\n",
      "AAACGAAAGAGCTCCC-1                         0                2708\n",
      "AAACGAAAGCGATACG-1                         0                1062\n",
      "AAACGAAAGGCTTCGC-1                         0                2145\n",
      "AAACGAAAGTACAACA-1                         0                 727\n",
      "AAACGAACAACGTACT-1                         0                1562\n",
      "AAACGAACAAGCGGTA-1                         0                 705\n",
      "                   blacklist_region_fragments peak_region_fragments\n",
      "AAACGAAAGAGCTCCC-1                          0                  4667\n",
      "AAACGAAAGCGATACG-1                          0                  2122\n",
      "AAACGAAAGGCTTCGC-1                          0                  3851\n",
      "AAACGAAAGTACAACA-1                          0                  1329\n",
      "AAACGAACAACGTACT-1                          0                  2945\n",
      "AAACGAACAAGCGGTA-1                          0                  1279\n",
      "                   peak_region_cutsites  proj\n",
      "AAACGAAAGAGCTCCC-1                 8189 Flt3l\n",
      "AAACGAAAGCGATACG-1                 3975 Flt3l\n",
      "AAACGAAAGGCTTCGC-1                 6992 Flt3l\n",
      "AAACGAAAGTACAACA-1                 2440 Flt3l\n",
      "AAACGAACAACGTACT-1                 5445 Flt3l\n",
      "AAACGAACAAGCGGTA-1                 2386 Flt3l\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "An object of class Seurat \n",
       "147518 features across 6543 samples within 1 assay \n",
       "Active assay: ATAC (147518 features, 0 variable features)\n",
       "\n",
       "[[2]]\n",
       "An object of class Seurat \n",
       "149495 features across 11949 samples within 1 assay \n",
       "Active assay: ATAC (149495 features, 0 variable features)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allSE = c() \n",
    "\n",
    "samples_df <- cbind(sample_names, samples)\n",
    "print('samples_df')\n",
    "print(samples_df)\n",
    "for (row in 1:nrow(samples_df)){\n",
    "    exp <- (samples_df[[row, \"samples\"]])\n",
    "    name <- (samples_df[[row, \"sample_names\"]]) \n",
    "    \n",
    "    print('exp')\n",
    "    print(exp)\n",
    "    print('name')\n",
    "    print(name)\n",
    "    barcode_path <- file.path(cellr_in, exp, \"outs\", \"filtered_peak_bc_matrix\", \"barcodes.tsv\")    \n",
    "    barcodes <- readr::read_tsv(barcode_path, col_names = F) # %>% tidyr::unite(barcode)\n",
    "    barcodes <- as.data.frame(barcodes) %>%  tibble::column_to_rownames(var=\"X1\") %>% tibble::add_column(proj=name)\n",
    "    frag_file <- file.path(cellr_in, exp, \"outs\", \"fragments.tsv.gz\")\n",
    "    \n",
    "    cells.meta.f <- file.path(cellr_in, exp, \"outs\", \"singlecell.csv\") \n",
    "    cells.meta <- as.data.frame(readr::read_csv(cells.meta.f)) %>% tibble::column_to_rownames(var=\"barcode\") %>% tibble::add_column(proj=name)\n",
    "    cells.meta <- cells.meta[rownames(cells.meta) %in% rownames(barcodes), ]\n",
    "\n",
    "    # quantify multiome peaks in the scATAC-seq dataset\n",
    "    \n",
    "    \n",
    "    print(\"Creating fragments object\")\n",
    "    frags.curr <- CreateFragmentObject(path = frag_file, cells= rownames(barcodes))\n",
    "    #print(frags.curr)\n",
    "    print(\"Quantifying peaks\")\n",
    "    ## Quantify peaks\n",
    "    curr.counts <- FeatureMatrix(\n",
    "      fragments = frags.curr,\n",
    "      features = combined.peaks,\n",
    "      cells = rownames(barcodes),\n",
    "      process_n = cores\n",
    "    )\n",
    "    \n",
    "    print(\"Creating chromatin assay\")\n",
    "    ## Create the objects and use simple filters\n",
    "    curr_assay <- CreateChromatinAssay(curr.counts, fragments = frags.curr, min.cells = 10, min.features = 200)\n",
    "    curr <- CreateSeuratObject(curr_assay, assay = \"ATAC\", project=name, meta.data=cells.meta)\n",
    "    print('curr_assay')\n",
    "    print(head(curr_assay))\n",
    "    print('curr')\n",
    "    print(head(curr[[]]))\n",
    "    allSE = c(allSE, curr)\n",
    "    #return(curr)\n",
    "}\n",
    "\n",
    "\n",
    "#allSE <- sapply(samples, create_frag, cellr_in=cellr_in)\n",
    "allSE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feba99cd",
   "metadata": {},
   "source": [
    "## QC metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a564ef8d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "qc <- function(se){\n",
    "    # extract gene annotations from EnsDb\n",
    "    annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)\n",
    "\n",
    "    # change to UCSC style since the data was mapped to hg19\n",
    "    seqlevelsStyle(annotations) <- 'UCSC'\n",
    "    genome(annotations) <- \"hg38\"\n",
    "    # add the gene information to the object\n",
    "    Annotation(se) <- annotations\n",
    "    gene.activities <- GeneActivity(se)\n",
    "    # compute nucleosome signal score per cell\n",
    "    se <- NucleosomeSignal(object = se)\n",
    "    \n",
    "    Annotation(se) <- annotations\n",
    "    # compute TSS enrichment score per cell\n",
    "    se <- TSSEnrichment(object = se, fast = FALSE)\n",
    "\n",
    "    # add blacklist ratio and fraction of reads in peaks\n",
    "    se$pct_reads_in_peaks <- se$peak_region_fragments / se$passed_filters * 100\n",
    "    se$blacklist_ratio <- se$blacklist_region_fragments / se$peak_region_fragments\n",
    "    se$high.tss <- ifelse(se$TSS.enrichment > 2, 'High', 'Low')\n",
    "    se$nucleosome_group <- ifelse(se$nucleosome_signal > 4, 'NS > 4', 'NS < 4')\n",
    "\n",
    "    return(se)\n",
    "}\n",
    "vPlot <- function(se){\n",
    "      vPlot <- VlnPlot(\n",
    "      object = se,\n",
    "      features = c('pct_reads_in_peaks', 'peak_region_fragments',\n",
    "                   'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal'),\n",
    "      pt.size = 0.1,\n",
    "      ncol = 5\n",
    "    )  \n",
    "    vPlot <- vPlot +    # Create grid of plots with title\n",
    "             plot_annotation(title = se$orig.ident[[1]]) & \n",
    "             theme(plot.title = element_text(hjust = 0.5, size=15))\n",
    "    #print(vPlot)\n",
    "    return(vPlot)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1422fd9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Fetching data...\n",
      "OK\n",
      "\n",
      "Parsing exons...\n",
      "OK\n",
      "Defining introns...\n",
      "OK\n",
      "Defining UTRs...\n",
      "OK\n",
      "Defining CDS...\n",
      "OK\n",
      "\n",
      "aggregating...\n",
      "\n",
      "Done\n",
      "\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .Seqinfo.mergexy(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n"
     ]
    }
   ],
   "source": [
    "allSE <- lapply(allSE, qc)\n",
    "saveRDS(allSE, file.path(outdir, paste0(\"allSamples.rds\")))\n",
    "lapply(allSE,vPlot)\n",
    "ggsave(file.path(outdir, paste0(\"QC_01.png\")))\n",
    "ggsave(file.path(outdir, paste0(\"QC_01.pdf\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ca89e0de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "An object of class Seurat \n",
       "147518 features across 6543 samples within 1 assay \n",
       "Active assay: ATAC (147518 features, 0 variable features)\n",
       "\n",
       "[[2]]\n",
       "An object of class Seurat \n",
       "149495 features across 11949 samples within 1 assay \n",
       "Active assay: ATAC (149495 features, 0 variable features)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "626a0538",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.3 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/isshamie/miniconda3/envs/ranno/lib/libopenblasp-r0.3.18.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] future_1.23.0             dplyr_1.0.7              \n",
       " [3] metap_1.6                 cowplot_1.1.1            \n",
       " [5] Rsamtools_2.10.0          Biostrings_2.62.0        \n",
       " [7] XVector_0.34.0            magrittr_2.0.1           \n",
       " [9] data.table_1.14.2         patchwork_1.1.1          \n",
       "[11] ggplot2_3.3.5             EnsDb.Hsapiens.v75_2.99.0\n",
       "[13] ensembldb_2.18.2          AnnotationFilter_1.18.0  \n",
       "[15] GenomicFeatures_1.46.1    AnnotationDbi_1.56.2     \n",
       "[17] Biobase_2.54.0            Signac_1.4.0             \n",
       "[19] SeuratObject_4.0.3        Seurat_4.0.5             \n",
       "[21] GenomicRanges_1.46.0      GenomeInfoDb_1.30.0      \n",
       "[23] IRanges_2.28.0            S4Vectors_0.32.2         \n",
       "[25] BiocGenerics_0.40.0       repr_1.1.3               \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] rappdirs_0.3.3              SnowballC_0.7.0            \n",
       "  [3] rtracklayer_1.54.0          pbdZMQ_0.3-6               \n",
       "  [5] scattermore_0.7             tidyr_1.1.4                \n",
       "  [7] bit64_4.0.5                 knitr_1.35                 \n",
       "  [9] irlba_2.3.3                 multcomp_1.4-17            \n",
       " [11] DelayedArray_0.20.0         rpart_4.1-15               \n",
       " [13] KEGGREST_1.34.0             RCurl_1.98-1.5             \n",
       " [15] generics_0.1.1              TH.data_1.1-0              \n",
       " [17] RSQLite_2.2.8               RANN_2.6.1                 \n",
       " [19] bit_4.0.4                   tzdb_0.2.0                 \n",
       " [21] mutoss_0.1-12               spatstat.data_2.1-0        \n",
       " [23] xml2_1.3.2                  httpuv_1.6.3               \n",
       " [25] SummarizedExperiment_1.24.0 assertthat_0.2.1           \n",
       " [27] xfun_0.28                   hms_1.1.1                  \n",
       " [29] evaluate_0.14               promises_1.2.0.1           \n",
       " [31] fansi_0.4.2                 restfulr_0.0.13            \n",
       " [33] progress_1.2.2              dbplyr_2.1.1               \n",
       " [35] igraph_1.2.8                DBI_1.1.1                  \n",
       " [37] tmvnsim_1.0-2               htmlwidgets_1.5.4          \n",
       " [39] sparsesvd_0.2               spatstat.geom_2.3-0        \n",
       " [41] purrr_0.3.4                 ellipsis_0.3.2             \n",
       " [43] backports_1.3.0             biomaRt_2.50.0             \n",
       " [45] deldir_1.0-6                MatrixGenerics_1.6.0       \n",
       " [47] vctrs_0.3.8                 ROCR_1.0-11                \n",
       " [49] abind_1.4-5                 cachem_1.0.6               \n",
       " [51] withr_2.4.2                 ggforce_0.3.3              \n",
       " [53] BSgenome_1.62.0             checkmate_2.0.0            \n",
       " [55] vroom_1.5.6                 sctransform_0.3.2          \n",
       " [57] GenomicAlignments_1.30.0    prettyunits_1.1.1          \n",
       " [59] goftest_1.2-3               mnormt_2.0.2               \n",
       " [61] cluster_2.1.2               IRdisplay_1.0              \n",
       " [63] lazyeval_0.2.2              crayon_1.4.2               \n",
       " [65] pkgconfig_2.0.3             slam_0.1-49                \n",
       " [67] tweenr_1.0.2                nlme_3.1-153               \n",
       " [69] ProtGenerics_1.26.0         nnet_7.3-16                \n",
       " [71] rlang_0.4.12                globals_0.14.0             \n",
       " [73] lifecycle_1.0.1             miniUI_0.1.1.1             \n",
       " [75] sandwich_3.0-1              filelock_1.0.2             \n",
       " [77] BiocFileCache_2.2.0         mathjaxr_1.4-0             \n",
       " [79] dichromat_2.0-0             polyclip_1.10-0            \n",
       " [81] matrixStats_0.61.0          lmtest_0.9-39              \n",
       " [83] Matrix_1.3-4                ggseqlogo_0.1              \n",
       " [85] IRkernel_1.2                zoo_1.8-9                  \n",
       " [87] base64enc_0.1-3             ggridges_0.5.3             \n",
       " [89] png_0.1-7                   viridisLite_0.4.0          \n",
       " [91] rjson_0.2.20                bitops_1.0-7               \n",
       " [93] KernSmooth_2.23-20          blob_1.2.2                 \n",
       " [95] stringr_1.4.0               parallelly_1.28.1          \n",
       " [97] readr_2.1.0                 jpeg_0.1-9                 \n",
       " [99] scales_1.1.1                memoise_2.0.0              \n",
       "[101] plyr_1.8.6                  ica_1.0-2                  \n",
       "[103] zlibbioc_1.40.0             compiler_4.1.1             \n",
       "[105] BiocIO_1.4.0                RColorBrewer_1.1-2         \n",
       "[107] plotrix_3.8-2               fitdistrplus_1.1-6         \n",
       "[109] cli_3.1.0                   listenv_0.8.0              \n",
       "[111] pbapply_1.5-0               htmlTable_2.3.0            \n",
       "[113] Formula_1.2-4               MASS_7.3-54                \n",
       "[115] mgcv_1.8-38                 tidyselect_1.1.1           \n",
       "[117] stringi_1.7.5               yaml_2.2.1                 \n",
       "[119] latticeExtra_0.6-29         ggrepel_0.9.1              \n",
       "[121] grid_4.1.1                  VariantAnnotation_1.40.0   \n",
       "[123] fastmatch_1.1-3             tools_4.1.1                \n",
       "[125] future.apply_1.8.1          parallel_4.1.1             \n",
       "[127] rstudioapi_0.13             uuid_1.0-3                 \n",
       "[129] foreign_0.8-81              lsa_0.73.2                 \n",
       "[131] gridExtra_2.3               farver_2.1.0               \n",
       "[133] Rtsne_0.15                  digest_0.6.28              \n",
       "[135] shiny_1.7.1                 qlcMatrix_0.9.7            \n",
       "[137] Rcpp_1.0.7                  later_1.2.0                \n",
       "[139] RcppAnnoy_0.0.19            httr_1.4.2                 \n",
       "[141] biovizBase_1.42.0           Rdpack_2.1.2               \n",
       "[143] colorspace_2.0-2            XML_3.99-0.8               \n",
       "[145] tensor_1.5                  reticulate_1.22            \n",
       "[147] splines_4.1.1               uwot_0.1.10                \n",
       "[149] RcppRoll_0.3.0              sn_2.0.0                   \n",
       "[151] spatstat.utils_2.2-0        multtest_2.50.0            \n",
       "[153] plotly_4.10.0               xtable_1.8-4               \n",
       "[155] jsonlite_1.7.2              R6_2.5.1                   \n",
       "[157] TFisher_0.2.0               Hmisc_4.6-0                \n",
       "[159] pillar_1.6.4                htmltools_0.5.2            \n",
       "[161] mime_0.12                   glue_1.5.0                 \n",
       "[163] fastmap_1.1.0               BiocParallel_1.28.0        \n",
       "[165] codetools_0.2-18            mvtnorm_1.1-3              \n",
       "[167] utf8_1.2.2                  lattice_0.20-45            \n",
       "[169] spatstat.sparse_2.0-0       tibble_3.1.6               \n",
       "[171] numDeriv_2016.8-1.1         curl_4.3.2                 \n",
       "[173] leiden_0.3.9                survival_3.2-13            \n",
       "[175] docopt_0.7.1                munsell_0.5.0              \n",
       "[177] GenomeInfoDbData_1.2.7      reshape2_1.4.4             \n",
       "[179] gtable_0.3.0                rbibutils_2.2.4            \n",
       "[181] spatstat.core_2.3-1        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2122599",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ceb7468",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
