{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "223d6f46",
   "metadata": {},
   "source": [
    "## Creates binarized Seurat objects, and merges them and finds Top Features (min.cutoff=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "235ae72f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Input info\n",
    "# cellr_in <- \"/data2/isshamie/mito_lineage/data/processed/mtscATAC/jan21_2021/MTblacklist\"\n",
    "# samples <- \"J2,P2\"\n",
    "# sample_names <- \"Flt3,Ctrl\"\n",
    "\n",
    "# # Saving\n",
    "# outdir <- \"/data/isshamie/mito_lineage/output/annotation/cd34norm/MTblacklist/mergedSamples\" #\"/data2/mito_lineage/Analysis/annotation/output/data/\"\n",
    "\n",
    "# # Parameters\n",
    "# nTop = 25000\n",
    "# assay=\"RNA\"\n",
    "\n",
    "# cores = 16\n",
    "\n",
    "# samples <- unlist(strsplit(samples, \",\")[[1]])\n",
    "# sample_names <- strsplit(sample_names, \",\")\n",
    "\n",
    "# samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0a97fcaa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-08-25 18:37:29--  https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi\n",
      "Resolving cf.10xgenomics.com (cf.10xgenomics.com)... 104.18.0.173, 104.18.1.173, 2606:4700::6812:ad, ...\n",
      "Connecting to cf.10xgenomics.com (cf.10xgenomics.com)|104.18.0.173|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1027204 (1003K) [binary/octet-stream]\n",
      "Saving to: ‘pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi’\n",
      "\n",
      "pbmc_granulocyte_so 100%[===================>]   1003K  --.-KB/s    in 0.1s    \n",
      "\n",
      "2021-08-25 18:37:30 (7.60 MB/s) - ‘pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi’ saved [1027204/1027204]\n",
      "\n",
      "--2021-08-25 18:37:30--  https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_10k_nextgem/atac_pbmc_10k_nextgem_fragments.tsv.gz\n",
      "Resolving cf.10xgenomics.com (cf.10xgenomics.com)... 104.18.0.173, 104.18.1.173, 2606:4700::6812:ad, ...\n",
      "Connecting to cf.10xgenomics.com (cf.10xgenomics.com)|104.18.0.173|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2173707527 (2.0G) [text/tab-separated-values]\n",
      "Saving to: ‘atac_pbmc_10k_nextgem_fragments.tsv.gz’\n",
      "\n",
      "atac_pbmc_10k_nextg 100%[===================>]   2.02G  41.2MB/s    in 59s     \n",
      "\n",
      "2021-08-25 18:38:30 (34.9 MB/s) - ‘atac_pbmc_10k_nextgem_fragments.tsv.gz’ saved [2173707527/2173707527]\n",
      "\n",
      "--2021-08-25 18:38:30--  https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_10k_nextgem/atac_pbmc_10k_nextgem_fragments.tsv.gz.tbi\n",
      "Resolving cf.10xgenomics.com (cf.10xgenomics.com)... 104.18.0.173, 104.18.1.173, 2606:4700::6812:ad, ...\n",
      "Connecting to cf.10xgenomics.com (cf.10xgenomics.com)|104.18.0.173|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1075367 (1.0M) [binary/octet-stream]\n",
      "Saving to: ‘atac_pbmc_10k_nextgem_fragments.tsv.gz.tbi’\n",
      "\n",
      "atac_pbmc_10k_nextg 100%[===================>]   1.03M  --.-KB/s    in 0.1s    \n",
      "\n",
      "2021-08-25 18:38:30 (8.75 MB/s) - ‘atac_pbmc_10k_nextgem_fragments.tsv.gz.tbi’ saved [1075367/1075367]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# # multiome\n",
    "# !wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5\n",
    "# #!wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz\n",
    "# !wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi\n",
    "# # scATAC\n",
    "# !wget https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_10k_nextgem/atac_pbmc_10k_nextgem_fragments.tsv.gz\n",
    "# !wget https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_10k_nextgem/atac_pbmc_10k_nextgem_fragments.tsv.gz.tbi\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20554931",
   "metadata": {},
   "source": [
    "\n",
    "## Preprocessing\n",
    "\n",
    "Here we'll load the PBMC multiome data pre-processed in our\n",
    "[multiome vignette](articles/pbmc_multiomic.html), and create a new object from\n",
    "the scATAC-seq data:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8705b530",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "07a86a75",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "Warning message in gzfile(file, \"rb\"):\n",
      "“cannot open compressed file '../vignette_data/pbmc_multiomic.rds', probable reason 'No such file or directory'”\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in gzfile(file, \"rb\"): cannot open the connection\n",
     "output_type": "error",
     "traceback": [
      "Error in gzfile(file, \"rb\"): cannot open the connection\nTraceback:\n",
      "1. readRDS(\"../vignette_data/pbmc_multiomic.rds\")",
      "2. gzfile(file, \"rb\")"
     ]
    }
   ],
   "source": [
    "\n",
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(ggplot2)\n",
    "# load the pre-processed multiome data\n",
    "pbmc.multi <- readRDS(\"../vignette_data/pbmc_multiomic.rds\")\n",
    "# process the scATAC data\n",
    "# first count fragments per cell\n",
    "fragpath <- \"~/data/10x_atac/pbmc10k_v2/atac_pbmc_10k_nextgem_fragments.tsv.gz\"\n",
    "fragcounts <- CountFragments(fragments = fragpath)\n",
    "atac.cells <- fragcounts[fragcounts$frequency_count > 2000, \"CB\"]\n",
    "# create the fragment object\n",
    "atac.frags <- CreateFragmentObject(path = fragpath, cells = atac.cells)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "356f651b",
   "metadata": {},
   "source": [
    "An important first step in any integrative analysis of single-cell chromatin data\n",
    "is to ensure that the same features are measured in each dataset. Here, we\n",
    "quantify the multiome peaks in the ATAC dataset to ensure that there are common features\n",
    "across the two datasets.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72ced4b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# quantify multiome peaks in the scATAC-seq dataset\n",
    "counts <- FeatureMatrix(\n",
    "  fragments = atac.frags,\n",
    "  features = granges(pbmc.multi),\n",
    "  cells = atac.cells\n",
    ")\n",
    "# create object\n",
    "atac.assay <- CreateChromatinAssay(\n",
    "  counts = counts,\n",
    "  min.features = 1000,\n",
    "  fragments = atac.frags\n",
    ")\n",
    "pbmc.atac <- CreateSeuratObject(counts = atac.assay, assay = \"peaks\")\n",
    "pbmc.atac <- subset(pbmc.atac, nCount_peaks > 2000 & nCount_peaks < 30000)\n",
    "# compute LSI\n",
    "pbmc.atac <- FindTopFeatures(pbmc.atac, min.cutoff = 10)\n",
    "pbmc.atac <- RunTFIDF(pbmc.atac)\n",
    "pbmc.atac <- RunSVD(pbmc.atac)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b23b642e",
   "metadata": {},
   "source": [
    "Next we can merge the multiome and scATAC datasets together and observe that\n",
    "there is a difference between them that appears to be due to the batch\n",
    "(experiment and technology-specific variation).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9ecddae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first add dataset-identifying metadata\n",
    "pbmc.atac$dataset <- \"ATAC\"\n",
    "pbmc.multi$dataset <- \"Multiome\"\n",
    "# merge\n",
    "pbmc.combined <- merge(pbmc.atac, pbmc.multi)\n",
    "# process the combined dataset\n",
    "pbmc.combined <- FindTopFeatures(pbmc.combined, min.cutoff = 10)\n",
    "pbmc.combined <- RunTFIDF(pbmc.combined)\n",
    "pbmc.combined <- RunSVD(pbmc.combined)\n",
    "pbmc.combined <- RunUMAP(pbmc.combined, reduction = \"lsi\", dims = 2:30)\n",
    "p1 <- DimPlot(pbmc.combined, group.by = \"dataset\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f86ce5a",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Integration\n",
    "To find integration anchors between the two datasets, we need to project them into\n",
    "a shared low-dimensional space. To do this, we'll use reciprocal LSI projection\n",
    "(projecting each dataset into the others LSI space) by setting `reduction=\"rlsi\"`.\n",
    "For more information about the data integration methods in Seurat, see our recent\n",
    "[paper](https://doi.org/10.1016/j.cell.2019.05.031)\n",
    "and the [Seurat website](https://satijalab.org/seurat/).\n",
    "\n",
    "Rather than integrating the normalized data matrix, as is typically done for \n",
    "scRNA-seq data, we'll integrate the low-dimensional cell embeddings (the LSI\n",
    "coordinates) across the datasets using the `IntegrateEmbeddings()` function.\n",
    "This is much better suited to scATAC-seq data,\n",
    "as we typically have a very sparse matrix with a large number of features. Note \n",
    "that this requires that we first compute an uncorrected LSI embedding using the\n",
    "merged dataset (as we did above).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "693b85b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "```{r message=FALSE, warning=FALSE}\n",
    "# find integration anchors\n",
    "integration.anchors <- FindIntegrationAnchors(\n",
    "  object.list = list(pbmc.multi, pbmc.atac),\n",
    "  anchor.features = rownames(pbmc.multi),\n",
    "  reduction = \"rlsi\",\n",
    "  dims = 2:30\n",
    ")\n",
    "# integrate LSI embeddings\n",
    "integrated <- IntegrateEmbeddings(\n",
    "  anchorset = integration.anchors,\n",
    "  reductions = pbmc.combined[[\"lsi\"]],\n",
    "  new.reduction.name = \"integrated_lsi\",\n",
    "  dims.to.integrate = 1:30\n",
    ")\n",
    "# create a new UMAP using the integrated embeddings\n",
    "integrated <- RunUMAP(integrated, reduction = \"integrated_lsi\", dims = 2:30)\n",
    "p2 <- DimPlot(integrated, group.by = \"dataset\")\n",
    "```\n",
    "\n",
    "Finally, we can compare the results of the merged and integrated datasets, and\n",
    "find that the integration has successfully removed the technology-specific variation\n",
    "in the dataset while retaining the cell-type-specific (biological) variation.\n",
    "\n",
    "```{r message=FALSE, warning=FALSE, fig.width=12}\n",
    "(p1 + ggtitle(\"Merged\")) | (p2 + ggtitle(\"Integrated\"))\n",
    "```\n",
    "\n",
    "Here we've demonstrated the integration method using two datasets, but the same\n",
    "workflow can be applied to integrate any number of datasets.\n",
    "\n",
    "## Reference mapping\n",
    "\n",
    "In cases where we have a large, high-quality dataset, or a dataset containing unique\n",
    "information not present in other datasets (cell type annotations or additional \n",
    "data modalities, for example), we often want to use that dataset as \n",
    "a reference and map queries onto it so that we can interpret these query datasets\n",
    "in the context of the existing reference.\n",
    "\n",
    "To demonstrate how to do this using single-cell chromatin reference and query\n",
    "datasets, we'll treat the PBMC multiome dataset here as a reference and map the\n",
    "scATAC-seq dataset to it using the `FindTransferAnchors()` and `MapQuery()`\n",
    "functions from Seurat.\n",
    "\n",
    "```{r message=FALSE, warning=FALSE}\n",
    "# compute UMAP and store the UMAP model\n",
    "pbmc.multi <- RunUMAP(pbmc.multi, reduction = \"lsi\", dims = 2:30, return.model = TRUE)\n",
    "# find transfer anchors\n",
    "transfer.anchors <- FindTransferAnchors(\n",
    "  reference = pbmc.multi,\n",
    "  query = pbmc.atac,\n",
    "  reference.reduction = \"lsi\",\n",
    "  reduction = \"lsiproject\",\n",
    "  dims = 2:30\n",
    ")\n",
    "# map query onto the reference dataset\n",
    "pbmc.atac <- MapQuery(\n",
    "  anchorset = transfer.anchors,\n",
    "  reference = pbmc.multi,\n",
    "  query = pbmc.atac,\n",
    "  refdata = pbmc.multi$predicted.id,\n",
    "  reference.reduction = \"lsi\",\n",
    "  new.reduction.name = \"ref.lsi\",\n",
    "  reduction.model = 'umap'\n",
    ")\n",
    "```\n",
    "\n",
    "<details>\n",
    "  <summary>**What is `MapQuery()` doing?**</summary>\n",
    "\n",
    "`MapQuery()` is a wrapper function that runs `TransferData()`, `IntegrateEmbeddings()`,\n",
    "and `ProjectUMAP()` for a query dataset, and sets sensible default parameters based\n",
    "on how the anchor object was generated. For finer control over the parameters used \n",
    "by each of these functions, you can pass parameters through `MapQuery()` to each function\n",
    "using the `transferdata.args`, `integrateembeddings.args`, and `projectumap.args` arguments\n",
    "for `MapQuery()`, or you can run each of the functions yourself. For example:\n",
    "\n",
    "```{r, eval=FALSE}\n",
    "pbmc.atac <- TransferData(\n",
    "  anchorset = transfer.anchors, \n",
    "  reference = pbmc.multi,\n",
    "  weight.reduction = \"lsiproject\",\n",
    "  query = pbmc.atac,\n",
    "  refdata = list(\n",
    "    celltype = \"predicted.id\",\n",
    "    predicted_RNA = \"RNA\")\n",
    ")\n",
    "pbmc.atac <- IntegrateEmbeddings(\n",
    "  anchorset = transfer.anchors,\n",
    "  reference = pbmc.multi,\n",
    "  query = pbmc.atac, \n",
    "  reductions = \"lsiproject\",\n",
    "  new.reduction.name = \"ref.lsi\"\n",
    ")\n",
    "pbmc.atac <- ProjectUMAP(\n",
    "  query = pbmc.atac, \n",
    "  query.reduction = \"ref.lsi\",\n",
    "  reference = pbmc.multi, \n",
    "  reference.reduction = \"lsi\",\n",
    "  reduction.model = \"umap\"\n",
    ")\n",
    "```\n",
    "\n",
    "</details>\n",
    "\n",
    "By running `MapQuery()`, we have mapped the scATAC-seq dataset onto the the\n",
    "multimodal reference, and enabled cell type labels to be transferred from reference\n",
    "to query. We can visualize these reference mapping results and the cell type \n",
    "labels now associated with the scATAC-seq dataset:\n",
    "\n",
    "```{r message=FALSE, warning=FALSE, fig.width=12, fig.height=6}\n",
    "p1 <- DimPlot(pbmc.multi, reduction = \"umap\", group.by = \"predicted.id\", label = TRUE, repel = TRUE) + NoLegend() + ggtitle(\"Reference\")\n",
    "p2 <- DimPlot(pbmc.atac, reduction = \"ref.umap\", group.by = \"predicted.id\", label = TRUE, repel = TRUE) + NoLegend() + ggtitle(\"Query\")\n",
    "p1 | p2\n",
    "```\n",
    "\n",
    "For more information about multimodal reference mapping, see the [Seurat vignette](https://satijalab.org/seurat/articles/multimodal_reference_mapping.html).\n",
    "\n",
    "## RNA imputation\n",
    "\n",
    "Above we transferred categorical information (the cell labels) and mapped the \n",
    "query data onto an existing reference UMAP. We can also transfer continuous data\n",
    "from the reference to the query in the same way. Here we demonstrate transferring\n",
    "the gene expression values from the PBMC multiome dataset (that measured DNA\n",
    "accessibility and gene expression in the same cells) to the PBMC scATAC-seq \n",
    "dataset (that measured DNA accessibility only). Note that we could also transfer\n",
    "these values using the `MapQuery()` function call above by setting the `refdata`\n",
    "parameter to a list of values.\n",
    "\n",
    "```{r message=FALSE, warning=FALSE}\n",
    "# predict gene expression values\n",
    "rna <- TransferData(\n",
    "  anchorset = transfer.anchors,\n",
    "  refdata = GetAssayData(pbmc.multi, assay = \"RNA\", slot = \"data\"),\n",
    "  weight.reduction = pbmc.atac[[\"lsi\"]],\n",
    "  dims = 2:30\n",
    ")\n",
    "# add predicted values as a new assay\n",
    "pbmc.atac[[\"predicted\"]] <- rna\n",
    "```\n",
    "\n",
    "We can look at some immune marker genes and see that the predicted expression\n",
    "patterns match our expectation based on known expression patterns.\n",
    "\n",
    "```{r message=FALSE, warning=FALSE, fig.width=12, fig.height=8}\n",
    "DefaultAssay(pbmc.atac) <- \"predicted\"\n",
    "FeaturePlot(\n",
    "  object = pbmc.atac,\n",
    "  features = c('MS4A1', 'CD3D', 'LEF1', 'NKG7', 'TREM1', 'LYZ'),\n",
    "  pt.size = 0.1,\n",
    "  max.cutoff = 'q95',\n",
    "  reduction = \"ref.umap\",\n",
    "  ncol = 3\n",
    ")\n",
    "```\n",
    "\n",
    "```{r include=FALSE}\n",
    "saveRDS(object = pbmc.atac, file = \"../vignette_data/pbmc_atac_integration.rds\")\n",
    "```\n",
    "\n",
    "<details>\n",
    "  <summary>**Session Info**</summary>\n",
    "  \n",
    "```{r}\n",
    "sessionInfo()\n",
    "```\n",
    "\n",
    "</details>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:mito_trace] *",
   "language": "R",
   "name": "conda-env-mito_trace-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
