{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-13T20:12:12.096903Z",
     "start_time": "2021-05-13T20:12:12.086904Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/md0/isshamie/Projects/Mito_Trace/workflow/notebooks/clone_af_dendrograms'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-13T20:12:12.105680Z",
     "start_time": "2021-05-13T20:12:12.099105Z"
    },
    "scrolled": true,
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "# INDIR = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/\"\n",
    "# OUTDIR = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/barcodes/btwnClones_dendro_dt_0.6/\"\n",
    "# DONOR = 0\n",
    "# dendroThresh=0.6\n",
    "\n",
    "#clones_meta = '/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/results/clones_summary/feb23_2022/clones_summary.csv'\n",
    "#clones_indir = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/\"\n",
    "#clone_sizes_in_f = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/gff_A2_black/annotation_clones/clone_counts/minCellConds_1/clone_sizes.csv\"\n",
    "#AF_INDIR = \"/data/Mito_Trace/output/pipeline/v04/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/enriched_barcodes/clones/variants_init/knn/kparam_30/donor0/\"\n",
    "\n",
    "INDIR = \"/data/Mito_Trace/output/pipeline/v04/CHIP_Input_nameFix_april08_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/enriched_barcodes/objs_high_cells_constraints_min_both_pct_0.5/clones/variants_prefilterMerge_impute/knn/kparam_30\"\n",
    "OUTDIR = \"/data/Mito_Trace/output/pipeline/v04/CHIP_Input_nameFix_april08_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/enriched_barcodes/objs_high_cells_constraints_min_both_pct_0.5/clones/variants_prefilterMerge_impute/knn/kparam_30/barcodes/btwnClones_dendro\"\n",
    "DONOR = 3\n",
    "\n",
    "\n",
    "# INDIR = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30\"\n",
    "# OUTDIR = \"/data/Mito_Trace/output/pipeline/v02/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_30/barcodes/btwnClones_dendro_dt_0.6\"\n",
    "# DONOR = 1\n",
    "#dendroThresh = 0.6\n",
    "\n",
    "verbose=False\n",
    "\n",
    "mt_ref=\"/data/Mito_Trace/data/processed/genomes/mtMasked/GRCh38_MT_blacklist_A2_2020/chrM.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-13T20:12:13.155526Z",
     "start_time": "2021-05-13T20:12:12.130665Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fig_utils from mplh\n",
      "0.5.6\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import mplh.cluster_help as ch\n",
    "import mplh.fig_utils as fu\n",
    "\n",
    "from os.path import join, exists, dirname\n",
    "from glob import glob\n",
    "import pickle\n",
    "import os\n",
    "import vireoSNP\n",
    "import numpy as np\n",
    "from scipy import sparse\n",
    "from scipy.io import mmread\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import hypergeom\n",
    "print(vireoSNP.__version__)\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from vireoSNP import Vireo\n",
    "np.set_printoptions(formatter={'float': lambda x: format(x, '.5f')})\n",
    "import scipy\n",
    "from itertools import combinations\n",
    "#from src.utils import utils  as ut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.utils.dendro_clust import run_dynamic\n",
    "import src.utils.variant_utils as vu\n",
    "import src.utils.dendro_clust as dc\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_f = join(OUTDIR, f\"donor{DONOR}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load cells_meta \n",
    "### Create clones_meta - get number of cells in a clone and donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_ATAC</th>\n",
       "      <th>nFeature_ATAC</th>\n",
       "      <th>total</th>\n",
       "      <th>duplicate</th>\n",
       "      <th>chimeric</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>lowmapq</th>\n",
       "      <th>mitochondrial</th>\n",
       "      <th>passed_filters</th>\n",
       "      <th>...</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>ID</th>\n",
       "      <th>Row.names</th>\n",
       "      <th>lineage</th>\n",
       "      <th>donor</th>\n",
       "      <th>condition</th>\n",
       "      <th>name</th>\n",
       "      <th>cluster_labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Input_AAACGAATCCCAGTAA-1</th>\n",
       "      <td>Input</td>\n",
       "      <td>2103.0</td>\n",
       "      <td>2103.0</td>\n",
       "      <td>30056.0</td>\n",
       "      <td>2764.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>832.0</td>\n",
       "      <td>22385.0</td>\n",
       "      <td>3888.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2540.0</td>\n",
       "      <td>2215.0</td>\n",
       "      <td>AAACGAATCCCAGTAA-1_Input</td>\n",
       "      <td>Input_AAACGAATCCCAGTAA-1</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>Input</td>\n",
       "      <td>0_25</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Input_AAACGAATCGTGAACT-1</th>\n",
       "      <td>Input</td>\n",
       "      <td>5996.0</td>\n",
       "      <td>5996.0</td>\n",
       "      <td>66340.0</td>\n",
       "      <td>10716.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>165.0</td>\n",
       "      <td>2305.0</td>\n",
       "      <td>39978.0</td>\n",
       "      <td>12814.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8689.0</td>\n",
       "      <td>5714.0</td>\n",
       "      <td>AAACGAATCGTGAACT-1_Input</td>\n",
       "      <td>Input_AAACGAATCGTGAACT-1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>Input</td>\n",
       "      <td>0_2</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Input_AAACTCGAGGTACACG-1</th>\n",
       "      <td>Input</td>\n",
       "      <td>4901.0</td>\n",
       "      <td>4901.0</td>\n",
       "      <td>49202.0</td>\n",
       "      <td>9164.0</td>\n",
       "      <td>265.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>1810.0</td>\n",
       "      <td>27219.0</td>\n",
       "      <td>10628.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6092.0</td>\n",
       "      <td>4446.0</td>\n",
       "      <td>AAACTCGAGGTACACG-1_Input</td>\n",
       "      <td>Input_AAACTCGAGGTACACG-1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>Input</td>\n",
       "      <td>0_4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Input_AAACTGCAGGATGTAT-1</th>\n",
       "      <td>Input</td>\n",
       "      <td>3421.0</td>\n",
       "      <td>3421.0</td>\n",
       "      <td>34970.0</td>\n",
       "      <td>5801.0</td>\n",
       "      <td>165.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>972.0</td>\n",
       "      <td>22347.0</td>\n",
       "      <td>5598.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3701.0</td>\n",
       "      <td>2984.0</td>\n",
       "      <td>AAACTGCAGGATGTAT-1_Input</td>\n",
       "      <td>Input_AAACTGCAGGATGTAT-1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>Input</td>\n",
       "      <td>0_4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Input_AAACTGCAGTCTCTAG-1</th>\n",
       "      <td>Input</td>\n",
       "      <td>3129.0</td>\n",
       "      <td>3129.0</td>\n",
       "      <td>31794.0</td>\n",
       "      <td>6835.0</td>\n",
       "      <td>210.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>1208.0</td>\n",
       "      <td>13249.0</td>\n",
       "      <td>10202.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5219.0</td>\n",
       "      <td>3174.0</td>\n",
       "      <td>AAACTGCAGTCTCTAG-1_Input</td>\n",
       "      <td>Input_AAACTGCAGTCTCTAG-1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>Input</td>\n",
       "      <td>0_2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_ATGTCTTGTGTCGTGC-1</th>\n",
       "      <td>Flt3l</td>\n",
       "      <td>8759.0</td>\n",
       "      <td>8759.0</td>\n",
       "      <td>155564.0</td>\n",
       "      <td>35841.0</td>\n",
       "      <td>999.0</td>\n",
       "      <td>490.0</td>\n",
       "      <td>3409.0</td>\n",
       "      <td>101832.0</td>\n",
       "      <td>12993.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11376.0</td>\n",
       "      <td>6631.0</td>\n",
       "      <td>ATGTCTTGTGTCGTGC-1_Flt3l</td>\n",
       "      <td>Flt3l_ATGTCTTGTGTCGTGC-1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>4_3</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_GTTGGTAAGCGAGAAA-1</th>\n",
       "      <td>Flt3l</td>\n",
       "      <td>7417.0</td>\n",
       "      <td>7417.0</td>\n",
       "      <td>159745.0</td>\n",
       "      <td>37109.0</td>\n",
       "      <td>832.0</td>\n",
       "      <td>641.0</td>\n",
       "      <td>3911.0</td>\n",
       "      <td>106447.0</td>\n",
       "      <td>10805.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9416.0</td>\n",
       "      <td>5770.0</td>\n",
       "      <td>GTTGGTAAGCGAGAAA-1_Flt3l</td>\n",
       "      <td>Flt3l_GTTGGTAAGCGAGAAA-1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>4_2</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_TGGGTGCAGTACCACT-1</th>\n",
       "      <td>Flt3l</td>\n",
       "      <td>8502.0</td>\n",
       "      <td>8502.0</td>\n",
       "      <td>187037.0</td>\n",
       "      <td>52900.0</td>\n",
       "      <td>784.0</td>\n",
       "      <td>1007.0</td>\n",
       "      <td>5934.0</td>\n",
       "      <td>112887.0</td>\n",
       "      <td>13525.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11193.0</td>\n",
       "      <td>6675.0</td>\n",
       "      <td>TGGGTGCAGTACCACT-1_Flt3l</td>\n",
       "      <td>Flt3l_TGGGTGCAGTACCACT-1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>4_3</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_TGGGTGCCAAGTTGCT-1</th>\n",
       "      <td>Flt3l</td>\n",
       "      <td>8723.0</td>\n",
       "      <td>8723.0</td>\n",
       "      <td>189721.0</td>\n",
       "      <td>45559.0</td>\n",
       "      <td>793.0</td>\n",
       "      <td>732.0</td>\n",
       "      <td>4374.0</td>\n",
       "      <td>124486.0</td>\n",
       "      <td>13777.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10792.0</td>\n",
       "      <td>6534.0</td>\n",
       "      <td>TGGGTGCCAAGTTGCT-1_Flt3l</td>\n",
       "      <td>Flt3l_TGGGTGCCAAGTTGCT-1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>4_3</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_TTTGAGGGTTGCAGAG-1</th>\n",
       "      <td>Flt3l</td>\n",
       "      <td>11111.0</td>\n",
       "      <td>11111.0</td>\n",
       "      <td>182438.0</td>\n",
       "      <td>70126.0</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>878.0</td>\n",
       "      <td>9648.0</td>\n",
       "      <td>79591.0</td>\n",
       "      <td>20956.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>16707.0</td>\n",
       "      <td>8735.0</td>\n",
       "      <td>TTTGAGGGTTGCAGAG-1_Flt3l</td>\n",
       "      <td>Flt3l_TTTGAGGGTTGCAGAG-1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>4_3</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>718 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         orig.ident  nCount_ATAC  nFeature_ATAC     total  \\\n",
       "Input_AAACGAATCCCAGTAA-1      Input       2103.0         2103.0   30056.0   \n",
       "Input_AAACGAATCGTGAACT-1      Input       5996.0         5996.0   66340.0   \n",
       "Input_AAACTCGAGGTACACG-1      Input       4901.0         4901.0   49202.0   \n",
       "Input_AAACTGCAGGATGTAT-1      Input       3421.0         3421.0   34970.0   \n",
       "Input_AAACTGCAGTCTCTAG-1      Input       3129.0         3129.0   31794.0   \n",
       "...                             ...          ...            ...       ...   \n",
       "Flt3l_ATGTCTTGTGTCGTGC-1      Flt3l       8759.0         8759.0  155564.0   \n",
       "Flt3l_GTTGGTAAGCGAGAAA-1      Flt3l       7417.0         7417.0  159745.0   \n",
       "Flt3l_TGGGTGCAGTACCACT-1      Flt3l       8502.0         8502.0  187037.0   \n",
       "Flt3l_TGGGTGCCAAGTTGCT-1      Flt3l       8723.0         8723.0  189721.0   \n",
       "Flt3l_TTTGAGGGTTGCAGAG-1      Flt3l      11111.0        11111.0  182438.0   \n",
       "\n",
       "                          duplicate  chimeric  unmapped  lowmapq  \\\n",
       "Input_AAACGAATCCCAGTAA-1     2764.0     124.0      63.0    832.0   \n",
       "Input_AAACGAATCGTGAACT-1    10716.0     362.0     165.0   2305.0   \n",
       "Input_AAACTCGAGGTACACG-1     9164.0     265.0     116.0   1810.0   \n",
       "Input_AAACTGCAGGATGTAT-1     5801.0     165.0      87.0    972.0   \n",
       "Input_AAACTGCAGTCTCTAG-1     6835.0     210.0      90.0   1208.0   \n",
       "...                             ...       ...       ...      ...   \n",
       "Flt3l_ATGTCTTGTGTCGTGC-1    35841.0     999.0     490.0   3409.0   \n",
       "Flt3l_GTTGGTAAGCGAGAAA-1    37109.0     832.0     641.0   3911.0   \n",
       "Flt3l_TGGGTGCAGTACCACT-1    52900.0     784.0    1007.0   5934.0   \n",
       "Flt3l_TGGGTGCCAAGTTGCT-1    45559.0     793.0     732.0   4374.0   \n",
       "Flt3l_TTTGAGGGTTGCAGAG-1    70126.0    1239.0     878.0   9648.0   \n",
       "\n",
       "                          mitochondrial  passed_filters  ... seurat_clusters  \\\n",
       "Input_AAACGAATCCCAGTAA-1        22385.0          3888.0  ...             2.0   \n",
       "Input_AAACGAATCGTGAACT-1        39978.0         12814.0  ...             4.0   \n",
       "Input_AAACTCGAGGTACACG-1        27219.0         10628.0  ...             1.0   \n",
       "Input_AAACTGCAGGATGTAT-1        22347.0          5598.0  ...             1.0   \n",
       "Input_AAACTGCAGTCTCTAG-1        13249.0         10202.0  ...             2.0   \n",
       "...                                 ...             ...  ...             ...   \n",
       "Flt3l_ATGTCTTGTGTCGTGC-1       101832.0         12993.0  ...             9.0   \n",
       "Flt3l_GTTGGTAAGCGAGAAA-1       106447.0         10805.0  ...             9.0   \n",
       "Flt3l_TGGGTGCAGTACCACT-1       112887.0         13525.0  ...             9.0   \n",
       "Flt3l_TGGGTGCCAAGTTGCT-1       124486.0         13777.0  ...             9.0   \n",
       "Flt3l_TTTGAGGGTTGCAGAG-1        79591.0         20956.0  ...             9.0   \n",
       "\n",
       "                          nCount_RNA  nFeature_RNA                        ID  \\\n",
       "Input_AAACGAATCCCAGTAA-1      2540.0        2215.0  AAACGAATCCCAGTAA-1_Input   \n",
       "Input_AAACGAATCGTGAACT-1      8689.0        5714.0  AAACGAATCGTGAACT-1_Input   \n",
       "Input_AAACTCGAGGTACACG-1      6092.0        4446.0  AAACTCGAGGTACACG-1_Input   \n",
       "Input_AAACTGCAGGATGTAT-1      3701.0        2984.0  AAACTGCAGGATGTAT-1_Input   \n",
       "Input_AAACTGCAGTCTCTAG-1      5219.0        3174.0  AAACTGCAGTCTCTAG-1_Input   \n",
       "...                              ...           ...                       ...   \n",
       "Flt3l_ATGTCTTGTGTCGTGC-1     11376.0        6631.0  ATGTCTTGTGTCGTGC-1_Flt3l   \n",
       "Flt3l_GTTGGTAAGCGAGAAA-1      9416.0        5770.0  GTTGGTAAGCGAGAAA-1_Flt3l   \n",
       "Flt3l_TGGGTGCAGTACCACT-1     11193.0        6675.0  TGGGTGCAGTACCACT-1_Flt3l   \n",
       "Flt3l_TGGGTGCCAAGTTGCT-1     10792.0        6534.0  TGGGTGCCAAGTTGCT-1_Flt3l   \n",
       "Flt3l_TTTGAGGGTTGCAGAG-1     16707.0        8735.0  TTTGAGGGTTGCAGAG-1_Flt3l   \n",
       "\n",
       "                                         Row.names  lineage  donor  condition  \\\n",
       "Input_AAACGAATCCCAGTAA-1  Input_AAACGAATCCCAGTAA-1       25      0      Input   \n",
       "Input_AAACGAATCGTGAACT-1  Input_AAACGAATCGTGAACT-1        2      0      Input   \n",
       "Input_AAACTCGAGGTACACG-1  Input_AAACTCGAGGTACACG-1        4      0      Input   \n",
       "Input_AAACTGCAGGATGTAT-1  Input_AAACTGCAGGATGTAT-1        4      0      Input   \n",
       "Input_AAACTGCAGTCTCTAG-1  Input_AAACTGCAGTCTCTAG-1        2      0      Input   \n",
       "...                                            ...      ...    ...        ...   \n",
       "Flt3l_ATGTCTTGTGTCGTGC-1  Flt3l_ATGTCTTGTGTCGTGC-1        3      4      Flt3l   \n",
       "Flt3l_GTTGGTAAGCGAGAAA-1  Flt3l_GTTGGTAAGCGAGAAA-1        2      4      Flt3l   \n",
       "Flt3l_TGGGTGCAGTACCACT-1  Flt3l_TGGGTGCAGTACCACT-1        3      4      Flt3l   \n",
       "Flt3l_TGGGTGCCAAGTTGCT-1  Flt3l_TGGGTGCCAAGTTGCT-1        3      4      Flt3l   \n",
       "Flt3l_TTTGAGGGTTGCAGAG-1  Flt3l_TTTGAGGGTTGCAGAG-1        3      4      Flt3l   \n",
       "\n",
       "                          name  cluster_labels  \n",
       "Input_AAACGAATCCCAGTAA-1  0_25             2.0  \n",
       "Input_AAACGAATCGTGAACT-1   0_2             4.0  \n",
       "Input_AAACTCGAGGTACACG-1   0_4             1.0  \n",
       "Input_AAACTGCAGGATGTAT-1   0_4             1.0  \n",
       "Input_AAACTGCAGTCTCTAG-1   0_2             2.0  \n",
       "...                        ...             ...  \n",
       "Flt3l_ATGTCTTGTGTCGTGC-1   4_3             9.0  \n",
       "Flt3l_GTTGGTAAGCGAGAAA-1   4_2             9.0  \n",
       "Flt3l_TGGGTGCAGTACCACT-1   4_3             9.0  \n",
       "Flt3l_TGGGTGCCAAGTTGCT-1   4_3             9.0  \n",
       "Flt3l_TTTGAGGGTTGCAGAG-1   4_3             9.0  \n",
       "\n",
       "[718 rows x 40 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cells_meta = pd.read_csv(join(INDIR, \"cells_meta.tsv\"), sep=\"\\t\", index_col=0)\n",
    "\n",
    "cells_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "cells_meta[\"lineage\"] = cells_meta.apply(lambda x: f'{x[\"donor\"]}_{x[\"lineage\"]}', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lineage</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0_10</th>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_11</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_12</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_13</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_14</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_15</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_16</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_17</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_18</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_19</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_2</th>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_20</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_21</th>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_22</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_23</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_24</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_25</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_26</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_27</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_28</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_29</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_3</th>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_30</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_4</th>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_5</th>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_6</th>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_7</th>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_8</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_9</th>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_13</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_15</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_16</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_19</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_28</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_32</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_33</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_10</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_16</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_7</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_0</th>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_2</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_3</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         size\n",
       "lineage      \n",
       "0_10       20\n",
       "0_11       10\n",
       "0_12        9\n",
       "0_13        7\n",
       "0_14        3\n",
       "0_15        2\n",
       "0_16        2\n",
       "0_17        9\n",
       "0_18        3\n",
       "0_19        9\n",
       "0_2       106\n",
       "0_20        2\n",
       "0_21       96\n",
       "0_22        2\n",
       "0_23        4\n",
       "0_24        2\n",
       "0_25       14\n",
       "0_26        5\n",
       "0_27        2\n",
       "0_28        2\n",
       "0_29        3\n",
       "0_3        72\n",
       "0_30        2\n",
       "0_4        52\n",
       "0_5        42\n",
       "0_6        40\n",
       "0_7        35\n",
       "0_8        30\n",
       "0_9        20\n",
       "1_13        2\n",
       "1_15        1\n",
       "1_16        2\n",
       "1_19        1\n",
       "1_28        2\n",
       "1_32        1\n",
       "1_33        1\n",
       "2_10        2\n",
       "2_16        2\n",
       "2_7         2\n",
       "3_0        88\n",
       "4_2         3\n",
       "4_3         6"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clones_meta_df = pd.DataFrame(cells_meta.groupby(\"lineage\").size(), columns=[\"size\"]) #, index=[\"size\"])\n",
    "clones_meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>size</th>\n",
       "      <th>donor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lineage</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0_10</th>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_11</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_12</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_13</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_14</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_15</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_16</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_17</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_18</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_19</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_2</th>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_20</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_21</th>\n",
       "      <td>96</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_22</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_23</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_24</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_25</th>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_26</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_27</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_28</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_29</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_3</th>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_30</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_4</th>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_5</th>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_6</th>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_7</th>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_8</th>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0_9</th>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_13</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_15</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_16</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_19</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_28</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_32</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_33</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_10</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_16</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2_7</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3_0</th>\n",
       "      <td>88</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_2</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4_3</th>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         size  donor\n",
       "lineage             \n",
       "0_10       20      0\n",
       "0_11       10      0\n",
       "0_12        9      0\n",
       "0_13        7      0\n",
       "0_14        3      0\n",
       "0_15        2      0\n",
       "0_16        2      0\n",
       "0_17        9      0\n",
       "0_18        3      0\n",
       "0_19        9      0\n",
       "0_2       106      0\n",
       "0_20        2      0\n",
       "0_21       96      0\n",
       "0_22        2      0\n",
       "0_23        4      0\n",
       "0_24        2      0\n",
       "0_25       14      0\n",
       "0_26        5      0\n",
       "0_27        2      0\n",
       "0_28        2      0\n",
       "0_29        3      0\n",
       "0_3        72      0\n",
       "0_30        2      0\n",
       "0_4        52      0\n",
       "0_5        42      0\n",
       "0_6        40      0\n",
       "0_7        35      0\n",
       "0_8        30      0\n",
       "0_9        20      0\n",
       "1_13        2      1\n",
       "1_15        1      1\n",
       "1_16        2      1\n",
       "1_19        1      1\n",
       "1_28        2      1\n",
       "1_32        1      1\n",
       "1_33        1      1\n",
       "2_10        2      2\n",
       "2_16        2      2\n",
       "2_7         2      2\n",
       "3_0        88      3\n",
       "4_2         3      4\n",
       "4_3         6      4"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clones_meta_df[\"donor\"] = [x.split(\"_\")[0] for x in clones_meta_df.index]\n",
    "clones_meta_df[\"donor\"] = clones_meta_df[\"donor\"].astype(int)\n",
    "clones_meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "curr_clones = clones_meta_df[clones_meta_df[\"donor\"]==DONOR].copy()\n",
    "# cells_meta = pd.read_csv(join(INDIR, \"cells_meta.tsv\"), sep='\\t')#.sort_values([\"donor\", \"lineage\"])\n",
    "curr_labels = cells_meta[cells_meta[\"donor\"]==DONOR].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_ATAC</th>\n",
       "      <th>nFeature_ATAC</th>\n",
       "      <th>total</th>\n",
       "      <th>duplicate</th>\n",
       "      <th>chimeric</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>lowmapq</th>\n",
       "      <th>mitochondrial</th>\n",
       "      <th>passed_filters</th>\n",
       "      <th>...</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>ID</th>\n",
       "      <th>Row.names</th>\n",
       "      <th>lineage</th>\n",
       "      <th>donor</th>\n",
       "      <th>condition</th>\n",
       "      <th>name</th>\n",
       "      <th>cluster_labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAGGATGTGCTGTGC-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACAAAGTCTTACGGA-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACCGATAGGTCGGTA-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACGTACGTATTCGCA-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATGGCTAGTGCAACG-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGGGTTAGTGGACAGT-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTAGGTGTCCATCTAT-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCAACTGTGCCGGTT-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCATCAGTGGAAAGA-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGAGGAGTCTCGAT-1_Flt3l</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_0</td>\n",
       "      <td>3</td>\n",
       "      <td>Flt3l</td>\n",
       "      <td>3_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         orig.ident  nCount_ATAC  nFeature_ATAC  total  \\\n",
       "AAAGGATGTGCTGTGC-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "AACAAAGTCTTACGGA-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "AACCGATAGGTCGGTA-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "AACGTACGTATTCGCA-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "AATGGCTAGTGCAACG-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "...                             ...          ...            ...    ...   \n",
       "TGGGTTAGTGGACAGT-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "TTAGGTGTCCATCTAT-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "TTCAACTGTGCCGGTT-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "TTCATCAGTGGAAAGA-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "TTTGAGGAGTCTCGAT-1_Flt3l        NaN          NaN            NaN    NaN   \n",
       "\n",
       "                          duplicate  chimeric  unmapped  lowmapq  \\\n",
       "AAAGGATGTGCTGTGC-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "AACAAAGTCTTACGGA-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "AACCGATAGGTCGGTA-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "AACGTACGTATTCGCA-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "AATGGCTAGTGCAACG-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "...                             ...       ...       ...      ...   \n",
       "TGGGTTAGTGGACAGT-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "TTAGGTGTCCATCTAT-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "TTCAACTGTGCCGGTT-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "TTCATCAGTGGAAAGA-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "TTTGAGGAGTCTCGAT-1_Flt3l        NaN       NaN       NaN      NaN   \n",
       "\n",
       "                          mitochondrial  passed_filters  ... seurat_clusters  \\\n",
       "AAAGGATGTGCTGTGC-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "AACAAAGTCTTACGGA-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "AACCGATAGGTCGGTA-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "AACGTACGTATTCGCA-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "AATGGCTAGTGCAACG-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "...                                 ...             ...  ...             ...   \n",
       "TGGGTTAGTGGACAGT-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "TTAGGTGTCCATCTAT-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "TTCAACTGTGCCGGTT-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "TTCATCAGTGGAAAGA-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "TTTGAGGAGTCTCGAT-1_Flt3l            NaN             NaN  ...             NaN   \n",
       "\n",
       "                          nCount_RNA  nFeature_RNA   ID  Row.names  lineage  \\\n",
       "AAAGGATGTGCTGTGC-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "AACAAAGTCTTACGGA-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "AACCGATAGGTCGGTA-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "AACGTACGTATTCGCA-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "AATGGCTAGTGCAACG-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "...                              ...           ...  ...        ...      ...   \n",
       "TGGGTTAGTGGACAGT-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "TTAGGTGTCCATCTAT-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "TTCAACTGTGCCGGTT-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "TTCATCAGTGGAAAGA-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "TTTGAGGAGTCTCGAT-1_Flt3l         NaN           NaN  NaN        NaN      3_0   \n",
       "\n",
       "                          donor  condition  name  cluster_labels  \n",
       "AAAGGATGTGCTGTGC-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "AACAAAGTCTTACGGA-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "AACCGATAGGTCGGTA-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "AACGTACGTATTCGCA-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "AATGGCTAGTGCAACG-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "...                         ...        ...   ...             ...  \n",
       "TGGGTTAGTGGACAGT-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "TTAGGTGTCCATCTAT-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "TTCAACTGTGCCGGTT-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "TTCATCAGTGGAAAGA-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "TTTGAGGAGTCTCGAT-1_Flt3l      3      Flt3l   3_0             NaN  \n",
       "\n",
       "[88 rows x 40 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curr_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# curr_clones = clones_meta_df[clones_meta_df[\"donor\"]==DONOR].set_index(\"lineage\")\n",
    "# curr_clones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load AF and DP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "af_indir = INDIR #join(AF_INDIR) #, \"sc_af\", f\"donor{DONOR}\")\n",
    "AF_df = pd.read_csv(join(af_indir, \"af.tsv\"), index_col=0, sep=\"\\t\").transpose()\n",
    "DP_df = pd.read_csv(join(af_indir, \"dp.tsv\"), index_col=0, sep=\"\\t\").transpose()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACGAATCCCAGTAA-1_Input</th>\n",
       "      <th>AAACGAATCGTGAACT-1_Input</th>\n",
       "      <th>AAACTCGAGGTACACG-1_Input</th>\n",
       "      <th>AAACTGCAGGATGTAT-1_Input</th>\n",
       "      <th>AAACTGCAGTCTCTAG-1_Input</th>\n",
       "      <th>AAACTGCTCCTCCCAA-1_Input</th>\n",
       "      <th>AAAGATGCATGGGACA-1_Input</th>\n",
       "      <th>AAAGGATGTGATGTGG-1_Input</th>\n",
       "      <th>AAAGGGCAGGCCAGTA-1_Input</th>\n",
       "      <th>AAAGGGCCACGATTAC-1_Input</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGAGGAGTCTCGAT-1_Flt3l</th>\n",
       "      <th>ACGTGGCAGCGATACG-1_Input</th>\n",
       "      <th>CTACTTATCGGGAATG-1_Input</th>\n",
       "      <th>TGGCCTTCAGTATCTG-1_Input</th>\n",
       "      <th>TGGGTTAGTCATTGCA-1_Input</th>\n",
       "      <th>ATGTCTTGTGTCGTGC-1_Flt3l</th>\n",
       "      <th>GTTGGTAAGCGAGAAA-1_Flt3l</th>\n",
       "      <th>TGGGTGCAGTACCACT-1_Flt3l</th>\n",
       "      <th>TGGGTGCCAAGTTGCT-1_Flt3l</th>\n",
       "      <th>TTTGAGGGTTGCAGAG-1_Flt3l</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10398G</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.016878</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10400T</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.017699</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.040816</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1040C</th>\n",
       "      <td>0.02381</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017937</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10559G</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1082G</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11204C</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16519C</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4117C</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11719A</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015025</td>\n",
       "      <td>0.004747</td>\n",
       "      <td>0.028714</td>\n",
       "      <td>0.013624</td>\n",
       "      <td>0.008435</td>\n",
       "      <td>0.001738</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012029</td>\n",
       "      <td>0.013947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2706G</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.019678</td>\n",
       "      <td>0.003490</td>\n",
       "      <td>0.065848</td>\n",
       "      <td>0.005935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035242</td>\n",
       "      <td>0.010922</td>\n",
       "      <td>0.009331</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 718 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        AAACGAATCCCAGTAA-1_Input  AAACGAATCGTGAACT-1_Input  \\\n",
       "10398G                   0.00000                  0.016878   \n",
       "10400T                   0.00000                  0.017699   \n",
       "1040C                    0.02381                  0.000000   \n",
       "10559G                   0.00000                  0.000000   \n",
       "1082G                    0.00000                  0.000000   \n",
       "...                          ...                       ...   \n",
       "11204C                   0.00000                  0.000000   \n",
       "16519C                   0.00000                  0.000000   \n",
       "4117C                    0.00000                  0.000000   \n",
       "11719A                   0.00000                  0.000000   \n",
       "2706G                    0.00000                  0.000000   \n",
       "\n",
       "        AAACTCGAGGTACACG-1_Input  AAACTGCAGGATGTAT-1_Input  \\\n",
       "10398G                       0.0                       0.0   \n",
       "10400T                       0.0                       0.0   \n",
       "1040C                        0.0                       0.0   \n",
       "10559G                       0.0                       0.0   \n",
       "1082G                        0.0                       0.0   \n",
       "...                          ...                       ...   \n",
       "11204C                       0.0                       0.0   \n",
       "16519C                       0.0                       0.0   \n",
       "4117C                        0.0                       0.0   \n",
       "11719A                       0.0                       0.0   \n",
       "2706G                        0.0                       0.0   \n",
       "\n",
       "        AAACTGCAGTCTCTAG-1_Input  AAACTGCTCCTCCCAA-1_Input  \\\n",
       "10398G                  0.000000                       0.0   \n",
       "10400T                  0.040816                       0.0   \n",
       "1040C                   0.000000                       0.0   \n",
       "10559G                  0.000000                       0.0   \n",
       "1082G                   0.000000                       0.0   \n",
       "...                          ...                       ...   \n",
       "11204C                  0.000000                       0.0   \n",
       "16519C                  0.000000                       0.0   \n",
       "4117C                   0.000000                       0.0   \n",
       "11719A                  0.000000                       0.0   \n",
       "2706G                   0.000000                       0.0   \n",
       "\n",
       "        AAAGATGCATGGGACA-1_Input  AAAGGATGTGATGTGG-1_Input  \\\n",
       "10398G                       0.0                       0.0   \n",
       "10400T                       0.0                       0.0   \n",
       "1040C                        0.0                       0.0   \n",
       "10559G                       0.0                       0.0   \n",
       "1082G                        0.0                       0.0   \n",
       "...                          ...                       ...   \n",
       "11204C                       0.0                       0.0   \n",
       "16519C                       0.0                       0.0   \n",
       "4117C                        0.0                       0.0   \n",
       "11719A                       0.0                       0.0   \n",
       "2706G                        0.0                       0.0   \n",
       "\n",
       "        AAAGGGCAGGCCAGTA-1_Input  AAAGGGCCACGATTAC-1_Input  ...  \\\n",
       "10398G                  0.000000                       0.0  ...   \n",
       "10400T                  0.000000                       0.0  ...   \n",
       "1040C                   0.017937                       0.0  ...   \n",
       "10559G                  0.019417                       0.0  ...   \n",
       "1082G                   0.000000                       0.0  ...   \n",
       "...                          ...                       ...  ...   \n",
       "11204C                  0.000000                       0.0  ...   \n",
       "16519C                  0.000000                       0.0  ...   \n",
       "4117C                   0.000000                       0.0  ...   \n",
       "11719A                  0.000000                       0.0  ...   \n",
       "2706G                   0.000000                       0.0  ...   \n",
       "\n",
       "        TTTGAGGAGTCTCGAT-1_Flt3l  ACGTGGCAGCGATACG-1_Input  \\\n",
       "10398G                       0.0                  0.000000   \n",
       "10400T                       0.0                  0.000000   \n",
       "1040C                        0.0                  0.000000   \n",
       "10559G                       0.0                  0.000000   \n",
       "1082G                        0.0                  0.000000   \n",
       "...                          ...                       ...   \n",
       "11204C                       0.0                  0.000000   \n",
       "16519C                       0.0                  0.000000   \n",
       "4117C                        0.0                  0.000000   \n",
       "11719A                       0.0                  0.015025   \n",
       "2706G                        0.0                  0.019678   \n",
       "\n",
       "        CTACTTATCGGGAATG-1_Input  TGGCCTTCAGTATCTG-1_Input  \\\n",
       "10398G                  0.000000                  0.000000   \n",
       "10400T                  0.000000                  0.000000   \n",
       "1040C                   0.000000                  0.000000   \n",
       "10559G                  0.000000                  0.000000   \n",
       "1082G                   0.000000                  0.000000   \n",
       "...                          ...                       ...   \n",
       "11204C                  0.000000                  0.000000   \n",
       "16519C                  0.000000                  0.000000   \n",
       "4117C                   0.000000                  0.000000   \n",
       "11719A                  0.004747                  0.028714   \n",
       "2706G                   0.003490                  0.065848   \n",
       "\n",
       "        TGGGTTAGTCATTGCA-1_Input  ATGTCTTGTGTCGTGC-1_Flt3l  \\\n",
       "10398G                  0.000000                  0.000000   \n",
       "10400T                  0.000000                  0.000000   \n",
       "1040C                   0.000000                  0.000000   \n",
       "10559G                  0.000000                  0.000000   \n",
       "1082G                   0.000000                  0.000000   \n",
       "...                          ...                       ...   \n",
       "11204C                  0.000000                  0.000000   \n",
       "16519C                  0.000000                  0.000000   \n",
       "4117C                   0.000000                  0.000000   \n",
       "11719A                  0.013624                  0.008435   \n",
       "2706G                   0.005935                  0.000000   \n",
       "\n",
       "        GTTGGTAAGCGAGAAA-1_Flt3l  TGGGTGCAGTACCACT-1_Flt3l  \\\n",
       "10398G                  0.000000                  0.000000   \n",
       "10400T                  0.000000                  0.000000   \n",
       "1040C                   0.000000                  0.000000   \n",
       "10559G                  0.000000                  0.000000   \n",
       "1082G                   0.000000                  0.000000   \n",
       "...                          ...                       ...   \n",
       "11204C                  0.000000                  0.000000   \n",
       "16519C                  0.000000                  0.000000   \n",
       "4117C                   0.000000                  0.000000   \n",
       "11719A                  0.001738                  0.000000   \n",
       "2706G                   0.035242                  0.010922   \n",
       "\n",
       "        TGGGTGCCAAGTTGCT-1_Flt3l  TTTGAGGGTTGCAGAG-1_Flt3l  \n",
       "10398G                  0.000000                  0.000000  \n",
       "10400T                  0.000000                  0.000000  \n",
       "1040C                   0.000000                  0.000000  \n",
       "10559G                  0.000000                  0.000000  \n",
       "1082G                   0.000000                  0.000000  \n",
       "...                          ...                       ...  \n",
       "11204C                  0.000000                  0.000000  \n",
       "16519C                  0.000000                  0.000000  \n",
       "4117C                   0.000000                  0.000000  \n",
       "11719A                  0.012029                  0.013947  \n",
       "2706G                   0.009331                  0.000000  \n",
       "\n",
       "[68 rows x 718 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AF_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions to get mean and median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_mean(cells, af):\n",
    "    print(cells.head())\n",
    "    return af.loc[:,cells[\"ID\"].values].mean(axis=1)\n",
    "\n",
    "def calc_median(cells, af):\n",
    "    #print(cells.head())\n",
    "    return af.loc[:,cells[\"ID\"].values].median(axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate clone-barcode mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "55"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(AF_df == 0).all().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID not set\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Flt3l_AAAGGATGTGCTGTGC-1</th>\n",
       "      <td>AAAGGATGTGCTGTGC-1_Flt3l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_AACAAAGTCTTACGGA-1</th>\n",
       "      <td>AACAAAGTCTTACGGA-1_Flt3l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_AACCGATAGGTCGGTA-1</th>\n",
       "      <td>AACCGATAGGTCGGTA-1_Flt3l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_AACGTACGTATTCGCA-1</th>\n",
       "      <td>AACGTACGTATTCGCA-1_Flt3l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_AATGGCTAGTGCAACG-1</th>\n",
       "      <td>AATGGCTAGTGCAACG-1_Flt3l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_TGGGTTAGTGGACAGT-1</th>\n",
       "      <td>TGGGTTAGTGGACAGT-1_Flt3l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_TTAGGTGTCCATCTAT-1</th>\n",
       "      <td>TTAGGTGTCCATCTAT-1_Flt3l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_TTCAACTGTGCCGGTT-1</th>\n",
       "      <td>TTCAACTGTGCCGGTT-1_Flt3l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_TTCATCAGTGGAAAGA-1</th>\n",
       "      <td>TTCATCAGTGGAAAGA-1_Flt3l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flt3l_TTTGAGGAGTCTCGAT-1</th>\n",
       "      <td>TTTGAGGAGTCTCGAT-1_Flt3l</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                ID\n",
       "Flt3l_AAAGGATGTGCTGTGC-1  AAAGGATGTGCTGTGC-1_Flt3l\n",
       "Flt3l_AACAAAGTCTTACGGA-1  AACAAAGTCTTACGGA-1_Flt3l\n",
       "Flt3l_AACCGATAGGTCGGTA-1  AACCGATAGGTCGGTA-1_Flt3l\n",
       "Flt3l_AACGTACGTATTCGCA-1  AACGTACGTATTCGCA-1_Flt3l\n",
       "Flt3l_AATGGCTAGTGCAACG-1  AATGGCTAGTGCAACG-1_Flt3l\n",
       "...                                            ...\n",
       "Flt3l_TGGGTTAGTGGACAGT-1  TGGGTTAGTGGACAGT-1_Flt3l\n",
       "Flt3l_TTAGGTGTCCATCTAT-1  TTAGGTGTCCATCTAT-1_Flt3l\n",
       "Flt3l_TTCAACTGTGCCGGTT-1  TTCAACTGTGCCGGTT-1_Flt3l\n",
       "Flt3l_TTCATCAGTGGAAAGA-1  TTCATCAGTGGAAAGA-1_Flt3l\n",
       "Flt3l_TTTGAGGAGTCTCGAT-1  TTTGAGGAGTCTCGAT-1_Flt3l\n",
       "\n",
       "[88 rows x 1 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def inverse_string(x):\n",
    "    spl = x.split(\"_\")\n",
    "    return f\"{spl[1]}_{spl[0]}\"\n",
    "if curr_labels[\"ID\"].isnull().all():\n",
    "    print(\"ID not set\")\n",
    "    curr_labels[\"ID\"] = curr_labels.index\n",
    "    curr_labels.index = [inverse_string(x) for x in curr_labels.index]\n",
    "    #curr_labels[\"ID\"] = [inverse_string(x) for x in curr_labels.index]\n",
    "curr_labels[[\"ID\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                         orig.ident  nCount_ATAC  nFeature_ATAC  total  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1        NaN          NaN            NaN    NaN   \n",
      "Flt3l_AACAAAGTCTTACGGA-1        NaN          NaN            NaN    NaN   \n",
      "Flt3l_AACCGATAGGTCGGTA-1        NaN          NaN            NaN    NaN   \n",
      "Flt3l_AACGTACGTATTCGCA-1        NaN          NaN            NaN    NaN   \n",
      "Flt3l_AATGGCTAGTGCAACG-1        NaN          NaN            NaN    NaN   \n",
      "\n",
      "                          duplicate  chimeric  unmapped  lowmapq  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1        NaN       NaN       NaN      NaN   \n",
      "Flt3l_AACAAAGTCTTACGGA-1        NaN       NaN       NaN      NaN   \n",
      "Flt3l_AACCGATAGGTCGGTA-1        NaN       NaN       NaN      NaN   \n",
      "Flt3l_AACGTACGTATTCGCA-1        NaN       NaN       NaN      NaN   \n",
      "Flt3l_AATGGCTAGTGCAACG-1        NaN       NaN       NaN      NaN   \n",
      "\n",
      "                          mitochondrial  passed_filters  ... seurat_clusters  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1            NaN             NaN  ...             NaN   \n",
      "Flt3l_AACAAAGTCTTACGGA-1            NaN             NaN  ...             NaN   \n",
      "Flt3l_AACCGATAGGTCGGTA-1            NaN             NaN  ...             NaN   \n",
      "Flt3l_AACGTACGTATTCGCA-1            NaN             NaN  ...             NaN   \n",
      "Flt3l_AATGGCTAGTGCAACG-1            NaN             NaN  ...             NaN   \n",
      "\n",
      "                          nCount_RNA  nFeature_RNA                        ID  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1         NaN           NaN  AAAGGATGTGCTGTGC-1_Flt3l   \n",
      "Flt3l_AACAAAGTCTTACGGA-1         NaN           NaN  AACAAAGTCTTACGGA-1_Flt3l   \n",
      "Flt3l_AACCGATAGGTCGGTA-1         NaN           NaN  AACCGATAGGTCGGTA-1_Flt3l   \n",
      "Flt3l_AACGTACGTATTCGCA-1         NaN           NaN  AACGTACGTATTCGCA-1_Flt3l   \n",
      "Flt3l_AATGGCTAGTGCAACG-1         NaN           NaN  AATGGCTAGTGCAACG-1_Flt3l   \n",
      "\n",
      "                          Row.names  lineage  donor  condition  name  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1        NaN      3_0      3      Flt3l   3_0   \n",
      "Flt3l_AACAAAGTCTTACGGA-1        NaN      3_0      3      Flt3l   3_0   \n",
      "Flt3l_AACCGATAGGTCGGTA-1        NaN      3_0      3      Flt3l   3_0   \n",
      "Flt3l_AACGTACGTATTCGCA-1        NaN      3_0      3      Flt3l   3_0   \n",
      "Flt3l_AATGGCTAGTGCAACG-1        NaN      3_0      3      Flt3l   3_0   \n",
      "\n",
      "                          cluster_labels  \n",
      "Flt3l_AAAGGATGTGCTGTGC-1             NaN  \n",
      "Flt3l_AACAAAGTCTTACGGA-1             NaN  \n",
      "Flt3l_AACCGATAGGTCGGTA-1             NaN  \n",
      "Flt3l_AACGTACGTATTCGCA-1             NaN  \n",
      "Flt3l_AATGGCTAGTGCAACG-1             NaN  \n",
      "\n",
      "[5 rows x 40 columns]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2291T</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lineage</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3_0</th>\n",
       "      <td>0.003272</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            2291T\n",
       "lineage          \n",
       "3_0      0.003272"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_af = curr_labels.groupby([\"lineage\"]).apply(calc_mean, af=AF_df.fillna(0))\n",
    "mean_af = mean_af.loc[~((mean_af == 0).all(axis=1)),~((mean_af == 0).all(axis=0))]\n",
    "\n",
    "mean_af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                         orig.ident  nCount_ATAC  nFeature_ATAC  total  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1        NaN          NaN            NaN    NaN   \n",
      "Flt3l_AACAAAGTCTTACGGA-1        NaN          NaN            NaN    NaN   \n",
      "Flt3l_AACCGATAGGTCGGTA-1        NaN          NaN            NaN    NaN   \n",
      "Flt3l_AACGTACGTATTCGCA-1        NaN          NaN            NaN    NaN   \n",
      "Flt3l_AATGGCTAGTGCAACG-1        NaN          NaN            NaN    NaN   \n",
      "\n",
      "                          duplicate  chimeric  unmapped  lowmapq  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1        NaN       NaN       NaN      NaN   \n",
      "Flt3l_AACAAAGTCTTACGGA-1        NaN       NaN       NaN      NaN   \n",
      "Flt3l_AACCGATAGGTCGGTA-1        NaN       NaN       NaN      NaN   \n",
      "Flt3l_AACGTACGTATTCGCA-1        NaN       NaN       NaN      NaN   \n",
      "Flt3l_AATGGCTAGTGCAACG-1        NaN       NaN       NaN      NaN   \n",
      "\n",
      "                          mitochondrial  passed_filters  ... seurat_clusters  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1            NaN             NaN  ...             NaN   \n",
      "Flt3l_AACAAAGTCTTACGGA-1            NaN             NaN  ...             NaN   \n",
      "Flt3l_AACCGATAGGTCGGTA-1            NaN             NaN  ...             NaN   \n",
      "Flt3l_AACGTACGTATTCGCA-1            NaN             NaN  ...             NaN   \n",
      "Flt3l_AATGGCTAGTGCAACG-1            NaN             NaN  ...             NaN   \n",
      "\n",
      "                          nCount_RNA  nFeature_RNA                        ID  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1         NaN           NaN  AAAGGATGTGCTGTGC-1_Flt3l   \n",
      "Flt3l_AACAAAGTCTTACGGA-1         NaN           NaN  AACAAAGTCTTACGGA-1_Flt3l   \n",
      "Flt3l_AACCGATAGGTCGGTA-1         NaN           NaN  AACCGATAGGTCGGTA-1_Flt3l   \n",
      "Flt3l_AACGTACGTATTCGCA-1         NaN           NaN  AACGTACGTATTCGCA-1_Flt3l   \n",
      "Flt3l_AATGGCTAGTGCAACG-1         NaN           NaN  AATGGCTAGTGCAACG-1_Flt3l   \n",
      "\n",
      "                          Row.names  lineage  donor  condition  name  \\\n",
      "Flt3l_AAAGGATGTGCTGTGC-1        NaN      3_0      3      Flt3l   3_0   \n",
      "Flt3l_AACAAAGTCTTACGGA-1        NaN      3_0      3      Flt3l   3_0   \n",
      "Flt3l_AACCGATAGGTCGGTA-1        NaN      3_0      3      Flt3l   3_0   \n",
      "Flt3l_AACGTACGTATTCGCA-1        NaN      3_0      3      Flt3l   3_0   \n",
      "Flt3l_AATGGCTAGTGCAACG-1        NaN      3_0      3      Flt3l   3_0   \n",
      "\n",
      "                          cluster_labels  \n",
      "Flt3l_AAAGGATGTGCTGTGC-1             NaN  \n",
      "Flt3l_AACAAAGTCTTACGGA-1             NaN  \n",
      "Flt3l_AACCGATAGGTCGGTA-1             NaN  \n",
      "Flt3l_AACGTACGTATTCGCA-1             NaN  \n",
      "Flt3l_AATGGCTAGTGCAACG-1             NaN  \n",
      "\n",
      "[5 rows x 40 columns]\n"
     ]
    }
   ],
   "source": [
    "mean_dp = curr_labels.groupby([\"lineage\"]).apply(calc_mean, af=np.log2(DP_df+1))\n",
    "mean_dp = mean_dp.loc[mean_af.index, mean_af.columns]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2291T</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lineage</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3_0</th>\n",
       "      <td>0.003272</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            2291T\n",
       "lineage          \n",
       "3_0      0.003272"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno = mean_af.reset_index()[[\"lineage\"]]\n",
    "\n",
    "# Remove clones that have all 0\n",
    "# print('mean_af shape', mean_af.shape)\n",
    "# mean_af = mean_af.loc[~((mean_af==0).all(axis=1))]\n",
    "# curr_clones = curr_clones.loc[mean_af.index]\n",
    "# print('mean_af shape after rm 0', mean_af.shape)\n",
    "\n",
    "\n",
    "#mean_dp = mean_dp.reset_index(drop=True)\n",
    "anno_dp = mean_dp.reset_index()[[\"lineage\"]]\n",
    "na_cells = curr_labels.loc[curr_labels[\"lineage\"].isnull()]\n",
    "na_mean_af = na_cells.groupby(\"condition\").apply(calc_mean, af=AF_df)\n",
    "\n",
    "if not (na_mean_af.shape[0] == 0):\n",
    "    print('na shape', na_mean_af.shape)\n",
    "    na_anno = pd.concat((anno,pd.DataFrame({\"lineage\":[\"None\"]*len(na_mean_af.index)})), ignore_index=True)\n",
    "    na_concat = pd.concat((mean_af.reset_index(drop=True), na_mean_af.reset_index(drop=True)), ignore_index=True)\n",
    "    na_mean_dp = na_cells.groupby(\"condition\").apply(calc_mean, af=np.log2(DP_df+1))\n",
    "    na_dp_concat = pd.concat((mean_dp, na_mean_dp.reset_index(drop=True)), ignore_index=True)\n",
    "else:\n",
    "    na_anno = anno\n",
    "    na_concat = mean_af.reset_index(drop=True)\n",
    "    na_dp_concat = mean_dp\n",
    "     \n",
    "mean_af  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get variant types\n",
    "variants_df = vu.add_ref_to_variants(variants=mean_af.columns.values,\n",
    "                                     mt_df=vu.load_mt_ref(mt_ref))\n",
    "\n",
    "variants_df = vu.type_of_variants(variants_df, to_preproc=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cluster "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "if not (mean_af.shape[0] == 1 or  mean_af.shape[1] == 1 ):\n",
    "    #den = dc.dendro_cluster(mean_af, g, d_thresh=0.6)\n",
    "    \n",
    "    #clusters, link = run_dynamic(mean_af, metric='euclidean', method=\"average\", minClusterSize=1)\n",
    "    clusters, link = dc.run_dynamic_hyper(np.sqrt(mean_af))\n",
    "#     is_comp = False\n",
    "#     for i in ([3,2,1]):\n",
    "#         try:\n",
    "#             clusters, link = run_dynamic(mean_af, metric='cosine', method=\"complete\", \n",
    "#                                          minClusterSize=i, deepSplit=3)\n",
    "#             is_comp = True\n",
    "#             break\n",
    "#         except IndexError:\n",
    "#             print(f'didnt work for {i}')\n",
    "#     print(\"min cluster size:\", i)\n",
    "    \n",
    "#     if not is_comp:\n",
    "#         print('min size 1 and deepsplit 1')\n",
    "#         clusters, link = run_dynamic(mean_af, metric='cosine', method=\"complete\", \n",
    "#                                      minClusterSize=1, deepSplit=1)\n",
    "        \n",
    "    curr_clones[\"den_clust\"] = curr_clones.apply(lambda x: f'C{clusters.loc[x.name, \"labels\"]}', axis=1)\n",
    "\n",
    "    size_pvals = dc.cluster_stats(curr_clones, out_f= out_f )\n",
    "\n",
    "    curr_clones[\"donor\"]= [x.split(\"_\")[0] for x in curr_clones.index]\n",
    "    curr_clones = curr_clones.loc[mean_af.index]\n",
    "\n",
    "    # with open(out_f + \".dendrogram_pvals.txt\", \"w\") as f:\n",
    "    #     #for l in size_pvals:\n",
    "    #     f.write(\"\\n\".join(size_pvals))\n",
    "\n",
    "    curr_clones\n",
    "else:\n",
    "    with open(out_f + \".dendrogram_pvals.txt\", \"w\") as f:\n",
    "        f.write(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot with dendro clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "conditions = cells_meta[\"condition\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "if not (mean_af.shape[0] == 1 or  mean_af.shape[1] == 1 ):\n",
    "    g = ch.plot_cluster(np.sqrt(mean_af),row_meta=curr_clones, to_row_clust=True, to_col_clust=True,\n",
    "                        col_meta=variants_df[[\"variant type\", \"variant change\"]], col_clr_schemes=\"categorical\",\n",
    "                        row_clr_schemes={\"size\":\"sequential\",\"donor\":\"categorical\",\"den_clust\":\"categorical\"}, \n",
    "                        vmax=0.2, vmin=0, row_linkage=link, metric=\"cosine\", method=\"complete\")\n",
    "    plt.suptitle(\"Donor sqrt allele-frequency heatmap with clone and condition labels (max AF=0.2)\")\n",
    "    #fu.helper_save(out_f+\"withEnrichmentFoldChange.NoCondition.clust.max2.AF.png\",to_pdf=False)\n",
    "    fu.helper_save(out_f+\".dendro.NoCondition.max2.AF.png\",to_svg=True, to_pdf=False)\n",
    "\n",
    "    inds = g.dendrogram_row.dendrogram[\"leaves\"]\n",
    "    cols = g.dendrogram_col.dendrogram[\"leaves\"]\n",
    "\n",
    "    g1d = ch.plot_cluster(mean_dp.iloc[inds, cols], row_meta=curr_clones, to_row_clust=False, to_col_clust=False,\n",
    "                          row_clr_schemes={\"size\":\"sequential\",\"donor\":\"categorical\",\"den_clust\":\"categorical\"}, \n",
    "                          col_meta=variants_df[[\"variant type\", \"variant change\"]], col_clr_schemes=\"categorical\",\n",
    "                         )\n",
    "    g1d.ax_cbar.set_title(\"log2(reads+1)\")\n",
    "    fu.helper_save(out_f+\".dendro.NoCondition.DP.png\", to_svg=True)#, f=g1d.figure)\n",
    "\n",
    "\n",
    "    mean_af = mean_af.iloc[inds,cols]\n",
    "    curr_clones = curr_clones.loc[mean_af.index]\n",
    "\n",
    "    ch.plot_cluster(np.sqrt(mean_af),row_meta=curr_clones, to_row_clust=False, to_col_clust=False,\n",
    "                    col_meta=variants_df[[\"variant type\", \"variant change\"]], col_clr_schemes=\"categorical\",\n",
    "                    row_clr_schemes={\"size\":\"sequential\",\"donor\":\"categorical\",\"den_clust\":\"categorical\"}, \n",
    "                    vmax=0.5, vmin=0)\n",
    "    plt.suptitle(\"Donor sqrt allele-frequency heatmap with clone and condition labels (max AF=0.5)\")\n",
    "    #fu.helper_save(out_f+\"withEnrichmentFoldChange.NoCondition.clust.max5.AF.png\",to_pdf=False)\n",
    "    fu.helper_save(out_f+\".dendro.NoCondition.max5.AF.png\",to_svg=True, to_pdf=False)\n",
    "\n",
    "else:\n",
    "    plt.figure()\n",
    "    fu.helper_save(out_f+\".dendro.NoCondition.max5.AF.png\",to_svg=True, to_pdf=False)\n",
    "    fu.helper_save(out_f+\".dendro.NoCondition.DP.png\", to_svg=True)#, f=g1d.figure)\n",
    "    fu.helper_save(out_f+\".dendro.NoCondition.max2.AF.png\",to_svg=True, to_pdf=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save mean_af and curr_clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_af.to_csv(out_f+\".mean.csv\")\n",
    "curr_clones.to_csv(out_f+\".clones_dendro.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/Mito_Trace/output/pipeline/v04/CHIP_Input_nameFix_april08_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/enriched_barcodes/objs_high_cells_constraints_min_both_pct_0.5/clones/variants_prefilterMerge_impute/knn/kparam_30/barcodes/btwnClones_dendro/donor3'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_f"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 5560.364131,
   "end_time": "2021-01-27T10:03:31.625455",
   "environment_variables": {},
   "exception": null,
   "input_path": "/data2/mito_lineage/src/vireo/vireoSNP_clones.ipynb",
   "output_path": "results/jan21_2021/chrM/P2_cellSNP_minC200_minAF0.01/lineage_chrM.ipynb",
   "parameters": {
    "AD_F": "data/jan21_2021/chrM/P2_cellSNP_minC200_minAF0.01/cellSNP.tag.AD.mtx",
    "DP_F": "data/jan21_2021/chrM/P2_cellSNP_minC200_minAF0.01/cellSNP.tag.DP.mtx"
   },
   "start_time": "2021-01-27T08:30:51.261324",
   "version": "2.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
