{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trying to understand how to properly process data from /data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim, which is amplified by MT primers and done with 5'-end sequencing with 10x genomics. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Libraries and files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:16:03.028652Z",
     "start_time": "2020-04-27T00:16:02.453066Z"
    }
   },
   "outputs": [],
   "source": [
    "from os.path import basename, join\n",
    "import os\n",
    "import yaml\n",
    "import pandas as pd\n",
    "from glob import glob\n",
    "import pysam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:16:03.044641Z",
     "start_time": "2020-04-27T00:16:03.030745Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'samples': 'parameters/samples.tsv',\n",
       " 'raw_folder': '/data2/isshamie/mito_lineage/data/raw/',\n",
       " 'dataset': '200403_A00953_0090_BHC23FDSXY',\n",
       " 'primers_list': '/data2/mito_lineage/BWA-Primers-MT/final_primers_with_tag.csv',\n",
       " 'adapter': 'AAGCAGTGGTATCAACGCAGAGTAC',\n",
       " 'ref_fa': '/data2/genome/human_GRCh38/cellranger/refdata-cellranger-GRCh38-3.0.0/fasta/genome.fa',\n",
       " 'transcriptome_dir': 'transcriptome=/data2/genome/human_GRCh38/cellranger/refdata-cellranger-GRCh38-3.0.0/'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "params = \"../parameters/trim_adaptors.yaml\"\n",
    "params = yaml.load(open(params), Loader=yaml.FullLoader)\n",
    "params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:16:03.055272Z",
     "start_time": "2020-04-27T00:16:03.046766Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/bc5prime_sc5p_r2_A_1/outs/possorted_genome_bam.bam\n"
     ]
    }
   ],
   "source": [
    "proc_dir = \"/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/\"\n",
    "#sample_name = \"bc5prime_sc5p_pe_lanes\"\n",
    "sample_name = \"bc5prime_sc5p_r2_A_1\"\n",
    "data = join(proc_dir, sample_name,\"outs/\")\n",
    "bam_f = glob(join(data, \"*.bam\"))[0]\n",
    "print(bam_f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:16:03.062139Z",
     "start_time": "2020-04-27T00:16:03.057146Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/picard/bc5prime_sc5p_r2_A_1.txt\n"
     ]
    }
   ],
   "source": [
    "outdir = join(proc_dir, \"picard\") #, sample_name)\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)\n",
    "    \n",
    "out_f = join(outdir, sample_name+'.txt')\n",
    "print(out_f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## After trimming the adaptor and running cellranger, check the bam file (which had no alignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A. looking at using the -SC5P R2 in 10x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check Picard stats on bam file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:16:03.071808Z",
     "start_time": "2020-04-27T00:16:03.063816Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "picard CollectAlignmentSummaryMetrics R=/data2/genome/human_GRCh38/cellranger/refdata-cellranger-GRCh38-3.0.0/fasta/genome.fa I=/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/bc5prime_sc5p_r2_A_1/outs/possorted_genome_bam.bam O=/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/picard/bc5prime_sc5p_r2_A_1.txt\n",
      "command already done\n"
     ]
    }
   ],
   "source": [
    "cmd = f\"picard CollectAlignmentSummaryMetrics \\\n",
    "R={params['ref_fa']} \\\n",
    "I={bam_f} \\\n",
    "O={out_f}\"\n",
    "print(cmd)\n",
    "\n",
    "if not os.path.exists(out_f):\n",
    "    os.system(cmd)\n",
    "else:\n",
    "    print('command already done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:16:03.142240Z",
     "start_time": "2020-04-27T00:16:03.073460Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CATEGORY</th>\n",
       "      <td>UNPAIRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TOTAL_READS</th>\n",
       "      <td>14207734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_READS</th>\n",
       "      <td>14207734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCT_PF_READS</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_NOISE_READS</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_READS_ALIGNED</th>\n",
       "      <td>51330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCT_PF_READS_ALIGNED</th>\n",
       "      <td>0.003613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_ALIGNED_BASES</th>\n",
       "      <td>4535922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_HQ_ALIGNED_READS</th>\n",
       "      <td>13418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_HQ_ALIGNED_BASES</th>\n",
       "      <td>1146468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_HQ_ALIGNED_Q20_BASES</th>\n",
       "      <td>696117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_HQ_MEDIAN_MISMATCHES</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_MISMATCH_RATE</th>\n",
       "      <td>0.016652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_HQ_ERROR_RATE</th>\n",
       "      <td>0.042517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_INDEL_RATE</th>\n",
       "      <td>8.7e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEAN_READ_LENGTH</th>\n",
       "      <td>77.0082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>READS_ALIGNED_IN_PAIRS</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCT_READS_ALIGNED_IN_PAIRS</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF_READS_IMPROPER_PAIRS</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCT_PF_READS_IMPROPER_PAIRS</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BAD_CYCLES</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRAND_BALANCE</th>\n",
       "      <td>0.17968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCT_CHIMERAS</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCT_ADAPTER</th>\n",
       "      <td>0.000292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIBRARY</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>READ_GROUP</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    0\n",
       "CATEGORY                     UNPAIRED\n",
       "TOTAL_READS                  14207734\n",
       "PF_READS                     14207734\n",
       "PCT_PF_READS                        1\n",
       "PF_NOISE_READS                      0\n",
       "PF_READS_ALIGNED                51330\n",
       "PCT_PF_READS_ALIGNED         0.003613\n",
       "PF_ALIGNED_BASES              4535922\n",
       "PF_HQ_ALIGNED_READS             13418\n",
       "PF_HQ_ALIGNED_BASES           1146468\n",
       "PF_HQ_ALIGNED_Q20_BASES        696117\n",
       "PF_HQ_MEDIAN_MISMATCHES             0\n",
       "PF_MISMATCH_RATE             0.016652\n",
       "PF_HQ_ERROR_RATE             0.042517\n",
       "PF_INDEL_RATE                 8.7e-05\n",
       "MEAN_READ_LENGTH              77.0082\n",
       "READS_ALIGNED_IN_PAIRS              0\n",
       "PCT_READS_ALIGNED_IN_PAIRS          0\n",
       "PF_READS_IMPROPER_PAIRS             0\n",
       "PCT_PF_READS_IMPROPER_PAIRS         0\n",
       "BAD_CYCLES                          0\n",
       "STRAND_BALANCE                0.17968\n",
       "PCT_CHIMERAS                        0\n",
       "PCT_ADAPTER                  0.000292\n",
       "SAMPLE                            NaN\n",
       "LIBRARY                           NaN\n",
       "READ_GROUP                        NaN"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "align_df = pd.read_csv(out_f, sep=\"\\t\",comment=\"#\")\n",
    "align_df.transpose()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in portion of sam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:16:03.148859Z",
     "start_time": "2020-04-27T00:16:03.144022Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A00953:90:HC23FDSXY:3:1559:18982:5400\t272\t0\t379486\t3\t9M253923N20M\t-1\t-1\t29\tTCTAACCATGGCCCGTATTTACCCTATAG\tarray('B', [37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 25, 37, 37, 37])\t[('NH', 2), ('HI', 2), ('AS', 21), ('nM', 0), ('RE', 'I'), ('li', 0), ('CR', 'CACCAAGCACCGAAAA'), ('CY', ',,FFFFFF:FFFFFFF'), ('UR', 'TACCCTAAAT'), ('UY', 'FFFFFFFFF:'), ('UB', 'TACCCTAAAT'), ('RG', 'bc5prime_sc5p_r2_A_1:0:1:HC23FDSXY:3')]\n"
     ]
    }
   ],
   "source": [
    "samfile = pysam.AlignmentFile(join(data,'tmp_100.sam'))\n",
    "\n",
    "for read in samfile.fetch():\n",
    "    print(read)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:52:23.705097Z",
     "start_time": "2020-04-27T00:52:23.660819Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>...</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:1559:18982:5400</td>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "      <td>379487</td>\n",
       "      <td>3</td>\n",
       "      <td>9M253923N20M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TCTAACCATGGCCCGTATTTACCCTATAG</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:21</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:CACCAAGCACCGAAAA</td>\n",
       "      <td>CY:Z:,,FFFFFF:FFFFFFF</td>\n",
       "      <td>UR:Z:TACCCTAAAT</td>\n",
       "      <td>UY:Z:FFFFFFFFF:</td>\n",
       "      <td>UB:Z:TACCCTAAAT</td>\n",
       "      <td>RG:Z:bc5prime_sc5p_r2_A_1:0:1:HC23FDSXY:3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:1559:18982:5400</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>607673</td>\n",
       "      <td>3</td>\n",
       "      <td>9M25737N20M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TCTAACCATGGCCCGTATTTACCCTATAG</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:21</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>xf:i:0</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:CACCAAGCACCGAAAA</td>\n",
       "      <td>CY:Z:,,FFFFFF:FFFFFFF</td>\n",
       "      <td>UR:Z:TACCCTAAAT</td>\n",
       "      <td>UY:Z:FFFFFFFFF:</td>\n",
       "      <td>UB:Z:TACCCTAAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:2608:15130:23218</td>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "      <td>629113</td>\n",
       "      <td>0</td>\n",
       "      <td>8S20M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTCTGCCTATCGAATACGCCGCAGGCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:19</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:CTGCCTATCGAATACG</td>\n",
       "      <td>CY:Z::FFFFFFFFFFFFFFF</td>\n",
       "      <td>UR:Z:CCGCAGGCCC</td>\n",
       "      <td>UY:Z:FFFFFFFFFF</td>\n",
       "      <td>UB:Z:CCGCAGGCCC</td>\n",
       "      <td>RG:Z:bc5prime_sc5p_r2_A_1:0:1:HC23FDSXY:3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:2608:13169:23484</td>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "      <td>629113</td>\n",
       "      <td>0</td>\n",
       "      <td>8S20M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTCTGCCTATCGAATACGCCGCAGGCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:19</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:CTGCCTATCGAATACG</td>\n",
       "      <td>CY:Z::FFFFFFFFFFFFFFF</td>\n",
       "      <td>UR:Z:CCGCAGGCCC</td>\n",
       "      <td>UY:Z:FFFFFFFFFF</td>\n",
       "      <td>UB:Z:CCGCAGGCCC</td>\n",
       "      <td>RG:Z:bc5prime_sc5p_r2_A_1:0:1:HC23FDSXY:3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:1668:20066:23343</td>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "      <td>629118</td>\n",
       "      <td>0</td>\n",
       "      <td>15M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ATACGCCGCAGGCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:14</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:TTTAAACAAAAAATAA</td>\n",
       "      <td>CY:Z:FFF:,:FFFF,:F::F</td>\n",
       "      <td>UR:Z:AAAACCCATC</td>\n",
       "      <td>UY:Z::FFFFFFFFF</td>\n",
       "      <td>UB:Z:AAAACCCATC</td>\n",
       "      <td>RG:Z:bc5prime_sc5p_r2_A_1:0:1:HC23FDSXY:3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:2619:22869:1297</td>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "      <td>631081</td>\n",
       "      <td>0</td>\n",
       "      <td>101M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CCGACCGTTGACTATTCTCTACAAACCACAAAGACATTGGAACACT...</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:99</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:AACTCTTGTCCTAGCG</td>\n",
       "      <td>CY:Z:FFFFFFFFFFFFFFFF</td>\n",
       "      <td>CB:Z:AACTCTTGTCCTAGCG-1</td>\n",
       "      <td>UR:Z:ACGAGCAAGG</td>\n",
       "      <td>UY:Z:FFFFFFFFFF</td>\n",
       "      <td>UB:Z:ACGAGCAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:1307:20980:30217</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>631084</td>\n",
       "      <td>3</td>\n",
       "      <td>101M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ACCGTTGACTATTCTCTACAAACCACAAAGACATTGGAACACTATA...</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:99</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>xf:i:0</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:GCCAAATCATTGGGCC</td>\n",
       "      <td>CY:Z:FFFFFFFFFFFFFFFF</td>\n",
       "      <td>CB:Z:GCCAAATCATTGGGCC-1</td>\n",
       "      <td>UR:Z:ACCTAATTGG</td>\n",
       "      <td>UY:Z:FFFFFFFFFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:2519:6587:10175</td>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "      <td>631090</td>\n",
       "      <td>0</td>\n",
       "      <td>101M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTATTCTCTACAAACCACAAAGACATTGGAACACTATACCTATT...</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:99</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:GGTATTGAGCTAAGAT</td>\n",
       "      <td>CY:Z:FFFFFFFFFFFFFFFF</td>\n",
       "      <td>CB:Z:GGTATTGAGCTAAGAT-1</td>\n",
       "      <td>UR:Z:TTAAATGGAT</td>\n",
       "      <td>UY:Z:FFFFFFFFFF</td>\n",
       "      <td>UB:Z:TTAAATGGAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:1459:29975:9815</td>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "      <td>631218</td>\n",
       "      <td>0</td>\n",
       "      <td>101M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GGTAACGACCACATCTACAACGTTATCGTCACAGCCCATGCATTTG...</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:99</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:CTCTACGCAAACGCGA</td>\n",
       "      <td>CY:Z:FFFFFFFFFFFFFFFF</td>\n",
       "      <td>CB:Z:CTCTACGCAAACGCGA-1</td>\n",
       "      <td>UR:Z:GTCAAAGTAA</td>\n",
       "      <td>UY:Z:FFFFFFFFFF</td>\n",
       "      <td>UB:Z:GTCAAAGTAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>A00953:90:HC23FDSXY:3:1419:9733:32346</td>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "      <td>631251</td>\n",
       "      <td>0</td>\n",
       "      <td>101M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GCCCATGCATTTGTAATAATCTTCTTCATAGTAATACCCATCATAA...</td>\n",
       "      <td>...</td>\n",
       "      <td>AS:i:99</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>RE:A:I</td>\n",
       "      <td>li:i:0</td>\n",
       "      <td>CR:Z:CTCACACCACAGCGTC</td>\n",
       "      <td>CY:Z:FFFFFFFFFFFFFFFF</td>\n",
       "      <td>CB:Z:CTCACACCACAGCGTC-1</td>\n",
       "      <td>UR:Z:AGGTGCGGGG</td>\n",
       "      <td>UY:Z:FFFFFFFFFF</td>\n",
       "      <td>UB:Z:AGGTGCGGGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         0    1  2       3  4             5  \\\n",
       "0    A00953:90:HC23FDSXY:3:1559:18982:5400  272  1  379487  3  9M253923N20M   \n",
       "1    A00953:90:HC23FDSXY:3:1559:18982:5400   16  1  607673  3   9M25737N20M   \n",
       "2   A00953:90:HC23FDSXY:3:2608:15130:23218  272  1  629113  0         8S20M   \n",
       "3   A00953:90:HC23FDSXY:3:2608:13169:23484  272  1  629113  0         8S20M   \n",
       "4   A00953:90:HC23FDSXY:3:1668:20066:23343  272  1  629118  0           15M   \n",
       "..                                     ...  ... ..     ... ..           ...   \n",
       "95   A00953:90:HC23FDSXY:3:2619:22869:1297  272  1  631081  0          101M   \n",
       "96  A00953:90:HC23FDSXY:3:1307:20980:30217    0  1  631084  3          101M   \n",
       "97   A00953:90:HC23FDSXY:3:2519:6587:10175  272  1  631090  0          101M   \n",
       "98   A00953:90:HC23FDSXY:3:1459:29975:9815  272  1  631218  0          101M   \n",
       "99   A00953:90:HC23FDSXY:3:1419:9733:32346  272  1  631251  0          101M   \n",
       "\n",
       "    6  7  8                                           Sequence  ...       13  \\\n",
       "0   *  0  0                      TCTAACCATGGCCCGTATTTACCCTATAG  ...  AS:i:21   \n",
       "1   *  0  0                      TCTAACCATGGCCCGTATTTACCCTATAG  ...  AS:i:21   \n",
       "2   *  0  0                       CTCTGCCTATCGAATACGCCGCAGGCCC  ...  AS:i:19   \n",
       "3   *  0  0                       CTCTGCCTATCGAATACGCCGCAGGCCC  ...  AS:i:19   \n",
       "4   *  0  0                                    ATACGCCGCAGGCCC  ...  AS:i:14   \n",
       ".. .. .. ..                                                ...  ...      ...   \n",
       "95  *  0  0  CCGACCGTTGACTATTCTCTACAAACCACAAAGACATTGGAACACT...  ...  AS:i:99   \n",
       "96  *  0  0  ACCGTTGACTATTCTCTACAAACCACAAAGACATTGGAACACTATA...  ...  AS:i:99   \n",
       "97  *  0  0  GACTATTCTCTACAAACCACAAAGACATTGGAACACTATACCTATT...  ...  AS:i:99   \n",
       "98  *  0  0  GGTAACGACCACATCTACAACGTTATCGTCACAGCCCATGCATTTG...  ...  AS:i:99   \n",
       "99  *  0  0  GCCCATGCATTTGTAATAATCTTCTTCATAGTAATACCCATCATAA...  ...  AS:i:99   \n",
       "\n",
       "        14      15      16                     17                     18  \\\n",
       "0   nM:i:0  RE:A:I  li:i:0  CR:Z:CACCAAGCACCGAAAA  CY:Z:,,FFFFFF:FFFFFFF   \n",
       "1   nM:i:0  RE:A:I  xf:i:0                 li:i:0  CR:Z:CACCAAGCACCGAAAA   \n",
       "2   nM:i:0  RE:A:I  li:i:0  CR:Z:CTGCCTATCGAATACG  CY:Z::FFFFFFFFFFFFFFF   \n",
       "3   nM:i:0  RE:A:I  li:i:0  CR:Z:CTGCCTATCGAATACG  CY:Z::FFFFFFFFFFFFFFF   \n",
       "4   nM:i:0  RE:A:I  li:i:0  CR:Z:TTTAAACAAAAAATAA  CY:Z:FFF:,:FFFF,:F::F   \n",
       "..     ...     ...     ...                    ...                    ...   \n",
       "95  nM:i:0  RE:A:I  li:i:0  CR:Z:AACTCTTGTCCTAGCG  CY:Z:FFFFFFFFFFFFFFFF   \n",
       "96  nM:i:0  RE:A:I  xf:i:0                 li:i:0  CR:Z:GCCAAATCATTGGGCC   \n",
       "97  nM:i:0  RE:A:I  li:i:0  CR:Z:GGTATTGAGCTAAGAT  CY:Z:FFFFFFFFFFFFFFFF   \n",
       "98  nM:i:0  RE:A:I  li:i:0  CR:Z:CTCTACGCAAACGCGA  CY:Z:FFFFFFFFFFFFFFFF   \n",
       "99  nM:i:0  RE:A:I  li:i:0  CR:Z:CTCACACCACAGCGTC  CY:Z:FFFFFFFFFFFFFFFF   \n",
       "\n",
       "                         19                       20               21  \\\n",
       "0           UR:Z:TACCCTAAAT          UY:Z:FFFFFFFFF:  UB:Z:TACCCTAAAT   \n",
       "1     CY:Z:,,FFFFFF:FFFFFFF          UR:Z:TACCCTAAAT  UY:Z:FFFFFFFFF:   \n",
       "2           UR:Z:CCGCAGGCCC          UY:Z:FFFFFFFFFF  UB:Z:CCGCAGGCCC   \n",
       "3           UR:Z:CCGCAGGCCC          UY:Z:FFFFFFFFFF  UB:Z:CCGCAGGCCC   \n",
       "4           UR:Z:AAAACCCATC          UY:Z::FFFFFFFFF  UB:Z:AAAACCCATC   \n",
       "..                      ...                      ...              ...   \n",
       "95  CB:Z:AACTCTTGTCCTAGCG-1          UR:Z:ACGAGCAAGG  UY:Z:FFFFFFFFFF   \n",
       "96    CY:Z:FFFFFFFFFFFFFFFF  CB:Z:GCCAAATCATTGGGCC-1  UR:Z:ACCTAATTGG   \n",
       "97  CB:Z:GGTATTGAGCTAAGAT-1          UR:Z:TTAAATGGAT  UY:Z:FFFFFFFFFF   \n",
       "98  CB:Z:CTCTACGCAAACGCGA-1          UR:Z:GTCAAAGTAA  UY:Z:FFFFFFFFFF   \n",
       "99  CB:Z:CTCACACCACAGCGTC-1          UR:Z:AGGTGCGGGG  UY:Z:FFFFFFFFFF   \n",
       "\n",
       "                                           22  \n",
       "0   RG:Z:bc5prime_sc5p_r2_A_1:0:1:HC23FDSXY:3  \n",
       "1                             UB:Z:TACCCTAAAT  \n",
       "2   RG:Z:bc5prime_sc5p_r2_A_1:0:1:HC23FDSXY:3  \n",
       "3   RG:Z:bc5prime_sc5p_r2_A_1:0:1:HC23FDSXY:3  \n",
       "4   RG:Z:bc5prime_sc5p_r2_A_1:0:1:HC23FDSXY:3  \n",
       "..                                        ...  \n",
       "95                            UB:Z:ACGAGCAAGG  \n",
       "96                            UY:Z:FFFFFFFFFF  \n",
       "97                            UB:Z:TTAAATGGAT  \n",
       "98                            UB:Z:GTCAAAGTAA  \n",
       "99                            UB:Z:AGGTGCGGGG  \n",
       "\n",
       "[100 rows x 23 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sam_df = pd.read_csv(join(data,'tmp_100.sam.txt'),sep=\"\\t\", header=None,usecols= list(range(23)))\n",
    "sam_df = sam_df.rename({9:\"Sequence\"},axis=1)\n",
    "sam_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## See if the fastq and bam contain the primers we designed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:16:09.503410Z",
     "start_time": "2020-04-27T00:16:09.416895Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence 5'</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Strand</th>\n",
       "      <th>MT Position</th>\n",
       "      <th>Sequence 5' with tag</th>\n",
       "      <th>Name with tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene60922_0</td>\n",
       "      <td>TGTTTATGGGGTGATGTGAG</td>\n",
       "      <td>gene60922</td>\n",
       "      <td>-</td>\n",
       "      <td>646</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACTGTTTATGGGGTGATGTGAG</td>\n",
       "      <td>tag_gene60922_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene60923_349</td>\n",
       "      <td>CTGAGCAAGAGGTGGTGAGG</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>-</td>\n",
       "      <td>1251</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACCTGAGCAAGAGGTGGTGAGG</td>\n",
       "      <td>tag_gene60923_349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene60923_830</td>\n",
       "      <td>TGCTGCGTGCTTGATGCTTG</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>-</td>\n",
       "      <td>770</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACTGCTGCGTGCTTGATGCTTG</td>\n",
       "      <td>tag_gene60923_830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene60923_499</td>\n",
       "      <td>TAGGGCTAAGCATAGTGGGG</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>-</td>\n",
       "      <td>1101</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACTAGGGCTAAGCATAGTGGGG</td>\n",
       "      <td>tag_gene60923_499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene60923_0</td>\n",
       "      <td>GTTCGTCCAAGTGCACTTTC</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>-</td>\n",
       "      <td>1600</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACGTTCGTCCAAGTGCACTTTC</td>\n",
       "      <td>tag_gene60923_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>gene60956_981</td>\n",
       "      <td>GCATGGCTAGGAATAGTCCT</td>\n",
       "      <td>gene60956</td>\n",
       "      <td>-</td>\n",
       "      <td>14905</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACGCATGGCTAGGAATAGTCCT</td>\n",
       "      <td>tag_gene60956_981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>gene60956_831</td>\n",
       "      <td>AATATAGGCCTCGCCCGATG</td>\n",
       "      <td>gene60956</td>\n",
       "      <td>-</td>\n",
       "      <td>15055</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACAATATAGGCCTCGCCCGATG</td>\n",
       "      <td>tag_gene60956_831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>gene60956_0</td>\n",
       "      <td>AGGCCCATTTGAGTATTTTG</td>\n",
       "      <td>gene60956</td>\n",
       "      <td>-</td>\n",
       "      <td>15886</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACAGGCCCATTTGAGTATTTTG</td>\n",
       "      <td>tag_gene60956_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>gene60957_0</td>\n",
       "      <td>TGTCCTTGGAAAAAGGTTTT</td>\n",
       "      <td>gene60957</td>\n",
       "      <td>-</td>\n",
       "      <td>15952</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACTGTCCTTGGAAAAAGGTTTT</td>\n",
       "      <td>tag_gene60957_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>gene60958_0</td>\n",
       "      <td>TCAGAGAAAAAGTCTTTAAC</td>\n",
       "      <td>gene60958</td>\n",
       "      <td>+</td>\n",
       "      <td>15955</td>\n",
       "      <td>AAGCAGTGGTATCAACGCAGAGTACTCAGAGAAAAAGTCTTTAAC</td>\n",
       "      <td>tag_gene60958_0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Name           Sequence 5'       Gene Strand  MT Position  \\\n",
       "0      gene60922_0  TGTTTATGGGGTGATGTGAG  gene60922      -          646   \n",
       "1    gene60923_349  CTGAGCAAGAGGTGGTGAGG  gene60923      -         1251   \n",
       "2    gene60923_830  TGCTGCGTGCTTGATGCTTG  gene60923      -          770   \n",
       "3    gene60923_499  TAGGGCTAAGCATAGTGGGG  gene60923      -         1101   \n",
       "4      gene60923_0  GTTCGTCCAAGTGCACTTTC  gene60923      -         1600   \n",
       "..             ...                   ...        ...    ...          ...   \n",
       "117  gene60956_981  GCATGGCTAGGAATAGTCCT  gene60956      -        14905   \n",
       "118  gene60956_831  AATATAGGCCTCGCCCGATG  gene60956      -        15055   \n",
       "119    gene60956_0  AGGCCCATTTGAGTATTTTG  gene60956      -        15886   \n",
       "120    gene60957_0  TGTCCTTGGAAAAAGGTTTT  gene60957      -        15952   \n",
       "121    gene60958_0  TCAGAGAAAAAGTCTTTAAC  gene60958      +        15955   \n",
       "\n",
       "                              Sequence 5' with tag      Name with tag  \n",
       "0    AAGCAGTGGTATCAACGCAGAGTACTGTTTATGGGGTGATGTGAG    tag_gene60922_0  \n",
       "1    AAGCAGTGGTATCAACGCAGAGTACCTGAGCAAGAGGTGGTGAGG  tag_gene60923_349  \n",
       "2    AAGCAGTGGTATCAACGCAGAGTACTGCTGCGTGCTTGATGCTTG  tag_gene60923_830  \n",
       "3    AAGCAGTGGTATCAACGCAGAGTACTAGGGCTAAGCATAGTGGGG  tag_gene60923_499  \n",
       "4    AAGCAGTGGTATCAACGCAGAGTACGTTCGTCCAAGTGCACTTTC    tag_gene60923_0  \n",
       "..                                             ...                ...  \n",
       "117  AAGCAGTGGTATCAACGCAGAGTACGCATGGCTAGGAATAGTCCT  tag_gene60956_981  \n",
       "118  AAGCAGTGGTATCAACGCAGAGTACAATATAGGCCTCGCCCGATG  tag_gene60956_831  \n",
       "119  AAGCAGTGGTATCAACGCAGAGTACAGGCCCATTTGAGTATTTTG    tag_gene60956_0  \n",
       "120  AAGCAGTGGTATCAACGCAGAGTACTGTCCTTGGAAAAAGGTTTT    tag_gene60957_0  \n",
       "121  AAGCAGTGGTATCAACGCAGAGTACTCAGAGAAAAAGTCTTTAAC    tag_gene60958_0  \n",
       "\n",
       "[122 rows x 7 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primers = pd.read_csv(params[\"primers_list\"])\n",
    "primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For each one, see if it's in the SAM strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T00:58:53.610807Z",
     "start_time": "2020-04-27T00:58:53.605029Z"
    }
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T01:00:51.734873Z",
     "start_time": "2020-04-27T01:00:49.830939Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here!\n",
      "Hi\n",
      "Here!\n",
      "Hi\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Here!\n",
      "Hi\n",
      "Hi\n",
      "Hi\n",
      "(100, 23)\n",
      "defaultdict(<class 'int'>, {'CTAGGACTATGAGAAT': 1, 'CTAGGACTATGAGAATCGAA': 1, 'TAAGGACTGCAAAACCCC': 2, 'TAAGCACCCTAATCAA': 1})\n",
      "defaultdict(<class 'int'>, {'CTAGGACTATGAGAATCGAA': 15, 'TAAGGACTGCAAAACCCCAC': 6})\n"
     ]
    }
   ],
   "source": [
    "count_primer = defaultdict(int)\n",
    "count_fq = defaultdict(int)\n",
    "for ind,val in primers.iterrows():\n",
    "    for ind2, val2 in sam_df.iterrows():\n",
    "        if val[\"Sequence 5'\"] in val2[\"Sequence\"]:\n",
    "            print(\"Here!\")\n",
    "            count_primer[val[\"Sequence 5'\"]] += 1\n",
    "        if  val2[\"Sequence\"] in val[\"Sequence 5'\"]:\n",
    "            print(\"Hi\")\n",
    "            count_fq[val2[\"Sequence\"]] += 1\n",
    "\n",
    "print(sam_df.shape)\n",
    "print(count_fq)\n",
    "print(count_primer)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T01:42:31.446793Z",
     "start_time": "2020-04-27T01:42:31.434357Z"
    }
   },
   "outputs": [],
   "source": [
    "def check_bam(sample_name):\n",
    "    #sample_name = \"bc5prime_sc5p_pe_lanes\"\n",
    "    #sample_name = \"bc5prime_sc5p_pe_lanes\"\n",
    "    data = join(proc_dir, sample_name,\"outs/\")\n",
    "    bam_f = glob(join(data, \"*.bam\"))[0]\n",
    "    print(bam_f)\n",
    "    outdir = join(proc_dir, \"picard\") #, sample_name)\n",
    "    if not os.path.exists(outdir):\n",
    "        os.mkdir(outdir)\n",
    "\n",
    "    out_f = join(outdir, sample_name+'.txt')\n",
    "    print(out_f)\n",
    "\n",
    "    cmd = f\"picard CollectAlignmentSummaryMetrics \\\n",
    "    R={params['ref_fa']} \\\n",
    "    I={bam_f} \\\n",
    "    O={out_f}\"\n",
    "    print(cmd)\n",
    "\n",
    "    if not os.path.exists(out_f):\n",
    "        os.system(cmd)\n",
    "    else:\n",
    "        print('command already done')\n",
    "        \n",
    "    align_df = pd.read_csv(out_f, sep=\"\\t\",comment=\"#\")\n",
    "    print(align_df.transpose())\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-04-27T01:42:59.162140Z",
     "start_time": "2020-04-27T01:42:43.116153Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/bc5prime_sc5p_pe_A_1/outs/possorted_genome_bam.bam\n",
      "/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/picard/bc5prime_sc5p_pe_A_1.txt\n",
      "picard CollectAlignmentSummaryMetrics     R=/data2/genome/human_GRCh38/cellranger/refdata-cellranger-GRCh38-3.0.0/fasta/genome.fa     I=/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/bc5prime_sc5p_pe_A_1/outs/possorted_genome_bam.bam     O=/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/picard/bc5prime_sc5p_pe_A_1.txt\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] File b'/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/picard/bc5prime_sc5p_pe_A_1.txt' does not exist: b'/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/picard/bc5prime_sc5p_pe_A_1.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-43-ccce3568ca26>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcheck_bam\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"bc5prime_sc5p_pe_A_1\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-42-aaa698e2120d>\u001b[0m in \u001b[0;36mcheck_bam\u001b[0;34m(sample_name)\u001b[0m\n\u001b[1;32m     23\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'command already done'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m     \u001b[0malign_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mout_f\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcomment\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"#\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malign_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtranspose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/software/anaconda2/envs/mito_trace/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    683\u001b[0m         )\n\u001b[1;32m    684\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 685\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    686\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    687\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/software/anaconda2/envs/mito_trace/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    455\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    456\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 457\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfp_or_buf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    458\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    459\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/software/anaconda2/envs/mito_trace/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    893\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    894\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 895\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    896\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    897\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/software/anaconda2/envs/mito_trace/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1133\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"c\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1134\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"c\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1135\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1136\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1137\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/software/anaconda2/envs/mito_trace/lib/python3.7/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1915\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"usecols\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1916\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1917\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1918\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1919\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] File b'/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/picard/bc5prime_sc5p_pe_A_1.txt' does not exist: b'/data2/isshamie/mito_lineage/data/processed/200403_A00953_0090_BHC23FDSXY/trim/picard/bc5prime_sc5p_pe_A_1.txt'"
     ]
    }
   ],
   "source": [
    "check_bam(\"bc5prime_sc5p_pe_A_1\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "/data/isshamie/software/cellranger-3.1.0/martian-cs/v3.2.3/adapters/python\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
