{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "41fdb08b",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# DE.out.path = \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0.1_pthresh_0.1/\"\n",
    "# export.path = \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0.1/GSEA_pthresh_0.1/\"\n",
    "# gsea_dir = \"/mnt/md0/isshamie/Projects/Mito_Trace/software/Bioinformatics_Tools/\"\n",
    "# pthresh = 0.1\n",
    "# stat_col = \"log2FoldChange\"\n",
    "# prefilter = TRUE\n",
    "# padjmethod = \"BH\"\n",
    "# gsea_pthresh = 0.1\n",
    "\n",
    "DE.out.path = \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust\"\n",
    "export.path = \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GO_pthresh.0.01/\"\n",
    "gsea_dir = \"/mnt/md0/isshamie/Projects/Mito_Trace/software/Bioinformatics_Tools/\"\n",
    "\n",
    "pthresh = 0.1 \n",
    "go_pthresh = 0.01 \n",
    "prefilter = FALSE \n",
    "stat_col = \"log2FoldChange\"\n",
    "padjmethod = \"BH\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b4fcc9a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(export.path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1b26d1e2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: topGO\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: graph\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: GO.db\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: SparseM\n",
      "\n",
      "\n",
      "Attaching package: ‘SparseM’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    backsolve\n",
      "\n",
      "\n",
      "\n",
      "groupGOTerms: \tGOBPTerm, GOMFTerm, GOCCTerm environments built.\n",
      "\n",
      "\n",
      "Attaching package: ‘topGO’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    members\n",
      "\n",
      "\n",
      "Loading required package: ALL\n",
      "\n",
      "Loading required package: org.Hs.eg.db\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: clusterProfiler\n",
      "\n",
      "clusterProfiler v4.2.0  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:clusterProfiler\"\n",
      "[1] \"package already installed:clusterProfiler\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GSEABase\n",
      "\n",
      "Loading required package: annotate\n",
      "\n",
      "Loading required package: XML\n",
      "\n",
      "\n",
      "Attaching package: ‘XML’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graph’:\n",
      "\n",
      "    addNode\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘GSEABase’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:topGO’:\n",
      "\n",
      "    ontology, phenotype\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:GSEABase\"\n",
      "[1] \"package already installed:GSEABase\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: biomaRt\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:biomaRt\"\n",
      "[1] \"package already installed:biomaRt\"\n",
      "[1] \"package already installed:org.Hs.eg.db\"\n",
      "[1] \"package already installed:org.Hs.eg.db\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: DOSE\n",
      "\n",
      "DOSE v3.20.0  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use DOSE in published research, please cite:\n",
      "Guangchuang Yu, Li-Gen Wang, Guang-Rong Yan, Qing-Yu He. DOSE: an R/Bioconductor package for Disease Ontology Semantic and Enrichment analysis. Bioinformatics 2015, 31(4):608-609\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:DOSE\"\n",
      "[1] \"package already installed:DOSE\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: gdata\n",
      "\n",
      "gdata: read.xls support for 'XLS' (Excel 97-2004) files ENABLED.\n",
      "\n",
      "\n",
      "\n",
      "gdata: read.xls support for 'XLSX' (Excel 2007+) files ENABLED.\n",
      "\n",
      "\n",
      "Attaching package: ‘gdata’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    startsWith, trim\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, first<-\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats4’:\n",
      "\n",
      "    nobs\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    nobs\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    object.size\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    startsWith\n",
      "\n",
      "\n",
      "Loading required package: data.table\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:gdata’:\n",
      "\n",
      "    first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "Loading required package: openxlsx\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:gdata’:\n",
      "\n",
      "    combine, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GSEABase’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graph’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘stringr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graph’:\n",
      "\n",
      "    boundary\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!require(\"topGO\", character.only = TRUE)) {\n",
    "    BiocManager::install(\"topGO\")   \n",
    "}\n",
    "if (!require(\"ALL\", character.only = TRUE)) {\n",
    "    BiocManager::install(\"ALL\")   \n",
    "}\n",
    "require(org.Hs.eg.db)\n",
    "source(paste0(gsea_dir,'R/Gsea.R'))\n",
    "library(topGO)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "55359f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "if ((prefilter==\"TRUE\") | (prefilter==TRUE)){\n",
    "    prefilter=TRUE\n",
    "}else{\n",
    "    prefilter=FALSE\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "68b43d85",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_0'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_1'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_10'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_11'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_12'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_13'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_14'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_15'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_2'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_3'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_4'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_5'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_6'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_7'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_8'</li><li>'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_9'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_0'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_1'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_10'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_11'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_12'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_13'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_14'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_15'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_2'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_3'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_4'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_5'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_6'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_7'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_8'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_9'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_0'\n",
       "2. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_1'\n",
       "3. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_10'\n",
       "4. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_11'\n",
       "5. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_12'\n",
       "6. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_13'\n",
       "7. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_14'\n",
       "8. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_15'\n",
       "9. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_2'\n",
       "10. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_3'\n",
       "11. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_4'\n",
       "12. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_5'\n",
       "13. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_6'\n",
       "14. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_7'\n",
       "15. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_8'\n",
       "16. '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_9'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_0\" \n",
       " [2] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_1\" \n",
       " [3] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_10\"\n",
       " [4] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_11\"\n",
       " [5] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_12\"\n",
       " [6] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_13\"\n",
       " [7] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_14\"\n",
       " [8] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_15\"\n",
       " [9] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_2\" \n",
       "[10] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_3\" \n",
       "[11] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_4\" \n",
       "[12] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_5\" \n",
       "[13] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_6\" \n",
       "[14] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_7\" \n",
       "[15] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_8\" \n",
       "[16] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_9\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "files <- list.files(path=DE.out.path,pattern = \"clust_*\", full.names = T, recursive = FALSE, include.dirs = TRUE)\n",
    "DE_dirs <- files[file.info(files)$isdir ]\n",
    "DE_dirs\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6467fe7c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "895b2b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols.prior <- c(\"Gene\",\"baseMean\",\"log2FoldChange\",\"lfcSE\",\"stat\",\"pvalue\",\"padj\")\n",
    "cols.needed <- c(\"Gene\",\"log2FoldChange\")\n",
    "cols.curr <- c(\"X\",\"p_val\",\"avg_log2FC\",\"pct.1\",\"pct.2\",\"p_val_adj_BH\") #first col is empty in csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0a26190",
   "metadata": {},
   "source": [
    "## Run for each results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e8f73b98",
   "metadata": {},
   "outputs": [],
   "source": [
    "run.go <- function(universe.genes, name, export.path, proc=\"MF\"){\n",
    "    selection <- function(allScore){ return(allScore < 0.01)} # function that returns TRUE/FALSE for p-values<0.05\n",
    "    \n",
    "    allGO2genes <- annFUN.org(whichOnto=proc, feasibleGenes=NULL, mapping=\"org.Hs.eg.db\", ID=\"symbol\")\n",
    "    GOdata <- new(\"topGOdata\",\n",
    "      ontology=proc,\n",
    "      allGenes=universe.genes,\n",
    "      annot=annFUN.GO2genes,\n",
    "      GO2genes=allGO2genes,\n",
    "      geneSel=selection,\n",
    "      nodeSize=10)\n",
    "    results.ks <- runTest(GOdata, algorithm=\"classic\", statistic=\"ks\")\n",
    "    goEnrichment <- GenTable(GOdata, KS=results.ks, orderBy=\"KS\", topNodes=20)\n",
    "    goEnrichment$KS <- as.numeric(goEnrichment$KS)\n",
    "    goEnrichment <- goEnrichment[goEnrichment$KS<0.05,]\n",
    "    goEnrichment <- goEnrichment[,c(\"GO.ID\",\"Term\",\"KS\")]\n",
    "    goEnrichment$Term <- gsub(\" [a-z]*\\\\.\\\\.\\\\.$\", \"\", goEnrichment$Term)\n",
    "    goEnrichment$Term <- gsub(\"\\\\.\\\\.\\\\.$\", \"\", goEnrichment$Term)\n",
    "    goEnrichment$Term <- paste(goEnrichment$GO.ID, goEnrichment$Term, sep=\", \")\n",
    "    goEnrichment$Term <- factor(goEnrichment$Term, levels=rev(goEnrichment$Term))\n",
    "    print(goEnrichment)\n",
    "    write.table(goEnrichment, file.path(export.path, paste0(name, \"_\", \"proc_\", proc, \".tsv\")), sep=\"\\t\")\n",
    "    return(goEnrichment)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "541f8ea4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_0/0_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 307   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 304   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 488 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 775 GO terms and 974 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 271 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 90 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term      KS\n",
      "1  GO:0003677                             GO:0003677, DNA binding 0.00082\n",
      "2  GO:0003676                    GO:0003676, nucleic acid binding 0.00103\n",
      "3  GO:0140110        GO:0140110, transcription regulator activity 0.00112\n",
      "4  GO:0043565           GO:0043565, sequence-specific DNA binding 0.00148\n",
      "5  GO:0019899                          GO:0019899, enzyme binding 0.00161\n",
      "6  GO:0001067 GO:0001067, transcription regulatory region nucleic 0.00483\n",
      "7  GO:0000976     GO:0000976, transcription cis-regulatory region 0.00542\n",
      "8  GO:0003690             GO:0003690, double-stranded DNA binding 0.00542\n",
      "9  GO:1990837   GO:1990837, sequence-specific double-stranded DNA 0.00542\n",
      "10 GO:0008270                        GO:0008270, zinc ion binding 0.01490\n",
      "11 GO:0046914            GO:0046914, transition metal ion binding 0.01490\n",
      "12 GO:0003779                           GO:0003779, actin binding 0.01921\n",
      "13 GO:0016740                    GO:0016740, transferase activity 0.02331\n",
      "14 GO:0000977         GO:0000977, RNA polymerase II transcription 0.02680\n",
      "15 GO:0003700        GO:0003700, DNA-binding transcription factor 0.02680\n",
      "16 GO:0016772      GO:0016772, transferase activity, transferring 0.02773\n",
      "17 GO:0097159         GO:0097159, organic cyclic compound binding 0.02806\n",
      "18 GO:1901363           GO:1901363, heterocyclic compound binding 0.02806\n",
      "19 GO:0000978 GO:0000978, RNA polymerase II cis-regulatory region 0.02921\n",
      "20 GO:0000987 GO:0000987, cis-regulatory region sequence-specific 0.02921\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1673 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 4479 GO terms and 9853 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 265 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 562 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term      KS\n",
      "1  GO:0060255    GO:0060255, regulation of macromolecule metabolic 1.1e-05\n",
      "2  GO:0009892 GO:0009892, negative regulation of metabolic process 2.9e-05\n",
      "3  GO:0019222          GO:0019222, regulation of metabolic process 3.1e-05\n",
      "4  GO:0010467                          GO:0010467, gene expression 7.0e-05\n",
      "5  GO:0010605     GO:0010605, negative regulation of macromolecule 7.0e-05\n",
      "6  GO:0031323 GO:0031323, regulation of cellular metabolic process 8.9e-05\n",
      "7  GO:0031324          GO:0031324, negative regulation of cellular 1.2e-04\n",
      "8  GO:0051171          GO:0051171, regulation of nitrogen compound 1.4e-04\n",
      "9  GO:0032270  GO:0032270, positive regulation of cellular protein 2.3e-04\n",
      "10 GO:0051247           GO:0051247, positive regulation of protein 2.3e-04\n",
      "11 GO:0043170          GO:0043170, macromolecule metabolic process 2.3e-04\n",
      "12 GO:0080090  GO:0080090, regulation of primary metabolic process 2.6e-04\n",
      "13 GO:0051172          GO:0051172, negative regulation of nitrogen 2.9e-04\n",
      "14 GO:0044260 GO:0044260, cellular macromolecule metabolic process 3.6e-04\n",
      "15 GO:0010468            GO:0010468, regulation of gene expression 3.9e-04\n",
      "16 GO:0051246  GO:0051246, regulation of protein metabolic process 4.3e-04\n",
      "17 GO:1902105  GO:1902105, regulation of leukocyte differentiation 6.4e-04\n",
      "18 GO:0009057          GO:0009057, macromolecule catabolic process 6.8e-04\n",
      "19 GO:0046632        GO:0046632, alpha-beta T cell differentiation 7.4e-04\n",
      "20 GO:0032268           GO:0032268, regulation of cellular protein 8.5e-04\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_1/1_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 714   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 617   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 799 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 1156 GO terms and 1485 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 554 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 138 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term       KS\n",
      "1  GO:0140110        GO:0140110, transcription regulator activity 0.000022\n",
      "2  GO:0043565           GO:0043565, sequence-specific DNA binding 0.000800\n",
      "3  GO:0000976     GO:0000976, transcription cis-regulatory region 0.000980\n",
      "4  GO:1990837   GO:1990837, sequence-specific double-stranded DNA 0.000980\n",
      "5  GO:0000977         GO:0000977, RNA polymerase II transcription 0.001470\n",
      "6  GO:0000981        GO:0000981, DNA-binding transcription factor 0.001650\n",
      "7  GO:0001067 GO:0001067, transcription regulatory region nucleic 0.001810\n",
      "8  GO:0003690             GO:0003690, double-stranded DNA binding 0.001810\n",
      "9  GO:0003700        GO:0003700, DNA-binding transcription factor 0.003260\n",
      "10 GO:0003712      GO:0003712, transcription coregulator activity 0.003630\n",
      "11 GO:0019899                          GO:0019899, enzyme binding 0.005020\n",
      "12 GO:0000987 GO:0000987, cis-regulatory region sequence-specific 0.005020\n",
      "13 GO:0003677                             GO:0003677, DNA binding 0.006210\n",
      "14 GO:0000978 GO:0000978, RNA polymerase II cis-regulatory region 0.007440\n",
      "15 GO:0003676                    GO:0003676, nucleic acid binding 0.009400\n",
      "16 GO:0042802               GO:0042802, identical protein binding 0.015450\n",
      "17 GO:0016740                    GO:0016740, transferase activity 0.016800\n",
      "18 GO:0019787      GO:0019787, ubiquitin-like protein transferase 0.018710\n",
      "19 GO:0140098       GO:0140098, catalytic activity, acting on RNA 0.018930\n",
      "20 GO:0003723                             GO:0003723, RNA binding 0.025280\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 2734 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 5961 GO terms and 13047 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 535 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 943 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term      KS\n",
      "1  GO:0044260 GO:0044260, cellular macromolecule metabolic process 3.3e-07\n",
      "2  GO:0048523  GO:0048523, negative regulation of cellular process 1.3e-06\n",
      "3  GO:0051171          GO:0051171, regulation of nitrogen compound 2.1e-06\n",
      "4  GO:0060255    GO:0060255, regulation of macromolecule metabolic 2.6e-06\n",
      "5  GO:0010605     GO:0010605, negative regulation of macromolecule 4.6e-06\n",
      "6  GO:0048519        GO:0048519, negative regulation of biological 6.8e-06\n",
      "7  GO:0044237               GO:0044237, cellular metabolic process 7.2e-06\n",
      "8  GO:0051172          GO:0051172, negative regulation of nitrogen 8.6e-06\n",
      "9  GO:0080090  GO:0080090, regulation of primary metabolic process 1.1e-05\n",
      "10 GO:0031323 GO:0031323, regulation of cellular metabolic process 1.4e-05\n",
      "11 GO:0009892 GO:0009892, negative regulation of metabolic process 1.5e-05\n",
      "12 GO:0031324          GO:0031324, negative regulation of cellular 1.6e-05\n",
      "13 GO:0043170          GO:0043170, macromolecule metabolic process 3.7e-05\n",
      "14 GO:0032268           GO:0032268, regulation of cellular protein 4.1e-05\n",
      "15 GO:0010558     GO:0010558, negative regulation of macromolecule 6.5e-05\n",
      "16 GO:0009890      GO:0009890, negative regulation of biosynthetic 7.8e-05\n",
      "17 GO:0031327          GO:0031327, negative regulation of cellular 7.8e-05\n",
      "18 GO:2000113          GO:2000113, negative regulation of cellular 1.0e-04\n",
      "19 GO:0051246  GO:0051246, regulation of protein metabolic process 1.1e-04\n",
      "20 GO:0006807      GO:0006807, nitrogen compound metabolic process 1.3e-04\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_10/10_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 3851    8\n",
      "[1] \"after pct removal\"\n",
      "[1] 2940    8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1832 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 2311 GO terms and 3013 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 2535 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 395 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term     KS\n",
      "1  GO:0001664       GO:0001664, G protein-coupled receptor binding 0.0054\n",
      "2  GO:0061629   GO:0061629, RNA polymerase II-specific DNA-binding 0.0062\n",
      "3  GO:0005261                  GO:0005261, cation channel activity 0.0088\n",
      "4  GO:0005216                     GO:0005216, ion channel activity 0.0131\n",
      "5  GO:0003682                        GO:0003682, chromatin binding 0.0138\n",
      "6  GO:0015318               GO:0015318, inorganic molecular entity 0.0152\n",
      "7  GO:0015075   GO:0015075, ion transmembrane transporter activity 0.0164\n",
      "8  GO:0140297 GO:0140297, DNA-binding transcription factor binding 0.0199\n",
      "9  GO:0015267                         GO:0015267, channel activity 0.0212\n",
      "10 GO:0022803        GO:0022803, passive transmembrane transporter 0.0212\n",
      "11 GO:0043169                           GO:0043169, cation binding 0.0282\n",
      "12 GO:0022857       GO:0022857, transmembrane transporter activity 0.0287\n",
      "13 GO:0004722     GO:0004722, protein serine/threonine phosphatase 0.0310\n",
      "14 GO:0046872                        GO:0046872, metal ion binding 0.0332\n",
      "15 GO:0016922                 GO:0016922, nuclear receptor binding 0.0384\n",
      "16 GO:0106306      GO:0106306, protein serine phosphatase activity 0.0406\n",
      "17 GO:0106307   GO:0106307, protein threonine phosphatase activity 0.0406\n",
      "18 GO:0008083                   GO:0008083, growth factor activity 0.0411\n",
      "19 GO:0042578      GO:0042578, phosphoric ester hydrolase activity 0.0425\n",
      "20 GO:0008408               GO:0008408, 3'-5' exonuclease activity 0.0448\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 6008 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 9989 GO terms and 22374 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 2464 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 2478 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                               Term      KS\n",
      "1  GO:0008284 GO:0008284, positive regulation of cell population 0.00024\n",
      "2  GO:0042692            GO:0042692, muscle cell differentiation 0.00079\n",
      "3  GO:0055001                GO:0055001, muscle cell development 0.00084\n",
      "4  GO:0019220      GO:0019220, regulation of phosphate metabolic 0.00087\n",
      "5  GO:0051174     GO:0051174, regulation of phosphorus metabolic 0.00087\n",
      "6  GO:0021954          GO:0021954, central nervous system neuron 0.00092\n",
      "7  GO:0035637     GO:0035637, multicellular organismal signaling 0.00104\n",
      "8  GO:0016311                      GO:0016311, dephosphorylation 0.00108\n",
      "9  GO:0044057           GO:0044057, regulation of system process 0.00137\n",
      "10 GO:0000079 GO:0000079, regulation of cyclin-dependent protein 0.00152\n",
      "11 GO:1904029 GO:1904029, regulation of cyclin-dependent protein 0.00152\n",
      "12 GO:0048660       GO:0048660, regulation of smooth muscle cell 0.00172\n",
      "13 GO:0035303        GO:0035303, regulation of dephosphorylation 0.00175\n",
      "14 GO:0007423              GO:0007423, sensory organ development 0.00178\n",
      "15 GO:0060348                       GO:0060348, bone development 0.00200\n",
      "16 GO:0010563      GO:0010563, negative regulation of phosphorus 0.00201\n",
      "17 GO:0045936       GO:0045936, negative regulation of phosphate 0.00201\n",
      "18 GO:0009888                     GO:0009888, tissue development 0.00238\n",
      "19 GO:0001657               GO:0001657, ureteric bud development 0.00244\n",
      "20 GO:0001823                GO:0001823, mesonephros development 0.00244\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_11/11_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 679   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 665   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 756 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 1124 GO terms and 1457 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 586 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 138 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                              Term       KS\n",
      "1  GO:1901363         GO:1901363, heterocyclic compound binding 0.000011\n",
      "2  GO:0097159       GO:0097159, organic cyclic compound binding 0.000017\n",
      "3  GO:0003676                  GO:0003676, nucleic acid binding 0.000079\n",
      "4  GO:0003729                          GO:0003729, mRNA binding 0.000320\n",
      "5  GO:0003723                           GO:0003723, RNA binding 0.000660\n",
      "6  GO:0005515                       GO:0005515, protein binding 0.007830\n",
      "7  GO:0044389 GO:0044389, ubiquitin-like protein ligase binding 0.011160\n",
      "8  GO:0001217   GO:0001217, DNA-binding transcription repressor 0.012350\n",
      "9  GO:0003824                    GO:0003824, catalytic activity 0.014360\n",
      "10 GO:0140098     GO:0140098, catalytic activity, acting on RNA 0.014880\n",
      "11 GO:0001227   GO:0001227, DNA-binding transcription repressor 0.020750\n",
      "12 GO:0000977       GO:0000977, RNA polymerase II transcription 0.039060\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 2513 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 5744 GO terms and 12708 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 574 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 918 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term      KS\n",
      "1  GO:0034641    GO:0034641, cellular nitrogen compound metabolic 0.00002\n",
      "2  GO:0035966     GO:0035966, response to topologically incorrect 0.00042\n",
      "3  GO:0044271 GO:0044271, cellular nitrogen compound biosynthetic 0.00052\n",
      "4  GO:1901360       GO:1901360, organic cyclic compound metabolic 0.00061\n",
      "5  GO:0006412                             GO:0006412, translation 0.00065\n",
      "6  GO:0044237              GO:0044237, cellular metabolic process 0.00068\n",
      "7  GO:0046483           GO:0046483, heterocycle metabolic process 0.00092\n",
      "8  GO:0043043            GO:0043043, peptide biosynthetic process 0.00115\n",
      "9  GO:0044249           GO:0044249, cellular biosynthetic process 0.00116\n",
      "10 GO:0031327         GO:0031327, negative regulation of cellular 0.00121\n",
      "11 GO:0006725    GO:0006725, cellular aromatic compound metabolic 0.00128\n",
      "12 GO:2000113         GO:2000113, negative regulation of cellular 0.00130\n",
      "13 GO:0009058                    GO:0009058, biosynthetic process 0.00146\n",
      "14 GO:0006518               GO:0006518, peptide metabolic process 0.00160\n",
      "15 GO:0006139          GO:0006139, nucleobase-containing compound 0.00178\n",
      "16 GO:1901576  GO:1901576, organic substance biosynthetic process 0.00204\n",
      "17 GO:0010558    GO:0010558, negative regulation of macromolecule 0.00221\n",
      "18 GO:0009890     GO:0009890, negative regulation of biosynthetic 0.00245\n",
      "19 GO:0043604              GO:0043604, amide biosynthetic process 0.00260\n",
      "20 GO:0010467                         GO:0010467, gene expression 0.00292\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_12/12_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 3237    8\n",
      "[1] \"after pct removal\"\n",
      "[1] 2739    8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1838 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 2336 GO terms and 3058 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 2393 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 381 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term        KS\n",
      "1  GO:0140640 GO:0140640, catalytic activity, acting on a nucleic 2.400e-07\n",
      "2  GO:0003676                    GO:0003676, nucleic acid binding 1.800e-06\n",
      "3  GO:0003677                             GO:0003677, DNA binding 6.400e-05\n",
      "4  GO:0140098       GO:0140098, catalytic activity, acting on RNA 8.200e-05\n",
      "5  GO:1901363           GO:1901363, heterocyclic compound binding 1.200e-04\n",
      "6  GO:0097159         GO:0097159, organic cyclic compound binding 2.900e-04\n",
      "7  GO:0003723                             GO:0003723, RNA binding 3.000e-04\n",
      "8  GO:0140097       GO:0140097, catalytic activity, acting on DNA 3.230e-03\n",
      "9  GO:0023023             GO:0023023, MHC protein complex binding 3.720e-03\n",
      "10 GO:0023026    GO:0023026, MHC class II protein complex binding 3.720e-03\n",
      "11 GO:0005515                         GO:0005515, protein binding 4.680e-03\n",
      "12 GO:0003727             GO:0003727, single-stranded RNA binding 5.660e-03\n",
      "13 GO:0031490                   GO:0031490, chromatin DNA binding 5.850e-03\n",
      "14 GO:0005488                                 GO:0005488, binding 6.070e-03\n",
      "15 GO:0016779         GO:0016779, nucleotidyltransferase activity 8.120e-03\n",
      "16 GO:0003725             GO:0003725, double-stranded RNA binding 9.170e-03\n",
      "17 GO:0004540                   GO:0004540, ribonuclease activity 1.091e-02\n",
      "18 GO:0003729                            GO:0003729, mRNA binding 1.167e-02\n",
      "19 GO:0004527                    GO:0004527, exonuclease activity 1.183e-02\n",
      "20 GO:0019888  GO:0019888, protein phosphatase regulator activity 1.191e-02\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 5885 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 9743 GO terms and 21752 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 2333 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 2516 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term      KS\n",
      "1  GO:0046483            GO:0046483, heterocycle metabolic process 4.6e-07\n",
      "2  GO:0090304           GO:0090304, nucleic acid metabolic process 7.1e-07\n",
      "3  GO:0006139           GO:0006139, nucleobase-containing compound 1.0e-06\n",
      "4  GO:0006725     GO:0006725, cellular aromatic compound metabolic 1.7e-06\n",
      "5  GO:1901360        GO:1901360, organic cyclic compound metabolic 2.6e-06\n",
      "6  GO:0044237               GO:0044237, cellular metabolic process 2.6e-06\n",
      "7  GO:0043170          GO:0043170, macromolecule metabolic process 5.6e-06\n",
      "8  GO:0006807      GO:0006807, nitrogen compound metabolic process 5.7e-06\n",
      "9  GO:0034641     GO:0034641, cellular nitrogen compound metabolic 5.8e-06\n",
      "10 GO:0008152                        GO:0008152, metabolic process 3.8e-05\n",
      "11 GO:0044238                GO:0044238, primary metabolic process 7.1e-05\n",
      "12 GO:0006396                           GO:0006396, RNA processing 8.6e-05\n",
      "13 GO:0016070                    GO:0016070, RNA metabolic process 1.4e-04\n",
      "14 GO:0071704      GO:0071704, organic substance metabolic process 1.6e-04\n",
      "15 GO:0044260 GO:0044260, cellular macromolecule metabolic process 1.6e-04\n",
      "16 GO:0034622      GO:0034622, cellular protein-containing complex 4.0e-04\n",
      "17 GO:0043933       GO:0043933, protein-containing complex subunit 4.3e-04\n",
      "18 GO:0034660                  GO:0034660, ncRNA metabolic process 7.0e-04\n",
      "19 GO:0010467                          GO:0010467, gene expression 7.0e-04\n",
      "20 GO:0022613     GO:0022613, ribonucleoprotein complex biogenesis 8.0e-04\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_13/13_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 1638    8\n",
      "[1] \"after pct removal\"\n",
      "[1] 1497    8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1283 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 1735 GO terms and 2260 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 1322 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 235 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                               Term      KS\n",
      "1  GO:0097159        GO:0097159, organic cyclic compound binding 0.00057\n",
      "2  GO:1901363          GO:1901363, heterocyclic compound binding 0.00071\n",
      "3  GO:0004175                 GO:0004175, endopeptidase activity 0.00640\n",
      "4  GO:0001221      GO:0001221, transcription coregulator binding 0.00693\n",
      "5  GO:0017124                     GO:0017124, SH3 domain binding 0.01749\n",
      "6  GO:0003676                   GO:0003676, nucleic acid binding 0.02148\n",
      "7  GO:0008233                     GO:0008233, peptidase activity 0.02637\n",
      "8  GO:0019904        GO:0019904, protein domain specific binding 0.03477\n",
      "9  GO:0016787                     GO:0016787, hydrolase activity 0.03623\n",
      "10 GO:0044389  GO:0044389, ubiquitin-like protein ligase binding 0.03701\n",
      "11 GO:0016747 GO:0016747, acyltransferase activity, transferring 0.03897\n",
      "12 GO:0000166                     GO:0000166, nucleotide binding 0.04055\n",
      "13 GO:1901265           GO:1901265, nucleoside phosphate binding 0.04055\n",
      "14 GO:0051213                   GO:0051213, dioxygenase activity 0.04229\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 4154 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 7784 GO terms and 17262 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 1273 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 1609 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term     KS\n",
      "1  GO:0006793            GO:0006793, phosphorus metabolic process 0.0033\n",
      "2  GO:0006397                         GO:0006397, mRNA processing 0.0038\n",
      "3  GO:0006796 GO:0006796, phosphate-containing compound metabolic 0.0040\n",
      "4  GO:0033135           GO:0033135, regulation of peptidyl-serine 0.0067\n",
      "5  GO:0000375   GO:0000375, RNA splicing, via transesterification 0.0073\n",
      "6  GO:0000377   GO:0000377, RNA splicing, via transesterification 0.0073\n",
      "7  GO:0000398          GO:0000398, mRNA splicing, via spliceosome 0.0073\n",
      "8  GO:0071704     GO:0071704, organic substance metabolic process 0.0105\n",
      "9  GO:0008152                       GO:0008152, metabolic process 0.0115\n",
      "10 GO:0008380                            GO:0008380, RNA splicing 0.0128\n",
      "11 GO:0018105         GO:0018105, peptidyl-serine phosphorylation 0.0130\n",
      "12 GO:0010604    GO:0010604, positive regulation of macromolecule 0.0136\n",
      "13 GO:0008630   GO:0008630, intrinsic apoptotic signaling pathway 0.0158\n",
      "14 GO:0010557    GO:0010557, positive regulation of macromolecule 0.0160\n",
      "15 GO:0016052          GO:0016052, carbohydrate catabolic process 0.0161\n",
      "16 GO:0098657                        GO:0098657, import into cell 0.0167\n",
      "17 GO:0098739           GO:0098739, import across plasma membrane 0.0167\n",
      "18 GO:0044237              GO:0044237, cellular metabolic process 0.0173\n",
      "19 GO:1903825    GO:1903825, organic acid transmembrane transport 0.0176\n",
      "20 GO:1905039 GO:1905039, carboxylic acid transmembrane transport 0.0176\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_14/14_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2493    8\n",
      "[1] \"after pct removal\"\n",
      "[1] 2293    8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1598 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 2065 GO terms and 2696 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 1993 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 328 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GO.ID                                               Term    KS\n",
      "1 GO:0016798 GO:0016798, hydrolase activity, acting on glycosyl 0.016\n",
      "2 GO:0003676                   GO:0003676, nucleic acid binding 0.030\n",
      "3 GO:0003723                            GO:0003723, RNA binding 0.036\n",
      "4 GO:0003714     GO:0003714, transcription corepressor activity 0.039\n",
      "5 GO:0019899                         GO:0019899, enzyme binding 0.040\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 5186 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 8967 GO terms and 20089 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 1922 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 2167 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term     KS\n",
      "1  GO:0000280                         GO:0000280, nuclear division 0.0015\n",
      "2  GO:0009135  GO:0009135, purine nucleoside diphosphate metabolic 0.0017\n",
      "3  GO:0009179        GO:0009179, purine ribonucleoside diphosphate 0.0017\n",
      "4  GO:0048469                          GO:0048469, cell maturation 0.0018\n",
      "5  GO:0009185     GO:0009185, ribonucleoside diphosphate metabolic 0.0027\n",
      "6  GO:0009132 GO:0009132, nucleoside diphosphate metabolic process 0.0037\n",
      "7  GO:0002376                    GO:0002376, immune system process 0.0038\n",
      "8  GO:0006096                       GO:0006096, glycolytic process 0.0039\n",
      "9  GO:0006757                  GO:0006757, ATP generation from ADP 0.0039\n",
      "10 GO:0002682      GO:0002682, regulation of immune system process 0.0042\n",
      "11 GO:0009150  GO:0009150, purine ribonucleotide metabolic process 0.0050\n",
      "12 GO:0009259         GO:0009259, ribonucleotide metabolic process 0.0054\n",
      "13 GO:0051348       GO:0051348, negative regulation of transferase 0.0054\n",
      "14 GO:0098542       GO:0098542, defense response to other organism 0.0055\n",
      "15 GO:0043170          GO:0043170, macromolecule metabolic process 0.0055\n",
      "16 GO:0021700                 GO:0021700, developmental maturation 0.0060\n",
      "17 GO:0016570                     GO:0016570, histone modification 0.0061\n",
      "18 GO:0072521     GO:0072521, purine-containing compound metabolic 0.0061\n",
      "19 GO:1903046               GO:1903046, meiotic cell cycle process 0.0064\n",
      "20 GO:0002683     GO:0002683, negative regulation of immune system 0.0066\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_15/15_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 5190    8\n",
      "[1] \"after pct removal\"\n",
      "[1] 4532    8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 2350 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 2869 GO terms and 3744 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 3951 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 532 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term     KS\n",
      "1  GO:0003676                     GO:0003676, nucleic acid binding 0.0018\n",
      "2  GO:1901363            GO:1901363, heterocyclic compound binding 0.0037\n",
      "3  GO:0097159          GO:0097159, organic cyclic compound binding 0.0051\n",
      "4  GO:0046872                        GO:0046872, metal ion binding 0.0083\n",
      "5  GO:0044183                GO:0044183, protein folding chaperone 0.0086\n",
      "6  GO:0043169                           GO:0043169, cation binding 0.0134\n",
      "7  GO:0000030             GO:0000030, mannosyltransferase activity 0.0134\n",
      "8  GO:0015179   GO:0015179, L-amino acid transmembrane transporter 0.0170\n",
      "9  GO:0003723                              GO:0003723, RNA binding 0.0180\n",
      "10 GO:0051082                 GO:0051082, unfolded protein binding 0.0188\n",
      "11 GO:0015171     GO:0015171, amino acid transmembrane transporter 0.0191\n",
      "12 GO:0030374    GO:0030374, nuclear receptor coactivator activity 0.0257\n",
      "13 GO:0031072               GO:0031072, heat shock protein binding 0.0386\n",
      "14 GO:0003727              GO:0003727, single-stranded RNA binding 0.0386\n",
      "15 GO:0001046      GO:0001046, core promoter sequence-specific DNA 0.0393\n",
      "16 GO:0003677                              GO:0003677, DNA binding 0.0405\n",
      "17 GO:0019902                      GO:0019902, phosphatase binding 0.0424\n",
      "18 GO:0005488                                  GO:0005488, binding 0.0474\n",
      "19 GO:0070273 GO:0070273, phosphatidylinositol-4-phosphate binding 0.0496\n",
      "20 GO:0005342   GO:0005342, organic acid transmembrane transporter 0.0499\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 7373 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 11325 GO terms and 25399 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 3799 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 3297 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term      KS\n",
      "1  GO:0007156     GO:0007156, homophilic cell adhesion via plasma 0.00043\n",
      "2  GO:0048010      GO:0048010, vascular endothelial growth factor 0.00055\n",
      "3  GO:0000122 GO:0000122, negative regulation of transcription by 0.00122\n",
      "4  GO:0090398                     GO:0090398, cellular senescence 0.00147\n",
      "5  GO:0030947      GO:0030947, regulation of vascular endothelial 0.00191\n",
      "6  GO:0070059   GO:0070059, intrinsic apoptotic signaling pathway 0.00222\n",
      "7  GO:0043484              GO:0043484, regulation of RNA splicing 0.00283\n",
      "8  GO:0043010             GO:0043010, camera-type eye development 0.00346\n",
      "9  GO:0006984            GO:0006984, ER-nucleus signaling pathway 0.00353\n",
      "10 GO:0035567     GO:0035567, non-canonical Wnt signaling pathway 0.00381\n",
      "11 GO:0090175   GO:0090175, regulation of establishment of planar 0.00431\n",
      "12 GO:0006725    GO:0006725, cellular aromatic compound metabolic 0.00462\n",
      "13 GO:0060071      GO:0060071, Wnt signaling pathway, planar cell 0.00497\n",
      "14 GO:1902235     GO:1902235, regulation of endoplasmic reticulum 0.00506\n",
      "15 GO:0090304          GO:0090304, nucleic acid metabolic process 0.00693\n",
      "16 GO:0098742  GO:0098742, cell-cell adhesion via plasma-membrane 0.00790\n",
      "17 GO:0034620   GO:0034620, cellular response to unfolded protein 0.00807\n",
      "18 GO:1903573      GO:1903573, negative regulation of response to 0.00864\n",
      "19 GO:1901360       GO:1901360, organic cyclic compound metabolic 0.01018\n",
      "20 GO:0046483           GO:0046483, heterocycle metabolic process 0.01045\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_2/2_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 326   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 322   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 492 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 791 GO terms and 1028 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 280 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 87 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GO.ID                                           Term     KS\n",
      "1 GO:0097159    GO:0097159, organic cyclic compound binding 0.0020\n",
      "2 GO:1901363      GO:1901363, heterocyclic compound binding 0.0029\n",
      "3 GO:0003723                        GO:0003723, RNA binding 0.0385\n",
      "4 GO:0003676               GO:0003676, nucleic acid binding 0.0412\n",
      "5 GO:0003712 GO:0003712, transcription coregulator activity 0.0416\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1770 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 4641 GO terms and 10177 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 269 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 591 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term     KS\n",
      "1  GO:0033365       GO:0033365, protein localization to organelle 0.0019\n",
      "2  GO:0051093    GO:0051093, negative regulation of developmental 0.0023\n",
      "3  GO:1904951 GO:1904951, positive regulation of establishment of 0.0026\n",
      "4  GO:0006518               GO:0006518, peptide metabolic process 0.0045\n",
      "5  GO:1903827          GO:1903827, regulation of cellular protein 0.0047\n",
      "6  GO:0032880      GO:0032880, regulation of protein localization 0.0048\n",
      "7  GO:0060341     GO:0060341, regulation of cellular localization 0.0076\n",
      "8  GO:0046907                 GO:0046907, intracellular transport 0.0087\n",
      "9  GO:0007005              GO:0007005, mitochondrion organization 0.0097\n",
      "10 GO:0034641    GO:0034641, cellular nitrogen compound metabolic 0.0117\n",
      "11 GO:1903829 GO:1903829, positive regulation of cellular protein 0.0119\n",
      "12 GO:0034613           GO:0034613, cellular protein localization 0.0133\n",
      "13 GO:0070727     GO:0070727, cellular macromolecule localization 0.0133\n",
      "14 GO:0061024                   GO:0061024, membrane organization 0.0135\n",
      "15 GO:0045596             GO:0045596, negative regulation of cell 0.0143\n",
      "16 GO:0043603        GO:0043603, cellular amide metabolic process 0.0169\n",
      "17 GO:0019953                     GO:0019953, sexual reproduction 0.0199\n",
      "18 GO:0048732                       GO:0048732, gland development 0.0238\n",
      "19 GO:0043933      GO:0043933, protein-containing complex subunit 0.0246\n",
      "20 GO:0010638        GO:0010638, positive regulation of organelle 0.0273\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_3/3_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 784   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 305   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 502 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 790 GO terms and 999 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 273 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 88 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GO.ID                                                 Term    KS\n",
      "1 GO:0140297 GO:0140297, DNA-binding transcription factor binding 0.011\n",
      "2 GO:0061629   GO:0061629, RNA polymerase II-specific DNA-binding 0.012\n",
      "3 GO:0003676                     GO:0003676, nucleic acid binding 0.026\n",
      "4 GO:0008134             GO:0008134, transcription factor binding 0.028\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1852 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 4702 GO terms and 10319 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 267 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 624 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term     KS\n",
      "1  GO:0044249           GO:0044249, cellular biosynthetic process 0.0011\n",
      "2  GO:1901576  GO:1901576, organic substance biosynthetic process 0.0011\n",
      "3  GO:0034645     GO:0034645, cellular macromolecule biosynthetic 0.0021\n",
      "4  GO:0009058                    GO:0009058, biosynthetic process 0.0022\n",
      "5  GO:0006979            GO:0006979, response to oxidative stress 0.0027\n",
      "6  GO:0009059      GO:0009059, macromolecule biosynthetic process 0.0030\n",
      "7  GO:0008152                       GO:0008152, metabolic process 0.0043\n",
      "8  GO:1901360       GO:1901360, organic cyclic compound metabolic 0.0051\n",
      "9  GO:2000112    GO:2000112, regulation of cellular macromolecule 0.0080\n",
      "10 GO:0044237              GO:0044237, cellular metabolic process 0.0107\n",
      "11 GO:0010556             GO:0010556, regulation of macromolecule 0.0110\n",
      "12 GO:0031326     GO:0031326, regulation of cellular biosynthetic 0.0132\n",
      "13 GO:0071704     GO:0071704, organic substance metabolic process 0.0143\n",
      "14 GO:0006725    GO:0006725, cellular aromatic compound metabolic 0.0145\n",
      "15 GO:0010033           GO:0010033, response to organic substance 0.0148\n",
      "16 GO:0006139          GO:0006139, nucleobase-containing compound 0.0165\n",
      "17 GO:0046483           GO:0046483, heterocycle metabolic process 0.0165\n",
      "18 GO:0030900                   GO:0030900, forebrain development 0.0174\n",
      "19 GO:0080090 GO:0080090, regulation of primary metabolic process 0.0174\n",
      "20 GO:0044238               GO:0044238, primary metabolic process 0.0175\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_4/4_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 269   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 268   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 416 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 689 GO terms and 866 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 237 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 76 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GO.ID                                                Term     KS\n",
      "1 GO:0003677                             GO:0003677, DNA binding 0.0023\n",
      "2 GO:0003676                    GO:0003676, nucleic acid binding 0.0103\n",
      "3 GO:0140110        GO:0140110, transcription regulator activity 0.0196\n",
      "4 GO:0043565           GO:0043565, sequence-specific DNA binding 0.0231\n",
      "5 GO:0005515                         GO:0005515, protein binding 0.0298\n",
      "6 GO:0001067 GO:0001067, transcription regulatory region nucleic 0.0336\n",
      "7 GO:0003690             GO:0003690, double-stranded DNA binding 0.0336\n",
      "8 GO:0000976     GO:0000976, transcription cis-regulatory region 0.0484\n",
      "9 GO:1990837   GO:1990837, sequence-specific double-stranded DNA 0.0484\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1538 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 4228 GO terms and 9315 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 233 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 515 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term      KS\n",
      "1  GO:0006974 GO:0006974, cellular response to DNA damage stimulus 0.00021\n",
      "2  GO:0090304           GO:0090304, nucleic acid metabolic process 0.00212\n",
      "3  GO:0031324          GO:0031324, negative regulation of cellular 0.00278\n",
      "4  GO:0006139           GO:0006139, nucleobase-containing compound 0.00294\n",
      "5  GO:0006725     GO:0006725, cellular aromatic compound metabolic 0.00300\n",
      "6  GO:0009889       GO:0009889, regulation of biosynthetic process 0.00363\n",
      "7  GO:0031326      GO:0031326, regulation of cellular biosynthetic 0.00363\n",
      "8  GO:0046483            GO:0046483, heterocycle metabolic process 0.00469\n",
      "9  GO:0031323 GO:0031323, regulation of cellular metabolic process 0.00526\n",
      "10 GO:0080090  GO:0080090, regulation of primary metabolic process 0.00609\n",
      "11 GO:0010556              GO:0010556, regulation of macromolecule 0.00610\n",
      "12 GO:1901360        GO:1901360, organic cyclic compound metabolic 0.00840\n",
      "13 GO:0051172          GO:0051172, negative regulation of nitrogen 0.00893\n",
      "14 GO:0016070                    GO:0016070, RNA metabolic process 0.00971\n",
      "15 GO:0009890      GO:0009890, negative regulation of biosynthetic 0.01035\n",
      "16 GO:0031327          GO:0031327, negative regulation of cellular 0.01035\n",
      "17 GO:0044237               GO:0044237, cellular metabolic process 0.01088\n",
      "18 GO:0045596              GO:0045596, negative regulation of cell 0.01113\n",
      "19 GO:0051649    GO:0051649, establishment of localization in cell 0.01235\n",
      "20 GO:0019219      GO:0019219, regulation of nucleobase-containing 0.01266\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_5/5_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 845   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 818   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 863 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 1270 GO terms and 1630 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 723 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 155 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term       KS\n",
      "1  GO:1901363            GO:1901363, heterocyclic compound binding 0.000088\n",
      "2  GO:0097159          GO:0097159, organic cyclic compound binding 0.000160\n",
      "3  GO:0003723                              GO:0003723, RNA binding 0.000220\n",
      "4  GO:0003676                     GO:0003676, nucleic acid binding 0.000990\n",
      "5  GO:0005515                          GO:0005515, protein binding 0.001350\n",
      "6  GO:0005488                                  GO:0005488, binding 0.005830\n",
      "7  GO:0004674 GO:0004674, protein serine/threonine kinase activity 0.014800\n",
      "8  GO:0016772       GO:0016772, transferase activity, transferring 0.016970\n",
      "9  GO:0043021        GO:0043021, ribonucleoprotein complex binding 0.019950\n",
      "10 GO:0016462                 GO:0016462, pyrophosphatase activity 0.025480\n",
      "11 GO:0016817       GO:0016817, hydrolase activity, acting on acid 0.025480\n",
      "12 GO:0016818       GO:0016818, hydrolase activity, acting on acid 0.025480\n",
      "13 GO:0017111       GO:0017111, nucleoside-triphosphatase activity 0.033590\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 3037 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 6416 GO terms and 14207 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 700 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 1125 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                              Term       KS\n",
      "1  GO:0006397                       GO:0006397, mRNA processing 0.000013\n",
      "2  GO:0046907               GO:0046907, intracellular transport 0.000020\n",
      "3  GO:0016570                  GO:0016570, histone modification 0.000200\n",
      "4  GO:0008380                          GO:0008380, RNA splicing 0.000210\n",
      "5  GO:0006396                        GO:0006396, RNA processing 0.000630\n",
      "6  GO:0044237            GO:0044237, cellular metabolic process 0.000970\n",
      "7  GO:0051649 GO:0051649, establishment of localization in cell 0.001120\n",
      "8  GO:0044248            GO:0044248, cellular catabolic process 0.001300\n",
      "9  GO:0009056                     GO:0009056, catabolic process 0.001350\n",
      "10 GO:0000375 GO:0000375, RNA splicing, via transesterification 0.001560\n",
      "11 GO:0000377 GO:0000377, RNA splicing, via transesterification 0.001560\n",
      "12 GO:0000398        GO:0000398, mRNA splicing, via spliceosome 0.001560\n",
      "13 GO:0030162             GO:0030162, regulation of proteolysis 0.001620\n",
      "14 GO:1901575   GO:1901575, organic substance catabolic process 0.001640\n",
      "15 GO:0016071                GO:0016071, mRNA metabolic process 0.001740\n",
      "16 GO:0022618    GO:0022618, ribonucleoprotein complex assembly 0.002060\n",
      "17 GO:0071826     GO:0071826, ribonucleoprotein complex subunit 0.002060\n",
      "18 GO:0050684         GO:0050684, regulation of mRNA processing 0.002730\n",
      "19 GO:0010498 GO:0010498, proteasomal protein catabolic process 0.003610\n",
      "20 GO:0031056    GO:0031056, regulation of histone modification 0.003900\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_6/6_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 301   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 300   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 480 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 769 GO terms and 977 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 263 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 85 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term        KS\n",
      "1  GO:0097159         GO:0097159, organic cyclic compound binding 0.0000015\n",
      "2  GO:1901363           GO:1901363, heterocyclic compound binding 0.0000026\n",
      "3  GO:0003676                    GO:0003676, nucleic acid binding 0.0000550\n",
      "4  GO:0043565           GO:0043565, sequence-specific DNA binding 0.0004300\n",
      "5  GO:0000977         GO:0000977, RNA polymerase II transcription 0.0015200\n",
      "6  GO:0003700        GO:0003700, DNA-binding transcription factor 0.0015200\n",
      "7  GO:0000976     GO:0000976, transcription cis-regulatory region 0.0035900\n",
      "8  GO:1990837   GO:1990837, sequence-specific double-stranded DNA 0.0035900\n",
      "9  GO:0000978 GO:0000978, RNA polymerase II cis-regulatory region 0.0040200\n",
      "10 GO:0000981        GO:0000981, DNA-binding transcription factor 0.0040200\n",
      "11 GO:0000987 GO:0000987, cis-regulatory region sequence-specific 0.0040200\n",
      "12 GO:0003677                             GO:0003677, DNA binding 0.0055100\n",
      "13 GO:0003723                             GO:0003723, RNA binding 0.0062500\n",
      "14 GO:0140110        GO:0140110, transcription regulator activity 0.0069700\n",
      "15 GO:0003690             GO:0003690, double-stranded DNA binding 0.0077100\n",
      "16 GO:0001067 GO:0001067, transcription regulatory region nucleic 0.0088200\n",
      "17 GO:0046914            GO:0046914, transition metal ion binding 0.0154700\n",
      "18 GO:0019899                          GO:0019899, enzyme binding 0.0205600\n",
      "19 GO:0035639      GO:0035639, purine ribonucleoside triphosphate 0.0256500\n",
      "20 GO:0043167                             GO:0043167, ion binding 0.0347700\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1643 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 4384 GO terms and 9687 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 260 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 530 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term      KS\n",
      "1  GO:0044238                GO:0044238, primary metabolic process 1.3e-05\n",
      "2  GO:0044260 GO:0044260, cellular macromolecule metabolic process 2.6e-05\n",
      "3  GO:0044271  GO:0044271, cellular nitrogen compound biosynthetic 3.1e-05\n",
      "4  GO:1901362     GO:1901362, organic cyclic compound biosynthetic 3.4e-05\n",
      "5  GO:0009889       GO:0009889, regulation of biosynthetic process 3.5e-05\n",
      "6  GO:0010556              GO:0010556, regulation of macromolecule 3.5e-05\n",
      "7  GO:0018130         GO:0018130, heterocycle biosynthetic process 5.2e-05\n",
      "8  GO:0019438   GO:0019438, aromatic compound biosynthetic process 5.2e-05\n",
      "9  GO:0034654           GO:0034654, nucleobase-containing compound 5.2e-05\n",
      "10 GO:0031326      GO:0031326, regulation of cellular biosynthetic 8.0e-05\n",
      "11 GO:0090304           GO:0090304, nucleic acid metabolic process 8.3e-05\n",
      "12 GO:0080090  GO:0080090, regulation of primary metabolic process 9.0e-05\n",
      "13 GO:0071704      GO:0071704, organic substance metabolic process 1.1e-04\n",
      "14 GO:2000112     GO:2000112, regulation of cellular macromolecule 1.2e-04\n",
      "15 GO:0006807      GO:0006807, nitrogen compound metabolic process 1.6e-04\n",
      "16 GO:0044237               GO:0044237, cellular metabolic process 1.9e-04\n",
      "17 GO:0006351             GO:0006351, transcription, DNA-templated 2.6e-04\n",
      "18 GO:0032774                 GO:0032774, RNA biosynthetic process 2.6e-04\n",
      "19 GO:0097659     GO:0097659, nucleic acid-templated transcription 2.6e-04\n",
      "20 GO:0008152                        GO:0008152, metabolic process 2.9e-04\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_7/7_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 345   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 335   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 520 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 847 GO terms and 1065 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 290 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 87 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GO.ID                                         Term     KS\n",
      "1 GO:0030234        GO:0030234, enzyme regulator activity 0.0011\n",
      "2 GO:0005515                  GO:0005515, protein binding 0.0086\n",
      "3 GO:0140110 GO:0140110, transcription regulator activity 0.0157\n",
      "4 GO:0003723                      GO:0003723, RNA binding 0.0212\n",
      "5 GO:0098772     GO:0098772, molecular function regulator 0.0313\n",
      "6 GO:0008092     GO:0008092, cytoskeletal protein binding 0.0348\n",
      "7 GO:0008134     GO:0008134, transcription factor binding 0.0366\n",
      "8 GO:0019899                   GO:0019899, enzyme binding 0.0461\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1659 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 4393 GO terms and 9640 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 285 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 550 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                Term      KS\n",
      "1  GO:1901575     GO:1901575, organic substance catabolic process 0.00084\n",
      "2  GO:0009056                       GO:0009056, catabolic process 0.00132\n",
      "3  GO:0006807     GO:0006807, nitrogen compound metabolic process 0.00186\n",
      "4  GO:0006325                  GO:0006325, chromatin organization 0.00308\n",
      "5  GO:0009725                     GO:0009725, response to hormone 0.00334\n",
      "6  GO:0048523 GO:0048523, negative regulation of cellular process 0.00346\n",
      "7  GO:1901564       GO:1901564, organonitrogen compound metabolic 0.00571\n",
      "8  GO:0044248              GO:0044248, cellular catabolic process 0.00586\n",
      "9  GO:1901565       GO:1901565, organonitrogen compound catabolic 0.00679\n",
      "10 GO:0009890     GO:0009890, negative regulation of biosynthetic 0.00720\n",
      "11 GO:0032880      GO:0032880, regulation of protein localization 0.00784\n",
      "12 GO:0010033           GO:0010033, response to organic substance 0.00804\n",
      "13 GO:0044238               GO:0044238, primary metabolic process 0.00827\n",
      "14 GO:0048519       GO:0048519, negative regulation of biological 0.00851\n",
      "15 GO:0010558    GO:0010558, negative regulation of macromolecule 0.00889\n",
      "16 GO:0031327         GO:0031327, negative regulation of cellular 0.00889\n",
      "17 GO:0051223         GO:0051223, regulation of protein transport 0.00892\n",
      "18 GO:0070201  GO:0070201, regulation of establishment of protein 0.00892\n",
      "19 GO:0046907                 GO:0046907, intracellular transport 0.01090\n",
      "20 GO:0080090 GO:0080090, regulation of primary metabolic process 0.01115\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_8/8_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 301   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 295   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 506 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 816 GO terms and 1042 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 271 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 87 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GO.ID                                                 Term     KS\n",
      "1 GO:0016772       GO:0016772, transferase activity, transferring 0.0064\n",
      "2 GO:0140297 GO:0140297, DNA-binding transcription factor binding 0.0109\n",
      "3 GO:0140110         GO:0140110, transcription regulator activity 0.0153\n",
      "4 GO:0008134             GO:0008134, transcription factor binding 0.0167\n",
      "5 GO:0016301                          GO:0016301, kinase activity 0.0172\n",
      "6 GO:0016773     GO:0016773, phosphotransferase activity, alcohol 0.0328\n",
      "7 GO:0097159          GO:0097159, organic cyclic compound binding 0.0428\n",
      "8 GO:1901363            GO:1901363, heterocyclic compound binding 0.0433\n",
      "9 GO:0001217      GO:0001217, DNA-binding transcription repressor 0.0445\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1784 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 4642 GO terms and 10249 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 259 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 611 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term      KS\n",
      "1  GO:0032268           GO:0032268, regulation of cellular protein 0.00035\n",
      "2  GO:0048523  GO:0048523, negative regulation of cellular process 0.00051\n",
      "3  GO:0048519        GO:0048519, negative regulation of biological 0.00054\n",
      "4  GO:0010608   GO:0010608, posttranscriptional regulation of gene 0.00067\n",
      "5  GO:0080090  GO:0080090, regulation of primary metabolic process 0.00071\n",
      "6  GO:0051246  GO:0051246, regulation of protein metabolic process 0.00076\n",
      "7  GO:0032270  GO:0032270, positive regulation of cellular protein 0.00079\n",
      "8  GO:0051247           GO:0051247, positive regulation of protein 0.00103\n",
      "9  GO:0060255    GO:0060255, regulation of macromolecule metabolic 0.00104\n",
      "10 GO:0051171          GO:0051171, regulation of nitrogen compound 0.00119\n",
      "11 GO:0031323 GO:0031323, regulation of cellular metabolic process 0.00151\n",
      "12 GO:0009892 GO:0009892, negative regulation of metabolic process 0.00181\n",
      "13 GO:0019222          GO:0019222, regulation of metabolic process 0.00236\n",
      "14 GO:0010467                          GO:0010467, gene expression 0.00239\n",
      "15 GO:0030217                   GO:0030217, T cell differentiation 0.00264\n",
      "16 GO:0010468            GO:0010468, regulation of gene expression 0.00351\n",
      "17 GO:0031399       GO:0031399, regulation of protein modification 0.00376\n",
      "18 GO:0031326      GO:0031326, regulation of cellular biosynthetic 0.00402\n",
      "19 GO:0009889       GO:0009889, regulation of biosynthetic process 0.00515\n",
      "20 GO:0010605     GO:0010605, negative regulation of macromolecule 0.00636\n",
      "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_9/9_DE.csv\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 387   8\n",
      "[1] \"after pct removal\"\n",
      "[1] 377   8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 488 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 766 GO terms and 975 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 318 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 95 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GO.ID                                        Term     KS\n",
      "1 GO:0003676            GO:0003676, nucleic acid binding 0.0012\n",
      "2 GO:0003723                     GO:0003723, RNA binding 0.0024\n",
      "3 GO:1901363   GO:1901363, heterocyclic compound binding 0.0038\n",
      "4 GO:0097159 GO:0097159, organic cyclic compound binding 0.0065\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Building most specific GOs .....\n",
      "\n",
      "\t( 1780 GO terms found. )\n",
      "\n",
      "\n",
      "Build GO DAG topology ..........\n",
      "\n",
      "\t( 4829 GO terms and 10679 relations. )\n",
      "\n",
      "\n",
      "Annotating nodes ...............\n",
      "\n",
      "\t( 311 genes annotated to the GO terms. )\n",
      "\n",
      "\n",
      "\t\t\t -- Classic Algorithm -- \n",
      "\n",
      "\t\t the algorithm is scoring 582 nontrivial nodes\n",
      "\t\t parameters: \n",
      "\t\t\t test statistic: ks\n",
      "\t\t\t score order: increasing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        GO.ID                                                 Term     KS\n",
      "1  GO:0010608   GO:0010608, posttranscriptional regulation of gene 0.0032\n",
      "2  GO:0051171          GO:0051171, regulation of nitrogen compound 0.0039\n",
      "3  GO:0080090  GO:0080090, regulation of primary metabolic process 0.0076\n",
      "4  GO:0031323 GO:0031323, regulation of cellular metabolic process 0.0110\n",
      "5  GO:0043170          GO:0043170, macromolecule metabolic process 0.0206\n",
      "6  GO:0044238                GO:0044238, primary metabolic process 0.0209\n",
      "7  GO:0009889       GO:0009889, regulation of biosynthetic process 0.0222\n",
      "8  GO:0051051         GO:0051051, negative regulation of transport 0.0233\n",
      "9  GO:0044237               GO:0044237, cellular metabolic process 0.0254\n",
      "10 GO:0010467                          GO:0010467, gene expression 0.0285\n",
      "11 GO:0044260 GO:0044260, cellular macromolecule metabolic process 0.0290\n",
      "12 GO:0006807      GO:0006807, nitrogen compound metabolic process 0.0301\n",
      "13 GO:0031326      GO:0031326, regulation of cellular biosynthetic 0.0325\n",
      "14 GO:0060255    GO:0060255, regulation of macromolecule metabolic 0.0341\n",
      "15 GO:0010468            GO:0010468, regulation of gene expression 0.0344\n",
      "16 GO:0010556              GO:0010556, regulation of macromolecule 0.0348\n",
      "17 GO:0006470                GO:0006470, protein dephosphorylation 0.0392\n",
      "18 GO:0000226    GO:0000226, microtubule cytoskeleton organization 0.0438\n",
      "19 GO:0050794           GO:0050794, regulation of cellular process 0.0451\n",
      "20 GO:0002009           GO:0002009, morphogenesis of an epithelium 0.0466\n"
     ]
    }
   ],
   "source": [
    "#for f in list.files(DE.path, pattern=\"*.csv\"){\n",
    "#DE_files <- list.files(path = DE.out.path,pattern=\"*.csv\", full.names=TRUE)\n",
    "for(d in DE_dirs){\n",
    "    f <- list.files(path=d, pattern=\"*_DE.csv\", full.names=T)\n",
    "    if (length(f)>1){\n",
    "        print(\"more than one file! check\")\n",
    "        print(f)\n",
    "    }\n",
    "    f = f[[1]]\n",
    "    print(f)\n",
    "    name <- basename(d)\n",
    "    curr <- read.csv(f)\n",
    "    f <- list.files(path=d, pattern=\"*_DE.csv\", full.names=T)[[1]]\n",
    "    if (!(nrow(curr)==0)){\n",
    "        colnames(curr) <- c(\"Gene\",\"pvalue\",\"log2FoldChange\",\"pct.1\",\"pct.2\",\"padj_bonf\", \"padj\")\n",
    "        curr$ensg <- mapIds(org.Hs.eg.db,\n",
    "                    keys=curr$Gene, \n",
    "                    column=\"ENSEMBL\",\n",
    "                    keytype=\"SYMBOL\",\n",
    "                    multiVals=\"first\")\n",
    "    }\n",
    "    print(dim(curr))\n",
    "    filt.curr <- curr[(curr$pct.1>0.1) & (curr$pct.2>0.1),]\n",
    "    print('after pct removal')\n",
    "    print(dim(filt.curr))\n",
    "    #filt.curr <- filt.curr[curr$padj<0.5,]\n",
    "    #print('after pval removal')\n",
    "    #print(dim(filt.curr))\n",
    "    filt.curr <- filt.curr[order(filt.curr$padj),]\n",
    "    filt.curr$logpadj <- -log10(filt.curr$padj)\n",
    "\n",
    "    sig.genes <- filt.curr[filt.curr$padj<go_pthresh, \"Gene\"]\n",
    "    universe.genes <- c()\n",
    "    for (row in row.names(filt.curr)){\n",
    "        universe.genes[filt.curr[row, \"Gene\"]] = filt.curr[row,\"padj\"]\n",
    "    }\n",
    "    try({\n",
    "        run.go(universe.genes, name=name, export.path=export.path, proc=\"MF\")\n",
    "        run.go(universe.genes, name=name, export.path=export.path, proc=\"BP\")\n",
    "    })\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b0a499c8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# require(ggplot2)\n",
    "# ggplot(goEnrichment, aes(x=Term, y=-log10(KS))) +\n",
    "#     stat_summary(geom = \"bar\", fun.y = mean, position = \"dodge\") +\n",
    "#     xlab(\"Biological process\") +\n",
    "#     ylab(\"Enrichment\") +\n",
    "#     ggtitle(\"Title\") +\n",
    "#     scale_y_continuous(breaks = round(seq(0, max(-log10(goEnrichment$KS)), by = 2), 1)) +\n",
    "#     theme_bw(base_size=24) +\n",
    "#     theme(\n",
    "#         legend.position='none',\n",
    "#         legend.background=element_rect(),\n",
    "#         plot.title=element_text(angle=0, size=24, face=\"bold\", vjust=1),\n",
    "#         axis.text.x=element_text(angle=0, size=18, face=\"bold\", hjust=1.10),\n",
    "#         axis.text.y=element_text(angle=0, size=18, face=\"bold\", vjust=0.5),\n",
    "#         axis.title=element_text(size=24, face=\"bold\"),\n",
    "#         legend.key=element_blank(),     #removes the border\n",
    "#         legend.key.size=unit(1, \"cm\"),      #Sets overall area/size of the legend\n",
    "#         legend.text=element_text(size=18),  #Text size\n",
    "#         title=element_text(size=18)) +\n",
    "#     guides(colour=guide_legend(override.aes=list(size=2.5))) +\n",
    "#     coord_flip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c8624952",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GO_pthresh.0.01/'"
      ],
      "text/latex": [
       "'/data/Mito\\_Trace/output/pipeline/TcellDupi\\_may17\\_2021/MTblacklist/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_hg38\\_1\\_2/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_p0.1/GO\\_pthresh.0.01/'"
      ],
      "text/markdown": [
       "'/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GO_pthresh.0.01/'"
      ],
      "text/plain": [
       "[1] \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GO_pthresh.0.01/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "export.path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8e3f179e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.3 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/isaac/miniconda3/envs/rgsea/lib/libopenblasp-r0.3.18.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] stringr_1.4.0         ggplot2_3.3.5         dplyr_1.0.7          \n",
       " [4] openxlsx_4.2.4        data.table_1.14.2     gdata_2.18.0         \n",
       " [7] DOSE_3.20.0           biomaRt_2.50.0        GSEABase_1.56.0      \n",
       "[10] annotate_1.72.0       XML_3.99-0.8          clusterProfiler_4.2.0\n",
       "[13] org.Hs.eg.db_3.14.0   ALL_1.36.0            topGO_2.46.0         \n",
       "[16] SparseM_1.81          GO.db_3.14.0          AnnotationDbi_1.56.1 \n",
       "[19] IRanges_2.28.0        S4Vectors_0.32.0      Biobase_2.54.0       \n",
       "[22] graph_1.72.0          BiocGenerics_0.40.0  \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] uuid_1.0-3             shadowtext_0.0.9       fastmatch_1.1-3       \n",
       "  [4] BiocFileCache_2.2.0    plyr_1.8.6             igraph_1.2.9          \n",
       "  [7] repr_1.1.3             lazyeval_0.2.2         splines_4.1.1         \n",
       " [10] BiocParallel_1.28.0    GenomeInfoDb_1.30.0    digest_0.6.29         \n",
       " [13] yulab.utils_0.0.4      htmltools_0.5.2        GOSemSim_2.20.0       \n",
       " [16] viridis_0.6.2          fansi_0.4.2            magrittr_2.0.1        \n",
       " [19] memoise_2.0.1          Biostrings_2.62.0      graphlayouts_0.7.2    \n",
       " [22] matrixStats_0.61.0     enrichplot_1.14.1      prettyunits_1.1.1     \n",
       " [25] colorspace_2.0-2       blob_1.2.2             rappdirs_0.3.3        \n",
       " [28] ggrepel_0.9.1          crayon_1.4.2           RCurl_1.98-1.5        \n",
       " [31] jsonlite_1.7.2         scatterpie_0.1.6       ape_5.5               \n",
       " [34] glue_1.5.1             polyclip_1.10-0        gtable_0.3.0          \n",
       " [37] zlibbioc_1.40.0        XVector_0.34.0         scales_1.1.1          \n",
       " [40] DBI_1.1.1              Rcpp_1.0.7             viridisLite_0.4.0     \n",
       " [43] xtable_1.8-4           progress_1.2.2         gridGraphics_0.5-1    \n",
       " [46] tidytree_0.3.6         bit_4.0.4              httr_1.4.2            \n",
       " [49] fgsea_1.20.0           RColorBrewer_1.1-2     ellipsis_0.3.2        \n",
       " [52] pkgconfig_2.0.3        farver_2.1.0           dbplyr_2.1.1          \n",
       " [55] utf8_1.2.2             ggplotify_0.1.0        tidyselect_1.1.1      \n",
       " [58] rlang_0.4.12           reshape2_1.4.4         munsell_0.5.0         \n",
       " [61] tools_4.1.1            cachem_1.0.6           downloader_0.4        \n",
       " [64] generics_0.1.1         RSQLite_2.2.8          evaluate_0.14         \n",
       " [67] fastmap_1.1.0          ggtree_3.2.0           bit64_4.0.5           \n",
       " [70] tidygraph_1.2.0        zip_2.2.0              purrr_0.3.4           \n",
       " [73] KEGGREST_1.34.0        ggraph_2.0.5           nlme_3.1-153          \n",
       " [76] aplot_0.1.1            DO.db_2.9              xml2_1.3.3            \n",
       " [79] compiler_4.1.1         filelock_1.0.2         curl_4.3.2            \n",
       " [82] png_0.1-7              treeio_1.18.0          tibble_3.1.6          \n",
       " [85] tweenr_1.0.2           stringi_1.7.6          lattice_0.20-45       \n",
       " [88] IRdisplay_1.0          Matrix_1.3-4           vctrs_0.3.8           \n",
       " [91] pillar_1.6.4           lifecycle_1.0.1        BiocManager_1.30.16   \n",
       " [94] bitops_1.0-7           patchwork_1.1.1        qvalue_2.26.0         \n",
       " [97] R6_2.5.1               gridExtra_2.3          MASS_7.3-54           \n",
       "[100] gtools_3.9.2           assertthat_0.2.1       withr_2.4.3           \n",
       "[103] GenomeInfoDbData_1.2.7 parallel_4.1.1         hms_1.1.1             \n",
       "[106] grid_4.1.1             ggfun_0.0.4            IRkernel_1.2          \n",
       "[109] tidyr_1.1.4            rvcheck_0.1.8          pbdZMQ_0.3-6          \n",
       "[112] ggforce_0.3.3          base64enc_0.1-3       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686d6487",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d0c01837",
   "metadata": {},
   "source": [
    "### "
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
