{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "41fdb08b",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# DE.out.path = \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0.1_pthresh_0.1/\"\n",
    "# export.path = \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_hg38_1_2/annotation_clones/de_btwncond_RNA/minPct_0_logfc0.1/GSEA_pthresh_0.1/\"\n",
    "# gsea_dir = \"/mnt/md0/isshamie/Projects/Mito_Trace/software/Bioinformatics_Tools/\"\n",
    "# pthresh = 0.1\n",
    "# stat_col = \"log2FoldChange\"\n",
    "# prefilter = TRUE\n",
    "# padjmethod = \"BH\"\n",
    "# gsea_pthresh = 0.1\n",
    "\n",
    "DE.out.path = \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust\"\n",
    "export.path = \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/\"\n",
    "gsea_dir = \"/mnt/md0/isshamie/Projects/Mito_Trace/software/Bioinformatics_Tools/\"\n",
    "\n",
    "pthresh = 0.1 \n",
    "gsea_pthresh = 0.1 \n",
    "prefilter = FALSE \n",
    "stat_col = \"log2FoldChange\"\n",
    "padjmethod = \"BH\"\n",
    "\n",
    "pct_thresh = 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "31043df1",
   "metadata": {},
   "outputs": [],
   "source": [
    "if ((prefilter==\"TRUE\") | (prefilter==TRUE)){\n",
    "    prefilter=TRUE\n",
    "}else{\n",
    "    prefilter=FALSE\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d59d5e98",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(export.path.prep):\n",
      "“'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep' already exists”\n"
     ]
    }
   ],
   "source": [
    "export.path.prep <- gsub(\"//\", \"/\", file.path(export.path, \"prep\"))\n",
    "dir.create(export.path.prep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "32f7ab3e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep'"
      ],
      "text/latex": [
       "'/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_p0.1/GSEA\\_pthresh.0.1\\_pref.FALSE\\_stat.log2FoldChange\\_padj.BH/prep'"
      ],
      "text/markdown": [
       "'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep'"
      ],
      "text/plain": [
       "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "export.path.prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4268e5d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#btwnClust.outdir <- file.path(outdir, \"btwnClust\")\n",
    "#DE.out.path <- file.path(DE.out.path, \"DE.out.path\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "68b43d85",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_0'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_1'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_10'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_11'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_12'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_13'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_14'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_2'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_3'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_4'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_5'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_6'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_7'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_8'</li><li>'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_9'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_0'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_1'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_10'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_11'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_12'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_13'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_14'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_2'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_3'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_4'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_5'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_6'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_7'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_8'\n",
       "\\item '/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_pthresh\\_0.1/btwnConds\\_inClust/clust\\_9'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_0'\n",
       "2. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_1'\n",
       "3. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_10'\n",
       "4. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_11'\n",
       "5. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_12'\n",
       "6. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_13'\n",
       "7. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_14'\n",
       "8. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_2'\n",
       "9. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_3'\n",
       "10. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_4'\n",
       "11. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_5'\n",
       "12. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_6'\n",
       "13. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_7'\n",
       "14. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_8'\n",
       "15. '/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_9'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_0\" \n",
       " [2] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_1\" \n",
       " [3] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_10\"\n",
       " [4] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_11\"\n",
       " [5] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_12\"\n",
       " [6] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_13\"\n",
       " [7] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_14\"\n",
       " [8] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_2\" \n",
       " [9] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_3\" \n",
       "[10] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_4\" \n",
       "[11] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_5\" \n",
       "[12] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_6\" \n",
       "[13] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_7\" \n",
       "[14] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_8\" \n",
       "[15] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_pthresh_0.1/btwnConds_inClust/clust_9\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "files <- list.files(path=DE.out.path,pattern = \"clust_*\", full.names = T, recursive = FALSE, include.dirs = TRUE)\n",
    "DE_dirs <- files[file.info(files)$isdir ]\n",
    "DE_dirs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ea4fc8d1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: clusterProfiler\n",
      "\n",
      "\n",
      "\n",
      "clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:clusterProfiler\"\n",
      "[1] \"package already installed:clusterProfiler\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GSEABase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: annotate\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Loading required package: XML\n",
      "\n",
      "Loading required package: graph\n",
      "\n",
      "\n",
      "Attaching package: ‘graph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:XML’:\n",
      "\n",
      "    addNode\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:GSEABase\"\n",
      "[1] \"package already installed:GSEABase\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: biomaRt\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:biomaRt\"\n",
      "[1] \"package already installed:biomaRt\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: org.Hs.eg.db\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:org.Hs.eg.db\"\n",
      "[1] \"package already installed:org.Hs.eg.db\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: DOSE\n",
      "\n",
      "DOSE v3.20.1  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use DOSE in published research, please cite:\n",
      "Guangchuang Yu, Li-Gen Wang, Guang-Rong Yan, Qing-Yu He. DOSE: an R/Bioconductor package for Disease Ontology Semantic and Enrichment analysis. Bioinformatics 2015, 31(4):608-609\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"package already installed:DOSE\"\n",
      "[1] \"package already installed:DOSE\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: gdata\n",
      "\n",
      "gdata: read.xls support for 'XLS' (Excel 97-2004) files ENABLED.\n",
      "\n",
      "\n",
      "\n",
      "gdata: Unable to load perl libaries needed by read.xls()\n",
      "gdata: to support 'XLSX' (Excel 2007+) files.\n",
      "\n",
      "\n",
      "\n",
      "gdata: Run the function 'installXLSXsupport()'\n",
      "gdata: to automatically download and install the perl\n",
      "gdata: libaries needed to support Excel XLS and XLSX formats.\n",
      "\n",
      "\n",
      "Attaching package: ‘gdata’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    startsWith, trim\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, first<-\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats4’:\n",
      "\n",
      "    nobs\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    nobs\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    object.size\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    startsWith\n",
      "\n",
      "\n",
      "Loading required package: data.table\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:gdata’:\n",
      "\n",
      "    first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "Loading required package: openxlsx\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:gdata’:\n",
      "\n",
      "    combine, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GSEABase’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graph’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘stringr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graph’:\n",
      "\n",
      "    boundary\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# library(\"AnnotationDbi\")\n",
    "#library(\"org.Hs.eg.db\")\n",
    "# library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "source(paste0(gsea_dir,'R/Gsea.R'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "895b2b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols.prior <- c(\"Gene\",\"baseMean\",\"log2FoldChange\",\"lfcSE\",\"stat\",\"pvalue\",\"padj\")\n",
    "cols.needed <- c(\"Gene\",\"log2FoldChange\")\n",
    "cols.curr <- c(\"X\",\"p_val\",\"avg_log2FC\",\"pct.1\",\"pct.2\",\"p_val_adj_BH\") #first col is empty in csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9270a190",
   "metadata": {},
   "source": [
    "## Prep files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "541f8ea4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#for f in list.files(DE.path, pattern=\"*.csv\"){\n",
    "\n",
    "DE_files <- list.files(path = DE.out.path,pattern=\"*.csv\", full.names=TRUE)\n",
    "for(d in DE_dirs){\n",
    "    f <- list.files(path=d, pattern=\"*_DE.csv\", full.names=T)\n",
    "    if (length(f)>1){\n",
    "        print(\"more than one file! check\")\n",
    "        print(f)\n",
    "    }\n",
    "    if (length(f)>0){\n",
    "        f = f[[1]]\n",
    "\n",
    "        curr <- read.csv(f)\n",
    "\n",
    "\n",
    "        if (!(nrow(curr)==0)){\n",
    "            colnames(curr) <- c(\"Gene\",\"pvalue\",\"log2FoldChange\",\"pct.1\",\"pct.2\",\"padj_bonf\", \"padj\")\n",
    "            curr$ensg <- mapIds(org.Hs.eg.db,\n",
    "                        keys=curr$Gene, \n",
    "                        column=\"ENSEMBL\",\n",
    "                        keytype=\"SYMBOL\",\n",
    "                        multiVals=\"first\")\n",
    "            write.table(curr, file.path(export.path.prep, paste0(strsplit(basename(f), \".csv\")[[1]], \".tsv\")),quote = FALSE,\n",
    "                        row.names=FALSE, sep='\\t')\n",
    "        }\n",
    "\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "06675344",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/'"
      ],
      "text/latex": [
       "'/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_p0.1/GSEA\\_pthresh.0.1\\_pref.FALSE\\_stat.log2FoldChange\\_padj.BH/'"
      ],
      "text/markdown": [
       "'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/'"
      ],
      "text/plain": [
       "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "export.path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c641fe0f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/0_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 109   7\n",
      "[1] 109   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "             pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "EPHA4  2.017617e-12      0.1240405 0.336 0.186 3.841341e-08 1.692874e-10\n",
      "CD226  3.106191e-12      0.1426756 0.284 0.151 5.913877e-08 1.692874e-10\n",
      "ABCA13 5.096611e-12      0.1549282 0.559 0.362 9.703437e-08 1.851769e-10\n",
      "EPHA5  9.747686e-12      0.1137204 0.246 0.125 1.855862e-07 2.656244e-10\n",
      "BMPR1B 3.800440e-11      0.1257817 0.404 0.242 7.235658e-07 8.284960e-10\n",
      "TP63   9.553686e-11      0.1025820 0.301 0.168 1.818926e-06 1.735586e-09\n",
      "                  ensg\n",
      "EPHA4  ENSG00000116106\n",
      "CD226  ENSG00000150637\n",
      "ABCA13 ENSG00000179869\n",
      "EPHA5  ENSG00000145242\n",
      "BMPR1B ENSG00000138696\n",
      "TP63   ENSG00000073282\n",
      "[1] \"here\"\n",
      "         Gene log2FoldChange\n",
      "ZEB2     ZEB2      0.1757190\n",
      "DACH1   DACH1      0.1718503\n",
      "ABCA13 ABCA13      0.1549282\n",
      "PLXDC2 PLXDC2      0.1496271\n",
      "CD226   CD226      0.1426756\n",
      "SORL1   SORL1      0.1374691\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/1_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 133   7\n",
      "[1] 133   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "              pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "SYTL2   1.922296e-13      0.1503209 0.285 0.143 3.659860e-09 2.556654e-11\n",
      "CDKL5   1.189174e-12      0.1079326 0.249 0.121 2.264068e-08 7.554433e-11\n",
      "AK5     2.551141e-12      0.1798796 0.547 0.349 4.857117e-08 7.554433e-11\n",
      "ST3GAL4 3.888905e-12      0.1438495 0.455 0.271 7.404087e-08 7.554433e-11\n",
      "FAM49A  4.209267e-12      0.1326743 0.357 0.200 8.014023e-08 7.554433e-11\n",
      "EBF1    4.939103e-12      0.1518110 0.550 0.343 9.403558e-08 7.554433e-11\n",
      "                   ensg\n",
      "SYTL2   ENSG00000137501\n",
      "CDKL5   ENSG00000008086\n",
      "AK5     ENSG00000154027\n",
      "ST3GAL4 ENSG00000110080\n",
      "FAM49A             <NA>\n",
      "EBF1    ENSG00000164330\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "AK5         AK5      0.1798796\n",
      "DTNA       DTNA      0.1645802\n",
      "EBF1       EBF1      0.1518110\n",
      "SYTL2     SYTL2      0.1503209\n",
      "MCC         MCC      0.1461866\n",
      "ST3GAL4 ST3GAL4      0.1438495\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/10_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 879   7\n",
      "[1] 790   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "               pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "PLD1     4.065966e-09      0.2618734 0.439 0.210 7.741193e-05 3.573984e-06\n",
      "ARHGAP32 6.639244e-08      0.2134041 0.426 0.205 1.264046e-03 2.214655e-05\n",
      "WSCD2    7.558548e-08      0.1375831 0.304 0.121 1.439072e-03 2.214655e-05\n",
      "SAMD4A   2.166277e-07      0.2727725 0.523 0.306 4.124374e-03 4.383664e-05\n",
      "P2RY12   2.493552e-07      0.1326471 0.308 0.129 4.747474e-03 4.383664e-05\n",
      "MTCL1    3.997472e-07      0.1368106 0.333 0.147 7.610786e-03 5.420226e-05\n",
      "                    ensg\n",
      "PLD1     ENSG00000075651\n",
      "ARHGAP32 ENSG00000134909\n",
      "WSCD2    ENSG00000075035\n",
      "SAMD4A   ENSG00000020577\n",
      "P2RY12   ENSG00000169313\n",
      "MTCL1    ENSG00000168502\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "SAMD4A   SAMD4A      0.2727725\n",
      "PLD1       PLD1      0.2618734\n",
      "ZC3HAV1 ZC3HAV1      0.2297555\n",
      "ATP2B2   ATP2B2      0.2204152\n",
      "MRTFB     MRTFB      0.2169843\n",
      "ACYP2     ACYP2      0.2158542\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.13% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.13% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.13% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.13% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/14_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 10522     7\n",
      "[1] 1861    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "            pvalue log2FoldChange pct.1 pct.2 padj_bonf      padj\n",
      "CHST11 0.001780282     -0.8566105 0.143 0.583         1 0.6004934\n",
      "IGF1R  0.005183336     -0.9056562 0.143 0.500         1 0.6004934\n",
      "MSI2   0.005644347     -0.9128558 0.179 0.542         1 0.6004934\n",
      "TTC7A  0.005728831     -0.5278580 0.143 0.542         1 0.6004934\n",
      "ZMIZ1  0.006156191     -1.0676660 0.250 0.583         1 0.6004934\n",
      "KDM4B  0.010408198     -0.4815826 0.107 0.458         1 0.6004934\n",
      "                  ensg\n",
      "CHST11 ENSG00000171310\n",
      "IGF1R  ENSG00000140443\n",
      "MSI2   ENSG00000153944\n",
      "TTC7A  ENSG00000068724\n",
      "ZMIZ1  ENSG00000108175\n",
      "KDM4B  ENSG00000127663\n",
      "[1] \"here\"\n",
      "             Gene log2FoldChange\n",
      "CATSPERG CATSPERG      0.9949531\n",
      "GRIA1       GRIA1      0.9093972\n",
      "SHC1         SHC1      0.8830331\n",
      "CKS1B       CKS1B      0.8830331\n",
      "MYO3B       MYO3B      0.8400258\n",
      "NCALD       NCALD      0.8391438\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (1.77% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (1.77% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (1.77% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (1.77% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/2_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 7717    7\n",
      "[1] 2422    7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "              pvalue log2FoldChange pct.1 pct.2 padj_bonf      padj\n",
      "LDLRAD3 0.0002749369      0.5950001 0.642 0.208         1 0.1525022\n",
      "NOL4L   0.0004325446      0.6319742 0.509 0.146         1 0.2042876\n",
      "NREP    0.0006221211      0.6668521 0.453 0.125         1 0.2286147\n",
      "SSBP2   0.0008256349      0.5212288 0.755 0.333         1 0.2633815\n",
      "MCPH1   0.0008361926      0.6418661 0.396 0.104         1 0.2633815\n",
      "RPS6KA5 0.0012254329      0.3099057 0.547 0.167         1 0.2699762\n",
      "                   ensg\n",
      "LDLRAD3 ENSG00000179241\n",
      "NOL4L   ENSG00000197183\n",
      "NREP    ENSG00000134986\n",
      "SSBP2   ENSG00000145687\n",
      "MCPH1   ENSG00000147316\n",
      "RPS6KA5 ENSG00000100784\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "NREP       NREP      0.6668521\n",
      "PBX3       PBX3      0.6537109\n",
      "MCPH1     MCPH1      0.6418661\n",
      "NOL4L     NOL4L      0.6319742\n",
      "MCUB       MCUB      0.6090708\n",
      "LDLRAD3 LDLRAD3      0.5950001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.33% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.33% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.33% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.33% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/4_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 410   7\n",
      "[1] 378   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "             pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "MFSD1  1.166910e-15      0.1992303 0.397 0.193 2.221680e-11 4.784332e-13\n",
      "XRCC5  1.207440e-12      0.1766518 0.397 0.209 2.298845e-08 2.475252e-10\n",
      "INTS4  3.966999e-12      0.1552789 0.319 0.156 7.552769e-08 4.128524e-10\n",
      "KCTD1  4.027829e-12      0.1693959 0.402 0.216 7.668583e-08 4.128524e-10\n",
      "UBE3C  5.870835e-12      0.1540731 0.377 0.198 1.117748e-07 4.814085e-10\n",
      "HS3ST5 7.200431e-12      0.1848314 0.298 0.144 1.370890e-07 4.920295e-10\n",
      "                  ensg\n",
      "MFSD1  ENSG00000118855\n",
      "XRCC5  ENSG00000079246\n",
      "INTS4  ENSG00000149262\n",
      "KCTD1  ENSG00000134504\n",
      "UBE3C  ENSG00000009335\n",
      "HS3ST5 ENSG00000249853\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "MFSD1     MFSD1      0.1992303\n",
      "HBS1L     HBS1L      0.1941069\n",
      "CMSS1     CMSS1      0.1936307\n",
      "HS3ST5   HS3ST5      0.1848314\n",
      "C2orf88 C2orf88      0.1792274\n",
      "PDE1A     PDE1A      0.1787444\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/5_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 153   7\n",
      "[1] 153   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "              pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "GMPS    1.522617e-13      0.1621664 0.408 0.197 2.898910e-09 2.329604e-11\n",
      "SMIM36  4.713900e-11      0.1243984 0.255 0.107 8.974794e-07 2.726975e-09\n",
      "PCSK5   5.608644e-11      0.1645686 0.582 0.341 1.067830e-06 2.726975e-09\n",
      "TRAK2   7.129346e-11      0.1402696 0.404 0.209 1.357356e-06 2.726975e-09\n",
      "ALDH1A1 1.146705e-10      0.1077136 0.492 0.262 2.183211e-06 3.508917e-09\n",
      "DRAM1   1.606987e-10      0.1022965 0.414 0.216 3.059543e-06 4.097817e-09\n",
      "                   ensg\n",
      "GMPS    ENSG00000163655\n",
      "SMIM36             <NA>\n",
      "PCSK5   ENSG00000099139\n",
      "TRAK2   ENSG00000115993\n",
      "ALDH1A1 ENSG00000165092\n",
      "DRAM1   ENSG00000136048\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "ASAP1     ASAP1      0.2160865\n",
      "PIP5K1B PIP5K1B      0.1839702\n",
      "PCSK5     PCSK5      0.1645686\n",
      "GMPS       GMPS      0.1621664\n",
      "ANAPC10 ANAPC10      0.1590871\n",
      "PDZRN4   PDZRN4      0.1548154\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/6_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 197   7\n",
      "[1] 192   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "             pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "ROR2   8.067654e-17      0.2294393 0.556 0.286 1.536001e-12 1.589328e-14\n",
      "KCNH1  2.414583e-13      0.2379400 0.610 0.364 4.597124e-09 2.151446e-11\n",
      "TTC13  3.276314e-13      0.1397417 0.293 0.121 6.237774e-09 2.151446e-11\n",
      "OLFM3  5.372290e-12      0.1098570 0.268 0.110 1.022830e-07 1.913467e-10\n",
      "MARCH8 5.850412e-12      0.1778300 0.487 0.265 1.113860e-07 1.913467e-10\n",
      "HS3ST5 1.374645e-11      0.1346201 0.326 0.150 2.617187e-07 3.385064e-10\n",
      "                  ensg\n",
      "ROR2   ENSG00000169071\n",
      "KCNH1  ENSG00000143473\n",
      "TTC13  ENSG00000143643\n",
      "OLFM3  ENSG00000118733\n",
      "MARCH8            <NA>\n",
      "HS3ST5 ENSG00000249853\n",
      "[1] \"here\"\n",
      "           Gene log2FoldChange\n",
      "KCNH1     KCNH1      0.2379400\n",
      "ROR2       ROR2      0.2294393\n",
      "COL25A1 COL25A1      0.2021173\n",
      "MARCH8   MARCH8      0.1778300\n",
      "SEMA4D   SEMA4D      0.1768760\n",
      "ABLIM1   ABLIM1      0.1689419\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/7_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 747   7\n",
      "[1] 674   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "              pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "DNER    3.485599e-08      0.1840395 0.450 0.235 0.0006636231 1.301871e-05\n",
      "SPATA16 2.635041e-07      0.1680560 0.325 0.160 0.0050168540 5.203633e-05\n",
      "THSD7A  3.450379e-07      0.2300477 0.447 0.258 0.0065691767 5.203633e-05\n",
      "ZNF608  3.556771e-07      0.2411030 0.434 0.243 0.0067717368 5.203633e-05\n",
      "ITGA1   5.165165e-07      0.1815900 0.407 0.227 0.0098339574 5.203633e-05\n",
      "RARB    5.390815e-07      0.2193483 0.488 0.288 0.0102635734 5.203633e-05\n",
      "                   ensg\n",
      "DNER    ENSG00000187957\n",
      "SPATA16 ENSG00000144962\n",
      "THSD7A  ENSG00000005108\n",
      "ZNF608  ENSG00000168916\n",
      "ITGA1   ENSG00000213949\n",
      "RARB    ENSG00000077092\n",
      "[1] \"here\"\n",
      "         Gene log2FoldChange\n",
      "ZNF608 ZNF608      0.2411030\n",
      "IKZF1   IKZF1      0.2410299\n",
      "THSD7A THSD7A      0.2300477\n",
      "ITGA4   ITGA4      0.2203764\n",
      "RARB     RARB      0.2193483\n",
      "KIF2A   KIF2A      0.2150421\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n",
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/8_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 396   7\n",
      "[1] 387   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "              pvalue log2FoldChange pct.1 pct.2    padj_bonf         padj\n",
      "UNC13B  2.830177e-09      0.1522827 0.430 0.211 5.388374e-05 1.120750e-06\n",
      "MORC1   1.570006e-08      0.1248317 0.290 0.124 2.989134e-04 2.234700e-06\n",
      "GNGT1   1.933499e-08      0.1528935 0.317 0.144 3.681189e-04 2.234700e-06\n",
      "PIK3CA  2.805257e-08      0.1815647 0.358 0.175 5.340929e-04 2.234700e-06\n",
      "COA1    2.821591e-08      0.1855147 0.484 0.266 5.372027e-04 2.234700e-06\n",
      "DYNC2H1 3.846067e-08      0.1661897 0.384 0.195 7.322527e-04 2.538404e-06\n",
      "                   ensg\n",
      "UNC13B  ENSG00000198722\n",
      "MORC1   ENSG00000114487\n",
      "GNGT1   ENSG00000127928\n",
      "PIK3CA  ENSG00000121879\n",
      "COA1    ENSG00000106603\n",
      "DYNC2H1 ENSG00000187240\n",
      "[1] \"here\"\n",
      "             Gene log2FoldChange\n",
      "ARHGAP10 ARHGAP10      0.2117806\n",
      "CACNA1E   CACNA1E      0.1869073\n",
      "COA1         COA1      0.1855147\n",
      "DOCK4       DOCK4      0.1826692\n",
      "PIK3CA     PIK3CA      0.1815647\n",
      "ELF2         ELF2      0.1727307\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.52% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.52% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.52% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "Warning message in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, :\n",
      "“There are ties in the preranked stats (0.52% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.”\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Here we go .... :/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/prep/9_DE.tsv\"\n",
      "[1] \"Myclass\"\n",
      "[1] 4034    7\n",
      "[1] 519   7\n",
      "[1] \"pvalue\"         \"log2FoldChange\" \"pct.1\"          \"pct.2\"         \n",
      "[5] \"padj_bonf\"      \"padj\"           \"ensg\"          \n",
      "              pvalue log2FoldChange pct.1 pct.2 padj_bonf       padj\n",
      "CERS6   1.648515e-05      0.3746728 0.380 0.228 0.3138608 0.03004872\n",
      "FRYL    5.447316e-04      0.3786708 0.244 0.143 1.0000000 0.12575794\n",
      "RNLS    7.992097e-04      0.4376675 0.227 0.135 1.0000000 0.13358737\n",
      "FAM172A 1.601268e-03      0.3229593 0.207 0.120 1.0000000 0.14629684\n",
      "ARMH4   1.652947e-03      0.3804059 0.310 0.207 1.0000000 0.14629684\n",
      "PIP5K1B 1.673178e-03      0.3738836 0.231 0.141 1.0000000 0.14629684\n",
      "                   ensg\n",
      "CERS6   ENSG00000172292\n",
      "FRYL    ENSG00000075539\n",
      "RNLS    ENSG00000184719\n",
      "FAM172A ENSG00000113391\n",
      "ARMH4   ENSG00000139971\n",
      "PIP5K1B ENSG00000107242\n",
      "[1] \"here\"\n",
      "         Gene log2FoldChange\n",
      "RNLS     RNLS      0.4376675\n",
      "FBXW7   FBXW7      0.4064176\n",
      "TG         TG      0.3929188\n",
      "LPIN1   LPIN1      0.3927306\n",
      "FAM13A FAM13A      0.3893987\n",
      "CPEB3   CPEB3      0.3840299\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "no term enriched under specific pvalueCutoff...\n",
      "\n",
      "preparing geneSet collections...\n",
      "\n",
      "GSEA analysis...\n",
      "\n",
      "leading edge analysis...\n",
      "\n",
      "done...\n",
      "\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n",
      "Warning message in is.na(egmt[[c]]):\n",
      "“is.na() applied to non-(list or vector) of type 'S4'”\n",
      "Warning message in summary(egmt[[c]]):\n",
      "“summary method to convert the object to data.frame is deprecated, please use as.data.frame instead.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"preExcel writing\"\n",
      "[1] \"postExcel writing\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>.Primitive(\"return\")</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       ".Primitive(\"return\")\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       ".Primitive(\"return\")\n",
       "```"
      ],
      "text/plain": [
       ".Primitive(\"return\")"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#rm(list=ls())\n",
    "\n",
    "########################################################################################\n",
    "##  gsea_ex1\n",
    "#---------------------------------------------------------------------------------------\n",
    "#' @description This example is designed to demonstrate how to run gsea analysis on the \n",
    "#'              input files (DESeq2 results) put in the same directory\n",
    "#' @details All the parameter definitions is in the below section of \"Setting Parameter\"\n",
    "#' @return out a list that summarizes the GSEA enrichment analysis\n",
    "#' \n",
    "#' @export exportExcel_fp an excel file of the 'out' can be exported\n",
    "#' @export exportPDF_fp an figure of the GSEA enrichment analysis\n",
    "#' --------------------------------------------------------------------------------------------------\n",
    "#' @author Austin W.T. Chiang\n",
    "#' @note   v1.4 -- 2017.02.15 Add description of ex1, ex2, and idmapping file\n",
    "#' @note   v1.3 -- 2017.02.08 Add function of displaying topN pathways and optimze code\n",
    "#' @note   v1.2 -- 2017.02.06 Add zebrafish2 human id mapping\n",
    "#' @note   v1.1 -- 2017.01.19 Midified for BEN\n",
    "#' @note   v1.0 -- 2016.12.15 File created\n",
    "#' --------------------------------------------------------------------------------------------------\n",
    "########################################################################################\n",
    "gsea_ex1 <- function(DE_out_path = NULL, export_path = NULL,\n",
    "                     IDmappingfile = NULL, convert = FALSE,\n",
    "                     gene_name = TRUE, stat.col=\"log2FoldChange\", GSEA_set = c('c0', 'c2', 'c3'),\n",
    "                     minGSSize = 30, pAdjustMethod = \"none\",\n",
    "                     pvalueCutoff = 0.01, export = \"BOTH\",\n",
    "                     ShowTop = TRUE, TopN = 5, gene_sets_dir=NULL){\n",
    "  DE_out_files <- list.files(path = DE_out_path,pattern=\"*.tsv\", full.names = TRUE)\n",
    "  for(f in DE_out_files){\n",
    "    print(paste0(\"Here we go .... :\", f))\n",
    "    #f <- \"res_geno_dobs_removed_C57BL6_J__M.tsv\"\n",
    "    ## Input the DESeq result file\n",
    "    myClass <- read.csv(f, row.names = 1,sep=\"\\t\")\n",
    "    if(convert == TRUE){\n",
    "      geneMapTable <- read.csv(IDmappingfile, sep = \"\\t\")\n",
    "      myClass <- useIDmappingFile(myClass=myClass,\n",
    "                                  geneMapTable=geneMapTable) #,\n",
    "                                  #Original_attribute=Original_attribute,\n",
    "                                 # Convert_attribute=Co nvert_attribute)\n",
    "    }\n",
    "    \n",
    "    #DEGTable <- subset(myClass, abs(log2FoldChange) >= .58 & (padj < 0.05) )\n",
    "    if (prefilter){\n",
    "        DEGTable <- subset(myClass, (padj < pthresh) )\n",
    "        }\n",
    "      \n",
    "    print(\"Myclass\")\n",
    "    print(dim(myClass))\n",
    "    DEGTable <- subset(myClass, (pct.1 > pct_thresh & pct.2 > pct_thresh) )\n",
    "    print(dim(DEGTable))\n",
    "\n",
    "    print(colnames(DEGTable))\n",
    "    print(head(DEGTable))\n",
    "    if(gene_name==TRUE){\n",
    "      DEGTable$Gene <- DEGTable$gene_name\n",
    "    }else{\n",
    "      print('here')\n",
    "      DEGTable$Gene <- row.names(DEGTable)\n",
    "    }\n",
    "    \n",
    "    DEGTable <- DEGTable[with(DEGTable, order(-log2FoldChange)), ] #order\n",
    "    DEGTable <- DEGTable[,c(\"Gene\",stat.col)]\n",
    "    if (stat.col==\"padj\"){\n",
    "        DEGTable[,stat.col] <- -log10(DEGTable[,stat.col])\n",
    "        DEGTable <- DEGTable[with(DEGTable, order(-padj)), ] #order\n",
    "    }\n",
    "    \n",
    "    \n",
    "\n",
    "    #DEGTable <- DEGTable[order(stat.col, ), ]\n",
    "      \n",
    "    print(head(DEGTable))\n",
    "    ## do the GSEA analysis\n",
    "    f1 <- gsub(\".csv\", \"\", basename(f))\n",
    "    exportExcel_fp = paste0(export_path,f1,\"_gsea.xlsx\")\n",
    "    exportPDF_fp = paste0(export_path,f1,\"_gsea.pdf\")\n",
    "    exportDOT_fp = file.path(export_path,paste0(f1,\"_gsea.dot\"))\n",
    "    GSEA_res <-  AnnotDEG_byGSEA(DEGTable = DEGTable, GSEA_set = GSEA_set,\n",
    "                                 export = export, ShowTop = ShowTop, TopN = TopN,\n",
    "                                 exportExcel_fp = exportExcel_fp,\n",
    "                                 exportPDF_fp = exportPDF_fp,\n",
    "                                 minGSSize = minGSSize, exportDOT_fp=exportDOT_fp,\n",
    "                                 pAdjustMethod = pAdjustMethod, stat.col=stat.col,\n",
    "                                 pvalueCutoff = pvalueCutoff, gene_sets_dir = gene_sets_dir)\n",
    "    \n",
    "    #dotplot(GSEA_res)  \n",
    "    #ggsave(file.path(export_path, paste0(f1, \"_gsea.dot.pdf\")))\n",
    "    \n",
    "  }\n",
    "    return\n",
    "}\n",
    "\n",
    "\n",
    "########################################################################################\n",
    "##  Setting Parameter\n",
    "########################################################################################\n",
    "#' @param DE_out_path the path of the DESeq2 output files \n",
    "#' @param export_path the path that user want to export the GSEA results\n",
    "#' @param convert 'TRUE': need to do id converion; 'FALSE': no id conversion needed  \n",
    "#' @param IDmappingfile the idmapping file show that user's id to human gene name\n",
    "#' @param gene_name 'TRUE': there is a columan \"gene_name\" in the input files (i.e. DESeq2 output files)\n",
    "#' @param GSEA_set = c('c0', 'c2', 'c3')\n",
    "#'                 co - hallmark gene sets;       c1 - positional gene sets; \n",
    "#'                 c2 - curated gene sets;        c3 - motif gene sets;\n",
    "#'                 c4 - computational gene sets;  c5 - GO gene sets;\n",
    "#'                 c6 - oncogenic signatures;     c7 - immunologic signatures. \n",
    "#'                 (http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C2)\n",
    "#' @param minGSSize minimum size of genes for a specific GSEA gene set\n",
    "#' @param pAdjustMethod one of \"holm\", \"hochberg\", \"hommel\", \"bonferroni\", \"BH\", \"BY\", \"fdr\", \"none\"\n",
    "#' @param pvalueCutoff cutoff of pvalue in the GSEA enrichment analysis \n",
    "#' @param export option for exporting GSEA result: \"EXCEL\", \"PDF\", or \"BOTH\"\n",
    "#' @param ShowTop 'TRUE': the visualization will display only the top and bottom N gene sets; \n",
    "#'                'FALSE': all of the significant gsea gene sets will be displayed \n",
    "#'                        --> Note that: it is usually results in too many gene sets, which are not easy to see anything    \n",
    "#' @param TopN this parameter indicate how many gene sets need to be displayed in the plots\n",
    "########################################################################################\n",
    "\n",
    "# DE_out_path <- DE.out.path\n",
    "# export_path <- \n",
    "#IDmappingfile <- \"\" #\"mouse2human.txt\" #\"Data/zebrafish2human_idmapping.txt\"\n",
    "convert = FALSE\n",
    "gene_name = FALSE \n",
    "GSEA_set = c('c0', 'c2', 'c5', 'c7')\n",
    "minGSSize = 10 #20 #30\n",
    "pAdjustMethod = \"BH\"\n",
    "pvalueCutoff = pthresh\n",
    "export = \"BOTH\"\n",
    "ShowTop = TRUE \n",
    "TopN = 10\n",
    "\n",
    "\n",
    "########################################################################################\n",
    "##  RUN gsea_ex1\n",
    "########################################################################################\n",
    "gsea_ex1(DE_out_path = export.path.prep, export_path = export.path, gene_name = gene_name,\n",
    "         convert = convert, GSEA_set = GSEA_set, minGSSize = minGSSize,\n",
    "         pAdjustMethod = padjmethod,  pvalueCutoff = gsea_pthresh,\n",
    "         export = export, ShowTop = ShowTop,  stat.col=stat_col ,\n",
    "         TopN = TopN, gene_sets_dir=gsea_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e4828454",
   "metadata": {},
   "outputs": [],
   "source": [
    "# require(DOSE)\n",
    "# dotplot(gse, showCategory=10, split=\".sign\") + facet_grid(.~.sign)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c8624952",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/'"
      ],
      "text/latex": [
       "'/data/Mito\\_Trace/output/pipeline/CHIP\\_b1/MTBlacklist\\_A2/data/merged/MT/cellr\\_True/numread\\_200/filters/minC10\\_minR50\\_topN0\\_hetT0.001\\_hetC10\\_hetCount5\\_bq20/mgatk/vireoIn/clones/variants\\_init/knn/kparam\\_3/concat/gff\\_A2\\_black/annotation\\_clones/de\\_btwncond\\_RNA/minPct\\_0\\_logfc0\\_p0.1/GSEA\\_pthresh.0.1\\_pref.FALSE\\_stat.log2FoldChange\\_padj.BH/'"
      ],
      "text/markdown": [
       "'/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/'"
      ],
      "text/plain": [
       "[1] \"/data/Mito_Trace/output/pipeline/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/clones/variants_init/knn/kparam_3/concat/gff_A2_black/annotation_clones/de_btwncond_RNA/minPct_0_logfc0_p0.1/GSEA_pthresh.0.1_pref.FALSE_stat.log2FoldChange_padj.BH/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "export.path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8e3f179e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.3 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/isaac/miniconda3/envs/mttrace/lib/libopenblasp-r0.3.18.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] stringr_1.4.0         ggplot2_3.3.5         dplyr_1.0.7          \n",
       " [4] openxlsx_4.2.4        data.table_1.14.2     gdata_2.18.0         \n",
       " [7] DOSE_3.20.1           org.Hs.eg.db_3.14.0   biomaRt_2.50.0       \n",
       "[10] GSEABase_1.56.0       graph_1.72.0          annotate_1.72.0      \n",
       "[13] XML_3.99-0.8          AnnotationDbi_1.56.1  IRanges_2.28.0       \n",
       "[16] S4Vectors_0.32.0      Biobase_2.54.0        BiocGenerics_0.40.0  \n",
       "[19] clusterProfiler_4.2.2\n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] fgsea_1.20.0           colorspace_2.0-2       ggtree_3.2.1          \n",
       "  [4] ellipsis_0.3.2         IRdisplay_1.0          qvalue_2.26.0         \n",
       "  [7] XVector_0.34.0         base64enc_0.1-3        aplot_0.1.2           \n",
       " [10] farver_2.1.0           graphlayouts_0.8.0     ggrepel_0.9.1         \n",
       " [13] bit64_4.0.5            fansi_0.4.2            scatterpie_0.1.7      \n",
       " [16] xml2_1.3.2             splines_4.1.1          cachem_1.0.6          \n",
       " [19] GOSemSim_2.20.0        polyclip_1.10-0        IRkernel_1.2          \n",
       " [22] jsonlite_1.7.2         dbplyr_2.1.1           GO.db_3.14.0          \n",
       " [25] png_0.1-7              ggforce_0.3.3          BiocManager_1.30.16   \n",
       " [28] compiler_4.1.1         httr_1.4.2             assertthat_0.2.1      \n",
       " [31] Matrix_1.3-4           fastmap_1.1.0          lazyeval_0.2.2        \n",
       " [34] tweenr_1.0.2           prettyunits_1.1.1      htmltools_0.5.2       \n",
       " [37] tools_4.1.1            igraph_1.2.9           gtable_0.3.0          \n",
       " [40] glue_1.5.1             GenomeInfoDbData_1.2.7 reshape2_1.4.4        \n",
       " [43] DO.db_2.9              rappdirs_0.3.3         fastmatch_1.1-3       \n",
       " [46] Rcpp_1.0.7             enrichplot_1.14.1      vctrs_0.3.8           \n",
       " [49] Biostrings_2.62.0      ape_5.6-1              nlme_3.1-153          \n",
       " [52] ggraph_2.0.5           lifecycle_1.0.1        gtools_3.9.2          \n",
       " [55] zlibbioc_1.40.0        MASS_7.3-54            scales_1.1.1          \n",
       " [58] tidygraph_1.2.0        hms_1.1.1              parallel_4.1.1        \n",
       " [61] RColorBrewer_1.1-2     curl_4.3.2             memoise_2.0.1         \n",
       " [64] gridExtra_2.3          downloader_0.4         ggfun_0.0.5           \n",
       " [67] yulab.utils_0.0.4      stringi_1.7.6          RSQLite_2.2.8         \n",
       " [70] tidytree_0.3.7         filelock_1.0.2         zip_2.2.0             \n",
       " [73] BiocParallel_1.28.0    repr_1.1.3             GenomeInfoDb_1.30.0   \n",
       " [76] rlang_0.4.12           pkgconfig_2.0.3        bitops_1.0-7          \n",
       " [79] evaluate_0.14          lattice_0.20-45        purrr_0.3.4           \n",
       " [82] labeling_0.4.2         treeio_1.18.1          patchwork_1.1.1       \n",
       " [85] shadowtext_0.1.1       bit_4.0.4              tidyselect_1.1.1      \n",
       " [88] plyr_1.8.6             magrittr_2.0.1         R6_2.5.1              \n",
       " [91] generics_0.1.1         pbdZMQ_0.3-6           DBI_1.1.1             \n",
       " [94] withr_2.4.2            pillar_1.6.4           KEGGREST_1.34.0       \n",
       " [97] RCurl_1.98-1.5         tibble_3.1.6           crayon_1.4.2          \n",
       "[100] uuid_1.0-3             utf8_1.2.2             BiocFileCache_2.2.0   \n",
       "[103] progress_1.2.2         viridis_0.6.2          grid_4.1.1            \n",
       "[106] blob_1.2.2             digest_0.6.28          xtable_1.8-4          \n",
       "[109] tidyr_1.1.4            gridGraphics_0.5-1     munsell_0.5.0         \n",
       "[112] viridisLite_0.4.0      ggplotify_0.1.0       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686d6487",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d0c01837",
   "metadata": {},
   "source": [
    "### "
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
