{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "151a4195",
   "metadata": {},
   "source": [
    "# Introduction: This analyses calls variants on the single-cell ATAC data at known donor CHIP mutation regions.\n",
    " \n",
    "### Steps:\n",
    "1. Create bed file for the known CHIP variants from excel file. \n",
    "    keep as hg19 here\n",
    "    Add +/- 1kb\n",
    "2. Run cellsnp for pileup near those regions\n",
    "3. See if any fragments locations overlap the bed file (they should be if there is pileup)\n",
    "4. Plot cells-by-CHIP mutants AF . Plot bulk for each mutant AF.\n",
    "5. Using the de-multiplexed cells, how many of the known called variants are found in one donor. Also do the same as "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b0dacd0",
   "metadata": {},
   "source": [
    "## Convert txt to bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "efa8d58b",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "config_f = \"config.yaml\"\n",
    "outdir=\"\"\n",
    "minCOUNT = 0.1\n",
    "minAF = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a554637b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Project Directory: /data2/mito_lineage\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "from src.utils.parse_config import read_config_file\n",
    "from src.utils.data_io import wrap_load_mtx_df\n",
    "\n",
    "from src.config import ROOT_DIR\n",
    "from os.path import join, dirname, exists\n",
    "import os\n",
    "import pandas as pd\n",
    "from glob import glob\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import mplh.cluster_help as ch\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "757c5573",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'global': 'PROJECT',\n",
       " 'variants_f': 'data/processed/gilead_chip_profile/20210222_Croker_CD34_sample_ship_MW.txt',\n",
       " 'indir': '/data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist',\n",
       " 'experiment': 'CHIP_april08_2021',\n",
       " 'genome': 'MTblacklist',\n",
       " 'ref_fa': '/data2/mito_lineage/data/external/GRCh38_MT_blacklist/fasta/genome.fa',\n",
       " 'samples': ['Control', 'Flt3l', 'Input']}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config = read_config_file(config_f)\n",
    "samples_dict = {ind+1: val for ind, val in enumerate(config[\"samples\"])}\n",
    "config\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "386e392a",
   "metadata": {},
   "source": [
    "### Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "87d05333",
   "metadata": {},
   "outputs": [],
   "source": [
    "if outdir == \"\":\n",
    "    outdir=\"./output\"\n",
    "\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)\n",
    "    \n",
    "out_bed = join(outdir, \"variants.hg19.1kb.bed\")\n",
    "out_hg19_bed = join(outdir, \"variants.hg19.bed\")\n",
    "out_pileup_d = join(outdir,\"variants_pileup_hg19\")\n",
    "\n",
    "REGION_VCF = join(outdir, \"variants.hg19.vcf\")\n",
    "REGION_VCF_1base = join(outdir, \"variants.hg19.1base.vcf\")\n",
    "\n",
    "# for exp in samples_dict.values():\n",
    "#     if not exists(join(out_pileup_d, exp)):\n",
    "#         os.mkdir(join(out_pileup_d, exp))\n",
    "if not exists(join(out_pileup_d)):\n",
    "    os.mkdir(join(out_pileup_d))\n",
    "\n",
    "# Aggregate\n",
    "curr_in = join(config[\"indir\"], \"aggregate\", \"outs\")\n",
    "peaks_bed_f = join(curr_in, \"peaks.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "490ff588",
   "metadata": {},
   "source": [
    "# 2. Create bed file for CHIP variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d72b2bc6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123196932</td>\n",
       "      <td>123196933</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>intronic</td>\n",
       "      <td>STAG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6065</td>\n",
       "      <td>34188</td>\n",
       "      <td>0.177401</td>\n",
       "      <td>G13</td>\n",
       "      <td>10386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>12022888</td>\n",
       "      <td>12022889</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>ETV6</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>ETV6:NM_001987:exon5:c.T995C:p.I332T</td>\n",
       "      <td>6525</td>\n",
       "      <td>19085</td>\n",
       "      <td>0.341892</td>\n",
       "      <td>G14</td>\n",
       "      <td>14263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25505475</td>\n",
       "      <td>25505476</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>DNMT3A:NM_001320892:exon4:c.G282A:p.E94E,DNMT3...</td>\n",
       "      <td>1760</td>\n",
       "      <td>15334</td>\n",
       "      <td>0.114778</td>\n",
       "      <td>G14</td>\n",
       "      <td>14263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>148544302</td>\n",
       "      <td>148544303</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>EZH2</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>EZH2:NM_001203247:exon2:c.A88C:p.K30Q,EZH2:NM_...</td>\n",
       "      <td>48</td>\n",
       "      <td>24238</td>\n",
       "      <td>0.001980</td>\n",
       "      <td>S03</td>\n",
       "      <td>RB1070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4</td>\n",
       "      <td>106196818</td>\n",
       "      <td>106196819</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>TET2</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>TET2:NM_001127208:exon11:c.G5152T:p.V1718L</td>\n",
       "      <td>14327</td>\n",
       "      <td>37742</td>\n",
       "      <td>0.379604</td>\n",
       "      <td>S03</td>\n",
       "      <td>RB1070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>76890276</td>\n",
       "      <td>76890277</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>intronic</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>44</td>\n",
       "      <td>14828</td>\n",
       "      <td>0.002967</td>\n",
       "      <td>S03</td>\n",
       "      <td>RB1070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrX</td>\n",
       "      <td>44950066</td>\n",
       "      <td>44950067</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>KDM6A</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>KDM6A:NM_001291418:exon24:c.G3599A:p.R1200Q,KD...</td>\n",
       "      <td>56</td>\n",
       "      <td>18167</td>\n",
       "      <td>0.003083</td>\n",
       "      <td>S03</td>\n",
       "      <td>RB1070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrX</td>\n",
       "      <td>44949080</td>\n",
       "      <td>44949081</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>KDM6A</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>KDM6A:NM_001291418:exon23:c.T3405G:p.P1135P,KD...</td>\n",
       "      <td>78</td>\n",
       "      <td>19177</td>\n",
       "      <td>0.004067</td>\n",
       "      <td>S03</td>\n",
       "      <td>RB1070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25457184</td>\n",
       "      <td>25457185</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>DNMT3A:NM_001320893:exon18:c.T2246G:p.L749R,DN...</td>\n",
       "      <td>290</td>\n",
       "      <td>19494</td>\n",
       "      <td>0.014876</td>\n",
       "      <td>S03</td>\n",
       "      <td>RB1070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133738339</td>\n",
       "      <td>133738340</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>ABL1:NM_005157:exon4:c.A740G:p.K247R,ABL1:NM_0...</td>\n",
       "      <td>23328</td>\n",
       "      <td>71887</td>\n",
       "      <td>0.324509</td>\n",
       "      <td>S10</td>\n",
       "      <td>RB1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123200259</td>\n",
       "      <td>123200260</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STAG2</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>STAG2:NM_006603:exon22:c.A2239T:p.K747X,STAG2:...</td>\n",
       "      <td>54</td>\n",
       "      <td>35697</td>\n",
       "      <td>0.001513</td>\n",
       "      <td>S10</td>\n",
       "      <td>RB1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr7</td>\n",
       "      <td>101740693</td>\n",
       "      <td>101740694</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>CUX1</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>CUX1:NM_001202544:exon4:c.C304G:p.R102G,CUX1:N...</td>\n",
       "      <td>5602</td>\n",
       "      <td>14098</td>\n",
       "      <td>0.397361</td>\n",
       "      <td>S10</td>\n",
       "      <td>RB1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr4</td>\n",
       "      <td>55599428</td>\n",
       "      <td>55599429</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>intronic</td>\n",
       "      <td>KIT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9319</td>\n",
       "      <td>18870</td>\n",
       "      <td>0.493853</td>\n",
       "      <td>S10</td>\n",
       "      <td>RB1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrX</td>\n",
       "      <td>44949080</td>\n",
       "      <td>44949081</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>KDM6A</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>KDM6A:NM_001291418:exon23:c.T3405G:p.P1135P,KD...</td>\n",
       "      <td>74</td>\n",
       "      <td>13750</td>\n",
       "      <td>0.005382</td>\n",
       "      <td>S10</td>\n",
       "      <td>RB1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25463540</td>\n",
       "      <td>25463541</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>DNMT3A:NM_001320893:exon13:c.C1685G:p.S562C,DN...</td>\n",
       "      <td>57</td>\n",
       "      <td>46443</td>\n",
       "      <td>0.001227</td>\n",
       "      <td>S10</td>\n",
       "      <td>RB1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25462021</td>\n",
       "      <td>25462022</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>DNMT3A:NM_001320893:exon15:c.G1929A:p.W643X,DN...</td>\n",
       "      <td>46</td>\n",
       "      <td>23142</td>\n",
       "      <td>0.001988</td>\n",
       "      <td>S10</td>\n",
       "      <td>RB1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7578367</td>\n",
       "      <td>7578368</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>intronic</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9161</td>\n",
       "      <td>40848</td>\n",
       "      <td>0.224270</td>\n",
       "      <td>S10</td>\n",
       "      <td>RB1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133747696</td>\n",
       "      <td>133747697</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>intronic</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>103</td>\n",
       "      <td>13651</td>\n",
       "      <td>0.007545</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chrX</td>\n",
       "      <td>129184781</td>\n",
       "      <td>129184782</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>intronic</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>186</td>\n",
       "      <td>5949</td>\n",
       "      <td>0.031266</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr4</td>\n",
       "      <td>106197283</td>\n",
       "      <td>106197284</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>TET2</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>TET2:NM_001127208:exon11:c.A5617C:p.I1873L</td>\n",
       "      <td>474</td>\n",
       "      <td>9417</td>\n",
       "      <td>0.050335</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr4</td>\n",
       "      <td>106196833</td>\n",
       "      <td>106196834</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>TET2</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>TET2:NM_001127208:exon11:c.C5167T:p.P1723S</td>\n",
       "      <td>7233</td>\n",
       "      <td>14516</td>\n",
       "      <td>0.498278</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr4</td>\n",
       "      <td>55561841</td>\n",
       "      <td>55561842</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>KIT</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>KIT:NM_000222:exon2:c.A232T:p.K78X,KIT:NM_0010...</td>\n",
       "      <td>44</td>\n",
       "      <td>27229</td>\n",
       "      <td>0.001616</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr4</td>\n",
       "      <td>55561839</td>\n",
       "      <td>55561840</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>KIT</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>KIT:NM_000222:exon2:c.A230T:p.N77I,KIT:NM_0010...</td>\n",
       "      <td>84</td>\n",
       "      <td>27243</td>\n",
       "      <td>0.003083</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr21</td>\n",
       "      <td>36206949</td>\n",
       "      <td>36206950</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>intronic</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99</td>\n",
       "      <td>1823</td>\n",
       "      <td>0.054306</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25467427</td>\n",
       "      <td>25467428</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>DNMT3A:NM_001320893:exon9:c.G1192A:p.G398R,DNM...</td>\n",
       "      <td>124</td>\n",
       "      <td>9590</td>\n",
       "      <td>0.012930</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25463522</td>\n",
       "      <td>25463523</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>DNMT3A:NM_001320893:exon13:c.G1703T:p.R568L,DN...</td>\n",
       "      <td>74</td>\n",
       "      <td>17322</td>\n",
       "      <td>0.004272</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7578367</td>\n",
       "      <td>7578368</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>intronic</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5313</td>\n",
       "      <td>22653</td>\n",
       "      <td>0.234538</td>\n",
       "      <td>S11</td>\n",
       "      <td>RB1096</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0          1          2  3  4         5       6                  7   \\\n",
       "0    chrX  123196932  123196933  T  C  intronic   STAG2                NaN   \n",
       "1   chr12   12022888   12022889  T  C    exonic    ETV6  nonsynonymous SNV   \n",
       "2    chr2   25505475   25505476  C  T    exonic  DNMT3A     synonymous SNV   \n",
       "3    chr7  148544302  148544303  T  G    exonic    EZH2  nonsynonymous SNV   \n",
       "4    chr4  106196818  106196819  G  T    exonic    TET2  nonsynonymous SNV   \n",
       "5    chrX   76890276   76890277  A  C  intronic    ATRX                NaN   \n",
       "6    chrX   44950066   44950067  G  A    exonic   KDM6A  nonsynonymous SNV   \n",
       "7    chrX   44949080   44949081  T  G    exonic   KDM6A     synonymous SNV   \n",
       "8    chr2   25457184   25457185  A  C    exonic  DNMT3A  nonsynonymous SNV   \n",
       "9    chr9  133738339  133738340  A  G    exonic    ABL1  nonsynonymous SNV   \n",
       "10   chrX  123200259  123200260  A  T    exonic   STAG2           stopgain   \n",
       "11   chr7  101740693  101740694  C  G    exonic    CUX1  nonsynonymous SNV   \n",
       "12   chr4   55599428   55599429  C  T  intronic     KIT                NaN   \n",
       "13   chrX   44949080   44949081  T  G    exonic   KDM6A     synonymous SNV   \n",
       "14   chr2   25463540   25463541  G  C    exonic  DNMT3A  nonsynonymous SNV   \n",
       "15   chr2   25462021   25462022  C  T    exonic  DNMT3A           stopgain   \n",
       "16  chr17    7578367    7578368  C  A  intronic    TP53                NaN   \n",
       "17   chr9  133747696  133747697  A  G  intronic    ABL1                NaN   \n",
       "18   chrX  129184781  129184782  T  G  intronic  BCORL1                NaN   \n",
       "19   chr4  106197283  106197284  A  C    exonic    TET2  nonsynonymous SNV   \n",
       "20   chr4  106196833  106196834  C  T    exonic    TET2  nonsynonymous SNV   \n",
       "21   chr4   55561841   55561842  A  T    exonic     KIT           stopgain   \n",
       "22   chr4   55561839   55561840  A  T    exonic     KIT  nonsynonymous SNV   \n",
       "23  chr21   36206949   36206950  G  C  intronic   RUNX1                NaN   \n",
       "24   chr2   25467427   25467428  C  T    exonic  DNMT3A  nonsynonymous SNV   \n",
       "25   chr2   25463522   25463523  C  A    exonic  DNMT3A  nonsynonymous SNV   \n",
       "26  chr17    7578367    7578368  C  A  intronic    TP53                NaN   \n",
       "\n",
       "                                                   8      9      10        11  \\\n",
       "0                                                 NaN   6065  34188  0.177401   \n",
       "1                ETV6:NM_001987:exon5:c.T995C:p.I332T   6525  19085  0.341892   \n",
       "2   DNMT3A:NM_001320892:exon4:c.G282A:p.E94E,DNMT3...   1760  15334  0.114778   \n",
       "3   EZH2:NM_001203247:exon2:c.A88C:p.K30Q,EZH2:NM_...     48  24238  0.001980   \n",
       "4          TET2:NM_001127208:exon11:c.G5152T:p.V1718L  14327  37742  0.379604   \n",
       "5                                                 NaN     44  14828  0.002967   \n",
       "6   KDM6A:NM_001291418:exon24:c.G3599A:p.R1200Q,KD...     56  18167  0.003083   \n",
       "7   KDM6A:NM_001291418:exon23:c.T3405G:p.P1135P,KD...     78  19177  0.004067   \n",
       "8   DNMT3A:NM_001320893:exon18:c.T2246G:p.L749R,DN...    290  19494  0.014876   \n",
       "9   ABL1:NM_005157:exon4:c.A740G:p.K247R,ABL1:NM_0...  23328  71887  0.324509   \n",
       "10  STAG2:NM_006603:exon22:c.A2239T:p.K747X,STAG2:...     54  35697  0.001513   \n",
       "11  CUX1:NM_001202544:exon4:c.C304G:p.R102G,CUX1:N...   5602  14098  0.397361   \n",
       "12                                                NaN   9319  18870  0.493853   \n",
       "13  KDM6A:NM_001291418:exon23:c.T3405G:p.P1135P,KD...     74  13750  0.005382   \n",
       "14  DNMT3A:NM_001320893:exon13:c.C1685G:p.S562C,DN...     57  46443  0.001227   \n",
       "15  DNMT3A:NM_001320893:exon15:c.G1929A:p.W643X,DN...     46  23142  0.001988   \n",
       "16                                                NaN   9161  40848  0.224270   \n",
       "17                                                NaN    103  13651  0.007545   \n",
       "18                                                NaN    186   5949  0.031266   \n",
       "19         TET2:NM_001127208:exon11:c.A5617C:p.I1873L    474   9417  0.050335   \n",
       "20         TET2:NM_001127208:exon11:c.C5167T:p.P1723S   7233  14516  0.498278   \n",
       "21  KIT:NM_000222:exon2:c.A232T:p.K78X,KIT:NM_0010...     44  27229  0.001616   \n",
       "22  KIT:NM_000222:exon2:c.A230T:p.N77I,KIT:NM_0010...     84  27243  0.003083   \n",
       "23                                                NaN     99   1823  0.054306   \n",
       "24  DNMT3A:NM_001320893:exon9:c.G1192A:p.G398R,DNM...    124   9590  0.012930   \n",
       "25  DNMT3A:NM_001320893:exon13:c.G1703T:p.R568L,DN...     74  17322  0.004272   \n",
       "26                                                NaN   5313  22653  0.234538   \n",
       "\n",
       "     12      13  \n",
       "0   G13   10386  \n",
       "1   G14   14263  \n",
       "2   G14   14263  \n",
       "3   S03  RB1070  \n",
       "4   S03  RB1070  \n",
       "5   S03  RB1070  \n",
       "6   S03  RB1070  \n",
       "7   S03  RB1070  \n",
       "8   S03  RB1070  \n",
       "9   S10  RB1003  \n",
       "10  S10  RB1003  \n",
       "11  S10  RB1003  \n",
       "12  S10  RB1003  \n",
       "13  S10  RB1003  \n",
       "14  S10  RB1003  \n",
       "15  S10  RB1003  \n",
       "16  S10  RB1003  \n",
       "17  S11  RB1096  \n",
       "18  S11  RB1096  \n",
       "19  S11  RB1096  \n",
       "20  S11  RB1096  \n",
       "21  S11  RB1096  \n",
       "22  S11  RB1096  \n",
       "23  S11  RB1096  \n",
       "24  S11  RB1096  \n",
       "25  S11  RB1096  \n",
       "26  S11  RB1096  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_df = pd.read_csv(join(ROOT_DIR, config[\"variants_f\"]), sep='\\t', header=None)\n",
    "bed_df[1]-=1\n",
    "bed_df[[0,1,2]].to_csv(out_hg19_bed, sep='\\t', header=None, index=False)\n",
    "meta_df = bed_df.drop([0,1,2], axis=1).copy().rename({3:\"REF\", 4:\"ALT\"}, axis=1)\n",
    "\n",
    "bed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3f13128e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123196932</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>12022888</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25505475</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>148544302</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4</td>\n",
       "      <td>106196818</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>76890276</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrX</td>\n",
       "      <td>44950066</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrX</td>\n",
       "      <td>44949080</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25457184</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133738339</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123200259</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr7</td>\n",
       "      <td>101740693</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr4</td>\n",
       "      <td>55599428</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrX</td>\n",
       "      <td>44949080</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25463540</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25462021</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7578367</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133747696</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chrX</td>\n",
       "      <td>129184781</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr4</td>\n",
       "      <td>106197283</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr4</td>\n",
       "      <td>106196833</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr4</td>\n",
       "      <td>55561841</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr4</td>\n",
       "      <td>55561839</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr21</td>\n",
       "      <td>36206949</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25467427</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr2</td>\n",
       "      <td>25463522</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7578367</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #CHROM        POS ID REF ALT QUAL FILTER INFO\n",
       "0    chrX  123196932  .   T   C    .   PASS     \n",
       "1   chr12   12022888  .   T   C    .   PASS     \n",
       "2    chr2   25505475  .   C   T    .   PASS     \n",
       "3    chr7  148544302  .   T   G    .   PASS     \n",
       "4    chr4  106196818  .   G   T    .   PASS     \n",
       "5    chrX   76890276  .   A   C    .   PASS     \n",
       "6    chrX   44950066  .   G   A    .   PASS     \n",
       "7    chrX   44949080  .   T   G    .   PASS     \n",
       "8    chr2   25457184  .   A   C    .   PASS     \n",
       "9    chr9  133738339  .   A   G    .   PASS     \n",
       "10   chrX  123200259  .   A   T    .   PASS     \n",
       "11   chr7  101740693  .   C   G    .   PASS     \n",
       "12   chr4   55599428  .   C   T    .   PASS     \n",
       "13   chrX   44949080  .   T   G    .   PASS     \n",
       "14   chr2   25463540  .   G   C    .   PASS     \n",
       "15   chr2   25462021  .   C   T    .   PASS     \n",
       "16  chr17    7578367  .   C   A    .   PASS     \n",
       "17   chr9  133747696  .   A   G    .   PASS     \n",
       "18   chrX  129184781  .   T   G    .   PASS     \n",
       "19   chr4  106197283  .   A   C    .   PASS     \n",
       "20   chr4  106196833  .   C   T    .   PASS     \n",
       "21   chr4   55561841  .   A   T    .   PASS     \n",
       "22   chr4   55561839  .   A   T    .   PASS     \n",
       "23  chr21   36206949  .   G   C    .   PASS     \n",
       "24   chr2   25467427  .   C   T    .   PASS     \n",
       "25   chr2   25463522  .   C   A    .   PASS     \n",
       "26  chr17    7578367  .   C   A    .   PASS     "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf = pd.concat((bed_df, meta_df), axis=1).rename({0:\"#CHROM\", 1:\"POS\"}, axis=1)\n",
    "vcf[\"ID\"] = \".\"\n",
    "vcf[\"QUAL\"] = \".\"\n",
    "vcf[\"FILTER\"] = \"PASS\"\n",
    "vcf[\"INFO\"] = \"\"\n",
    "#vcf[[\"#CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\", \"INFO\"]].to_csv(REGION_VCF, sep='\\t', index=False)\n",
    "\n",
    "\n",
    "vcf[[\"#CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\"]].to_csv(REGION_VCF, sep='\\t', index=False)\n",
    "\n",
    "\n",
    "vcf[[\"#CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\", \"INFO\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b043666",
   "metadata": {},
   "source": [
    "### Convert bed to (dummy)-vcf  for cellsnp-lite input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa2eba8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open a file in write mode to add the comment\n",
    "# Then close the file and reopen it with pandas in append mode\n",
    "with open(REGION_VCF, 'w') as f:\n",
    "    f.write(f'##fileformat=VCFv4.2\\n##reference=file://{config[\"ref_fa\"]}\\n')\n",
    "vcf[[\"#CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\"]].to_csv(REGION_VCF, sep='\\t', index=False, mode='a')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "117977d1",
   "metadata": {},
   "source": [
    "## +/- 500bps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5df475f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_df[1]-=500\n",
    "bed_df[2]+=500\n",
    "bed_df.to_csv(out_bed, sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07497298",
   "metadata": {},
   "source": [
    "# 2. Run cellsnp\n",
    "(since bam is separate for each experiment, need to run as that)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "715c4e68",
   "metadata": {},
   "source": [
    "### Run with exclFLAG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b69da684",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Control\n",
      "cellsnp-lite -s /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Control/outs/possorted_bam.bam -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Control/outs/filtered_peak_bc_matrix/barcodes.tsv -O ./output/variants_pileup_hg19/Control_1base -R ./output/variants.hg19.vcf --exclFLAG 772  -p 20 --minMAF 0 --minCOUNT 0.1 --UMItag None --cellTAG CB \n",
      "Flt3l\n",
      "cellsnp-lite -s /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Flt3l/outs/possorted_bam.bam -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Flt3l/outs/filtered_peak_bc_matrix/barcodes.tsv -O ./output/variants_pileup_hg19/Flt3l_1base -R ./output/variants.hg19.vcf --exclFLAG 772  -p 20 --minMAF 0 --minCOUNT 0.1 --UMItag None --cellTAG CB \n",
      "Input\n",
      "cellsnp-lite -s /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Input/outs/possorted_bam.bam -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Input/outs/filtered_peak_bc_matrix/barcodes.tsv -O ./output/variants_pileup_hg19/Input_1base -R ./output/variants.hg19.vcf --exclFLAG 772  -p 20 --minMAF 0 --minCOUNT 0.1 --UMItag None --cellTAG CB \n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "for exp in samples_dict.values():\n",
    "    print(exp)\n",
    "    BAM = join(config[\"indir\"], exp, \"outs/possorted_bam.bam\")\n",
    "    BARCODE = join(config[\"indir\"], exp, \"outs/filtered_peak_bc_matrix/barcodes.tsv\")\n",
    "    OUT_DIR = join(out_pileup_d, exp+\"_1base\")\n",
    "    cmd = f\"cellsnp-lite -s {BAM} -b {BARCODE} -O {OUT_DIR} -R {REGION_VCF} --exclFLAG 772  -p 20 --minMAF {minAF} --minCOUNT {minCOUNT} --UMItag None --cellTAG CB \"\n",
    "    print(cmd)\n",
    "    os.system(cmd)\n",
    "    AD, DP = wrap_load_mtx_df(OUT_DIR, oth_f=False, prefix=\"cellSNP.tag\",\n",
    "                         columns=('Variant', 'Cell', 'integer'), inc_af=False,\n",
    "                         as_dense=True, var_names=False, vcf_prefix=\"cellSNP.base\")\n",
    "\n",
    "    AF=AD/DP.fillna(0)\n",
    "    AF\n",
    "    print((AD>0).sum().sum())\n",
    "    print((AF>0).sum(axis=1))\n",
    "    f, ax = plt.subplots(ncols=2, figsize=(10,8))\n",
    "    sns.heatmap(AF.fillna(0), ax=ax[0], vmin=0)\n",
    "    ax[0].set_title(\"Variant Allele Frequency\")\n",
    "    sns.heatmap(DP, ax=ax[1], vmin=0)\n",
    "    ax[1].set_title(\"Read Coverage at variant position\")\n",
    "    f.suptitle(exp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac030aee",
   "metadata": {},
   "source": [
    "## Run 1base just in case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ce8abde",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Control\n",
      "cellsnp-lite -s /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Control/outs/possorted_bam.bam -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Control/outs/filtered_peak_bc_matrix/barcodes.tsv -O ./output/variants_pileup_hg19/Control -R ./output/variants.hg19.1base.vcf --exclFLAG 772  -p 20 --minMAF 0 --minCOUNT 0.1 --UMItag None --cellTAG CB \n",
      "Flt3l\n",
      "cellsnp-lite -s /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Flt3l/outs/possorted_bam.bam -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Flt3l/outs/filtered_peak_bc_matrix/barcodes.tsv -O ./output/variants_pileup_hg19/Flt3l -R ./output/variants.hg19.1base.vcf --exclFLAG 772  -p 20 --minMAF 0 --minCOUNT 0.1 --UMItag None --cellTAG CB \n",
      "Input\n",
      "cellsnp-lite -s /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Input/outs/possorted_bam.bam -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Input/outs/filtered_peak_bc_matrix/barcodes.tsv -O ./output/variants_pileup_hg19/Input -R ./output/variants.hg19.1base.vcf --exclFLAG 772  -p 20 --minMAF 0 --minCOUNT 0.1 --UMItag None --cellTAG CB \n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "def copy_df(df):\n",
    "    return df.copy()\n",
    "def pos_add(df):\n",
    "    df['POS']+= 1\n",
    "    return df\n",
    "def vcf_save(df, fname, config):\n",
    "    # Then close the file and reopen it with pandas in append mode\n",
    "    with open(fname, 'w') as f:\n",
    "        f.write(f'##fileformat=VCFv4.2\\n##reference=file://{config[\"ref_fa\"]}\\n')\n",
    "    df[[\"#CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\"]].to_csv(fname, sep='\\t', index=False, mode='a')\n",
    "    return\n",
    "\n",
    "vcf_save(vcf.pipe(copy_df).pipe(pos_add),REGION_VCF_1base, config)\n",
    "\n",
    "for exp in samples_dict.values():\n",
    "    print(exp)\n",
    "    BAM = join(config[\"indir\"], exp, \"outs/possorted_bam.bam\")\n",
    "    BARCODE = join(config[\"indir\"], exp, \"outs/filtered_peak_bc_matrix/barcodes.tsv\")\n",
    "    OUT_DIR = join(out_pileup_d, exp)\n",
    "    cmd = f\"cellsnp-lite -s {BAM} -b {BARCODE} -O {OUT_DIR} -R {REGION_VCF_1base} --exclFLAG 772  -p 20 --minMAF {minAF} --minCOUNT {minCOUNT} --UMItag None --cellTAG CB \"\n",
    "    #shell: \"cellsnp-lite -s {input.bam_f} --exclFLAG {params.exclFLAG}  -b {input.barcode_f} -O {output} -p 20 --minMAF {params.minMAF} --minCOUNT {params.minCOUNT} --gzip --chrom chrM --printSkipSNPs --UMItag None --cellTAG CB 2> {log}\"#chrM\"\n",
    "    print(cmd)\n",
    "    os.system(cmd)\n",
    "    AD, DP = wrap_load_mtx_df(OUT_DIR, oth_f=False, prefix=\"cellSNP.tag\",\n",
    "                         columns=('Variant', 'Cell', 'integer'), inc_af=False,\n",
    "                         as_dense=True, var_names=False, vcf_prefix=\"cellSNP.base\")\n",
    "\n",
    "    AF=AD/DP.fillna(0)\n",
    "    AF\n",
    "    print((AD>0).sum().sum())\n",
    "    print((AF>0).sum(axis=1))\n",
    "    f, ax = plt.subplots(ncols=2, figsize=(10,8))\n",
    "    sns.heatmap(AF.fillna(0), ax=ax[0], vmin=0)\n",
    "    ax[0].set_title(\"Variant Allele Frequency\")\n",
    "    sns.heatmap(DP, ax=ax[1], vmin=0)\n",
    "    ax[1].set_title(\"Read Coverage at variant position\")\n",
    "    f.suptitle(exp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec29ae75",
   "metadata": {},
   "source": [
    "## Run w/out exclflag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7624e1de",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Control\n",
      "cellsnp-lite -s /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Control/outs/possorted_bam.bam -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Control/outs/filtered_peak_bc_matrix/barcodes.tsv -O ./output/variants_pileup_hg19/Control_noEXCL -R ./output/variants.hg19.vcf  -p 20 --minMAF 0 --minCOUNT 0.1 --UMItag None --cellTAG CB \n",
      "Flt3l\n",
      "cellsnp-lite -s /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Flt3l/outs/possorted_bam.bam -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Flt3l/outs/filtered_peak_bc_matrix/barcodes.tsv -O ./output/variants_pileup_hg19/Flt3l_noEXCL -R ./output/variants.hg19.vcf  -p 20 --minMAF 0 --minCOUNT 0.1 --UMItag None --cellTAG CB \n",
      "Input\n",
      "cellsnp-lite -s /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Input/outs/possorted_bam.bam -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/Input/outs/filtered_peak_bc_matrix/barcodes.tsv -O ./output/variants_pileup_hg19/Input_noEXCL -R ./output/variants.hg19.vcf  -p 20 --minMAF 0 --minCOUNT 0.1 --UMItag None --cellTAG CB \n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "for exp in samples_dict.values():\n",
    "    print(exp)\n",
    "    BAM = join(config[\"indir\"], exp, \"outs/possorted_bam.bam\")\n",
    "    BARCODE = join(config[\"indir\"], exp, \"outs/filtered_peak_bc_matrix/barcodes.tsv\")\n",
    "    OUT_DIR = join(out_pileup_d, exp+\"_noEXCL\")\n",
    "    cmd = f\"cellsnp-lite -s {BAM} -b {BARCODE} -O {OUT_DIR} -R {REGION_VCF}  -p 20 --minMAF {minAF} --minCOUNT {minCOUNT} --UMItag None --cellTAG CB \"\n",
    "    print(cmd)\n",
    "    os.system(cmd)\n",
    "    AD, DP = wrap_load_mtx_df(OUT_DIR, oth_f=False, prefix=\"cellSNP.tag\",\n",
    "                         columns=('Variant', 'Cell', 'integer'), inc_af=False,\n",
    "                         as_dense=True, var_names=False, vcf_prefix=\"cellSNP.base\")\n",
    "\n",
    "    AF=AD/DP.fillna(0)\n",
    "    AF\n",
    "    print((AD>0).sum().sum())\n",
    "    print((AF>0).sum(axis=1))\n",
    "    f, ax = plt.subplots(ncols=2, figsize=(10,8))\n",
    "    sns.heatmap(AF.fillna(0), ax=ax[0], vmin=0)\n",
    "    ax[0].set_title(\"Variant Allele Frequency\")\n",
    "    sns.heatmap(DP, ax=ax[1], vmin=0)\n",
    "    ax[1].set_title(\"Read Coverage at variant position\")\n",
    "    f.suptitle(exp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08b87124",
   "metadata": {},
   "source": [
    "# 3. See if aggregate peak location overlaps with variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "185126d9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9920</td>\n",
       "      <td>10570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>96482</td>\n",
       "      <td>96714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>180685</td>\n",
       "      <td>181125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>191595</td>\n",
       "      <td>191964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>267775</td>\n",
       "      <td>268263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145453</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56763251</td>\n",
       "      <td>56763800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145454</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56825902</td>\n",
       "      <td>56825911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145455</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56826567</td>\n",
       "      <td>56826798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145456</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56828629</td>\n",
       "      <td>56851740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145457</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56870879</td>\n",
       "      <td>56871130</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>145458 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2\n",
       "0       chr1      9920     10570\n",
       "1       chr1     96482     96714\n",
       "2       chr1    180685    181125\n",
       "3       chr1    191595    191964\n",
       "4       chr1    267775    268263\n",
       "...      ...       ...       ...\n",
       "145453  chrY  56763251  56763800\n",
       "145454  chrY  56825902  56825911\n",
       "145455  chrY  56826567  56826798\n",
       "145456  chrY  56828629  56851740\n",
       "145457  chrY  56870879  56871130\n",
       "\n",
       "[145458 rows x 3 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## peaks location\n",
    "peaks_df = pd.read_csv(peaks_bed_f, sep='\\t', header=None)\n",
    "peaks_df\n",
    "\n",
    "# CBs = pd.read_csv(join(curr_in, \"singlecell.csv\"))\n",
    "# CBs.index= CBs.index+1\n",
    "# CBs[\"Condition ID\"] = CBs[\"barcode\"].apply(lambda x: x.split(\"-\")[-1])\n",
    "# CBs[\"Condition\"] = CBs[\"Condition ID\"].astype(int).map(samples_dict)\n",
    "# CBs.head()\n",
    "\n",
    "# ## Load peak annotations\n",
    "# peak_annotations = pd.read_csv(join(curr_in,'peak_annotation.tsv'),sep='\\t')\n",
    "# peak_annotations.index=peak_annotations.index+1\n",
    "# peak_annotations[\"gene\"] = peak_annotations[\"gene\"].str.upper()\n",
    "# peak_annotations[\"Peak\"]= peak_annotations.index\n",
    "\n",
    "# ## Load peaks matrix\n",
    "# filename = join(curr_in,'filtered_peak_bc_matrix/matrix.mtx')\n",
    "# peaks_sparse_mtx = pd.read_csv(filename,sep=' ',skiprows=2, header=None).iloc[1:].reset_index(drop=True)\n",
    "# peaks_sparse_mtx.columns = [\"Peak\", \"Cell\", \"Count\"]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "77e36aef",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bedtools intersect -a ./output/variants.hg19.1kb.bed -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/aggregate/outs/peaks.bed \n",
      "chr12\t12022533\t12022883\tT\tC\texonic\tETV6\tnonsynonymous SNV\tETV6:NM_001987:exon5:c.T995C:p.I332T\t6525\t19085\t0.341891538\tG14\t14263\n",
      "chr9\t133737970\t133738676\tA\tG\texonic\tABL1\tnonsynonymous SNV\tABL1:NM_005157:exon4:c.A740G:p.K247R,ABL1:NM_007313:exon4:c.A797G:p.K266R\t23328\t71887\t0.324509299\tS10\tRB1003\n",
      "chr7\t101740809\t101741194\tC\tG\texonic\tCUX1\tnonsynonymous SNV\tCUX1:NM_001202544:exon4:c.C304G:p.R102G,CUX1:NM_001202546:exon4:c.C241G:p.R81G,CUX1:NM_001202543:exon5:c.C352G:p.R118G,CUX1:NM_001913:exon5:c.C352G:p.R118G,CUX1:NM_181500:exon5:c.C352G:p.R118G,CUX1:NM_181552:exon5:c.C319G:p.R107G\t5602\t14098\t0.39736132799999996\tS10\tRB1003\n",
      "chr17\t7577867\t7578868\tC\tA\tintronic\tTP53\t\t\t9161\t40848\t0.224270466\tS10\tRB1003\n",
      "chr4\t55561884\t55562334\tA\tT\texonic\tKIT\tstopgain\tKIT:NM_000222:exon2:c.A232T:p.K78X,KIT:NM_001093772:exon2:c.A232T:p.K78X\t44\t27229\t0.001615924\tS11\tRB1096\n",
      "chr4\t55561884\t55562334\tA\tT\texonic\tKIT\tnonsynonymous SNV\tKIT:NM_000222:exon2:c.A230T:p.N77I,KIT:NM_001093772:exon2:c.A230T:p.N77I\t84\t27243\t0.0030833609999999997\tS11\tRB1096\n",
      "chr17\t7577867\t7578868\tC\tA\tintronic\tTP53\t\t\t5313\t22653\t0.234538472\tS11\tRB1096\n"
     ]
    }
   ],
   "source": [
    "cmd = f\"bedtools intersect -a {out_bed} -b {peaks_bed_f} \"\n",
    "print(cmd)\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4a34596e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bedtools intersect -wo -a ./output/variants.hg19.bed -b /data2/isshamie/mito_lineage/data/processed/mtscATAC/CHIP_april08_2021_Croker/MTblacklist/aggregate/outs/peaks.bed \n",
      "chr9\t133738339\t133738340\tchr9\t133737970\t133738676\t1\n",
      "chr17\t7578367\t7578368\tchr17\t7577701\t7582561\t1\n",
      "chr17\t7578367\t7578368\tchr17\t7577701\t7582561\t1\n"
     ]
    }
   ],
   "source": [
    "cmd = f\"bedtools intersect -wo -a {out_hg19_bed} -b {peaks_bed_f} \"\n",
    "print(cmd)\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "127b5f05",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 ./output/variants_pileup/Control/cellSNP.tag.AD.mtx\n",
      "6 ./output/variants_pileup/Flt3l/cellSNP.tag.AD.mtx\n",
      "4 ./output/variants_pileup/Input/cellSNP.tag.AD.mtx\n",
      "11 ./output/variants_pileup/Flt3l_1base/cellSNP.tag.AD.mtx\n",
      "6 ./output/variants_pileup/Control_noEXCL/cellSNP.tag.AD.mtx\n",
      "4 ./output/variants_pileup/Input_1base/cellSNP.tag.AD.mtx\n",
      "4 ./output/variants_pileup/Input_noEXCL/cellSNP.tag.AD.mtx\n",
      "7 ./output/variants_pileup/Control_1base/cellSNP.tag.AD.mtx\n",
      "11 ./output/variants_pileup/Flt3l_noEXCL/cellSNP.tag.AD.mtx\n",
      "5 ./output/variants_pileup_hg19/Control/cellSNP.tag.AD.mtx\n",
      "6 ./output/variants_pileup_hg19/Flt3l/cellSNP.tag.AD.mtx\n",
      "3 ./output/variants_pileup_hg19/Input/cellSNP.tag.AD.mtx\n",
      "3 ./output/variants_pileup_hg19/Flt3l_1base/cellSNP.tag.AD.mtx\n",
      "3 ./output/variants_pileup_hg19/Control_noEXCL/cellSNP.tag.AD.mtx\n",
      "3 ./output/variants_pileup_hg19/Input_1base/cellSNP.tag.AD.mtx\n",
      "3 ./output/variants_pileup_hg19/Input_noEXCL/cellSNP.tag.AD.mtx\n",
      "3 ./output/variants_pileup_hg19/Control_1base/cellSNP.tag.AD.mtx\n",
      "3 ./output/variants_pileup_hg19/Flt3l_noEXCL/cellSNP.tag.AD.mtx\n"
     ]
    }
   ],
   "source": [
    "cmd = f'find {outdir} -name \"*AD.mtx\" -exec wc -l {{}} \\;'\n",
    "!{cmd}\n",
    "#!find output/ -name \"*AD.mtx\" -exec wc -l {} \\;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fcdfc7dc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 ./output/variants_pileup/Control/cellSNP.tag.DP.mtx\n",
      "6 ./output/variants_pileup/Flt3l/cellSNP.tag.DP.mtx\n",
      "4 ./output/variants_pileup/Input/cellSNP.tag.DP.mtx\n",
      "15 ./output/variants_pileup/Flt3l_1base/cellSNP.tag.DP.mtx\n",
      "9 ./output/variants_pileup/Control_noEXCL/cellSNP.tag.DP.mtx\n",
      "8 ./output/variants_pileup/Input_1base/cellSNP.tag.DP.mtx\n",
      "7 ./output/variants_pileup/Input_noEXCL/cellSNP.tag.DP.mtx\n",
      "9 ./output/variants_pileup/Control_1base/cellSNP.tag.DP.mtx\n",
      "15 ./output/variants_pileup/Flt3l_noEXCL/cellSNP.tag.DP.mtx\n",
      "5 ./output/variants_pileup_hg19/Control/cellSNP.tag.DP.mtx\n",
      "7 ./output/variants_pileup_hg19/Flt3l/cellSNP.tag.DP.mtx\n",
      "5 ./output/variants_pileup_hg19/Input/cellSNP.tag.DP.mtx\n",
      "24 ./output/variants_pileup_hg19/Flt3l_1base/cellSNP.tag.DP.mtx\n",
      "4 ./output/variants_pileup_hg19/Control_noEXCL/cellSNP.tag.DP.mtx\n",
      "6 ./output/variants_pileup_hg19/Input_1base/cellSNP.tag.DP.mtx\n",
      "6 ./output/variants_pileup_hg19/Input_noEXCL/cellSNP.tag.DP.mtx\n",
      "4 ./output/variants_pileup_hg19/Control_1base/cellSNP.tag.DP.mtx\n",
      "24 ./output/variants_pileup_hg19/Flt3l_noEXCL/cellSNP.tag.DP.mtx\n"
     ]
    }
   ],
   "source": [
    "cmd = f'find {outdir} -name \"*DP.mtx\" -exec wc -l {{}} \\;'\n",
    "!{cmd}\n",
    "#!find output/ -name \"*DP.mtx\" -exec wc -l {} \\;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7010d46b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "print((AD>0).sum().sum())\n",
    "print((AF>0).sum(axis=1))\n",
    "f, ax = plt.subplots(ncols=2, figsize=(10,8))\n",
    "sns.heatmap(AF.fillna(0), ax=ax[0], vmin=0)\n",
    "sns.heatmap(DP, ax=ax[1], vmin=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51227ddf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9a49847c",
   "metadata": {},
   "source": [
    "## plot"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
