{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8f3eabb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from os.path import join, isdir\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "from icecream import ic\n",
    "from src.utils import variant_utils as vu\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e27fb5e4",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "mult_dir = \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex\"\n",
    "don_dir = \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex/clones_init/\"\n",
    "samples = \"preB,postB\"\n",
    "mt_genome=\"/data/Mito_Trace/data/external/GRCh38_MT_blacklist/chrM.fasta\"\n",
    "t=0.1\n",
    "rt=50\n",
    "n_donors = 4\n",
    "outdir = \"/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex/mt_clones/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6b798949",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cda898e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex/clones_init/donor0',\n",
       " 1: '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex/clones_init/donor1'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "don_dirs = [d for d in glob(don_dir + \"/donor*\") if isdir(d)]\n",
    "don_dirs = {int(x.split(\"donor\")[-1]):x for x in don_dirs}\n",
    "don_dirs = {x:don_dirs[x] for x in don_dirs if x in range(n_donors)}\n",
    "don_dirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "734b446b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# thresholds = [0.01, 0.1, 0.25, 0.5, 0.75, 0.9]\n",
    "# read_thresh = [2, 10, 25, 50, 100]\n",
    "# thresholds = [0.1, 0.25]#\n",
    "# read_thresh = [25, 50]\n",
    "\n",
    "thresholds = [t]\n",
    "read_thresh=[rt]\n",
    "cells_thresh=10\n",
    "ref_cells_thresh=500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "51bd08d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = samples.split(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f92e1ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_af_in_cell(cell_var, t, rt=None, is_af_above=True, check_cov=False):\n",
    "    \"\"\"\n",
    "    Checks which cells pass the thresholds using the af and cov columns\n",
    "    \"\"\"\n",
    "    if check_cov:\n",
    "        return cell_var[cell_var[\"cov\"]>rt].index\n",
    "    if is_af_above:\n",
    "        if rt is None:\n",
    "            return cell_var[cell_var[\"af\"] > t].index\n",
    "        else:\n",
    "            return cell_var[(cell_var[\"af\"] > t) & (cell_var[\"cov\"] > rt)].index\n",
    "    else:\n",
    "        if rt is None:\n",
    "            return cell_var[cell_var[\"af\"] < t].index\n",
    "        else:\n",
    "            return cell_var[(cell_var[\"af\"] < t) & (cell_var[\"cov\"] > rt)].index        \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "72706400",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_variant_descriptives(cells_df, name, prefix=\"\"):#, t, rt, \n",
    "    vars_ser = {}\n",
    "    vars_ser[f\"{prefix}{name}_median\"] = cells_df[\"af\"].median()\n",
    "    vars_ser[f\"{prefix}{name}_mean\"] = cells_df[\"af\"].mean(axis=0)\n",
    "    vars_ser[f\"{prefix}{name}_coverage_median\"] = cells_df[\"cov\"].median()\n",
    "    vars_ser[f\"{prefix}{name}_coverage_mean\"] = cells_df[\"cov\"].mean()\n",
    "    vars_ser[f\"{prefix}{name}_nCells\"] = len(cells_df)\n",
    "    return pd.Series(vars_ser)\n",
    "\n",
    "\n",
    "def set_keys(cols, curr):\n",
    "    keys = [\"_median\", \"_mean\", \"_nCells\", \"_coverage_median\", \"_coverage_mean\"]\n",
    "    for k in keys:\n",
    "        cols.append(f\"{curr}{k}\")\n",
    "    return cols\n",
    "\n",
    "\n",
    "def add_columns(columns, thresholds, read_thresh, sample=None):\n",
    "    if sample is None:\n",
    "        prefix = \"\"\n",
    "    else:\n",
    "        prefix=f\"cond{sample}_\"\n",
    "\n",
    "    for t in thresholds:\n",
    "        #columns.append(t)\n",
    "        columns = set_keys(columns, f\"{prefix}{t}\" )\n",
    "        for rt in read_thresh:\n",
    "            columns = set_keys(columns, f\"{prefix}{t}_{rt}\" )\n",
    "            columns = set_keys(columns, f\"{prefix}{t}_{rt}_refPassCov\" )\n",
    "    return columns\n",
    "\n",
    "#########################\n",
    "## Add in sample information\n",
    "#########################\n",
    "def get_variant_info(curr_af, curr_cov, columns, variants, prefix=\"\" ):\n",
    "    variants_cell_index_d = {}\n",
    "    variants_df_init = {}\n",
    "\n",
    "    for v in variants:\n",
    "        #print('variant', v)\n",
    "        curr_v_d = {}\n",
    "        variants_d = {}\n",
    "        \n",
    "        \n",
    "        # create cell-by-(af,cov) df to pass into summary functions\n",
    "        curr_v_df = pd.concat((curr_af.loc[:,[v]].rename({v:\"af\"}, axis=1),\n",
    "                               curr_cov.loc[:,[v]].rename({v:\"cov\"}, axis=1)), axis=1)\n",
    "        for t in thresholds:\n",
    "            if prefix == \"\":\n",
    "                t_key = t\n",
    "            else:\n",
    "                t_key = (t, prefix)\n",
    "            curr_v_d[t_key] = is_af_in_cell(curr_v_df, t, rt=None, is_af_above=True, check_cov=False)\n",
    "\n",
    "            variants_d[t_key] = calc_variant_descriptives(curr_v_df.loc[curr_v_d[t_key]], name=f\"{prefix}{t}\")\n",
    "            for rt in read_thresh:\n",
    "                if prefix == \"\":\n",
    "                    trt_key = (t,rt)\n",
    "                    reftrt_key = (t, rt, \"ref_pass_cov\")\n",
    "                else:\n",
    "                    trt_key = (t,rt, prefix) \n",
    "                    reftrt_key = (t, rt, prefix, \"ref_pass_cov\")\n",
    "                \n",
    "                curr_v_d[trt_key] = is_af_in_cell(curr_v_df, t, rt=rt, is_af_above=True, check_cov=False)\n",
    "                variants_d[trt_key] = calc_variant_descriptives(curr_v_df.loc[curr_v_d[trt_key]], \n",
    "                                                                     name=f\"{prefix}{t}_{rt}\")\n",
    "                #ic(trt_key)\n",
    "                #ic(variants_d[trt_key].head())\n",
    "                curr_v_d[reftrt_key] = is_af_in_cell(curr_v_df, t, rt=rt, \n",
    "                                                                 is_af_above=False, check_cov=False)\n",
    "                variants_d[reftrt_key] = calc_variant_descriptives(curr_v_df.loc[curr_v_d[reftrt_key]], \n",
    "                                                                               name=f\"{prefix}{t}_{rt}_refPassCov\"\n",
    "                                                                               )\n",
    "                    \n",
    "        for rt in read_thresh:\n",
    "            if prefix == \"\":\n",
    "                rt_key = (rt, \"pass_cov\")\n",
    "            else:\n",
    "                rt_key = (rt,  \"pass_cov\", prefix)\n",
    "            curr_v_d[rt_key] = is_af_in_cell(curr_v_df, t=None, rt=rt, is_af_above=False, check_cov=True)\n",
    "            variants_d[rt_key] = calc_variant_descriptives(curr_v_df.loc[curr_v_d[rt_key]], \n",
    "                                                                                 name=f\"{prefix}{rt}_passCov\")\n",
    "        variants_cell_index_d[v] = curr_v_d \n",
    "        variants_df_init[v] = variants_d    \n",
    "     \n",
    "\n",
    "\n",
    "    # Combine all the variants from the dictionary\n",
    "    variants_df = pd.DataFrame(index=variants, columns = columns)\n",
    "    for v in variants_df_init:\n",
    "        variants_df.loc[v] = pd.concat([val for val in variants_df_init[v].values()])\n",
    "\n",
    "    variants_df[f\"{prefix}median\"] = curr_af.median()\n",
    "    variants_df[f\"{prefix}mean\"] = curr_af.mean()\n",
    "    variants_df[f\"{prefix}median_cov\"] = curr_cov.median()\n",
    "    variants_df[f\"{prefix}mean_cov\"] = curr_cov.mean()\n",
    "    return variants_df, variants_cell_index_d, variants_df_init\n",
    "\n",
    "def create_cluster_size(variants_df, cells_thresh, ref_cells_thresh):\n",
    "    \"\"\"Get clone sizes using clones that passed the filter\n",
    "    \n",
    "    \"\"\"\n",
    "    inds = [f\"{t}_{rt}_nCells\" for t in thresholds for rt in read_thresh]\n",
    "    num_vars_df = pd.DataFrame(index=inds, columns=[\"nClones\", \"nClones_and_nRef\"])#, \"fold_nClones_nRef\"])\n",
    "    clone_sizes_d = {}\n",
    "    for t in thresholds:\n",
    "        for rt in read_thresh:\n",
    "            num_vars_df.loc[f\"{t}_{rt}_nCells\",\"nClones\"] = (variants_df[f\"{t}_{rt}_nCells\"] > cells_thresh).sum()\n",
    "                                                      \n",
    "            num_vars_df.loc[f\"{t}_{rt}_nCells\",\"nClones_and_nRef\"] = ((variants_df[f\"{t}_{rt}_nCells\"] > cells_thresh) & \n",
    "                                                                  (variants_df[f\"{t}_{rt}_refPassCov_nCells\"] > ref_cells_thresh)).sum() \n",
    "            \n",
    "            passed_vars = variants_df.loc[((variants_df[f\"{t}_{rt}_nCells\"] > cells_thresh) & \n",
    "                                          (variants_df[f\"{t}_{rt}_refPassCov_nCells\"] > ref_cells_thresh))].index \n",
    "            curr_clone_sizes_d = pd.DataFrame(index=passed_vars, columns=[\"clone_sizes\", \"ref_sizes\", \"param\"])\n",
    "            curr_clone_sizes_d[\"param\"] = f\"{t}_{rt}\"\n",
    "            curr_clone_sizes_d[\"clone_sizes\"] =  variants_df.loc[passed_vars, f\"{t}_{rt}_nCells\"]\n",
    "            curr_clone_sizes_d[\"ref_sizes\"] =  variants_df.loc[passed_vars, f\"{t}_{rt}_refPassCov_nCells\"]\n",
    "            curr_clone_sizes_d[\"clone_plus_ref_sizes\"] = curr_clone_sizes_d[\"clone_sizes\"] + curr_clone_sizes_d[\"ref_sizes\"]\n",
    "            curr_clone_sizes_d[\"relative_size\"] = curr_clone_sizes_d[\"clone_sizes\"]/curr_clone_sizes_d[\"ref_sizes\"]\n",
    "            curr_clone_sizes_d[\"variant\"] = passed_vars\n",
    "            clone_sizes_d[(t, rt)] = curr_clone_sizes_d \n",
    "    return clone_sizes_d, num_vars_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad7e2a12",
   "metadata": {},
   "source": [
    "## For each parameter, rank clones by i) highest nCells and ii) highest nRef\n",
    "## Create df where index is cell and columns are 'variant,' 'isClone', 'af' \n",
    "### To do this, use variants dict of passed indices\n",
    "## For each variant plot the distribution and hue=isClone\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6da329a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_distributions(t, rt, variants_cell_index_d, af, clone_thresh, ref_cells_thresh, prefix=\"\"):\n",
    "    ic(t,rt)\n",
    "    clone_var_distributions = pd.DataFrame(columns=[\"variant\", \"isClone\", \"af\"])\n",
    "    for v in variants_cell_index_d:\n",
    "        curr_v = variants_cell_index_d[v]\n",
    "        curr_cells = curr_v[(t,rt)]\n",
    "        curr_ref_cells = curr_v[(t,rt, \"ref_pass_cov\")]\n",
    "            \n",
    "        if len(curr_cells) < clone_thresh or len(curr_ref_cells) < ref_cells_thresh:\n",
    "            #ic(f\"variant {v} not pass threshold\")\n",
    "            continue \n",
    "        curr_clone = pd.DataFrame(index=curr_cells, columns=[\"variant\", \"isClone\", \"af\", \"param\"])\n",
    "        curr_clone[\"isClone\"] = True\n",
    "        curr_clone[\"variant\"] = v\n",
    "        curr_clone[\"param\"] = f\"{t}_{rt}\"\n",
    "        curr_clone[\"af\"] = af.loc[curr_cells, v]\n",
    "        curr_clone[\"cell\"] = af.loc[curr_cells].index\n",
    "        clone_var_distributions = pd.concat((clone_var_distributions, \n",
    "                                             curr_clone),axis=0, ignore_index=True)\n",
    "        curr_ref = pd.DataFrame(index=curr_ref_cells, columns=[\"variant\", \"isClone\", \"af\"])\n",
    "        curr_ref[\"isClone\"] = False\n",
    "        curr_ref[\"variant\"] = v\n",
    "        curr_ref[\"param\"] = f\"{t}_{rt}\"\n",
    "        curr_ref[\"af\"] = af.loc[curr_ref_cells, v]\n",
    "        curr_ref[\"cell\"] = af.loc[curr_ref_cells].index\n",
    "        clone_var_distributions = pd.concat((clone_var_distributions, \n",
    "                                             curr_ref),axis=0, ignore_index=True)          \n",
    "    return clone_var_distributions\n",
    "\n",
    "\n",
    "def create_distributions_samples(t, rt, variants_cell_index_d, af, clone_thresh, ref_cells_thresh, samples):\n",
    "    ic(t,rt)\n",
    "    clone_var_distributions = pd.DataFrame(columns=[\"variant\", \"sample\", \"af\", \"param\"])\n",
    "    for v in variants_cell_index_d:\n",
    "        curr_v = variants_cell_index_d[v]\n",
    "        curr_cells = curr_v[(t,rt, f\"cond{samples[0]}_\")]\n",
    "        curr_ref_cells = curr_v[(t,rt, f\"cond{samples[1]}_\")]\n",
    "            \n",
    "        if len(curr_cells) < clone_thresh or len(curr_ref_cells) < ref_cells_thresh:\n",
    "            #ic(f\"variant {v} not pass threshold\")\n",
    "            continue \n",
    "        curr_clone = pd.DataFrame(index=curr_cells, columns=[\"variant\", \"sample\", \"af\", \"param\"])\n",
    "        curr_clone[\"sample\"] = samples[0]\n",
    "        curr_clone[\"variant\"] = v\n",
    "        curr_clone[\"param\"] = f\"{t}_{rt}\"\n",
    "        curr_clone[\"af\"] = af.loc[curr_cells, v]\n",
    "        curr_clone[\"cell\"] = af.loc[curr_cells].index\n",
    "        clone_var_distributions = pd.concat((clone_var_distributions, \n",
    "                                             curr_clone),axis=0, ignore_index=True)\n",
    "        curr_ref = pd.DataFrame(index=curr_ref_cells, columns=[\"variant\", \"isClone\", \"af\"])\n",
    "        curr_ref[\"sample\"] = samples[1]\n",
    "        curr_ref[\"variant\"] = v\n",
    "        curr_ref[\"param\"] = f\"{t}_{rt}\"\n",
    "        curr_ref[\"af\"] = af.loc[curr_ref_cells, v]\n",
    "        curr_ref[\"cell\"] = af.loc[curr_ref_cells].index\n",
    "        clone_var_distributions = pd.concat((clone_var_distributions, \n",
    "                                             curr_ref),axis=0, ignore_index=True)          \n",
    "    return clone_var_distributions\n",
    "\n",
    "\n",
    "\n",
    "## Number of clones across different AF+coverage thresholds\n",
    "#(2 other parameters, min_cells=10 and ref_cells_thresh=500 are used here, which are the thresholds for the cluster sizes and the reference variant being covered by a sufficient number of cells)\n",
    "\n",
    "def plot_violin(clone_var_dist, col_order, x=\"isClone\"):\n",
    "    g = sns.FacetGrid(data=clone_var_dist, col=\"variant\", col_wrap=8, \n",
    "                      sharex=False, \n",
    "                      sharey=False,col_order=col_order)\n",
    "    g.map_dataframe(sns.violinplot, x=x, y=\"af\")    \n",
    "    return g\n",
    "\n",
    "def plot_hist(clone_var_dist, col_order, hue_col=\"isClone\"):\n",
    "    g = sns.FacetGrid(data=clone_var_distributions, col=\"variant\", col_wrap=8, hue=hue_col,\n",
    "                      sharex=False, \n",
    "                      sharey=False, col_order=col_order)\n",
    "    g.map_dataframe(sns.histplot, x=\"af\", stat=\"probability\")\n",
    "    return\n",
    "\n",
    "def get_var_order(clone_var_dist, variants_df, t, rt):\n",
    "    curr_vars = set(clone_var_dist[\"variant\"])\n",
    "    var_order = variants_df.loc[curr_vars].sort_values([f\"{t}_{rt}_nCells\", f\"{t}_{rt}_refPassCov_nCells\"], ascending=False).index\n",
    "    return var_order\n",
    "\n",
    "# def get_var_order(clone_var_dist, variants_df, t, rt):\n",
    "#     curr_vars = set(clone_var_dist[\"variant\"])\n",
    "#     var_order = variants_df.loc[curr_vars].sort_values([f\"{t}_{rt}_nCells\", f\"{t}_{rt}_refPassCov_nCells\"], ascending=False).index\n",
    "#     return var_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "24274c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_don(don_dir):\n",
    "    vcf = pd.read_csv(join(don_dir, \"cellSNP.base.vcf\"), sep=\"\\t\")\n",
    "    variants_meta = vcf[[\"#CHROM\", \"POS\", \"REF\", \"ALT\"]]\n",
    "    variants_meta.index = variants_meta.apply(lambda x: f\"{x['POS']}{x['REF'][0]}>{x['ALT']}\", axis=1)\n",
    "    variants_meta = vu.type_of_variants(variants_meta.index)\n",
    "    variants_meta[\"ID\"] = variants_meta.index\n",
    "    variants_meta.index = variants_meta.apply(lambda x: f\"{x['position']}{x['alt']}\", axis=1)\n",
    "    af = pd.read_csv(join(don_dir, \"af.tsv\"), sep=\"\\t\", index_col=0)\n",
    "    cov = pd.read_csv(join(don_dir, \"dp.tsv\"), sep=\"\\t\", index_col=0)                   \n",
    "    cells_meta = pd.read_csv(join(don_dir,\"cells_meta.tsv\"), sep=\"\\t\").set_index(\"ID\")\n",
    "    ic(cells_meta.shape)\n",
    "    cells_meta = cells_meta.loc[af.index]\n",
    "    ic(\"After filtering on af indices\", cells_meta.shape)\n",
    "    \n",
    "    variants = set(af.columns)\n",
    "    variants_cell_index_d = {}\n",
    "    #variants_df_init = {}\n",
    "    columns = add_columns([], thresholds, read_thresh)\n",
    "    variants_df, variants_cell_index_d, _ = get_variant_info(af, cov, columns=columns, prefix=\"\", variants=variants)\n",
    "\n",
    "    for sample, curr_cells_meta in cells_meta.groupby(\"condition\"):\n",
    "        print(sample)\n",
    "        print(curr_cells_meta.shape)\n",
    "\n",
    "        curr_af = af.loc[curr_cells_meta.index]\n",
    "        curr_cov = cov.loc[curr_cells_meta.index]\n",
    "        curr_cols = add_columns([], thresholds, read_thresh, sample=sample)\n",
    "        columns += curr_cols\n",
    "        curr_variants_df, curr_index_d, curr_variants_df_init = get_variant_info(curr_af, curr_cov, columns=curr_cols,\n",
    "                                                                   prefix=f\"cond{sample}_\", variants=variants)\n",
    "        variants_df = pd.concat((variants_df, curr_variants_df), axis=1)\n",
    "\n",
    "        for v in curr_index_d.keys(): \n",
    "            for k in curr_index_d[v].keys():\n",
    "                assert(k not in variants_cell_index_d[v].keys())\n",
    "                variants_cell_index_d[v][k] = curr_index_d[v][k]\n",
    "    variants_df = pd.merge(variants_meta, variants_df, left_index=True, right_index=True, how=\"right\")\n",
    "\n",
    "    ## Remove high variants\n",
    "    variants_df = variants_df[~(variants_df[\"median\"] > 0.6)]\n",
    "\n",
    "    clone_sizes_d, num_vars_df = create_cluster_size(variants_df, cells_thresh=10, ref_cells_thresh=500)\n",
    "    clone_sizes_d = pd.concat(clone_sizes_d, ignore_index=True)\n",
    "\n",
    "    num_vars_mat_df = pd.DataFrame(index=thresholds,columns=read_thresh, dtype=int)\n",
    "    for t in thresholds:\n",
    "        for rt in read_thresh:\n",
    "            num_vars_mat_df.loc[t,rt] = num_vars_df.loc[f\"{t}_{rt}_nCells\",\"nClones\"]\n",
    "    \n",
    "    # Create binary clone output for each param\n",
    "#     t=0.1\n",
    "#     rt=25\n",
    "#     clone_var_distributions = create_distributions(t,rt, variants_cell_index_d, af, cells_thresh, ref_cells_thresh) \n",
    "    clone_var_distributions = create_distributions(t,rt, variants_cell_index_d, af, cells_thresh, ref_cells_thresh)\n",
    "    print('clone_var_distributions')\n",
    "    print(clone_var_distributions.head())\n",
    "    clone_out_df = clone_var_distributions.pivot(index=\"cell\", columns=\"variant\", values=\"af\").fillna(0)\n",
    "    return clone_out_df,  variants_cell_index_d, clone_var_distributions, clone_sizes_d, variants_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "19e1db43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex/clones_init/donor0',\n",
       " 1: '/data/Mito_Trace/output/pipeline/TcellDupi_may17_2021/MTblacklist/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex/clones_init/donor1'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "don_dirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dc85d6c9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ic| cells_meta.shape: (6746, 6)\n",
      "ic| 'After filtering on af indices', cells_meta.shape: (6746, 6)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "post\n",
      "(4133, 6)\n",
      "pre\n",
      "(2613, 6)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ic| t: 0.1, rt: 50\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clone_var_distributions\n",
      "  variant isClone        af   param                    cell\n",
      "0   3847C    True  0.141176  0.1_50  AAACTGCCAAGTCTCA-1_pre\n",
      "1   3847C    True  0.176101  0.1_50  AATGCCATCTATCCTA-1_pre\n",
      "2   3847C    True  0.333333  0.1_50  CAAGCTATCGATCGCG-1_pre\n",
      "3   3847C    True  0.112676  0.1_50  CAAGGCCAGTGGTGTG-1_pre\n",
      "4   3847C    True  0.106383  0.1_50  CACCTTGCAAGACTAA-1_pre\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ic| cells_meta.shape: (8017, 6)\n",
      "ic| 'After filtering on af indices', cells_meta.shape: (8017, 6)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "post\n",
      "(5297, 6)\n",
      "pre\n",
      "(2720, 6)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ic| t: 0.1, rt: 50\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clone_var_distributions\n",
      "  variant isClone        af   param                     cell\n",
      "0   6456A    True  0.120482  0.1_50   GGTCATATCCCGGGTA-1_pre\n",
      "1   6456A    True  0.181818  0.1_50   TCAGTCCCATGGGAAC-1_pre\n",
      "2   6456A    True  0.105263  0.1_50  AACAAAGGTCACCCGA-1_post\n",
      "3   6456A    True  0.115385  0.1_50  AAGATAGCAAAGCATA-1_post\n",
      "4   6456A    True  0.142857  0.1_50  CTTAATCTCCAGGCCA-1_post\n"
     ]
    }
   ],
   "source": [
    "out_donors = {}\n",
    "for d in don_dirs: \n",
    "    clone_out_df, variants_cell_index_d, clone_var_distributions, clone_sizes_d, variants_df = run_don(don_dirs[d])\n",
    "    out_donors[d] = clone_out_df.fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a5930df",
   "metadata": {},
   "source": [
    "## Save cells-by-MT matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0e7f132b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>variant</th>\n",
       "      <th>10646T</th>\n",
       "      <th>12295C</th>\n",
       "      <th>13188T</th>\n",
       "      <th>14568A</th>\n",
       "      <th>14766T</th>\n",
       "      <th>150T</th>\n",
       "      <th>16126C</th>\n",
       "      <th>16362C</th>\n",
       "      <th>16537T</th>\n",
       "      <th>2351C</th>\n",
       "      <th>...</th>\n",
       "      <th>57C</th>\n",
       "      <th>58C</th>\n",
       "      <th>6093A</th>\n",
       "      <th>6456A</th>\n",
       "      <th>66T</th>\n",
       "      <th>6776C</th>\n",
       "      <th>7595A</th>\n",
       "      <th>7775A</th>\n",
       "      <th>8770T</th>\n",
       "      <th>9087A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGCGTTGCC-1_post</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>0.060606</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.026667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGGCTCCTG-1_pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGGTTAACA-1_post</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.016129</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGTGAATAC-1_pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGTGCTAGG-1_post</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTGTTCTTTCA-1_post</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCCGGAAAG-1_post</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.029851</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCCTGACCT-1_post</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCGCGATGC-1_pre</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.072581</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTGTTCTTAATCC-1_post</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14760 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "variant                  10646T  12295C    13188T  14568A    14766T      150T  \\\n",
       "cell                                                                            \n",
       "AAACGAAAGCGTTGCC-1_post     0.0     0.0  0.000000     0.0  0.215686  0.060606   \n",
       "AAACGAAAGGCTCCTG-1_pre      0.0     0.0  0.000000     0.0  0.000000  0.000000   \n",
       "AAACGAAAGGTTAACA-1_post     0.0     0.0  0.016129     0.0  0.000000  0.000000   \n",
       "AAACGAAAGTGAATAC-1_pre      0.0     0.0  0.000000     0.0  0.000000  0.000000   \n",
       "AAACGAAAGTGCTAGG-1_post     0.0     0.0  0.036364     0.0  0.000000  0.000000   \n",
       "...                         ...     ...       ...     ...       ...       ...   \n",
       "TTTGTGTGTTCTTTCA-1_post     NaN     NaN       NaN     NaN       NaN       NaN   \n",
       "TTTGTGTTCCGGAAAG-1_post     NaN     NaN       NaN     NaN       NaN       NaN   \n",
       "TTTGTGTTCCTGACCT-1_post     NaN     NaN       NaN     NaN       NaN       NaN   \n",
       "TTTGTGTTCGCGATGC-1_pre      NaN     NaN       NaN     NaN       NaN       NaN   \n",
       "TTTGTGTTCTTAATCC-1_post     NaN     NaN       NaN     NaN       NaN       NaN   \n",
       "\n",
       "variant                  16126C    16362C  16537T  2351C  ...       57C  58C  \\\n",
       "cell                                                      ...                  \n",
       "AAACGAAAGCGTTGCC-1_post     0.0  0.026667     0.0    0.0  ...       NaN  NaN   \n",
       "AAACGAAAGGCTCCTG-1_pre      0.0  0.000000     0.0    0.0  ...       NaN  NaN   \n",
       "AAACGAAAGGTTAACA-1_post     0.0  0.035714     0.0    0.0  ...       NaN  NaN   \n",
       "AAACGAAAGTGAATAC-1_pre      0.0  0.000000     0.0    0.0  ...       NaN  NaN   \n",
       "AAACGAAAGTGCTAGG-1_post     0.0  0.000000     0.0    0.0  ...       NaN  NaN   \n",
       "...                         ...       ...     ...    ...  ...       ...  ...   \n",
       "TTTGTGTGTTCTTTCA-1_post     NaN       NaN     NaN    NaN  ...  0.000000  0.0   \n",
       "TTTGTGTTCCGGAAAG-1_post     NaN       NaN     NaN    NaN  ...  0.000000  0.0   \n",
       "TTTGTGTTCCTGACCT-1_post     NaN       NaN     NaN    NaN  ...  0.000000  0.0   \n",
       "TTTGTGTTCGCGATGC-1_pre      NaN       NaN     NaN    NaN  ...  0.072581  0.0   \n",
       "TTTGTGTTCTTAATCC-1_post     NaN       NaN     NaN    NaN  ...  0.000000  0.0   \n",
       "\n",
       "variant                  6093A  6456A       66T     6776C  7595A  7775A  \\\n",
       "cell                                                                      \n",
       "AAACGAAAGCGTTGCC-1_post    NaN    NaN       NaN       NaN    NaN    NaN   \n",
       "AAACGAAAGGCTCCTG-1_pre     NaN    NaN       NaN       NaN    NaN    NaN   \n",
       "AAACGAAAGGTTAACA-1_post    NaN    NaN       NaN       NaN    NaN    NaN   \n",
       "AAACGAAAGTGAATAC-1_pre     NaN    NaN       NaN       NaN    NaN    NaN   \n",
       "AAACGAAAGTGCTAGG-1_post    NaN    NaN       NaN       NaN    NaN    NaN   \n",
       "...                        ...    ...       ...       ...    ...    ...   \n",
       "TTTGTGTGTTCTTTCA-1_post    0.0    0.0  0.000000  0.000000    0.0    0.0   \n",
       "TTTGTGTTCCGGAAAG-1_post    0.0    0.0  0.029851  0.000000    0.0    0.0   \n",
       "TTTGTGTTCCTGACCT-1_post    0.0    0.0  0.000000  0.033333    0.0    0.0   \n",
       "TTTGTGTTCGCGATGC-1_pre     0.0    0.0  0.000000  0.000000    0.0    0.0   \n",
       "TTTGTGTTCTTAATCC-1_post    0.0    0.0  0.000000  0.000000    0.0    0.0   \n",
       "\n",
       "variant                  8770T  9087A  \n",
       "cell                                   \n",
       "AAACGAAAGCGTTGCC-1_post    NaN    NaN  \n",
       "AAACGAAAGGCTCCTG-1_pre     NaN    NaN  \n",
       "AAACGAAAGGTTAACA-1_post    NaN    NaN  \n",
       "AAACGAAAGTGAATAC-1_pre     NaN    NaN  \n",
       "AAACGAAAGTGCTAGG-1_post    NaN    NaN  \n",
       "...                        ...    ...  \n",
       "TTTGTGTGTTCTTTCA-1_post    0.0    0.0  \n",
       "TTTGTGTTCCGGAAAG-1_post    0.0    0.0  \n",
       "TTTGTGTTCCTGACCT-1_post    0.0    0.0  \n",
       "TTTGTGTTCGCGATGC-1_pre     0.0    0.0  \n",
       "TTTGTGTTCTTAATCC-1_post    0.0    0.0  \n",
       "\n",
       "[14760 rows x 44 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_donors_df = pd.concat([d for d in out_donors.values()])\n",
    "out_donors_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b0147740",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_donors_df.to_csv(join(outdir, \"cells_mt.tsv\"), sep=\"\\t\")\n",
    "(out_donors_df>t).astype(int).to_csv(join(outdir, \"cells_meta.tsv\"), sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d5e86502",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([3.233e+03, 5.800e+01, 1.700e+01, 4.000e+00, 1.000e+00, 7.000e+00,\n",
       "        7.000e+00, 4.000e+00, 8.000e+00, 5.000e+00]),\n",
       " array([ 1. ,  2.2,  3.4,  4.6,  5.8,  7. ,  8.2,  9.4, 10.6, 11.8, 13. ]),\n",
       " <BarContainer object of 10 artists>)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD4CAYAAAAAczaOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAARAklEQVR4nO3df8ydZX3H8fdnrUP8QYTwQGrbrMx0TiCzjKbrRrI40dGpsfiHSc0mTUZSQ3DTxWQD94fujy4s88dGNliqMkrGII0/QqPi7DoXY4LiA2OUUhmNMHhsRx81TtwSttbv/ngukrNyeH73nJbr/UpO7vt8z3Wd+3vT9tO717nPQ6oKSVIffmbcDUiSRsfQl6SOGPqS1BFDX5I6YuhLUkdWjruBuZx//vm1bt26cbchSWeUBx544PtVNXFy/bQP/XXr1jE5OTnuNiTpjJLk34fVXd6RpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOnPbfyF2KdTd8aSzHffKmt4/luJI0F6/0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHVkztBP8vIk9yf51yQHk/xJq5+XZF+Sx9v23IE5NyY5nOSxJFcN1C9PcqC9dnOSnJrTkiQNM58r/eeAN1fVG4ENwJYkm4EbgP1VtR7Y356T5GJgG3AJsAW4JcmK9l63AjuA9e2xZflORZI0lzlDv2b8pD19WXsUsBXY3eq7gavb/lbg7qp6rqqeAA4Dm5KsAs6pqvuqqoA7BuZIkkZgXmv6SVYkeQg4Buyrqm8BF1bVUYC2vaANXw08PTB9qtVWt/2T65KkEZlX6FfViaraAKxh5qr90lmGD1unr1nqL3yDZEeSySST09PT82lRkjQPC7p7p6p+BPwzM2vxz7QlG9r2WBs2BawdmLYGONLqa4bUhx1nV1VtrKqNExMTC2lRkjSL+dy9M5HkNW3/bOAtwHeAvcD2Nmw7cE/b3wtsS3JWkouY+cD2/rYE9GySze2unWsG5kiSRmA+/+esVcDudgfOzwB7quqLSe4D9iS5FngKeDdAVR1Msgd4FDgOXF9VJ9p7XQfcDpwN3NsekqQRmTP0q+ph4LIh9R8AV77InJ3AziH1SWC2zwMkSaeQ38iVpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1ZM7QT7I2ydeSHEpyMMkHWv2jSb6X5KH2eNvAnBuTHE7yWJKrBuqXJznQXrs5SU7NaUmShlk5jzHHgQ9V1YNJXg08kGRfe+2TVfWxwcFJLga2AZcArwX+MckvVNUJ4FZgB/BN4MvAFuDe5TkVSdJc5rzSr6qjVfVg238WOASsnmXKVuDuqnquqp4ADgObkqwCzqmq+6qqgDuAq5d6ApKk+VvQmn6SdcBlwLda6f1JHk5yW5JzW2018PTAtKlWW932T64PO86OJJNJJqenpxfSoiRpFvMO/SSvAj4HfLCqfszMUs3rgA3AUeDjzw8dMr1mqb+wWLWrqjZW1caJiYn5tihJmsO8Qj/Jy5gJ/Dur6vMAVfVMVZ2oqp8CnwI2teFTwNqB6WuAI62+ZkhdkjQi87l7J8BngENV9YmB+qqBYe8CHmn7e4FtSc5KchGwHri/qo4CzybZ3N7zGuCeZToPSdI8zOfunSuA9wIHkjzUah8G3pNkAzNLNE8C7wOoqoNJ9gCPMnPnz/Xtzh2A64DbgbOZuWvHO3ckaYTmDP2q+gbD1+O/PMucncDOIfVJ4NKFNChJWj5+I1eSOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0JekjswZ+knWJvlakkNJDib5QKufl2Rfksfb9tyBOTcmOZzksSRXDdQvT3KgvXZzkpya05IkDTOfK/3jwIeq6g3AZuD6JBcDNwD7q2o9sL89p722DbgE2ALckmRFe69bgR3A+vbYsoznIkmaw5yhX1VHq+rBtv8scAhYDWwFdrdhu4Gr2/5W4O6qeq6qngAOA5uSrALOqar7qqqAOwbmSJJGYEFr+knWAZcB3wIurKqjMPMXA3BBG7YaeHpg2lSrrW77J9eHHWdHkskkk9PT0wtpUZI0i3mHfpJXAZ8DPlhVP55t6JBazVJ/YbFqV1VtrKqNExMT821RkjSHeYV+kpcxE/h3VtXnW/mZtmRD2x5r9Slg7cD0NcCRVl8zpC5JGpH53L0T4DPAoar6xMBLe4HtbX87cM9AfVuSs5JcxMwHtve3JaBnk2xu73nNwBxJ0gisnMeYK4D3AgeSPNRqHwZuAvYkuRZ4Cng3QFUdTLIHeJSZO3+ur6oTbd51wO3A2cC97SFJGpE5Q7+qvsHw9XiAK19kzk5g55D6JHDpQhqUJC0fv5ErSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkfmDP0ktyU5luSRgdpHk3wvyUPt8baB125McjjJY0muGqhfnuRAe+3mJFn+05EkzWY+V/q3A1uG1D9ZVRva48sASS4GtgGXtDm3JFnRxt8K7ADWt8ew95QknUJzhn5VfR344Tzfbytwd1U9V1VPAIeBTUlWAedU1X1VVcAdwNWL7FmStEhLWdN/f5KH2/LPua22Gnh6YMxUq61u+yfXh0qyI8lkksnp6ekltChJGrTY0L8VeB2wATgKfLzVh63T1yz1oapqV1VtrKqNExMTi2xRknSyRYV+VT1TVSeq6qfAp4BN7aUpYO3A0DXAkVZfM6QuSRqhRYV+W6N/3ruA5+/s2QtsS3JWkouY+cD2/qo6CjybZHO7a+ca4J4l9C1JWoSVcw1IchfwJuD8JFPAR4A3JdnAzBLNk8D7AKrqYJI9wKPAceD6qjrR3uo6Zu4EOhu4tz0kSSM0Z+hX1XuGlD8zy/idwM4h9Ung0gV1J0laVn4jV5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOzBn6SW5LcizJIwO185LsS/J425478NqNSQ4neSzJVQP1y5McaK/dnCTLfzqSpNnM50r/dmDLSbUbgP1VtR7Y356T5GJgG3BJm3NLkhVtzq3ADmB9e5z8npKkU2zO0K+qrwM/PKm8Fdjd9ncDVw/U766q56rqCeAwsCnJKuCcqrqvqgq4Y2COJGlEFrumf2FVHQVo2wtafTXw9MC4qVZb3fZPrg+VZEeSySST09PTi2xRknSy5f4gd9g6fc1SH6qqdlXVxqraODExsWzNSVLvFhv6z7QlG9r2WKtPAWsHxq0BjrT6miF1SdIILTb09wLb2/524J6B+rYkZyW5iJkPbO9vS0DPJtnc7tq5ZmCOJGlEVs41IMldwJuA85NMAR8BbgL2JLkWeAp4N0BVHUyyB3gUOA5cX1Un2ltdx8ydQGcD97aHJGmE5gz9qnrPi7x05YuM3wnsHFKfBC5dUHeSpGXlN3IlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6siSQj/Jk0kOJHkoyWSrnZdkX5LH2/bcgfE3Jjmc5LEkVy21eUnSwizHlf5vVNWGqtrYnt8A7K+q9cD+9pwkFwPbgEuALcAtSVYsw/ElSfN0KpZ3tgK72/5u4OqB+t1V9VxVPQEcBjadguNLkl7EUkO/gK8meSDJjla7sKqOArTtBa2+Gnh6YO5Uq71Akh1JJpNMTk9PL7FFSdLzVi5x/hVVdSTJBcC+JN+ZZWyG1GrYwKraBewC2Lhx49AxkqSFW9KVflUdadtjwBeYWa55JskqgLY91oZPAWsHpq8Bjizl+JKkhVl06Cd5ZZJXP78P/CbwCLAX2N6GbQfuaft7gW1JzkpyEbAeuH+xx5ckLdxSlncuBL6Q5Pn3+fuq+kqSbwN7klwLPAW8G6CqDibZAzwKHAeur6oTS+pekrQgiw79qvou8MYh9R8AV77InJ3AzsUeU5K0NH4jV5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdWTluBt4KVp3w5fGduwnb3r72I4t6fQ38iv9JFuSPJbkcJIbRn18SerZSEM/yQrgr4HfAi4G3pPk4lH2IEk9G/XyzibgcFV9FyDJ3cBW4NER9/GSNa6lpR6Xlca5jDcu4/p19r/18hl16K8Gnh54PgX8ysmDkuwAdrSnP0ny2Ah6W6zzge+Pu4llsuhzyZ8tcydL91L5dTmtzmOJv86n1bks0Sk/l2X4M/Vzw4qjDv0MqdULClW7gF2nvp2lSzJZVRvH3cdy8FxOPy+V8wDP5XQx6g9yp4C1A8/XAEdG3IMkdWvUof9tYH2Si5L8LLAN2DviHiSpWyNd3qmq40neD/wDsAK4raoOjrKHU+CMWIaaJ8/l9PNSOQ/wXE4LqXrBkrok6SXKH8MgSR0x9CWpI4b+IiVZm+RrSQ4lOZjkA+PuaSmSrEjyL0m+OO5eliLJa5J8Nsl32q/Nr467p8VK8gft99YjSe5K8vJx9zRfSW5LcizJIwO185LsS/J42547zh7n60XO5c/b77GHk3whyWvG2OKCGPqLdxz4UFW9AdgMXH+G/0iJDwCHxt3EMvhL4CtV9YvAGzlDzynJauD3gY1VdSkzNz5sG29XC3I7sOWk2g3A/qpaD+xvz88Et/PCc9kHXFpVvwT8G3DjqJtaLEN/karqaFU92PafZSZcVo+3q8VJsgZ4O/DpcfeyFEnOAX4d+AxAVf1PVf1orE0tzUrg7CQrgVdwBn2npaq+DvzwpPJWYHfb3w1cPcqeFmvYuVTVV6vqeHv6TWa+c3RGMPSXQZJ1wGXAt8bcymL9BfCHwE/H3MdS/TwwDfxtW6r6dJJXjrupxaiq7wEfA54CjgL/WVVfHW9XS3ZhVR2FmYsm4IIx97Ncfhe4d9xNzJehv0RJXgV8DvhgVf143P0sVJJ3AMeq6oFx97IMVgK/DNxaVZcB/8WZs4Tw/7T17q3ARcBrgVcm+Z3xdqWTJfljZpZ67xx3L/Nl6C9BkpcxE/h3VtXnx93PIl0BvDPJk8DdwJuT/N14W1q0KWCqqp7/F9dnmflL4Ez0FuCJqpquqv8FPg/82ph7WqpnkqwCaNtjY+5nSZJsB94B/HadQV94MvQXKUmYWTs+VFWfGHc/i1VVN1bVmqpax8wHhf9UVWfkFWVV/QfwdJLXt9KVnLk/tvspYHOSV7Tfa1dyhn4oPWAvsL3tbwfuGWMvS5JkC/BHwDur6r/H3c9CGPqLdwXwXmaujB9qj7eNuynxe8CdSR4GNgB/Ot52Fqf9a+WzwIPAAWb+rJ4xX/1PchdwH/D6JFNJrgVuAt6a5HHgre35ae9FzuWvgFcD+9qf/b8Za5ML4I9hkKSOeKUvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JH/g/sOJAZeKAfMAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_vars_per_cell = (out_donors_df>t).sum(axis=1)\n",
    "#remove 0s\n",
    "plt.hist(num_vars_per_cell[num_vars_per_cell!=0])\n",
    "#plt.xscale('log')"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
