{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e64b4b89",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Snakemake Script. Don't need to import snakemake.\"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "from os.path import dirname, join\n",
    "import numpy as np\n",
    "#import click\n",
    "#import snakemake\n",
    "from src.utils.data_io import af_to_vireo\n",
    "from icecream import ic\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17ff0fcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cells_meta = \"/mnt/md0/isshamie/Projects/Mito_Trace/output/pipeline/v03/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex/cells_meta.tsv\"\n",
    "# sample = [\"Input\",\"Control\",Flt3l\"]\n",
    "# ref_mt = \"/data/Mito_Trace/data/external/chrM_refAllele.txt\"\n",
    "\n",
    "# cov = [\"/data/Mito_Trace/output/pipeline/v03/CHIP_b1/MTBlacklist_A2/data/Input/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/Input.coverage.txt\",\n",
    "#        \"/data/Mito_Trace/output/pipeline/v03/CHIP_b1/MTBlacklist_A2/data/Control/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/Control.coverage.txt\",\n",
    "#        \"/data/Mito_Trace/output/pipeline/v03/CHIP_b1/MTBlacklist_A2/data/Flt3l/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/Flt3l.coverage.txt\"]\n",
    "\n",
    "# output_fs = [\"/mnt/md0/isshamie/Projects/Mito_Trace/output/pipeline/v03/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex/clones_init/donor0\",\n",
    "#             \"/mnt/md0/isshamie/Projects/Mito_Trace/output/pipeline/v03/CHIP_b1/MTBlacklist_A2/data/merged/MT/cellr_True/numread_200/filters/minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20/mgatk/vireoIn/multiplex/clones_init/donor1\"]\n",
    "\n",
    "# var_type=\"init\"\n",
    "\n",
    "#minC10_minR50_topN0_hetT0.001_hetC10_hetCount5_bq20\n",
    "cov = snakemake.input.cov\n",
    "output_fs = snakemake.output.af\n",
    "cells_meta = snakemake.params.cells_meta\n",
    "sample = snakemake.params.sample\n",
    "ref_mt = snakemake.params.ref_mt\n",
    "var_type=snakemake.params.var_type\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aada4df",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cells_meta = pd.read_csv(cells_meta , sep='\\t')\n",
    "#sample = sample.split(',')\n",
    "ref_fa = pd.read_csv(ref_mt, sep='\\t', header=None, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a00738b3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for d, curr_donor in cells_meta.groupby('donor'):\n",
    "    d = int(d)\n",
    "    ic('donor', d)\n",
    "    vars_conds = []\n",
    "    af_donor_conds = []\n",
    "    dp_donor_conds = []\n",
    "\n",
    "    # Loop through each condition and filter cells for the donor ones\n",
    "    for ind, curr_cov_f in enumerate(cov): #enumerate(snakemake.input.cov):\n",
    "        \n",
    "        ## AF\n",
    "        af = pd.read_csv(join(dirname(curr_cov_f), \"af_by_cell.tsv\"), sep='\\t',\n",
    "                         header=0, index_col=0) #cells-by-variants\n",
    "        condition = sample[ind]\n",
    "        ic(condition)\n",
    "        \n",
    "        af.index = af.index + \"_\" + condition\n",
    "#         print('af')\n",
    "#         print(af.head())\n",
    "        curr_samp_donor_df = curr_donor[curr_donor[\"condition\"] == condition]  # Condition-donors\n",
    "        af_donor_samp = af.loc[af.index.isin(curr_samp_donor_df[\"ID\"].values)]\n",
    "        ic(af_donor_samp.shape)\n",
    "        af_donor_samp = af_donor_samp.loc[((af_donor_samp > 0).sum(axis=1)>10), ((af_donor_samp > 0).sum(axis=0)>10)] #af_donor_samp.loc[(af_donor_samp.sum(axis=1) > 0), (af_donor_samp.sum(axis=0) > 0)]\n",
    "        vars_conds.append(set(af_donor_samp.columns))\n",
    "        af_donor_conds.append(af_donor_samp)\n",
    "        ic(curr_cov_f)\n",
    "        ic(af_donor_samp.shape)\n",
    "        ## Coverage\n",
    "        curr_samp_cov = pd.read_csv(curr_cov_f,\n",
    "                                    header=None)  # load sample coverage\n",
    "        curr_samp_cov[1] = curr_samp_cov[1] + \"_\" + sample[ind]\n",
    "        curr_samp_cov = curr_samp_cov.loc[\n",
    "            curr_samp_cov[1].isin(curr_samp_donor_df[\"ID\"])]\n",
    "        curr_samp_cov = curr_samp_cov.rename({1: \"Cell\", 0: \"Position\", 2: \"Value\"}, axis=1).pivot(index=\"Cell\",\n",
    "                                                                                                  columns=\"Position\",\n",
    "                                                                                                  values=\"Value\")\n",
    "        dp_donor_samp = pd.DataFrame(index=af_donor_samp.index, columns=af_donor_samp.columns)\n",
    "\n",
    "        def fill_dp(x,y):\n",
    "            #print('x', x)\n",
    "            #print('y', y.head())\n",
    "            return y.loc[y.index.isin(x.index), int(x.name[:-1])].values\n",
    "\n",
    "        dp_donor_samp = dp_donor_samp.apply(fill_dp, args=(curr_samp_cov,))\n",
    "\n",
    "        print('null dp', dp_donor_samp.isnull().sum().sum())\n",
    "        #dp_donor_samp = dp_donor_samp.loc[dp_donor_samp.index.isin(af_donor_samp.index), dp_donor_samp.columns.isin([int(x[:-1] ) for x in af_donor_samp.columns])]\n",
    "        dp_donor_conds.append(dp_donor_samp)\n",
    "\n",
    "    # Get overlap variants and remove cells/vars with 0s\n",
    "    if var_type == \"simpleUnion\":\n",
    "        join_type = \"outer\"\n",
    "        overlap_vars = vars_conds[0].union(*vars_conds)\n",
    "    else:\n",
    "        join_type=\"inner\"\n",
    "        overlap_vars = vars_conds[0].intersection(*vars_conds)\n",
    "    print(join_type)\n",
    "    #overlap_vars = vars_conds[0].union(*vars_conds)\n",
    "    print('overlap_vars', len(overlap_vars))\n",
    "    af_donor_df = pd.concat(af_donor_conds, join=join_type, axis=0,\n",
    "                                                ignore_index=False)\n",
    "    print('af_donor_df', af_donor_df.shape)\n",
    "    af_donor_df = af_donor_df.loc[:, overlap_vars]\n",
    "    print('af_donor_df after using overlap vars', af_donor_df.shape)\n",
    "    af_donor_df = af_donor_df.loc[(af_donor_df.sum(axis=1)>0), (af_donor_df.sum(axis=0)>0)]\n",
    "    print('af_donor_df after 0 removal', af_donor_df.shape)\n",
    "\n",
    "    #print('af_donor_df')\n",
    "    #print(af_donor_df.head())\n",
    "    dp_donor_df = pd.concat(dp_donor_conds, join=join_type, axis=0,\n",
    "                                                ignore_index=False).fillna(0)\n",
    "    dp_donor_df = dp_donor_df.loc[dp_donor_df.index.isin(af_donor_df.index), dp_donor_df.columns.isin(af_donor_df.columns)]\n",
    "    #print('dp_donor_df')\n",
    "\n",
    "    # Save allele + depth matrix + cells_meta\n",
    "    af_donor_df.fillna(0).to_csv(join(dirname(output_fs[int(d)]), \"af.tsv\"),\n",
    "                       sep='\\t', index=True)\n",
    "    dp_donor_df.fillna(0).to_csv(join(dirname(output_fs[int(d)]), \"dp.tsv\"),\n",
    "                       sep='\\t', index=True)\n",
    "\n",
    "    curr_cells_meta = cells_meta[cells_meta[\"ID\"].isin(af_donor_df.index)].copy()\n",
    "    curr_cells_meta.to_csv(join(dirname(output_fs[int(d)]), \"cells_meta.tsv\"),\n",
    "                          sep='\\t', index=False)\n",
    "\n",
    "    ##  Create VireoIn\n",
    "    \n",
    "    # Requires AD sparse matrix\n",
    "    af_donor_df = af_donor_df.fillna(0)\n",
    "    dp_donor_df = dp_donor_df.fillna(0)\n",
    "    af_to_vireo(af_donor_df.transpose(), dp_donor_df.transpose(),\n",
    "                outdir=dirname(output_fs[int(d)]),\n",
    "                out_name=\"cellSNP\")\n",
    "\n",
    "    curr_vars = af_donor_df.columns\n",
    "\n",
    "    af_meta = pd.DataFrame({\"POS\": [int(x[:-1]) for x in curr_vars],\n",
    "                                                    \"ALT\": [x[-1] for x in curr_vars],\n",
    "                                                    \"index\": np.arange(1,len(curr_vars)+1),\n",
    "                                                    \"REF\": [f\"{ref_fa.loc[int(x[:-1]),1]}{x[:-1]}\" for x in curr_vars]\n",
    "                                                }\n",
    "                                            )\n",
    "    af_meta[\"#CHROM\"] = \"MT\"\n",
    "    af_meta[[\"#CHROM\", \"POS\", \"REF\",\"ALT\", \"index\"]].to_csv(join(dirname(output_fs[d]), \"cellSNP.base.vcf\"),\n",
    "                   sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86add433",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(dp_donor_df.shape)\n",
    "# dp_donor_df.loc[(dp_donor_df.sum(axis=1)>0), (dp_donor_df.sum(axis=0)>0)].shape\n",
    "# (dp_donor_df.sum(axis=1) > 1000).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6876dd2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(af_donor_df.shape)\n",
    "# af_donor_df.loc[(af_donor_df.sum(axis=1)>0), (af_donor_df.sum(axis=0)>0)].shape\n",
    "# (af_donor_df.sum(axis=1) > 0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7977f27",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
