{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.787609Z",
     "start_time": "2020-02-15T21:37:05.205202Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import gzip\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "from tqdm import tqdm\n",
    "from os.path import join\n",
    "from Bio import SeqIO\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.794826Z",
     "start_time": "2020-02-15T21:37:05.790084Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from alignment import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.802640Z",
     "start_time": "2020-02-15T21:37:05.797093Z"
    }
   },
   "outputs": [],
   "source": [
    "data_path = '/data2/genome/human_GRCh38/ncbi/'\n",
    "mito_chrom = 'NC_012920.1'\n",
    "\n",
    "window =150 #bps covered for sequencing\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load genomic info and extract mitochondria information\n",
    "- fasta sequence\n",
    "- gff file\n",
    "- genes as fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.822859Z",
     "start_time": "2020-02-15T21:37:05.804336Z"
    }
   },
   "outputs": [],
   "source": [
    "mt_fasta = join(data_path, 'GCF_000001405.38_GRCh38.p12_genomic.fna')\n",
    "\n",
    "if not os.path.exists(\"MT.fasta\"):\n",
    "    ref_seq = SeqIO.index(mt_fasta, \"fasta\")\n",
    "\n",
    "    mt_seq = ref_seq[mito_chrom]\n",
    "    with open(\"MT.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(mt_seq, output_handle, \"fasta\")\n",
    "    mt_seq_biopython = mt_seq.seq\n",
    "    mt_seq = str(mt_seq.seq)\n",
    "else:\n",
    "    mt_seq = str(list(SeqIO.parse(\"MT.fasta\", format='fasta'))[0].seq)\n",
    "    mt_seq_biopython = list(SeqIO.parse(\"MT.fasta\", format='fasta'))[0].seq\n",
    "    #mt_seq_biopython_reverse = mt_seq_biopython.reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.857820Z",
     "start_time": "2020-02-15T21:37:05.824830Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading line 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>id1959760</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>16569</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene60922</td>\n",
       "      <td>gene</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna171173</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>id1959761</td>\n",
       "      <td>exon</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene60923</td>\n",
       "      <td>gene</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>id1959784</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>gene60958</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>rna171196</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>id1959785</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>id1959786</td>\n",
       "      <td>D_loop</td>\n",
       "      <td>16024</td>\n",
       "      <td>17145</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Id    Type  Start    End Strand     Location\n",
       "0    id1959760  region      1  16569      +  NC_012920.1\n",
       "1    gene60922    gene    577    647      +  NC_012920.1\n",
       "2    rna171173    tRNA    577    647      +  NC_012920.1\n",
       "3    id1959761    exon    577    647      +  NC_012920.1\n",
       "4    gene60923    gene    648   1601      +  NC_012920.1\n",
       "..         ...     ...    ...    ...    ...          ...\n",
       "96   id1959784    exon  15888  15953      +  NC_012920.1\n",
       "97   gene60958    gene  15956  16023      -  NC_012920.1\n",
       "98   rna171196    tRNA  15956  16023      -  NC_012920.1\n",
       "99   id1959785    exon  15956  16023      -  NC_012920.1\n",
       "100  id1959786  D_loop  16024  17145      -  NC_012920.1\n",
       "\n",
       "[101 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_gff_filename = join(data_path,'GRCh38_latest_genomic.gff')\n",
    "gff_filename = \"MT.gff\"\n",
    "if not os.path.exists(gff_filename):\n",
    "    cmd = f\"grep {mito_chrom} {full_gff_filename} > {gff_filename}\"\n",
    "    print(cmd)\n",
    "    os.system(cmd)\n",
    "\n",
    "gff_info = read_gff(gff_filename, file_format=None)\n",
    "## Analyzing GFF\n",
    "gff_info.Type.unique()\n",
    "\n",
    "gff_info = gff_info.loc[gff_info.Location == mito_chrom].reset_index(drop=True)\n",
    "gff_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.863643Z",
     "start_time": "2020-02-15T21:37:05.859735Z"
    }
   },
   "outputs": [],
   "source": [
    "#cmd = f\"bedtools getfasta -fi MT.fasta -bed {gff_filename} -name -fo MT.genes.fasta\"\n",
    "if not os.path.exists('MT.genes.fasta'):\n",
    "    cmd = 'gff2bed < MT.gff > MT.bed'\n",
    "    print(cmd)\n",
    "    os.system(cmd)\n",
    "    cmd = f\"bedtools getfasta -fi MT.fasta -bed MT.bed -name -fo MT.genes.fasta\"\n",
    "    print(cmd)\n",
    "    os.system(cmd) #gff_filename = data_path + 'MT.gff'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T20:44:57.459207Z",
     "start_time": "2020-02-15T20:44:28.280266Z"
    },
    "deletable": false,
    "editable": false,
    "run_control": {
     "frozen": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /data2/genome/human_GRCh38/ncbi/GCF_000001405.38_GRCh38.p12_genomic.fna\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTGCGAGACGCTGGAGCCGGAGCACCCTATGTCGCAGTATCTGTCTTTGATTCCTGCCTCATCCTATTATTTATCGCACCTACGTTCAATATTACAGGCGAACATACTTACTAAAGTGTGTTAATTAATTAATGCTTGTAGGACATAATAATAACAATTGAATGTCTGCACAGCCACTTTCCACACAGACATCATAACAAAAAATTTCCACCAAACCCCCCCTCCCCCGCTTCTGGCCACAGCACTTAAACACATCTCTGCCAAACCCCAAAAACAAAGAACCCTAACACCAGCCTAACCAGATTTCAAATTTTATCTTTTGGCGGTATGCACTTTTAACAGTCACCCCCCAACTAACACATTATTTTCCCCTCCCACTCCCATACTACTAATCTCATCAATACAACCCCCGCCCATCCTACCCAGCACACACACACCGCTGCTAACCCCATACCCCGAACCAACCAAACCCCAAAGACACCCCCCACAGTTTATGTAGCTTACCTCCTCAAAGCAATACACTGAAAATGTTTAGACGGGCTCACATCACCCCATAAACAAATAGGTTTGGTCCTAGCCTTTCTATTAGCTCTTAGTAAGATTACACATGCAAGCATCCCCGTTCCAGTGAGTTCACCCTCTAAATCACCACGATCAAAAGGAACAAGCATCAAGCACGCAGCAATGCAGCTCAAAACGCTTAGCCTAGCCACACCCCCACGGGAAACAGCAGTGATTAACCTTTAGCAATAAACGAAAGTTTAACTAAGCTATACTAACCCCAGGGTTGGTCAATTTCGTGCCAGCCACCGCGGTCACACGATTAACCCAAGTCAATAGAAGCCGGCGTAAAGAGTGTTTTAGATCACCCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCCAGTTGACACAAAATAGACTACGAAAGTGGCTTTAACATATCTGAACACACAATAGCTAAGACCCAAACTGGGATTAGATACCCCACTATGCTTAGCCCTAAACCTCAACAGTTAAATCAACAAAACTGCTCGCCAGAACACTACGAGCCACAGCTTAAAACTCAAAGGACCTGGCGGTGCTTCATATCCCTCTAGAGGAGCCTGTTCTGTAATCGATAAACCCCGATCAACCTCACCACCTCTTGCTCAGCCTATATACCGCCATCTTCAGCAAACCCTGATGAAGGCTACAAAGTAAGCGCAAGTACCCACGTAAAGACGTTAGGTCAAGGTGTAGCCCATGAGGTGGCAAGAAATGGGCTACATTTTCTACCCCAGAAAACTACGATAGCCCTTATGAAACTTAAGGGTCGAAGGTGGATTTAGCAGTAAACTAAGAGTAGAGTGCTTAGTTGAACAGGGCCCTGAAGCGCGTACACACCGCCCGTCACCCTCCTCAAGTATACTTCAAAGGACATTTAACTAAAACCCCTACGCATTTATATAGAGGAGACAAGTCGTAACATGGTAAGTGTACTGGAAAGTGCACTTGGACGAACCAGAGTGTAGCTTAACACAAAGCACCCAACTTACACTTAGGAGATTTCAACTTAACTTGACCGCTCTGAGCTAAACCTAGCCCCAAACCCACTCCACCTTACTACCAGACAACCTTAGCCAAACCATTTACCCAAATAAAGTATAGGCGATAGAAATTGAAACCTGGCGCAATAGATATAGTACCGCAAGGGAAAGATGAAAAATTATAACCAAGCATAATATAGCAAGGACTAACCCCTATACCTTCTGCATAATGAATTAACTAGAAATAACTTTGCAAGGAGAGCCAAAGCTAAGACCCCCGAAACCAGACGAGCTACCTAAGAACAGCTAAAAGAGCACACCCGTCTATGTAGCAAAATAGTGGGAAGATTTATAGGTAGAGGCGACAAACCTACCGAGCCTGGTGATAGCTGGTTGTCCAAGATAGAATCTTAGTTCAACTTTAAATTTGCCCACAGAACCCTCTAAATCCCCTTGTAAATTTAACTGTTAGTCCAAAGAGGAACAGCTCTTTGGACACTAGGAAAAAACCTTGTAGAGAGAGTAAAAAATTTAACACCCATAGTAGGCCTAAAAGCAGCCACCAATTAAGAAAGCGTTCAAGCTCAACACCCACTACCTAAAAAATCCCAAACATATAACTGAACTCCTCACACCCAATTGGACCAATCTATCACCCTATAGAAGAACTAATGTTAGTATAAGTAACATGAAAACATTCTCCTCCGCATAAGCCTGCGTCAGATTAAAACACTGAACTGACAATTAACAGCCCAATATCTACAATCAACCAACAAGTCATTATTACCCTCACTGTCAACCCAACACAGGCATGCTCATAAGGAAAGGTTAAAAAAAGTAAAAGGAACTCGGCAAATCTTACCCCGCCTGTTTACCAAAAACATCACCTCTAGCATCACCAGTATTAGAGGCACCGCCTGCCCAGTGACACATGTTTAACGGCCGCGGTACCCTAACCGTGCAAAGGTAGCATAATCACTTGTTCCTTAAATAGGGACCTGTATGAATGGCTCCACGAGGGTTCAGCTGTCTCTTACTTTTAACCAGTGAAATTGACCTGCCCGTGAAGAGGCGGGCATAACACAGCAAGACGAGAAGACCCTATGGAGCTTTAATTTATTAATGCAAACAGTACCTAACAAACCCACAGGTCCTAAACTACCAAACCTGCATTAAAAATTTCGGTTGGGGCGACCTCGGAGCAGAACCCAACCTCCGAGCAGTACATGCTAAGACTTCACCAGTCAAAGCGAACTACTATACTCAATTGATCCAATAACTTGACCAACGGAACAAGTTACCCTAGGGATAACAGCGCAATCCTATTCTAGAGTCCATATCAACAATAGGGTTTACGACCTCGATGTTGGATCAGGACATCCCGATGGTGCAGCCGCTATTAAAGGTTCGTTTGTTCAACGATTAAAGTCCTACGTGATCTGAGTTCAGACCGGAGTAATCCAGGTCGGTTTCTATCTACNTTCAAATTCCTCCCTGTACGAAAGGACAAGAGAAATAAGGCCTACTTCACAAAGCGCCTTCCCCCGTAAATGATATCATCTCAACTTAGTATTATACCCACACCCACCCAAGAACAGGGTTTGTTAAGATGGCAGAGCCCGGTAATCGCATAAAACTTAAAACTTTACAGTCAGAGGTTCAATTCCTCTTCTTAACAACATACCCATGGCCAACCTCCTACTCCTCATTGTACCCATTCTAATCGCAATGGCATTCCTAATGCTTACCGAACGAAAAATTCTAGGCTATATACAACTACGCAAAGGCCCCAACGTTGTAGGCCCCTACGGGCTACTACAACCCTTCGCTGACGCCATAAAACTCTTCACCAAAGAGCCCCTAAAACCCGCCACATCTACCATCACCCTCTACATCACCGCCCCGACCTTAGCTCTCACCATCGCTCTTCTACTATGAACCCCCCTCCCCATACCCAACCCCCTGGTCAACCTCAACCTAGGCCTCCTATTTATTCTAGCCACCTCTAGCCTAGCCGTTTACTCAATCCTCTGATCAGGGTGAGCATCAAACTCAAACTACGCCCTGATCGGCGCACTGCGAGCAGTAGCCCAAACAATCTCATATGAAGTCACCCTAGCCATCATTCTACTATCAACATTACTAATAAGTGGCTCCTTTAACCTCTCCACCCTTATCACAACACAAGAACACCTCTGATTACTCCTGCCATCATGACCCTTGGCCATAATATGATTTATCTCCACACTAGCAGAGACCAACCGAACCCCCTTCGACCTTGCCGAAGGGGAGTCCGAACTAGTCTCAGGCTTCAACATCGAATACGCCGCAGGCCCCTTCGCCCTATTCTTCATAGCCGAATACACAAACATTATTATAATAAACACCCTCACCACTACAATCTTCCTAGGAACAACATATGACGCACTCTCCCCTGAACTCTACACAACATATTTTGTCACCAAGACCCTACTTCTAACCTCCCTGTTCTTATGAATTCGAACAGCATACCCCCGATTCCGCTACGACCAACTCATACACCTCCTATGAAAAAACTTCCTACCACTCACCCTAGCATTACTTATATGATATGTCTCCATACCCATTACAATCTCCAGCATTCCCCCTCAAACCTAAGAAATATGTCTGATAAAAGAGTTACTTTGATAGAGTAAATAATAGGAGCTTAAACCCCCTTATTTCTAGGACTATGAGAATCGAACCCATCCCTGAGAATCCAAAATTCTCCGTGCCACCTATCACACCCCATCCTAAAGTAAGGTCAGCTAAATAAGCTATCGGGCCCATACCCCGAAAATGTTGGTTATACCCTTCCCGTACTAATTAATCCCCTGGCCCAACCCGTCATCTACTCTACCATCTTTGCAGGCACACTCATCACAGCGCTAAGCTCGCACTGATTTTTTACCTGAGTAGGCCTAGAAATAAACATGCTAGCTTTTATTCCAGTTCTAACCAAAAAAATAAACCCTCGTTCCACAGAAGCTGCCATCAAGTATTTCCTCACGCAAGCAACCGCATCCATAATCCTTCTAATAGCTATCCTCTTCAACAATATACTCTCCGGACAATGAACCATAACCAATACTACCAATCAATACTCATCATTAATAATCATAATAGCTATAGCAATAAAACTAGGAATAGCCCCCTTTCACTTCTGAGTCCCAGAGGTTACCCAAGGCACCCCTCTGACATCCGGCCTGCTTCTTCTCACATGACAAAAACTAGCCCCCATCTCAATCATATACCAAATCTCTCCCTCACTAAACGTAAGCCTTCTCCTCACTCTCTCAATCTTATCCATCATAGCAGGCAGTTGAGGTGGATTAAACCAAACCCAGCTACGCAAAATCTTAGCATACTCCTCAATTACCCACATAGGATGAATAATAGCAGTTCTACCGTACAACCCTAACATAACCATTCTTAATTTAACTATTTATATTATCCTAACTACTACCGCATTCCTACTACTCAACTTAAACTCCAGCACCACGACCCTACTACTATCTCGCACCTGAAACAAGCTAACATGACTAACACCCTTAATTCCATCCACCCTCCTCTCCCTAGGAGGCCTGCCCCCGCTAACCGGCTTTTTGCCCAAATGGGCCATTATCGAAGAATTCACAAAAAACAATAGCCTCATCATCCCCACCATCATAGCCACCATCACCCTCCTTAACCTCTACTTCTACCTACGCCTAATCTACTCCACCTCAATCACACTACTCCCCATATCTAACAACGTAAAAATAAAATGACAGTTTGAACATACAAAACCCACCCCATTCCTCCCCACACTCATCGCCCTTACCACGCTACTCCTACCTATCTCCCCTTTTATACTAATAATCTTATAGAAATTTAGGTTAAATACAGACCAAGAGCCTTCAAAGCCCTCAGTAAGTTGCAATACTTAATTTCTGTAACAGCTAAGGACTGCAAAACCCCACTCTGCATCAACTGAACGCAAATCAGCCACTTTAATTAAGCTAAGCCCTTACTAGACCAATGGGACTTAAACCCACAAACACTTAGTTAACAGCTAAGCACCCTAATCAACTGGCTTCAATCTACTTCTCCCGCCGCCGGGAAAAAAGGCGGGAGAAGCCCCGGCAGGTTTGAAGCTGCTTCTTCGAATTTGCAATTCAATATGAAAATCACCTCGGAGCTGGTAAAAAGAGGCCTAACCCCTGTCTTTAGATTTACAGTCCAATGCTTCACTCAGCCATTTTACCTCACCCCCACTGATGTTCGCCGACCGTTGACTATTCTCTACAAACCACAAAGACATTGGAACACTATACCTATTATTCGGCGCATGAGCTGGAGTCCTAGGCACAGCTCTAAGCCTCCTTATTCGAGCCGAGCTGGGCCAGCCAGGCAACCTTCTAGGTAACGACCACATCTACAACGTTATCGTCACAGCCCATGCATTTGTAATAATCTTCTTCATAGTAATACCCATCATAATCGGAGGCTTTGGCAACTGACTAGTTCCCCTAATAATCGGTGCCCCCGATATGGCGTTTCCCCGCATAAACAACATAAGCTTCTGACTCTTACCTCCCTCTCTCCTACTCCTGCTCGCATCTGCTATAGTGGAGGCCGGAGCAGGAACAGGTTGAACAGTCTACCCTCCCTTAGCAGGGAACTACTCCCACCCTGGAGCCTCCGTAGACCTAACCATCTTCTCCTTACACCTAGCAGGTGTCTCCTCTATCTTAGGGGCCATCAATTTCATCACAACAATTATCAATATAAAACCCCCTGCCATAACCCAATACCAAACGCCCCTCTTCGTCTGATCCGTCCTAATCACAGCAGTCCTACTTCTCCTATCTCTCCCAGTCCTAGCTGCTGGCATCACTATACTACTAACAGACCGCAACCTCAACACCACCTTCTTCGACCCCGCCGGAGGAGGAGACCCCATTCTATACCAACACCTATTCTGATTTTTCGGTCACCCTGAAGTTTATATTCTTATCCTACCAGGCTTCGGAATAATCTCCCATATTGTAACTTACTACTCCGGAAAAAAAGAACCATTTGGATACATAGGTATGGTCTGAGCTATGATATCAATTGGCTTCCTAGGGTTTATCGTGTGAGCACACCATATATTTACAGTAGGAATAGACGTAGACACACGAGCATATTTCACCTCCGCTACCATAATCATCGCTATCCCCACCGGCGTCAAAGTATTTAGCTGACTCGCCACACTCCACGGAAGCAATATGAAATGATCTGCTGCAGTGCTCTGAGCCCTAGGATTCATCTTTCTTTTCACCGTAGGTGGCCTGACTGGCATTGTATTAGCAAACTCATCACTAGACATCGTACTACACGACACGTACTACGTTGTAGCCCACTTCCACTATGTCCTATCAATAGGAGCTGTATTTGCCATCATAGGAGGCTTCATTCACTGATTTCCCCTATTCTCAGGCTACACCCTAGACCAAACCTACGCCAAAATCCATTTCACTATCATATTCATCGGCGTAAATCTAACTTTCTTCCCACAACACTTTCTCGGCCTATCCGGAATGCCCCGACGTTACTCGGACTACCCCGATGCATACACCACATGAAACATCCTATCATCTGTAGGCTCATTCATTTCTCTAACAGCAGTAATATTAATAATTTTCATGATTTGAGAAGCCTTCGCTTCGAAGCGAAAAGTCCTAATAGTAGAAGAACCCTCCATAAACCTGGAGTGACTATATGGATGCCCCCCACCCTACCACACATTCGAAGAACCCGTATACATAAAATCTAGACAAAAAAGGAAGGAATCGAACCCCCCAAAGCTGGTTTCAAGCCAACCCCATGGCCTCCATGACTTTTTCAAAAAGGTATTAGAAAAACCATTTCATAACTTTGTCAAAGTTAAATTATAGGCTAAATCCTATATATCTTAATGGCACATGCAGCGCAAGTAGGTCTACAAGACGCTACTTCCCCTATCATAGAAGAGCTTATCACCTTTCATGATCACGCCCTCATAATCATTTTCCTTATCTGCTTCCTAGTCCTGTATGCCCTTTTCCTAACACTCACAACAAAACTAACTAATACTAACATCTCAGACGCTCAGGAAATAGAAACCGTCTGAACTATCCTGCCCGCCATCATCCTAGTCCTCATCGCCCTCCCATCCCTACGCATCCTTTACATAACAGACGAGGTCAACGATCCCTCCCTTACCATCAAATCAATTGGCCACCAATGGTACTGAACCTACGAGTACACCGACTACGGCGGACTAATCTTCAACTCCTACATACTTCCCCCATTATTCCTAGAACCAGGCGACCTGCGACTCCTTGACGTTGACAATCGAGTAGTACTCCCGATTGAAGCCCCCATTCGTATAATAATTACATCACAAGACGTCTTGCACTCATGAGCTGTCCCCACATTAGGCTTAAAAACAGATGCAATTCCCGGACGTCTAAACCAAACCACTTTCACCGCTACACGACCGGGGGTATACTACGGTCAATGCTCTGAAATCTGTGGAGCAAACCACAGTTTCATGCCCATCGTCCTAGAATTAATTCCCCTAAAAATCTTTGAAATAGGGCCCGTATTTACCCTATAGCACCCCCTCTACCCCCTCTAGAGCCCACTGTAAAGCTAACTTAGCATTAACCTTTTAAGTTAAAGATTAAGAGAACCAACACCTCTTTACAGTGAAATGCCCCAACTAAATACTACCGTATGGCCCACCATAATTACCCCCATACTCCTTACACTATTCCTCATCACCCAACTAAAAATATTAAACACAAACTACCACCTACCTCCCTCACCAAAGCCCATAAAAATAAAAAATTATAACAAACCCTGAGAACCAAAATGAACGAAAATCTGTTCGCTTCATTCATTGCCCCCACAATCCTAGGCCTACCCGCCGCAGTACTGATCATTCTATTTCCCCCTCTATTGATCCCCACCTCCAAATATCTCATCAACAACCGACTAATCACCACCCAACAATGACTAATCAAACTAACCTCAAAACAAATGATAACCATACACAACACTAAAGGACGAACCTGATCTCTTATACTAGTATCCTTAATCATTTTTATTGCCACAACTAACCTCCTCGGACTCCTGCCTCACTCATTTACACCAACCACCCAACTATCTATAAACCTAGCCATGGCCATCCCCTTATGAGCGGGCACAGTGATTATAGGCTTTCGCTCTAAGATTAAAAATGCCCTAGCCCACTTCTTACCACAAGGCACACCTACACCCCTTATCCCCATACTAGTTATTATCGAAACCATCAGCCTACTCATTCAACCAATAGCCCTGGCCGTACGCCTAACCGCTAACATTACTGCAGGCCACCTACTCATGCACCTAATTGGAAGCGCCACCCTAGCAATATCAACCATTAACCTTCCCTCTACACTTATCATCTTCACAATTCTAATTCTACTGACTATCCTAGAAATCGCTGTCGCCTTAATCCAAGCCTACGTTTTCACACTTCTAGTAAGCCTCTACCTGCACGACAACACATAATGACCCACCAATCACATGCCTATCATATAGTAAAACCCAGCCCATGACCCCTAACAGGGGCCCTCTCAGCCCTCCTAATGACCTCCGGCCTAGCCATGTGATTTCACTTCCACTCCATAACGCTCCTCATACTAGGCCTACTAACCAACACACTAACCATATACCAATGATGGCGCGATGTAACACGAGAAAGCACATACCAAGGCCACCACACACCACCTGTCCAAAAAGGCCTTCGATACGGGATAATCCTATTTATTACCTCAGAAGTTTTTTTCTTCGCAGGATTTTTCTGAGCCTTTTACCACTCCAGCCTAGCCCCTACCCCCCAATTAGGAGGGCACTGGCCCCCAACAGGCATCACCCCGCTAAATCCCCTAGAAGTCCCACTCCTAAACACATCCGTATTACTCGCATCAGGAGTATCAATCACCTGAGCTCACCATAGTCTAATAGAAAACAACCGAAACCAAATAATTCAAGCACTGCTTATTACAATTTTACTGGGTCTCTATTTTACCCTCCTACAAGCCTCAGAGTACTTCGAGTCTCCCTTCACCATTTCCGACGGCATCTACGGCTCAACATTTTTTGTAGCCACAGGCTTCCACGGACTTCACGTCATTATTGGCTCAACTTTCCTCACTATCTGCTTCATCCGCCAACTAATATTTCACTTTACATCCAAACATCACTTTGGCTTCGAAGCCGCCGCCTGATACTGGCATTTTGTAGATGTGGTTTGACTATTTCTGTATGTCTCCATCTATTGATGAGGGTCTTACTCTTTTAGTATAAATAGTACCGTTAACTTCCAATTAACTAGTTTTGACAACATTCAAAAAAGAGTAATAAACTTCGCCTTAATTTTAATAATCAACACCCTCCTAGCCTTACTACTAATAATTATTACATTTTGACTACCACAACTCAACGGCTACATAGAAAAATCCACCCCTTACGAGTGCGGCTTCGACCCTATATCCCCCGCCCGCGTCCCTTTCTCCATAAAATTCTTCTTAGTAGCTATTACCTTCTTATTATTTGATCTAGAAATTGCCCTCCTTTTACCCCTACCATGAGCCCTACAAACAACTAACCTGCCACTAATAGTTATGTCATCCCTCTTATTAATCATCATCCTAGCCCTAAGTCTGGCCTATGAGTGACTACAAAAAGGATTAGACTGAACCGAATTGGTATATAGTTTAAACAAAACGAATGATTTCGACTCATTAAATTATGATAATCATATTTACCAAATGCCCCTCATTTACATAAATATTATACTAGCATTTACCATCTCACTTCTAGGAATACTAGTATATCGCTCACACCTCATATCCTCCCTACTATGCCTAGAAGGAATAATACTATCGCTGTTCATTATAGCTACTCTCATAACCCTCAACACCCACTCCCTCTTAGCCAATATTGTGCCTATTGCCATACTAGTCTTTGCCGCCTGCGAAGCAGCGGTGGGCCTAGCCCTACTAGTCTCAATCTCCAACACATATGGCCTAGACTACGTACATAACCTAAACCTACTCCAATGCTAAAACTAATCGTCCCAACAATTATATTACTACCACTGACATGACTTTCCAAAAAACACATAATTTGAATCAACACAACCACCCACAGCCTAATTATTAGCATCATCCCTCTACTATTTTTTAACCAAATCAACAACAACCTATTTAGCTGTTCCCCAACCTTTTCCTCCGACCCCCTAACAACCCCCCTCCTAATACTAACTACCTGACTCCTACCCCTCACAATCATGGCAAGCCAACGCCACTTATCCAGTGAACCACTATCACGAAAAAAACTCTACCTCTCTATACTAATCTCCCTACAAATCTCCTTAATTATAACATTCACAGCCACAGAACTAATCATATTTTATATCTTCTTCGAAACCACACTTATCCCCACCTTGGCTATCATCACCCGATGAGGCAACCAGCCAGAACGCCTGAACGCAGGCACATACTTCCTATTCTACACCCTAGTAGGCTCCCTTCCCCTACTCATCGCACTAATTTACACTCACAACACCCTAGGCTCACTAAACATTCTACTACTCACTCTCACTGCCCAAGAACTATCAAACTCCTGAGCCAACAACTTAATATGACTAGCTTACACAATAGCTTTTATAGTAAAGATACCTCTTTACGGACTCCACTTATGACTCCCTAAAGCCCATGTCGAAGCCCCCATCGCTGGGTCAATAGTACTTGCCGCAGTACTCTTAAAACTAGGCGGCTATGGTATAATACGCCTCACACTCATTCTCAACCCCCTGACAAAACACATAGCCTACCCCTTCCTTGTACTATCCCTATGAGGCATAATTATAACAAGCTCCATCTGCCTACGACAAACAGACCTAAAATCGCTCATTGCATACTCTTCAATCAGCCACATAGCCCTCGTAGTAACAGCCATTCTCATCCAAACCCCCTGAAGCTTCACCGGCGCAGTCATTCTCATAATCGCCCACGGGCTTACATCCTCATTACTATTCTGCCTAGCAAACTCAAACTACGAACGCACTCACAGTCGCATCATAATCCTCTCTCAAGGACTTCAAACTCTACTCCCACTAATAGCTTTTTGATGACTTCTAGCAAGCCTCGCTAACCTCGCCTTACCCCCCACTATTAACCTACTGGGAGAACTCTCTGTGCTAGTAACCACGTTCTCCTGATCAAATATCACTCTCCTACTTACAGGACTCAACATACTAGTCACAGCCCTATACTCCCTCTACATATTTACCACAACACAATGGGGCTCACTCACCCACCACATTAACAACATAAAACCCTCATTCACACGAGAAAACACCCTCATGTTCATACACCTATCCCCCATTCTCCTCCTATCCCTCAACCCCGACATCATTACCGGGTTTTCCTCTTGTAAATATAGTTTAACCAAAACATCAGATTGTGAATCTGACAACAGAGGCTTACGACCCCTTATTTACCGAGAAAGCTCACAAGAACTGCTAACTCATGCCCCCATGTCTAACAACATGGCTTTCTCAACTTTTAAAGGATAACAGCTATCCATTGGTCTTAGGCCCCAAAAATTTTGGTGCAACTCCAAATAAAAGTAATAACCATGCACACTACTATAACCACCCTAACCCTGACTTCCCTAATTCCCCCCATCCTTACCACCCTCGTTAACCCTAACAAAAAAAACTCATACCCCCATTATGTAAAATCCATTGTCGCATCCACCTTTATTATCAGTCTCTTCCCCACAACAATATTCATGTGCCTAGACCAAGAAGTTATTATCTCGAACTGACACTGAGCCACAACCCAAACAACCCAGCTCTCCCTAAGCTTCAAACTAGACTACTTCTCCATAATATTCATCCCTGTAGCATTGTTCGTTACATGGTCCATCATAGAATTCTCACTGTGATATATAAACTCAGACCCAAACATTAATCAGTTCTTCAAATATCTACTCATCTTCCTAATTACCATACTAATCTTAGTTACCGCTAACAACCTATTCCAACTGTTCATCGGCTGAGAGGGCGTAGGAATTATATCCTTCTTGCTCATCAGTTGATGATACGCCCGAGCAGATGCCAACACAGCAGCCATTCAAGCAATCCTATACAACCGTATCGGCGATATCGGTTTCATCCTCGCCTTAGCATGATTTATCCTACACTCCAACTCATGAGACCCACAACAAATAGCCCTTCTAAACGCTAATCCAAGCCTCACCCCACTACTAGGCCTCCTCCTAGCAGCAGCAGGCAAATCAGCCCAATTAGGTCTCCACCCCTGACTCCCCTCAGCCATAGAAGGCCCCACCCCAGTCTCAGCCCTACTCCACTCAAGCACTATAGTTGTAGCAGGAATCTTCTTACTCATCCGCTTCCACCCCCTAGCAGAAAATAGCCCACTAATCCAAACTCTAACACTATGCTTAGGCGCTATCACCACTCTGTTCGCAGCAGTCTGCGCCCTTACACAAAATGACATCAAAAAAATCGTAGCCTTCTCCACTTCAAGTCAACTAGGACTCATAATAGTTACAATCGGCATCAACCAACCACACCTAGCATTCCTGCACATCTGTACCCACGCCTTCTTCAAAGCCATACTATTTATGTGCTCCGGGTCCATCATCCACAACCTTAACAATGAACAAGATATTCGAAAAATAGGAGGACTACTCAAAACCATACCTCTCACTTCAACCTCCCTCACCATTGGCAGCCTAGCATTAGCAGGAATACCTTTCCTCACAGGTTTCTACTCCAAAGACCACATCATCGAAACCGCAAACATATCATACACAAACGCCTGAGCCCTATCTATTACTCTCATCGCTACCTCCCTGACAAGCGCCTATAGCACTCGAATAATTCTTCTCACCCTAACAGGTCAACCTCGCTTCCCCACCCTTACTAACATTAACGAAAATAACCCCACCCTACTAAACCCCATTAAACGCCTGGCAGCCGGAAGCCTATTCGCAGGATTTCTCATTACTAACAACATTTCCCCCGCATCCCCCTTCCAAACAACAATCCCCCTCTACCTAAAACTCACAGCCCTCGCTGTCACTTTCCTAGGACTTCTAACAGCCCTAGACCTCAACTACCTAACCAACAAACTTAAAATAAAATCCCCACTATGCACATTTTATTTCTCCAACATACTCGGATTCTACCCTAGCATCACACACCGCACAATCCCCTATCTAGGCCTTCTTACGAGCCAAAACCTGCCCCTACTCCTCCTAGACCTAACCTGACTAGAAAAGCTATTACCTAAAACAATTTCACAGCACCAAATCTCCACCTCCATCATCACCTCAACCCAAAAAGGCATAATTAAACTTTACTTCCTCTCTTTCTTCTTCCCACTCATCCTAACCCTACTCCTAATCACATAACCTATTCCCCCGAGCAATCTCAATTACAATATATACACCAACAAACAATGTTCAACCAGTAACTACTACTAATCAACGCCCATAATCATACAAAGCCCCCGCACCAATAGGATCCTCCCGAATCAACCCTGACCCCTCTCCTTCATAAATTATTCAGCTTCCTACACTATTAAAGTTTACCACAACCACCACCCCATCATACTCTTTCACCCACAGCACCAATCCTACCTCCATCGCTAACCCCACTAAAACACTCACCAAGACCTCAACCCCTGACCCCCATGCCTCAGGATACTCCTCAATAGCCATCGCTGTAGTATATCCAAAGACAACCATCATTCCCCCTAAATAAATTAAAAAAACTATTAAACCCATATAACCTCCCCCAAAATTCAGAATAATAACACACCCGACCACACCGCTAACAATCAATACTAAACCCCCATAAATAGGAGAAGGCTTAGAAGAAAACCCCACAAACCCCATTACTAAACCCACACTCAACAGAAACAAAGCATACATCATTATTCTCGCACGGACTACAACCACGACCAATGATATGAAAAACCATCGTTGTATTTCAACTACAAGAACACCAATGACCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACCACTCATTCATCGACCTCCCCACCCCATCCAACATCTCCGCATGATGAAACTTCGGCTCACTCCTTGGCGCCTGCCTGATCCTCCAAATCACCACAGGACTATTCCTAGCCATGCACTACTCACCAGACGCCTCAACCGCCTTTTCATCAATCGCCCACATCACTCGAGACGTAAATTATGGCTGAATCATCCGCTACCTTCACGCCAATGGCGCCTCAATATTCTTTATCTGCCTCTTCCTACACATCGGGCGAGGCCTATATTACGGATCATTTCTCTACTCAGAAACCTGAAACATCGGCATTATCCTCCTGCTTGCAACTATAGCAACAGCCTTCATAGGCTATGTCCTCCCGTGAGGCCAAATATCATTCTGAGGGGCCACAGTAATTACAAACTTACTATCCGCCATCCCATACATTGGGACAGACCTAGTTCAATGAATCTGAGGAGGCTACTCAGTAGACAGTCCCACCCTCACACGATTCTTTACCTTTCACTTCATCTTGCCCTTCATTATTGCAGCCCTAGCAACACTCCACCTCCTATTCTTGCACGAAACGGGATCAAACAACCCCCTAGGAATCACCTCCCATTCCGATAAAATCACCTTCCACCCTTACTACACAATCAAAGACGCCCTCGGCTTACTTCTCTTCCTTCTCTCCTTAATGACATTAACACTATTCTCACCAGACCTCCTAGGCGACCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCACATCAAGCCCGAATGATATTTCCTATTCGCCTACACAATTCTCCGATCCGTCCCTAACAAACTAGGAGGCGTCCTTGCCCTATTACTATCCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAACAAAGCATAATATTTCGCCCACTAAGCCAATCACTTTATTGACTCCTAGCCGCAGACCTCCTCATTCTAACCTGAATCGGAGGACAACCAGTAAGCTACCCTTTTACCATCATTGGACAAGTAGCATCCGTACTATACTTCACAACAATCCTAATCCTAATACCAACTATCTCCCTAATTGAAAACAAAATACTCAAATGGGCCTGTCCTTGTAGTATAAACTAATACACCAGTCTTGTAAACCGGAGATGAAAACCTTTTTCCAAGGACAAATCAGAGAAAAAGTCTTTAACTCCACCATTAGCACCCAAAGCTAAGATTCTAATTTAAACTATTCTCTGTTCTTTCATGGGGAAGCAGATTTGGGTACCACCCAAGTATTGACTCACCCATCAACAACCGCTATGTATTTCGTACATTACTGCCAGCCACCATGAATATTGTACGGTACCATAAATACTTGACCACCTGTAGTACATAAAAACCCAATCCACATCAAAACCCCCTCCCCATGCTTACAAGCAAGTACAGCAATCAACCCTCAACTATCACACATCAACTGCAACTCCAAAGCCACCCCTCACCCACTAGGATACCAACAAACCTACCCACCCTTAACAGTACATAGTACATAAAGCCATTTACCGTACATAGCACATTACAGTCAAATCCCTTCTCGTCCCCATGGATGACCCCCCTCAGATAGGGGTCCCTTGACCACCATCCTCCGTGAAATCAATATCCCGCACAAGAGTGCTACTCTCCTCGCTCCGGGCCCATAACACTTGGGGGTAGCTAAAGTGAACTGTATCCGACATCTGGTTCCTACTTCAGGGTCATAAAGCCTAAATAGCCCACACGTTCCCCTTAAATAAGACATCACGATG'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt_fasta = join(data_path, 'GCF_000001405.38_GRCh38.p12_genomic.fna')\n",
    "ref_seq = get_read_sequences(mt_fasta, file_format='fasta', compression=None)\n",
    "mt_seq = ref_seq[mito_chrom]\n",
    "mt_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "run_control": {
     "frozen": true
    }
   },
   "outputs": [],
   "source": [
    "gff_filename = join(data_path,'GRCh38_latest_genomic.gff')\n",
    "gff_info = read_gff(gff_filename, file_format=None)\n",
    "## Analyzing GFF\n",
    "gff_info.Type.unique()\n",
    "\n",
    "gff_info = gff_info.loc[gff_info.Location == mito_chrom].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.899657Z",
     "start_time": "2020-02-15T21:37:05.866636Z"
    }
   },
   "outputs": [],
   "source": [
    "primers = pd.read_excel('96-Primers.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.910527Z",
     "start_time": "2020-02-15T21:37:05.902141Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence 5'</th>\n",
       "      <th>Corresponding fragment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FampA</td>\n",
       "      <td>AAATCTTACCCCGCCTGTTT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RampA</td>\n",
       "      <td>AATTAGGCTGTGGGTGGTTG</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FampB</td>\n",
       "      <td>GCCATACTAGTCTTTGCCGC</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RampB</td>\n",
       "      <td>GGCAGGTCAATTTCACTGGT</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>F1</td>\n",
       "      <td>CCGCTTCTGGCCACAGCACT</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Name           Sequence 5' Corresponding fragment\n",
       "0  FampA  AAATCTTACCCCGCCTGTTT                      A\n",
       "1  RampA  AATTAGGCTGTGGGTGGTTG                      A\n",
       "2  FampB  GCCATACTAGTCTTTGCCGC                      B\n",
       "3  RampB  GGCAGGTCAATTTCACTGGT                      B\n",
       "4     F1  CCGCTTCTGGCCACAGCACT                      B"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primers.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.929969Z",
     "start_time": "2020-02-15T21:37:05.912231Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence 5'</th>\n",
       "      <th>Corresponding fragment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>F1</td>\n",
       "      <td>CCGCTTCTGGCCACAGCACT</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>F2</td>\n",
       "      <td>GGTTGGTCAATTTCGTGCCAG</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>R1</td>\n",
       "      <td>ACTTGGGTTAATCGTGTGACC</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>F3</td>\n",
       "      <td>CATCAAGCACGCAGCAATG</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>F4</td>\n",
       "      <td>CTCACCACCTCTTGCTCAGC</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>F69</td>\n",
       "      <td>CGGCTTCGACCCTATATCC</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>R24</td>\n",
       "      <td>GGTAAAAGGAGGGCAATTTCT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>F70</td>\n",
       "      <td>CTACTCTCATAACCCTCAACACC</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>F71</td>\n",
       "      <td>ATTAAACCAGACCCAGCTACG</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>F72</td>\n",
       "      <td>AGCATATTTCACCTCCGCTAC</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Name              Sequence 5' Corresponding fragment\n",
       "4    F1     CCGCTTCTGGCCACAGCACT                      B\n",
       "5    F2    GGTTGGTCAATTTCGTGCCAG                      B\n",
       "6    R1    ACTTGGGTTAATCGTGTGACC                      B\n",
       "7    F3      CATCAAGCACGCAGCAATG                      B\n",
       "8    F4     CTCACCACCTCTTGCTCAGC                      B\n",
       "..  ...                      ...                    ...\n",
       "95  F69      CGGCTTCGACCCTATATCC                      A\n",
       "96  R24    GGTAAAAGGAGGGCAATTTCT                      A\n",
       "97  F70  CTACTCTCATAACCCTCAACACC                      A\n",
       "98  F71    ATTAAACCAGACCCAGCTACG                      A\n",
       "99  F72    AGCATATTTCACCTCCGCTAC                      A\n",
       "\n",
       "[96 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## remove the first 4, which are restriction enzymes\n",
    "primers = primers[~(primers[\"Name\"].str.contains(\"amp\"))]\n",
    "primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.961280Z",
     "start_time": "2020-02-15T21:37:05.931835Z"
    }
   },
   "outputs": [],
   "source": [
    "with open('96-Primers.fasta', mode='w') as f:\n",
    "    for row in primers.iterrows():\n",
    "        print('>Primer-{}|MT|Fragment-{}'.format(row[1]['Name'], row[1]['Corresponding fragment']), file=f)\n",
    "        print(row[1][\"Sequence 5'\"], file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.967430Z",
     "start_time": "2020-02-15T21:37:05.963421Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "96"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(primers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.979900Z",
     "start_time": "2020-02-15T21:37:05.969047Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CCGCTTCTGGCCACAGCACT': 'F1',\n",
       " 'GGTTGGTCAATTTCGTGCCAG': 'F2',\n",
       " 'ACTTGGGTTAATCGTGTGACC': 'R1',\n",
       " 'CATCAAGCACGCAGCAATG': 'F3',\n",
       " 'CTCACCACCTCTTGCTCAGC': 'F4',\n",
       " 'CTTGACCGCTCTGAGCTAAAC': 'F5',\n",
       " 'AAGCTAAGACCCCCGAAACC': 'F6',\n",
       " 'AAACCTACCGAGCCTGGTG': 'F7',\n",
       " 'GAGGAACAGCTCTTTGGACAC': 'F8',\n",
       " 'TCGTCCCAACAATTATATTACTACCA': 'F9',\n",
       " 'CTGTTTGTCGTAGGCAGATGG': 'R2',\n",
       " 'AACGCCACTTATCCAGTGAACC': 'F10',\n",
       " 'GACTCCCTAAAGCCCATGTCG': 'F11',\n",
       " 'CATCTGCCTACGACAAACA': 'F12',\n",
       " 'ACAGCCATTCTCATCCAAACCC': 'F13',\n",
       " 'AACCACGTTCTCCTGATCAAA': 'F14',\n",
       " 'GATATCGCCGATACGGTTG': 'R3',\n",
       " 'AGCGGATGAGTAAGAAGATTCC': 'R4',\n",
       " 'TTGAAGAAGGCGTGGGTACAG': 'R5',\n",
       " 'TTCATCCCTGTAGCATTGTTCG': 'F15',\n",
       " 'TTGCTCATCAGTTGATGATACG': 'F16',\n",
       " 'CACTCTGTTCGCAGCAGTATG': 'F17',\n",
       " 'CATCATCGAAACCGCAAAC': 'F18',\n",
       " 'TTTCTCCAACATACTCGGATTC': 'F19',\n",
       " 'ACAAACAATGGTCAACCAGTAAC': 'F20',\n",
       " 'TCCAAAGACAACCATCATTCC': 'F21',\n",
       " 'TTATCGGAATGGGAGGTGATTC': 'R6',\n",
       " 'TACTCACCAGACGCCTCAACCG': 'F22',\n",
       " 'AGTCCCACCCTCACACGATTC': 'F23',\n",
       " 'CGCCTACACAATTCTCCGATC': 'F24',\n",
       " 'CGGTTGTTGATGGGTGAGTC': 'R7',\n",
       " 'AAATGGGCCTGTCCTTGTAG': 'F25',\n",
       " 'TCATAAGGGCTATCGTAGTTTTC': 'R8',\n",
       " 'GTGGCAAGAAATGGGCTAC': 'F26',\n",
       " 'AACATATAACTGAACTCCTCACACC': 'F27',\n",
       " 'GCCGCAGTACTCTTAAAACTAGG': 'F28',\n",
       " 'AGGACTCAACATACTAGTCACAGC': 'F29',\n",
       " 'GCCATACTAGTCTTTGCCGC': 'F30',\n",
       " 'GCTGTGTTGGCATCTGCTC': 'R9',\n",
       " 'AAAGACCACATCATCGAAACC': 'F31',\n",
       " 'CTAACAGGTCAACCTCGCTTCC': 'F32',\n",
       " 'CCTTCATAAATTATTCAGCTTCCT': 'F33',\n",
       " 'CAATGATATGAAAAACCATCGTT': 'F34',\n",
       " 'GGATGGCGGATAGTAAGTTTGT': 'R10',\n",
       " 'CAGGGTTGGTCAATTTCGT': 'F35',\n",
       " 'AATGGTTTGGCTAAGGTTGT': 'F36',\n",
       " 'ACGAACAATGCTACAGGGATG': 'R11',\n",
       " 'GGCATTATCCTCCTGCTTGCAACTAT': 'F37',\n",
       " 'ATGTCCTGATCCAACATCGAG': 'R12',\n",
       " 'AGAAGAGCGATGGTGAGAGC': 'R13',\n",
       " 'CGACCTCGATGTTGGATCAGGACA': 'F38',\n",
       " 'AGATGGCAGAGCCCGGTAATC': 'F39',\n",
       " 'ACTACAACCCTTCGCTGACG': 'F40',\n",
       " 'CCCTAGCATTACTTATATGATATGTCTCCATACCCATTACAATCTCC': 'F41',\n",
       " 'TCAGGCTTCAACATCGAATACG': 'F42',\n",
       " 'CCCATCCTAAAGTAAGGTCAGC': 'F43',\n",
       " 'CCCTTTCACTTCTGAGTCCCAG': 'F44',\n",
       " 'CACCATCACCCTCCTTAACC': 'F45',\n",
       " 'GCTGAGTGAAGCATTGGACTG': 'R14',\n",
       " 'TAAGCACCCTAATCAACTGGC': 'F46',\n",
       " 'ATAGTGATGCCAGCAGCTAGG': 'R15',\n",
       " 'CGCATCTGCTATAGTGGAGG': 'F47',\n",
       " 'TTTCATGTGGTGTATGCATCG': 'R16',\n",
       " 'GCCATAACCCAATACCAAACG': 'F48',\n",
       " 'GAGGCTTCATTCACTGATTTCC': 'F49',\n",
       " 'GGGCAGGATAGTTCAGACGG': 'R17',\n",
       " 'TTCCCACAACACTTTCTCGGCC': 'F50',\n",
       " 'AAGTTAGCTTTACAGTGGGCTCTAG': 'R18',\n",
       " 'CGGTCAATGCTCTGAAATCTGTG': 'F51',\n",
       " 'CTGTTCGCTTCATTCATTGCC': 'F52',\n",
       " 'GTGGCGCTTCCAATTAGGTG': 'R19',\n",
       " 'GTGCTTTCTCGTGTTACATCG': 'R20',\n",
       " 'GAAAGTTGAGCCAATAATGACG': 'R21',\n",
       " 'TTTCACTTCCACTCCATAACGC': 'F53',\n",
       " 'CCTGATACTGGCATTTTGTAGATGTGG': 'F54',\n",
       " 'ACTACCACAACTCAACGGCTAC': 'F55',\n",
       " 'CTAACCGTGCAAAGGTAGCA': 'F56',\n",
       " 'GCAATTCCCGGACGTCTAAACCAAA': 'F57',\n",
       " 'GCCATAATATGATTTATCTCCACA': 'F58',\n",
       " 'AAACCCTCGTTCCACAGAA': 'F59',\n",
       " 'GATGAATAATAGCAGTTCTACCGT': 'F60',\n",
       " 'CAACGTAAAAATAAAATGACAGTT': 'F61',\n",
       " 'ATATGAAAATCACCTCGGAGC': 'F62',\n",
       " 'AGTTACAATATGGGAGATTATTCC': 'R22',\n",
       " 'CGCAAGTAGGTCTACAAGACG': 'F63',\n",
       " 'CTAATCTTCAACTCCTACATACTTCC': 'F64',\n",
       " 'ATCTGTTTTTAAGCCTAATGTGG': 'R23',\n",
       " 'AAGATTAAGAGAACCAACACCTCT': 'F65',\n",
       " 'AACAACCGACTAATCACCACCCAACAATG': 'F66',\n",
       " 'TCATCTTCACAATTCTAATTCTACTG': 'F67',\n",
       " 'TCGAGTCTCCCTTCACCATT': 'F68',\n",
       " 'CGGCTTCGACCCTATATCC': 'F69',\n",
       " 'GGTAAAAGGAGGGCAATTTCT': 'R24',\n",
       " 'CTACTCTCATAACCCTCAACACC': 'F70',\n",
       " 'ATTAAACCAGACCCAGCTACG': 'F71',\n",
       " 'AGCATATTTCACCTCCGCTAC': 'F72'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels = dict(zip(primers['Sequence 5\\''].values.tolist(), primers['Name'].values.tolist())) \n",
    "labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BWA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:05.988691Z",
     "start_time": "2020-02-15T21:37:05.981562Z"
    }
   },
   "outputs": [],
   "source": [
    "complementarity = {'A' : 'T',\n",
    "                   'C' : 'G',\n",
    "                   'G' : 'C',\n",
    "                   'T' : 'A',\n",
    "                   'N' : 'N'\n",
    "                   }\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.000976Z",
     "start_time": "2020-02-15T21:37:05.990697Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CCGCTTCTGGCCACAGCACT',\n",
       " 'GGTTGGTCAATTTCGTGCCAG',\n",
       " 'ACTTGGGTTAATCGTGTGACC',\n",
       " 'CATCAAGCACGCAGCAATG',\n",
       " 'CTCACCACCTCTTGCTCAGC',\n",
       " 'CTTGACCGCTCTGAGCTAAAC',\n",
       " 'AAGCTAAGACCCCCGAAACC',\n",
       " 'AAACCTACCGAGCCTGGTG',\n",
       " 'GAGGAACAGCTCTTTGGACAC',\n",
       " 'TCGTCCCAACAATTATATTACTACCA',\n",
       " 'CTGTTTGTCGTAGGCAGATGG',\n",
       " 'AACGCCACTTATCCAGTGAACC',\n",
       " 'GACTCCCTAAAGCCCATGTCG',\n",
       " 'CATCTGCCTACGACAAACA',\n",
       " 'ACAGCCATTCTCATCCAAACCC',\n",
       " 'AACCACGTTCTCCTGATCAAA',\n",
       " 'GATATCGCCGATACGGTTG',\n",
       " 'AGCGGATGAGTAAGAAGATTCC',\n",
       " 'TTGAAGAAGGCGTGGGTACAG',\n",
       " 'TTCATCCCTGTAGCATTGTTCG',\n",
       " 'TTGCTCATCAGTTGATGATACG',\n",
       " 'CACTCTGTTCGCAGCAGTATG',\n",
       " 'CATCATCGAAACCGCAAAC',\n",
       " 'TTTCTCCAACATACTCGGATTC',\n",
       " 'ACAAACAATGGTCAACCAGTAAC',\n",
       " 'TCCAAAGACAACCATCATTCC',\n",
       " 'TTATCGGAATGGGAGGTGATTC',\n",
       " 'TACTCACCAGACGCCTCAACCG',\n",
       " 'AGTCCCACCCTCACACGATTC',\n",
       " 'CGCCTACACAATTCTCCGATC',\n",
       " 'CGGTTGTTGATGGGTGAGTC',\n",
       " 'AAATGGGCCTGTCCTTGTAG',\n",
       " 'TCATAAGGGCTATCGTAGTTTTC',\n",
       " 'GTGGCAAGAAATGGGCTAC',\n",
       " 'AACATATAACTGAACTCCTCACACC',\n",
       " 'GCCGCAGTACTCTTAAAACTAGG',\n",
       " 'AGGACTCAACATACTAGTCACAGC',\n",
       " 'GCCATACTAGTCTTTGCCGC',\n",
       " 'GCTGTGTTGGCATCTGCTC',\n",
       " 'AAAGACCACATCATCGAAACC',\n",
       " 'CTAACAGGTCAACCTCGCTTCC',\n",
       " 'CCTTCATAAATTATTCAGCTTCCT',\n",
       " 'CAATGATATGAAAAACCATCGTT',\n",
       " 'GGATGGCGGATAGTAAGTTTGT',\n",
       " 'CAGGGTTGGTCAATTTCGT',\n",
       " 'AATGGTTTGGCTAAGGTTGT',\n",
       " 'ACGAACAATGCTACAGGGATG',\n",
       " 'GGCATTATCCTCCTGCTTGCAACTAT',\n",
       " 'ATGTCCTGATCCAACATCGAG',\n",
       " 'AGAAGAGCGATGGTGAGAGC',\n",
       " 'CGACCTCGATGTTGGATCAGGACA',\n",
       " 'AGATGGCAGAGCCCGGTAATC',\n",
       " 'ACTACAACCCTTCGCTGACG',\n",
       " 'CCCTAGCATTACTTATATGATATGTCTCCATACCCATTACAATCTCC',\n",
       " 'TCAGGCTTCAACATCGAATACG',\n",
       " 'CCCATCCTAAAGTAAGGTCAGC',\n",
       " 'CCCTTTCACTTCTGAGTCCCAG',\n",
       " 'CACCATCACCCTCCTTAACC',\n",
       " 'GCTGAGTGAAGCATTGGACTG',\n",
       " 'TAAGCACCCTAATCAACTGGC',\n",
       " 'ATAGTGATGCCAGCAGCTAGG',\n",
       " 'CGCATCTGCTATAGTGGAGG',\n",
       " 'TTTCATGTGGTGTATGCATCG',\n",
       " 'GCCATAACCCAATACCAAACG',\n",
       " 'GAGGCTTCATTCACTGATTTCC',\n",
       " 'GGGCAGGATAGTTCAGACGG',\n",
       " 'TTCCCACAACACTTTCTCGGCC',\n",
       " 'AAGTTAGCTTTACAGTGGGCTCTAG',\n",
       " 'CGGTCAATGCTCTGAAATCTGTG',\n",
       " 'CTGTTCGCTTCATTCATTGCC',\n",
       " 'GTGGCGCTTCCAATTAGGTG',\n",
       " 'GTGCTTTCTCGTGTTACATCG',\n",
       " 'GAAAGTTGAGCCAATAATGACG',\n",
       " 'TTTCACTTCCACTCCATAACGC',\n",
       " 'CCTGATACTGGCATTTTGTAGATGTGG',\n",
       " 'ACTACCACAACTCAACGGCTAC',\n",
       " 'CTAACCGTGCAAAGGTAGCA',\n",
       " 'GCAATTCCCGGACGTCTAAACCAAA',\n",
       " 'GCCATAATATGATTTATCTCCACA',\n",
       " 'AAACCCTCGTTCCACAGAA',\n",
       " 'GATGAATAATAGCAGTTCTACCGT',\n",
       " 'CAACGTAAAAATAAAATGACAGTT',\n",
       " 'ATATGAAAATCACCTCGGAGC',\n",
       " 'AGTTACAATATGGGAGATTATTCC',\n",
       " 'CGCAAGTAGGTCTACAAGACG',\n",
       " 'CTAATCTTCAACTCCTACATACTTCC',\n",
       " 'ATCTGTTTTTAAGCCTAATGTGG',\n",
       " 'AAGATTAAGAGAACCAACACCTCT',\n",
       " 'AACAACCGACTAATCACCACCCAACAATG',\n",
       " 'TCATCTTCACAATTCTAATTCTACTG',\n",
       " 'TCGAGTCTCCCTTCACCATT',\n",
       " 'CGGCTTCGACCCTATATCC',\n",
       " 'GGTAAAAGGAGGGCAATTTCT',\n",
       " 'CTACTCTCATAACCCTCAACACC',\n",
       " 'ATTAAACCAGACCCAGCTACG',\n",
       " 'AGCATATTTCACCTCCGCTAC']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primers[\"Sequence 5'\"].values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.641314Z",
     "start_time": "2020-02-15T21:37:06.002698Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "result = search_positions(primers[\"Sequence 5'\"].values.tolist(), mt_seq, K=100)\n",
    "result = result.dropna().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.653599Z",
     "start_time": "2020-02-15T21:37:06.643293Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pattern</th>\n",
       "      <th>Start-Position</th>\n",
       "      <th>End-Position</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCGCTTCTGGCCACAGCACT</td>\n",
       "      <td>314.0</td>\n",
       "      <td>333.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGTTGGTCAATTTCGTGCCAG</td>\n",
       "      <td>873.0</td>\n",
       "      <td>893.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CATCAAGCACGCAGCAATG</td>\n",
       "      <td>756.0</td>\n",
       "      <td>774.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTCACCACCTCTTGCTCAGC</td>\n",
       "      <td>1234.0</td>\n",
       "      <td>1253.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CTTGACCGCTCTGAGCTAAAC</td>\n",
       "      <td>1657.0</td>\n",
       "      <td>1677.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAGCTAAGACCCCCGAAACC</td>\n",
       "      <td>1892.0</td>\n",
       "      <td>1911.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AAACCTACCGAGCCTGGTG</td>\n",
       "      <td>1993.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GAGGAACAGCTCTTTGGACAC</td>\n",
       "      <td>2105.0</td>\n",
       "      <td>2125.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TCGTCCCAACAATTATATTACTACCA</td>\n",
       "      <td>10773.0</td>\n",
       "      <td>10798.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AACGCCACTTATCCAGTGAACC</td>\n",
       "      <td>11001.0</td>\n",
       "      <td>11022.0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Pattern  Start-Position  End-Position Strand\n",
       "0        CCGCTTCTGGCCACAGCACT           314.0         333.0      +\n",
       "1       GGTTGGTCAATTTCGTGCCAG           873.0         893.0      +\n",
       "2         CATCAAGCACGCAGCAATG           756.0         774.0      +\n",
       "3        CTCACCACCTCTTGCTCAGC          1234.0        1253.0      +\n",
       "4       CTTGACCGCTCTGAGCTAAAC          1657.0        1677.0      +\n",
       "5        AAGCTAAGACCCCCGAAACC          1892.0        1911.0      +\n",
       "6         AAACCTACCGAGCCTGGTG          1993.0        2011.0      +\n",
       "7       GAGGAACAGCTCTTTGGACAC          2105.0        2125.0      +\n",
       "8  TCGTCCCAACAATTATATTACTACCA         10773.0       10798.0      +\n",
       "9      AACGCCACTTATCCAGTGAACC         11001.0       11022.0      +"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.664831Z",
     "start_time": "2020-02-15T21:37:06.655281Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(93,)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result['Pattern'].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.689128Z",
     "start_time": "2020-02-15T21:37:06.666603Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pattern</th>\n",
       "      <th>Start-Position</th>\n",
       "      <th>End-Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCGCTTCTGGCCACAGCACT</td>\n",
       "      <td>314.0</td>\n",
       "      <td>333.0</td>\n",
       "      <td>+</td>\n",
       "      <td>F1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GGTTGGTCAATTTCGTGCCAG</td>\n",
       "      <td>873.0</td>\n",
       "      <td>893.0</td>\n",
       "      <td>+</td>\n",
       "      <td>F2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CATCAAGCACGCAGCAATG</td>\n",
       "      <td>756.0</td>\n",
       "      <td>774.0</td>\n",
       "      <td>+</td>\n",
       "      <td>F3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTCACCACCTCTTGCTCAGC</td>\n",
       "      <td>1234.0</td>\n",
       "      <td>1253.0</td>\n",
       "      <td>+</td>\n",
       "      <td>F4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CTTGACCGCTCTGAGCTAAAC</td>\n",
       "      <td>1657.0</td>\n",
       "      <td>1677.0</td>\n",
       "      <td>+</td>\n",
       "      <td>F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>GTGCTTTCTCGTGTTACATCG</td>\n",
       "      <td>9382.0</td>\n",
       "      <td>9403.0</td>\n",
       "      <td>-</td>\n",
       "      <td>R20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>GAAAGTTGAGCCAATAATGACG</td>\n",
       "      <td>9826.0</td>\n",
       "      <td>9848.0</td>\n",
       "      <td>-</td>\n",
       "      <td>R21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>AGTTACAATATGGGAGATTATTCC</td>\n",
       "      <td>6656.0</td>\n",
       "      <td>6680.0</td>\n",
       "      <td>-</td>\n",
       "      <td>R22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>ATCTGTTTTTAAGCCTAATGTGG</td>\n",
       "      <td>8081.0</td>\n",
       "      <td>8104.0</td>\n",
       "      <td>-</td>\n",
       "      <td>R23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>GGTAAAAGGAGGGCAATTTCT</td>\n",
       "      <td>10258.0</td>\n",
       "      <td>10279.0</td>\n",
       "      <td>-</td>\n",
       "      <td>R24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Pattern  Start-Position  End-Position Strand Label\n",
       "0       CCGCTTCTGGCCACAGCACT           314.0         333.0      +    F1\n",
       "1      GGTTGGTCAATTTCGTGCCAG           873.0         893.0      +    F2\n",
       "2        CATCAAGCACGCAGCAATG           756.0         774.0      +    F3\n",
       "3       CTCACCACCTCTTGCTCAGC          1234.0        1253.0      +    F4\n",
       "4      CTTGACCGCTCTGAGCTAAAC          1657.0        1677.0      +    F5\n",
       "..                       ...             ...           ...    ...   ...\n",
       "88     GTGCTTTCTCGTGTTACATCG          9382.0        9403.0      -   R20\n",
       "89    GAAAGTTGAGCCAATAATGACG          9826.0        9848.0      -   R21\n",
       "90  AGTTACAATATGGGAGATTATTCC          6656.0        6680.0      -   R22\n",
       "91   ATCTGTTTTTAAGCCTAATGTGG          8081.0        8104.0      -   R23\n",
       "92     GGTAAAAGGAGGGCAATTTCT         10258.0       10279.0      -   R24\n",
       "\n",
       "[93 rows x 5 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result['Label'] = result.Pattern.map(labels)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.891901Z",
     "start_time": "2020-02-15T21:37:06.691045Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "result.to_excel('96-Primers-Alignment.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Map positions into metadata in GFF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyzing GFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.899830Z",
     "start_time": "2020-02-15T21:37:06.894301Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['region', 'gene', 'tRNA', 'exon', 'rRNA', 'sequence_feature',\n",
       "       'CDS', 'D_loop'], dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_info.Type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.910398Z",
     "start_time": "2020-02-15T21:37:06.901413Z"
    }
   },
   "outputs": [],
   "source": [
    "gff_info = gff_info.loc[gff_info.Location == mito_chrom].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.935220Z",
     "start_time": "2020-02-15T21:37:06.914325Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene60922</td>\n",
       "      <td>gene</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene60923</td>\n",
       "      <td>gene</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gene60924</td>\n",
       "      <td>gene</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>gene60925</td>\n",
       "      <td>gene</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>gene60926</td>\n",
       "      <td>gene</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>gene60927</td>\n",
       "      <td>gene</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>gene60928</td>\n",
       "      <td>gene</td>\n",
       "      <td>4263</td>\n",
       "      <td>4331</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>gene60929</td>\n",
       "      <td>gene</td>\n",
       "      <td>4329</td>\n",
       "      <td>4400</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>gene60930</td>\n",
       "      <td>gene</td>\n",
       "      <td>4402</td>\n",
       "      <td>4469</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>gene60931</td>\n",
       "      <td>gene</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>gene60932</td>\n",
       "      <td>gene</td>\n",
       "      <td>5512</td>\n",
       "      <td>5579</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>gene60933</td>\n",
       "      <td>gene</td>\n",
       "      <td>5587</td>\n",
       "      <td>5655</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>gene60934</td>\n",
       "      <td>gene</td>\n",
       "      <td>5657</td>\n",
       "      <td>5729</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>gene60935</td>\n",
       "      <td>gene</td>\n",
       "      <td>5761</td>\n",
       "      <td>5826</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>gene60936</td>\n",
       "      <td>gene</td>\n",
       "      <td>5826</td>\n",
       "      <td>5891</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>gene60937</td>\n",
       "      <td>gene</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>gene60938</td>\n",
       "      <td>gene</td>\n",
       "      <td>7446</td>\n",
       "      <td>7514</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>gene60939</td>\n",
       "      <td>gene</td>\n",
       "      <td>7518</td>\n",
       "      <td>7585</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>gene60940</td>\n",
       "      <td>gene</td>\n",
       "      <td>7586</td>\n",
       "      <td>8269</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>gene60941</td>\n",
       "      <td>gene</td>\n",
       "      <td>8295</td>\n",
       "      <td>8364</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>gene60942</td>\n",
       "      <td>gene</td>\n",
       "      <td>8366</td>\n",
       "      <td>8572</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>gene60943</td>\n",
       "      <td>gene</td>\n",
       "      <td>8527</td>\n",
       "      <td>9207</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>gene60944</td>\n",
       "      <td>gene</td>\n",
       "      <td>9207</td>\n",
       "      <td>9990</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>gene60945</td>\n",
       "      <td>gene</td>\n",
       "      <td>9991</td>\n",
       "      <td>10058</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>gene60946</td>\n",
       "      <td>gene</td>\n",
       "      <td>10059</td>\n",
       "      <td>10404</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>gene60947</td>\n",
       "      <td>gene</td>\n",
       "      <td>10405</td>\n",
       "      <td>10469</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>gene60948</td>\n",
       "      <td>gene</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>gene60949</td>\n",
       "      <td>gene</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>gene60950</td>\n",
       "      <td>gene</td>\n",
       "      <td>12138</td>\n",
       "      <td>12206</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>gene60951</td>\n",
       "      <td>gene</td>\n",
       "      <td>12207</td>\n",
       "      <td>12265</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>gene60952</td>\n",
       "      <td>gene</td>\n",
       "      <td>12266</td>\n",
       "      <td>12336</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>gene60953</td>\n",
       "      <td>gene</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>gene60954</td>\n",
       "      <td>gene</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>gene60955</td>\n",
       "      <td>gene</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>gene60956</td>\n",
       "      <td>gene</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>gene60957</td>\n",
       "      <td>gene</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>gene60958</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Id  Type  Start    End Strand     Location\n",
       "1   gene60922  gene    577    647      +  NC_012920.1\n",
       "4   gene60923  gene    648   1601      +  NC_012920.1\n",
       "7   gene60924  gene   1602   1670      +  NC_012920.1\n",
       "10  gene60925  gene   1671   3229      +  NC_012920.1\n",
       "14  gene60926  gene   3230   3304      +  NC_012920.1\n",
       "17  gene60927  gene   3307   4262      +  NC_012920.1\n",
       "19  gene60928  gene   4263   4331      +  NC_012920.1\n",
       "22  gene60929  gene   4329   4400      -  NC_012920.1\n",
       "25  gene60930  gene   4402   4469      +  NC_012920.1\n",
       "28  gene60931  gene   4470   5511      +  NC_012920.1\n",
       "30  gene60932  gene   5512   5579      +  NC_012920.1\n",
       "33  gene60933  gene   5587   5655      -  NC_012920.1\n",
       "36  gene60934  gene   5657   5729      -  NC_012920.1\n",
       "39  gene60935  gene   5761   5826      -  NC_012920.1\n",
       "42  gene60936  gene   5826   5891      -  NC_012920.1\n",
       "45  gene60937  gene   5904   7445      +  NC_012920.1\n",
       "47  gene60938  gene   7446   7514      -  NC_012920.1\n",
       "50  gene60939  gene   7518   7585      +  NC_012920.1\n",
       "53  gene60940  gene   7586   8269      +  NC_012920.1\n",
       "55  gene60941  gene   8295   8364      +  NC_012920.1\n",
       "58  gene60942  gene   8366   8572      +  NC_012920.1\n",
       "60  gene60943  gene   8527   9207      +  NC_012920.1\n",
       "62  gene60944  gene   9207   9990      +  NC_012920.1\n",
       "64  gene60945  gene   9991  10058      +  NC_012920.1\n",
       "67  gene60946  gene  10059  10404      +  NC_012920.1\n",
       "69  gene60947  gene  10405  10469      +  NC_012920.1\n",
       "72  gene60948  gene  10470  10766      +  NC_012920.1\n",
       "74  gene60949  gene  10760  12137      +  NC_012920.1\n",
       "76  gene60950  gene  12138  12206      +  NC_012920.1\n",
       "79  gene60951  gene  12207  12265      +  NC_012920.1\n",
       "82  gene60952  gene  12266  12336      +  NC_012920.1\n",
       "85  gene60953  gene  12337  14148      +  NC_012920.1\n",
       "87  gene60954  gene  14149  14673      -  NC_012920.1\n",
       "89  gene60955  gene  14674  14742      -  NC_012920.1\n",
       "92  gene60956  gene  14747  15887      +  NC_012920.1\n",
       "94  gene60957  gene  15888  15953      +  NC_012920.1\n",
       "97  gene60958  gene  15956  16023      -  NC_012920.1"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_info.loc[gff_info.Type == 'gene']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Â Map GFF elements into positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:06.943701Z",
     "start_time": "2020-02-15T21:37:06.937315Z"
    }
   },
   "outputs": [],
   "source": [
    "pos_info = defaultdict(lambda: {'elements' : set(), 'primers' : set()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:55.399365Z",
     "start_time": "2020-02-15T21:37:06.945555Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 16569/16569 [00:48<00:00, 342.08it/s]\n"
     ]
    }
   ],
   "source": [
    "gff_transcript = gff_info.loc[gff_info.Type == 'gene'].reset_index(drop=True)\n",
    "\n",
    "for i in tqdm(range(1, len(mt_seq)+1)):\n",
    "    mapped_df = gff_transcript.loc[gff_transcript.apply(lambda row: between(row['Start'], row['End'], i),\n",
    "                                                        axis=1)]\n",
    "    pos_info[i]['elements'] = pos_info[i]['elements'].union(set(mapped_df.Id.unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:55.408658Z",
     "start_time": "2020-02-15T21:37:55.402129Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'elements': {'gene60937'}, 'primers': set()}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_info[7444]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map Primers into positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:56.190506Z",
     "start_time": "2020-02-15T21:37:55.410357Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "87\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Primers that are complementary to the transcripts\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Primer F3 located in gene gene60923 at 756\n",
      "Primer F35 located in gene gene60923 at 870\n",
      "Primer F2 located in gene gene60923 at 873\n",
      "Primer R1 located in gene gene60923 at 900\n",
      "Primer F4 located in gene gene60923 at 1234\n",
      "Primer F26 located in gene gene60923 at 1348\n",
      "Primer R8 located in gene gene60923 at 1379\n",
      "Primer F36 located in gene gene60925 at 1709\n",
      "Primer F6 located in gene gene60925 at 1892\n",
      "Primer F7 located in gene gene60925 at 1993\n",
      "Primer F8 located in gene gene60925 at 2105\n",
      "Primer F27 located in gene gene60925 at 2238\n",
      "Primer F56 located in gene gene60925 at 2579\n",
      "Primer F38 located in gene gene60925 at 2982\n",
      "Primer R12 located in gene gene60925 at 2985\n",
      "Primer F39 located in gene gene60926 at 3234\n",
      "Primer F40 located in gene gene60927 at 3441\n",
      "Primer R13 located in gene gene60927 at 3537\n",
      "Primer F58 located in gene gene60927 at 3850\n",
      "Primer F42 located in gene gene60927 at 3931\n",
      "Primer F41 located in gene gene60927 at 4196\n",
      "Primer F59 located in gene gene60931 at 4613\n",
      "Primer F44 located in gene gene60931 at 4797\n",
      "Primer F60 located in gene gene60931 at 5032\n",
      "Primer F45 located in gene gene60931 at 5318\n",
      "Primer F61 located in gene gene60931 at 5396\n",
      "Primer F46 located in gene gene60934 at 5700\n",
      "Primer F62 located in gene gene60935 at 5805\n",
      "Primer R14 located in gene gene60936 at 5861\n",
      "Primer F47 located in gene gene60937 at 6242\n",
      "Primer F48 located in gene gene60937 at 6426\n",
      "Primer R15 located in gene gene60937 at 6505\n",
      "Primer R22 located in gene gene60937 at 6656\n",
      "Primer F72 located in gene gene60937 at 6809\n",
      "Primer F49 located in gene gene60937 at 7075\n",
      "Primer F50 located in gene gene60937 at 7179\n",
      "Primer R16 located in gene gene60937 at 7234\n",
      "Primer F63 located in gene gene60940 at 7599\n",
      "Primer R17 located in gene gene60940 at 7772\n",
      "Primer F64 located in gene gene60940 at 7931\n",
      "Primer R23 located in gene gene60940 at 8081\n",
      "Primer F57 located in gene gene60940 at 8105\n",
      "Primer F51 located in gene gene60940 at 8164\n",
      "Primer F65 located in gene gene60941 at 8332\n",
      "Primer F52 located in gene gene60942 at 8539\n",
      "Primer F66 located in gene gene60943 at 8641\n",
      "Primer R19 located in gene gene60943 at 9039\n",
      "Primer F67 located in gene gene60943 at 9098\n",
      "Primer F53 located in gene gene60944 at 9309\n",
      "Primer R20 located in gene gene60944 at 9382\n",
      "Primer F68 located in gene gene60944 at 9751\n",
      "Primer R21 located in gene gene60944 at 9826\n",
      "Primer F54 located in gene gene60944 at 9922\n",
      "Primer F55 located in gene gene60946 at 10127\n",
      "Primer F69 located in gene gene60946 at 10175\n",
      "Primer R24 located in gene gene60946 at 10258\n",
      "Primer F70 located in gene gene60948 at 10600\n",
      "Primer F30 located in gene gene60948 at 10653\n",
      "Primer F9 located in gene gene60949 at 10773\n",
      "Primer F10 located in gene gene60949 at 11001\n",
      "Primer F11 located in gene gene60949 at 11403\n",
      "Primer F28 located in gene gene60949 at 11453\n",
      "Primer R2 located in gene gene60949 at 11579\n",
      "Primer F12 located in gene gene60949 at 11581\n",
      "Primer F13 located in gene gene60949 at 11654\n",
      "Primer F14 located in gene gene60949 at 11908\n",
      "Primer F29 located in gene gene60949 at 11947\n",
      "Primer F15 located in gene gene60953 at 12601\n",
      "Primer R11 located in gene gene60953 at 12602\n",
      "Primer F16 located in gene gene60953 at 12793\n",
      "Primer R9 located in gene gene60953 at 12817\n",
      "Primer R3 located in gene gene60953 at 12857\n",
      "Primer R4 located in gene gene60953 at 13101\n",
      "Primer R5 located in gene gene60953 at 13322\n",
      "Primer F31 located in gene gene60953 at 13510\n",
      "Primer F18 located in gene gene60953 at 13518\n",
      "Primer F32 located in gene gene60953 at 13627\n",
      "Primer F19 located in gene gene60953 at 13899\n",
      "Primer F33 located in gene gene60954 at 14288\n",
      "Primer F21 located in gene gene60954 at 14470\n",
      "Primer F34 located in gene gene60955 at 14701\n",
      "Primer F22 located in gene gene60956 at 14909\n",
      "Primer F37 located in gene gene60956 at 15092\n",
      "Primer R10 located in gene gene60956 at 15184\n",
      "Primer F23 located in gene gene60956 at 15260\n",
      "Primer R6 located in gene gene60956 at 15374\n",
      "Primer F24 located in gene gene60956 at 15574\n"
     ]
    }
   ],
   "source": [
    "location_info = []\n",
    "for row in result.iterrows():\n",
    "    r = row[1]\n",
    "    # See if primer is located in a gene and in the opposite strand (complementary to the transcript)\n",
    "    df = map_element_with_gff(r['Start-Position'],\n",
    "                              r['End-Position'],\n",
    "                              gff_transcript,\n",
    "                              #strand=r['Strand']\n",
    "                             )\n",
    "    if len(df) != 0:\n",
    "        # When gene is in - and primer in +\n",
    "        if r['Strand'] == '+':\n",
    "            start = int(r['Start-Position'])\n",
    "            end = start + window - 1 \n",
    "        elif r['Strand'] == '-':\n",
    "            end = int(r['End-Position'])\n",
    "            start = end - window +1\n",
    "            \n",
    "        for i in range(start, end+1):\n",
    "            pos_info[i]['primers'].add(r['Label'])\n",
    "            \n",
    "        location_info.append((int(r['Start-Position']), r['Label'], df.loc[0, 'Id']))\n",
    "\n",
    "print(len(location_info))\n",
    "print('-'*100)\n",
    "print('Primers that are complementary to the transcripts')\n",
    "print('-'*100)\n",
    "for l in sorted(location_info):\n",
    "        print('Primer {} located in gene {} at {}'.format(l[1], l[2], l[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:56.196142Z",
     "start_time": "2020-02-15T21:37:56.192176Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'elements': {'gene60953'}, 'primers': {'R4'}}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_info[13123]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract covered and uncovered regions when using primers for sequencing the transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:56.225451Z",
     "start_time": "2020-02-15T21:37:56.197651Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 16569/16569 [00:00<00:00, 1128685.49it/s]\n"
     ]
    }
   ],
   "source": [
    "regions = dict()\n",
    "\n",
    "tmp_cov = []\n",
    "tmp_uncov = []\n",
    "for i in tqdm(range(1, len(mt_seq)+1)):\n",
    "    if len(pos_info[i]['primers']) > 0:\n",
    "        tmp_cov.append(i)\n",
    "    elif len(pos_info[i]['elements']) > 0:\n",
    "        tmp_uncov.append(i)\n",
    "        \n",
    "regions['covered'] = list(interval_extract(tmp_cov))\n",
    "regions['uncovered'] = list(interval_extract(tmp_uncov))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:56.244213Z",
     "start_time": "2020-02-15T21:37:56.227062Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------------------------------------------\n",
      "Covered regions\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Fragment size: 267 - Located at 756-1022\n",
      "Fragment size: 264 - Located at 1234-1497\n",
      "Fragment size: 150 - Located at 1580-1729\n",
      "Fragment size: 496 - Located at 1892-2387\n",
      "Fragment size: 150 - Located at 2579-2728\n",
      "Fragment size: 275 - Located at 2857-3131\n",
      "Fragment size: 150 - Located at 3234-3383\n",
      "Fragment size: 183 - Located at 3408-3590\n",
      "Fragment size: 231 - Located at 3850-4080\n",
      "Fragment size: 150 - Located at 4196-4345\n",
      "Fragment size: 150 - Located at 4613-4762\n",
      "Fragment size: 150 - Located at 4797-4946\n",
      "Fragment size: 150 - Located at 5032-5181\n",
      "Fragment size: 228 - Located at 5318-5545\n",
      "Fragment size: 255 - Located at 5700-5954\n",
      "Fragment size: 439 - Located at 6242-6680\n",
      "Fragment size: 150 - Located at 6809-6958\n",
      "Fragment size: 254 - Located at 7075-7328\n",
      "Fragment size: 194 - Located at 7599-7792\n",
      "Fragment size: 383 - Located at 7931-8313\n",
      "Fragment size: 150 - Located at 8332-8481\n",
      "Fragment size: 252 - Located at 8539-8790\n",
      "Fragment size: 150 - Located at 8910-9059\n",
      "Fragment size: 150 - Located at 9098-9247\n",
      "Fragment size: 205 - Located at 9254-9458\n",
      "Fragment size: 202 - Located at 9699-9900\n",
      "Fragment size: 150 - Located at 9922-10071\n",
      "Fragment size: 198 - Located at 10127-10324\n",
      "Fragment size: 323 - Located at 10600-10922\n",
      "Fragment size: 150 - Located at 11001-11150\n",
      "Fragment size: 401 - Located at 11403-11803\n",
      "Fragment size: 189 - Located at 11908-12096\n",
      "Fragment size: 469 - Located at 12474-12942\n",
      "Fragment size: 150 - Located at 12974-13123\n",
      "Fragment size: 150 - Located at 13194-13343\n",
      "Fragment size: 267 - Located at 13510-13776\n",
      "Fragment size: 150 - Located at 13899-14048\n",
      "Fragment size: 150 - Located at 14288-14437\n",
      "Fragment size: 150 - Located at 14470-14619\n",
      "Fragment size: 150 - Located at 14701-14850\n",
      "Fragment size: 333 - Located at 14909-15241\n",
      "Fragment size: 163 - Located at 15247-15409\n",
      "Fragment size: 150 - Located at 15574-15723\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Uncovered regions\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Fragment size: 179 - Located at 577-755\n",
      "Fragment size: 211 - Located at 1023-1233\n",
      "Fragment size: 82 - Located at 1498-1579\n",
      "Fragment size: 162 - Located at 1730-1891\n",
      "Fragment size: 191 - Located at 2388-2578\n",
      "Fragment size: 128 - Located at 2729-2856\n",
      "Fragment size: 102 - Located at 3132-3233\n",
      "Fragment size: 24 - Located at 3384-3407\n",
      "Fragment size: 259 - Located at 3591-3849\n",
      "Fragment size: 115 - Located at 4081-4195\n",
      "Fragment size: 55 - Located at 4346-4400\n",
      "Fragment size: 211 - Located at 4402-4612\n",
      "Fragment size: 34 - Located at 4763-4796\n",
      "Fragment size: 85 - Located at 4947-5031\n",
      "Fragment size: 136 - Located at 5182-5317\n",
      "Fragment size: 34 - Located at 5546-5579\n",
      "Fragment size: 69 - Located at 5587-5655\n",
      "Fragment size: 43 - Located at 5657-5699\n",
      "Fragment size: 287 - Located at 5955-6241\n",
      "Fragment size: 128 - Located at 6681-6808\n",
      "Fragment size: 116 - Located at 6959-7074\n",
      "Fragment size: 186 - Located at 7329-7514\n",
      "Fragment size: 81 - Located at 7518-7598\n",
      "Fragment size: 138 - Located at 7793-7930\n",
      "Fragment size: 18 - Located at 8314-8331\n",
      "Fragment size: 57 - Located at 8482-8538\n",
      "Fragment size: 119 - Located at 8791-8909\n",
      "Fragment size: 38 - Located at 9060-9097\n",
      "Fragment size: 6 - Located at 9248-9253\n",
      "Fragment size: 240 - Located at 9459-9698\n",
      "Fragment size: 21 - Located at 9901-9921\n",
      "Fragment size: 55 - Located at 10072-10126\n",
      "Fragment size: 275 - Located at 10325-10599\n",
      "Fragment size: 78 - Located at 10923-11000\n",
      "Fragment size: 252 - Located at 11151-11402\n",
      "Fragment size: 104 - Located at 11804-11907\n",
      "Fragment size: 377 - Located at 12097-12473\n",
      "Fragment size: 31 - Located at 12943-12973\n",
      "Fragment size: 70 - Located at 13124-13193\n",
      "Fragment size: 166 - Located at 13344-13509\n",
      "Fragment size: 122 - Located at 13777-13898\n",
      "Fragment size: 239 - Located at 14049-14287\n",
      "Fragment size: 32 - Located at 14438-14469\n",
      "Fragment size: 81 - Located at 14620-14700\n",
      "Fragment size: 58 - Located at 14851-14908\n",
      "Fragment size: 5 - Located at 15242-15246\n",
      "Fragment size: 164 - Located at 15410-15573\n",
      "Fragment size: 230 - Located at 15724-15953\n",
      "Fragment size: 68 - Located at 15956-16023\n"
     ]
    }
   ],
   "source": [
    "print('-'*100)\n",
    "print('Covered regions')\n",
    "print('-'*100)\n",
    "for r in regions['covered']:\n",
    "    size = r[1] - r[0] + 1\n",
    "    print('Fragment size: {} - Located at {}-{}'.format(size, r[0], r[1]))\n",
    "    \n",
    "print('-'*100)\n",
    "print('Uncovered regions')\n",
    "print('-'*100)\n",
    "for r in regions['uncovered']:\n",
    "    size = r[1] - r[0] + 1\n",
    "    print('Fragment size: {} - Located at {}-{}'.format(size, r[0], r[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construct additional primers for regions that are not covered (only if minimum of 25 bp uncovered)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:56.296196Z",
     "start_time": "2020-02-15T21:37:56.245764Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'F_p577': ['TTTATGTAGCTTACCTCCTC', 'gene60922'],\n",
       " 'F_p1023': ['GAAAGTGGCTTTAACATATC', 'gene60923'],\n",
       " 'F_p1498': ['TCAAGTATACTTCAAAGGAC', 'gene60923'],\n",
       " 'F_p1730': ['ACCCAAATAAAGTATAGGCG', 'gene60925'],\n",
       " 'F_p2388': ['CAATCAACCAACAAGTCATT', 'gene60925'],\n",
       " 'F_p2729': ['ATGGAGCTTTAATTTATTAA', 'gene60925'],\n",
       " 'F_p3132': ['ACAAGAGAAATAAGGCCTAC', 'gene60925'],\n",
       " 'F_p3591': ['GTCAACCTCAACCTAGGCCT', 'gene60927'],\n",
       " 'F_p4081': ['TTGTCACCAAGACCCTACTT', 'gene60927'],\n",
       " 'R_p4400': ['TAGGATGGGGTGTGATAGGT', 'gene60929'],\n",
       " 'F_p4402': ['GTAAGGTCAGCTAAATAAGC', 'gene60930'],\n",
       " 'F_p4763': ['ATAATAGCTATAGCAATAAA', 'gene60931'],\n",
       " 'F_p4947': ['TATCCATCATAGCAGGCAGT', 'gene60931'],\n",
       " 'F_p5182': ['ATGACTAACACCCTTAATTC', 'gene60931'],\n",
       " 'F_p5546': ['AAGCCCTCAGTAAGTTGCAA', 'gene60932'],\n",
       " 'R_p5655': ['AAGGGCTTAGCTTAATTAAA', 'gene60933'],\n",
       " 'R_p5699': ['GCTGTTAACTAAGTGTTTGT', 'gene60934'],\n",
       " 'F_p5955': ['TATACCTATTATTCGGCGCA', 'gene60937'],\n",
       " 'F_p6681': ['ACTACTCCGGAAAAAAAGAA', 'gene60937'],\n",
       " 'F_p6959': ['CTGACTGGCATTGTATTAGC', 'gene60937'],\n",
       " 'F_p7329': ['TCGCTTCGAAGCGAAAAGTC', 'gene60937'],\n",
       " 'F_p7518': ['AGGTATTAGAAAAACCATTT', 'gene60939'],\n",
       " 'F_p7793': ['CCATCATCCTAGTCCTCATC', 'gene60940'],\n",
       " 'F_p8482': ['AAGCCCATAAAAATAAAAAA', 'gene60942'],\n",
       " 'F_p8791': ['CTCACTCATTTACACCAACC', 'gene60943'],\n",
       " 'F_p9060': ['CTAGCAATATCAACCATTAA', 'gene60943'],\n",
       " 'F_p9459': ['TATTTATTACCTCAGAAGTT', 'gene60944'],\n",
       " 'F_p10072': ['AATTTTAATAATCAACACCC', 'gene60946'],\n",
       " 'F_p10325': ['TCATCCCTCTTATTAATCAT', 'gene60946'],\n",
       " 'F_p10923': ['CTTTTCCTCCGACCCCCTAA', 'gene60949'],\n",
       " 'F_p11151': ['TATCATCACCCGATGAGGCA', 'gene60949'],\n",
       " 'F_p11804': ['AAACTCTACTCCCACTAATA', 'gene60949'],\n",
       " 'F_p12097': ['CTATCCCTCAACCCCGACAT', 'gene60949'],\n",
       " 'F_p12943': ['TTCTAAACGCTAATCCAAGC', 'gene60953'],\n",
       " 'F_p13124': ['CCACCCCCTAGCAGAAAATA', 'gene60953'],\n",
       " 'F_p13344': ['GCCATACTATTTATGTGCTC', 'gene60953'],\n",
       " 'F_p13777': ['CAATCCCCCTCTACCTAAAA', 'gene60953'],\n",
       " 'F_p14049': ['TCCACCTCCATCATCACCTC', 'gene60953'],\n",
       " 'R_p14469': ['TATACTACAGCGATGGCTAT', 'gene60954'],\n",
       " 'R_p14700': ['GTCGTGGTTGTAGTCCGTGC', 'gene60954'],\n",
       " 'F_p14851': ['CTCCTTGGCGCCTGCCTGAT', 'gene60956'],\n",
       " 'F_p15410': ['CTTACTACACAATCAAAGAC', 'gene60956'],\n",
       " 'F_p15724': ['CTCCTAGCCGCAGACCTCCT', 'gene60956'],\n",
       " 'R_p16023': ['CAGAGAATAGTTTAAATTAG', 'gene60958']}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Loop through uncovered regions, if positive strand take the start plus the first 20 bp,\n",
    "##                                 if negative strand take the reverse complement of the end - 20bp\n",
    "##                                 e.g. start=20, end=150, take 130-150 then reverse complement\n",
    "new_primers = dict()\n",
    "for i in regions[\"uncovered\"]:\n",
    "    if i[1] - i[0] + 1 >= 25:\n",
    "        curr_gene = list(pos_info[i[0]+2][\"elements\"])[0]\n",
    "        curr = gff_transcript[gff_transcript['Id']==list(pos_info[i[0]]['elements'])[0]]\n",
    "        if curr[\"Strand\"].values[0] == '+':\n",
    "            new_primers[\"F_p\" + str(i[0])  ] = [str(mt_seq_biopython[i[0]:i[0]+20]), curr_gene]\n",
    "        elif curr[\"Strand\"].values[0] == \"-\":\n",
    "            new_primers[\"R_p\" + str(i[1])] = [str(mt_seq_biopython[i[1]-20:i[1]].reverse_complement()), curr_gene]\n",
    "        else:\n",
    "            print(\"No strand information\")\n",
    "print(len(new_primers))\n",
    "new_primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combine the primers that map and the additional primers and save to csv (Name, Sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:37:56.533790Z",
     "start_time": "2020-02-15T21:37:56.297816Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>F2</th>\n",
       "      <td>GGTTGGTCAATTTCGTGCCAG</td>\n",
       "      <td>gene60923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F3</th>\n",
       "      <td>CATCAAGCACGCAGCAATG</td>\n",
       "      <td>gene60923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F4</th>\n",
       "      <td>CTCACCACCTCTTGCTCAGC</td>\n",
       "      <td>gene60923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F6</th>\n",
       "      <td>AAGCTAAGACCCCCGAAACC</td>\n",
       "      <td>gene60925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F7</th>\n",
       "      <td>AAACCTACCGAGCCTGGTG</td>\n",
       "      <td>gene60925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_p14700</th>\n",
       "      <td>GTCGTGGTTGTAGTCCGTGC</td>\n",
       "      <td>gene60954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_p14851</th>\n",
       "      <td>CTCCTTGGCGCCTGCCTGAT</td>\n",
       "      <td>gene60956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_p15410</th>\n",
       "      <td>CTTACTACACAATCAAAGAC</td>\n",
       "      <td>gene60956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F_p15724</th>\n",
       "      <td>CTCCTAGCCGCAGACCTCCT</td>\n",
       "      <td>gene60956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R_p16023</th>\n",
       "      <td>CAGAGAATAGTTTAAATTAG</td>\n",
       "      <td>gene60958</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>131 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Sequence       Gene\n",
       "F2        GGTTGGTCAATTTCGTGCCAG  gene60923\n",
       "F3          CATCAAGCACGCAGCAATG  gene60923\n",
       "F4         CTCACCACCTCTTGCTCAGC  gene60923\n",
       "F6         AAGCTAAGACCCCCGAAACC  gene60925\n",
       "F7          AAACCTACCGAGCCTGGTG  gene60925\n",
       "...                         ...        ...\n",
       "R_p14700   GTCGTGGTTGTAGTCCGTGC  gene60954\n",
       "F_p14851   CTCCTTGGCGCCTGCCTGAT  gene60956\n",
       "F_p15410   CTTACTACACAATCAAAGAC  gene60956\n",
       "F_p15724   CTCCTAGCCGCAGACCTCCT  gene60956\n",
       "R_p16023   CAGAGAATAGTTTAAATTAG  gene60958\n",
       "\n",
       "[131 rows x 2 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_primers = pd.DataFrame(columns=[\"Sequence\", \"Gene\"])#, index=\"Name\")\n",
    "for l in location_info:\n",
    "    curr_seq = primers[primers[\"Name\"] == l[1]][\"Sequence 5'\"].values[0]\n",
    "    final_primers = pd.concat((final_primers, \n",
    "                               pd.DataFrame({\"Sequence\":curr_seq, \"Gene\":l[2]},index=[l[1]])))\n",
    "final_primers\n",
    "\n",
    "for p in new_primers:\n",
    "    final_primers = pd.concat((final_primers, \n",
    "                               pd.DataFrame({\"Sequence\":new_primers[p][0], \"Gene\":new_primers[p][1]},\n",
    "                                            index=[p])))\n",
    "final_primers.sort_index().to_csv(\"final_primers.csv\")\n",
    "final_primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map to MT regardless genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T20:06:02.720040Z",
     "start_time": "2020-02-15T20:06:02.670484Z"
    },
    "deletable": false,
    "editable": false,
    "run_control": {
     "frozen": true
    }
   },
   "outputs": [],
   "source": [
    "location_info2 = []\n",
    "for row in result.iterrows():\n",
    "    r = row[1]\n",
    "\n",
    "    # When gene is in - and primer in +\n",
    "    if r['Strand'] == '+':\n",
    "        start = int(r['Start-Position'])\n",
    "        end = start + window - 1 \n",
    "    elif r['Strand'] == '-':\n",
    "        end = int(r['End-Position'])\n",
    "        start = end - window +1\n",
    "\n",
    "    for i in range(start, end+1):\n",
    "        location_info2.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "run_control": {
     "frozen": true
    }
   },
   "outputs": [],
   "source": [
    "regions2 = dict()\n",
    "\n",
    "tmp_cov = []\n",
    "tmp_uncov = []\n",
    "for i in tqdm(range(1, len(mt_seq)+1)):\n",
    "    if i in location_info2:\n",
    "        tmp_cov.append(i)\n",
    "    else:\n",
    "        tmp_uncov.append(i)\n",
    "        \n",
    "regions2['covered'] = list(interval_extract(tmp_cov))\n",
    "regions2['uncovered'] = list(interval_extract(tmp_uncov))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "run_control": {
     "frozen": true
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print('-'*100)\n",
    "print('Covered regions')\n",
    "print('-'*100)\n",
    "for r in regions2['covered']:\n",
    "    size = r[1] - r[0] + 1\n",
    "    print('Fragment size: {} - Located at {}-{}'.format(size, r[0], r[1]))\n",
    "    \n",
    "print('-'*100)\n",
    "print('Uncovered regions')\n",
    "print('-'*100)\n",
    "for r in regions2['uncovered']:\n",
    "    size = r[1] - r[0] + 1\n",
    "    print('Fragment size: {} - Located at {}-{}'.format(size, r[0], r[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Double check by loading in the genes fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T22:55:09.519544Z",
     "start_time": "2020-02-15T22:55:08.796842Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 100/100 [00:00<00:00, 142.66it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'F21',\n",
       " 'F33',\n",
       " 'F34',\n",
       " 'F36',\n",
       " 'F46',\n",
       " 'F62',\n",
       " 'R1',\n",
       " 'R10',\n",
       " 'R11',\n",
       " 'R12',\n",
       " 'R13',\n",
       " 'R15',\n",
       " 'R16',\n",
       " 'R17',\n",
       " 'R19',\n",
       " 'R2',\n",
       " 'R20',\n",
       " 'R21',\n",
       " 'R22',\n",
       " 'R23',\n",
       " 'R24',\n",
       " 'R3',\n",
       " 'R4',\n",
       " 'R5',\n",
       " 'R6',\n",
       " 'R8',\n",
       " 'R9'}"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mt_genes = SeqIO.to_dict(SeqIO.parse(\"MT.genes.fasta\", \"fasta\"))\n",
    "mt_genes\n",
    "\n",
    "primers_used = set()\n",
    "genes_seen = set()\n",
    "\n",
    "pos_primers_used = set()\n",
    "pos_genes_seen = set()\n",
    "for gene in tqdm(mt_genes):\n",
    "    if 'gene' in gene:\n",
    "        curr_for = mt_genes[gene].seq\n",
    "        curr_rev = mt_genes[gene].seq.reverse_complement()\n",
    "        #print(curr_rev)\n",
    "        for p_ind, p_val in final_primers.iterrows():\n",
    "    #             print('p_ind',p_ind)\n",
    "    #             print('p_val',p_val[\"Sequence\"])\n",
    "            if p_val[\"Sequence\"] in curr_rev:\n",
    "                primers_used.add(p_ind)\n",
    "                genes_seen.add(gene)\n",
    "\n",
    "            if p_val[\"Sequence\"] in curr_for:\n",
    "                pos_primers_used.add(p_ind)\n",
    "                pos_genes_seen.add(gene)\n",
    "\n",
    "primers_used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T22:55:19.104558Z",
     "start_time": "2020-02-15T22:55:19.099172Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(primers_used)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T22:55:21.409089Z",
     "start_time": "2020-02-15T22:55:21.401224Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'F10',\n",
       " 'F11',\n",
       " 'F12',\n",
       " 'F13',\n",
       " 'F14',\n",
       " 'F15',\n",
       " 'F16',\n",
       " 'F18',\n",
       " 'F19',\n",
       " 'F2',\n",
       " 'F22',\n",
       " 'F23',\n",
       " 'F24',\n",
       " 'F26',\n",
       " 'F27',\n",
       " 'F28',\n",
       " 'F29',\n",
       " 'F3',\n",
       " 'F30',\n",
       " 'F31',\n",
       " 'F32',\n",
       " 'F35',\n",
       " 'F37',\n",
       " 'F38',\n",
       " 'F39',\n",
       " 'F4',\n",
       " 'F40',\n",
       " 'F41',\n",
       " 'F42',\n",
       " 'F44',\n",
       " 'F45',\n",
       " 'F47',\n",
       " 'F48',\n",
       " 'F49',\n",
       " 'F50',\n",
       " 'F51',\n",
       " 'F52',\n",
       " 'F53',\n",
       " 'F54',\n",
       " 'F55',\n",
       " 'F56',\n",
       " 'F57',\n",
       " 'F58',\n",
       " 'F59',\n",
       " 'F6',\n",
       " 'F60',\n",
       " 'F61',\n",
       " 'F63',\n",
       " 'F64',\n",
       " 'F65',\n",
       " 'F66',\n",
       " 'F67',\n",
       " 'F68',\n",
       " 'F69',\n",
       " 'F7',\n",
       " 'F70',\n",
       " 'F72',\n",
       " 'F8',\n",
       " 'F9',\n",
       " 'F_p10072',\n",
       " 'F_p1023',\n",
       " 'F_p10325',\n",
       " 'F_p10923',\n",
       " 'F_p11151',\n",
       " 'F_p11804',\n",
       " 'F_p12097',\n",
       " 'F_p12943',\n",
       " 'F_p13124',\n",
       " 'F_p13344',\n",
       " 'F_p13777',\n",
       " 'F_p14049',\n",
       " 'F_p14851',\n",
       " 'F_p1498',\n",
       " 'F_p15410',\n",
       " 'F_p15724',\n",
       " 'F_p1730',\n",
       " 'F_p2388',\n",
       " 'F_p2729',\n",
       " 'F_p3132',\n",
       " 'F_p3591',\n",
       " 'F_p4081',\n",
       " 'F_p4402',\n",
       " 'F_p4763',\n",
       " 'F_p4947',\n",
       " 'F_p5182',\n",
       " 'F_p5546',\n",
       " 'F_p577',\n",
       " 'F_p5955',\n",
       " 'F_p6681',\n",
       " 'F_p6959',\n",
       " 'F_p7329',\n",
       " 'F_p7518',\n",
       " 'F_p7793',\n",
       " 'F_p8482',\n",
       " 'F_p8791',\n",
       " 'F_p9060',\n",
       " 'F_p9459',\n",
       " 'R14',\n",
       " 'R_p14469',\n",
       " 'R_p14700',\n",
       " 'R_p16023',\n",
       " 'R_p4400',\n",
       " 'R_p5655',\n",
       " 'R_p5699'}"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_primers_used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T22:55:35.140703Z",
     "start_time": "2020-02-15T22:55:35.134743Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "104"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pos_primers_used)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T22:55:35.150725Z",
     "start_time": "2020-02-15T22:55:35.142375Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pos_genes_seen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:49:33.395889Z",
     "start_time": "2020-02-15T21:49:33.392079Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "131"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(final_primers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-15T21:50:30.211749Z",
     "start_time": "2020-02-15T21:50:30.207455Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(primers_used)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
