{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.663682Z",
     "start_time": "2020-02-17T00:27:27.151593Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gzip\n",
    "from collections import defaultdict\n",
    "from tqdm import tqdm\n",
    "from os.path import join\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.670516Z",
     "start_time": "2020-02-17T00:27:27.665856Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from alignment import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.677616Z",
     "start_time": "2020-02-17T00:27:27.672114Z"
    }
   },
   "outputs": [],
   "source": [
    "data_path = '/data2/genome/human_GRCh38/ncbi/'\n",
    "mito_chrom = 'NC_012920.1'\n",
    "\n",
    "window =150 #bps covered for sequencing\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load genomic info and extract mitochondria information\n",
    "- MT fasta sequence\n",
    "- gff file\n",
    "- MT genes in fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.696144Z",
     "start_time": "2020-02-17T00:27:27.679147Z"
    }
   },
   "outputs": [],
   "source": [
    "mt_fasta = join(data_path, 'GCF_000001405.38_GRCh38.p12_genomic.fna')\n",
    "\n",
    "if not os.path.exists(\"MT.fasta\"):\n",
    "    ref_seq = SeqIO.index(mt_fasta, \"fasta\")\n",
    "\n",
    "    mt_seq = ref_seq[mito_chrom]\n",
    "    with open(\"MT.fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(mt_seq, output_handle, \"fasta\")\n",
    "    mt_seq_biopython = mt_seq.seq\n",
    "    mt_seq = str(mt_seq.seq)\n",
    "else:\n",
    "    mt_seq = str(list(SeqIO.parse(\"MT.fasta\", format='fasta'))[0].seq)\n",
    "    mt_seq_biopython = list(SeqIO.parse(\"MT.fasta\", format='fasta'))[0].seq\n",
    "    #mt_seq_biopython_reverse = mt_seq_biopython.reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.727435Z",
     "start_time": "2020-02-17T00:27:27.697540Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading line 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>id1959760</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>16569</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene60922</td>\n",
       "      <td>gene</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna171173</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>id1959761</td>\n",
       "      <td>exon</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene60923</td>\n",
       "      <td>gene</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>id1959784</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>gene60958</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>rna171196</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>id1959785</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>id1959786</td>\n",
       "      <td>D_loop</td>\n",
       "      <td>16024</td>\n",
       "      <td>17145</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Id    Type  Start    End Strand     Location\n",
       "0    id1959760  region      1  16569      +  NC_012920.1\n",
       "1    gene60922    gene    577    647      +  NC_012920.1\n",
       "2    rna171173    tRNA    577    647      +  NC_012920.1\n",
       "3    id1959761    exon    577    647      +  NC_012920.1\n",
       "4    gene60923    gene    648   1601      +  NC_012920.1\n",
       "..         ...     ...    ...    ...    ...          ...\n",
       "96   id1959784    exon  15888  15953      +  NC_012920.1\n",
       "97   gene60958    gene  15956  16023      -  NC_012920.1\n",
       "98   rna171196    tRNA  15956  16023      -  NC_012920.1\n",
       "99   id1959785    exon  15956  16023      -  NC_012920.1\n",
       "100  id1959786  D_loop  16024  17145      -  NC_012920.1\n",
       "\n",
       "[101 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_gff_filename = join(data_path,'GRCh38_latest_genomic.gff')\n",
    "gff_filename = \"MT.gff\"\n",
    "if not os.path.exists(gff_filename):\n",
    "    cmd = f\"grep {mito_chrom} {full_gff_filename} > {gff_filename}\"\n",
    "    print(cmd)\n",
    "    os.system(cmd)\n",
    "\n",
    "gff_info = read_gff(gff_filename, file_format=None)\n",
    "## Analyzing GFF\n",
    "gff_info.Type.unique()\n",
    "\n",
    "gff_info = gff_info.loc[gff_info.Location == mito_chrom].reset_index(drop=True)\n",
    "gff_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.745146Z",
     "start_time": "2020-02-17T00:27:27.728963Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene60922</td>\n",
       "      <td>gene</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene60923</td>\n",
       "      <td>gene</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gene60924</td>\n",
       "      <td>gene</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>gene60925</td>\n",
       "      <td>gene</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>gene60926</td>\n",
       "      <td>gene</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>gene60927</td>\n",
       "      <td>gene</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>gene60928</td>\n",
       "      <td>gene</td>\n",
       "      <td>4263</td>\n",
       "      <td>4331</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>gene60929</td>\n",
       "      <td>gene</td>\n",
       "      <td>4329</td>\n",
       "      <td>4400</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>gene60930</td>\n",
       "      <td>gene</td>\n",
       "      <td>4402</td>\n",
       "      <td>4469</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>gene60931</td>\n",
       "      <td>gene</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>gene60932</td>\n",
       "      <td>gene</td>\n",
       "      <td>5512</td>\n",
       "      <td>5579</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>gene60933</td>\n",
       "      <td>gene</td>\n",
       "      <td>5587</td>\n",
       "      <td>5655</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>gene60934</td>\n",
       "      <td>gene</td>\n",
       "      <td>5657</td>\n",
       "      <td>5729</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>gene60935</td>\n",
       "      <td>gene</td>\n",
       "      <td>5761</td>\n",
       "      <td>5826</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>gene60936</td>\n",
       "      <td>gene</td>\n",
       "      <td>5826</td>\n",
       "      <td>5891</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>gene60937</td>\n",
       "      <td>gene</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>gene60938</td>\n",
       "      <td>gene</td>\n",
       "      <td>7446</td>\n",
       "      <td>7514</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>gene60939</td>\n",
       "      <td>gene</td>\n",
       "      <td>7518</td>\n",
       "      <td>7585</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>gene60940</td>\n",
       "      <td>gene</td>\n",
       "      <td>7586</td>\n",
       "      <td>8269</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>gene60941</td>\n",
       "      <td>gene</td>\n",
       "      <td>8295</td>\n",
       "      <td>8364</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>gene60942</td>\n",
       "      <td>gene</td>\n",
       "      <td>8366</td>\n",
       "      <td>8572</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>gene60943</td>\n",
       "      <td>gene</td>\n",
       "      <td>8527</td>\n",
       "      <td>9207</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>gene60944</td>\n",
       "      <td>gene</td>\n",
       "      <td>9207</td>\n",
       "      <td>9990</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>gene60945</td>\n",
       "      <td>gene</td>\n",
       "      <td>9991</td>\n",
       "      <td>10058</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>gene60946</td>\n",
       "      <td>gene</td>\n",
       "      <td>10059</td>\n",
       "      <td>10404</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>gene60947</td>\n",
       "      <td>gene</td>\n",
       "      <td>10405</td>\n",
       "      <td>10469</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>gene60948</td>\n",
       "      <td>gene</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>gene60949</td>\n",
       "      <td>gene</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>gene60950</td>\n",
       "      <td>gene</td>\n",
       "      <td>12138</td>\n",
       "      <td>12206</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>gene60951</td>\n",
       "      <td>gene</td>\n",
       "      <td>12207</td>\n",
       "      <td>12265</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>gene60952</td>\n",
       "      <td>gene</td>\n",
       "      <td>12266</td>\n",
       "      <td>12336</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>gene60953</td>\n",
       "      <td>gene</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>gene60954</td>\n",
       "      <td>gene</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>gene60955</td>\n",
       "      <td>gene</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>gene60956</td>\n",
       "      <td>gene</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>gene60957</td>\n",
       "      <td>gene</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>gene60958</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_012920.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Id  Type  Start    End Strand     Location\n",
       "1   gene60922  gene    577    647      +  NC_012920.1\n",
       "4   gene60923  gene    648   1601      +  NC_012920.1\n",
       "7   gene60924  gene   1602   1670      +  NC_012920.1\n",
       "10  gene60925  gene   1671   3229      +  NC_012920.1\n",
       "14  gene60926  gene   3230   3304      +  NC_012920.1\n",
       "17  gene60927  gene   3307   4262      +  NC_012920.1\n",
       "19  gene60928  gene   4263   4331      +  NC_012920.1\n",
       "22  gene60929  gene   4329   4400      -  NC_012920.1\n",
       "25  gene60930  gene   4402   4469      +  NC_012920.1\n",
       "28  gene60931  gene   4470   5511      +  NC_012920.1\n",
       "30  gene60932  gene   5512   5579      +  NC_012920.1\n",
       "33  gene60933  gene   5587   5655      -  NC_012920.1\n",
       "36  gene60934  gene   5657   5729      -  NC_012920.1\n",
       "39  gene60935  gene   5761   5826      -  NC_012920.1\n",
       "42  gene60936  gene   5826   5891      -  NC_012920.1\n",
       "45  gene60937  gene   5904   7445      +  NC_012920.1\n",
       "47  gene60938  gene   7446   7514      -  NC_012920.1\n",
       "50  gene60939  gene   7518   7585      +  NC_012920.1\n",
       "53  gene60940  gene   7586   8269      +  NC_012920.1\n",
       "55  gene60941  gene   8295   8364      +  NC_012920.1\n",
       "58  gene60942  gene   8366   8572      +  NC_012920.1\n",
       "60  gene60943  gene   8527   9207      +  NC_012920.1\n",
       "62  gene60944  gene   9207   9990      +  NC_012920.1\n",
       "64  gene60945  gene   9991  10058      +  NC_012920.1\n",
       "67  gene60946  gene  10059  10404      +  NC_012920.1\n",
       "69  gene60947  gene  10405  10469      +  NC_012920.1\n",
       "72  gene60948  gene  10470  10766      +  NC_012920.1\n",
       "74  gene60949  gene  10760  12137      +  NC_012920.1\n",
       "76  gene60950  gene  12138  12206      +  NC_012920.1\n",
       "79  gene60951  gene  12207  12265      +  NC_012920.1\n",
       "82  gene60952  gene  12266  12336      +  NC_012920.1\n",
       "85  gene60953  gene  12337  14148      +  NC_012920.1\n",
       "87  gene60954  gene  14149  14673      -  NC_012920.1\n",
       "89  gene60955  gene  14674  14742      -  NC_012920.1\n",
       "92  gene60956  gene  14747  15887      +  NC_012920.1\n",
       "94  gene60957  gene  15888  15953      +  NC_012920.1\n",
       "97  gene60958  gene  15956  16023      -  NC_012920.1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_info_genes = gff_info[gff_info[\"Type\"]=='gene']\n",
    "gff_info_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.754116Z",
     "start_time": "2020-02-17T00:27:27.747810Z"
    }
   },
   "outputs": [],
   "source": [
    "#cmd = f\"bedtools getfasta -fi MT.fasta -bed {gff_filename} -name -fo MT.genes.fasta\"\n",
    "if not os.path.exists('MT.genes.fasta'):\n",
    "    cmd = 'gff2bed < MT.gff > MT.bed'\n",
    "    print(cmd)\n",
    "    os.system(cmd)\n",
    "    cmd = f\"bedtools getfasta -fi MT.fasta -bed MT.bed -s -name -fo MT.genes.fasta\"\n",
    "    print(cmd)\n",
    "    os.system(cmd) #gff_filename = data_path + 'MT.gff'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.769740Z",
     "start_time": "2020-02-17T00:27:27.756189Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene60922': SeqRecord(seq=Seq('GTTTATGTAGCTTACCTCCTCAAAGCAATACACTGAAAATGTTTAGACGGGCTC...ACA', SingleLetterAlphabet()), id='gene60922::NC_012920.1:576-647(+)', name='gene60922::NC_012920.1:576-647(+)', description='gene60922::NC_012920.1:576-647(+)', dbxrefs=[]),\n",
       " 'gene60923': SeqRecord(seq=Seq('AATAGGTTTGGTCCTAGCCTTTCTATTAGCTCTTAGTAAGATTACACATGCAAG...AAC', SingleLetterAlphabet()), id='gene60923::NC_012920.1:647-1601(+)', name='gene60923::NC_012920.1:647-1601(+)', description='gene60923::NC_012920.1:647-1601(+)', dbxrefs=[]),\n",
       " 'gene60924': SeqRecord(seq=Seq('CAGAGTGTAGCTTAACACAAAGCACCCAACTTACACTTAGGAGATTTCAACTTA...TGA', SingleLetterAlphabet()), id='gene60924::NC_012920.1:1601-1670(+)', name='gene60924::NC_012920.1:1601-1670(+)', description='gene60924::NC_012920.1:1601-1670(+)', dbxrefs=[]),\n",
       " 'gene60925': SeqRecord(seq=Seq('GCTAAACCTAGCCCCAAACCCACTCCACCTTACTACCAGACAACCTTAGCCAAA...TTT', SingleLetterAlphabet()), id='gene60925::NC_012920.1:1670-3229(+)', name='gene60925::NC_012920.1:1670-3229(+)', description='gene60925::NC_012920.1:1670-3229(+)', dbxrefs=[]),\n",
       " 'gene60926': SeqRecord(seq=Seq('GTTAAGATGGCAGAGCCCGGTAATCGCATAAAACTTAAAACTTTACAGTCAGAG...ACA', SingleLetterAlphabet()), id='gene60926::NC_012920.1:3229-3304(+)', name='gene60926::NC_012920.1:3229-3304(+)', description='gene60926::NC_012920.1:3229-3304(+)', dbxrefs=[]),\n",
       " 'gene60927': SeqRecord(seq=Seq('ATACCCATGGCCAACCTCCTACTCCTCATTGTACCCATTCTAATCGCAATGGCA...CTA', SingleLetterAlphabet()), id='gene60927::NC_012920.1:3306-4262(+)', name='gene60927::NC_012920.1:3306-4262(+)', description='gene60927::NC_012920.1:3306-4262(+)', dbxrefs=[]),\n",
       " 'gene60928': SeqRecord(seq=Seq('AGAAATATGTCTGATAAAAGAGTTACTTTGATAGAGTAAATAATAGGAGCTTAA...CTA', SingleLetterAlphabet()), id='gene60928::NC_012920.1:4262-4331(+)', name='gene60928::NC_012920.1:4262-4331(+)', description='gene60928::NC_012920.1:4262-4331(+)', dbxrefs=[]),\n",
       " 'gene60929': SeqRecord(seq=Seq('TAGGATGGGGTGTGATAGGTGGCACGGAGAATTTTGGATTCTCAGGGATGGGTT...TAG', SingleLetterAlphabet()), id='gene60929::NC_012920.1:4328-4400(-)', name='gene60929::NC_012920.1:4328-4400(-)', description='gene60929::NC_012920.1:4328-4400(-)', dbxrefs=[]),\n",
       " 'gene60930': SeqRecord(seq=Seq('AGTAAGGTCAGCTAAATAAGCTATCGGGCCCATACCCCGAAAATGTTGGTTATA...CTA', SingleLetterAlphabet()), id='gene60930::NC_012920.1:4401-4469(+)', name='gene60930::NC_012920.1:4401-4469(+)', description='gene60930::NC_012920.1:4401-4469(+)', dbxrefs=[]),\n",
       " 'gene60931': SeqRecord(seq=Seq('ATTAATCCCCTGGCCCAACCCGTCATCTACTCTACCATCTTTGCAGGCACACTC...TAT', SingleLetterAlphabet()), id='gene60931::NC_012920.1:4469-5511(+)', name='gene60931::NC_012920.1:4469-5511(+)', description='gene60931::NC_012920.1:4469-5511(+)', dbxrefs=[]),\n",
       " 'gene60932': SeqRecord(seq=Seq('AGAAATTTAGGTTAAATACAGACCAAGAGCCTTCAAAGCCCTCAGTAAGTTGCA...CTG', SingleLetterAlphabet()), id='gene60932::NC_012920.1:5511-5579(+)', name='gene60932::NC_012920.1:5511-5579(+)', description='gene60932::NC_012920.1:5511-5579(+)', dbxrefs=[]),\n",
       " 'gene60933': SeqRecord(seq=Seq('AAGGGCTTAGCTTAATTAAAGTGGCTGATTTGCGTTCAGTTGATGCAGAGTGGG...TTA', SingleLetterAlphabet()), id='gene60933::NC_012920.1:5586-5655(-)', name='gene60933::NC_012920.1:5586-5655(-)', description='gene60933::NC_012920.1:5586-5655(-)', dbxrefs=[]),\n",
       " 'gene60934': SeqRecord(seq=Seq('TAGATTGAAGCCAGTTGATTAGGGTGCTTAGCTGTTAACTAAGTGTTTGTGGGT...TAG', SingleLetterAlphabet()), id='gene60934::NC_012920.1:5656-5729(-)', name='gene60934::NC_012920.1:5656-5729(-)', description='gene60934::NC_012920.1:5656-5729(-)', dbxrefs=[]),\n",
       " 'gene60935': SeqRecord(seq=Seq('AGCTCCGAGGTGATTTTCATATTGAATTGCAAATTCGAAGAAGCAGCTTCAAAC...CTT', SingleLetterAlphabet()), id='gene60935::NC_012920.1:5760-5826(-)', name='gene60935::NC_012920.1:5760-5826(-)', description='gene60935::NC_012920.1:5760-5826(-)', dbxrefs=[]),\n",
       " 'gene60936': SeqRecord(seq=Seq('GGTAAAATGGCTGAGTGAAGCATTGGACTGTAAATCTAAAGACAGGGGTTAGGC...CCA', SingleLetterAlphabet()), id='gene60936::NC_012920.1:5825-5891(-)', name='gene60936::NC_012920.1:5825-5891(-)', description='gene60936::NC_012920.1:5825-5891(-)', dbxrefs=[]),\n",
       " 'gene60937': SeqRecord(seq=Seq('ATGTTCGCCGACCGTTGACTATTCTCTACAAACCACAAAGACATTGGAACACTA...AGA', SingleLetterAlphabet()), id='gene60937::NC_012920.1:5903-7445(+)', name='gene60937::NC_012920.1:5903-7445(+)', description='gene60937::NC_012920.1:5903-7445(+)', dbxrefs=[]),\n",
       " 'gene60938': SeqRecord(seq=Seq('GAAAAAGTCATGGAGGCCATGGGGTTGGCTTGAAACCAGCTTTGGGGGGTTCGA...TTG', SingleLetterAlphabet()), id='gene60938::NC_012920.1:7445-7514(-)', name='gene60938::NC_012920.1:7445-7514(-)', description='gene60938::NC_012920.1:7445-7514(-)', dbxrefs=[]),\n",
       " 'gene60939': SeqRecord(seq=Seq('AAGGTATTAGAAAAACCATTTCATAACTTTGTCAAAGTTAAATTATAGGCTAAA...TTA', SingleLetterAlphabet()), id='gene60939::NC_012920.1:7517-7585(+)', name='gene60939::NC_012920.1:7517-7585(+)', description='gene60939::NC_012920.1:7517-7585(+)', dbxrefs=[]),\n",
       " 'gene60940': SeqRecord(seq=Seq('ATGGCACATGCAGCGCAAGTAGGTCTACAAGACGCTACTTCCCCTATCATAGAA...TAG', SingleLetterAlphabet()), id='gene60940::NC_012920.1:7585-8269(+)', name='gene60940::NC_012920.1:7585-8269(+)', description='gene60940::NC_012920.1:7585-8269(+)', dbxrefs=[]),\n",
       " 'gene60941': SeqRecord(seq=Seq('CACTGTAAAGCTAACTTAGCATTAACCTTTTAAGTTAAAGATTAAGAGAACCAA...TGA', SingleLetterAlphabet()), id='gene60941::NC_012920.1:8294-8364(+)', name='gene60941::NC_012920.1:8294-8364(+)', description='gene60941::NC_012920.1:8294-8364(+)', dbxrefs=[]),\n",
       " 'gene60942': SeqRecord(seq=Seq('ATGCCCCAACTAAATACTACCGTATGGCCCACCATAATTACCCCCATACTCCTT...TAG', SingleLetterAlphabet()), id='gene60942::NC_012920.1:8365-8572(+)', name='gene60942::NC_012920.1:8365-8572(+)', description='gene60942::NC_012920.1:8365-8572(+)', dbxrefs=[]),\n",
       " 'gene60943': SeqRecord(seq=Seq('ATGAACGAAAATCTGTTCGCTTCATTCATTGCCCCCACAATCCTAGGCCTACCC...TAA', SingleLetterAlphabet()), id='gene60943::NC_012920.1:8526-9207(+)', name='gene60943::NC_012920.1:8526-9207(+)', description='gene60943::NC_012920.1:8526-9207(+)', dbxrefs=[]),\n",
       " 'gene60944': SeqRecord(seq=Seq('ATGACCCACCAATCACATGCCTATCATATAGTAAAACCCAGCCCATGACCCCTA...CTT', SingleLetterAlphabet()), id='gene60944::NC_012920.1:9206-9990(+)', name='gene60944::NC_012920.1:9206-9990(+)', description='gene60944::NC_012920.1:9206-9990(+)', dbxrefs=[]),\n",
       " 'gene60945': SeqRecord(seq=Seq('ACTCTTTTAGTATAAATAGTACCGTTAACTTCCAATTAACTAGTTTTGACAACA...GTA', SingleLetterAlphabet()), id='gene60945::NC_012920.1:9990-10058(+)', name='gene60945::NC_012920.1:9990-10058(+)', description='gene60945::NC_012920.1:9990-10058(+)', dbxrefs=[]),\n",
       " 'gene60946': SeqRecord(seq=Seq('ATAAACTTCGCCTTAATTTTAATAATCAACACCCTCCTAGCCTTACTACTAATA...AAT', SingleLetterAlphabet()), id='gene60946::NC_012920.1:10058-10404(+)', name='gene60946::NC_012920.1:10058-10404(+)', description='gene60946::NC_012920.1:10058-10404(+)', dbxrefs=[]),\n",
       " 'gene60947': SeqRecord(seq=Seq('TGGTATATAGTTTAAACAAAACGAATGATTTCGACTCATTAAATTATGATAATC...CAA', SingleLetterAlphabet()), id='gene60947::NC_012920.1:10404-10469(+)', name='gene60947::NC_012920.1:10404-10469(+)', description='gene60947::NC_012920.1:10404-10469(+)', dbxrefs=[]),\n",
       " 'gene60948': SeqRecord(seq=Seq('ATGCCCCTCATTTACATAAATATTATACTAGCATTTACCATCTCACTTCTAGGA...TAA', SingleLetterAlphabet()), id='gene60948::NC_012920.1:10469-10766(+)', name='gene60948::NC_012920.1:10469-10766(+)', description='gene60948::NC_012920.1:10469-10766(+)', dbxrefs=[]),\n",
       " 'gene60949': SeqRecord(seq=Seq('ATGCTAAAACTAATCGTCCCAACAATTATATTACTACCACTGACATGACTTTCC...CTT', SingleLetterAlphabet()), id='gene60949::NC_012920.1:10759-12137(+)', name='gene60949::NC_012920.1:10759-12137(+)', description='gene60949::NC_012920.1:10759-12137(+)', dbxrefs=[]),\n",
       " 'gene60950': SeqRecord(seq=Seq('GTAAATATAGTTTAACCAAAACATCAGATTGTGAATCTGACAACAGAGGCTTAC...ACC', SingleLetterAlphabet()), id='gene60950::NC_012920.1:12137-12206(+)', name='gene60950::NC_012920.1:12137-12206(+)', description='gene60950::NC_012920.1:12137-12206(+)', dbxrefs=[]),\n",
       " 'gene60951': SeqRecord(seq=Seq('GAGAAAGCTCACAAGAACTGCTAACTCATGCCCCCATGTCTAACAACATGGCTTTCTCA', SingleLetterAlphabet()), id='gene60951::NC_012920.1:12206-12265(+)', name='gene60951::NC_012920.1:12206-12265(+)', description='gene60951::NC_012920.1:12206-12265(+)', dbxrefs=[]),\n",
       " 'gene60952': SeqRecord(seq=Seq('ACTTTTAAAGGATAACAGCTATCCATTGGTCTTAGGCCCCAAAAATTTTGGTGC...GTA', SingleLetterAlphabet()), id='gene60952::NC_012920.1:12265-12336(+)', name='gene60952::NC_012920.1:12265-12336(+)', description='gene60952::NC_012920.1:12265-12336(+)', dbxrefs=[]),\n",
       " 'gene60953': SeqRecord(seq=Seq('ATAACCATGCACACTACTATAACCACCCTAACCCTGACTTCCCTAATTCCCCCC...TAA', SingleLetterAlphabet()), id='gene60953::NC_012920.1:12336-14148(+)', name='gene60953::NC_012920.1:12336-14148(+)', description='gene60953::NC_012920.1:12336-14148(+)', dbxrefs=[]),\n",
       " 'gene60954': SeqRecord(seq=Seq('ATGATGTATGCTTTGTTTCTGTTGAGTGTGGGTTTAGTAATGGGGTTTGTGGGG...AGG', SingleLetterAlphabet()), id='gene60954::NC_012920.1:14148-14673(-)', name='gene60954::NC_012920.1:14148-14673(-)', description='gene60954::NC_012920.1:14148-14673(-)', dbxrefs=[]),\n",
       " 'gene60955': SeqRecord(seq=Seq('GTTCTTGTAGTTGAAATACAACGATGGTTTTTCATATCATTGGTCGTGGTTGTA...ATA', SingleLetterAlphabet()), id='gene60955::NC_012920.1:14673-14742(-)', name='gene60955::NC_012920.1:14673-14742(-)', description='gene60955::NC_012920.1:14673-14742(-)', dbxrefs=[]),\n",
       " 'gene60956': SeqRecord(seq=Seq('ATGACCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACCACTCATTC...CCT', SingleLetterAlphabet()), id='gene60956::NC_012920.1:14746-15887(+)', name='gene60956::NC_012920.1:14746-15887(+)', description='gene60956::NC_012920.1:14746-15887(+)', dbxrefs=[]),\n",
       " 'gene60957': SeqRecord(seq=Seq('GTCCTTGTAGTATAAACTAATACACCAGTCTTGTAAACCGGAGATGAAAACCTT...ACA', SingleLetterAlphabet()), id='gene60957::NC_012920.1:15887-15953(+)', name='gene60957::NC_012920.1:15887-15953(+)', description='gene60957::NC_012920.1:15887-15953(+)', dbxrefs=[]),\n",
       " 'gene60958': SeqRecord(seq=Seq('CAGAGAATAGTTTAAATTAGAATCTTAGCTTTGGGTGCTAATGGTGGAGTTAAA...TGA', SingleLetterAlphabet()), id='gene60958::NC_012920.1:15955-16023(-)', name='gene60958::NC_012920.1:15955-16023(-)', description='gene60958::NC_012920.1:15955-16023(-)', dbxrefs=[])}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_seq = SeqIO.to_dict(SeqIO.parse('MT.genes.fasta',format='fasta'))\n",
    "\n",
    "## Clean up genes_seq names\n",
    "genes_dict = dict()\n",
    "for g in genes_seq:\n",
    "    if \"gene\" in g:\n",
    "        \n",
    "        genes_dict[g.split(\":\")[0]] = genes_seq[g]\n",
    "genes_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ensure the gff + means the mt_seq sequence and negative is reverse "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.959774Z",
     "start_time": "2020-02-17T00:27:27.771242Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Positive strand\n",
      "          Id  Type  Start  End Strand     Location\n",
      "1  gene60922  gene    577  647      +  NC_012920.1\n",
      "MT seq fasta   GTTTATGTAGCTTACCTCCTCAAAGCAATACACTGAAAATGTTTAGACGGGCTCACATCACCCCATAAACA\n",
      "MT genes fasta GTTTATGTAGCTTACCTCCTCAAAGCAATACACTGAAAATGTTTAGACGGGCTCACATCACCCCATAAACA\n",
      "          Id  Type  Start   End Strand     Location\n",
      "4  gene60923  gene    648  1601      +  NC_012920.1\n",
      "MT seq fasta   AATAGGTTTGGTCCTAGCCTTTCTATTAGCTCTTAGTAAGATTACACATGCAAGCATCCCCGTTCCAGTGAGTTCACCCTCTAAATCACCACGATCAAAAGGAACAAGCATCAAGCACGCAGCAATGCAGCTCAAAACGCTTAGCCTAGCCACACCCCCACGGGAAACAGCAGTGATTAACCTTTAGCAATAAACGAAAGTTTAACTAAGCTATACTAACCCCAGGGTTGGTCAATTTCGTGCCAGCCACCGCGGTCACACGATTAACCCAAGTCAATAGAAGCCGGCGTAAAGAGTGTTTTAGATCACCCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCCAGTTGACACAAAATAGACTACGAAAGTGGCTTTAACATATCTGAACACACAATAGCTAAGACCCAAACTGGGATTAGATACCCCACTATGCTTAGCCCTAAACCTCAACAGTTAAATCAACAAAACTGCTCGCCAGAACACTACGAGCCACAGCTTAAAACTCAAAGGACCTGGCGGTGCTTCATATCCCTCTAGAGGAGCCTGTTCTGTAATCGATAAACCCCGATCAACCTCACCACCTCTTGCTCAGCCTATATACCGCCATCTTCAGCAAACCCTGATGAAGGCTACAAAGTAAGCGCAAGTACCCACGTAAAGACGTTAGGTCAAGGTGTAGCCCATGAGGTGGCAAGAAATGGGCTACATTTTCTACCCCAGAAAACTACGATAGCCCTTATGAAACTTAAGGGTCGAAGGTGGATTTAGCAGTAAACTAAGAGTAGAGTGCTTAGTTGAACAGGGCCCTGAAGCGCGTACACACCGCCCGTCACCCTCCTCAAGTATACTTCAAAGGACATTTAACTAAAACCCCTACGCATTTATATAGAGGAGACAAGTCGTAACATGGTAAGTGTACTGGAAAGTGCACTTGGACGAAC\n",
      "MT genes fasta AATAGGTTTGGTCCTAGCCTTTCTATTAGCTCTTAGTAAGATTACACATGCAAGCATCCCCGTTCCAGTGAGTTCACCCTCTAAATCACCACGATCAAAAGGAACAAGCATCAAGCACGCAGCAATGCAGCTCAAAACGCTTAGCCTAGCCACACCCCCACGGGAAACAGCAGTGATTAACCTTTAGCAATAAACGAAAGTTTAACTAAGCTATACTAACCCCAGGGTTGGTCAATTTCGTGCCAGCCACCGCGGTCACACGATTAACCCAAGTCAATAGAAGCCGGCGTAAAGAGTGTTTTAGATCACCCCCTCCCCAATAAAGCTAAAACTCACCTGAGTTGTAAAAAACTCCAGTTGACACAAAATAGACTACGAAAGTGGCTTTAACATATCTGAACACACAATAGCTAAGACCCAAACTGGGATTAGATACCCCACTATGCTTAGCCCTAAACCTCAACAGTTAAATCAACAAAACTGCTCGCCAGAACACTACGAGCCACAGCTTAAAACTCAAAGGACCTGGCGGTGCTTCATATCCCTCTAGAGGAGCCTGTTCTGTAATCGATAAACCCCGATCAACCTCACCACCTCTTGCTCAGCCTATATACCGCCATCTTCAGCAAACCCTGATGAAGGCTACAAAGTAAGCGCAAGTACCCACGTAAAGACGTTAGGTCAAGGTGTAGCCCATGAGGTGGCAAGAAATGGGCTACATTTTCTACCCCAGAAAACTACGATAGCCCTTATGAAACTTAAGGGTCGAAGGTGGATTTAGCAGTAAACTAAGAGTAGAGTGCTTAGTTGAACAGGGCCCTGAAGCGCGTACACACCGCCCGTCACCCTCCTCAAGTATACTTCAAAGGACATTTAACTAAAACCCCTACGCATTTATATAGAGGAGACAAGTCGTAACATGGTAAGTGTACTGGAAAGTGCACTTGGACGAAC\n",
      "          Id  Type  Start   End Strand     Location\n",
      "7  gene60924  gene   1602  1670      +  NC_012920.1\n",
      "MT seq fasta   CAGAGTGTAGCTTAACACAAAGCACCCAACTTACACTTAGGAGATTTCAACTTAACTTGACCGCTCTGA\n",
      "MT genes fasta CAGAGTGTAGCTTAACACAAAGCACCCAACTTACACTTAGGAGATTTCAACTTAACTTGACCGCTCTGA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "10  gene60925  gene   1671  3229      +  NC_012920.1\n",
      "MT seq fasta   GCTAAACCTAGCCCCAAACCCACTCCACCTTACTACCAGACAACCTTAGCCAAACCATTTACCCAAATAAAGTATAGGCGATAGAAATTGAAACCTGGCGCAATAGATATAGTACCGCAAGGGAAAGATGAAAAATTATAACCAAGCATAATATAGCAAGGACTAACCCCTATACCTTCTGCATAATGAATTAACTAGAAATAACTTTGCAAGGAGAGCCAAAGCTAAGACCCCCGAAACCAGACGAGCTACCTAAGAACAGCTAAAAGAGCACACCCGTCTATGTAGCAAAATAGTGGGAAGATTTATAGGTAGAGGCGACAAACCTACCGAGCCTGGTGATAGCTGGTTGTCCAAGATAGAATCTTAGTTCAACTTTAAATTTGCCCACAGAACCCTCTAAATCCCCTTGTAAATTTAACTGTTAGTCCAAAGAGGAACAGCTCTTTGGACACTAGGAAAAAACCTTGTAGAGAGAGTAAAAAATTTAACACCCATAGTAGGCCTAAAAGCAGCCACCAATTAAGAAAGCGTTCAAGCTCAACACCCACTACCTAAAAAATCCCAAACATATAACTGAACTCCTCACACCCAATTGGACCAATCTATCACCCTATAGAAGAACTAATGTTAGTATAAGTAACATGAAAACATTCTCCTCCGCATAAGCCTGCGTCAGATTAAAACACTGAACTGACAATTAACAGCCCAATATCTACAATCAACCAACAAGTCATTATTACCCTCACTGTCAACCCAACACAGGCATGCTCATAAGGAAAGGTTAAAAAAAGTAAAAGGAACTCGGCAAATCTTACCCCGCCTGTTTACCAAAAACATCACCTCTAGCATCACCAGTATTAGAGGCACCGCCTGCCCAGTGACACATGTTTAACGGCCGCGGTACCCTAACCGTGCAAAGGTAGCATAATCACTTGTTCCTTAAATAGGGACCTGTATGAATGGCTCCACGAGGGTTCAGCTGTCTCTTACTTTTAACCAGTGAAATTGACCTGCCCGTGAAGAGGCGGGCATAACACAGCAAGACGAGAAGACCCTATGGAGCTTTAATTTATTAATGCAAACAGTACCTAACAAACCCACAGGTCCTAAACTACCAAACCTGCATTAAAAATTTCGGTTGGGGCGACCTCGGAGCAGAACCCAACCTCCGAGCAGTACATGCTAAGACTTCACCAGTCAAAGCGAACTACTATACTCAATTGATCCAATAACTTGACCAACGGAACAAGTTACCCTAGGGATAACAGCGCAATCCTATTCTAGAGTCCATATCAACAATAGGGTTTACGACCTCGATGTTGGATCAGGACATCCCGATGGTGCAGCCGCTATTAAAGGTTCGTTTGTTCAACGATTAAAGTCCTACGTGATCTGAGTTCAGACCGGAGTAATCCAGGTCGGTTTCTATCTACNTTCAAATTCCTCCCTGTACGAAAGGACAAGAGAAATAAGGCCTACTTCACAAAGCGCCTTCCCCCGTAAATGATATCATCTCAACTTAGTATTATACCCACACCCACCCAAGAACAGGGTTT\n",
      "MT genes fasta GCTAAACCTAGCCCCAAACCCACTCCACCTTACTACCAGACAACCTTAGCCAAACCATTTACCCAAATAAAGTATAGGCGATAGAAATTGAAACCTGGCGCAATAGATATAGTACCGCAAGGGAAAGATGAAAAATTATAACCAAGCATAATATAGCAAGGACTAACCCCTATACCTTCTGCATAATGAATTAACTAGAAATAACTTTGCAAGGAGAGCCAAAGCTAAGACCCCCGAAACCAGACGAGCTACCTAAGAACAGCTAAAAGAGCACACCCGTCTATGTAGCAAAATAGTGGGAAGATTTATAGGTAGAGGCGACAAACCTACCGAGCCTGGTGATAGCTGGTTGTCCAAGATAGAATCTTAGTTCAACTTTAAATTTGCCCACAGAACCCTCTAAATCCCCTTGTAAATTTAACTGTTAGTCCAAAGAGGAACAGCTCTTTGGACACTAGGAAAAAACCTTGTAGAGAGAGTAAAAAATTTAACACCCATAGTAGGCCTAAAAGCAGCCACCAATTAAGAAAGCGTTCAAGCTCAACACCCACTACCTAAAAAATCCCAAACATATAACTGAACTCCTCACACCCAATTGGACCAATCTATCACCCTATAGAAGAACTAATGTTAGTATAAGTAACATGAAAACATTCTCCTCCGCATAAGCCTGCGTCAGATTAAAACACTGAACTGACAATTAACAGCCCAATATCTACAATCAACCAACAAGTCATTATTACCCTCACTGTCAACCCAACACAGGCATGCTCATAAGGAAAGGTTAAAAAAAGTAAAAGGAACTCGGCAAATCTTACCCCGCCTGTTTACCAAAAACATCACCTCTAGCATCACCAGTATTAGAGGCACCGCCTGCCCAGTGACACATGTTTAACGGCCGCGGTACCCTAACCGTGCAAAGGTAGCATAATCACTTGTTCCTTAAATAGGGACCTGTATGAATGGCTCCACGAGGGTTCAGCTGTCTCTTACTTTTAACCAGTGAAATTGACCTGCCCGTGAAGAGGCGGGCATAACACAGCAAGACGAGAAGACCCTATGGAGCTTTAATTTATTAATGCAAACAGTACCTAACAAACCCACAGGTCCTAAACTACCAAACCTGCATTAAAAATTTCGGTTGGGGCGACCTCGGAGCAGAACCCAACCTCCGAGCAGTACATGCTAAGACTTCACCAGTCAAAGCGAACTACTATACTCAATTGATCCAATAACTTGACCAACGGAACAAGTTACCCTAGGGATAACAGCGCAATCCTATTCTAGAGTCCATATCAACAATAGGGTTTACGACCTCGATGTTGGATCAGGACATCCCGATGGTGCAGCCGCTATTAAAGGTTCGTTTGTTCAACGATTAAAGTCCTACGTGATCTGAGTTCAGACCGGAGTAATCCAGGTCGGTTTCTATCTACNTTCAAATTCCTCCCTGTACGAAAGGACAAGAGAAATAAGGCCTACTTCACAAAGCGCCTTCCCCCGTAAATGATATCATCTCAACTTAGTATTATACCCACACCCACCCAAGAACAGGGTTT\n",
      "           Id  Type  Start   End Strand     Location\n",
      "14  gene60926  gene   3230  3304      +  NC_012920.1\n",
      "MT seq fasta   GTTAAGATGGCAGAGCCCGGTAATCGCATAAAACTTAAAACTTTACAGTCAGAGGTTCAATTCCTCTTCTTAACA\n",
      "MT genes fasta GTTAAGATGGCAGAGCCCGGTAATCGCATAAAACTTAAAACTTTACAGTCAGAGGTTCAATTCCTCTTCTTAACA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "17  gene60927  gene   3307  4262      +  NC_012920.1\n",
      "MT seq fasta   ATACCCATGGCCAACCTCCTACTCCTCATTGTACCCATTCTAATCGCAATGGCATTCCTAATGCTTACCGAACGAAAAATTCTAGGCTATATACAACTACGCAAAGGCCCCAACGTTGTAGGCCCCTACGGGCTACTACAACCCTTCGCTGACGCCATAAAACTCTTCACCAAAGAGCCCCTAAAACCCGCCACATCTACCATCACCCTCTACATCACCGCCCCGACCTTAGCTCTCACCATCGCTCTTCTACTATGAACCCCCCTCCCCATACCCAACCCCCTGGTCAACCTCAACCTAGGCCTCCTATTTATTCTAGCCACCTCTAGCCTAGCCGTTTACTCAATCCTCTGATCAGGGTGAGCATCAAACTCAAACTACGCCCTGATCGGCGCACTGCGAGCAGTAGCCCAAACAATCTCATATGAAGTCACCCTAGCCATCATTCTACTATCAACATTACTAATAAGTGGCTCCTTTAACCTCTCCACCCTTATCACAACACAAGAACACCTCTGATTACTCCTGCCATCATGACCCTTGGCCATAATATGATTTATCTCCACACTAGCAGAGACCAACCGAACCCCCTTCGACCTTGCCGAAGGGGAGTCCGAACTAGTCTCAGGCTTCAACATCGAATACGCCGCAGGCCCCTTCGCCCTATTCTTCATAGCCGAATACACAAACATTATTATAATAAACACCCTCACCACTACAATCTTCCTAGGAACAACATATGACGCACTCTCCCCTGAACTCTACACAACATATTTTGTCACCAAGACCCTACTTCTAACCTCCCTGTTCTTATGAATTCGAACAGCATACCCCCGATTCCGCTACGACCAACTCATACACCTCCTATGAAAAAACTTCCTACCACTCACCCTAGCATTACTTATATGATATGTCTCCATACCCATTACAATCTCCAGCATTCCCCCTCAAACCTA\n",
      "MT genes fasta ATACCCATGGCCAACCTCCTACTCCTCATTGTACCCATTCTAATCGCAATGGCATTCCTAATGCTTACCGAACGAAAAATTCTAGGCTATATACAACTACGCAAAGGCCCCAACGTTGTAGGCCCCTACGGGCTACTACAACCCTTCGCTGACGCCATAAAACTCTTCACCAAAGAGCCCCTAAAACCCGCCACATCTACCATCACCCTCTACATCACCGCCCCGACCTTAGCTCTCACCATCGCTCTTCTACTATGAACCCCCCTCCCCATACCCAACCCCCTGGTCAACCTCAACCTAGGCCTCCTATTTATTCTAGCCACCTCTAGCCTAGCCGTTTACTCAATCCTCTGATCAGGGTGAGCATCAAACTCAAACTACGCCCTGATCGGCGCACTGCGAGCAGTAGCCCAAACAATCTCATATGAAGTCACCCTAGCCATCATTCTACTATCAACATTACTAATAAGTGGCTCCTTTAACCTCTCCACCCTTATCACAACACAAGAACACCTCTGATTACTCCTGCCATCATGACCCTTGGCCATAATATGATTTATCTCCACACTAGCAGAGACCAACCGAACCCCCTTCGACCTTGCCGAAGGGGAGTCCGAACTAGTCTCAGGCTTCAACATCGAATACGCCGCAGGCCCCTTCGCCCTATTCTTCATAGCCGAATACACAAACATTATTATAATAAACACCCTCACCACTACAATCTTCCTAGGAACAACATATGACGCACTCTCCCCTGAACTCTACACAACATATTTTGTCACCAAGACCCTACTTCTAACCTCCCTGTTCTTATGAATTCGAACAGCATACCCCCGATTCCGCTACGACCAACTCATACACCTCCTATGAAAAAACTTCCTACCACTCACCCTAGCATTACTTATATGATATGTCTCCATACCCATTACAATCTCCAGCATTCCCCCTCAAACCTA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "19  gene60928  gene   4263  4331      +  NC_012920.1\n",
      "MT seq fasta   AGAAATATGTCTGATAAAAGAGTTACTTTGATAGAGTAAATAATAGGAGCTTAAACCCCCTTATTTCTA\n",
      "MT genes fasta AGAAATATGTCTGATAAAAGAGTTACTTTGATAGAGTAAATAATAGGAGCTTAAACCCCCTTATTTCTA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "25  gene60930  gene   4402  4469      +  NC_012920.1\n",
      "MT seq fasta   AGTAAGGTCAGCTAAATAAGCTATCGGGCCCATACCCCGAAAATGTTGGTTATACCCTTCCCGTACTA\n",
      "MT genes fasta AGTAAGGTCAGCTAAATAAGCTATCGGGCCCATACCCCGAAAATGTTGGTTATACCCTTCCCGTACTA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "28  gene60931  gene   4470  5511      +  NC_012920.1\n",
      "MT seq fasta   ATTAATCCCCTGGCCCAACCCGTCATCTACTCTACCATCTTTGCAGGCACACTCATCACAGCGCTAAGCTCGCACTGATTTTTTACCTGAGTAGGCCTAGAAATAAACATGCTAGCTTTTATTCCAGTTCTAACCAAAAAAATAAACCCTCGTTCCACAGAAGCTGCCATCAAGTATTTCCTCACGCAAGCAACCGCATCCATAATCCTTCTAATAGCTATCCTCTTCAACAATATACTCTCCGGACAATGAACCATAACCAATACTACCAATCAATACTCATCATTAATAATCATAATAGCTATAGCAATAAAACTAGGAATAGCCCCCTTTCACTTCTGAGTCCCAGAGGTTACCCAAGGCACCCCTCTGACATCCGGCCTGCTTCTTCTCACATGACAAAAACTAGCCCCCATCTCAATCATATACCAAATCTCTCCCTCACTAAACGTAAGCCTTCTCCTCACTCTCTCAATCTTATCCATCATAGCAGGCAGTTGAGGTGGATTAAACCAAACCCAGCTACGCAAAATCTTAGCATACTCCTCAATTACCCACATAGGATGAATAATAGCAGTTCTACCGTACAACCCTAACATAACCATTCTTAATTTAACTATTTATATTATCCTAACTACTACCGCATTCCTACTACTCAACTTAAACTCCAGCACCACGACCCTACTACTATCTCGCACCTGAAACAAGCTAACATGACTAACACCCTTAATTCCATCCACCCTCCTCTCCCTAGGAGGCCTGCCCCCGCTAACCGGCTTTTTGCCCAAATGGGCCATTATCGAAGAATTCACAAAAAACAATAGCCTCATCATCCCCACCATCATAGCCACCATCACCCTCCTTAACCTCTACTTCTACCTACGCCTAATCTACTCCACCTCAATCACACTACTCCCCATATCTAACAACGTAAAAATAAAATGACAGTTTGAACATACAAAACCCACCCCATTCCTCCCCACACTCATCGCCCTTACCACGCTACTCCTACCTATCTCCCCTTTTATACTAATAATCTTAT\n",
      "MT genes fasta ATTAATCCCCTGGCCCAACCCGTCATCTACTCTACCATCTTTGCAGGCACACTCATCACAGCGCTAAGCTCGCACTGATTTTTTACCTGAGTAGGCCTAGAAATAAACATGCTAGCTTTTATTCCAGTTCTAACCAAAAAAATAAACCCTCGTTCCACAGAAGCTGCCATCAAGTATTTCCTCACGCAAGCAACCGCATCCATAATCCTTCTAATAGCTATCCTCTTCAACAATATACTCTCCGGACAATGAACCATAACCAATACTACCAATCAATACTCATCATTAATAATCATAATAGCTATAGCAATAAAACTAGGAATAGCCCCCTTTCACTTCTGAGTCCCAGAGGTTACCCAAGGCACCCCTCTGACATCCGGCCTGCTTCTTCTCACATGACAAAAACTAGCCCCCATCTCAATCATATACCAAATCTCTCCCTCACTAAACGTAAGCCTTCTCCTCACTCTCTCAATCTTATCCATCATAGCAGGCAGTTGAGGTGGATTAAACCAAACCCAGCTACGCAAAATCTTAGCATACTCCTCAATTACCCACATAGGATGAATAATAGCAGTTCTACCGTACAACCCTAACATAACCATTCTTAATTTAACTATTTATATTATCCTAACTACTACCGCATTCCTACTACTCAACTTAAACTCCAGCACCACGACCCTACTACTATCTCGCACCTGAAACAAGCTAACATGACTAACACCCTTAATTCCATCCACCCTCCTCTCCCTAGGAGGCCTGCCCCCGCTAACCGGCTTTTTGCCCAAATGGGCCATTATCGAAGAATTCACAAAAAACAATAGCCTCATCATCCCCACCATCATAGCCACCATCACCCTCCTTAACCTCTACTTCTACCTACGCCTAATCTACTCCACCTCAATCACACTACTCCCCATATCTAACAACGTAAAAATAAAATGACAGTTTGAACATACAAAACCCACCCCATTCCTCCCCACACTCATCGCCCTTACCACGCTACTCCTACCTATCTCCCCTTTTATACTAATAATCTTAT\n",
      "           Id  Type  Start   End Strand     Location\n",
      "30  gene60932  gene   5512  5579      +  NC_012920.1\n",
      "MT seq fasta   AGAAATTTAGGTTAAATACAGACCAAGAGCCTTCAAAGCCCTCAGTAAGTTGCAATACTTAATTTCTG\n",
      "MT genes fasta AGAAATTTAGGTTAAATACAGACCAAGAGCCTTCAAAGCCCTCAGTAAGTTGCAATACTTAATTTCTG\n",
      "           Id  Type  Start   End Strand     Location\n",
      "45  gene60937  gene   5904  7445      +  NC_012920.1\n",
      "MT seq fasta   ATGTTCGCCGACCGTTGACTATTCTCTACAAACCACAAAGACATTGGAACACTATACCTATTATTCGGCGCATGAGCTGGAGTCCTAGGCACAGCTCTAAGCCTCCTTATTCGAGCCGAGCTGGGCCAGCCAGGCAACCTTCTAGGTAACGACCACATCTACAACGTTATCGTCACAGCCCATGCATTTGTAATAATCTTCTTCATAGTAATACCCATCATAATCGGAGGCTTTGGCAACTGACTAGTTCCCCTAATAATCGGTGCCCCCGATATGGCGTTTCCCCGCATAAACAACATAAGCTTCTGACTCTTACCTCCCTCTCTCCTACTCCTGCTCGCATCTGCTATAGTGGAGGCCGGAGCAGGAACAGGTTGAACAGTCTACCCTCCCTTAGCAGGGAACTACTCCCACCCTGGAGCCTCCGTAGACCTAACCATCTTCTCCTTACACCTAGCAGGTGTCTCCTCTATCTTAGGGGCCATCAATTTCATCACAACAATTATCAATATAAAACCCCCTGCCATAACCCAATACCAAACGCCCCTCTTCGTCTGATCCGTCCTAATCACAGCAGTCCTACTTCTCCTATCTCTCCCAGTCCTAGCTGCTGGCATCACTATACTACTAACAGACCGCAACCTCAACACCACCTTCTTCGACCCCGCCGGAGGAGGAGACCCCATTCTATACCAACACCTATTCTGATTTTTCGGTCACCCTGAAGTTTATATTCTTATCCTACCAGGCTTCGGAATAATCTCCCATATTGTAACTTACTACTCCGGAAAAAAAGAACCATTTGGATACATAGGTATGGTCTGAGCTATGATATCAATTGGCTTCCTAGGGTTTATCGTGTGAGCACACCATATATTTACAGTAGGAATAGACGTAGACACACGAGCATATTTCACCTCCGCTACCATAATCATCGCTATCCCCACCGGCGTCAAAGTATTTAGCTGACTCGCCACACTCCACGGAAGCAATATGAAATGATCTGCTGCAGTGCTCTGAGCCCTAGGATTCATCTTTCTTTTCACCGTAGGTGGCCTGACTGGCATTGTATTAGCAAACTCATCACTAGACATCGTACTACACGACACGTACTACGTTGTAGCCCACTTCCACTATGTCCTATCAATAGGAGCTGTATTTGCCATCATAGGAGGCTTCATTCACTGATTTCCCCTATTCTCAGGCTACACCCTAGACCAAACCTACGCCAAAATCCATTTCACTATCATATTCATCGGCGTAAATCTAACTTTCTTCCCACAACACTTTCTCGGCCTATCCGGAATGCCCCGACGTTACTCGGACTACCCCGATGCATACACCACATGAAACATCCTATCATCTGTAGGCTCATTCATTTCTCTAACAGCAGTAATATTAATAATTTTCATGATTTGAGAAGCCTTCGCTTCGAAGCGAAAAGTCCTAATAGTAGAAGAACCCTCCATAAACCTGGAGTGACTATATGGATGCCCCCCACCCTACCACACATTCGAAGAACCCGTATACATAAAATCTAGA\n",
      "MT genes fasta ATGTTCGCCGACCGTTGACTATTCTCTACAAACCACAAAGACATTGGAACACTATACCTATTATTCGGCGCATGAGCTGGAGTCCTAGGCACAGCTCTAAGCCTCCTTATTCGAGCCGAGCTGGGCCAGCCAGGCAACCTTCTAGGTAACGACCACATCTACAACGTTATCGTCACAGCCCATGCATTTGTAATAATCTTCTTCATAGTAATACCCATCATAATCGGAGGCTTTGGCAACTGACTAGTTCCCCTAATAATCGGTGCCCCCGATATGGCGTTTCCCCGCATAAACAACATAAGCTTCTGACTCTTACCTCCCTCTCTCCTACTCCTGCTCGCATCTGCTATAGTGGAGGCCGGAGCAGGAACAGGTTGAACAGTCTACCCTCCCTTAGCAGGGAACTACTCCCACCCTGGAGCCTCCGTAGACCTAACCATCTTCTCCTTACACCTAGCAGGTGTCTCCTCTATCTTAGGGGCCATCAATTTCATCACAACAATTATCAATATAAAACCCCCTGCCATAACCCAATACCAAACGCCCCTCTTCGTCTGATCCGTCCTAATCACAGCAGTCCTACTTCTCCTATCTCTCCCAGTCCTAGCTGCTGGCATCACTATACTACTAACAGACCGCAACCTCAACACCACCTTCTTCGACCCCGCCGGAGGAGGAGACCCCATTCTATACCAACACCTATTCTGATTTTTCGGTCACCCTGAAGTTTATATTCTTATCCTACCAGGCTTCGGAATAATCTCCCATATTGTAACTTACTACTCCGGAAAAAAAGAACCATTTGGATACATAGGTATGGTCTGAGCTATGATATCAATTGGCTTCCTAGGGTTTATCGTGTGAGCACACCATATATTTACAGTAGGAATAGACGTAGACACACGAGCATATTTCACCTCCGCTACCATAATCATCGCTATCCCCACCGGCGTCAAAGTATTTAGCTGACTCGCCACACTCCACGGAAGCAATATGAAATGATCTGCTGCAGTGCTCTGAGCCCTAGGATTCATCTTTCTTTTCACCGTAGGTGGCCTGACTGGCATTGTATTAGCAAACTCATCACTAGACATCGTACTACACGACACGTACTACGTTGTAGCCCACTTCCACTATGTCCTATCAATAGGAGCTGTATTTGCCATCATAGGAGGCTTCATTCACTGATTTCCCCTATTCTCAGGCTACACCCTAGACCAAACCTACGCCAAAATCCATTTCACTATCATATTCATCGGCGTAAATCTAACTTTCTTCCCACAACACTTTCTCGGCCTATCCGGAATGCCCCGACGTTACTCGGACTACCCCGATGCATACACCACATGAAACATCCTATCATCTGTAGGCTCATTCATTTCTCTAACAGCAGTAATATTAATAATTTTCATGATTTGAGAAGCCTTCGCTTCGAAGCGAAAAGTCCTAATAGTAGAAGAACCCTCCATAAACCTGGAGTGACTATATGGATGCCCCCCACCCTACCACACATTCGAAGAACCCGTATACATAAAATCTAGA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "50  gene60939  gene   7518  7585      +  NC_012920.1\n",
      "MT seq fasta   AAGGTATTAGAAAAACCATTTCATAACTTTGTCAAAGTTAAATTATAGGCTAAATCCTATATATCTTA\n",
      "MT genes fasta AAGGTATTAGAAAAACCATTTCATAACTTTGTCAAAGTTAAATTATAGGCTAAATCCTATATATCTTA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "53  gene60940  gene   7586  8269      +  NC_012920.1\n",
      "MT seq fasta   ATGGCACATGCAGCGCAAGTAGGTCTACAAGACGCTACTTCCCCTATCATAGAAGAGCTTATCACCTTTCATGATCACGCCCTCATAATCATTTTCCTTATCTGCTTCCTAGTCCTGTATGCCCTTTTCCTAACACTCACAACAAAACTAACTAATACTAACATCTCAGACGCTCAGGAAATAGAAACCGTCTGAACTATCCTGCCCGCCATCATCCTAGTCCTCATCGCCCTCCCATCCCTACGCATCCTTTACATAACAGACGAGGTCAACGATCCCTCCCTTACCATCAAATCAATTGGCCACCAATGGTACTGAACCTACGAGTACACCGACTACGGCGGACTAATCTTCAACTCCTACATACTTCCCCCATTATTCCTAGAACCAGGCGACCTGCGACTCCTTGACGTTGACAATCGAGTAGTACTCCCGATTGAAGCCCCCATTCGTATAATAATTACATCACAAGACGTCTTGCACTCATGAGCTGTCCCCACATTAGGCTTAAAAACAGATGCAATTCCCGGACGTCTAAACCAAACCACTTTCACCGCTACACGACCGGGGGTATACTACGGTCAATGCTCTGAAATCTGTGGAGCAAACCACAGTTTCATGCCCATCGTCCTAGAATTAATTCCCCTAAAAATCTTTGAAATAGGGCCCGTATTTACCCTATAG\n",
      "MT genes fasta ATGGCACATGCAGCGCAAGTAGGTCTACAAGACGCTACTTCCCCTATCATAGAAGAGCTTATCACCTTTCATGATCACGCCCTCATAATCATTTTCCTTATCTGCTTCCTAGTCCTGTATGCCCTTTTCCTAACACTCACAACAAAACTAACTAATACTAACATCTCAGACGCTCAGGAAATAGAAACCGTCTGAACTATCCTGCCCGCCATCATCCTAGTCCTCATCGCCCTCCCATCCCTACGCATCCTTTACATAACAGACGAGGTCAACGATCCCTCCCTTACCATCAAATCAATTGGCCACCAATGGTACTGAACCTACGAGTACACCGACTACGGCGGACTAATCTTCAACTCCTACATACTTCCCCCATTATTCCTAGAACCAGGCGACCTGCGACTCCTTGACGTTGACAATCGAGTAGTACTCCCGATTGAAGCCCCCATTCGTATAATAATTACATCACAAGACGTCTTGCACTCATGAGCTGTCCCCACATTAGGCTTAAAAACAGATGCAATTCCCGGACGTCTAAACCAAACCACTTTCACCGCTACACGACCGGGGGTATACTACGGTCAATGCTCTGAAATCTGTGGAGCAAACCACAGTTTCATGCCCATCGTCCTAGAATTAATTCCCCTAAAAATCTTTGAAATAGGGCCCGTATTTACCCTATAG\n",
      "           Id  Type  Start   End Strand     Location\n",
      "55  gene60941  gene   8295  8364      +  NC_012920.1\n",
      "MT seq fasta   CACTGTAAAGCTAACTTAGCATTAACCTTTTAAGTTAAAGATTAAGAGAACCAACACCTCTTTACAGTGA\n",
      "MT genes fasta CACTGTAAAGCTAACTTAGCATTAACCTTTTAAGTTAAAGATTAAGAGAACCAACACCTCTTTACAGTGA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "58  gene60942  gene   8366  8572      +  NC_012920.1\n",
      "MT seq fasta   ATGCCCCAACTAAATACTACCGTATGGCCCACCATAATTACCCCCATACTCCTTACACTATTCCTCATCACCCAACTAAAAATATTAAACACAAACTACCACCTACCTCCCTCACCAAAGCCCATAAAAATAAAAAATTATAACAAACCCTGAGAACCAAAATGAACGAAAATCTGTTCGCTTCATTCATTGCCCCCACAATCCTAG\n",
      "MT genes fasta ATGCCCCAACTAAATACTACCGTATGGCCCACCATAATTACCCCCATACTCCTTACACTATTCCTCATCACCCAACTAAAAATATTAAACACAAACTACCACCTACCTCCCTCACCAAAGCCCATAAAAATAAAAAATTATAACAAACCCTGAGAACCAAAATGAACGAAAATCTGTTCGCTTCATTCATTGCCCCCACAATCCTAG\n",
      "           Id  Type  Start   End Strand     Location\n",
      "60  gene60943  gene   8527  9207      +  NC_012920.1\n",
      "MT seq fasta   ATGAACGAAAATCTGTTCGCTTCATTCATTGCCCCCACAATCCTAGGCCTACCCGCCGCAGTACTGATCATTCTATTTCCCCCTCTATTGATCCCCACCTCCAAATATCTCATCAACAACCGACTAATCACCACCCAACAATGACTAATCAAACTAACCTCAAAACAAATGATAACCATACACAACACTAAAGGACGAACCTGATCTCTTATACTAGTATCCTTAATCATTTTTATTGCCACAACTAACCTCCTCGGACTCCTGCCTCACTCATTTACACCAACCACCCAACTATCTATAAACCTAGCCATGGCCATCCCCTTATGAGCGGGCACAGTGATTATAGGCTTTCGCTCTAAGATTAAAAATGCCCTAGCCCACTTCTTACCACAAGGCACACCTACACCCCTTATCCCCATACTAGTTATTATCGAAACCATCAGCCTACTCATTCAACCAATAGCCCTGGCCGTACGCCTAACCGCTAACATTACTGCAGGCCACCTACTCATGCACCTAATTGGAAGCGCCACCCTAGCAATATCAACCATTAACCTTCCCTCTACACTTATCATCTTCACAATTCTAATTCTACTGACTATCCTAGAAATCGCTGTCGCCTTAATCCAAGCCTACGTTTTCACACTTCTAGTAAGCCTCTACCTGCACGACAACACATAA\n",
      "MT genes fasta ATGAACGAAAATCTGTTCGCTTCATTCATTGCCCCCACAATCCTAGGCCTACCCGCCGCAGTACTGATCATTCTATTTCCCCCTCTATTGATCCCCACCTCCAAATATCTCATCAACAACCGACTAATCACCACCCAACAATGACTAATCAAACTAACCTCAAAACAAATGATAACCATACACAACACTAAAGGACGAACCTGATCTCTTATACTAGTATCCTTAATCATTTTTATTGCCACAACTAACCTCCTCGGACTCCTGCCTCACTCATTTACACCAACCACCCAACTATCTATAAACCTAGCCATGGCCATCCCCTTATGAGCGGGCACAGTGATTATAGGCTTTCGCTCTAAGATTAAAAATGCCCTAGCCCACTTCTTACCACAAGGCACACCTACACCCCTTATCCCCATACTAGTTATTATCGAAACCATCAGCCTACTCATTCAACCAATAGCCCTGGCCGTACGCCTAACCGCTAACATTACTGCAGGCCACCTACTCATGCACCTAATTGGAAGCGCCACCCTAGCAATATCAACCATTAACCTTCCCTCTACACTTATCATCTTCACAATTCTAATTCTACTGACTATCCTAGAAATCGCTGTCGCCTTAATCCAAGCCTACGTTTTCACACTTCTAGTAAGCCTCTACCTGCACGACAACACATAA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "62  gene60944  gene   9207  9990      +  NC_012920.1\n",
      "MT seq fasta   ATGACCCACCAATCACATGCCTATCATATAGTAAAACCCAGCCCATGACCCCTAACAGGGGCCCTCTCAGCCCTCCTAATGACCTCCGGCCTAGCCATGTGATTTCACTTCCACTCCATAACGCTCCTCATACTAGGCCTACTAACCAACACACTAACCATATACCAATGATGGCGCGATGTAACACGAGAAAGCACATACCAAGGCCACCACACACCACCTGTCCAAAAAGGCCTTCGATACGGGATAATCCTATTTATTACCTCAGAAGTTTTTTTCTTCGCAGGATTTTTCTGAGCCTTTTACCACTCCAGCCTAGCCCCTACCCCCCAATTAGGAGGGCACTGGCCCCCAACAGGCATCACCCCGCTAAATCCCCTAGAAGTCCCACTCCTAAACACATCCGTATTACTCGCATCAGGAGTATCAATCACCTGAGCTCACCATAGTCTAATAGAAAACAACCGAAACCAAATAATTCAAGCACTGCTTATTACAATTTTACTGGGTCTCTATTTTACCCTCCTACAAGCCTCAGAGTACTTCGAGTCTCCCTTCACCATTTCCGACGGCATCTACGGCTCAACATTTTTTGTAGCCACAGGCTTCCACGGACTTCACGTCATTATTGGCTCAACTTTCCTCACTATCTGCTTCATCCGCCAACTAATATTTCACTTTACATCCAAACATCACTTTGGCTTCGAAGCCGCCGCCTGATACTGGCATTTTGTAGATGTGGTTTGACTATTTCTGTATGTCTCCATCTATTGATGAGGGTCTT\n",
      "MT genes fasta ATGACCCACCAATCACATGCCTATCATATAGTAAAACCCAGCCCATGACCCCTAACAGGGGCCCTCTCAGCCCTCCTAATGACCTCCGGCCTAGCCATGTGATTTCACTTCCACTCCATAACGCTCCTCATACTAGGCCTACTAACCAACACACTAACCATATACCAATGATGGCGCGATGTAACACGAGAAAGCACATACCAAGGCCACCACACACCACCTGTCCAAAAAGGCCTTCGATACGGGATAATCCTATTTATTACCTCAGAAGTTTTTTTCTTCGCAGGATTTTTCTGAGCCTTTTACCACTCCAGCCTAGCCCCTACCCCCCAATTAGGAGGGCACTGGCCCCCAACAGGCATCACCCCGCTAAATCCCCTAGAAGTCCCACTCCTAAACACATCCGTATTACTCGCATCAGGAGTATCAATCACCTGAGCTCACCATAGTCTAATAGAAAACAACCGAAACCAAATAATTCAAGCACTGCTTATTACAATTTTACTGGGTCTCTATTTTACCCTCCTACAAGCCTCAGAGTACTTCGAGTCTCCCTTCACCATTTCCGACGGCATCTACGGCTCAACATTTTTTGTAGCCACAGGCTTCCACGGACTTCACGTCATTATTGGCTCAACTTTCCTCACTATCTGCTTCATCCGCCAACTAATATTTCACTTTACATCCAAACATCACTTTGGCTTCGAAGCCGCCGCCTGATACTGGCATTTTGTAGATGTGGTTTGACTATTTCTGTATGTCTCCATCTATTGATGAGGGTCTT\n",
      "           Id  Type  Start    End Strand     Location\n",
      "64  gene60945  gene   9991  10058      +  NC_012920.1\n",
      "MT seq fasta   ACTCTTTTAGTATAAATAGTACCGTTAACTTCCAATTAACTAGTTTTGACAACATTCAAAAAAGAGTA\n",
      "MT genes fasta ACTCTTTTAGTATAAATAGTACCGTTAACTTCCAATTAACTAGTTTTGACAACATTCAAAAAAGAGTA\n",
      "           Id  Type  Start    End Strand     Location\n",
      "67  gene60946  gene  10059  10404      +  NC_012920.1\n",
      "MT seq fasta   ATAAACTTCGCCTTAATTTTAATAATCAACACCCTCCTAGCCTTACTACTAATAATTATTACATTTTGACTACCACAACTCAACGGCTACATAGAAAAATCCACCCCTTACGAGTGCGGCTTCGACCCTATATCCCCCGCCCGCGTCCCTTTCTCCATAAAATTCTTCTTAGTAGCTATTACCTTCTTATTATTTGATCTAGAAATTGCCCTCCTTTTACCCCTACCATGAGCCCTACAAACAACTAACCTGCCACTAATAGTTATGTCATCCCTCTTATTAATCATCATCCTAGCCCTAAGTCTGGCCTATGAGTGACTACAAAAAGGATTAGACTGAACCGAAT\n",
      "MT genes fasta ATAAACTTCGCCTTAATTTTAATAATCAACACCCTCCTAGCCTTACTACTAATAATTATTACATTTTGACTACCACAACTCAACGGCTACATAGAAAAATCCACCCCTTACGAGTGCGGCTTCGACCCTATATCCCCCGCCCGCGTCCCTTTCTCCATAAAATTCTTCTTAGTAGCTATTACCTTCTTATTATTTGATCTAGAAATTGCCCTCCTTTTACCCCTACCATGAGCCCTACAAACAACTAACCTGCCACTAATAGTTATGTCATCCCTCTTATTAATCATCATCCTAGCCCTAAGTCTGGCCTATGAGTGACTACAAAAAGGATTAGACTGAACCGAAT\n",
      "           Id  Type  Start    End Strand     Location\n",
      "69  gene60947  gene  10405  10469      +  NC_012920.1\n",
      "MT seq fasta   TGGTATATAGTTTAAACAAAACGAATGATTTCGACTCATTAAATTATGATAATCATATTTACCAA\n",
      "MT genes fasta TGGTATATAGTTTAAACAAAACGAATGATTTCGACTCATTAAATTATGATAATCATATTTACCAA\n",
      "           Id  Type  Start    End Strand     Location\n",
      "72  gene60948  gene  10470  10766      +  NC_012920.1\n",
      "MT seq fasta   ATGCCCCTCATTTACATAAATATTATACTAGCATTTACCATCTCACTTCTAGGAATACTAGTATATCGCTCACACCTCATATCCTCCCTACTATGCCTAGAAGGAATAATACTATCGCTGTTCATTATAGCTACTCTCATAACCCTCAACACCCACTCCCTCTTAGCCAATATTGTGCCTATTGCCATACTAGTCTTTGCCGCCTGCGAAGCAGCGGTGGGCCTAGCCCTACTAGTCTCAATCTCCAACACATATGGCCTAGACTACGTACATAACCTAAACCTACTCCAATGCTAA\n",
      "MT genes fasta ATGCCCCTCATTTACATAAATATTATACTAGCATTTACCATCTCACTTCTAGGAATACTAGTATATCGCTCACACCTCATATCCTCCCTACTATGCCTAGAAGGAATAATACTATCGCTGTTCATTATAGCTACTCTCATAACCCTCAACACCCACTCCCTCTTAGCCAATATTGTGCCTATTGCCATACTAGTCTTTGCCGCCTGCGAAGCAGCGGTGGGCCTAGCCCTACTAGTCTCAATCTCCAACACATATGGCCTAGACTACGTACATAACCTAAACCTACTCCAATGCTAA\n",
      "           Id  Type  Start    End Strand     Location\n",
      "74  gene60949  gene  10760  12137      +  NC_012920.1\n",
      "MT seq fasta   ATGCTAAAACTAATCGTCCCAACAATTATATTACTACCACTGACATGACTTTCCAAAAAACACATAATTTGAATCAACACAACCACCCACAGCCTAATTATTAGCATCATCCCTCTACTATTTTTTAACCAAATCAACAACAACCTATTTAGCTGTTCCCCAACCTTTTCCTCCGACCCCCTAACAACCCCCCTCCTAATACTAACTACCTGACTCCTACCCCTCACAATCATGGCAAGCCAACGCCACTTATCCAGTGAACCACTATCACGAAAAAAACTCTACCTCTCTATACTAATCTCCCTACAAATCTCCTTAATTATAACATTCACAGCCACAGAACTAATCATATTTTATATCTTCTTCGAAACCACACTTATCCCCACCTTGGCTATCATCACCCGATGAGGCAACCAGCCAGAACGCCTGAACGCAGGCACATACTTCCTATTCTACACCCTAGTAGGCTCCCTTCCCCTACTCATCGCACTAATTTACACTCACAACACCCTAGGCTCACTAAACATTCTACTACTCACTCTCACTGCCCAAGAACTATCAAACTCCTGAGCCAACAACTTAATATGACTAGCTTACACAATAGCTTTTATAGTAAAGATACCTCTTTACGGACTCCACTTATGACTCCCTAAAGCCCATGTCGAAGCCCCCATCGCTGGGTCAATAGTACTTGCCGCAGTACTCTTAAAACTAGGCGGCTATGGTATAATACGCCTCACACTCATTCTCAACCCCCTGACAAAACACATAGCCTACCCCTTCCTTGTACTATCCCTATGAGGCATAATTATAACAAGCTCCATCTGCCTACGACAAACAGACCTAAAATCGCTCATTGCATACTCTTCAATCAGCCACATAGCCCTCGTAGTAACAGCCATTCTCATCCAAACCCCCTGAAGCTTCACCGGCGCAGTCATTCTCATAATCGCCCACGGGCTTACATCCTCATTACTATTCTGCCTAGCAAACTCAAACTACGAACGCACTCACAGTCGCATCATAATCCTCTCTCAAGGACTTCAAACTCTACTCCCACTAATAGCTTTTTGATGACTTCTAGCAAGCCTCGCTAACCTCGCCTTACCCCCCACTATTAACCTACTGGGAGAACTCTCTGTGCTAGTAACCACGTTCTCCTGATCAAATATCACTCTCCTACTTACAGGACTCAACATACTAGTCACAGCCCTATACTCCCTCTACATATTTACCACAACACAATGGGGCTCACTCACCCACCACATTAACAACATAAAACCCTCATTCACACGAGAAAACACCCTCATGTTCATACACCTATCCCCCATTCTCCTCCTATCCCTCAACCCCGACATCATTACCGGGTTTTCCTCTT\n",
      "MT genes fasta ATGCTAAAACTAATCGTCCCAACAATTATATTACTACCACTGACATGACTTTCCAAAAAACACATAATTTGAATCAACACAACCACCCACAGCCTAATTATTAGCATCATCCCTCTACTATTTTTTAACCAAATCAACAACAACCTATTTAGCTGTTCCCCAACCTTTTCCTCCGACCCCCTAACAACCCCCCTCCTAATACTAACTACCTGACTCCTACCCCTCACAATCATGGCAAGCCAACGCCACTTATCCAGTGAACCACTATCACGAAAAAAACTCTACCTCTCTATACTAATCTCCCTACAAATCTCCTTAATTATAACATTCACAGCCACAGAACTAATCATATTTTATATCTTCTTCGAAACCACACTTATCCCCACCTTGGCTATCATCACCCGATGAGGCAACCAGCCAGAACGCCTGAACGCAGGCACATACTTCCTATTCTACACCCTAGTAGGCTCCCTTCCCCTACTCATCGCACTAATTTACACTCACAACACCCTAGGCTCACTAAACATTCTACTACTCACTCTCACTGCCCAAGAACTATCAAACTCCTGAGCCAACAACTTAATATGACTAGCTTACACAATAGCTTTTATAGTAAAGATACCTCTTTACGGACTCCACTTATGACTCCCTAAAGCCCATGTCGAAGCCCCCATCGCTGGGTCAATAGTACTTGCCGCAGTACTCTTAAAACTAGGCGGCTATGGTATAATACGCCTCACACTCATTCTCAACCCCCTGACAAAACACATAGCCTACCCCTTCCTTGTACTATCCCTATGAGGCATAATTATAACAAGCTCCATCTGCCTACGACAAACAGACCTAAAATCGCTCATTGCATACTCTTCAATCAGCCACATAGCCCTCGTAGTAACAGCCATTCTCATCCAAACCCCCTGAAGCTTCACCGGCGCAGTCATTCTCATAATCGCCCACGGGCTTACATCCTCATTACTATTCTGCCTAGCAAACTCAAACTACGAACGCACTCACAGTCGCATCATAATCCTCTCTCAAGGACTTCAAACTCTACTCCCACTAATAGCTTTTTGATGACTTCTAGCAAGCCTCGCTAACCTCGCCTTACCCCCCACTATTAACCTACTGGGAGAACTCTCTGTGCTAGTAACCACGTTCTCCTGATCAAATATCACTCTCCTACTTACAGGACTCAACATACTAGTCACAGCCCTATACTCCCTCTACATATTTACCACAACACAATGGGGCTCACTCACCCACCACATTAACAACATAAAACCCTCATTCACACGAGAAAACACCCTCATGTTCATACACCTATCCCCCATTCTCCTCCTATCCCTCAACCCCGACATCATTACCGGGTTTTCCTCTT\n",
      "           Id  Type  Start    End Strand     Location\n",
      "76  gene60950  gene  12138  12206      +  NC_012920.1\n",
      "MT seq fasta   GTAAATATAGTTTAACCAAAACATCAGATTGTGAATCTGACAACAGAGGCTTACGACCCCTTATTTACC\n",
      "MT genes fasta GTAAATATAGTTTAACCAAAACATCAGATTGTGAATCTGACAACAGAGGCTTACGACCCCTTATTTACC\n",
      "           Id  Type  Start    End Strand     Location\n",
      "79  gene60951  gene  12207  12265      +  NC_012920.1\n",
      "MT seq fasta   GAGAAAGCTCACAAGAACTGCTAACTCATGCCCCCATGTCTAACAACATGGCTTTCTCA\n",
      "MT genes fasta GAGAAAGCTCACAAGAACTGCTAACTCATGCCCCCATGTCTAACAACATGGCTTTCTCA\n",
      "           Id  Type  Start    End Strand     Location\n",
      "82  gene60952  gene  12266  12336      +  NC_012920.1\n",
      "MT seq fasta   ACTTTTAAAGGATAACAGCTATCCATTGGTCTTAGGCCCCAAAAATTTTGGTGCAACTCCAAATAAAAGTA\n",
      "MT genes fasta ACTTTTAAAGGATAACAGCTATCCATTGGTCTTAGGCCCCAAAAATTTTGGTGCAACTCCAAATAAAAGTA\n",
      "           Id  Type  Start    End Strand     Location\n",
      "85  gene60953  gene  12337  14148      +  NC_012920.1\n",
      "MT seq fasta   ATAACCATGCACACTACTATAACCACCCTAACCCTGACTTCCCTAATTCCCCCCATCCTTACCACCCTCGTTAACCCTAACAAAAAAAACTCATACCCCCATTATGTAAAATCCATTGTCGCATCCACCTTTATTATCAGTCTCTTCCCCACAACAATATTCATGTGCCTAGACCAAGAAGTTATTATCTCGAACTGACACTGAGCCACAACCCAAACAACCCAGCTCTCCCTAAGCTTCAAACTAGACTACTTCTCCATAATATTCATCCCTGTAGCATTGTTCGTTACATGGTCCATCATAGAATTCTCACTGTGATATATAAACTCAGACCCAAACATTAATCAGTTCTTCAAATATCTACTCATCTTCCTAATTACCATACTAATCTTAGTTACCGCTAACAACCTATTCCAACTGTTCATCGGCTGAGAGGGCGTAGGAATTATATCCTTCTTGCTCATCAGTTGATGATACGCCCGAGCAGATGCCAACACAGCAGCCATTCAAGCAATCCTATACAACCGTATCGGCGATATCGGTTTCATCCTCGCCTTAGCATGATTTATCCTACACTCCAACTCATGAGACCCACAACAAATAGCCCTTCTAAACGCTAATCCAAGCCTCACCCCACTACTAGGCCTCCTCCTAGCAGCAGCAGGCAAATCAGCCCAATTAGGTCTCCACCCCTGACTCCCCTCAGCCATAGAAGGCCCCACCCCAGTCTCAGCCCTACTCCACTCAAGCACTATAGTTGTAGCAGGAATCTTCTTACTCATCCGCTTCCACCCCCTAGCAGAAAATAGCCCACTAATCCAAACTCTAACACTATGCTTAGGCGCTATCACCACTCTGTTCGCAGCAGTCTGCGCCCTTACACAAAATGACATCAAAAAAATCGTAGCCTTCTCCACTTCAAGTCAACTAGGACTCATAATAGTTACAATCGGCATCAACCAACCACACCTAGCATTCCTGCACATCTGTACCCACGCCTTCTTCAAAGCCATACTATTTATGTGCTCCGGGTCCATCATCCACAACCTTAACAATGAACAAGATATTCGAAAAATAGGAGGACTACTCAAAACCATACCTCTCACTTCAACCTCCCTCACCATTGGCAGCCTAGCATTAGCAGGAATACCTTTCCTCACAGGTTTCTACTCCAAAGACCACATCATCGAAACCGCAAACATATCATACACAAACGCCTGAGCCCTATCTATTACTCTCATCGCTACCTCCCTGACAAGCGCCTATAGCACTCGAATAATTCTTCTCACCCTAACAGGTCAACCTCGCTTCCCCACCCTTACTAACATTAACGAAAATAACCCCACCCTACTAAACCCCATTAAACGCCTGGCAGCCGGAAGCCTATTCGCAGGATTTCTCATTACTAACAACATTTCCCCCGCATCCCCCTTCCAAACAACAATCCCCCTCTACCTAAAACTCACAGCCCTCGCTGTCACTTTCCTAGGACTTCTAACAGCCCTAGACCTCAACTACCTAACCAACAAACTTAAAATAAAATCCCCACTATGCACATTTTATTTCTCCAACATACTCGGATTCTACCCTAGCATCACACACCGCACAATCCCCTATCTAGGCCTTCTTACGAGCCAAAACCTGCCCCTACTCCTCCTAGACCTAACCTGACTAGAAAAGCTATTACCTAAAACAATTTCACAGCACCAAATCTCCACCTCCATCATCACCTCAACCCAAAAAGGCATAATTAAACTTTACTTCCTCTCTTTCTTCTTCCCACTCATCCTAACCCTACTCCTAATCACATAA\n",
      "MT genes fasta ATAACCATGCACACTACTATAACCACCCTAACCCTGACTTCCCTAATTCCCCCCATCCTTACCACCCTCGTTAACCCTAACAAAAAAAACTCATACCCCCATTATGTAAAATCCATTGTCGCATCCACCTTTATTATCAGTCTCTTCCCCACAACAATATTCATGTGCCTAGACCAAGAAGTTATTATCTCGAACTGACACTGAGCCACAACCCAAACAACCCAGCTCTCCCTAAGCTTCAAACTAGACTACTTCTCCATAATATTCATCCCTGTAGCATTGTTCGTTACATGGTCCATCATAGAATTCTCACTGTGATATATAAACTCAGACCCAAACATTAATCAGTTCTTCAAATATCTACTCATCTTCCTAATTACCATACTAATCTTAGTTACCGCTAACAACCTATTCCAACTGTTCATCGGCTGAGAGGGCGTAGGAATTATATCCTTCTTGCTCATCAGTTGATGATACGCCCGAGCAGATGCCAACACAGCAGCCATTCAAGCAATCCTATACAACCGTATCGGCGATATCGGTTTCATCCTCGCCTTAGCATGATTTATCCTACACTCCAACTCATGAGACCCACAACAAATAGCCCTTCTAAACGCTAATCCAAGCCTCACCCCACTACTAGGCCTCCTCCTAGCAGCAGCAGGCAAATCAGCCCAATTAGGTCTCCACCCCTGACTCCCCTCAGCCATAGAAGGCCCCACCCCAGTCTCAGCCCTACTCCACTCAAGCACTATAGTTGTAGCAGGAATCTTCTTACTCATCCGCTTCCACCCCCTAGCAGAAAATAGCCCACTAATCCAAACTCTAACACTATGCTTAGGCGCTATCACCACTCTGTTCGCAGCAGTCTGCGCCCTTACACAAAATGACATCAAAAAAATCGTAGCCTTCTCCACTTCAAGTCAACTAGGACTCATAATAGTTACAATCGGCATCAACCAACCACACCTAGCATTCCTGCACATCTGTACCCACGCCTTCTTCAAAGCCATACTATTTATGTGCTCCGGGTCCATCATCCACAACCTTAACAATGAACAAGATATTCGAAAAATAGGAGGACTACTCAAAACCATACCTCTCACTTCAACCTCCCTCACCATTGGCAGCCTAGCATTAGCAGGAATACCTTTCCTCACAGGTTTCTACTCCAAAGACCACATCATCGAAACCGCAAACATATCATACACAAACGCCTGAGCCCTATCTATTACTCTCATCGCTACCTCCCTGACAAGCGCCTATAGCACTCGAATAATTCTTCTCACCCTAACAGGTCAACCTCGCTTCCCCACCCTTACTAACATTAACGAAAATAACCCCACCCTACTAAACCCCATTAAACGCCTGGCAGCCGGAAGCCTATTCGCAGGATTTCTCATTACTAACAACATTTCCCCCGCATCCCCCTTCCAAACAACAATCCCCCTCTACCTAAAACTCACAGCCCTCGCTGTCACTTTCCTAGGACTTCTAACAGCCCTAGACCTCAACTACCTAACCAACAAACTTAAAATAAAATCCCCACTATGCACATTTTATTTCTCCAACATACTCGGATTCTACCCTAGCATCACACACCGCACAATCCCCTATCTAGGCCTTCTTACGAGCCAAAACCTGCCCCTACTCCTCCTAGACCTAACCTGACTAGAAAAGCTATTACCTAAAACAATTTCACAGCACCAAATCTCCACCTCCATCATCACCTCAACCCAAAAAGGCATAATTAAACTTTACTTCCTCTCTTTCTTCTTCCCACTCATCCTAACCCTACTCCTAATCACATAA\n",
      "           Id  Type  Start    End Strand     Location\n",
      "92  gene60956  gene  14747  15887      +  NC_012920.1\n",
      "MT seq fasta   ATGACCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACCACTCATTCATCGACCTCCCCACCCCATCCAACATCTCCGCATGATGAAACTTCGGCTCACTCCTTGGCGCCTGCCTGATCCTCCAAATCACCACAGGACTATTCCTAGCCATGCACTACTCACCAGACGCCTCAACCGCCTTTTCATCAATCGCCCACATCACTCGAGACGTAAATTATGGCTGAATCATCCGCTACCTTCACGCCAATGGCGCCTCAATATTCTTTATCTGCCTCTTCCTACACATCGGGCGAGGCCTATATTACGGATCATTTCTCTACTCAGAAACCTGAAACATCGGCATTATCCTCCTGCTTGCAACTATAGCAACAGCCTTCATAGGCTATGTCCTCCCGTGAGGCCAAATATCATTCTGAGGGGCCACAGTAATTACAAACTTACTATCCGCCATCCCATACATTGGGACAGACCTAGTTCAATGAATCTGAGGAGGCTACTCAGTAGACAGTCCCACCCTCACACGATTCTTTACCTTTCACTTCATCTTGCCCTTCATTATTGCAGCCCTAGCAACACTCCACCTCCTATTCTTGCACGAAACGGGATCAAACAACCCCCTAGGAATCACCTCCCATTCCGATAAAATCACCTTCCACCCTTACTACACAATCAAAGACGCCCTCGGCTTACTTCTCTTCCTTCTCTCCTTAATGACATTAACACTATTCTCACCAGACCTCCTAGGCGACCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCACATCAAGCCCGAATGATATTTCCTATTCGCCTACACAATTCTCCGATCCGTCCCTAACAAACTAGGAGGCGTCCTTGCCCTATTACTATCCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAACAAAGCATAATATTTCGCCCACTAAGCCAATCACTTTATTGACTCCTAGCCGCAGACCTCCTCATTCTAACCTGAATCGGAGGACAACCAGTAAGCTACCCTTTTACCATCATTGGACAAGTAGCATCCGTACTATACTTCACAACAATCCTAATCCTAATACCAACTATCTCCCTAATTGAAAACAAAATACTCAAATGGGCCT\n",
      "MT genes fasta ATGACCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACCACTCATTCATCGACCTCCCCACCCCATCCAACATCTCCGCATGATGAAACTTCGGCTCACTCCTTGGCGCCTGCCTGATCCTCCAAATCACCACAGGACTATTCCTAGCCATGCACTACTCACCAGACGCCTCAACCGCCTTTTCATCAATCGCCCACATCACTCGAGACGTAAATTATGGCTGAATCATCCGCTACCTTCACGCCAATGGCGCCTCAATATTCTTTATCTGCCTCTTCCTACACATCGGGCGAGGCCTATATTACGGATCATTTCTCTACTCAGAAACCTGAAACATCGGCATTATCCTCCTGCTTGCAACTATAGCAACAGCCTTCATAGGCTATGTCCTCCCGTGAGGCCAAATATCATTCTGAGGGGCCACAGTAATTACAAACTTACTATCCGCCATCCCATACATTGGGACAGACCTAGTTCAATGAATCTGAGGAGGCTACTCAGTAGACAGTCCCACCCTCACACGATTCTTTACCTTTCACTTCATCTTGCCCTTCATTATTGCAGCCCTAGCAACACTCCACCTCCTATTCTTGCACGAAACGGGATCAAACAACCCCCTAGGAATCACCTCCCATTCCGATAAAATCACCTTCCACCCTTACTACACAATCAAAGACGCCCTCGGCTTACTTCTCTTCCTTCTCTCCTTAATGACATTAACACTATTCTCACCAGACCTCCTAGGCGACCCAGACAATTATACCCTAGCCAACCCCTTAAACACCCCTCCCCACATCAAGCCCGAATGATATTTCCTATTCGCCTACACAATTCTCCGATCCGTCCCTAACAAACTAGGAGGCGTCCTTGCCCTATTACTATCCATCCTCATCCTAGCAATAATCCCCATCCTCCATATATCCAAACAACAAAGCATAATATTTCGCCCACTAAGCCAATCACTTTATTGACTCCTAGCCGCAGACCTCCTCATTCTAACCTGAATCGGAGGACAACCAGTAAGCTACCCTTTTACCATCATTGGACAAGTAGCATCCGTACTATACTTCACAACAATCCTAATCCTAATACCAACTATCTCCCTAATTGAAAACAAAATACTCAAATGGGCCT\n",
      "           Id  Type  Start    End Strand     Location\n",
      "94  gene60957  gene  15888  15953      +  NC_012920.1\n",
      "MT seq fasta   GTCCTTGTAGTATAAACTAATACACCAGTCTTGTAAACCGGAGATGAAAACCTTTTTCCAAGGACA\n",
      "MT genes fasta GTCCTTGTAGTATAAACTAATACACCAGTCTTGTAAACCGGAGATGAAAACCTTTTTCCAAGGACA\n",
      "Negative strand\n",
      "           Id  Type  Start   End Strand     Location\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22  gene60929  gene   4329  4400      -  NC_012920.1\n",
      "MT seq fasta   TAGGATGGGGTGTGATAGGTGGCACGGAGAATTTTGGATTCTCAGGGATGGGTTCGATTCTCATAGTCCTAG\n",
      "MT genes fasta TAGGATGGGGTGTGATAGGTGGCACGGAGAATTTTGGATTCTCAGGGATGGGTTCGATTCTCATAGTCCTAG\n",
      "           Id  Type  Start   End Strand     Location\n",
      "33  gene60933  gene   5587  5655      -  NC_012920.1\n",
      "MT seq fasta   AAGGGCTTAGCTTAATTAAAGTGGCTGATTTGCGTTCAGTTGATGCAGAGTGGGGTTTTGCAGTCCTTA\n",
      "MT genes fasta AAGGGCTTAGCTTAATTAAAGTGGCTGATTTGCGTTCAGTTGATGCAGAGTGGGGTTTTGCAGTCCTTA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "36  gene60934  gene   5657  5729      -  NC_012920.1\n",
      "MT seq fasta   TAGATTGAAGCCAGTTGATTAGGGTGCTTAGCTGTTAACTAAGTGTTTGTGGGTTTAAGTCCCATTGGTCTAG\n",
      "MT genes fasta TAGATTGAAGCCAGTTGATTAGGGTGCTTAGCTGTTAACTAAGTGTTTGTGGGTTTAAGTCCCATTGGTCTAG\n",
      "           Id  Type  Start   End Strand     Location\n",
      "39  gene60935  gene   5761  5826      -  NC_012920.1\n",
      "MT seq fasta   AGCTCCGAGGTGATTTTCATATTGAATTGCAAATTCGAAGAAGCAGCTTCAAACCTGCCGGGGCTT\n",
      "MT genes fasta AGCTCCGAGGTGATTTTCATATTGAATTGCAAATTCGAAGAAGCAGCTTCAAACCTGCCGGGGCTT\n",
      "           Id  Type  Start   End Strand     Location\n",
      "42  gene60936  gene   5826  5891      -  NC_012920.1\n",
      "MT seq fasta   GGTAAAATGGCTGAGTGAAGCATTGGACTGTAAATCTAAAGACAGGGGTTAGGCCTCTTTTTACCA\n",
      "MT genes fasta GGTAAAATGGCTGAGTGAAGCATTGGACTGTAAATCTAAAGACAGGGGTTAGGCCTCTTTTTACCA\n",
      "           Id  Type  Start   End Strand     Location\n",
      "47  gene60938  gene   7446  7514      -  NC_012920.1\n",
      "MT seq fasta   GAAAAAGTCATGGAGGCCATGGGGTTGGCTTGAAACCAGCTTTGGGGGGTTCGATTCCTTCCTTTTTTG\n",
      "MT genes fasta GAAAAAGTCATGGAGGCCATGGGGTTGGCTTGAAACCAGCTTTGGGGGGTTCGATTCCTTCCTTTTTTG\n",
      "           Id  Type  Start    End Strand     Location\n",
      "87  gene60954  gene  14149  14673      -  NC_012920.1\n",
      "MT seq fasta   ATGATGTATGCTTTGTTTCTGTTGAGTGTGGGTTTAGTAATGGGGTTTGTGGGGTTTTCTTCTAAGCCTTCTCCTATTTATGGGGGTTTAGTATTGATTGTTAGCGGTGTGGTCGGGTGTGTTATTATTCTGAATTTTGGGGGAGGTTATATGGGTTTAATAGTTTTTTTAATTTATTTAGGGGGAATGATGGTTGTCTTTGGATATACTACAGCGATGGCTATTGAGGAGTATCCTGAGGCATGGGGGTCAGGGGTTGAGGTCTTGGTGAGTGTTTTAGTGGGGTTAGCGATGGAGGTAGGATTGGTGCTGTGGGTGAAAGAGTATGATGGGGTGGTGGTTGTGGTAAACTTTAATAGTGTAGGAAGCTGAATAATTTATGAAGGAGAGGGGTCAGGGTTGATTCGGGAGGATCCTATTGGTGCGGGGGCTTTGTATGATTATGGGCGTTGATTAGTAGTAGTTACTGGTTGAACATTGTTTGTTGGTGTATATATTGTAATTGAGATTGCTCGGGGGAATAGG\n",
      "MT genes fasta ATGATGTATGCTTTGTTTCTGTTGAGTGTGGGTTTAGTAATGGGGTTTGTGGGGTTTTCTTCTAAGCCTTCTCCTATTTATGGGGGTTTAGTATTGATTGTTAGCGGTGTGGTCGGGTGTGTTATTATTCTGAATTTTGGGGGAGGTTATATGGGTTTAATAGTTTTTTTAATTTATTTAGGGGGAATGATGGTTGTCTTTGGATATACTACAGCGATGGCTATTGAGGAGTATCCTGAGGCATGGGGGTCAGGGGTTGAGGTCTTGGTGAGTGTTTTAGTGGGGTTAGCGATGGAGGTAGGATTGGTGCTGTGGGTGAAAGAGTATGATGGGGTGGTGGTTGTGGTAAACTTTAATAGTGTAGGAAGCTGAATAATTTATGAAGGAGAGGGGTCAGGGTTGATTCGGGAGGATCCTATTGGTGCGGGGGCTTTGTATGATTATGGGCGTTGATTAGTAGTAGTTACTGGTTGAACATTGTTTGTTGGTGTATATATTGTAATTGAGATTGCTCGGGGGAATAGG\n",
      "           Id  Type  Start    End Strand     Location\n",
      "89  gene60955  gene  14674  14742      -  NC_012920.1\n",
      "MT seq fasta   GTTCTTGTAGTTGAAATACAACGATGGTTTTTCATATCATTGGTCGTGGTTGTAGTCCGTGCGAGAATA\n",
      "MT genes fasta GTTCTTGTAGTTGAAATACAACGATGGTTTTTCATATCATTGGTCGTGGTTGTAGTCCGTGCGAGAATA\n",
      "           Id  Type  Start    End Strand     Location\n",
      "97  gene60958  gene  15956  16023      -  NC_012920.1\n",
      "MT seq fasta   CAGAGAATAGTTTAAATTAGAATCTTAGCTTTGGGTGCTAATGGTGGAGTTAAAGACTTTTTCTCTGA\n",
      "MT genes fasta CAGAGAATAGTTTAAATTAGAATCTTAGCTTTGGGTGCTAATGGTGGAGTTAAAGACTTTTTCTCTGA\n"
     ]
    }
   ],
   "source": [
    "print(\"Positive strand\")\n",
    "for g in genes_seq:\n",
    "    if 'gene' in g:\n",
    "        curr_gene = g.split(\":\")[0]\n",
    "        curr_gff = gff_info_genes[gff_info_genes[\"Id\"] == curr_gene]\n",
    "        if curr_gff[\"Strand\"].values[0] == '+':\n",
    "            \n",
    "            print(curr_gff)\n",
    "            print(\"MT seq fasta  \", mt_seq[int(curr_gff[\"Start\"])-1:int(curr_gff[\"End\"])])\n",
    "            print(\"MT genes fasta\", genes_seq[g].seq)\n",
    "\n",
    "print(\"Negative strand\")\n",
    "for g in genes_seq:\n",
    "    if 'gene' in g:\n",
    "        curr_gene = g.split(\":\")[0]\n",
    "        curr_gff = gff_info_genes[gff_info_genes[\"Id\"] == curr_gene]\n",
    "        if curr_gff[\"Strand\"].values[0] == '-':\n",
    "            print(curr_gff)\n",
    "            print(\"MT seq fasta  \", mt_seq_biopython[int(curr_gff[\"Start\"]-1):int(curr_gff[\"End\"])].reverse_complement())\n",
    "            print(\"MT genes fasta\", genes_seq[g].seq)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load previous primer data\n",
    "10.1186/1471-2164-10-139"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:27.995521Z",
     "start_time": "2020-02-17T00:27:27.961334Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence 5'</th>\n",
       "      <th>Corresponding fragment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>F1</td>\n",
       "      <td>CCGCTTCTGGCCACAGCACT</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>F2</td>\n",
       "      <td>GGTTGGTCAATTTCGTGCCAG</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>R1</td>\n",
       "      <td>ACTTGGGTTAATCGTGTGACC</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>F3</td>\n",
       "      <td>CATCAAGCACGCAGCAATG</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>F4</td>\n",
       "      <td>CTCACCACCTCTTGCTCAGC</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>F69</td>\n",
       "      <td>CGGCTTCGACCCTATATCC</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>R24</td>\n",
       "      <td>GGTAAAAGGAGGGCAATTTCT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>F70</td>\n",
       "      <td>CTACTCTCATAACCCTCAACACC</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>F71</td>\n",
       "      <td>ATTAAACCAGACCCAGCTACG</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>F72</td>\n",
       "      <td>AGCATATTTCACCTCCGCTAC</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Name              Sequence 5' Corresponding fragment\n",
       "4    F1     CCGCTTCTGGCCACAGCACT                      B\n",
       "5    F2    GGTTGGTCAATTTCGTGCCAG                      B\n",
       "6    R1    ACTTGGGTTAATCGTGTGACC                      B\n",
       "7    F3      CATCAAGCACGCAGCAATG                      B\n",
       "8    F4     CTCACCACCTCTTGCTCAGC                      B\n",
       "..  ...                      ...                    ...\n",
       "95  F69      CGGCTTCGACCCTATATCC                      A\n",
       "96  R24    GGTAAAAGGAGGGCAATTTCT                      A\n",
       "97  F70  CTACTCTCATAACCCTCAACACC                      A\n",
       "98  F71    ATTAAACCAGACCCAGCTACG                      A\n",
       "99  F72    AGCATATTTCACCTCCGCTAC                      A\n",
       "\n",
       "[96 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primers = pd.read_excel('96-Primers.xlsx')\n",
    "## remove the first 4, which are restriction enzymes\n",
    "primers = primers[~(primers[\"Name\"].str.contains(\"amp\"))]\n",
    "primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add to primer dataframe gene and what positions of the gene is covered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:28.512196Z",
     "start_time": "2020-02-17T00:27:27.997015Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 37/37 [00:00<00:00, 74.17it/s]\n"
     ]
    }
   ],
   "source": [
    "primers_ext = primers.copy()\n",
    "primers_ext[\"Gene\"] = \"\"\n",
    "primers_ext[\"Start\"] = -1\n",
    "primers_ext[\"End\"] = -1\n",
    "for g in tqdm(genes_dict):\n",
    "    #curr_gene = str(genes_dict[g].seq)\n",
    "    curr_gene_reverse = str(genes_dict[g].seq.reverse_complement())\n",
    "    for p_ind,p_val in primers_ext.iterrows():\n",
    "        pos = curr_gene_reverse.find(p_val[\"Sequence 5'\"])\n",
    "        if pos != -1:\n",
    "            if primers_ext.loc[p_ind,\"Gene\"] == \"\":\n",
    "                primers_ext.loc[p_ind,\"Gene\"] = g\n",
    "                primers_ext.loc[p_ind,\"Start\"] = pos\n",
    "                primers_ext.loc[p_ind,\"End\"] = pos + window\n",
    "            else:\n",
    "                if not \"Gene_B\" in primers_ext.columns.values:\n",
    "                    primers_ext[\"Gene_B\"] = \"\"\n",
    "                    primers_ext[\"Start_B\"] = -1\n",
    "                    primers_ext[\"End_B\"] = -1\n",
    "                if primers_ext.loc[p_ind,\"Gene_B\"] == \"\":\n",
    "                    primers_ext.loc[p_ind,\"Start_B\"] = pos\n",
    "                    primers_ext.loc[p_ind,\"End_B\"] = pos + window\n",
    "                else:\n",
    "                    print(\"More than 2 genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:28.529425Z",
     "start_time": "2020-02-17T00:27:28.513999Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence 5'</th>\n",
       "      <th>Corresponding fragment</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>R1</td>\n",
       "      <td>ACTTGGGTTAATCGTGTGACC</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>680</td>\n",
       "      <td>830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>R2</td>\n",
       "      <td>CTGTTTGTCGTAGGCAGATGG</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60949</td>\n",
       "      <td>537</td>\n",
       "      <td>687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>R3</td>\n",
       "      <td>GATATCGCCGATACGGTTG</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>1272</td>\n",
       "      <td>1422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>R4</td>\n",
       "      <td>AGCGGATGAGTAAGAAGATTCC</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>1025</td>\n",
       "      <td>1175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>R5</td>\n",
       "      <td>TTGAAGAAGGCGTGGGTACAG</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>805</td>\n",
       "      <td>955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>F21</td>\n",
       "      <td>TCCAAAGACAACCATCATTCC</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60954</td>\n",
       "      <td>321</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>R6</td>\n",
       "      <td>TTATCGGAATGGGAGGTGATTC</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60956</td>\n",
       "      <td>491</td>\n",
       "      <td>641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>R8</td>\n",
       "      <td>TCATAAGGGCTATCGTAGTTTTC</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>199</td>\n",
       "      <td>349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>R9</td>\n",
       "      <td>GCTGTGTTGGCATCTGCTC</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>1312</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>F33</td>\n",
       "      <td>CCTTCATAAATTATTCAGCTTCCT</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60954</td>\n",
       "      <td>139</td>\n",
       "      <td>289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>F34</td>\n",
       "      <td>CAATGATATGAAAAACCATCGTT</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60955</td>\n",
       "      <td>27</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>R10</td>\n",
       "      <td>GGATGGCGGATAGTAAGTTTGT</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60956</td>\n",
       "      <td>681</td>\n",
       "      <td>831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>F36</td>\n",
       "      <td>AATGGTTTGGCTAAGGTTGT</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60925</td>\n",
       "      <td>1500</td>\n",
       "      <td>1650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>R11</td>\n",
       "      <td>ACGAACAATGCTACAGGGATG</td>\n",
       "      <td>B</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>1525</td>\n",
       "      <td>1675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>R12</td>\n",
       "      <td>ATGTCCTGATCCAACATCGAG</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60925</td>\n",
       "      <td>223</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>R13</td>\n",
       "      <td>AGAAGAGCGATGGTGAGAGC</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60927</td>\n",
       "      <td>705</td>\n",
       "      <td>855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>F46</td>\n",
       "      <td>TAAGCACCCTAATCAACTGGC</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60934</td>\n",
       "      <td>43</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>R15</td>\n",
       "      <td>ATAGTGATGCCAGCAGCTAGG</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60937</td>\n",
       "      <td>919</td>\n",
       "      <td>1069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>R16</td>\n",
       "      <td>TTTCATGTGGTGTATGCATCG</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60937</td>\n",
       "      <td>190</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>R17</td>\n",
       "      <td>GGGCAGGATAGTTCAGACGG</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60940</td>\n",
       "      <td>477</td>\n",
       "      <td>627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>R19</td>\n",
       "      <td>GTGGCGCTTCCAATTAGGTG</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60943</td>\n",
       "      <td>148</td>\n",
       "      <td>298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>R20</td>\n",
       "      <td>GTGCTTTCTCGTGTTACATCG</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60944</td>\n",
       "      <td>587</td>\n",
       "      <td>737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>R21</td>\n",
       "      <td>GAAAGTTGAGCCAATAATGACG</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60944</td>\n",
       "      <td>142</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>F62</td>\n",
       "      <td>ATATGAAAATCACCTCGGAGC</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60935</td>\n",
       "      <td>44</td>\n",
       "      <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>R22</td>\n",
       "      <td>AGTTACAATATGGGAGATTATTCC</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60937</td>\n",
       "      <td>765</td>\n",
       "      <td>915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>R23</td>\n",
       "      <td>ATCTGTTTTTAAGCCTAATGTGG</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60940</td>\n",
       "      <td>165</td>\n",
       "      <td>315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>R24</td>\n",
       "      <td>GGTAAAAGGAGGGCAATTTCT</td>\n",
       "      <td>A</td>\n",
       "      <td>gene60946</td>\n",
       "      <td>125</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Name               Sequence 5' Corresponding fragment       Gene  Start  \\\n",
       "6    R1     ACTTGGGTTAATCGTGTGACC                      B  gene60923    680   \n",
       "14   R2     CTGTTTGTCGTAGGCAGATGG                      B  gene60949    537   \n",
       "20   R3       GATATCGCCGATACGGTTG                      B  gene60953   1272   \n",
       "21   R4    AGCGGATGAGTAAGAAGATTCC                      B  gene60953   1025   \n",
       "22   R5     TTGAAGAAGGCGTGGGTACAG                      B  gene60953    805   \n",
       "29  F21     TCCAAAGACAACCATCATTCC                      B  gene60954    321   \n",
       "30   R6    TTATCGGAATGGGAGGTGATTC                      B  gene60956    491   \n",
       "36   R8   TCATAAGGGCTATCGTAGTTTTC                      B  gene60923    199   \n",
       "42   R9       GCTGTGTTGGCATCTGCTC                      B  gene60953   1312   \n",
       "45  F33  CCTTCATAAATTATTCAGCTTCCT                      B  gene60954    139   \n",
       "46  F34   CAATGATATGAAAAACCATCGTT                      B  gene60955     27   \n",
       "47  R10    GGATGGCGGATAGTAAGTTTGT                      B  gene60956    681   \n",
       "49  F36      AATGGTTTGGCTAAGGTTGT                      B  gene60925   1500   \n",
       "50  R11     ACGAACAATGCTACAGGGATG                      B  gene60953   1525   \n",
       "52  R12     ATGTCCTGATCCAACATCGAG                      A  gene60925    223   \n",
       "53  R13      AGAAGAGCGATGGTGAGAGC                      A  gene60927    705   \n",
       "63  F46     TAAGCACCCTAATCAACTGGC                      A  gene60934     43   \n",
       "64  R15     ATAGTGATGCCAGCAGCTAGG                      A  gene60937    919   \n",
       "66  R16     TTTCATGTGGTGTATGCATCG                      A  gene60937    190   \n",
       "69  R17      GGGCAGGATAGTTCAGACGG                      A  gene60940    477   \n",
       "74  R19      GTGGCGCTTCCAATTAGGTG                      A  gene60943    148   \n",
       "75  R20     GTGCTTTCTCGTGTTACATCG                      A  gene60944    587   \n",
       "76  R21    GAAAGTTGAGCCAATAATGACG                      A  gene60944    142   \n",
       "86  F62     ATATGAAAATCACCTCGGAGC                      A  gene60935     44   \n",
       "87  R22  AGTTACAATATGGGAGATTATTCC                      A  gene60937    765   \n",
       "90  R23   ATCTGTTTTTAAGCCTAATGTGG                      A  gene60940    165   \n",
       "96  R24     GGTAAAAGGAGGGCAATTTCT                      A  gene60946    125   \n",
       "\n",
       "     End  \n",
       "6    830  \n",
       "14   687  \n",
       "20  1422  \n",
       "21  1175  \n",
       "22   955  \n",
       "29   471  \n",
       "30   641  \n",
       "36   349  \n",
       "42  1462  \n",
       "45   289  \n",
       "46   177  \n",
       "47   831  \n",
       "49  1650  \n",
       "50  1675  \n",
       "52   373  \n",
       "53   855  \n",
       "63   193  \n",
       "64  1069  \n",
       "66   340  \n",
       "69   627  \n",
       "74   298  \n",
       "75   737  \n",
       "76   292  \n",
       "86   194  \n",
       "87   915  \n",
       "90   315  \n",
       "96   275  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(primers_ext[\"Gene\"] != \"\").sum()\n",
    "primers_ext[primers_ext[\"Gene\"] != \"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:28.540200Z",
     "start_time": "2020-02-17T00:27:28.530972Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(primers_ext[\"Gene\"] == '').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:28.550858Z",
     "start_time": "2020-02-17T00:27:28.541992Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(primers_ext[\"Gene\"] != '').sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine the region of genes that are covered or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:28.572200Z",
     "start_time": "2020-02-17T00:27:28.552666Z"
    }
   },
   "outputs": [],
   "source": [
    "def cover_regions(primers,genes_dict,frag_size= 50 ):\n",
    "    print(\"here\")\n",
    "    genes_covered = defaultdict(list)\n",
    "    \n",
    "    # For each gene, determine which regions are covered by primer, and where they start and stop\n",
    "    for g in genes_dict:\n",
    "        genes_covered[g] = []\n",
    "        matches = primers[primers[\"Gene\"] == g]\n",
    "        for m_ind, m_val in matches.iterrows():\n",
    "            genes_covered[g].append((m_val[\"Start\"],m_val[\"End\"]))\n",
    "\n",
    "    # For each gene create a boolean vector for each position if it is covered or not\n",
    "    gene_positions = dict()\n",
    "    for g in genes_dict:\n",
    "        gene_positions[g] = [False]*len(genes_dict[g].seq)\n",
    "        for regions in genes_covered[g]:\n",
    "            gene_positions[g][regions[0]:regions[1]] = [True]*(regions[1]-regions[0])    \n",
    "\n",
    "    #Determine the uncovered regoins \n",
    "    genes_uncovered = defaultdict(list)\n",
    "    for g in gene_positions:\n",
    "        curr_start = 0\n",
    "        curr_bool = gene_positions[g][0]\n",
    "        for ind,val in enumerate(gene_positions[g]):        \n",
    "            if not val == curr_bool:\n",
    "                if curr_bool == False:\n",
    "                    genes_uncovered[g].append((curr_start,ind))\n",
    "                curr_bool = val\n",
    "                curr_start = ind\n",
    "        #If end is uncovered\n",
    "        if curr_bool == False:\n",
    "            genes_uncovered[g].append((curr_start,len(gene_positions[g])))\n",
    "    \n",
    "    print(\"Regions not covered\")\n",
    "    count = 0\n",
    "    for g in genes_uncovered:\n",
    "        print(g)\n",
    "        for p in genes_uncovered[g]:\n",
    "            print(p)\n",
    "            print(\"Fragment size:\", p[1]-p[0])\n",
    "            count += 1\n",
    "    print(f\"Number of uncovered regions: {count}\")\n",
    "    \n",
    "    ## Add primers in uncovered regions at the beginning of each gene\n",
    "    new_primers = dict()\n",
    "    for g in genes_uncovered:\n",
    "        curr_seq_reverse = str(genes_dict[g].seq.reverse_complement())\n",
    "        for r in genes_uncovered[g]:\n",
    "            if r[1]-r[0] >= frag_size:\n",
    "                new_primers[f\"{g}_{r[0]}\"] = curr_seq_reverse[r[0]:r[0]+20]\n",
    "\n",
    "    \n",
    "    final_primers = primers[primers[\"Gene\"] != \"\"]\n",
    "    for p in new_primers:\n",
    "        final_primers = pd.concat((final_primers, \n",
    "                                   pd.DataFrame({\"Sequence 5'\":new_primers[p], \"Gene\":p.split(\"_\")[0],\"Name\":p,\n",
    "                                                \"Start\":int(p.split(\"_\")[1]), \"End\":int(p.split(\"_\")[1])+window},\n",
    "                                                index=[p])),sort=False)\n",
    "    \n",
    "    \n",
    "    if len(primers) == len(final_primers):\n",
    "        return final_primers, genes_uncovered\n",
    "    else: # If there is more to change, run again. \n",
    "        return cover_regions(final_primers,genes_dict,frag_size= 50 )\n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.371845Z",
     "start_time": "2020-02-17T00:27:28.573929Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "here\n",
      "Regions not covered\n",
      "gene60922\n",
      "(0, 71)\n",
      "Fragment size: 71\n",
      "gene60923\n",
      "(0, 199)\n",
      "Fragment size: 199\n",
      "(349, 680)\n",
      "Fragment size: 331\n",
      "(830, 954)\n",
      "Fragment size: 124\n",
      "gene60924\n",
      "(0, 69)\n",
      "Fragment size: 69\n",
      "gene60925\n",
      "(0, 223)\n",
      "Fragment size: 223\n",
      "(373, 1500)\n",
      "Fragment size: 1127\n",
      "gene60926\n",
      "(0, 75)\n",
      "Fragment size: 75\n",
      "gene60927\n",
      "(0, 705)\n",
      "Fragment size: 705\n",
      "(855, 956)\n",
      "Fragment size: 101\n",
      "gene60928\n",
      "(0, 69)\n",
      "Fragment size: 69\n",
      "gene60929\n",
      "(0, 72)\n",
      "Fragment size: 72\n",
      "gene60930\n",
      "(0, 68)\n",
      "Fragment size: 68\n",
      "gene60931\n",
      "(0, 1042)\n",
      "Fragment size: 1042\n",
      "gene60932\n",
      "(0, 68)\n",
      "Fragment size: 68\n",
      "gene60933\n",
      "(0, 69)\n",
      "Fragment size: 69\n",
      "gene60934\n",
      "(0, 43)\n",
      "Fragment size: 43\n",
      "gene60935\n",
      "(0, 44)\n",
      "Fragment size: 44\n",
      "gene60936\n",
      "(0, 66)\n",
      "Fragment size: 66\n",
      "gene60937\n",
      "(0, 190)\n",
      "Fragment size: 190\n",
      "(340, 765)\n",
      "Fragment size: 425\n",
      "(915, 919)\n",
      "Fragment size: 4\n",
      "(1069, 1542)\n",
      "Fragment size: 473\n",
      "gene60938\n",
      "(0, 69)\n",
      "Fragment size: 69\n",
      "gene60939\n",
      "(0, 68)\n",
      "Fragment size: 68\n",
      "gene60940\n",
      "(0, 165)\n",
      "Fragment size: 165\n",
      "(315, 477)\n",
      "Fragment size: 162\n",
      "(627, 684)\n",
      "Fragment size: 57\n",
      "gene60941\n",
      "(0, 70)\n",
      "Fragment size: 70\n",
      "gene60942\n",
      "(0, 207)\n",
      "Fragment size: 207\n",
      "gene60943\n",
      "(0, 148)\n",
      "Fragment size: 148\n",
      "(298, 681)\n",
      "Fragment size: 383\n",
      "gene60944\n",
      "(0, 142)\n",
      "Fragment size: 142\n",
      "(292, 587)\n",
      "Fragment size: 295\n",
      "(737, 784)\n",
      "Fragment size: 47\n",
      "gene60945\n",
      "(0, 68)\n",
      "Fragment size: 68\n",
      "gene60946\n",
      "(0, 125)\n",
      "Fragment size: 125\n",
      "(275, 346)\n",
      "Fragment size: 71\n",
      "gene60947\n",
      "(0, 65)\n",
      "Fragment size: 65\n",
      "gene60948\n",
      "(0, 297)\n",
      "Fragment size: 297\n",
      "gene60949\n",
      "(0, 537)\n",
      "Fragment size: 537\n",
      "(687, 1378)\n",
      "Fragment size: 691\n",
      "gene60950\n",
      "(0, 69)\n",
      "Fragment size: 69\n",
      "gene60951\n",
      "(0, 59)\n",
      "Fragment size: 59\n",
      "gene60952\n",
      "(0, 71)\n",
      "Fragment size: 71\n",
      "gene60953\n",
      "(0, 805)\n",
      "Fragment size: 805\n",
      "(955, 1025)\n",
      "Fragment size: 70\n",
      "(1175, 1272)\n",
      "Fragment size: 97\n",
      "(1462, 1525)\n",
      "Fragment size: 63\n",
      "(1675, 1812)\n",
      "Fragment size: 137\n",
      "gene60954\n",
      "(0, 139)\n",
      "Fragment size: 139\n",
      "(289, 321)\n",
      "Fragment size: 32\n",
      "(471, 525)\n",
      "Fragment size: 54\n",
      "gene60955\n",
      "(0, 27)\n",
      "Fragment size: 27\n",
      "gene60956\n",
      "(0, 491)\n",
      "Fragment size: 491\n",
      "(641, 681)\n",
      "Fragment size: 40\n",
      "(831, 1141)\n",
      "Fragment size: 310\n",
      "gene60957\n",
      "(0, 66)\n",
      "Fragment size: 66\n",
      "gene60958\n",
      "(0, 68)\n",
      "Fragment size: 68\n",
      "Number of uncovered regions: 59\n",
      "here\n",
      "Regions not covered\n",
      "gene60923\n",
      "(150, 199)\n",
      "Fragment size: 49\n",
      "(499, 680)\n",
      "Fragment size: 181\n",
      "gene60925\n",
      "(150, 223)\n",
      "Fragment size: 73\n",
      "(523, 1500)\n",
      "Fragment size: 977\n",
      "gene60927\n",
      "(150, 705)\n",
      "Fragment size: 555\n",
      "gene60931\n",
      "(150, 1042)\n",
      "Fragment size: 892\n",
      "gene60934\n",
      "(0, 43)\n",
      "Fragment size: 43\n",
      "gene60935\n",
      "(0, 44)\n",
      "Fragment size: 44\n",
      "gene60937\n",
      "(150, 190)\n",
      "Fragment size: 40\n",
      "(490, 765)\n",
      "Fragment size: 275\n",
      "(915, 919)\n",
      "Fragment size: 4\n",
      "(1219, 1542)\n",
      "Fragment size: 323\n",
      "gene60940\n",
      "(150, 165)\n",
      "Fragment size: 15\n",
      "(465, 477)\n",
      "Fragment size: 12\n",
      "gene60942\n",
      "(150, 207)\n",
      "Fragment size: 57\n",
      "gene60943\n",
      "(448, 681)\n",
      "Fragment size: 233\n",
      "gene60944\n",
      "(442, 587)\n",
      "Fragment size: 145\n",
      "(737, 784)\n",
      "Fragment size: 47\n",
      "gene60948\n",
      "(150, 297)\n",
      "Fragment size: 147\n",
      "gene60949\n",
      "(150, 537)\n",
      "Fragment size: 387\n",
      "(837, 1378)\n",
      "Fragment size: 541\n",
      "gene60953\n",
      "(150, 805)\n",
      "Fragment size: 655\n",
      "gene60954\n",
      "(289, 321)\n",
      "Fragment size: 32\n",
      "gene60955\n",
      "(0, 27)\n",
      "Fragment size: 27\n",
      "gene60956\n",
      "(150, 491)\n",
      "Fragment size: 341\n",
      "(641, 681)\n",
      "Fragment size: 40\n",
      "(981, 1141)\n",
      "Fragment size: 160\n",
      "Number of uncovered regions: 27\n",
      "here\n",
      "Regions not covered\n",
      "gene60923\n",
      "(150, 199)\n",
      "Fragment size: 49\n",
      "(649, 680)\n",
      "Fragment size: 31\n",
      "gene60925\n",
      "(673, 1500)\n",
      "Fragment size: 827\n",
      "gene60927\n",
      "(300, 705)\n",
      "Fragment size: 405\n",
      "gene60931\n",
      "(300, 1042)\n",
      "Fragment size: 742\n",
      "gene60934\n",
      "(0, 43)\n",
      "Fragment size: 43\n",
      "gene60935\n",
      "(0, 44)\n",
      "Fragment size: 44\n",
      "gene60937\n",
      "(150, 190)\n",
      "Fragment size: 40\n",
      "(640, 765)\n",
      "Fragment size: 125\n",
      "(915, 919)\n",
      "Fragment size: 4\n",
      "(1369, 1542)\n",
      "Fragment size: 173\n",
      "gene60940\n",
      "(150, 165)\n",
      "Fragment size: 15\n",
      "(465, 477)\n",
      "Fragment size: 12\n",
      "gene60943\n",
      "(598, 681)\n",
      "Fragment size: 83\n",
      "gene60944\n",
      "(737, 784)\n",
      "Fragment size: 47\n",
      "gene60949\n",
      "(300, 537)\n",
      "Fragment size: 237\n",
      "(987, 1378)\n",
      "Fragment size: 391\n",
      "gene60953\n",
      "(300, 805)\n",
      "Fragment size: 505\n",
      "gene60954\n",
      "(289, 321)\n",
      "Fragment size: 32\n",
      "gene60955\n",
      "(0, 27)\n",
      "Fragment size: 27\n",
      "gene60956\n",
      "(300, 491)\n",
      "Fragment size: 191\n",
      "(641, 681)\n",
      "Fragment size: 40\n",
      "(1131, 1141)\n",
      "Fragment size: 10\n",
      "Number of uncovered regions: 23\n",
      "here\n",
      "Regions not covered\n",
      "gene60923\n",
      "(150, 199)\n",
      "Fragment size: 49\n",
      "(649, 680)\n",
      "Fragment size: 31\n",
      "gene60925\n",
      "(823, 1500)\n",
      "Fragment size: 677\n",
      "gene60927\n",
      "(450, 705)\n",
      "Fragment size: 255\n",
      "gene60931\n",
      "(450, 1042)\n",
      "Fragment size: 592\n",
      "gene60934\n",
      "(0, 43)\n",
      "Fragment size: 43\n",
      "gene60935\n",
      "(0, 44)\n",
      "Fragment size: 44\n",
      "gene60937\n",
      "(150, 190)\n",
      "Fragment size: 40\n",
      "(915, 919)\n",
      "Fragment size: 4\n",
      "(1519, 1542)\n",
      "Fragment size: 23\n",
      "gene60940\n",
      "(150, 165)\n",
      "Fragment size: 15\n",
      "(465, 477)\n",
      "Fragment size: 12\n",
      "gene60944\n",
      "(737, 784)\n",
      "Fragment size: 47\n",
      "gene60949\n",
      "(450, 537)\n",
      "Fragment size: 87\n",
      "(1137, 1378)\n",
      "Fragment size: 241\n",
      "gene60953\n",
      "(450, 805)\n",
      "Fragment size: 355\n",
      "gene60954\n",
      "(289, 321)\n",
      "Fragment size: 32\n",
      "gene60955\n",
      "(0, 27)\n",
      "Fragment size: 27\n",
      "gene60956\n",
      "(450, 491)\n",
      "Fragment size: 41\n",
      "(641, 681)\n",
      "Fragment size: 40\n",
      "(1131, 1141)\n",
      "Fragment size: 10\n",
      "Number of uncovered regions: 21\n",
      "here\n",
      "Regions not covered\n",
      "gene60923\n",
      "(150, 199)\n",
      "Fragment size: 49\n",
      "(649, 680)\n",
      "Fragment size: 31\n",
      "gene60925\n",
      "(973, 1500)\n",
      "Fragment size: 527\n",
      "gene60927\n",
      "(600, 705)\n",
      "Fragment size: 105\n",
      "gene60931\n",
      "(600, 1042)\n",
      "Fragment size: 442\n",
      "gene60934\n",
      "(0, 43)\n",
      "Fragment size: 43\n",
      "gene60935\n",
      "(0, 44)\n",
      "Fragment size: 44\n",
      "gene60937\n",
      "(150, 190)\n",
      "Fragment size: 40\n",
      "(915, 919)\n",
      "Fragment size: 4\n",
      "(1519, 1542)\n",
      "Fragment size: 23\n",
      "gene60940\n",
      "(150, 165)\n",
      "Fragment size: 15\n",
      "(465, 477)\n",
      "Fragment size: 12\n",
      "gene60944\n",
      "(737, 784)\n",
      "Fragment size: 47\n",
      "gene60949\n",
      "(1287, 1378)\n",
      "Fragment size: 91\n",
      "gene60953\n",
      "(600, 805)\n",
      "Fragment size: 205\n",
      "gene60954\n",
      "(289, 321)\n",
      "Fragment size: 32\n",
      "gene60955\n",
      "(0, 27)\n",
      "Fragment size: 27\n",
      "gene60956\n",
      "(450, 491)\n",
      "Fragment size: 41\n",
      "(641, 681)\n",
      "Fragment size: 40\n",
      "(1131, 1141)\n",
      "Fragment size: 10\n",
      "Number of uncovered regions: 20\n",
      "here\n",
      "Regions not covered\n",
      "gene60923\n",
      "(150, 199)\n",
      "Fragment size: 49\n",
      "(649, 680)\n",
      "Fragment size: 31\n",
      "gene60925\n",
      "(1123, 1500)\n",
      "Fragment size: 377\n",
      "gene60931\n",
      "(750, 1042)\n",
      "Fragment size: 292\n",
      "gene60934\n",
      "(0, 43)\n",
      "Fragment size: 43\n",
      "gene60935\n",
      "(0, 44)\n",
      "Fragment size: 44\n",
      "gene60937\n",
      "(150, 190)\n",
      "Fragment size: 40\n",
      "(915, 919)\n",
      "Fragment size: 4\n",
      "(1519, 1542)\n",
      "Fragment size: 23\n",
      "gene60940\n",
      "(150, 165)\n",
      "Fragment size: 15\n",
      "(465, 477)\n",
      "Fragment size: 12\n",
      "gene60944\n",
      "(737, 784)\n",
      "Fragment size: 47\n",
      "gene60953\n",
      "(750, 805)\n",
      "Fragment size: 55\n",
      "gene60954\n",
      "(289, 321)\n",
      "Fragment size: 32\n",
      "gene60955\n",
      "(0, 27)\n",
      "Fragment size: 27\n",
      "gene60956\n",
      "(450, 491)\n",
      "Fragment size: 41\n",
      "(641, 681)\n",
      "Fragment size: 40\n",
      "(1131, 1141)\n",
      "Fragment size: 10\n",
      "Number of uncovered regions: 18\n",
      "here\n",
      "Regions not covered\n",
      "gene60923\n",
      "(150, 199)\n",
      "Fragment size: 49\n",
      "(649, 680)\n",
      "Fragment size: 31\n",
      "gene60925\n",
      "(1273, 1500)\n",
      "Fragment size: 227\n",
      "gene60931\n",
      "(900, 1042)\n",
      "Fragment size: 142\n",
      "gene60934\n",
      "(0, 43)\n",
      "Fragment size: 43\n",
      "gene60935\n",
      "(0, 44)\n",
      "Fragment size: 44\n",
      "gene60937\n",
      "(150, 190)\n",
      "Fragment size: 40\n",
      "(915, 919)\n",
      "Fragment size: 4\n",
      "(1519, 1542)\n",
      "Fragment size: 23\n",
      "gene60940\n",
      "(150, 165)\n",
      "Fragment size: 15\n",
      "(465, 477)\n",
      "Fragment size: 12\n",
      "gene60944\n",
      "(737, 784)\n",
      "Fragment size: 47\n",
      "gene60954\n",
      "(289, 321)\n",
      "Fragment size: 32\n",
      "gene60955\n",
      "(0, 27)\n",
      "Fragment size: 27\n",
      "gene60956\n",
      "(450, 491)\n",
      "Fragment size: 41\n",
      "(641, 681)\n",
      "Fragment size: 40\n",
      "(1131, 1141)\n",
      "Fragment size: 10\n",
      "Number of uncovered regions: 17\n",
      "here\n",
      "Regions not covered\n",
      "gene60923\n",
      "(150, 199)\n",
      "Fragment size: 49\n",
      "(649, 680)\n",
      "Fragment size: 31\n",
      "gene60925\n",
      "(1423, 1500)\n",
      "Fragment size: 77\n",
      "gene60934\n",
      "(0, 43)\n",
      "Fragment size: 43\n",
      "gene60935\n",
      "(0, 44)\n",
      "Fragment size: 44\n",
      "gene60937\n",
      "(150, 190)\n",
      "Fragment size: 40\n",
      "(915, 919)\n",
      "Fragment size: 4\n",
      "(1519, 1542)\n",
      "Fragment size: 23\n",
      "gene60940\n",
      "(150, 165)\n",
      "Fragment size: 15\n",
      "(465, 477)\n",
      "Fragment size: 12\n",
      "gene60944\n",
      "(737, 784)\n",
      "Fragment size: 47\n",
      "gene60954\n",
      "(289, 321)\n",
      "Fragment size: 32\n",
      "gene60955\n",
      "(0, 27)\n",
      "Fragment size: 27\n",
      "gene60956\n",
      "(450, 491)\n",
      "Fragment size: 41\n",
      "(641, 681)\n",
      "Fragment size: 40\n",
      "(1131, 1141)\n",
      "Fragment size: 10\n",
      "Number of uncovered regions: 16\n",
      "here\n",
      "Regions not covered\n",
      "gene60923\n",
      "(150, 199)\n",
      "Fragment size: 49\n",
      "(649, 680)\n",
      "Fragment size: 31\n",
      "gene60934\n",
      "(0, 43)\n",
      "Fragment size: 43\n",
      "gene60935\n",
      "(0, 44)\n",
      "Fragment size: 44\n",
      "gene60937\n",
      "(150, 190)\n",
      "Fragment size: 40\n",
      "(915, 919)\n",
      "Fragment size: 4\n",
      "(1519, 1542)\n",
      "Fragment size: 23\n",
      "gene60940\n",
      "(150, 165)\n",
      "Fragment size: 15\n",
      "(465, 477)\n",
      "Fragment size: 12\n",
      "gene60944\n",
      "(737, 784)\n",
      "Fragment size: 47\n",
      "gene60954\n",
      "(289, 321)\n",
      "Fragment size: 32\n",
      "gene60955\n",
      "(0, 27)\n",
      "Fragment size: 27\n",
      "gene60956\n",
      "(450, 491)\n",
      "Fragment size: 41\n",
      "(641, 681)\n",
      "Fragment size: 40\n",
      "(1131, 1141)\n",
      "Fragment size: 10\n",
      "Number of uncovered regions: 15\n"
     ]
    }
   ],
   "source": [
    "final_primers, genes_uncovered = cover_regions(primers_ext,genes_dict,frag_size= 50 )\n",
    "final_primers = final_primers[[\"Name\",\"Sequence 5'\", \"Start\",\"End\", \"Gene\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.379703Z",
     "start_time": "2020-02-17T00:27:29.374012Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {'gene60923': [(150, 199), (649, 680)],\n",
       "             'gene60934': [(0, 43)],\n",
       "             'gene60935': [(0, 44)],\n",
       "             'gene60937': [(150, 190), (915, 919), (1519, 1542)],\n",
       "             'gene60940': [(150, 165), (465, 477)],\n",
       "             'gene60944': [(737, 784)],\n",
       "             'gene60954': [(289, 321)],\n",
       "             'gene60955': [(0, 27)],\n",
       "             'gene60956': [(450, 491), (641, 681), (1131, 1141)]})"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_uncovered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add in strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.563955Z",
     "start_time": "2020-02-17T00:27:29.381442Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence 5'</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>R1</td>\n",
       "      <td>ACTTGGGTTAATCGTGTGACC</td>\n",
       "      <td>680</td>\n",
       "      <td>830</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>R2</td>\n",
       "      <td>CTGTTTGTCGTAGGCAGATGG</td>\n",
       "      <td>537</td>\n",
       "      <td>687</td>\n",
       "      <td>gene60949</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>R3</td>\n",
       "      <td>GATATCGCCGATACGGTTG</td>\n",
       "      <td>1272</td>\n",
       "      <td>1422</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>R4</td>\n",
       "      <td>AGCGGATGAGTAAGAAGATTCC</td>\n",
       "      <td>1025</td>\n",
       "      <td>1175</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>R5</td>\n",
       "      <td>TTGAAGAAGGCGTGGGTACAG</td>\n",
       "      <td>805</td>\n",
       "      <td>955</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60931_750</th>\n",
       "      <td>gene60931_750</td>\n",
       "      <td>TTATTAATGATGAGTATTGA</td>\n",
       "      <td>750</td>\n",
       "      <td>900</td>\n",
       "      <td>gene60931</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60953_750</th>\n",
       "      <td>gene60953_750</td>\n",
       "      <td>TTGTTCATTGTTAAGGTTGT</td>\n",
       "      <td>750</td>\n",
       "      <td>900</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60925_1273</th>\n",
       "      <td>gene60925_1273</td>\n",
       "      <td>ACATAGACGGGTGTGCTCTT</td>\n",
       "      <td>1273</td>\n",
       "      <td>1423</td>\n",
       "      <td>gene60925</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60931_900</th>\n",
       "      <td>gene60931_900</td>\n",
       "      <td>TTTTTTTGGTTAGAACTGGA</td>\n",
       "      <td>900</td>\n",
       "      <td>1050</td>\n",
       "      <td>gene60931</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60925_1423</th>\n",
       "      <td>gene60925_1423</td>\n",
       "      <td>ATTTTTCATCTTTCCCTTGC</td>\n",
       "      <td>1423</td>\n",
       "      <td>1573</td>\n",
       "      <td>gene60925</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Name             Sequence 5'  Start   End  \\\n",
       "6                           R1   ACTTGGGTTAATCGTGTGACC    680   830   \n",
       "14                          R2   CTGTTTGTCGTAGGCAGATGG    537   687   \n",
       "20                          R3     GATATCGCCGATACGGTTG   1272  1422   \n",
       "21                          R4  AGCGGATGAGTAAGAAGATTCC   1025  1175   \n",
       "22                          R5   TTGAAGAAGGCGTGGGTACAG    805   955   \n",
       "...                        ...                     ...    ...   ...   \n",
       "gene60931_750    gene60931_750    TTATTAATGATGAGTATTGA    750   900   \n",
       "gene60953_750    gene60953_750    TTGTTCATTGTTAAGGTTGT    750   900   \n",
       "gene60925_1273  gene60925_1273    ACATAGACGGGTGTGCTCTT   1273  1423   \n",
       "gene60931_900    gene60931_900    TTTTTTTGGTTAGAACTGGA    900  1050   \n",
       "gene60925_1423  gene60925_1423    ATTTTTCATCTTTCCCTTGC   1423  1573   \n",
       "\n",
       "                     Gene Strand  \n",
       "6               gene60923      -  \n",
       "14              gene60949      -  \n",
       "20              gene60953      -  \n",
       "21              gene60953      -  \n",
       "22              gene60953      -  \n",
       "...                   ...    ...  \n",
       "gene60931_750   gene60931      -  \n",
       "gene60953_750   gene60953      -  \n",
       "gene60925_1273  gene60925      -  \n",
       "gene60931_900   gene60931      -  \n",
       "gene60925_1423  gene60925      -  \n",
       "\n",
       "[122 rows x 6 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_primers[\"Strand\"] = \"\"\n",
    "for ind,p in final_primers.iterrows():\n",
    "    strand = gff_info_genes.set_index(\"Id\").loc[p[\"Gene\"],\"Strand\"]\n",
    "    if (strand) == \"+\":\n",
    "        final_primers.loc[ind,\"Strand\"] = \"-\"\n",
    "    else:\n",
    "        final_primers.loc[ind,\"Strand\"] = \"+\"\n",
    "final_primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add in MT position (and if its negative strand, then position is reverse complement position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.667181Z",
     "start_time": "2020-02-17T00:27:29.565444Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence 5'</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Strand</th>\n",
       "      <th>MT Position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>R1</td>\n",
       "      <td>ACTTGGGTTAATCGTGTGACC</td>\n",
       "      <td>680</td>\n",
       "      <td>830</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>-</td>\n",
       "      <td>920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>R2</td>\n",
       "      <td>CTGTTTGTCGTAGGCAGATGG</td>\n",
       "      <td>537</td>\n",
       "      <td>687</td>\n",
       "      <td>gene60949</td>\n",
       "      <td>-</td>\n",
       "      <td>11599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>R3</td>\n",
       "      <td>GATATCGCCGATACGGTTG</td>\n",
       "      <td>1272</td>\n",
       "      <td>1422</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>12875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>R4</td>\n",
       "      <td>AGCGGATGAGTAAGAAGATTCC</td>\n",
       "      <td>1025</td>\n",
       "      <td>1175</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>13122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>R5</td>\n",
       "      <td>TTGAAGAAGGCGTGGGTACAG</td>\n",
       "      <td>805</td>\n",
       "      <td>955</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>13342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60931_750</th>\n",
       "      <td>gene60931_750</td>\n",
       "      <td>TTATTAATGATGAGTATTGA</td>\n",
       "      <td>750</td>\n",
       "      <td>900</td>\n",
       "      <td>gene60931</td>\n",
       "      <td>-</td>\n",
       "      <td>4760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60953_750</th>\n",
       "      <td>gene60953_750</td>\n",
       "      <td>TTGTTCATTGTTAAGGTTGT</td>\n",
       "      <td>750</td>\n",
       "      <td>900</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>13397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60925_1273</th>\n",
       "      <td>gene60925_1273</td>\n",
       "      <td>ACATAGACGGGTGTGCTCTT</td>\n",
       "      <td>1273</td>\n",
       "      <td>1423</td>\n",
       "      <td>gene60925</td>\n",
       "      <td>-</td>\n",
       "      <td>1955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60931_900</th>\n",
       "      <td>gene60931_900</td>\n",
       "      <td>TTTTTTTGGTTAGAACTGGA</td>\n",
       "      <td>900</td>\n",
       "      <td>1050</td>\n",
       "      <td>gene60931</td>\n",
       "      <td>-</td>\n",
       "      <td>4610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60925_1423</th>\n",
       "      <td>gene60925_1423</td>\n",
       "      <td>ATTTTTCATCTTTCCCTTGC</td>\n",
       "      <td>1423</td>\n",
       "      <td>1573</td>\n",
       "      <td>gene60925</td>\n",
       "      <td>-</td>\n",
       "      <td>1805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Name             Sequence 5'  Start   End  \\\n",
       "6                           R1   ACTTGGGTTAATCGTGTGACC    680   830   \n",
       "14                          R2   CTGTTTGTCGTAGGCAGATGG    537   687   \n",
       "20                          R3     GATATCGCCGATACGGTTG   1272  1422   \n",
       "21                          R4  AGCGGATGAGTAAGAAGATTCC   1025  1175   \n",
       "22                          R5   TTGAAGAAGGCGTGGGTACAG    805   955   \n",
       "...                        ...                     ...    ...   ...   \n",
       "gene60931_750    gene60931_750    TTATTAATGATGAGTATTGA    750   900   \n",
       "gene60953_750    gene60953_750    TTGTTCATTGTTAAGGTTGT    750   900   \n",
       "gene60925_1273  gene60925_1273    ACATAGACGGGTGTGCTCTT   1273  1423   \n",
       "gene60931_900    gene60931_900    TTTTTTTGGTTAGAACTGGA    900  1050   \n",
       "gene60925_1423  gene60925_1423    ATTTTTCATCTTTCCCTTGC   1423  1573   \n",
       "\n",
       "                     Gene Strand  MT Position  \n",
       "6               gene60923      -          920  \n",
       "14              gene60949      -        11599  \n",
       "20              gene60953      -        12875  \n",
       "21              gene60953      -        13122  \n",
       "22              gene60953      -        13342  \n",
       "...                   ...    ...          ...  \n",
       "gene60931_750   gene60931      -         4760  \n",
       "gene60953_750   gene60953      -        13397  \n",
       "gene60925_1273  gene60925      -         1955  \n",
       "gene60931_900   gene60931      -         4610  \n",
       "gene60925_1423  gene60925      -         1805  \n",
       "\n",
       "[122 rows x 7 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_primers[\"MT Position\"] = -1\n",
    "for ind,p in final_primers.iterrows():\n",
    "    if p[\"Strand\"] == \"+\":\n",
    "        pos = mt_seq.find(p[\"Sequence 5'\"])\n",
    "    else:\n",
    "        pos = mt_seq_biopython.reverse_complement().find(p[\"Sequence 5'\"])\n",
    "        if pos == -1:\n",
    "            print(\"Not here!\")\n",
    "        else:\n",
    "            pos = len(mt_seq) - pos - 1\n",
    "    if pos == -1:\n",
    "        print(\"Not here!\")\n",
    "    else:\n",
    "        final_primers.loc[ind,\"MT Position\"] = pos\n",
    "        \n",
    "final_primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SAVE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.684163Z",
     "start_time": "2020-02-17T00:27:29.668652Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "final_primers[[\"Name\",\"Sequence 5'\",\"Gene\", \"Strand\", \"MT Position\"]].sort_values(\"Gene\").to_csv(\"final_primers.csv\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.708243Z",
     "start_time": "2020-02-17T00:27:29.685992Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence 5'</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Strand</th>\n",
       "      <th>MT Position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>R1</td>\n",
       "      <td>ACTTGGGTTAATCGTGTGACC</td>\n",
       "      <td>680</td>\n",
       "      <td>830</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>-</td>\n",
       "      <td>920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>R2</td>\n",
       "      <td>CTGTTTGTCGTAGGCAGATGG</td>\n",
       "      <td>537</td>\n",
       "      <td>687</td>\n",
       "      <td>gene60949</td>\n",
       "      <td>-</td>\n",
       "      <td>11599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>R3</td>\n",
       "      <td>GATATCGCCGATACGGTTG</td>\n",
       "      <td>1272</td>\n",
       "      <td>1422</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>12875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>R4</td>\n",
       "      <td>AGCGGATGAGTAAGAAGATTCC</td>\n",
       "      <td>1025</td>\n",
       "      <td>1175</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>13122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>R5</td>\n",
       "      <td>TTGAAGAAGGCGTGGGTACAG</td>\n",
       "      <td>805</td>\n",
       "      <td>955</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>13342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60931_750</th>\n",
       "      <td>gene60931_750</td>\n",
       "      <td>TTATTAATGATGAGTATTGA</td>\n",
       "      <td>750</td>\n",
       "      <td>900</td>\n",
       "      <td>gene60931</td>\n",
       "      <td>-</td>\n",
       "      <td>4760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60953_750</th>\n",
       "      <td>gene60953_750</td>\n",
       "      <td>TTGTTCATTGTTAAGGTTGT</td>\n",
       "      <td>750</td>\n",
       "      <td>900</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>13397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60925_1273</th>\n",
       "      <td>gene60925_1273</td>\n",
       "      <td>ACATAGACGGGTGTGCTCTT</td>\n",
       "      <td>1273</td>\n",
       "      <td>1423</td>\n",
       "      <td>gene60925</td>\n",
       "      <td>-</td>\n",
       "      <td>1955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60931_900</th>\n",
       "      <td>gene60931_900</td>\n",
       "      <td>TTTTTTTGGTTAGAACTGGA</td>\n",
       "      <td>900</td>\n",
       "      <td>1050</td>\n",
       "      <td>gene60931</td>\n",
       "      <td>-</td>\n",
       "      <td>4610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene60925_1423</th>\n",
       "      <td>gene60925_1423</td>\n",
       "      <td>ATTTTTCATCTTTCCCTTGC</td>\n",
       "      <td>1423</td>\n",
       "      <td>1573</td>\n",
       "      <td>gene60925</td>\n",
       "      <td>-</td>\n",
       "      <td>1805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Name             Sequence 5'  Start   End  \\\n",
       "6                           R1   ACTTGGGTTAATCGTGTGACC    680   830   \n",
       "14                          R2   CTGTTTGTCGTAGGCAGATGG    537   687   \n",
       "20                          R3     GATATCGCCGATACGGTTG   1272  1422   \n",
       "21                          R4  AGCGGATGAGTAAGAAGATTCC   1025  1175   \n",
       "22                          R5   TTGAAGAAGGCGTGGGTACAG    805   955   \n",
       "...                        ...                     ...    ...   ...   \n",
       "gene60931_750    gene60931_750    TTATTAATGATGAGTATTGA    750   900   \n",
       "gene60953_750    gene60953_750    TTGTTCATTGTTAAGGTTGT    750   900   \n",
       "gene60925_1273  gene60925_1273    ACATAGACGGGTGTGCTCTT   1273  1423   \n",
       "gene60931_900    gene60931_900    TTTTTTTGGTTAGAACTGGA    900  1050   \n",
       "gene60925_1423  gene60925_1423    ATTTTTCATCTTTCCCTTGC   1423  1573   \n",
       "\n",
       "                     Gene Strand  MT Position  \n",
       "6               gene60923      -          920  \n",
       "14              gene60949      -        11599  \n",
       "20              gene60953      -        12875  \n",
       "21              gene60953      -        13122  \n",
       "22              gene60953      -        13342  \n",
       "...                   ...    ...          ...  \n",
       "gene60931_750   gene60931      -         4760  \n",
       "gene60953_750   gene60953      -        13397  \n",
       "gene60925_1273  gene60925      -         1955  \n",
       "gene60931_900   gene60931      -         4610  \n",
       "gene60925_1423  gene60925      -         1805  \n",
       "\n",
       "[122 rows x 7 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.741259Z",
     "start_time": "2020-02-17T00:27:29.711776Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1509\n",
      "14223\n"
     ]
    }
   ],
   "source": [
    "# Regions covered in DNA \n",
    "pos_mt = [False]*(len(mt_seq))\n",
    "neg_mt = [False]*(len(mt_seq))\n",
    "for ind,val in final_primers.iterrows(): \n",
    "    if val[\"Strand\"] == \"-\":\n",
    "        neg_mt[val[\"MT Position\"]:min(val[\"MT Position\"]+150,len(mt_seq))] = [True]*(min(val[\"MT Position\"]+150,len(mt_seq))-val[\"MT Position\"])\n",
    "    elif val[\"Strand\"] == \"+\" :\n",
    "        pos_mt[val[\"MT Position\"]:min(val[\"MT Position\"]+150,len(mt_seq))] = [True]*(min(val[\"MT Position\"]+150,len(mt_seq))-val[\"MT Position\"])\n",
    "    \n",
    "print(np.array(pos_mt).sum())\n",
    "print(np.array(neg_mt).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.774725Z",
     "start_time": "2020-02-17T00:27:29.743052Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14260\n",
      "1068\n"
     ]
    }
   ],
   "source": [
    "gene_pos_mt = [False]*(len(mt_seq))\n",
    "gene_neg_mt = [False]*(len(mt_seq))\n",
    "for ind, val in gff_info_genes.iterrows():\n",
    "    if val[\"Strand\"] == \"-\":\n",
    "        gene_neg_mt[val[\"Start\"]:val[\"End\"]] = [True]*(val[\"End\"]-val[\"Start\"])\n",
    "    else:\n",
    "        gene_pos_mt[val[\"Start\"]:val[\"End\"]] = [True]*(val[\"End\"]-val[\"Start\"])\n",
    "    \n",
    "\n",
    "print(np.array(gene_pos_mt).sum())\n",
    "print(np.array(gene_neg_mt).sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Include all the missing primers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find the primers used and not used in old set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.788180Z",
     "start_time": "2020-02-17T00:27:29.776270Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'F1',\n",
       " 'F10',\n",
       " 'F11',\n",
       " 'F12',\n",
       " 'F13',\n",
       " 'F14',\n",
       " 'F15',\n",
       " 'F16',\n",
       " 'F17',\n",
       " 'F18',\n",
       " 'F19',\n",
       " 'F2',\n",
       " 'F20',\n",
       " 'F22',\n",
       " 'F23',\n",
       " 'F24',\n",
       " 'F25',\n",
       " 'F26',\n",
       " 'F27',\n",
       " 'F28',\n",
       " 'F29',\n",
       " 'F3',\n",
       " 'F30',\n",
       " 'F31',\n",
       " 'F32',\n",
       " 'F35',\n",
       " 'F37',\n",
       " 'F38',\n",
       " 'F39',\n",
       " 'F4',\n",
       " 'F40',\n",
       " 'F41',\n",
       " 'F42',\n",
       " 'F43',\n",
       " 'F44',\n",
       " 'F45',\n",
       " 'F47',\n",
       " 'F48',\n",
       " 'F49',\n",
       " 'F5',\n",
       " 'F50',\n",
       " 'F51',\n",
       " 'F52',\n",
       " 'F53',\n",
       " 'F54',\n",
       " 'F55',\n",
       " 'F56',\n",
       " 'F57',\n",
       " 'F58',\n",
       " 'F59',\n",
       " 'F6',\n",
       " 'F60',\n",
       " 'F61',\n",
       " 'F63',\n",
       " 'F64',\n",
       " 'F65',\n",
       " 'F66',\n",
       " 'F67',\n",
       " 'F68',\n",
       " 'F69',\n",
       " 'F7',\n",
       " 'F70',\n",
       " 'F71',\n",
       " 'F72',\n",
       " 'F8',\n",
       " 'F9',\n",
       " 'R14',\n",
       " 'R18',\n",
       " 'R7'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primers_used = set()\n",
    "primers_not_used = set()\n",
    "for p in primers[\"Name\"].values:\n",
    "    if p in final_primers[\"Name\"].values:\n",
    "        primers_used.add(p)\n",
    "    else:\n",
    "        primers_not_used.add(p)\n",
    "primers_used\n",
    "primers_not_used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.801353Z",
     "start_time": "2020-02-17T00:27:29.789727Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "69"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(primers_not_used)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:29.817626Z",
     "start_time": "2020-02-17T00:27:29.803264Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(primers_used)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:30.101035Z",
     "start_time": "2020-02-17T00:27:29.819642Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "F7\n",
      "F28\n",
      "F5\n",
      "F57\n",
      "R18\n",
      "F8\n",
      "F6\n",
      "F24\n",
      "F29\n",
      "F70\n",
      "F35\n",
      "F44\n",
      "F11\n",
      "F58\n",
      "F4\n",
      "F69\n",
      "F9\n",
      "F65\n",
      "F52\n",
      "F54\n",
      "F22\n",
      "F45\n",
      "F53\n",
      "F13\n",
      "F16\n",
      "F56\n",
      "F27\n",
      "F3\n",
      "F67\n",
      "F49\n",
      "F19\n",
      "F2\n",
      "F1\n",
      "F41\n",
      "F39\n",
      "F40\n",
      "F59\n",
      "F37\n",
      "F71\n",
      "F17\n",
      "F10\n",
      "F51\n",
      "F50\n",
      "F12\n",
      "F42\n",
      "R7\n",
      "F23\n",
      "F14\n",
      "F30\n",
      "F32\n",
      "F25\n",
      "R14\n",
      "F47\n",
      "F48\n",
      "F61\n",
      "F31\n",
      "F68\n",
      "F72\n",
      "F18\n",
      "F60\n",
      "F55\n",
      "F63\n",
      "F43\n",
      "F20\n",
      "F26\n",
      "F38\n",
      "F66\n",
      "F15\n",
      "F64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Sequence 5'</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Strand</th>\n",
       "      <th>MT Position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>R1</td>\n",
       "      <td>ACTTGGGTTAATCGTGTGACC</td>\n",
       "      <td>680.0</td>\n",
       "      <td>830.0</td>\n",
       "      <td>gene60923</td>\n",
       "      <td>-</td>\n",
       "      <td>920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>R2</td>\n",
       "      <td>CTGTTTGTCGTAGGCAGATGG</td>\n",
       "      <td>537.0</td>\n",
       "      <td>687.0</td>\n",
       "      <td>gene60949</td>\n",
       "      <td>-</td>\n",
       "      <td>11599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>R3</td>\n",
       "      <td>GATATCGCCGATACGGTTG</td>\n",
       "      <td>1272.0</td>\n",
       "      <td>1422.0</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>12875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>R4</td>\n",
       "      <td>AGCGGATGAGTAAGAAGATTCC</td>\n",
       "      <td>1025.0</td>\n",
       "      <td>1175.0</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>13122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>R5</td>\n",
       "      <td>TTGAAGAAGGCGTGGGTACAG</td>\n",
       "      <td>805.0</td>\n",
       "      <td>955.0</td>\n",
       "      <td>gene60953</td>\n",
       "      <td>-</td>\n",
       "      <td>13342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F26</th>\n",
       "      <td>F26</td>\n",
       "      <td>GTGGCAAGAAATGGGCTAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>1347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F38</th>\n",
       "      <td>F38</td>\n",
       "      <td>CGACCTCGATGTTGGATCAGGACA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>2981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F66</th>\n",
       "      <td>F66</td>\n",
       "      <td>AACAACCGACTAATCACCACCCAACAATG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>8640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F15</th>\n",
       "      <td>F15</td>\n",
       "      <td>TTCATCCCTGTAGCATTGTTCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>12600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F64</th>\n",
       "      <td>F64</td>\n",
       "      <td>CTAATCTTCAACTCCTACATACTTCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>7930</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Name                    Sequence 5'   Start     End       Gene Strand  \\\n",
       "6     R1          ACTTGGGTTAATCGTGTGACC   680.0   830.0  gene60923      -   \n",
       "14    R2          CTGTTTGTCGTAGGCAGATGG   537.0   687.0  gene60949      -   \n",
       "20    R3            GATATCGCCGATACGGTTG  1272.0  1422.0  gene60953      -   \n",
       "21    R4         AGCGGATGAGTAAGAAGATTCC  1025.0  1175.0  gene60953      -   \n",
       "22    R5          TTGAAGAAGGCGTGGGTACAG   805.0   955.0  gene60953      -   \n",
       "..   ...                            ...     ...     ...        ...    ...   \n",
       "F26  F26            GTGGCAAGAAATGGGCTAC     NaN     NaN        NaN      +   \n",
       "F38  F38       CGACCTCGATGTTGGATCAGGACA     NaN     NaN        NaN      +   \n",
       "F66  F66  AACAACCGACTAATCACCACCCAACAATG     NaN     NaN        NaN      +   \n",
       "F15  F15         TTCATCCCTGTAGCATTGTTCG     NaN     NaN        NaN      +   \n",
       "F64  F64     CTAATCTTCAACTCCTACATACTTCC     NaN     NaN        NaN      +   \n",
       "\n",
       "     MT Position  \n",
       "6            920  \n",
       "14         11599  \n",
       "20         12875  \n",
       "21         13122  \n",
       "22         13342  \n",
       "..           ...  \n",
       "F26         1347  \n",
       "F38         2981  \n",
       "F66         8640  \n",
       "F15        12600  \n",
       "F64         7930  \n",
       "\n",
       "[191 rows x 7 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_primers_withAllOld = final_primers.copy()\n",
    "primers_ext = primers_ext.set_index(\"Name\", drop=False)\n",
    "\n",
    "for p in primers_not_used:    \n",
    "    print(p)\n",
    "    curr_seq = primers_ext.loc[p,\"Sequence 5'\"]\n",
    "    pos = mt_seq.find(curr_seq)\n",
    "    if pos == -1:\n",
    "        #Try negative\n",
    "        pos = str(mt_seq_biopython.reverse_complement()).find(curr_seq)\n",
    "        if pos != -1:\n",
    "            strand = \"-\"\n",
    "            pos = len(mt_seq) - pos - 1\n",
    "    else:\n",
    "        strand = \"+\"\n",
    "        \n",
    "    final_primers_withAllOld = pd.concat((final_primers_withAllOld, \n",
    "                       pd.DataFrame({\"Sequence 5'\":primers_ext.loc[p,\"Sequence 5'\"],\"Name\":p,\n",
    "                                    \"Strand\":strand,\"MT Position\": pos}, index=[p])),sort=False)\n",
    "\n",
    "\n",
    "final_primers_withAllOld[[\"Name\",\"Sequence 5'\",\"Gene\", \"Strand\", \"MT Position\"]].sort_values(\"Gene\").to_csv(\"final_primers_withAllOld.csv\", index=None)\n",
    "\n",
    "final_primers_withAllOld"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-17T00:27:30.147410Z",
     "start_time": "2020-02-17T00:27:30.102654Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9862\n",
      "14504\n"
     ]
    }
   ],
   "source": [
    "# Regions covered in DNA \n",
    "pos_mt = [False]*(len(mt_seq))\n",
    "neg_mt = [False]*(len(mt_seq))\n",
    "for ind,val in final_primers_withAllOld.iterrows(): \n",
    "    if val[\"Strand\"] == \"-\":\n",
    "        neg_mt[val[\"MT Position\"]:min(val[\"MT Position\"]+150,len(mt_seq))] = [True]*(min(val[\"MT Position\"]+150,len(mt_seq))-val[\"MT Position\"])\n",
    "    elif val[\"Strand\"] == \"+\" :\n",
    "        pos_mt[val[\"MT Position\"]:min(val[\"MT Position\"]+150,len(mt_seq))] = [True]*(min(val[\"MT Position\"]+150,len(mt_seq))-val[\"MT Position\"])\n",
    "    \n",
    "print(np.array(pos_mt).sum())\n",
    "print(np.array(neg_mt).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
